// @univerjs/docs-hyper-link-ui/lib/cjs/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
let _univerjs_core = require("@univerjs/core");
let _univerjs_docs_hyper_link = require("@univerjs/docs-hyper-link");
let _univerjs_engine_render = require("@univerjs/engine-render");
let _univerjs_docs = require("@univerjs/docs");
let _univerjs_docs_ui = require("@univerjs/docs-ui");
let rxjs = require("rxjs");
let _univerjs_design = require("@univerjs/design");
let _univerjs_ui = require("@univerjs/ui");
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");
let _univerjs_icons = require("@univerjs/icons");

//#region package.json
var name = "@univerjs/docs-hyper-link-ui";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
const DOCS_HYPER_LINK_UI_PLUGIN_CONFIG_KEY = "docs-hyper-link-ui.config";
const configSymbol = Symbol(DOCS_HYPER_LINK_UI_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/commands/commands/add-link.command.ts
const AddDocHyperLinkCommand = {
	type: _univerjs_core.CommandType.COMMAND,
	id: "docs.command.add-hyper-link",
	async handler(accessor, params) {
		if (!params) return false;
		const { payload, unitId, selections } = params;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const doMutation = (0, _univerjs_docs.addCustomRangeBySelectionFactory)(accessor, {
			rangeId: (0, _univerjs_core.generateRandomId)(),
			rangeType: _univerjs_core.CustomRangeType.HYPERLINK,
			properties: { url: payload },
			unitId,
			selections
		});
		if (doMutation) return commandService.syncExecuteCommand(doMutation.id, doMutation.params);
		return false;
	}
};

//#endregion
//#region src/commands/commands/update-link.command.ts
const UpdateDocHyperLinkCommand = {
	id: "docs.command.update-hyper-link",
	type: _univerjs_core.CommandType.COMMAND,
	handler(accessor, params) {
		var _oldBody$textRuns;
		if (!params) return false;
		const { unitId, payload, segmentId, linkId } = params;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const univerInstanceService = accessor.get(_univerjs_core.IUniverInstanceService);
		const currentSelection = accessor.get(_univerjs_docs.DocSelectionManagerService).getActiveTextRange();
		const doc = univerInstanceService.getUnit(unitId, _univerjs_core.UniverInstanceType.UNIVER_DOC);
		if (!currentSelection || !doc) return false;
		const textRun = (_oldBody$textRuns = (0, _univerjs_core.getBodySlice)(doc.getSelfOrHeaderFooterModel(segmentId).getBody(), currentSelection.startOffset, currentSelection.endOffset).textRuns) === null || _oldBody$textRuns === void 0 ? void 0 : _oldBody$textRuns[0];
		if (textRun) textRun.ed = params.label.length + 1;
		const replaceSelection = (0, _univerjs_docs.replaceSelectionFactory)(accessor, {
			unitId,
			body: {
				dataStream: `${params.label}`,
				customRanges: [{
					rangeId: linkId,
					rangeType: _univerjs_core.CustomRangeType.HYPERLINK,
					startIndex: 0,
					endIndex: params.label.length + 1,
					properties: { url: payload }
				}],
				textRuns: textRun ? [textRun] : void 0
			},
			selection: {
				startOffset: currentSelection.startOffset,
				endOffset: currentSelection.endOffset,
				collapsed: false,
				segmentId
			}
		});
		if (!replaceSelection) return false;
		return commandService.syncExecuteCommand(replaceSelection.id, replaceSelection.params);
	}
};

//#endregion
//#region src/views/hyper-link-edit/utils.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
function isBlankInput(value) {
	return value.trim().length === 0;
}

//#endregion
//#region src/views/hyper-link-edit/index.tsx
function hasProtocol(urlString) {
	return /^[a-zA-Z]+:\/\//.test(urlString);
}
function isEmail(url) {
	return /^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/.test(url);
}
function transformUrl(urlStr) {
	return hasProtocol(urlStr) ? urlStr : isEmail(urlStr) ? `mailto://${urlStr}` : `https://${urlStr}`;
}
const DocHyperLinkEdit = () => {
	const hyperLinkService = (0, _univerjs_ui.useDependency)(DocHyperLinkPopupService);
	const localeService = (0, _univerjs_ui.useDependency)(_univerjs_core.LocaleService);
	const editing = (0, _univerjs_ui.useObservable)(hyperLinkService.editingLink$);
	const commandService = (0, _univerjs_ui.useDependency)(_univerjs_core.ICommandService);
	const univerInstanceService = (0, _univerjs_ui.useDependency)(_univerjs_core.IUniverInstanceService);
	const docSelectionManagerService = (0, _univerjs_ui.useDependency)(_univerjs_docs.DocSelectionManagerService);
	const [link, setLink] = (0, react.useState)("");
	const [label, setLabel] = (0, react.useState)("");
	const [showError, setShowError] = (0, react.useState)(false);
	const isLegal = _univerjs_core.Tools.isLegalUrl(link);
	const doc = editing ? univerInstanceService.getUnit(editing.unitId, _univerjs_core.UniverInstanceType.UNIVER_DOC) : univerInstanceService.getCurrentUnitOfType(_univerjs_core.UniverInstanceType.UNIVER_DOC);
	(0, react.useEffect)(() => {
		var _doc$getSelfOrHeaderF2, _BuildTextUtils$custo, _body$customRanges2;
		const activeRange = docSelectionManagerService.getActiveTextRange();
		if (!activeRange) return;
		if (editing) {
			var _doc$getSelfOrHeaderF, _body$customRanges;
			const body = doc === null || doc === void 0 || (_doc$getSelfOrHeaderF = doc.getSelfOrHeaderFooterModel(editing.segmentId)) === null || _doc$getSelfOrHeaderF === void 0 ? void 0 : _doc$getSelfOrHeaderF.getBody();
			const matchedRange = body === null || body === void 0 || (_body$customRanges = body.customRanges) === null || _body$customRanges === void 0 ? void 0 : _body$customRanges.find((i) => (editing === null || editing === void 0 ? void 0 : editing.linkId) === i.rangeId && i.startIndex === editing.startIndex && i.endIndex === editing.endIndex);
			if (doc && matchedRange) {
				var _matchedRange$propert, _matchedRange$propert2;
				setLink((_matchedRange$propert = (_matchedRange$propert2 = matchedRange.properties) === null || _matchedRange$propert2 === void 0 ? void 0 : _matchedRange$propert2.url) !== null && _matchedRange$propert !== void 0 ? _matchedRange$propert : "");
				setLabel(_univerjs_core.BuildTextUtils.transform.getPlainText((0, _univerjs_core.getBodySlice)(body, matchedRange.startIndex, matchedRange.endIndex + 1).dataStream));
			}
			return;
		}
		const body = doc === null || doc === void 0 || (_doc$getSelfOrHeaderF2 = doc.getSelfOrHeaderFooterModel(activeRange.segmentId)) === null || _doc$getSelfOrHeaderF2 === void 0 ? void 0 : _doc$getSelfOrHeaderF2.getBody();
		const selection = body ? activeRange : null;
		const matchedRange = selection && ((_BuildTextUtils$custo = _univerjs_core.BuildTextUtils.customRange.getCustomRangesInterestsWithSelection(selection, (_body$customRanges2 = body === null || body === void 0 ? void 0 : body.customRanges) !== null && _body$customRanges2 !== void 0 ? _body$customRanges2 : [])) === null || _BuildTextUtils$custo === void 0 ? void 0 : _BuildTextUtils$custo[0]);
		if (doc && matchedRange) {
			var _matchedRange$propert3, _matchedRange$propert4;
			setLink((_matchedRange$propert3 = matchedRange === null || matchedRange === void 0 || (_matchedRange$propert4 = matchedRange.properties) === null || _matchedRange$propert4 === void 0 ? void 0 : _matchedRange$propert4.url) !== null && _matchedRange$propert3 !== void 0 ? _matchedRange$propert3 : "");
		}
	}, [
		doc,
		editing,
		docSelectionManagerService,
		univerInstanceService
	]);
	const handleCancel = () => {
		hyperLinkService.hideEditPopup();
	};
	const handleConfirm = () => {
		setShowError(true);
		if (!isLegal || !doc) return;
		const linkFinal = transformUrl(link);
		if (!editing) commandService.executeCommand(AddDocHyperLinkCommand.id, {
			unitId: doc.getUnitId(),
			payload: linkFinal
		});
		else {
			if (isBlankInput(label)) return;
			commandService.executeCommand(UpdateDocHyperLinkCommand.id, {
				unitId: doc.getUnitId(),
				payload: linkFinal,
				linkId: editing.linkId,
				label,
				segmentId: editing.segmentId
			});
		}
		hyperLinkService.hideEditPopup();
	};
	if (!doc) return;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, _univerjs_design.clsx)("univer-box-border univer-w-[328px] univer-rounded-xl univer-bg-white univer-px-6 univer-py-5 univer-shadow dark:!univer-bg-gray-900", _univerjs_design.borderClassName),
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", { children: [editing ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.FormLayout, {
			label: localeService.t("docs-hyper-link-ui.edit.label"),
			error: showError && isBlankInput(label) ? localeService.t("docs-hyper-link-ui.edit.labelError") : "",
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Input, {
				value: label,
				onChange: setLabel,
				autoFocus: true,
				onKeyDown: (evt) => {
					if (evt.keyCode === _univerjs_ui.KeyCode.ENTER) handleConfirm();
				}
			})
		}) : null, /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.FormLayout, {
			label: localeService.t("docs-hyper-link-ui.edit.address"),
			error: showError && !isLegal ? localeService.t("docs-hyper-link-ui.edit.addressError") : "",
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Input, {
				value: link,
				onChange: setLink,
				autoFocus: true,
				onKeyDown: (evt) => {
					if (evt.keyCode === _univerjs_ui.KeyCode.ENTER) handleConfirm();
				}
			})
		})] }), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "univer-flex univer-justify-end univer-gap-3",
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Button, {
				onClick: handleCancel,
				children: localeService.t("docs-hyper-link-ui.edit.cancel")
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Button, {
				variant: "primary",
				disabled: isBlankInput(link),
				onClick: handleConfirm,
				children: localeService.t("docs-hyper-link-ui.edit.confirm")
			})]
		})]
	});
};
DocHyperLinkEdit.componentKey = "docs-hyper-link-edit";

//#endregion
//#region src/commands/commands/delete-link.command.ts
const DeleteDocHyperLinkCommand = {
	type: _univerjs_core.CommandType.COMMAND,
	id: "docs.command.delete-hyper-link",
	async handler(accessor, params) {
		if (!params) return false;
		const { unitId, linkId, segmentId } = params;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const doMutation = (0, _univerjs_docs.deleteCustomRangeFactory)(accessor, {
			unitId,
			rangeId: linkId,
			segmentId
		});
		if (!doMutation) return false;
		return await commandService.syncExecuteCommand(doMutation.id, doMutation.params);
	}
};

//#endregion
//#region src/commands/operations/popup.operation.ts
const shouldDisableAddLink = (accessor) => {
	const textSelectionService = accessor.get(_univerjs_docs.DocSelectionManagerService);
	const univerInstanceService = accessor.get(_univerjs_core.IUniverInstanceService);
	const textRanges = textSelectionService.getTextRanges();
	if (!(textRanges === null || textRanges === void 0 ? void 0 : textRanges.length)) return true;
	const activeRange = textRanges[0];
	if (!univerInstanceService.getCurrentUnitOfType(_univerjs_core.UniverInstanceType.UNIVER_DOC) || !activeRange || activeRange.collapsed) return true;
	return false;
};
const ShowDocHyperLinkEditPopupOperation = {
	type: _univerjs_core.CommandType.OPERATION,
	id: "doc.operation.show-hyper-link-edit-popup",
	handler(accessor, params) {
		var _univerInstanceServic;
		const linkInfo = params === null || params === void 0 ? void 0 : params.link;
		const univerInstanceService = accessor.get(_univerjs_core.IUniverInstanceService);
		if (shouldDisableAddLink(accessor) && !linkInfo) return false;
		const hyperLinkService = accessor.get(DocHyperLinkPopupService);
		const unitId = (linkInfo === null || linkInfo === void 0 ? void 0 : linkInfo.unitId) || ((_univerInstanceServic = univerInstanceService.getCurrentUnitOfType(_univerjs_core.UniverInstanceType.UNIVER_DOC)) === null || _univerInstanceServic === void 0 ? void 0 : _univerInstanceServic.getUnitId());
		if (!unitId) return false;
		hyperLinkService.showEditPopup(unitId, linkInfo);
		return true;
	}
};
const ToggleDocHyperLinkInfoPopupOperation = {
	type: _univerjs_core.CommandType.OPERATION,
	id: "doc.operation.toggle-hyper-link-info-popup",
	handler(accessor, params) {
		const hyperLinkService = accessor.get(DocHyperLinkPopupService);
		if (!params) {
			hyperLinkService.hideInfoPopup();
			return true;
		}
		hyperLinkService.showInfoPopup(params);
		return true;
	}
};
const ClickDocHyperLinkOperation = {
	type: _univerjs_core.CommandType.OPERATION,
	id: "doc.operation.click-hyper-link",
	handler(accessor, params) {
		var _body$customRanges;
		if (!params) return false;
		const { unitId, linkId, segmentId } = params;
		const doc = accessor.get(_univerjs_core.IUniverInstanceService).getUnit(unitId, _univerjs_core.UniverInstanceType.UNIVER_DOC);
		const body = doc === null || doc === void 0 ? void 0 : doc.getSelfOrHeaderFooterModel(segmentId).getBody();
		const link = body === null || body === void 0 || (_body$customRanges = body.customRanges) === null || _body$customRanges === void 0 || (_body$customRanges = _body$customRanges.find((range) => range.rangeId === linkId && range.rangeType === _univerjs_core.CustomRangeType.HYPERLINK)) === null || _body$customRanges === void 0 || (_body$customRanges = _body$customRanges.properties) === null || _body$customRanges === void 0 ? void 0 : _body$customRanges.url;
		if (!(0, _univerjs_core.isSafeUrl)(link)) return false;
		window.open(link, "_blank", "noopener noreferrer");
		return true;
	}
};

//#endregion
//#region src/views/hyper-link-popup/index.tsx
const DocLinkPopup = () => {
	var _body$customRanges, _link$properties;
	const hyperLinkService = (0, _univerjs_ui.useDependency)(DocHyperLinkPopupService);
	const commandService = (0, _univerjs_ui.useDependency)(_univerjs_core.ICommandService);
	const messageService = (0, _univerjs_ui.useDependency)(_univerjs_ui.IMessageService);
	const localeService = (0, _univerjs_ui.useDependency)(_univerjs_core.LocaleService);
	const currentPopup = (0, _univerjs_ui.useObservable)(hyperLinkService.showingLink$);
	const univerInstanceService = (0, _univerjs_ui.useDependency)(_univerjs_core.IUniverInstanceService);
	if (!currentPopup) return null;
	const { unitId, linkId, segmentId, startIndex, endIndex } = currentPopup;
	const doc = univerInstanceService.getUnit(unitId, _univerjs_core.UniverInstanceType.UNIVER_DOC);
	const body = doc === null || doc === void 0 ? void 0 : doc.getSelfOrHeaderFooterModel(segmentId).getBody();
	const link = body === null || body === void 0 || (_body$customRanges = body.customRanges) === null || _body$customRanges === void 0 ? void 0 : _body$customRanges.find((range) => range.rangeId === linkId && range.rangeType === _univerjs_core.CustomRangeType.HYPERLINK && range.startIndex === startIndex && range.endIndex === endIndex);
	if (!link) return null;
	const url = (_link$properties = link.properties) === null || _link$properties === void 0 ? void 0 : _link$properties.url;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, _univerjs_design.clsx)("univer-box-border univer-flex univer-max-w-80 univer-items-center univer-justify-between univer-overflow-hidden univer-rounded-lg univer-bg-white univer-p-3 univer-shadow dark:!univer-bg-gray-900", _univerjs_design.borderClassName),
		onClick: () => {
			hyperLinkService.hideInfoPopup();
		},
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "univer-flex univer-h-6 univer-flex-1 univer-cursor-pointer univer-items-center univer-truncate univer-text-sm univer-leading-5 univer-text-primary-500",
			onClick: () => window.open(url, void 0, "noopener noreferrer"),
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "univer-mr-2 univer-flex univer-size-5 univer-flex-[0_0_auto] univer-items-center univer-justify-center univer-text-base univer-text-gray-900 dark:!univer-text-white",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_icons.LinkIcon, {})
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Tooltip, {
				showIfEllipsis: true,
				title: url,
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
					className: "univer-flex-1 univer-truncate",
					children: url
				})
			})]
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "univer-flex univer-h-6 univer-flex-[0_0_auto] univer-items-center univer-justify-center",
			children: [
				/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
					className: "univer-ml-2 univer-flex univer-size-6 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded univer-text-base",
					onClick: () => {
						navigator.clipboard.writeText(url);
						messageService.show({
							content: localeService.t("docs-hyper-link-ui.info.coped"),
							type: _univerjs_design.MessageType.Info
						});
					},
					children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Tooltip, {
						placement: "bottom",
						title: localeService.t("docs-hyper-link-ui.info.copy"),
						children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_icons.CopyIcon, {})
					})
				}),
				/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
					className: "univer-ml-2 univer-flex univer-size-6 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded univer-text-base",
					onClick: () => {
						commandService.executeCommand(ShowDocHyperLinkEditPopupOperation.id, { link: currentPopup });
					},
					children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Tooltip, {
						placement: "bottom",
						title: localeService.t("docs-hyper-link-ui.info.edit"),
						children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_icons.WriteIcon, {})
					})
				}),
				/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
					className: "univer-ml-2 univer-flex univer-size-6 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded univer-text-base",
					onClick: () => {
						commandService.executeCommand(DeleteDocHyperLinkCommand.id, {
							unitId,
							linkId: link.rangeId,
							segmentId
						});
					},
					children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Tooltip, {
						placement: "bottom",
						title: localeService.t("docs-hyper-link-ui.info.cancel"),
						children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_icons.UnlinkIcon, {})
					})
				})
			]
		})]
	});
};
DocLinkPopup.componentKey = "univer.doc.link-info-popup";

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/services/hyper-link-popup.service.ts
let DocHyperLinkPopupService = class DocHyperLinkPopupService extends _univerjs_core.Disposable {
	constructor(_docCanvasPopupManagerService, _textSelectionManagerService, _univerInstanceService) {
		super();
		this._docCanvasPopupManagerService = _docCanvasPopupManagerService;
		this._textSelectionManagerService = _textSelectionManagerService;
		this._univerInstanceService = _univerInstanceService;
		_defineProperty(this, "_editingLink$", new rxjs.BehaviorSubject(null));
		_defineProperty(this, "_showingLink$", new rxjs.BehaviorSubject(null));
		_defineProperty(this, "editingLink$", this._editingLink$.asObservable());
		_defineProperty(this, "showingLink$", this._showingLink$.asObservable());
		_defineProperty(this, "_editPopup", null);
		_defineProperty(this, "_infoPopup", null);
		this.disposeWithMe(() => {
			this._editingLink$.complete();
			this._showingLink$.complete();
		});
	}
	get editing() {
		return this._editingLink$.value;
	}
	get showing() {
		return this._showingLink$.value;
	}
	showEditPopup(unitId, linkInfo) {
		if (this._editPopup) this._editPopup.dispose();
		this._editingLink$.next(linkInfo);
		const textRanges = this._textSelectionManagerService.getTextRanges({
			unitId,
			subUnitId: unitId
		});
		let activeRange = textRanges === null || textRanges === void 0 ? void 0 : textRanges[textRanges.length - 1];
		if (linkInfo) {
			const { segmentId, segmentPage, startIndex, endIndex } = linkInfo;
			activeRange = {
				collapsed: false,
				startOffset: startIndex,
				endOffset: endIndex + 1,
				segmentId,
				segmentPage
			};
			this._textSelectionManagerService.replaceDocRanges([{
				startOffset: startIndex,
				endOffset: endIndex + 1
			}]);
		}
		if (activeRange) {
			this._editPopup = this._docCanvasPopupManagerService.attachPopupToRange(activeRange, {
				componentKey: DocHyperLinkEdit.componentKey,
				direction: "bottom"
			}, unitId);
			return this._editPopup;
		}
		return null;
	}
	hideEditPopup() {
		var _this$_editPopup;
		this._editingLink$.next(null);
		(_this$_editPopup = this._editPopup) === null || _this$_editPopup === void 0 || _this$_editPopup.dispose();
	}
	showInfoPopup(info) {
		var _this$showing, _this$showing2, _this$showing3, _this$showing4, _this$showing5, _this$showing6;
		const { linkId, unitId, segmentId, segmentPage, startIndex, endIndex } = info;
		if (((_this$showing = this.showing) === null || _this$showing === void 0 ? void 0 : _this$showing.linkId) === linkId && ((_this$showing2 = this.showing) === null || _this$showing2 === void 0 ? void 0 : _this$showing2.unitId) === unitId && ((_this$showing3 = this.showing) === null || _this$showing3 === void 0 ? void 0 : _this$showing3.segmentId) === segmentId && ((_this$showing4 = this.showing) === null || _this$showing4 === void 0 ? void 0 : _this$showing4.segmentPage) === segmentPage && ((_this$showing5 = this.showing) === null || _this$showing5 === void 0 ? void 0 : _this$showing5.startIndex) === startIndex && ((_this$showing6 = this.showing) === null || _this$showing6 === void 0 ? void 0 : _this$showing6.endIndex) === endIndex) return;
		if (this._infoPopup) this._infoPopup.dispose();
		if (!this._univerInstanceService.getUnit(unitId, _univerjs_core.UniverInstanceType.UNIVER_DOC)) return;
		this._showingLink$.next({
			unitId,
			linkId,
			segmentId,
			segmentPage,
			startIndex,
			endIndex
		});
		this._infoPopup = this._docCanvasPopupManagerService.attachPopupToRange({
			collapsed: false,
			startOffset: startIndex,
			endOffset: endIndex + 1,
			segmentId,
			segmentPage
		}, {
			componentKey: DocLinkPopup.componentKey,
			direction: "top-center",
			multipleDirection: "top",
			onClickOutside: () => {
				this.hideInfoPopup();
			}
		}, unitId);
		return this._infoPopup;
	}
	hideInfoPopup() {
		var _this$_infoPopup;
		this._showingLink$.next(null);
		(_this$_infoPopup = this._infoPopup) === null || _this$_infoPopup === void 0 || _this$_infoPopup.dispose();
	}
};
DocHyperLinkPopupService = __decorate([
	__decorateParam(0, (0, _univerjs_core.Inject)(_univerjs_docs_ui.DocCanvasPopManagerService)),
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_docs.DocSelectionManagerService)),
	__decorateParam(2, _univerjs_core.IUniverInstanceService)
], DocHyperLinkPopupService);

//#endregion
//#region src/controllers/doc-hyper-link-selection.controller.ts
let DocHyperLinkSelectionController = class DocHyperLinkSelectionController extends _univerjs_core.Disposable {
	constructor(_commandService, _univerInstanceService, _docHyperLinkService) {
		super();
		this._commandService = _commandService;
		this._univerInstanceService = _univerInstanceService;
		this._docHyperLinkService = _docHyperLinkService;
		this._initSelectionChange();
	}
	_initSelectionChange() {
		this.disposeWithMe(this._commandService.onCommandExecuted((commandInfo) => {
			if (commandInfo.id === _univerjs_docs.SetTextSelectionsOperation.id) {
				const { unitId, ranges, segmentId } = commandInfo.params;
				const doc = this._univerInstanceService.getUnit(unitId, _univerjs_core.UniverInstanceType.UNIVER_DOC);
				const primary = ranges[0];
				if (primary && doc) {
					var _doc$getSelfOrHeaderF;
					const { startOffset, endOffset, collapsed, segmentPage } = primary;
					const customRanges = (_doc$getSelfOrHeaderF = doc.getSelfOrHeaderFooterModel(segmentId)) === null || _doc$getSelfOrHeaderF === void 0 || (_doc$getSelfOrHeaderF = _doc$getSelfOrHeaderF.getBody()) === null || _doc$getSelfOrHeaderF === void 0 ? void 0 : _doc$getSelfOrHeaderF.customRanges;
					if (collapsed) {
						var _customRanges$findInd;
						const index = (_customRanges$findInd = customRanges === null || customRanges === void 0 ? void 0 : customRanges.findIndex((value) => value.startIndex < startOffset && value.endIndex > endOffset - 1)) !== null && _customRanges$findInd !== void 0 ? _customRanges$findInd : -1;
						if (index > -1) {
							const customRange = customRanges[index];
							this._docHyperLinkService.showInfoPopup({
								unitId,
								linkId: customRange.rangeId,
								segmentId,
								segmentPage,
								startIndex: customRange.startIndex,
								endIndex: customRange.endIndex
							});
							return;
						}
					} else if (customRanges === null || customRanges === void 0 ? void 0 : customRanges.find((value) => value.startIndex <= startOffset && value.endIndex >= endOffset - 1)) return;
				}
				this._docHyperLinkService.hideInfoPopup();
				this._docHyperLinkService.hideEditPopup();
			}
		}));
	}
};
DocHyperLinkSelectionController = __decorate([
	__decorateParam(0, _univerjs_core.ICommandService),
	__decorateParam(1, _univerjs_core.IUniverInstanceService),
	__decorateParam(2, (0, _univerjs_core.Inject)(DocHyperLinkPopupService))
], DocHyperLinkSelectionController);

//#endregion
//#region src/controllers/render-controllers/hyper-link-event.render-controller.ts
let DocHyperLinkEventRenderController = class DocHyperLinkEventRenderController extends _univerjs_core.Disposable {
	get _skeleton() {
		return this._docSkeletonManagerService.getSkeleton();
	}
	constructor(_context, _docEventManagerService, _commandService, _hyperLinkPopupService, _docSkeletonManagerService, _docSelectionManagerService) {
		super();
		this._context = _context;
		this._docEventManagerService = _docEventManagerService;
		this._commandService = _commandService;
		this._hyperLinkPopupService = _hyperLinkPopupService;
		this._docSkeletonManagerService = _docSkeletonManagerService;
		this._docSelectionManagerService = _docSelectionManagerService;
		if (this._context.unitId === _univerjs_core.DOCS_ZEN_EDITOR_UNIT_ID_KEY || this._context.unitId === _univerjs_core.DOCS_NORMAL_EDITOR_UNIT_ID_KEY) return;
		this._initHover();
		this._initClick();
	}
	_hideInfoPopup() {
		if (this._hyperLinkPopupService.showing) this._commandService.executeCommand(ToggleDocHyperLinkInfoPopupOperation.id);
	}
	_initHover() {
		this.disposeWithMe(this._docEventManagerService.hoverCustomRanges$.subscribe((ranges) => {
			var _activeRanges$, _link$segmentId;
			const link = ranges.find((range) => range.range.rangeType === _univerjs_core.CustomRangeType.HYPERLINK);
			const activeRanges = this._docSelectionManagerService.getTextRanges();
			const currentSegmentId = activeRanges === null || activeRanges === void 0 || (_activeRanges$ = activeRanges[0]) === null || _activeRanges$ === void 0 ? void 0 : _activeRanges$.segmentId;
			if (((_link$segmentId = link === null || link === void 0 ? void 0 : link.segmentId) !== null && _link$segmentId !== void 0 ? _link$segmentId : "") !== currentSegmentId) {
				this._hideInfoPopup();
				return;
			}
			if (link) this._commandService.executeCommand(ToggleDocHyperLinkInfoPopupOperation.id, {
				unitId: this._context.unitId,
				linkId: link.range.rangeId,
				segmentId: link.segmentId,
				segmentPage: link.segmentPageIndex,
				rangeId: link.range.rangeId,
				startIndex: link.range.startIndex,
				endIndex: link.range.endIndex
			});
			else this._hideInfoPopup();
		}));
	}
	_initClick() {
		this.disposeWithMe(this._docEventManagerService.clickCustomRanges$.subscribe((range) => {
			const link = range.range;
			if (link) this._commandService.executeCommand(ClickDocHyperLinkOperation.id, {
				unitId: this._context.unitId,
				linkId: link.rangeId,
				segmentId: range.segmentId
			});
		}));
	}
};
DocHyperLinkEventRenderController = __decorate([
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_docs_ui.DocEventManagerService)),
	__decorateParam(2, _univerjs_core.ICommandService),
	__decorateParam(3, (0, _univerjs_core.Inject)(DocHyperLinkPopupService)),
	__decorateParam(4, (0, _univerjs_core.Inject)(_univerjs_docs.DocSkeletonManagerService)),
	__decorateParam(5, (0, _univerjs_core.Inject)(_univerjs_docs.DocSelectionManagerService))
], DocHyperLinkEventRenderController);

//#endregion
//#region src/controllers/render-controllers/render.controller.ts
let DocHyperLinkRenderController = class DocHyperLinkRenderController extends _univerjs_core.Disposable {
	constructor(_context, _docInterceptorService, _hyperLinkService, _docRenderController) {
		super();
		this._context = _context;
		this._docInterceptorService = _docInterceptorService;
		this._hyperLinkService = _hyperLinkService;
		this._docRenderController = _docRenderController;
		this._init();
		this._initReRender();
	}
	_init() {
		this._docInterceptorService.intercept(_univerjs_docs.DOC_INTERCEPTOR_POINT.CUSTOM_RANGE, { handler: (data, pos, next) => {
			if (!data) return next(data);
			const { unitId, index } = pos;
			const activeLink = this._hyperLinkService.showing;
			if (!activeLink) return next({
				...data,
				active: false
			});
			const { linkId, unitId: linkUnitId, startIndex, endIndex } = activeLink;
			const isActive = linkUnitId === unitId && data.rangeId === linkId && index >= startIndex && index <= endIndex;
			return next({
				...data,
				active: isActive
			});
		} });
	}
	_initReRender() {
		this.disposeWithMe(this._hyperLinkService.showingLink$.pipe((0, rxjs.distinctUntilChanged)((prev, aft) => (prev === null || prev === void 0 ? void 0 : prev.linkId) === (aft === null || aft === void 0 ? void 0 : aft.linkId) && (prev === null || prev === void 0 ? void 0 : prev.unitId) === (aft === null || aft === void 0 ? void 0 : aft.unitId) && (prev === null || prev === void 0 ? void 0 : prev.startIndex) === (aft === null || aft === void 0 ? void 0 : aft.startIndex)), (0, rxjs.pairwise)()).subscribe(([preLink, link]) => {
			if (link) {
				if (link.unitId === this._context.unitId) this._docRenderController.reRender(link.unitId);
			} else if (preLink && preLink.unitId === this._context.unitId) this._docRenderController.reRender(preLink.unitId);
		}));
	}
};
DocHyperLinkRenderController = __decorate([
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_docs.DocInterceptorService)),
	__decorateParam(2, (0, _univerjs_core.Inject)(DocHyperLinkPopupService)),
	__decorateParam(3, (0, _univerjs_core.Inject)(_univerjs_docs_ui.DocRenderController))
], DocHyperLinkRenderController);

//#endregion
//#region src/menu/menu.ts
const DOC_LINK_ICON = "doc-hyper-link-icon";
function AddHyperLinkMenuItemFactory(accessor) {
	return {
		id: ShowDocHyperLinkEditPopupOperation.id,
		type: _univerjs_ui.MenuItemType.BUTTON,
		icon: DOC_LINK_ICON,
		title: "docs-hyper-link-ui.menu.tooltip",
		tooltip: "docs-hyper-link-ui.menu.tooltip",
		hidden$: (0, _univerjs_ui.getMenuHiddenObservable)(accessor, _univerjs_core.UniverInstanceType.UNIVER_DOC),
		disabled$: new rxjs.Observable(function(subscribe) {
			const observer = accessor.get(_univerjs_docs.DocSelectionManagerService).textSelection$.pipe((0, rxjs.debounceTime)(16)).subscribe(() => {
				subscribe.next(shouldDisableAddLink(accessor));
			});
			return () => {
				observer.unsubscribe();
			};
		})
	};
}
const addLinkShortcut = {
	id: ShowDocHyperLinkEditPopupOperation.id,
	binding: _univerjs_ui.MetaKeys.CTRL_COMMAND | _univerjs_ui.KeyCode.K,
	description: "docs-hyper-link-ui.menu.tooltip",
	preconditions: _univerjs_docs_ui.whenDocAndEditorFocused
};

//#endregion
//#region src/menu/schema.ts
const menuSchema = {
	[_univerjs_ui.RibbonInsertGroup.MEDIA]: { [ShowDocHyperLinkEditPopupOperation.id]: {
		order: 1,
		menuItemFactory: AddHyperLinkMenuItemFactory
	} },
	[_univerjs_ui.ContextMenuPosition.MAIN_AREA]: { [_univerjs_ui.ContextMenuGroup.DATA]: { [ShowDocHyperLinkEditPopupOperation.id]: {
		order: 0,
		menuItemFactory: AddHyperLinkMenuItemFactory
	} } },
	[_univerjs_ui.ContextMenuPosition.PARAGRAPH]: {
		[_univerjs_ui.ContextMenuGroup.LAYOUT]: { [_univerjs_docs_ui.INSERT_BELLOW_MENU_ID]: { [ShowDocHyperLinkEditPopupOperation.id]: {
			order: 6,
			menuItemFactory: AddHyperLinkMenuItemFactory
		} } },
		[_univerjs_docs_ui.EMPTY_PARAGRAPH_MENU_ID]: { [_univerjs_ui.ContextMenuGroup.LAYOUT]: { [ShowDocHyperLinkEditPopupOperation.id]: {
			order: 6,
			menuItemFactory: AddHyperLinkMenuItemFactory
		} } }
	}
};

//#endregion
//#region src/controllers/ui.controller.ts
let DocHyperLinkUIController = class DocHyperLinkUIController extends _univerjs_core.Disposable {
	constructor(_componentManager, _commandService, _menuManagerService, _shortcutService) {
		super();
		this._componentManager = _componentManager;
		this._commandService = _commandService;
		this._menuManagerService = _menuManagerService;
		this._shortcutService = _shortcutService;
		this._initComponents();
		this._initCommands();
		this._initMenus();
		this._initShortcut();
	}
	_initComponents() {
		[
			[DocHyperLinkEdit.componentKey, DocHyperLinkEdit],
			[DocLinkPopup.componentKey, DocLinkPopup],
			[DOC_LINK_ICON, _univerjs_icons.LinkIcon]
		].forEach(([key, comp]) => {
			this.disposeWithMe(this._componentManager.register(key, comp));
		});
	}
	_initCommands() {
		[
			AddDocHyperLinkCommand,
			UpdateDocHyperLinkCommand,
			DeleteDocHyperLinkCommand,
			ShowDocHyperLinkEditPopupOperation,
			ToggleDocHyperLinkInfoPopupOperation,
			ClickDocHyperLinkOperation
		].forEach((command) => {
			this._commandService.registerCommand(command);
		});
	}
	_initShortcut() {
		[addLinkShortcut].forEach((shortcut) => {
			this._shortcutService.registerShortcut(shortcut);
		});
	}
	_initMenus() {
		this._menuManagerService.mergeMenu(menuSchema);
	}
};
DocHyperLinkUIController = __decorate([
	__decorateParam(0, (0, _univerjs_core.Inject)(_univerjs_ui.ComponentManager)),
	__decorateParam(1, _univerjs_core.ICommandService),
	__decorateParam(2, _univerjs_ui.IMenuManagerService),
	__decorateParam(3, _univerjs_ui.IShortcutService)
], DocHyperLinkUIController);

//#endregion
//#region src/types/const/index.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const DOC_HYPER_LINK_UI_PLUGIN = "DOC_HYPER_LINK_UI_PLUGIN";

//#endregion
//#region src/plugin.ts
let UniverDocsHyperLinkUIPlugin = class UniverDocsHyperLinkUIPlugin extends _univerjs_core.Plugin {
	constructor(_config = defaultPluginConfig, _injector, _renderManagerSrv, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._renderManagerSrv = _renderManagerSrv;
		this._configService = _configService;
		const { menu, ...rest } = (0, _univerjs_core.merge)({}, defaultPluginConfig, this._config);
		if (menu) this._configService.setConfig("menu", menu, { merge: true });
		this._configService.setConfig(DOCS_HYPER_LINK_UI_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		[
			[DocHyperLinkPopupService],
			[DocHyperLinkUIController],
			[DocHyperLinkSelectionController]
		].forEach((dep) => {
			this._injector.add(dep);
		});
		this._injector.get(DocHyperLinkUIController);
	}
	onReady() {
		this._injector.get(DocHyperLinkSelectionController);
	}
	onRendered() {
		this._initRenderModule();
	}
	_initRenderModule() {
		[[DocHyperLinkRenderController], [DocHyperLinkEventRenderController]].forEach((dep) => {
			this._renderManagerSrv.registerRenderModule(_univerjs_core.UniverInstanceType.UNIVER_DOC, dep);
		});
	}
};
_defineProperty(UniverDocsHyperLinkUIPlugin, "pluginName", DOC_HYPER_LINK_UI_PLUGIN);
_defineProperty(UniverDocsHyperLinkUIPlugin, "packageName", name);
_defineProperty(UniverDocsHyperLinkUIPlugin, "version", version);
_defineProperty(UniverDocsHyperLinkUIPlugin, "type", _univerjs_core.UniverInstanceType.UNIVER_DOC);
UniverDocsHyperLinkUIPlugin = __decorate([
	(0, _univerjs_core.DependentOn)(_univerjs_docs_hyper_link.UniverDocsHyperLinkPlugin),
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector)),
	__decorateParam(2, _univerjs_engine_render.IRenderManagerService),
	__decorateParam(3, _univerjs_core.IConfigService)
], UniverDocsHyperLinkUIPlugin);

//#endregion
Object.defineProperty(exports, 'UniverDocsHyperLinkUIPlugin', {
  enumerable: true,
  get: function () {
    return UniverDocsHyperLinkUIPlugin;
  }
});
