// @univerjs/sheets-table/lib/index.js
import { BorderStyleTypes, CellValueType, CommandType, Disposable, ICommandService, IConfigService, ILogService, IResourceManagerService, IUndoRedoService, IUniverInstanceService, Inject, Injector, InterceptorEffectEnum, LocaleService, Plugin, RTree, Rectangle, UniverInstanceType, createREGEXFromWildChar, customNameCharacterCheck, generateRandomId, merge, registerDependencies, sequenceExecute, touchDependencies } from "@univerjs/core";
import { IDefinedNamesService, RemoveSuperTableMutation, SetSuperTableMutation } from "@univerjs/engine-formula";
import { AddRangeThemeMutation, CopySheetCommand, IExclusiveRangeService, INTERCEPTOR_POINT, InsertColCommand, InsertColMutation, InsertRowCommand, InsertRowMutation, RangeThemeStyle, RefRangeService, RemoveColCommand, RemoveColMutation, RemoveRangeThemeMutation, RemoveRowCommand, RemoveRowMutation, RemoveSheetCommand, SheetInterceptorService, SheetRangeThemeModel, SheetRangeThemeService, SheetsSelectionsService, ZebraCrossingCacheController, getMoveRangeUndoRedoMutations, getSheetCommandTarget } from "@univerjs/sheets";
import { BehaviorSubject, Subject, filter, switchMap } from "rxjs";

//#region src/types/enum.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
let TableColumnDataTypeEnum = /* @__PURE__ */ function(TableColumnDataTypeEnum) {
	TableColumnDataTypeEnum["None"] = "none";
	TableColumnDataTypeEnum["String"] = "string";
	TableColumnDataTypeEnum["Number"] = "number";
	TableColumnDataTypeEnum["Date"] = "date";
	TableColumnDataTypeEnum["Bool"] = "bool";
	TableColumnDataTypeEnum["Checkbox"] = "checkbox";
	TableColumnDataTypeEnum["List"] = "list";
	return TableColumnDataTypeEnum;
}({});
let TableColumnFilterTypeEnum = /* @__PURE__ */ function(TableColumnFilterTypeEnum) {
	TableColumnFilterTypeEnum["manual"] = "manual";
	TableColumnFilterTypeEnum["condition"] = "condition";
	return TableColumnFilterTypeEnum;
}({});
let TableConditionTypeEnum = /* @__PURE__ */ function(TableConditionTypeEnum) {
	TableConditionTypeEnum["Date"] = "date";
	TableConditionTypeEnum["Number"] = "number";
	TableConditionTypeEnum["String"] = "string";
	TableConditionTypeEnum["Logic"] = "logic";
	return TableConditionTypeEnum;
}({});
let TableNumberCompareTypeEnum = /* @__PURE__ */ function(TableNumberCompareTypeEnum) {
	TableNumberCompareTypeEnum["Equal"] = "equal";
	TableNumberCompareTypeEnum["NotEqual"] = "notEqual";
	TableNumberCompareTypeEnum["GreaterThan"] = "greaterThan";
	TableNumberCompareTypeEnum["GreaterThanOrEqual"] = "greaterThanOrEqual";
	TableNumberCompareTypeEnum["LessThan"] = "lessThan";
	TableNumberCompareTypeEnum["LessThanOrEqual"] = "lessThanOrEqual";
	TableNumberCompareTypeEnum["Between"] = "between";
	TableNumberCompareTypeEnum["NotBetween"] = "notBetween";
	TableNumberCompareTypeEnum["Above"] = "above";
	TableNumberCompareTypeEnum["Below"] = "below";
	TableNumberCompareTypeEnum["TopN"] = "topN";
	return TableNumberCompareTypeEnum;
}({});
let TableStringCompareTypeEnum = /* @__PURE__ */ function(TableStringCompareTypeEnum) {
	TableStringCompareTypeEnum["Equal"] = "equal";
	TableStringCompareTypeEnum["NotEqual"] = "notEqual";
	TableStringCompareTypeEnum["Contains"] = "contains";
	TableStringCompareTypeEnum["NotContains"] = "notContains";
	TableStringCompareTypeEnum["StartsWith"] = "startsWith";
	TableStringCompareTypeEnum["EndsWith"] = "endsWith";
	return TableStringCompareTypeEnum;
}({});
let TableDateCompareTypeEnum = /* @__PURE__ */ function(TableDateCompareTypeEnum) {
	TableDateCompareTypeEnum["Equal"] = "equal";
	TableDateCompareTypeEnum["NotEqual"] = "notEqual";
	TableDateCompareTypeEnum["After"] = "after";
	TableDateCompareTypeEnum["AfterOrEqual"] = "afterOrEqual";
	TableDateCompareTypeEnum["Before"] = "before";
	TableDateCompareTypeEnum["BeforeOrEqual"] = "beforeOrEqual";
	TableDateCompareTypeEnum["Between"] = "between";
	TableDateCompareTypeEnum["NotBetween"] = "notBetween";
	TableDateCompareTypeEnum["Today"] = "today";
	TableDateCompareTypeEnum["Yesterday"] = "yesterday";
	TableDateCompareTypeEnum["Tomorrow"] = "tomorrow";
	TableDateCompareTypeEnum["ThisWeek"] = "thisWeek";
	TableDateCompareTypeEnum["LastWeek"] = "lastWeek";
	TableDateCompareTypeEnum["NextWeek"] = "nextWeek";
	TableDateCompareTypeEnum["ThisMonth"] = "thisMonth";
	TableDateCompareTypeEnum["LastMonth"] = "lastMonth";
	TableDateCompareTypeEnum["NextMonth"] = "nextMonth";
	TableDateCompareTypeEnum["ThisQuarter"] = "thisQuarter";
	TableDateCompareTypeEnum["LastQuarter"] = "lastQuarter";
	TableDateCompareTypeEnum["NextQuarter"] = "nextQuarter";
	TableDateCompareTypeEnum["ThisYear"] = "thisYear";
	TableDateCompareTypeEnum["LastYear"] = "lastYear";
	TableDateCompareTypeEnum["NextYear"] = "nextYear";
	TableDateCompareTypeEnum["YearToDate"] = "yearToDate";
	TableDateCompareTypeEnum["Quarter"] = "quarter";
	TableDateCompareTypeEnum["Month"] = "month";
	TableDateCompareTypeEnum["M1"] = "m1";
	TableDateCompareTypeEnum["M2"] = "m2";
	TableDateCompareTypeEnum["M3"] = "m3";
	TableDateCompareTypeEnum["M4"] = "m4";
	TableDateCompareTypeEnum["M5"] = "m5";
	TableDateCompareTypeEnum["M6"] = "m6";
	TableDateCompareTypeEnum["M7"] = "m7";
	TableDateCompareTypeEnum["M8"] = "m8";
	TableDateCompareTypeEnum["M9"] = "m9";
	TableDateCompareTypeEnum["M10"] = "m10";
	TableDateCompareTypeEnum["M11"] = "m11";
	TableDateCompareTypeEnum["M12"] = "m12";
	TableDateCompareTypeEnum["Q1"] = "q1";
	TableDateCompareTypeEnum["Q2"] = "q2";
	TableDateCompareTypeEnum["Q3"] = "q3";
	TableDateCompareTypeEnum["Q4"] = "q4";
	return TableDateCompareTypeEnum;
}({});
/**
* Represents the pivot cell style type enum
*/
let SheetsTableButtonStateEnum = /* @__PURE__ */ function(SheetsTableButtonStateEnum) {
	SheetsTableButtonStateEnum[SheetsTableButtonStateEnum["FilteredSortNone"] = 1] = "FilteredSortNone";
	SheetsTableButtonStateEnum[SheetsTableButtonStateEnum["FilteredSortAsc"] = 2] = "FilteredSortAsc";
	SheetsTableButtonStateEnum[SheetsTableButtonStateEnum["FilteredSortDesc"] = 3] = "FilteredSortDesc";
	SheetsTableButtonStateEnum[SheetsTableButtonStateEnum["FilterNoneSortNone"] = 4] = "FilterNoneSortNone";
	SheetsTableButtonStateEnum[SheetsTableButtonStateEnum["FilterNoneSortAsc"] = 5] = "FilterNoneSortAsc";
	SheetsTableButtonStateEnum[SheetsTableButtonStateEnum["FilterNoneSortDesc"] = 6] = "FilterNoneSortDesc";
	return SheetsTableButtonStateEnum;
}({});
let SheetsTableSortStateEnum = /* @__PURE__ */ function(SheetsTableSortStateEnum) {
	SheetsTableSortStateEnum["Asc"] = "asc";
	SheetsTableSortStateEnum["Desc"] = "desc";
	SheetsTableSortStateEnum["None"] = "none";
	return SheetsTableSortStateEnum;
}({});

//#endregion
//#region src/util.ts
function getColumnName(columnIndex, columnText) {
	return `${columnText} ${columnIndex}`;
}
const BooleanTrue = "TRUE";
const BooleanFalse = "FALSE";
const getStringFromDataStream$1 = (data) => {
	var _data$body;
	return ((_data$body = data.body) === null || _data$body === void 0 ? void 0 : _data$body.dataStream.replace(/\r\n$/, "")) || "";
};
/**
*  transform cell data to dimension name
* @param cellData the sheet cell data
* @param styles workBook styles collection
* @param patternInfoRecord The cache record for pattern info
* @returns {string} The dimension name
*/
function convertCellDataToString(cellData) {
	if (cellData) {
		const { v, t, p } = cellData;
		if (p) return getStringFromDataStream$1(p);
		if ((t === CellValueType.FORCE_STRING || t === CellValueType.STRING) && v !== void 0 && v !== null) return String(v);
		else if (t === CellValueType.BOOLEAN) return v ? BooleanTrue : BooleanFalse;
		else if (t === CellValueType.NUMBER) return String(v);
		else {
			if (typeof v === "boolean") return v ? BooleanTrue : BooleanFalse;
			return v === void 0 || v === null ? "" : String(v);
		}
	}
	return "";
}
function getTableFilterState(tableFilter, sortState) {
	if (tableFilter !== void 0 && tableFilter !== null) switch (sortState) {
		case "asc": return 2;
		case "desc": return 3;
		default: return 1;
	}
	else switch (sortState) {
		case "asc": return 5;
		case "desc": return 6;
		default: return 4;
	}
}
function isConditionFilter(filter) {
	if (!filter) return false;
	return filter.filterType === "condition";
}
function isManualTableFilter(filter) {
	if (!filter) return false;
	return filter.filterType === "manual";
}
/**
* Get existing names including sheet names, table names and defined names to check for duplicates table name.
*/
function getExistingNamesSet(unitId, options) {
	const { univerInstanceService, tableManager, definedNamesService } = options;
	const existingNamesSet = /* @__PURE__ */ new Set();
	const workbook = univerInstanceService === null || univerInstanceService === void 0 ? void 0 : univerInstanceService.getUnit(unitId, UniverInstanceType.UNIVER_SHEET);
	if (workbook) workbook.getSheets().forEach((sheet) => {
		existingNamesSet.add(sheet.getName().toLowerCase());
	});
	const tableList = tableManager === null || tableManager === void 0 ? void 0 : tableManager.getTableList(unitId);
	if (tableList && tableList.length > 0) tableList.forEach((tableItem) => {
		existingNamesSet.add(tableItem.name.toLowerCase());
	});
	const definedNames = definedNamesService === null || definedNamesService === void 0 ? void 0 : definedNamesService.getDefinedNameMap(unitId);
	if (definedNames) Object.values(definedNames).forEach((definedName) => {
		existingNamesSet.add(definedName.name.toLowerCase());
	});
	return existingNamesSet;
}

//#endregion
//#region src/config/config.ts
const tableDefaultBorderStyle = {
	s: BorderStyleTypes.THIN,
	cl: { rgb: "rgb(95, 101, 116)" }
};
const SHEETS_TABLE_PLUGIN_CONFIG_KEY = "sheets-table.config";
const configSymbol = Symbol(SHEETS_TABLE_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/controllers/table-theme.factory.ts
const customEmptyThemeWithBorderStyle = {
	headerRowStyle: { bd: { t: tableDefaultBorderStyle } },
	headerColumnStyle: { bd: { l: tableDefaultBorderStyle } },
	lastColumnStyle: { bd: { r: tableDefaultBorderStyle } },
	lastRowStyle: { bd: { b: tableDefaultBorderStyle } }
};
const processStyleWithBorderStyle = (key, style) => {
	if (key === "headerRowStyle") {
		if (!style.bd) return {
			...style,
			bd: { t: tableDefaultBorderStyle }
		};
	} else if (key === "lastRowStyle") {
		if (!style.bd) return {
			...style,
			bd: { b: tableDefaultBorderStyle }
		};
	} else if (key === "lastColumnStyle") {
		if (!style.bd) return {
			...style,
			bd: { r: tableDefaultBorderStyle }
		};
	} else if (key === "headerColumnStyle") {
		if (!style.bd) return {
			...style,
			bd: { l: tableDefaultBorderStyle }
		};
	}
	return style;
};
const tableDefaultThemeStyleArr = [
	[[
		"#6280F9",
		"#FFFFFF",
		"#EEF2FF",
		"#DCE4FF"
	], ["#fff"]],
	[[
		"#16BDCA",
		"#FFFFFF",
		"#EDFAFA",
		"#AFECEF"
	], ["#000"]],
	[[
		"#31C48D",
		"#FFFFFF",
		"#F3FAF7",
		"#BCF0DA"
	], ["#fff"]],
	[[
		"#AC94FA",
		"#FFFFFF",
		"#F6F5FF",
		"#EDEBFE"
	], ["#fff"]],
	[[
		"#F17EBB",
		"#FFFFFF",
		"#FDF2F8",
		"#FCE8F3"
	], ["#fff"]],
	[[
		"#F98080",
		"#FFFFFF",
		"#FDF2F2",
		"#FDE8E8"
	], ["#fff"]]
];
const tableThemeConfig = tableDefaultThemeStyleArr.map((item, index) => {
	const [backgroundArr, colorArr] = item;
	const [headerRowBg, firstRowBg, secondRowBg, lastRowBg] = backgroundArr;
	const [headerCl] = colorArr;
	return {
		name: `table-default-${index}`,
		style: {
			headerRowStyle: {
				bg: { rgb: headerRowBg },
				cl: { rgb: headerCl },
				bd: { t: tableDefaultBorderStyle }
			},
			headerColumnStyle: { bd: { l: tableDefaultBorderStyle } },
			firstRowStyle: { bg: { rgb: firstRowBg } },
			secondRowStyle: { bg: { rgb: secondRowBg } },
			lastRowStyle: {
				bg: { rgb: lastRowBg },
				bd: { b: tableDefaultBorderStyle }
			},
			lastColumnStyle: { bd: { r: tableDefaultBorderStyle } }
		}
	};
});

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region src/model/table-column.ts
var TableColumn = class {
	constructor(id, name) {
		_defineProperty(this, "dataType", void 0);
		_defineProperty(this, "id", void 0);
		_defineProperty(this, "displayName", void 0);
		_defineProperty(this, "formula", void 0);
		_defineProperty(this, "meta", void 0);
		_defineProperty(this, "style", void 0);
		this.id = id;
		this.displayName = name;
		this.dataType = "string";
		this.formula = "";
		this.meta = {};
		this.style = {};
	}
	getMeta() {
		return this.meta;
	}
	setMeta(meta) {
		this.meta = meta;
	}
	getDisplayName() {
		return this.displayName;
	}
	toJSON() {
		return {
			id: this.id,
			displayName: this.displayName,
			dataType: this.dataType,
			formula: this.formula,
			meta: this.meta,
			style: this.style
		};
	}
	fromJSON(json) {
		this.id = json.id;
		this.displayName = json.displayName;
		this.dataType = json.dataType;
		this.formula = json.formula;
		this.meta = json.meta;
		this.style = json.style;
	}
};

//#endregion
//#region src/const.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const PLUGIN_NAME = "SHEET_TABLE_PLUGIN";
const FEATURE_TABLE_ID = "SHEET_TABLE";
const SHEET_TABLE_CUSTOM_THEME_PREFIX = "table-custom";
const TABLE_FILTER_EMPTY_VALUE = "__UNIVER_TABLE_FILTER_EMPTY__";

//#endregion
//#region src/model/filter-util/date-filter-util.ts
/**
* The provided date is a date in Q1 of the year.
* @param {Date} date - The date to compare.
* @returns {boolean} return the date is match
*/
const dateQ1 = (date) => {
	return date.getMonth() <= 2;
};
/**
* The provided date is a date in Q2 of the year.
* @param {Date} date - The date to compare.
* @returns {boolean} return the date is match
*/
const dateQ2 = (date) => {
	const month = date.getMonth();
	return month > 2 && month <= 5;
};
/**
* The provided date is a date in Q3 of the year.
* @param {Date} date - The date to compare.
* @returns {boolean} return the date is match
*/
const dateQ3 = (date) => {
	const month = date.getMonth();
	return month > 5 && month <= 8;
};
/**
* The provided date is a date in Q4 of the year.
* @param {Date} date - The date to compare.
* @returns {boolean} return the date is match
*/
const dateQ4 = (date) => {
	const month = date.getMonth();
	return month > 8 && month <= 11;
};
/**
* The provided date is a date in January.
* @param {Date} date - The date to compare.
* @returns {boolean} return the date is match
*/
const dateM1 = (date) => {
	return date.getMonth() === 0;
};
/**
* The provided date is a date in February.
* @param {Date} date - The date to compare.
* @returns {boolean} return the date is match
*/
const dateM2 = (date) => {
	return date.getMonth() === 1;
};
/**
* The provided date is a date in March.
* @param {Date} date - The date to compare.
* @returns {boolean} return the date is match
*/
const dateM3 = (date) => {
	return date.getMonth() === 2;
};
/**
* The provided date is a date in April.
* @param {Date} date - The date to compare.
* @returns {boolean} return the date is match
*/
const dateM4 = (date) => {
	return date.getMonth() === 3;
};
/**
* The provided date is a date in May.
* @param {Date} date - The date to compare.
* @returns {boolean} return the date is match
*/
const dateM5 = (date) => {
	return date.getMonth() === 4;
};
/**
* The provided date is a date in June.
* @param {Date} date - The date to compare.
* @returns {boolean} return the date is match
*/
const dateM6 = (date) => {
	return date.getMonth() === 5;
};
/**
* The provided date is a date in July.
* @param {Date} date - The date to compare.
* @returns {boolean} return the date is match
*/
const dateM7 = (date) => {
	return date.getMonth() === 6;
};
/**
* The provided date is a date in August.
* @param {Date} date - The date to compare.
* @returns {boolean} return the date is match
*/
const dateM8 = (date) => {
	return date.getMonth() === 7;
};
/**
* The provided date is a date in September.
* @param {Date} date - The date to compare.
* @returns {boolean} return the date is match
*/
const dateM9 = (date) => {
	return date.getMonth() === 8;
};
/**
* The provided date is a date in October.
* @param {Date} date - The date to compare.
* @returns {boolean} return the date is match
*/
const dateM10 = (date) => {
	return date.getMonth() === 9;
};
/**
* The provided date is a date in November.
* @param {Date} date - The date to compare.
* @returns {boolean} return the date is match
*/
const dateM11 = (date) => {
	return date.getMonth() === 10;
};
/**
* The provided date is a date in December.
* @param {Date} date - The date to compare.
* @returns {boolean} return the date is match
*/
const dateM12 = (date) => {
	return date.getMonth() === 11;
};
/**
* The provided date is today.
* @param {Date} expectedDate - The date to compare.
* @param {Date} [anchorTime] - The reference date.
* @returns {boolean} return the date is match
*/
const today = (expectedDate, anchorTime = /* @__PURE__ */ new Date()) => {
	return expectedDate.toDateString() === anchorTime.toDateString();
};
/**
* The provided date is tomorrow.
* @param {Date} date - The date to compare.
* @param {Date} [anchorTime] - The reference date.
* @returns {boolean} return the date is match
*/
const tomorrow = (date, anchorTime = /* @__PURE__ */ new Date()) => {
	const tomorrow = new Date(anchorTime);
	tomorrow.setDate(tomorrow.getDate() + 1);
	return date.toDateString() === tomorrow.toDateString();
};
/**
* The provided date is yesterday.
* @param {Date} date - The date to compare.
* @param {Date} [anchorTime] - The reference date.
* @returns {boolean} return the date is match
*/
const yesterday = (date, anchorTime = /* @__PURE__ */ new Date()) => {
	const yesterday = new Date(anchorTime);
	yesterday.setDate(yesterday.getDate() - 1);
	return date.toDateString() === yesterday.toDateString();
};
/**
* Get the start date of the week for the provided date.
* @param {Date} date - The date to get the week start.
* @returns {Date} The start date of the week.
*/
const getWeekStart = (date) => {
	const day = date.getDay();
	const diff = date.getDate() - day + (day === 0 ? -6 : 1);
	const weekStart = new Date(date);
	weekStart.setDate(diff);
	return weekStart;
};
const perWeek = 10080 * 60 * 1e3;
/**
* The provided date is in the current week.
* @param {Date} date - The date to compare.
* @param {Date} [anchorTime] - The reference date.
* @returns {boolean} return the date is match
*/
const thisWeek = (date, anchorTime = /* @__PURE__ */ new Date()) => {
	const weekStart = getWeekStart(date);
	const anchorTimeWeekStart = getWeekStart(anchorTime);
	return weekStart.toDateString() === anchorTimeWeekStart.toDateString();
};
/**
* The provided date is in the next week.
* @param {Date} date - The date to compare.
* @param {Date} [anchorTime] - The reference date.
* @returns {boolean} return the date is match
*/
const nextWeek = (date, anchorTime = /* @__PURE__ */ new Date()) => {
	const weekStart = getWeekStart(date);
	const anchorTimeNextWeekStart = new Date(getWeekStart(anchorTime).getTime() + perWeek);
	return weekStart.toDateString() === anchorTimeNextWeekStart.toDateString();
};
/**
* The provided date is in the last week.
* @param {Date} date - The date to compare.
* @param {Date} [anchorTime] - The reference date.
* @returns {boolean} return the date is match
*/
const lastWeek = (date, anchorTime = /* @__PURE__ */ new Date()) => {
	const weekStart = getWeekStart(date);
	const anchorTimeLastWeekStart = /* @__PURE__ */ new Date(getWeekStart(anchorTime).getTime() - perWeek);
	return weekStart.toDateString() === anchorTimeLastWeekStart.toDateString();
};
/**
* The provided date is in the current month.
* @param {Date} date - The date to compare.
* @param {Date} [anchorTime] - The reference date.
* @returns {boolean} return the date is match
*/
const thisMonth = (date, anchorTime = /* @__PURE__ */ new Date()) => {
	return date.getFullYear() === anchorTime.getFullYear() && date.getMonth() === anchorTime.getMonth();
};
/**
* Get the start date of the month for the provided date.
* @param {Date} date - The date to get the month start.
* @returns {Date} The start date of the month.
*/
const getMonthStart = (date) => {
	const monthStart = new Date(date);
	monthStart.setHours(0, 0, 0, 0);
	monthStart.setDate(1);
	return monthStart;
};
/**
* The provided date is in the next month.
* @param {Date} date - The date to compare.
* @param {Date} [anchorTime] - The reference date.
* @returns {boolean} return the date is match
*/
const nextMonth = (date, anchorTime = /* @__PURE__ */ new Date()) => {
	const nextMonthStart = new Date(anchorTime);
	nextMonthStart.setHours(0, 0, 0, 0);
	nextMonthStart.setMonth(nextMonthStart.getMonth() + 1, 1);
	const monthEnd = new Date(nextMonthStart);
	monthEnd.setMonth(monthEnd.getMonth() + 1, 0);
	const dateTime = date.getTime();
	return dateTime >= nextMonthStart.getTime() && dateTime < monthEnd.getTime();
};
/**
* The provided date is in the last month.
* @param {Date} date - The date to compare.
* @param {Date} [anchorTime] - The reference date.
* @returns {boolean} return the date is match
*/
const lastMonth = (date, anchorTime = /* @__PURE__ */ new Date()) => {
	const lastMonthStart = getMonthStart(anchorTime);
	const monthEnd = new Date(lastMonthStart);
	monthEnd.setMonth(monthEnd.getMonth() + 1, 0);
	const dateTime = date.getTime();
	return dateTime >= lastMonthStart.getTime() && dateTime < monthEnd.getTime();
};
/**
* Get the start date of the quarter for the provided date.
* @param {Date} date - The date to get the quarter start.
* @returns {Date} The start date of the quarter.
*/
const getQuarterStart = (date) => {
	const quarterStart = new Date(date);
	quarterStart.setHours(0, 0, 0, 0);
	quarterStart.setDate(1);
	const month = quarterStart.getMonth();
	quarterStart.setMonth(month - month % 3);
	return quarterStart;
};
/**
* The provided date is in the current quarter.
* @param {Date} date - The date to compare.
* @param {Date} [anchorTime] - The reference date.
* @returns {boolean} return the date is match
*/
const thisQuarter = (date, anchorTime = /* @__PURE__ */ new Date()) => {
	const quarterStart = getQuarterStart(anchorTime);
	const nextQuarterStart = new Date(quarterStart);
	nextQuarterStart.setMonth(nextQuarterStart.getMonth() + 3);
	const dateTime = date.getTime();
	return dateTime >= quarterStart.getTime() && dateTime < nextQuarterStart.getTime();
};
/**
* The provided date is in the next quarter.
* @param {Date} date - The date to compare.
* @param {Date} [anchorTime] - The reference date.
* @returns {boolean} return the date is match
*/
const nextQuarter = (date, anchorTime = /* @__PURE__ */ new Date()) => {
	const quarterStart = getQuarterStart(anchorTime);
	const nextQuarterStart = new Date(quarterStart);
	nextQuarterStart.setMonth(nextQuarterStart.getMonth() + 3);
	const nextQuarterEnd = new Date(nextQuarterStart);
	nextQuarterEnd.setMonth(nextQuarterEnd.getMonth() + 3, 0);
	const dateTime = date.getTime();
	return dateTime >= nextQuarterStart.getTime() && dateTime < nextQuarterEnd.getTime();
};
/**
* The provided date is in the last quarter.
* @param {Date} date - The date to compare.
* @param {Date} [anchorTime] - The reference date.
* @returns {boolean} return the date is match
*/
const lastQuarter = (date, anchorTime = /* @__PURE__ */ new Date()) => {
	const quarterStart = getQuarterStart(anchorTime);
	const lastQuarterStart = new Date(quarterStart);
	lastQuarterStart.setMonth(lastQuarterStart.getMonth() - 3);
	const lastQuarterEnd = new Date(quarterStart);
	lastQuarterEnd.setDate(0);
	const dateTime = date.getTime();
	return dateTime >= lastQuarterStart.getTime() && dateTime < lastQuarterEnd.getTime();
};
/**
* The provided date is in the current year.
* @param {Date} date - The date to compare.
* @param {Date} [anchorTime] - The reference date.
* @returns {boolean} return the date is match
*/
const thisYear = (date, anchorTime = /* @__PURE__ */ new Date()) => {
	return date.getFullYear() === anchorTime.getFullYear();
};
/**
* The provided date is in the next year.
* @param {Date} date - The date to compare.
* @param {Date} [anchorTime] - The reference date.
* @returns {boolean} return the date is match
*/
const nextYear = (date, anchorTime = /* @__PURE__ */ new Date()) => {
	return date.getFullYear() === anchorTime.getFullYear() + 1;
};
/**
* The provided date is in the last year.
* @param {Date} date - The date to compare.
* @param {Date} [anchorTime] - The reference date.
* @returns {boolean} return the date is match
*/
const lastYear = (date, anchorTime = /* @__PURE__ */ new Date()) => {
	return date.getFullYear() === anchorTime.getFullYear() - 1;
};
/**
* The provided date is in the year to date.
* @param {Date} date - The date to compare.
* @param {Date} [anchorTime] - The reference date.
* @returns {boolean} return the date is match
*/
const yearToDate = (date, anchorTime = /* @__PURE__ */ new Date()) => {
	const yearStart = new Date(anchorTime);
	yearStart.setHours(0, 0, 0, 0);
	yearStart.setMonth(0, 1);
	const dateTime = date.getTime();
	return dateTime >= yearStart.getTime() && dateTime < anchorTime.getTime();
};
function getDateFilterExecuteFunc(filterInfo) {
	switch (filterInfo.compareType) {
		case "equal": {
			const expected = new Date(filterInfo.expectedValue);
			return (date) => date.getTime() === expected.getTime();
		}
		case "notEqual": {
			const expected = new Date(filterInfo.expectedValue);
			return (date) => date.getTime() !== expected.getTime();
		}
		case "after": {
			const expected = new Date(filterInfo.expectedValue);
			return (date) => date.getTime() > expected.getTime();
		}
		case "before": {
			const expected = new Date(filterInfo.expectedValue);
			return (date) => date.getTime() < expected.getTime();
		}
		case "afterOrEqual": {
			const expected = new Date(filterInfo.expectedValue);
			return (date) => date.getTime() >= expected.getTime();
		}
		case "beforeOrEqual": {
			const expected = new Date(filterInfo.expectedValue);
			return (date) => date.getTime() <= expected.getTime();
		}
		case "between": return (date) => {
			const [start, end] = filterInfo.expectedValue;
			return date.getTime() >= new Date(start).getTime() && date.getTime() <= new Date(end).getTime();
		};
		case "notBetween": return (date) => {
			const [start, end] = filterInfo.expectedValue;
			return date.getTime() < new Date(start).getTime() || date.getTime() > new Date(end).getTime();
		};
		case "today": {
			const anchorTime = filterInfo.anchorTime ? new Date(filterInfo.anchorTime) : /* @__PURE__ */ new Date();
			return (date) => today(date, anchorTime);
		}
		case "yesterday": {
			const anchorTime = filterInfo.anchorTime ? new Date(filterInfo.anchorTime) : /* @__PURE__ */ new Date();
			return (date) => yesterday(date, anchorTime);
		}
		case "tomorrow": {
			const anchorTime = filterInfo.anchorTime ? new Date(filterInfo.anchorTime) : /* @__PURE__ */ new Date();
			return (date) => tomorrow(date, anchorTime);
		}
		case "thisWeek": {
			const anchorTime = filterInfo.anchorTime ? new Date(filterInfo.anchorTime) : /* @__PURE__ */ new Date();
			return (date) => thisWeek(date, anchorTime);
		}
		case "lastWeek": {
			const anchorTime = filterInfo.anchorTime ? new Date(filterInfo.anchorTime) : /* @__PURE__ */ new Date();
			return (date) => lastWeek(date, anchorTime);
		}
		case "nextWeek": {
			const anchorTime = filterInfo.anchorTime ? new Date(filterInfo.anchorTime) : /* @__PURE__ */ new Date();
			return (date) => nextWeek(date, anchorTime);
		}
		case "thisMonth": {
			const anchorTime = filterInfo.anchorTime ? new Date(filterInfo.anchorTime) : /* @__PURE__ */ new Date();
			return (date) => thisMonth(date, anchorTime);
		}
		case "lastMonth": {
			const anchorTime = filterInfo.anchorTime ? new Date(filterInfo.anchorTime) : /* @__PURE__ */ new Date();
			return (date) => lastMonth(date, anchorTime);
		}
		case "nextMonth": {
			const anchorTime = filterInfo.anchorTime ? new Date(filterInfo.anchorTime) : /* @__PURE__ */ new Date();
			return (date) => nextMonth(date, anchorTime);
		}
		case "thisQuarter": {
			const anchorTime = filterInfo.anchorTime ? new Date(filterInfo.anchorTime) : /* @__PURE__ */ new Date();
			return (date) => thisQuarter(date, anchorTime);
		}
		case "lastQuarter": {
			const anchorTime = filterInfo.anchorTime ? new Date(filterInfo.anchorTime) : /* @__PURE__ */ new Date();
			return (date) => lastQuarter(date, anchorTime);
		}
		case "nextQuarter": {
			const anchorTime = filterInfo.anchorTime ? new Date(filterInfo.anchorTime) : /* @__PURE__ */ new Date();
			return (date) => nextQuarter(date, anchorTime);
		}
		case "thisYear": {
			const anchorTime = filterInfo.anchorTime ? new Date(filterInfo.anchorTime) : /* @__PURE__ */ new Date();
			return (date) => thisYear(date, anchorTime);
		}
		case "lastYear": {
			const anchorTime = filterInfo.anchorTime ? new Date(filterInfo.anchorTime) : /* @__PURE__ */ new Date();
			return (date) => lastYear(date, anchorTime);
		}
		case "nextYear": {
			const anchorTime = filterInfo.anchorTime ? new Date(filterInfo.anchorTime) : /* @__PURE__ */ new Date();
			return (date) => nextYear(date, anchorTime);
		}
		case "yearToDate": {
			const anchorTime = filterInfo.anchorTime ? new Date(filterInfo.anchorTime) : /* @__PURE__ */ new Date();
			return (date) => yearToDate(date, anchorTime);
		}
		case "m1": return dateM1;
		case "m2": return dateM2;
		case "m3": return dateM3;
		case "m4": return dateM4;
		case "m5": return dateM5;
		case "m6": return dateM6;
		case "m7": return dateM7;
		case "m8": return dateM8;
		case "m9": return dateM9;
		case "m10": return dateM10;
		case "m11": return dateM11;
		case "m12": return dateM12;
		case "q1": return dateQ1;
		case "q2": return dateQ2;
		case "q3": return dateQ3;
		case "q4": return dateQ4;
		default: throw new Error("Unsupported compare type");
	}
}

//#endregion
//#region src/model/filter-util/top-n.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
/**
* @description Represents a heap data structure.
* A heap is a complete binary tree that satisfies the heap property.
* The heap property states that the parent node is always smaller or bigger than its children.
* The root node is the smallest or biggest element in the heap.
* The heap is used to find the kth largest or smallest element in a list.
* A array can be used to represent a heap.
* The root node is at index 0, and the left child of a node at index i is at index 2i + 1.
* The right child of a node at index i is at index 2i + 2.
* The parent node of a node at index i is at index (i - 1) / 2.
* @example
* const maxHeap = [5, 3, 4, 1, 2];
* // const minHeap = [6, 8, 7, 10, 9];
*  // the root node in maxHeap is 5
*  // the root child is 3, 4, then the child of 3 is 1, 2
*/
var Heap = class {
	/**
	* Initializes a new instance of the Heap class.
	*/
	constructor() {
		_defineProperty(this, "heap", void 0);
		this.heap = [];
	}
	/**
	* Swaps the elements at the given indices in the heap.
	* @param index1 The index of the first element.
	* @param index2 The index of the second element.
	*/
	swap(index1, index2) {
		const temp = this.heap[index1];
		this.heap[index1] = this.heap[index2];
		this.heap[index2] = temp;
	}
	/**
	* Returns the index of the parent node for the given index.
	* @param index The index of the node.
	* @returns The index of the parent node.
	*/
	getParentIndex(index) {
		return Math.floor((index - 1) / 2);
	}
	/**
	* Returns the index of the left child node for the given index.
	* @param index The index of the node.
	* @returns The index of the left child node.
	*/
	getLeftIndex(index) {
		return index * 2 + 1;
	}
	/**
	* Returns the index of the right child node for the given index.
	* @param index The index of the node.
	* @returns The index of the right child node.
	*/
	getRightIndex(index) {
		return index * 2 + 2;
	}
	/**
	* Returns the number of elements in the heap.
	* @returns The number of elements in the heap.
	*/
	size() {
		return this.heap.length;
	}
	/**
	* Returns the minimum value in the heap without removing it.
	* @returns The minimum value in the heap.
	*/
	peek() {
		return this.heap[0];
	}
	/**
	* @description Returns whether the heap includes the given value.
	* @param {number} value  The value to be checked.
	* @returns {boolean} return true if the heap includes the given value
	*/
	include(value) {
		return this.heap.includes(value);
	}
};
/**
* @description Represents a min heap data structure.
* in MinHeap, the parent node is always smaller than its children.
* The root node is the smallest element in the heap.
* The min heap is used to find the kth largest element in a list.
*/
var MinHeap = class extends Heap {
	/**
	* Initializes a new instance of the MinHeap class.
	*/
	constructor() {
		super();
	}
	/**
	* Moves the element at the given index up the heap until it satisfies the min heap property.
	* @param index The index of the element to be shifted up.
	*/
	shiftUp(index) {
		if (index === 0) return;
		const parentIndex = this.getParentIndex(index);
		if (this.heap[parentIndex] > this.heap[index]) {
			this.swap(parentIndex, index);
			this.shiftUp(parentIndex);
		}
	}
	/**
	* Moves the element at the given index down the heap until it satisfies the min heap property.
	* @param index The index of the element to be shifted down.
	*/
	shiftDown(index) {
		const leftIndex = this.getLeftIndex(index);
		const rightIndex = this.getRightIndex(index);
		if (this.heap[leftIndex] < this.heap[index]) {
			this.swap(leftIndex, index);
			this.shiftDown(leftIndex);
		}
		if (this.heap[rightIndex] < this.heap[index]) {
			this.swap(rightIndex, index);
			this.shiftDown(rightIndex);
		}
	}
	/**
	* Inserts a new value into the min heap.
	* @param value The value to be inserted.
	*/
	insert(value) {
		this.heap.push(value);
		this.shiftUp(this.heap.length - 1);
	}
	/**
	* Removes and returns the minimum value from the min heap.
	*/
	pop() {
		this.heap[0] = this.heap.pop();
		this.shiftDown(0);
	}
};
/**
* Returns the kth largest element from the given list using a min heap.
* @param {number[]} list The list of numbers.
* @param {number} k The value of k.
* @returns The kth largest element.
*/
const getLargestK = (list, k) => {
	const minHeap = new MinHeap();
	for (const item of list) {
		minHeap.insert(item);
		if (minHeap.size() > k) minHeap.pop();
	}
	return minHeap.heap;
};

//#endregion
//#region src/model/filter-util/number-filter-util.ts
/**
* @description Checks if a value is above the average.
* @param {number} value - The value to check.
* @param {number} average - The average value.
* @returns {boolean} A boolean value indicating whether the value is above the average.
*/
const above = (value, average) => {
	return value > average;
};
/**
* @description Checks if a value is below the average.
* @param {number} value - The value to check.
* @param {number} average - The average value.
* @returns {boolean} A boolean value indicating whether the value is below the average.
*/
const below = (value, average) => {
	return value < average;
};
/**
* @description Gets the largest N values from a list and checks if the expected value is included.
* @param {number[]} list - The list of numbers.
* @param {number} top - The number of top values to retrieve.
* @param {number} expectedValue - The expected value to check for inclusion.
* @returns {boolean} A boolean value indicating whether the expected value is included in the top N values.
*/
const getTopN = (list, top, expectedValue) => {
	return getLargestK(list, top).includes(expectedValue);
};
function getNumberFilterExecuteFunc(filter, calculatedOptions) {
	switch (filter.compareType) {
		case "equal": {
			const expectedValue = Number(filter.expectedValue);
			return (value) => value === expectedValue;
		}
		case "notEqual": {
			const expectedValue = Number(filter.expectedValue);
			return (value) => value !== expectedValue;
		}
		case "greaterThan": {
			const expectedValue = Number(filter.expectedValue);
			return (value) => value > expectedValue;
		}
		case "greaterThanOrEqual": {
			const expectedValue = Number(filter.expectedValue);
			return (value) => value >= expectedValue;
		}
		case "lessThan": {
			const expectedValue = Number(filter.expectedValue);
			return (value) => value < expectedValue;
		}
		case "lessThanOrEqual": {
			const expectedValue = Number(filter.expectedValue);
			return (value) => value <= expectedValue;
		}
		case "between": {
			const [min, max] = filter.expectedValue;
			const minValue = Number(min);
			const maxValue = Number(max);
			if (minValue > maxValue) return (value) => value >= maxValue && value <= minValue;
			return (value) => value >= minValue && value <= maxValue;
		}
		case "notBetween": {
			const [min, max] = filter.expectedValue;
			const minValue = Number(min);
			const maxValue = Number(max);
			if (minValue > maxValue) return (value) => value < maxValue || value > minValue;
			return (value) => value < minValue || value > maxValue;
		}
		case "above": {
			const average = calculatedOptions.average;
			return (value) => above(value, average);
		}
		case "below": {
			const average = calculatedOptions.average;
			return (value) => below(value, average);
		}
		case "topN": {
			const list = calculatedOptions.list;
			const top = Number(filter.expectedValue);
			return (value) => getTopN(list, top, value);
		}
	}
}

//#endregion
//#region src/model/filter-util/text-filter-util.ts
const textEqual = (compareValue, expectedValue) => {
	return createREGEXFromWildChar(expectedValue).test(compareValue);
};
const textNotEqual = (compareValue, expectedValue) => {
	return !createREGEXFromWildChar(expectedValue).test(compareValue);
};
const textContain = (compareValue, expectedValue) => {
	return createREGEXFromWildChar(`*${expectedValue}*`).test(compareValue);
};
const textNotContain = (compareValue, expectedValue) => {
	return !createREGEXFromWildChar(`*${expectedValue}*`).test(compareValue);
};
const textStartWith = (compareValue, expectedValue) => {
	return createREGEXFromWildChar(`${expectedValue}*`).test(compareValue);
};
const textEndWith = (compareValue, expectedValue) => {
	return createREGEXFromWildChar(`*${expectedValue}`).test(compareValue);
};
function getTextFilterExecuteFunc(filter) {
	switch (filter.compareType) {
		case "equal": return (value) => textEqual(value, filter.expectedValue);
		case "notEqual": return (value) => textNotEqual(value, filter.expectedValue);
		case "contains": return (value) => textContain(value, filter.expectedValue);
		case "notContains": return (value) => textNotContain(value, filter.expectedValue);
		case "startsWith": return (value) => textStartWith(value, filter.expectedValue);
		case "endsWith": return (value) => textEndWith(value, filter.expectedValue);
		default:
			console.error(`Unknown filter operator: ${filter.compareType}`);
			return (value) => true;
	}
}

//#endregion
//#region src/model/filter-util/condition.ts
const NumberDynamicFilterCompareTypeSet = new Set([
	"above",
	"below",
	"topN"
]);
const DateDynamicFilterCompareTypeSet = new Set([
	"today",
	"yesterday",
	"tomorrow",
	"thisWeek",
	"lastWeek",
	"nextWeek",
	"thisMonth",
	"lastMonth",
	"nextMonth",
	"thisQuarter",
	"lastQuarter",
	"nextQuarter",
	"nextYear",
	"thisYear",
	"lastYear",
	"yearToDate"
]);
function isNumberDynamicFilter(compareType) {
	return NumberDynamicFilterCompareTypeSet.has(compareType);
}
function getConditionExecuteFunc(filter, calculatedOptions) {
	if (isNumberDynamicFilter(filter.filterInfo.compareType)) return (value) => true;
	else switch (filter.filterInfo.conditionType) {
		case "date": return getDateFilterExecuteFunc(filter.filterInfo);
		case "number": return getNumberFilterExecuteFunc(filter.filterInfo, calculatedOptions);
		case "string": return getTextFilterExecuteFunc(filter.filterInfo);
		case "logic":
		default: return (value) => true;
	}
}
function getCellValueWithConditionType(sheet, row, col, conditionType) {
	switch (conditionType) {
		case "date": {
			const dateNumber = getNumberCellValue(sheet, row, col);
			return dateNumber ? excelSerialToDateTime(dateNumber) : null;
		}
		case "number": return getNumberCellValue(sheet, row, col);
		case "string":
		default: return getStringCellValue(sheet, row, col);
	}
}
const getStringFromDataStream = (data) => {
	var _data$body;
	return ((_data$body = data.body) === null || _data$body === void 0 ? void 0 : _data$body.dataStream.replace(/\r\n$/, "")) || "";
};
function getStringCellValue(sheet, row, col) {
	const cellData = sheet.getCell(row, col);
	if (!cellData) return null;
	const { v, t, p } = cellData;
	if (p) return getStringFromDataStream(p);
	if (typeof v === "string") {
		if (t === CellValueType.BOOLEAN) return v.toUpperCase();
		return v;
	}
	if (typeof v === "number") {
		if (t === CellValueType.BOOLEAN) return v ? "TRUE" : "FALSE";
		return v;
	}
	if (typeof v === "boolean") return v ? "TRUE" : "FALSE";
	if (v === void 0) return;
	return String(v);
}
function getNumberCellValue(sheet, row, col) {
	const cellData = sheet.getCell(row, col);
	if (!cellData) return null;
	const { v, t, p } = cellData;
	if (p) return null;
	if (typeof v === "string" && t === CellValueType.NUMBER) return Number(sheet.getCellRaw(row, col).v);
	return Number(v);
}
function excelSerialToDateTime(serial) {
	const baseDate = new Date(Date.UTC(1900, 0, 1, 0, 0, 0));
	const leapDayDate = new Date(Date.UTC(1900, 1, 28, 0, 0, 0));
	let dayDifference = serial - 1;
	if (dayDifference > (leapDayDate.getTime() - baseDate.getTime()) / (1e3 * 3600 * 24)) dayDifference -= 1;
	if (dayDifference < 0) dayDifference = serial;
	return new Date(baseDate.getTime() + dayDifference * (1e3 * 3600 * 24));
}

//#endregion
//#region src/model/table-filter.ts
var TableFilters = class {
	constructor() {
		_defineProperty(this, "_tableColumnFilterList", void 0);
		_defineProperty(this, "_tableSortInfo", void 0);
		_defineProperty(this, "_filterOutRows", void 0);
		this._tableColumnFilterList = [];
	}
	setColumnFilter(columnIndex, filter) {
		if (!filter) this._tableColumnFilterList[columnIndex] = void 0;
		else this._tableColumnFilterList[columnIndex] = filter;
	}
	setSortState(columnIndex, sortState) {
		this._tableSortInfo = {
			columnIndex,
			sortState
		};
	}
	getColumnFilter(columnIndex) {
		return this._tableColumnFilterList[columnIndex];
	}
	getFilterState(columnIndex) {
		var _this$_tableSortInfo;
		const sortState = ((_this$_tableSortInfo = this._tableSortInfo) === null || _this$_tableSortInfo === void 0 ? void 0 : _this$_tableSortInfo.columnIndex) === columnIndex ? this._tableSortInfo.sortState : "none";
		return getTableFilterState(this._tableColumnFilterList[columnIndex], sortState);
	}
	getSortState() {
		var _this$_tableSortInfo2;
		return (_this$_tableSortInfo2 = this._tableSortInfo) !== null && _this$_tableSortInfo2 !== void 0 ? _this$_tableSortInfo2 : {};
	}
	getFilterStates(range) {
		const states = [];
		const { startColumn, endColumn } = range;
		for (let i = startColumn; i <= endColumn; i++) states.push(this.getFilterState(i - startColumn));
		return states;
	}
	getFilterOutRows() {
		return this._filterOutRows;
	}
	doFilter(sheet, range) {
		const filterOutRows = /* @__PURE__ */ new Set();
		const tableColumnFilterList = this._tableColumnFilterList;
		for (let i = 0; i < tableColumnFilterList.length; i++) if (tableColumnFilterList[i]) this.doColumnFilter(sheet, range, i, filterOutRows);
		this._filterOutRows = filterOutRows;
		return filterOutRows;
	}
	doColumnFilter(sheet, range, columnIndex, filterOutRows) {
		const filter = this._tableColumnFilterList[columnIndex];
		if (filter && sheet) {
			const { startRow, endRow, startColumn } = range;
			const column = startColumn + columnIndex;
			const executeFunc = this.getExecuteFunc(sheet, range, columnIndex, filter);
			for (let row = startRow; row <= endRow; row++) {
				const conditionType = isConditionFilter(filter) ? filter.filterInfo.conditionType : "string";
				const cellValue = getCellValueWithConditionType(sheet, row, column, conditionType);
				if (cellValue === null && !executeFunc(cellValue)) filterOutRows.add(row);
				else if (!executeFunc(getCellValueWithConditionType(sheet, row, column, conditionType))) filterOutRows.add(row);
			}
		}
	}
	_getNumberCalculatedOptions(sheet, range, columnIndex) {
		const { startRow, endRow, startColumn } = range;
		const column = startColumn + columnIndex;
		const list = [];
		let count = 0;
		let sum = 0;
		for (let row = startRow; row <= endRow; row++) {
			const val = getCellValueWithConditionType(sheet, row, column, "number");
			if (val !== null) {
				list.push(val);
				count++;
				sum += val;
			}
		}
		return {
			list,
			average: count > 0 ? sum / count : 0
		};
	}
	getExecuteFunc(sheet, range, columnIndex, filter) {
		if (filter.filterType === "manual") {
			const valuesSet = new Set(filter.values);
			return (value) => {
				if (value == null) return valuesSet.has(TABLE_FILTER_EMPTY_VALUE);
				return valuesSet.has(value);
			};
		} else if (filter.filterType === "condition") return getConditionExecuteFunc(filter, isNumberDynamicFilter(filter.filterInfo.compareType) ? this._getNumberCalculatedOptions(sheet, range, columnIndex) : void 0);
		else return (value) => {
			return true;
		};
	}
	toJSON() {
		return {
			tableColumnFilterList: this._tableColumnFilterList,
			tableSortInfo: this._tableSortInfo
		};
	}
	fromJSON(json) {
		var _json$tableColumnFilt;
		this._tableColumnFilterList = (_json$tableColumnFilt = json.tableColumnFilterList) !== null && _json$tableColumnFilt !== void 0 ? _json$tableColumnFilt : [];
		if (json.tableSortInfo) this._tableSortInfo = json.tableSortInfo;
	}
	dispose() {
		this._tableColumnFilterList = [];
	}
};

//#endregion
//#region src/model/table.ts
var Table = class {
	constructor(id, name, range, header, options = {}) {
		_defineProperty(this, "_id", void 0);
		_defineProperty(this, "_name", void 0);
		_defineProperty(this, "_tableStyleId", void 0);
		_defineProperty(this, "_showHeader", void 0);
		_defineProperty(this, "_showFooter", void 0);
		_defineProperty(this, "_range", void 0);
		_defineProperty(this, "_columns", /* @__PURE__ */ new Map());
		_defineProperty(this, "_columnOrder", []);
		_defineProperty(this, "tableMeta", void 0);
		_defineProperty(this, "_tableFilters", void 0);
		_defineProperty(this, "_subUnitId", void 0);
		this._id = id;
		this._range = range;
		this._name = name;
		this._tableFilters = new TableFilters();
		this._init(header, options);
	}
	_init(header, options) {
		var _options$showHeader;
		this._tableStyleId = options === null || options === void 0 ? void 0 : options.tableStyleId;
		this._showHeader = (_options$showHeader = options === null || options === void 0 ? void 0 : options.showHeader) !== null && _options$showHeader !== void 0 ? _options$showHeader : true;
		this._showFooter = false;
		const range = this.getRange();
		const startColumn = range.startColumn;
		const endColumn = range.endColumn;
		for (let i = startColumn; i <= endColumn; i++) {
			var _options$columns;
			const index = i - startColumn;
			let id;
			let columnName;
			if ((_options$columns = options.columns) === null || _options$columns === void 0 ? void 0 : _options$columns[index]) {
				id = options.columns[index].id;
				columnName = options.columns[index].displayName;
			} else {
				id = generateRandomId();
				columnName = header[i - startColumn];
			}
			const column = new TableColumn(id, columnName);
			this._columns.set(id, column);
			this._columnOrder.push(id);
		}
		if (options.filters) options.filters.forEach((filter, index) => {
			if (filter) this._tableFilters.setColumnFilter(index, filter);
		});
	}
	setTableFilterColumn(columnIndex, filter) {
		this._tableFilters.setColumnFilter(columnIndex, filter);
	}
	getTableFilterColumn(columnIndex) {
		return this._tableFilters.getColumnFilter(columnIndex);
	}
	getTableFilters() {
		return this._tableFilters;
	}
	getTableFilterRange() {
		const tableRange = this.getRange();
		const showHeader = this.isShowHeader();
		const isShowFooter = this.isShowFooter();
		const { startRow, startColumn, endRow, endColumn } = tableRange;
		return {
			startRow: showHeader ? startRow + 1 : startRow,
			startColumn,
			endRow: isShowFooter ? endRow - 1 : endRow,
			endColumn
		};
	}
	setColumns(columns) {
		this._columns.clear();
		this._columnOrder = [];
		columns.forEach((columnJson) => {
			const column = new TableColumn(columnJson.id, columnJson.displayName);
			column.fromJSON(columnJson);
			this._columns.set(columnJson.id, column);
			this._columnOrder.push(columnJson.id);
		});
	}
	getColumnsCount() {
		return this._columnOrder.length;
	}
	insertColumn(index, column) {
		const columnId = column.id;
		this._columns.set(columnId, column);
		this._columnOrder.splice(index, 0, columnId);
	}
	removeColumn(index) {
		const columnId = this._columnOrder[index];
		this._columns.delete(columnId);
		this._columnOrder.splice(index, 1);
	}
	setTableMeta(meta) {
		this.tableMeta = meta;
	}
	getTableMeta() {
		return this.tableMeta;
	}
	getColumn(columnId) {
		return this._columns.get(columnId);
	}
	getTableColumnByIndex(index) {
		const id = this._columnOrder[index];
		return this.getColumn(id);
	}
	getColumnNameByIndex(index) {
		var _this$getColumn;
		const id = this._columnOrder[index];
		return ((_this$getColumn = this.getColumn(id)) === null || _this$getColumn === void 0 ? void 0 : _this$getColumn.getDisplayName()) || "";
	}
	getId() {
		return this._id;
	}
	getRangeInfo() {
		return { ...this._range };
	}
	getRange() {
		return { ...this._range };
	}
	setRange(range) {
		this._range = range;
	}
	setDisplayName(name) {
		this._name = name;
	}
	getDisplayName() {
		return this._name;
	}
	getSubunitId() {
		return this._subUnitId;
	}
	setSubunitId(subUnitId) {
		this._subUnitId = subUnitId;
	}
	getTableStyleId() {
		var _this$_tableStyleId;
		return (_this$_tableStyleId = this._tableStyleId) !== null && _this$_tableStyleId !== void 0 ? _this$_tableStyleId : tableThemeConfig[0].name;
	}
	setTableStyleId(tableStyleId) {
		this._tableStyleId = tableStyleId;
	}
	isShowHeader() {
		var _this$_showHeader;
		return (_this$_showHeader = this._showHeader) !== null && _this$_showHeader !== void 0 ? _this$_showHeader : true;
	}
	setShowHeader(showHeader) {
		this._showHeader = showHeader;
	}
	isShowFooter() {
		var _this$_showFooter;
		return (_this$_showFooter = this._showFooter) !== null && _this$_showFooter !== void 0 ? _this$_showFooter : false;
	}
	getTableInfo() {
		return {
			id: this._id,
			subUnitId: this._subUnitId,
			name: this._name,
			range: this.getRangeInfo(),
			meta: this.tableMeta,
			showHeader: this._showHeader,
			columns: this._columnOrder.map((columnId) => this._columns.get(columnId).toJSON())
		};
	}
	getTableConfig() {
		return {
			name: this.getDisplayName(),
			range: this.getRangeInfo(),
			options: {
				showHeader: this._showHeader,
				showFooter: this._showFooter
			},
			tableStyleId: this._tableStyleId
		};
	}
	getFilterStates(range) {
		return this._tableFilters.getFilterStates(range);
	}
	toJSON() {
		const columns = [];
		this._columns.forEach((column) => {
			columns.push(column.toJSON());
		});
		return {
			id: this._id,
			name: this._name,
			range: this.getRangeInfo(),
			options: {
				showHeader: this._showHeader,
				showFooter: this._showFooter,
				tableStyleId: this._tableStyleId
			},
			filters: this._tableFilters.toJSON(),
			columns,
			meta: this.tableMeta
		};
	}
	fromJSON(json) {
		var _json$options$showHea, _json$options$showFoo;
		this._id = json.id;
		this._name = json.name;
		this._range = json.range;
		this.tableMeta = json.meta;
		this._tableStyleId = json.options.tableStyleId || "";
		this._showHeader = (_json$options$showHea = json.options.showHeader) !== null && _json$options$showHea !== void 0 ? _json$options$showHea : true;
		this._showFooter = (_json$options$showFoo = json.options.showFooter) !== null && _json$options$showFoo !== void 0 ? _json$options$showFoo : true;
		json.columns.forEach((column) => {
			const tableColumn = new TableColumn(column.id, column.displayName);
			tableColumn.fromJSON(column);
			this._columns.set(column.id, tableColumn);
			this._columnOrder.push(column.id);
		});
		this._tableFilters = new TableFilters();
		this._tableFilters.fromJSON(json.filters);
	}
	dispose() {
		this._id = "";
		this._name = "";
		this._tableStyleId = "";
		this._showHeader = true;
		this._showFooter = true;
		delete this._range;
		this._columns.clear();
		this._columnOrder = [];
	}
};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/model/table-manager.ts
let TableManager = class TableManager extends Disposable {
	constructor(_univerInstanceService, _localeService) {
		super();
		this._univerInstanceService = _univerInstanceService;
		this._localeService = _localeService;
		_defineProperty(this, "_tableMap", void 0);
		_defineProperty(this, "_tableAdd$", new Subject());
		_defineProperty(this, "tableAdd$", this._tableAdd$.asObservable());
		_defineProperty(this, "_tableDelete$", new Subject());
		_defineProperty(this, "tableDelete$", this._tableDelete$.asObservable());
		_defineProperty(this, "_tableNameChanged$", new Subject());
		_defineProperty(this, "tableNameChanged$", this._tableNameChanged$.asObservable());
		_defineProperty(this, "_tableRangeChanged$", new Subject());
		_defineProperty(this, "tableRangeChanged$", this._tableRangeChanged$.asObservable());
		_defineProperty(this, "_tableThemeChanged$", new Subject());
		_defineProperty(this, "tableThemeChanged$", this._tableThemeChanged$.asObservable());
		_defineProperty(this, "_tableFilterChanged$", new Subject());
		_defineProperty(this, "tableFilterChanged$", this._tableFilterChanged$.asObservable());
		_defineProperty(this, "_tableInitStatus", new BehaviorSubject(false));
		_defineProperty(this, "tableInitStatus$", this._tableInitStatus.asObservable());
		this._tableMap = /* @__PURE__ */ new Map();
	}
	_ensureUnit(unitId) {
		if (!this._tableMap.has(unitId)) this._tableMap.set(unitId, /* @__PURE__ */ new Map());
		return this._tableMap.get(unitId);
	}
	getColumnHeader(unitId, subUnitId, range, prefixText) {
		var _this$_univerInstance;
		const worksheet = (_this$_univerInstance = this._univerInstanceService.getUnit(unitId)) === null || _this$_univerInstance === void 0 ? void 0 : _this$_univerInstance.getSheetBySheetId(subUnitId);
		const { startRow, startColumn, endColumn } = range;
		const header = [];
		const columnText = prefixText !== null && prefixText !== void 0 ? prefixText : "Column";
		for (let i = startColumn; i <= endColumn; i++) header.push(convertCellDataToString(worksheet === null || worksheet === void 0 ? void 0 : worksheet.getCell(startRow, i)) || getColumnName(i - startColumn + 1, columnText));
		return header;
	}
	/**
	* Add a table to univer.
	* @param {string} unitId The unit id of the table.
	* @param {string} subUnitId The subunit id of the table.
	* @param {string} name The table name, it should be unique in the unit or it will be appended with a number.
	* @param {ITableRange} range The range of the table, it contains the unit id and subunit id.
	* @param {string[]} [header] The header of the table, if not provided, it will be generated based on the range.
	* @param {string} [initId] The initial id of the table, if not provided, a random id will be generated.
	* @param {ITableOptions} [options] Other options of the table.
	* @returns {string} The table id.
	*/
	addTable(unitId, subUnitId, name, range, header, initId, options) {
		const id = initId !== null && initId !== void 0 ? initId : generateRandomId();
		const table = new Table(id, name, range, header || this.getColumnHeader(unitId, subUnitId, range), options);
		table.setSubunitId(subUnitId);
		this._ensureUnit(unitId).set(id, table);
		this._tableAdd$.next({
			unitId,
			subUnitId,
			range,
			tableName: name,
			tableId: id,
			tableStyleId: options === null || options === void 0 ? void 0 : options.tableStyleId
		});
		if (options === null || options === void 0 ? void 0 : options.filters) {
			var _this$_univerInstance2;
			const worksheet = (_this$_univerInstance2 = this._univerInstanceService.getUnit(unitId)) === null || _this$_univerInstance2 === void 0 ? void 0 : _this$_univerInstance2.getSheetBySheetId(subUnitId);
			table.getTableFilters().doFilter(worksheet, range);
			this._tableFilterChanged$.next({
				unitId,
				subUnitId,
				tableId: id
			});
		}
		return id;
	}
	addFilter(unitId, tableId, column, filter) {
		const table = this.getTable(unitId, tableId);
		if (table) {
			table.getTableFilters().setColumnFilter(column, filter);
			const subUnitId = table.getSubunitId();
			this._tableFilterChanged$.next({
				unitId,
				subUnitId,
				tableId
			});
		}
	}
	getFilterRanges(unitId, subUnitId) {
		const unitMap = this._tableMap.get(unitId);
		if (!unitMap) return [];
		const filterRanges = [];
		unitMap.forEach((table) => {
			if (table.getSubunitId() === subUnitId && table.isShowHeader()) filterRanges.push(table.getRange());
		});
		return filterRanges;
	}
	getSheetFilterRangeWithState(unitId, subUnitId) {
		const unitMap = this._tableMap.get(unitId);
		if (!unitMap) return [];
		const filterRanges = [];
		unitMap.forEach((table) => {
			if (table.getSubunitId() === subUnitId && table.isShowHeader()) filterRanges.push({
				tableId: table.getId(),
				range: table.getRange(),
				states: table.getFilterStates(table.getRange())
			});
		});
		return filterRanges;
	}
	getTable(unitId, tableId) {
		const unitMap = this._tableMap.get(unitId);
		if (!unitMap) return;
		return unitMap.get(tableId);
	}
	/**
	* Get the unique table name, in excel, the table name should be unique because it is used as a reference.
	* @param {string} unitId The unit id of the table.
	* @param {string} baseName The base name of the table.
	* @returns {string} The unique table name
	*/
	getUniqueTableName(unitId, baseName) {
		const unitMap = this._tableMap.get(unitId);
		if (!unitMap) return baseName;
		const tableNamesSet = new Set(Array.from(unitMap.values()).map((table) => table.getDisplayName()));
		let newName = baseName;
		let count = 1;
		while (tableNamesSet.has(newName)) {
			newName = `${baseName}-${count}`;
			count++;
		}
		return newName;
	}
	/**
	* Get table by unit id and table id.
	* @param {string} unitId  The unit id of the table.
	* @param {string} tableId The table id.
	* @returns {Table} The table.
	*/
	getTableById(unitId, tableId) {
		return this.getTable(unitId, tableId);
	}
	getTableList(unitId) {
		const unitMap = this._tableMap.get(unitId);
		if (!unitMap) return [];
		return Array.from(unitMap.values()).map((table) => {
			return {
				...table.getTableInfo(),
				unitId
			};
		});
	}
	/**
	* Get the table list by unit id and subunit id.
	* @param {string} unitId The unit id of the table.
	* @param {string} subUnitId The subunit id of the table.
	* @returns {Table[]} The table list.
	*/
	getTablesBySubunitId(unitId, subUnitId) {
		const unitMap = this._tableMap.get(unitId);
		if (!unitMap) return [];
		return Array.from(unitMap.values()).filter((table) => table.getSubunitId() === subUnitId);
	}
	getTablesInfoBySubunitId(unitId, subUnitId) {
		return this.getTablesBySubunitId(unitId, subUnitId).map((table) => {
			return {
				id: table.getId(),
				name: table.getDisplayName(),
				range: table.getRange()
			};
		});
	}
	deleteTable(unitId, tableId) {
		const unitMap = this._tableMap.get(unitId);
		const table = unitMap === null || unitMap === void 0 ? void 0 : unitMap.get(tableId);
		if (table) {
			const tableInfo = table.getTableInfo();
			const tableStyleId = table.getTableStyleId();
			unitMap === null || unitMap === void 0 || unitMap.delete(tableId);
			const { subUnitId, range, name } = tableInfo;
			this._tableDelete$.next({
				unitId,
				subUnitId,
				tableId,
				range,
				tableName: name,
				tableStyleId
			});
		}
	}
	operationTableRowCol(unitId, tableId, config) {
		const table = this.getTableById(unitId, tableId);
		if (!table) return;
		const { operationType, rowColType, index, count, columnsJson } = config;
		const oldRange = table.getRange();
		const newRange = { ...oldRange };
		if (operationType === "insert") {
			if (rowColType === "row") newRange.endRow += count;
			else if (rowColType === "column") {
				newRange.endColumn += count;
				for (let i = 0; i < count; i++) {
					const columnPrefix = this._localeService.t("sheets-table.columnPrefix");
					const column = new TableColumn(generateRandomId(), getColumnName(table.getColumnsCount() + 1 + i, columnPrefix));
					if (columnsJson === null || columnsJson === void 0 ? void 0 : columnsJson[i]) column.fromJSON(columnsJson[i]);
					const columnIndex = index + i - oldRange.startColumn;
					table.insertColumn(columnIndex, column);
				}
			}
		} else if (rowColType === "row") newRange.endRow -= count;
		else if (rowColType === "column") {
			newRange.endColumn -= count;
			for (let i = count - 1; i >= 0; i--) {
				const columnIndex = index + i - oldRange.startColumn;
				table.removeColumn(columnIndex);
			}
		}
		table.setRange(newRange);
		this._tableRangeChanged$.next({
			unitId,
			subUnitId: table.getSubunitId(),
			tableId,
			range: newRange,
			oldRange
		});
	}
	updateTableRange(unitId, tableId, config) {
		const table = this.getTableById(unitId, tableId);
		if (!table) return;
		const oldRange = table.getRange();
		const newRange = config.newRange;
		if (newRange.startColumn < oldRange.startColumn) {
			const diff = oldRange.startColumn - newRange.startColumn;
			const columnPrefix = this._localeService.t("sheets-table.columnPrefix");
			for (let i = 0; i < diff; i++) table.insertColumn(oldRange.startColumn, new TableColumn(generateRandomId(), getColumnName(table.getColumnsCount() + 1, columnPrefix)));
		} else if (newRange.startColumn > oldRange.startColumn) {
			const diff = newRange.startColumn - oldRange.startColumn;
			for (let i = diff - 1; i >= 0; i--) {
				const columnIndex = newRange.startColumn + i - oldRange.startColumn;
				table.removeColumn(columnIndex);
			}
		}
		if (newRange.endColumn < oldRange.endColumn) {
			const diff = oldRange.endColumn - newRange.endColumn;
			for (let i = diff - 1; i >= 0; i--) {
				const columnIndex = newRange.endColumn + i - oldRange.startColumn;
				table.removeColumn(columnIndex);
			}
		} else if (newRange.endColumn > oldRange.endColumn) {
			const diff = newRange.endColumn - oldRange.endColumn;
			const columnPrefix = this._localeService.t("sheets-table.columnPrefix");
			for (let i = 0; i < diff; i++) table.insertColumn(oldRange.endColumn, new TableColumn(generateRandomId(), getColumnName(table.getColumnsCount() + 1, columnPrefix)));
		}
		table.setRange(newRange);
		this._tableRangeChanged$.next({
			unitId,
			subUnitId: table.getSubunitId(),
			tableId,
			range: newRange,
			oldRange
		});
	}
	setTableByConfig(unitId, tableId, config) {
		const unitMap = this._tableMap.get(unitId);
		const table = unitMap === null || unitMap === void 0 ? void 0 : unitMap.get(tableId);
		if (!table) return;
		const subUnitId = table.getSubunitId();
		const { name, updateRange, rowColOperation, theme, options } = config;
		if (name) {
			const oldTableName = table.getDisplayName();
			table.setDisplayName(name);
			this._tableNameChanged$.next({
				unitId,
				subUnitId,
				tableId,
				tableName: name,
				oldTableName
			});
		}
		if (rowColOperation) this.operationTableRowCol(unitId, tableId, rowColOperation);
		if (updateRange) this.updateTableRange(unitId, tableId, updateRange);
		if (theme) {
			var _table$getTableStyleI;
			const oldTheme = (_table$getTableStyleI = table.getTableStyleId()) !== null && _table$getTableStyleI !== void 0 ? _table$getTableStyleI : "default";
			table.setTableStyleId(theme);
			this._tableThemeChanged$.next({
				unitId,
				subUnitId,
				tableId,
				theme,
				oldTheme
			});
		}
		if (options) {
			if (options.showHeader !== void 0) table.setShowHeader(options.showHeader);
		}
	}
	toJSON(unitId) {
		const result = {};
		const unitMap = this._tableMap.get(unitId);
		if (unitMap) unitMap.forEach((table) => {
			const subUnitId = table.getSubunitId();
			if (!result[subUnitId]) {
				const tableFilteredOutRows = /* @__PURE__ */ new Set();
				this.getTablesBySubunitId(unitId, subUnitId).forEach((table) => {
					const tableFilteredRows = table.getTableFilters().getFilterOutRows();
					if (!tableFilteredRows) return;
					for (const row of tableFilteredRows) tableFilteredOutRows.add(row);
				});
				result[subUnitId] = {
					tables: [],
					tableFilteredOutRows: Array.from(tableFilteredOutRows)
				};
			}
			result[subUnitId].tables.push(table.toJSON());
		});
		return result;
	}
	fromJSON(unitId, data) {
		const unitMap = this._ensureUnit(unitId);
		Object.keys(data).forEach((subUnitId) => {
			const target = getSheetCommandTarget(this._univerInstanceService, {
				unitId,
				subUnitId
			});
			if (!target) return;
			const sheet = target.worksheet;
			let tables;
			if (data[subUnitId].tables) tables = data[subUnitId].tables;
			else if (Array.isArray(data[subUnitId])) tables = data[subUnitId];
			if (!tables) return;
			tables.forEach((table) => {
				const header = this.getColumnHeader(unitId, subUnitId, table.range);
				const tableInstance = new Table(table.id, table.name, table.range, header, table.options);
				tableInstance.setTableMeta(table.meta);
				if (table.columns.length) tableInstance.setColumns(table.columns);
				if (table.filters) {
					const tableFilter = tableInstance.getTableFilters();
					tableFilter.fromJSON(table.filters);
					tableFilter.doFilter(sheet, tableInstance.getTableFilterRange());
				}
				tableInstance.setSubunitId(subUnitId);
				unitMap.set(table.id, tableInstance);
				this._tableAdd$.next({
					unitId,
					subUnitId,
					range: table.range,
					tableName: table.name,
					tableId: table.id
				});
			});
		});
		this._tableInitStatus.next(true);
	}
	deleteUnitId(unitId) {
		const unitMap = this._tableMap.get(unitId);
		if (unitMap) unitMap.forEach((table) => table.dispose());
		this._tableMap.delete(unitId);
	}
	dispose() {
		super.dispose();
		this._tableAdd$.complete();
		this._tableDelete$.complete();
		this._tableNameChanged$.complete();
		this._tableRangeChanged$.complete();
		this._tableThemeChanged$.complete();
		this._tableFilterChanged$.complete();
		this._tableInitStatus.complete();
		this._tableMap.forEach((unitMap) => {
			unitMap.forEach((table) => table.dispose());
			unitMap.clear();
		});
		this._tableMap.clear();
	}
};
TableManager = __decorate([__decorateParam(0, IUniverInstanceService), __decorateParam(1, Inject(LocaleService))], TableManager);

//#endregion
//#region src/services/table.service.ts
let SheetTableService = class SheetTableService extends Disposable {
	constructor(_tableManager) {
		super();
		this._tableManager = _tableManager;
	}
	getTableInfo(unitId, tableId) {
		const table = this._tableManager.getTable(unitId, tableId);
		if (!table) return;
		return {
			unitId,
			...table.getTableInfo()
		};
	}
	getTableList(unitId) {
		return this._tableManager.getTableList(unitId);
	}
	addTable(unitId, subUnitId, tableName, rangeInfo, tableHeader, tableId, options) {
		return this._tableManager.addTable(unitId, subUnitId, tableName, rangeInfo, tableHeader, tableId, options);
	}
	deleteTable(unitId, subUnitId, tableId) {
		this._tableManager.deleteTable(unitId, tableId);
	}
	getTableMeta(unitId, tableId) {
		var _this$_tableManager$g;
		return (_this$_tableManager$g = this._tableManager.getTable(unitId, tableId)) === null || _this$_tableManager$g === void 0 ? void 0 : _this$_tableManager$g.getTableMeta();
	}
	setTableMeta(unitId, tableId, meta) {
		var _this$_tableManager$g2;
		(_this$_tableManager$g2 = this._tableManager.getTable(unitId, tableId)) === null || _this$_tableManager$g2 === void 0 || _this$_tableManager$g2.setTableMeta(meta);
	}
	getTableColumnMeta(unitId, tableId, index) {
		var _this$_tableManager$g3;
		return (_this$_tableManager$g3 = this._tableManager.getTable(unitId, tableId)) === null || _this$_tableManager$g3 === void 0 || (_this$_tableManager$g3 = _this$_tableManager$g3.getTableColumnByIndex(index)) === null || _this$_tableManager$g3 === void 0 ? void 0 : _this$_tableManager$g3.getMeta();
	}
	selTableColumnMeta(unitId, tableId, index, meta) {
		var _this$_tableManager$g4;
		(_this$_tableManager$g4 = this._tableManager.getTable(unitId, tableId)) === null || _this$_tableManager$g4 === void 0 || (_this$_tableManager$g4 = _this$_tableManager$g4.getTableColumnByIndex(index)) === null || _this$_tableManager$g4 === void 0 || _this$_tableManager$g4.setMeta(meta);
	}
	addFilter(unitId, tableId, column, filter) {
		this._tableManager.addFilter(unitId, tableId, column, filter);
	}
	getCellValueWithConditionType(sheet, row, col, conditionType = "string") {
		return getCellValueWithConditionType(sheet, row, col, conditionType);
	}
};
SheetTableService = __decorate([__decorateParam(0, Inject(TableManager))], SheetTableService);

//#endregion
//#region src/commands/mutations/add-sheet-table.mutation.ts
const AddSheetTableMutation = {
	id: "sheet.mutation.add-table",
	type: CommandType.MUTATION,
	handler: (accessor, params) => {
		const { tableId, unitId, subUnitId, name, range, header, options } = params;
		accessor.get(SheetTableService).addTable(unitId, subUnitId, name, range, header, tableId, options);
		return true;
	}
};

//#endregion
//#region src/commands/mutations/delete-sheet-table.mutation.ts
const DeleteSheetTableMutation = {
	id: "sheet.mutation.delete-table",
	type: CommandType.MUTATION,
	handler: (accessor, params) => {
		const { unitId, subUnitId, tableId } = params;
		accessor.get(SheetTableService).deleteTable(unitId, subUnitId, tableId);
		return true;
	}
};

//#endregion
//#region src/commands/commands/add-sheet-table.command.ts
const AddSheetTableCommand = {
	id: "sheet.command.add-table",
	type: CommandType.COMMAND,
	handler: (accessor, params) => {
		var _params$id;
		if (!params) return false;
		const undoRedoService = accessor.get(IUndoRedoService);
		const commandService = accessor.get(ICommandService);
		const localeService = accessor.get(LocaleService);
		const tableManager = accessor.get(TableManager);
		const tableId = (_params$id = params.id) !== null && _params$id !== void 0 ? _params$id : generateRandomId();
		const existingNamesSet = getExistingNamesSet(params.unitId, {
			univerInstanceService: accessor.get(IUniverInstanceService),
			tableManager,
			definedNamesService: accessor.get(IDefinedNamesService)
		});
		let tableName = params.name;
		if (!tableName || !customNameCharacterCheck(tableName.toLowerCase(), existingNamesSet)) {
			const prefix = localeService.t("sheets-table.tablePrefix");
			let index = tableManager.getTableList(params.unitId).length + 1;
			for (const name of existingNamesSet) if (name.startsWith(prefix.toLowerCase())) {
				const n = Number(name.slice(prefix.length));
				if (Number.isInteger(n) && n >= index) index = n + 1;
			}
			tableName = `${prefix}${index}`;
		}
		const redos = [];
		const undos = [];
		const { unitId, subUnitId, range } = params;
		const header = tableManager.getColumnHeader(unitId, subUnitId, range, localeService.t("sheets-table.columnPrefix"));
		redos.push({
			id: AddSheetTableMutation.id,
			params: {
				...params,
				tableId,
				name: tableName,
				header
			}
		});
		undos.push({
			id: DeleteSheetTableMutation.id,
			params: {
				tableId,
				unitId: params.unitId
			}
		});
		if (sequenceExecute(redos, commandService).result) {
			undoRedoService.pushUndoRedo({
				unitID: params.unitId,
				undoMutations: undos,
				redoMutations: redos
			});
			return true;
		}
		return false;
	}
};

//#endregion
//#region src/commands/mutations/set-sheet-table.mutation.ts
const SetSheetTableMutation = {
	id: "sheet.mutation.set-sheet-table",
	type: CommandType.MUTATION,
	handler: (accessor, params) => {
		if (!params) return false;
		const { unitId, tableId, config } = params;
		accessor.get(TableManager).setTableByConfig(unitId, tableId, config);
		return true;
	}
};

//#endregion
//#region src/commands/commands/add-table-theme.command.ts
const AddTableThemeCommand = {
	id: "sheet.command.add-table-theme",
	type: CommandType.COMMAND,
	handler: (accessor, params) => {
		if (!params) return false;
		const tableManager = accessor.get(TableManager);
		const { unitId, tableId, themeStyle } = params;
		const redos = [];
		const undos = [];
		const table = tableManager.getTableById(unitId, tableId);
		if (!table) return false;
		const subUnitId = table.getSubunitId();
		redos.push({
			id: AddRangeThemeMutation.id,
			params: {
				unitId,
				subUnitId,
				styleJSON: themeStyle.toJson()
			}
		});
		redos.push({
			id: SetSheetTableMutation.id,
			params: {
				unitId,
				subUnitId,
				tableId,
				config: { theme: themeStyle.getName() }
			}
		});
		undos.push({
			id: SetSheetTableMutation.id,
			params: {
				unitId,
				subUnitId,
				tableId,
				config: { themeStyle: table.getTableStyleId() }
			}
		});
		undos.push({
			id: RemoveRangeThemeMutation.id,
			params: {
				unitId,
				subUnitId,
				styleName: themeStyle.getName()
			}
		});
		if (sequenceExecute(redos, accessor.get(ICommandService)).result) {
			accessor.get(IUndoRedoService).pushUndoRedo({
				unitID: unitId,
				undoMutations: undos,
				redoMutations: redos
			});
			return true;
		}
		return false;
	}
};

//#endregion
//#region src/commands/commands/delete-sheet-table.command.ts
const DeleteSheetTableCommand = {
	id: "sheet.command.delete-table",
	type: CommandType.COMMAND,
	handler: (accessor, params) => {
		var _interceptorCommands$, _interceptorCommands$2;
		if (!params) return false;
		const undoRedoService = accessor.get(IUndoRedoService);
		const commandService = accessor.get(ICommandService);
		const sheetTableManager = accessor.get(TableManager);
		const logService = accessor.get(ILogService);
		const redos = [];
		const undos = [];
		const tableInstance = sheetTableManager.getTable(params.unitId, params.tableId);
		const tableConfig = tableInstance === null || tableInstance === void 0 ? void 0 : tableInstance.toJSON();
		if (!tableConfig) {
			logService.error("[TableManager]: Table not found");
			return false;
		}
		const interceptorCommands = accessor.get(SheetInterceptorService).onCommandExecute({
			id: DeleteSheetTableCommand.id,
			params: {
				...params,
				tableName: tableConfig.name
			}
		});
		redos.push(...(_interceptorCommands$ = interceptorCommands.preRedos) !== null && _interceptorCommands$ !== void 0 ? _interceptorCommands$ : []);
		redos.push({
			id: DeleteSheetTableMutation.id,
			params: { ...params }
		});
		redos.push(...interceptorCommands.redos);
		undos.push(...(_interceptorCommands$2 = interceptorCommands.preUndos) !== null && _interceptorCommands$2 !== void 0 ? _interceptorCommands$2 : []);
		undos.push({
			id: AddSheetTableMutation.id,
			params: {
				unitId: params.unitId,
				subUnitId: params.subUnitId,
				tableId: params.tableId,
				name: tableConfig.name,
				range: tableConfig.range,
				options: tableConfig.options
			}
		});
		undos.push(...interceptorCommands.undos);
		if (sequenceExecute(redos, commandService).result) {
			undoRedoService.pushUndoRedo({
				unitID: params.unitId,
				undoMutations: undos,
				redoMutations: redos
			});
			return true;
		}
		return false;
	}
};

//#endregion
//#region src/commands/commands/remove-table-theme.command.ts
const RemoveTableThemeCommand = {
	id: "sheet.command.remove-table-theme",
	type: CommandType.COMMAND,
	handler: (accessor, params) => {
		if (!params) return false;
		const { unitId, tableId, themeName } = params;
		const tableManager = accessor.get(TableManager);
		const rangeThemeModel = accessor.get(SheetRangeThemeModel);
		const table = tableManager.getTableById(unitId, tableId);
		if (!table) return false;
		const subUnitId = table.getSubunitId();
		const redos = [];
		const undos = [];
		const defaultRangeThemes = rangeThemeModel.getRegisteredRangeThemes().filter((item) => item === null || item === void 0 ? void 0 : item.startsWith("table-default"));
		let shouldBeSelectedTheme = rangeThemeModel.getRegisteredRangeThemes().filter((item) => item === null || item === void 0 ? void 0 : item.startsWith(SHEET_TABLE_CUSTOM_THEME_PREFIX)).find((item) => item !== themeName);
		if (!shouldBeSelectedTheme) shouldBeSelectedTheme = defaultRangeThemes[0];
		redos.push({
			id: SetSheetTableMutation.id,
			params: {
				unitId,
				subUnitId,
				tableId,
				config: { theme: shouldBeSelectedTheme }
			}
		});
		redos.push({
			id: RemoveRangeThemeMutation.id,
			params: {
				unitId,
				subUnitId,
				styleName: themeName
			}
		});
		const themeStyle = rangeThemeModel.getDefaultRangeThemeStyle(themeName);
		if (themeStyle) {
			undos.push({
				id: AddRangeThemeMutation.id,
				params: {
					unitId,
					subUnitId,
					styleJSON: themeStyle.toJson()
				}
			});
			undos.push({
				id: SetSheetTableMutation.id,
				params: {
					unitId,
					subUnitId,
					tableId,
					config: { theme: themeName }
				}
			});
		}
		if (sequenceExecute(redos, accessor.get(ICommandService)).result) {
			accessor.get(IUndoRedoService).pushUndoRedo({
				unitID: unitId,
				redoMutations: redos,
				undoMutations: undos
			});
			return true;
		}
		return false;
	}
};

//#endregion
//#region src/util/table-name.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
function validateSheetTableName(name, existingNamesSet) {
	const trimmedName = name.trim();
	if (!trimmedName) return {
		valid: false,
		reason: "empty"
	};
	const normalizedExistingNames = new Set(Array.from(existingNamesSet, (item) => item.toLowerCase()));
	if (!customNameCharacterCheck(trimmedName.toLowerCase(), normalizedExistingNames)) return {
		valid: false,
		reason: "invalid"
	};
	return { valid: true };
}

//#endregion
//#region src/commands/commands/set-sheet-table.command.ts
const SetSheetTableCommand = {
	id: "sheet.command.set-table-config",
	type: CommandType.COMMAND,
	handler: (accessor, params) => {
		var _interceptorCommands$, _interceptorCommands$2;
		if (!params) return false;
		const { unitId, tableId, name, updateRange, rowColOperation, theme } = params;
		const tableManager = accessor.get(TableManager);
		const table = tableManager.getTableById(unitId, tableId);
		if (!table) return false;
		const oldTableConfig = {};
		const newTableConfig = {};
		const localeService = accessor.get(LocaleService);
		const existingNamesSet = getExistingNamesSet(unitId, {
			univerInstanceService: accessor.get(IUniverInstanceService),
			tableManager,
			definedNamesService: accessor.get(IDefinedNamesService)
		});
		if (name) {
			if (!validateSheetTableName(name, existingNamesSet).valid) {
				accessor.get(ILogService).warn(localeService.t("sheets-table.tableNameError"));
				return false;
			}
			oldTableConfig.name = table.getDisplayName();
			newTableConfig.name = name;
		}
		if (rowColOperation) {
			oldTableConfig.rowColOperation = {
				operationType: rowColOperation.operationType === "insert" ? "delete" : "insert",
				rowColType: rowColOperation.rowColType,
				index: rowColOperation.index,
				count: rowColOperation.count
			};
			newTableConfig.rowColOperation = rowColOperation;
		}
		if (updateRange) {
			oldTableConfig.updateRange = { newRange: table.getRange() };
			newTableConfig.updateRange = updateRange;
		}
		if (theme) {
			oldTableConfig.theme = table.getTableStyleId();
			newTableConfig.theme = theme;
		}
		const redoParams = {
			unitId,
			subUnitId: table.getSubunitId(),
			tableId,
			config: newTableConfig
		};
		const undoParams = {
			unitId,
			subUnitId: table.getSubunitId(),
			tableId,
			config: oldTableConfig
		};
		const interceptorCommands = accessor.get(SheetInterceptorService).onCommandExecute({
			id: SetSheetTableCommand.id,
			params: {
				...params,
				oldTableName: oldTableConfig.name
			}
		});
		const redos = [
			...(_interceptorCommands$ = interceptorCommands.preRedos) !== null && _interceptorCommands$ !== void 0 ? _interceptorCommands$ : [],
			{
				id: SetSheetTableMutation.id,
				params: redoParams
			},
			...interceptorCommands.redos
		];
		const undos = [
			...(_interceptorCommands$2 = interceptorCommands.preUndos) !== null && _interceptorCommands$2 !== void 0 ? _interceptorCommands$2 : [],
			{
				id: SetSheetTableMutation.id,
				params: undoParams
			},
			...interceptorCommands.undos
		];
		const commandService = accessor.get(ICommandService);
		redos.forEach((mutation) => {
			commandService.executeCommand(mutation.id, mutation.params);
		});
		accessor.get(IUndoRedoService).pushUndoRedo({
			unitID: unitId,
			undoMutations: undos,
			redoMutations: redos
		});
		return true;
	}
};

//#endregion
//#region src/commands/mutations/set-table-filter.mutation.ts
const SetSheetTableFilterMutation = {
	id: "sheet.mutation.set-table-filter",
	type: CommandType.MUTATION,
	handler: (accessor, params) => {
		const { tableId, unitId, column, tableFilter } = params;
		accessor.get(TableManager).addFilter(unitId, tableId, column, tableFilter);
		return true;
	}
};

//#endregion
//#region src/commands/commands/set-table-filter.command.ts
const SetSheetTableFilterCommand = {
	id: "sheet.command.set-table-filter",
	type: CommandType.COMMAND,
	handler: (accessor, params) => {
		if (!params) return false;
		const undoRedoService = accessor.get(IUndoRedoService);
		const commandService = accessor.get(ICommandService);
		const tableId = params.tableId || generateRandomId();
		const redos = [];
		const undos = [];
		redos.push({
			id: SetSheetTableFilterMutation.id,
			params: {
				...params,
				tableId
			}
		});
		undos.push({
			id: SetSheetTableFilterMutation.id,
			params: {
				...params,
				tableId,
				tableFilter: void 0
			}
		});
		if (sequenceExecute(redos, commandService).result) {
			undoRedoService.pushUndoRedo({
				unitID: params.unitId,
				undoMutations: undos,
				redoMutations: redos
			});
			return true;
		}
		return false;
	}
};

//#endregion
//#region src/controllers/sheets-table.controller.ts
let SheetsTableController = class SheetsTableController extends Disposable {
	constructor(_univerInstanceService, _sheetInterceptorService, _tableManager, _resourceManagerService) {
		super();
		this._univerInstanceService = _univerInstanceService;
		this._sheetInterceptorService = _sheetInterceptorService;
		this._tableManager = _tableManager;
		this._resourceManagerService = _resourceManagerService;
		_defineProperty(this, "_tableRangeRTree", /* @__PURE__ */ new Map());
		this._initSnapshot();
		this._initSheetChange();
		this.registerTableChangeEvent();
		this.registerTableHeaderInterceptor();
	}
	getContainerTableWithRange(unitId, subUnitId, range) {
		const rTree = this._ensureTableRangeRTree(unitId);
		const wrapperTableId = Array.from(rTree.bulkSearch([{
			unitId,
			sheetId: subUnitId,
			range
		}])).find((id) => {
			const table = this._tableManager.getTable(unitId, String(id));
			if (table) return Rectangle.contains(table.getRange(), range);
			return false;
		});
		if (wrapperTableId) return this._tableManager.getTable(unitId, String(wrapperTableId));
	}
	_ensureTableRangeRTree(unitId) {
		if (!this._tableRangeRTree.has(unitId)) this._tableRangeRTree.set(unitId, new RTree());
		return this._tableRangeRTree.get(unitId);
	}
	registerTableChangeEvent() {
		this.disposeWithMe(this._tableManager.tableAdd$.subscribe((event) => {
			const { range, tableId, unitId, subUnitId } = event;
			this._ensureTableRangeRTree(unitId).insert({
				unitId,
				sheetId: subUnitId,
				id: tableId,
				range: { ...range }
			});
		}));
		this.disposeWithMe(this._tableManager.tableRangeChanged$.subscribe((event) => {
			const { range, tableId, unitId, subUnitId, oldRange } = event;
			const rTree = this._ensureTableRangeRTree(unitId);
			rTree.remove({
				unitId,
				sheetId: subUnitId,
				id: tableId,
				range: { ...oldRange }
			});
			rTree.insert({
				unitId,
				sheetId: subUnitId,
				id: tableId,
				range: { ...range }
			});
		}));
		this.disposeWithMe(this._tableManager.tableDelete$.subscribe((event) => {
			const { tableId, unitId, subUnitId, range } = event;
			this._ensureTableRangeRTree(unitId).remove({
				unitId,
				sheetId: subUnitId,
				id: tableId,
				range: { ...range }
			});
		}));
	}
	registerTableHeaderInterceptor() {
		this.disposeWithMe(this._sheetInterceptorService.intercept(INTERCEPTOR_POINT.CELL_CONTENT, {
			effect: InterceptorEffectEnum.Value,
			handler: (cell, context, next) => {
				const { row, col, unitId, subUnitId } = context;
				const rTree = this._ensureTableRangeRTree(unitId);
				if ((cell === null || cell === void 0 ? void 0 : cell.v) === void 0 && rTree) {
					const ids = Array.from(rTree.bulkSearch([{
						unitId,
						sheetId: subUnitId,
						range: {
							startColumn: col,
							endColumn: col,
							startRow: row,
							endRow: row
						}
					}]));
					if (ids.length > 0) {
						const table = this._tableManager.getTable(unitId, ids[0]);
						if (table) {
							const tableRange = table.getRange();
							const index = col - tableRange.startColumn;
							if (tableRange.startRow === row) {
								const columnName = table.getColumnNameByIndex(index);
								if (!cell || cell === context.rawData) cell = { ...context.rawData };
								cell.v = columnName;
								return next(cell);
							}
						}
					}
				}
				return next(cell);
			}
		}));
	}
	_toJson(unitId) {
		return this._tableManager.toJSON(unitId);
	}
	_fromJSON(unitId, resources) {
		return this._tableManager.fromJSON(unitId, resources);
	}
	_deleteUnitId(unitId) {
		this._tableManager.deleteUnitId(unitId);
	}
	_initSnapshot() {
		this.disposeWithMe(this._resourceManagerService.registerPluginResource({
			toJson: (unitId) => {
				return JSON.stringify(this._toJson(unitId));
			},
			parseJson: (json) => {
				if (!json) return {};
				try {
					return JSON.parse(json);
				} catch (error) {
					return {};
				}
			},
			businesses: [UniverInstanceType.UNIVER_SHEET],
			pluginName: PLUGIN_NAME,
			onLoad: (unitId, resources) => {
				this._fromJSON(unitId, resources);
			},
			onUnLoad: (unitId) => {
				this._deleteUnitId(unitId);
			}
		}));
	}
	_initSheetChange() {
		this.disposeWithMe(this._sheetInterceptorService.interceptCommand({ getMutations: (commandInfo) => {
			if (commandInfo.id === RemoveSheetCommand.id) {
				var _getActiveSheet;
				const params = commandInfo.params;
				const unitId = params.unitId || this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_SHEET).getUnitId();
				const subUnitId = params.subUnitId || ((_getActiveSheet = this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_SHEET).getActiveSheet()) === null || _getActiveSheet === void 0 ? void 0 : _getActiveSheet.getSheetId());
				if (!unitId || !subUnitId) return {
					redos: [],
					undos: []
				};
				const tables = this._tableManager.getTablesBySubunitId(unitId, subUnitId);
				if (tables.length === 0) return {
					redos: [],
					undos: []
				};
				const redos = [];
				const undos = [];
				tables.forEach((table) => {
					const tableJson = table.toJSON();
					redos.push({
						id: DeleteSheetTableMutation.id,
						params: {
							unitId,
							subUnitId,
							tableId: tableJson.id
						}
					});
					undos.push({
						id: AddSheetTableMutation.id,
						params: {
							unitId,
							subUnitId,
							name: tableJson.name,
							range: tableJson.range,
							tableId: tableJson.id,
							options: {
								...tableJson.options,
								columns: tableJson.columns,
								filters: tableJson.filters.tableColumnFilterList
							}
						}
					});
				});
				return {
					redos,
					undos
				};
			} else if (commandInfo.id === CopySheetCommand.id) {
				const { unitId, subUnitId, targetSubUnitId } = commandInfo.params;
				if (!unitId || !subUnitId || !targetSubUnitId) return {
					redos: [],
					undos: []
				};
				const tables = this._tableManager.getTablesBySubunitId(unitId, subUnitId);
				if (tables.length === 0) return {
					redos: [],
					undos: []
				};
				const redos = [];
				const undos = [];
				tables.forEach((table) => {
					const tableJson = table.toJSON();
					const tableId = generateRandomId();
					redos.push({
						id: AddSheetTableMutation.id,
						params: {
							unitId,
							subUnitId: targetSubUnitId,
							name: tableJson.name,
							range: {
								...tableJson.range,
								sheetId: targetSubUnitId
							},
							tableId,
							options: {
								...tableJson.options,
								columns: tableJson.columns,
								filters: tableJson.filters.tableColumnFilterList
							}
						}
					});
					undos.push({
						id: DeleteSheetTableMutation.id,
						params: {
							unitId,
							subUnitId: targetSubUnitId,
							tableId
						}
					});
				});
				return {
					redos,
					undos
				};
			}
			return {
				redos: [],
				undos: []
			};
		} }));
	}
	dispose() {
		super.dispose();
		this._tableRangeRTree.clear();
	}
};
SheetsTableController = __decorate([
	__decorateParam(0, Inject(IUniverInstanceService)),
	__decorateParam(1, Inject(SheetInterceptorService)),
	__decorateParam(2, Inject(TableManager)),
	__decorateParam(3, Inject(IResourceManagerService))
], SheetsTableController);

//#endregion
//#region src/commands/commands/sheet-table-row-col.command.ts
function executeTableMutationSequence(accessor, unitId, redos, undos) {
	if (sequenceExecute(redos, accessor.get(ICommandService)).result) {
		accessor.get(IUndoRedoService).pushUndoRedo({
			unitID: unitId,
			undoMutations: undos,
			redoMutations: redos
		});
		return true;
	}
	return false;
}
const SheetTableInsertRowCommand = {
	id: "sheet.command.table-insert-row",
	type: CommandType.COMMAND,
	handler: (accessor) => {
		const target = getSheetCommandTarget(accessor.get(IUniverInstanceService));
		if (!target) return false;
		const { workbook, worksheet, unitId, subUnitId } = target;
		const selections = accessor.get(SheetsSelectionsService).getCurrentSelections();
		if (!selections.length || selections.length > 1) return false;
		accessor.get(TableManager);
		const range = selections[0].range;
		const table = accessor.get(SheetsTableController).getContainerTableWithRange(unitId, subUnitId, range);
		if (!table) return false;
		const insertRowCount = range.endRow - range.startRow + 1;
		const worksheetLastRowIndex = worksheet.getRowCount() - 1;
		const rowContentIndex = worksheet.getCellMatrix().getDataRange().endRow;
		const redos = [];
		const undos = [];
		if (worksheetLastRowIndex - rowContentIndex < insertRowCount) {
			redos.push({
				id: InsertRowMutation.id,
				params: {
					unitId,
					subUnitId,
					range: { ...range }
				}
			});
			redos.push({
				id: SetSheetTableMutation.id,
				params: {
					unitId,
					subUnitId,
					tableId: table.getId(),
					config: { updateRange: { newRange: {
						...table.getRange(),
						endRow: table.getRange().endRow + insertRowCount
					} } }
				}
			});
			undos.push({
				id: SetSheetTableMutation.id,
				params: {
					unitId,
					subUnitId,
					tableId: table.getId(),
					config: { updateRange: { newRange: table.getRange() } }
				}
			});
			undos.push({
				id: RemoveRowMutation.id,
				params: {
					unitId,
					subUnitId,
					range: { ...range }
				}
			});
		} else {
			const oldRange = { ...table.getRange() };
			redos.push({
				id: SetSheetTableMutation.id,
				params: {
					unitId,
					subUnitId,
					tableId: table.getId(),
					config: { updateRange: { newRange: {
						...oldRange,
						endRow: oldRange.endRow + insertRowCount
					} } }
				}
			});
			undos.push({
				id: SetSheetTableMutation.id,
				params: {
					unitId,
					subUnitId,
					tableId: table.getId(),
					config: { updateRange: { newRange: { ...oldRange } } }
				}
			});
			const moveRangeMutations = getMoveRangeUndoRedoMutations(accessor, {
				unitId,
				subUnitId,
				range: {
					startRow: range.startRow,
					endRow: rowContentIndex,
					startColumn: oldRange.startColumn,
					endColumn: oldRange.endColumn
				}
			}, {
				unitId,
				subUnitId,
				range: {
					startRow: range.startRow + insertRowCount,
					endRow: rowContentIndex + insertRowCount,
					startColumn: oldRange.startColumn,
					endColumn: oldRange.endColumn
				}
			});
			if (moveRangeMutations) {
				redos.push(...moveRangeMutations.redos);
				undos.push(...moveRangeMutations.undos);
			}
		}
		if (sequenceExecute(redos, accessor.get(ICommandService)).result) {
			accessor.get(IUndoRedoService).pushUndoRedo({
				unitID: unitId,
				undoMutations: undos,
				redoMutations: redos
			});
			return true;
		}
		return false;
	}
};
const SheetTableInsertRowAtCommand = {
	id: "sheet.command.table-insert-row-at",
	type: CommandType.COMMAND,
	handler: (accessor, params) => {
		if (!params) return false;
		const { unitId, subUnitId, tableId, index, count = 1 } = params;
		if (count <= 0) return false;
		const target = getSheetCommandTarget(accessor.get(IUniverInstanceService), {
			unitId,
			subUnitId
		});
		if (!target) return false;
		const table = accessor.get(TableManager).getTableById(unitId, tableId);
		if (!table || table.getSubunitId() !== subUnitId) return false;
		const oldRange = table.getRange();
		if (index <= oldRange.startRow || index > oldRange.endRow + 1) return false;
		const redos = [{
			id: SetSheetTableMutation.id,
			params: {
				unitId,
				subUnitId,
				tableId,
				config: { updateRange: { newRange: {
					...oldRange,
					endRow: oldRange.endRow + count
				} } }
			}
		}];
		const undos = [{
			id: SetSheetTableMutation.id,
			params: {
				unitId,
				subUnitId,
				tableId,
				config: { updateRange: { newRange: { ...oldRange } } }
			}
		}];
		const rowContentIndex = target.worksheet.getCellMatrix().getDataRange().endRow;
		const moveRangeMutations = getMoveRangeUndoRedoMutations(accessor, {
			unitId,
			subUnitId,
			range: {
				startRow: index,
				endRow: rowContentIndex,
				startColumn: oldRange.startColumn,
				endColumn: oldRange.endColumn
			}
		}, {
			unitId,
			subUnitId,
			range: {
				startRow: index + count,
				endRow: rowContentIndex + count,
				startColumn: oldRange.startColumn,
				endColumn: oldRange.endColumn
			}
		});
		if (moveRangeMutations) {
			redos.push(...moveRangeMutations.redos);
			undos.push(...moveRangeMutations.undos);
		}
		return executeTableMutationSequence(accessor, unitId, redos, undos);
	}
};
const SheetTableInsertColCommand = {
	id: "sheet.command.table-insert-col",
	type: CommandType.COMMAND,
	handler: (accessor) => {
		const target = getSheetCommandTarget(accessor.get(IUniverInstanceService));
		if (!target) return false;
		const { worksheet, unitId, subUnitId } = target;
		const selections = accessor.get(SheetsSelectionsService).getCurrentSelections();
		if (!selections.length || selections.length > 1) return false;
		const range = selections[0].range;
		const table = accessor.get(SheetsTableController).getContainerTableWithRange(unitId, subUnitId, range);
		if (!table) return false;
		const insertColCount = range.endColumn - range.startColumn + 1;
		const worksheetLastColIndex = worksheet.getColumnCount() - 1;
		const colContentIndex = worksheet.getCellMatrix().getDataRange().endColumn;
		const redos = [];
		const undos = [];
		if (worksheetLastColIndex - colContentIndex < insertColCount) {
			redos.push({
				id: InsertColMutation.id,
				params: {
					unitId,
					subUnitId,
					range: { ...range }
				}
			});
			redos.push({
				id: SetSheetTableMutation.id,
				params: {
					unitId,
					subUnitId,
					tableId: table.getId(),
					config: { rowColOperation: {
						operationType: "insert",
						rowColType: "column",
						index: range.startColumn,
						count: insertColCount
					} }
				}
			});
			undos.push({
				id: SetSheetTableMutation.id,
				params: {
					unitId,
					subUnitId,
					tableId: table.getId(),
					config: { rowColOperation: {
						operationType: "delete",
						rowColType: "column",
						index: range.startColumn,
						count: insertColCount
					} }
				}
			});
			undos.push({
				id: RemoveColMutation.id,
				params: {
					unitId,
					subUnitId,
					range: { ...range }
				}
			});
		} else {
			const oldRange = table.getRange();
			redos.push({
				id: SetSheetTableMutation.id,
				params: {
					unitId,
					subUnitId,
					tableId: table.getId(),
					config: { rowColOperation: {
						operationType: "insert",
						rowColType: "column",
						index: range.startColumn,
						count: insertColCount
					} }
				}
			});
			undos.push({
				id: SetSheetTableMutation.id,
				params: {
					unitId,
					subUnitId,
					tableId: table.getId(),
					config: { rowColOperation: {
						operationType: "delete",
						rowColType: "column",
						index: range.startColumn,
						count: insertColCount
					} }
				}
			});
			const moveRangeMutations = getMoveRangeUndoRedoMutations(accessor, {
				unitId,
				subUnitId,
				range: {
					startRow: oldRange.startRow,
					endRow: oldRange.endRow,
					startColumn: range.startColumn,
					endColumn: colContentIndex
				}
			}, {
				unitId,
				subUnitId,
				range: {
					startRow: oldRange.startRow,
					endRow: oldRange.endRow,
					startColumn: range.startColumn + insertColCount,
					endColumn: colContentIndex + insertColCount
				}
			});
			if (moveRangeMutations) {
				redos.push(...moveRangeMutations.redos);
				undos.push(...moveRangeMutations.undos);
			}
		}
		if (sequenceExecute(redos, accessor.get(ICommandService)).result) {
			accessor.get(IUndoRedoService).pushUndoRedo({
				unitID: unitId,
				undoMutations: undos,
				redoMutations: redos
			});
			return true;
		}
		return false;
	}
};
const SheetTableInsertColumnAtCommand = {
	id: "sheet.command.table-insert-column-at",
	type: CommandType.COMMAND,
	handler: (accessor, params) => {
		if (!params) return false;
		const { unitId, subUnitId, tableId, index, count = 1 } = params;
		if (count <= 0) return false;
		const target = getSheetCommandTarget(accessor.get(IUniverInstanceService), {
			unitId,
			subUnitId
		});
		if (!target) return false;
		const table = accessor.get(TableManager).getTableById(unitId, tableId);
		if (!table || table.getSubunitId() !== subUnitId) return false;
		const oldRange = table.getRange();
		if (index < oldRange.startColumn || index > oldRange.endColumn + 1) return false;
		const redos = [{
			id: SetSheetTableMutation.id,
			params: {
				unitId,
				subUnitId,
				tableId,
				config: { rowColOperation: {
					operationType: "insert",
					rowColType: "column",
					index,
					count
				} }
			}
		}];
		const undos = [{
			id: SetSheetTableMutation.id,
			params: {
				unitId,
				subUnitId,
				tableId,
				config: { rowColOperation: {
					operationType: "delete",
					rowColType: "column",
					index,
					count
				} }
			}
		}];
		const colContentIndex = target.worksheet.getCellMatrix().getDataRange().endColumn;
		if (index <= colContentIndex) {
			const moveRangeMutations = getMoveRangeUndoRedoMutations(accessor, {
				unitId,
				subUnitId,
				range: {
					startRow: oldRange.startRow,
					endRow: oldRange.endRow,
					startColumn: index,
					endColumn: colContentIndex
				}
			}, {
				unitId,
				subUnitId,
				range: {
					startRow: oldRange.startRow,
					endRow: oldRange.endRow,
					startColumn: index + count,
					endColumn: colContentIndex + count
				}
			});
			if (moveRangeMutations) {
				redos.push(...moveRangeMutations.redos);
				undos.push(...moveRangeMutations.undos);
			}
		}
		return executeTableMutationSequence(accessor, unitId, redos, undos);
	}
};
const SheetTableRemoveColumnAtCommand = {
	id: "sheet.command.table-remove-column-at",
	type: CommandType.COMMAND,
	handler: (accessor, params) => {
		var _interceptorCommands$, _interceptorCommands$2;
		if (!params) return false;
		const { unitId, subUnitId, tableId, index, count = 1 } = params;
		if (count <= 0) return false;
		const target = getSheetCommandTarget(accessor.get(IUniverInstanceService), {
			unitId,
			subUnitId
		});
		if (!target) return false;
		const table = accessor.get(TableManager).getTableById(unitId, tableId);
		if (!table || table.getSubunitId() !== subUnitId) return false;
		const oldRange = table.getRange();
		if (index < oldRange.startColumn || index + count - 1 > oldRange.endColumn || count >= oldRange.endColumn - oldRange.startColumn + 1) return false;
		const tableInfo = table.getTableInfo();
		const columns = [];
		const gap = index - oldRange.startColumn;
		for (let i = 0; i < count; i++) {
			const column = tableInfo.columns[gap + i];
			if (column) columns.push(column);
		}
		const interceptorCommands = accessor.get(SheetInterceptorService).onCommandExecute({
			id: SheetTableRemoveColumnAtCommand.id,
			params: {
				...params,
				tableName: tableInfo.name,
				removedColumnNames: columns.map((column) => column.displayName)
			}
		});
		const redos = [
			...(_interceptorCommands$ = interceptorCommands.preRedos) !== null && _interceptorCommands$ !== void 0 ? _interceptorCommands$ : [],
			{
				id: SetSheetTableMutation.id,
				params: {
					unitId,
					subUnitId,
					tableId,
					config: { rowColOperation: {
						operationType: "delete",
						rowColType: "column",
						index,
						count
					} }
				}
			},
			...interceptorCommands.redos
		];
		const undos = [
			...(_interceptorCommands$2 = interceptorCommands.preUndos) !== null && _interceptorCommands$2 !== void 0 ? _interceptorCommands$2 : [],
			{
				id: SetSheetTableMutation.id,
				params: {
					unitId,
					subUnitId,
					tableId,
					config: { rowColOperation: {
						operationType: "insert",
						rowColType: "column",
						index,
						count,
						columnsJson: columns
					} }
				}
			},
			...interceptorCommands.undos
		];
		const colContentIndex = target.worksheet.getCellMatrix().getDataRange().endColumn;
		if (index + count <= colContentIndex) {
			const moveRangeMutations = getMoveRangeUndoRedoMutations(accessor, {
				unitId,
				subUnitId,
				range: {
					startRow: oldRange.startRow,
					endRow: oldRange.endRow,
					startColumn: index + count,
					endColumn: colContentIndex
				}
			}, {
				unitId,
				subUnitId,
				range: {
					startRow: oldRange.startRow,
					endRow: oldRange.endRow,
					startColumn: index,
					endColumn: colContentIndex - count
				}
			});
			if (moveRangeMutations) {
				redos.push(...moveRangeMutations.redos);
				undos.push(...moveRangeMutations.undos);
			}
		}
		return executeTableMutationSequence(accessor, unitId, redos, undos);
	}
};
const SheetTableRemoveRowCommand = {
	id: "sheet.command.table-remove-row",
	type: CommandType.COMMAND,
	handler: (accessor, params) => {
		if (!params) return false;
		const target = getSheetCommandTarget(accessor.get(IUniverInstanceService), params);
		if (!target) return false;
		const { unitId, subUnitId } = target;
		const selections = accessor.get(SheetsSelectionsService).getCurrentSelections();
		if (!selections.length || selections.length > 1) return false;
		const range = selections[0].range;
		const table = accessor.get(SheetsTableController).getContainerTableWithRange(unitId, subUnitId, range);
		if (!table) return false;
		const removeRowCount = range.endRow - range.startRow + 1;
		const redos = [];
		const undos = [];
		const oldRange = table.getRange();
		redos.push({
			id: SetSheetTableMutation.id,
			params: {
				unitId,
				subUnitId,
				tableId: table.getId(),
				config: { updateRange: { newRange: {
					...oldRange,
					endRow: oldRange.endRow - removeRowCount
				} } }
			}
		});
		undos.push({
			id: SetSheetTableMutation.id,
			params: {
				unitId,
				subUnitId,
				tableId: table.getId(),
				config: { updateRange: { newRange: { ...oldRange } } }
			}
		});
		const rowContentIndex = target.worksheet.getCellMatrix().getDataRange().endRow;
		const moveRangeMutations = getMoveRangeUndoRedoMutations(accessor, {
			unitId,
			subUnitId,
			range: {
				startRow: range.endRow + 1,
				endRow: rowContentIndex,
				startColumn: oldRange.startColumn,
				endColumn: oldRange.endColumn
			}
		}, {
			unitId,
			subUnitId,
			range: {
				startRow: range.startRow,
				endRow: rowContentIndex - removeRowCount,
				startColumn: oldRange.startColumn,
				endColumn: oldRange.endColumn
			}
		});
		if (moveRangeMutations) {
			redos.push(...moveRangeMutations.redos);
			undos.push(...moveRangeMutations.undos);
		}
		if (sequenceExecute(redos, accessor.get(ICommandService)).result) {
			accessor.get(IUndoRedoService).pushUndoRedo({
				unitID: unitId,
				undoMutations: undos,
				redoMutations: redos
			});
			return true;
		}
		return false;
	}
};
const SheetTableRemoveColCommand = {
	id: "sheet.command.table-remove-col",
	type: CommandType.COMMAND,
	handler: (accessor, params) => {
		var _interceptorCommands$3, _interceptorCommands$4;
		if (!params) return false;
		const target = getSheetCommandTarget(accessor.get(IUniverInstanceService), params);
		if (!target) return false;
		const { workbook, unitId, subUnitId } = target;
		accessor.get(TableManager);
		const selections = accessor.get(SheetsSelectionsService).getCurrentSelections();
		if (!selections.length || selections.length > 1) return false;
		const range = selections[0].range;
		const table = accessor.get(SheetsTableController).getContainerTableWithRange(unitId, subUnitId, range);
		if (!table) return false;
		const removeColCount = range.endColumn - range.startColumn + 1;
		const redos = [];
		const undos = [];
		const oldRange = table.getRange();
		redos.push({
			id: SetSheetTableMutation.id,
			params: {
				unitId,
				subUnitId,
				tableId: table.getId(),
				config: { rowColOperation: {
					operationType: "delete",
					rowColType: "column",
					index: range.startColumn,
					count: removeColCount
				} }
			}
		});
		const tableInfo = table.getTableInfo();
		const columns = [];
		const gap = range.startColumn - oldRange.startColumn;
		for (let i = 0; i < removeColCount; i++) {
			const column = tableInfo.columns[gap + i];
			if (column) columns.push(column);
		}
		const interceptorCommands = accessor.get(SheetInterceptorService).onCommandExecute({
			id: SheetTableRemoveColCommand.id,
			params: {
				...params,
				tableName: tableInfo.name,
				removedColumnNames: columns.map((column) => column.displayName)
			}
		});
		redos.unshift(...(_interceptorCommands$3 = interceptorCommands.preRedos) !== null && _interceptorCommands$3 !== void 0 ? _interceptorCommands$3 : []);
		redos.push(...interceptorCommands.redos);
		undos.unshift(...(_interceptorCommands$4 = interceptorCommands.preUndos) !== null && _interceptorCommands$4 !== void 0 ? _interceptorCommands$4 : []);
		undos.push({
			id: SetSheetTableMutation.id,
			params: {
				unitId,
				subUnitId,
				tableId: table.getId(),
				config: { rowColOperation: {
					operationType: "insert",
					rowColType: "column",
					index: range.startColumn,
					count: removeColCount,
					columnsJson: columns
				} }
			}
		});
		undos.push(...interceptorCommands.undos);
		const colContentIndex = target.worksheet.getCellMatrix().getDataRange().endColumn;
		const moveRangeMutations = getMoveRangeUndoRedoMutations(accessor, {
			unitId,
			subUnitId,
			range: {
				startRow: oldRange.startRow,
				endRow: oldRange.endRow,
				startColumn: range.endColumn + 1,
				endColumn: colContentIndex
			}
		}, {
			unitId,
			subUnitId,
			range: {
				startRow: oldRange.startRow,
				endRow: oldRange.endRow,
				startColumn: range.startColumn,
				endColumn: colContentIndex - removeColCount
			}
		});
		if (moveRangeMutations) {
			redos.push(...moveRangeMutations.redos);
			undos.push(...moveRangeMutations.undos);
		}
		if (sequenceExecute(redos, accessor.get(ICommandService)).result) {
			accessor.get(IUndoRedoService).pushUndoRedo({
				unitID: unitId,
				undoMutations: undos,
				redoMutations: redos
			});
			return true;
		}
		return false;
	}
};

//#endregion
//#region package.json
var name = "@univerjs/sheets-table";
var version = "0.25.1";

//#endregion
//#region src/controllers/sheet-table-formula.controller.ts
let SheetTableFormulaController = class SheetTableFormulaController extends Disposable {
	constructor(_tableManager, _commandService) {
		super();
		this._tableManager = _tableManager;
		this._commandService = _commandService;
		this._initRangeListener();
	}
	_initRangeListener() {
		this.disposeWithMe(this._tableManager.tableRangeChanged$.subscribe((event) => {
			const { tableId, unitId } = event;
			const table = this._tableManager.getTableById(unitId, tableId);
			if (!table) return;
			this._updateSuperTable(unitId, table);
		}));
		this.disposeWithMe(this._tableManager.tableAdd$.subscribe((event) => {
			const { tableId, unitId } = event;
			const table = this._tableManager.getTableById(unitId, tableId);
			if (!table) return;
			this._updateSuperTable(unitId, table);
		}));
		this.disposeWithMe(this._tableManager.tableDelete$.subscribe((event) => {
			const { unitId, tableName } = event;
			this._commandService.executeCommand(RemoveSuperTableMutation.id, {
				unitId,
				tableName
			});
		}));
		this.disposeWithMe(this._tableManager.tableNameChanged$.subscribe((event) => {
			const { tableId, unitId, oldTableName } = event;
			this._commandService.executeCommand(RemoveSuperTableMutation.id, {
				unitId,
				tableName: oldTableName
			});
			const table = this._tableManager.getTableById(unitId, tableId);
			if (!table) return;
			this._updateSuperTable(unitId, table, oldTableName);
		}));
	}
	_updateSuperTable(unitId, table, oldTableName) {
		const tableInfo = table.getTableInfo();
		const name = tableInfo.name;
		const columns = tableInfo.columns;
		const titleMap = /* @__PURE__ */ new Map();
		columns.forEach((column, index) => {
			titleMap.set(column.displayName, index);
		});
		this._commandService.executeCommand(SetSuperTableMutation.id, {
			unitId,
			tableName: name,
			oldTableName,
			reference: {
				range: tableInfo.range,
				sheetId: tableInfo.subUnitId,
				titleMap
			}
		});
	}
};
SheetTableFormulaController = __decorate([__decorateParam(0, Inject(TableManager)), __decorateParam(1, ICommandService)], SheetTableFormulaController);

//#endregion
//#region src/controllers/sheet-table-range.controller.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
let SheetTableRangeController = class SheetTableRangeController extends Disposable {
	constructor(_tableManager, _exclusiveRangeService) {
		super();
		this._tableManager = _tableManager;
		this._exclusiveRangeService = _exclusiveRangeService;
		this._initRangeListener();
	}
	_initRangeListener() {
		this.disposeWithMe(this._tableManager.tableRangeChanged$.subscribe((event) => {
			const { range, tableId, unitId, subUnitId } = event;
			this._exclusiveRangeService.clearExclusiveRangesByGroupId(unitId, subUnitId, FEATURE_TABLE_ID, tableId);
			this._exclusiveRangeService.addExclusiveRange(unitId, subUnitId, FEATURE_TABLE_ID, [{
				range: { ...range },
				groupId: tableId
			}]);
		}));
		this.disposeWithMe(this._tableManager.tableAdd$.subscribe((event) => {
			const { tableId, unitId, subUnitId, range } = event;
			this._exclusiveRangeService.addExclusiveRange(unitId, subUnitId, FEATURE_TABLE_ID, [{
				range: { ...range },
				groupId: tableId
			}]);
		}));
		this.disposeWithMe(this._tableManager.tableDelete$.subscribe((event) => {
			const { tableId, unitId, subUnitId } = event;
			this._exclusiveRangeService.clearExclusiveRangesByGroupId(unitId, subUnitId, FEATURE_TABLE_ID, tableId);
		}));
	}
};
SheetTableRangeController = __decorate([__decorateParam(0, Inject(TableManager)), __decorateParam(1, Inject(IExclusiveRangeService))], SheetTableRangeController);

//#endregion
//#region src/controllers/sheet-table-ref-range.controller.ts
const SHEET_TABLE_REMOVE_COL_COMMAND_ID = "sheet.command.table-remove-col";
const DELETE_SHEET_TABLE_COMMAND_ID = "sheet.command.delete-table";
let SheetTableRefRangeController = class SheetTableRefRangeController extends Disposable {
	constructor(_commandService, _refRangeService, _univerInstanceService, _injector, _sheetInterceptorService, _tableManager, _localeService) {
		super();
		this._commandService = _commandService;
		this._refRangeService = _refRangeService;
		this._univerInstanceService = _univerInstanceService;
		this._injector = _injector;
		this._sheetInterceptorService = _sheetInterceptorService;
		this._tableManager = _tableManager;
		this._localeService = _localeService;
		this._initCommandInterceptor();
		this._initCommandListener();
	}
	_initCommandInterceptor() {
		const self = this;
		this._sheetInterceptorService.interceptCommand({
			priority: -1,
			getMutations(commandInfo) {
				const defaultReturn = {
					redos: [],
					undos: []
				};
				const { id, params } = commandInfo;
				switch (id) {
					case InsertRowCommand.id: return self._generateTableMutationWithInsertRow(params);
					case InsertColCommand.id: return self._generateTableMutationWithInsertCol(params);
					case RemoveRowCommand.id: return self._generateTableMutationWithRemoveRow(params);
					case RemoveColCommand.id: return self._generateTableMutationWithRemoveCol(params);
				}
				return defaultReturn;
			}
		});
	}
	_generateTableMutationWithInsertRow(insertParams) {
		const undos = [];
		const redos = [];
		const target = getSheetCommandTarget(this._univerInstanceService, insertParams);
		if (!target) return {
			undos,
			redos
		};
		const { unitId, subUnitId } = target;
		const allSubUnitTables = this._tableManager.getTablesBySubunitId(unitId, subUnitId);
		if (!allSubUnitTables.length) return {
			undos,
			redos
		};
		const { range } = insertParams;
		allSubUnitTables.forEach((table) => {
			const tableRange = table.getRange();
			if (range.startRow > tableRange.startRow && range.startRow <= tableRange.endRow) {
				const insertRowCount = range.endRow - range.startRow + 1;
				redos.push({
					id: SetSheetTableMutation.id,
					params: {
						unitId,
						subUnitId,
						tableId: table.getId(),
						config: { updateRange: { newRange: {
							...tableRange,
							endRow: tableRange.endRow + insertRowCount
						} } }
					}
				});
				undos.push({
					id: SetSheetTableMutation.id,
					params: {
						unitId,
						subUnitId,
						tableId: table.getId(),
						config: { updateRange: { newRange: { ...tableRange } } }
					}
				});
			}
		});
		return {
			undos,
			redos
		};
	}
	_generateTableMutationWithInsertCol(insertParams) {
		const undos = [];
		const redos = [];
		const target = getSheetCommandTarget(this._univerInstanceService, insertParams);
		if (!target) return {
			undos,
			redos
		};
		const { unitId, subUnitId } = target;
		const allSubUnitTables = this._tableManager.getTablesBySubunitId(unitId, subUnitId);
		if (!allSubUnitTables.length) return {
			undos,
			redos
		};
		const { range } = insertParams;
		allSubUnitTables.forEach((table) => {
			const tableRange = table.getRange();
			if (range.startColumn > tableRange.startColumn && range.startColumn <= tableRange.endColumn) {
				const insertColCount = range.endColumn - range.startColumn + 1;
				redos.push({
					id: SetSheetTableMutation.id,
					params: {
						unitId,
						subUnitId,
						tableId: table.getId(),
						config: { rowColOperation: {
							operationType: "insert",
							rowColType: "column",
							index: range.startColumn,
							count: insertColCount
						} }
					}
				});
				undos.push({
					id: SetSheetTableMutation.id,
					params: {
						unitId,
						subUnitId,
						tableId: table.getId(),
						config: { rowColOperation: {
							operationType: "delete",
							rowColType: "column",
							index: range.startColumn,
							count: insertColCount
						} }
					}
				});
			}
		});
		return {
			undos,
			redos
		};
	}
	_generateTableMutationWithRemoveRow(removeParams) {
		const undos = [];
		const redos = [];
		const preRedos = [];
		const preUndos = [];
		const target = getSheetCommandTarget(this._univerInstanceService);
		if (!target) return {
			undos,
			redos,
			preRedos,
			preUndos
		};
		const { unitId, subUnitId } = target;
		const allSubUnitTables = this._tableManager.getTablesBySubunitId(unitId, subUnitId);
		if (!allSubUnitTables.length) return {
			undos,
			redos,
			preRedos,
			preUndos
		};
		const { range } = removeParams;
		const removeRowCount = range.endRow - range.startRow + 1;
		allSubUnitTables.forEach((table) => {
			const tableRange = table.getRange();
			if (Rectangle.intersects(tableRange, range)) {
				if (range.startRow <= tableRange.startRow && range.endRow >= tableRange.startRow) {
					preRedos.push({
						id: DeleteSheetTableMutation.id,
						params: {
							unitId,
							subUnitId,
							tableId: table.getId()
						}
					});
					const tableJson = table.toJSON();
					undos.push({
						id: AddSheetTableMutation.id,
						params: {
							unitId,
							subUnitId,
							tableId: tableJson.id,
							name: tableJson.name,
							range: tableJson.range,
							options: tableJson.options
						}
					});
				} else if (range.startRow > tableRange.startRow && range.startRow <= tableRange.endRow) {
					redos.push({
						id: SetSheetTableMutation.id,
						params: {
							unitId,
							subUnitId,
							tableId: table.getId(),
							config: { updateRange: { newRange: {
								...tableRange,
								endRow: tableRange.endRow - removeRowCount
							} } }
						}
					});
					undos.push({
						id: SetSheetTableMutation.id,
						params: {
							unitId,
							subUnitId,
							tableId: table.getId(),
							config: { updateRange: { newRange: { ...tableRange } } }
						}
					});
				} else if (range.startRow < tableRange.endRow && range.endRow >= tableRange.endRow) {
					redos.push({
						id: SetSheetTableMutation.id,
						params: {
							unitId,
							subUnitId,
							tableId: table.getId(),
							config: { updateRange: { newRange: {
								...tableRange,
								endRow: range.startRow - 1
							} } }
						}
					});
					undos.push({
						id: SetSheetTableMutation.id,
						params: {
							unitId,
							subUnitId,
							tableId: table.getId(),
							config: { updateRange: { newRange: { ...tableRange } } }
						}
					});
				}
			}
		});
		return {
			undos,
			redos,
			preRedos,
			preUndos
		};
	}
	_generateTableMutationWithRemoveCol(removeParams) {
		const undos = [];
		const redos = [];
		const preRedos = [];
		const preUndos = [];
		const target = getSheetCommandTarget(this._univerInstanceService);
		if (!target) return {
			undos,
			redos,
			preRedos,
			preUndos
		};
		const { unitId, subUnitId } = target;
		const allSubUnitTables = this._tableManager.getTablesBySubunitId(unitId, subUnitId);
		if (!allSubUnitTables.length) return {
			undos,
			redos,
			preRedos,
			preUndos
		};
		const { range } = removeParams;
		const removeColCount = range.endColumn - range.startColumn + 1;
		allSubUnitTables.forEach((table) => {
			const tableRange = table.getRange();
			if (Rectangle.intersects(tableRange, range)) {
				if (range.startColumn <= tableRange.startColumn && range.endColumn >= tableRange.endColumn) {
					var _formulaMutations$pre, _formulaMutations$pre2;
					const tableInfo = table.getTableInfo();
					const formulaMutations = this._sheetInterceptorService.onCommandExecute({
						id: DELETE_SHEET_TABLE_COMMAND_ID,
						params: {
							unitId,
							subUnitId,
							tableId: table.getId(),
							tableName: tableInfo.name
						}
					});
					preRedos.push(...(_formulaMutations$pre = formulaMutations.preRedos) !== null && _formulaMutations$pre !== void 0 ? _formulaMutations$pre : [], ...formulaMutations.redos);
					preRedos.push({
						id: DeleteSheetTableMutation.id,
						params: {
							unitId,
							subUnitId,
							tableId: table.getId()
						}
					});
					const tableJson = table.toJSON();
					const { startRow, startColumn, endColumn } = tableJson.range;
					const workbook = this._univerInstanceService.getUnit(unitId);
					const worksheet = workbook === null || workbook === void 0 ? void 0 : workbook.getSheetBySheetId(subUnitId);
					if (!worksheet) return {
						undos,
						redos,
						preRedos,
						preUndos
					};
					const header = [];
					for (let i = startColumn; i <= endColumn; i++) header.push(convertCellDataToString(worksheet === null || worksheet === void 0 ? void 0 : worksheet.getCell(startRow, i)) || getColumnName(i - startColumn + 1, this._localeService.t("sheets-table.columnPrefix")));
					undos.push({
						id: AddSheetTableMutation.id,
						params: {
							unitId,
							subUnitId,
							tableId: tableJson.id,
							name: tableJson.name,
							header,
							range: tableJson.range,
							options: tableJson.options
						}
					});
					undos.push(...(_formulaMutations$pre2 = formulaMutations.preUndos) !== null && _formulaMutations$pre2 !== void 0 ? _formulaMutations$pre2 : [], ...formulaMutations.undos);
				} else if (range.startColumn <= tableRange.startColumn && range.endColumn >= tableRange.startColumn) {
					const tableJson = table.toJSON();
					const removeColumnCount = range.endColumn - tableRange.startColumn + 1;
					redos.push({
						id: SetSheetTableMutation.id,
						params: {
							unitId,
							subUnitId,
							tableId: table.getId(),
							config: { rowColOperation: {
								operationType: "delete",
								rowColType: "column",
								index: tableRange.startColumn,
								count: removeColumnCount
							} }
						}
					});
					const columns = [];
					for (let i = 0; i < removeColumnCount; i++) {
						const column = table.getTableColumnByIndex(i);
						if (column) columns.push(column.toJSON());
					}
					preUndos.push(this._getDeleteTableMutation(unitId, subUnitId, table.getId()));
					undos.push(this._getAddTableMutation(unitId, subUnitId, tableJson));
					this._appendTableColumnFormulaMutations(redos, undos, {
						unitId,
						subUnitId,
						tableId: table.getId(),
						tableName: table.getTableInfo().name,
						range,
						columns
					});
				} else if (range.startColumn > tableRange.startColumn && range.endColumn > tableRange.endColumn) {
					const tableJson = table.toJSON();
					const removeColumnCount = tableRange.endColumn - range.startColumn + 1;
					redos.push({
						id: SetSheetTableMutation.id,
						params: {
							unitId,
							subUnitId,
							tableId: table.getId(),
							config: { rowColOperation: {
								operationType: "delete",
								rowColType: "column",
								index: range.startColumn,
								count: removeColumnCount
							} }
						}
					});
					const columns = [];
					const gap = range.startColumn - tableRange.startColumn;
					for (let i = 0; i < removeColumnCount; i++) {
						const column = table.getTableColumnByIndex(i + gap);
						if (column) columns.push(column.toJSON());
					}
					preUndos.push(this._getDeleteTableMutation(unitId, subUnitId, table.getId()));
					undos.push(this._getAddTableMutation(unitId, subUnitId, tableJson));
					this._appendTableColumnFormulaMutations(redos, undos, {
						unitId,
						subUnitId,
						tableId: table.getId(),
						tableName: table.getTableInfo().name,
						range,
						columns
					});
				} else if (range.startColumn > tableRange.startColumn && range.endColumn <= tableRange.endColumn) {
					const tableJson = table.toJSON();
					redos.push({
						id: SetSheetTableMutation.id,
						params: {
							unitId,
							subUnitId,
							tableId: table.getId(),
							config: { rowColOperation: {
								operationType: "delete",
								rowColType: "column",
								index: range.startColumn,
								count: removeColCount
							} }
						}
					});
					const columns = [];
					const gap = range.startColumn - tableRange.startColumn;
					for (let i = 0; i < removeColCount; i++) {
						const column = table.getTableColumnByIndex(i + gap);
						if (column) columns.push(column.toJSON());
					}
					preUndos.push(this._getDeleteTableMutation(unitId, subUnitId, table.getId()));
					undos.push(this._getAddTableMutation(unitId, subUnitId, tableJson));
					this._appendTableColumnFormulaMutations(redos, undos, {
						unitId,
						subUnitId,
						tableId: table.getId(),
						tableName: table.getTableInfo().name,
						range,
						columns
					});
				}
			}
		});
		return {
			undos,
			redos,
			preRedos,
			preUndos
		};
	}
	_getDeleteTableMutation(unitId, subUnitId, tableId) {
		return {
			id: DeleteSheetTableMutation.id,
			params: {
				unitId,
				subUnitId,
				tableId
			}
		};
	}
	_getAddTableMutation(unitId, subUnitId, tableJson) {
		const header = tableJson.columns.map((column) => column.displayName);
		return {
			id: AddSheetTableMutation.id,
			params: {
				unitId,
				subUnitId,
				tableId: tableJson.id,
				name: tableJson.name,
				header,
				range: tableJson.range,
				options: {
					...tableJson.options,
					columns: tableJson.columns,
					filters: tableJson.filters.tableColumnFilterList
				}
			}
		};
	}
	_appendTableColumnFormulaMutations(redos, undos, info) {
		var _formulaMutations$pre3, _formulaMutations$pre4;
		const removedColumnNames = info.columns.map((column) => column.displayName);
		if (!removedColumnNames.length) return;
		const formulaMutations = this._sheetInterceptorService.onCommandExecute({
			id: SHEET_TABLE_REMOVE_COL_COMMAND_ID,
			params: {
				unitId: info.unitId,
				subUnitId: info.subUnitId,
				tableId: info.tableId,
				tableName: info.tableName,
				range: info.range,
				removedColumnNames
			}
		});
		redos.splice(Math.max(redos.length - 1, 0), 0, ...(_formulaMutations$pre3 = formulaMutations.preRedos) !== null && _formulaMutations$pre3 !== void 0 ? _formulaMutations$pre3 : [], ...formulaMutations.redos);
		undos.push(...(_formulaMutations$pre4 = formulaMutations.preUndos) !== null && _formulaMutations$pre4 !== void 0 ? _formulaMutations$pre4 : [], ...formulaMutations.undos);
	}
	_initCommandListener() {
		this._commandService.onCommandExecuted((commandInfo) => {
			if (commandInfo.id === InsertRowMutation.id) {
				const { unitId, subUnitId, range } = commandInfo.params;
				const insertCount = range.endRow - range.startRow + 1;
				this._tableManager.getTablesBySubunitId(unitId, subUnitId).forEach((table) => {
					const tableRange = table.getRange();
					if (range.startRow <= tableRange.startRow) this._tableManager.updateTableRange(unitId, table.getId(), { newRange: {
						...tableRange,
						startRow: tableRange.startRow + insertCount,
						endRow: tableRange.endRow + insertCount
					} });
				});
			} else if (commandInfo.id === InsertColMutation.id) {
				const { unitId, subUnitId, range } = commandInfo.params;
				const insertCount = range.endColumn - range.startColumn + 1;
				this._tableManager.getTablesBySubunitId(unitId, subUnitId).forEach((table) => {
					const tableRange = table.getRange();
					if (range.startColumn <= tableRange.startColumn) this._tableManager.updateTableRange(unitId, table.getId(), { newRange: {
						...tableRange,
						startColumn: tableRange.startColumn + insertCount,
						endColumn: tableRange.endColumn + insertCount
					} });
				});
			} else if (commandInfo.id === RemoveRowMutation.id) {
				const { unitId, subUnitId, range } = commandInfo.params;
				const removeCount = range.endRow - range.startRow + 1;
				this._tableManager.getTablesBySubunitId(unitId, subUnitId).forEach((table) => {
					const tableRange = table.getRange();
					if (range.startRow < tableRange.startRow) this._tableManager.updateTableRange(unitId, table.getId(), { newRange: {
						...tableRange,
						startRow: tableRange.startRow - removeCount,
						endRow: tableRange.endRow - removeCount
					} });
				});
			} else if (commandInfo.id === RemoveColMutation.id) {
				const { unitId, subUnitId, range } = commandInfo.params;
				const removeCount = range.endColumn - range.startColumn + 1;
				this._tableManager.getTablesBySubunitId(unitId, subUnitId).forEach((table) => {
					const tableRange = table.getRange();
					if (range.startColumn < tableRange.startColumn) this._tableManager.updateTableRange(unitId, table.getId(), { newRange: {
						...tableRange,
						startColumn: tableRange.startColumn - removeCount,
						endColumn: tableRange.endColumn - removeCount
					} });
				});
			}
		});
	}
};
SheetTableRefRangeController = __decorate([
	__decorateParam(0, Inject(ICommandService)),
	__decorateParam(1, Inject(RefRangeService)),
	__decorateParam(2, Inject(IUniverInstanceService)),
	__decorateParam(3, Inject(Injector)),
	__decorateParam(4, Inject(SheetInterceptorService)),
	__decorateParam(5, Inject(TableManager)),
	__decorateParam(6, Inject(LocaleService))
], SheetTableRefRangeController);

//#endregion
//#region src/controllers/sheet-table-theme.controller.ts
let SheetsTableThemeController = class SheetsTableThemeController extends Disposable {
	constructor(_tableManager, _sheetRangeThemeService, _sheetRangeThemeModel, _configService) {
		super();
		this._tableManager = _tableManager;
		this._sheetRangeThemeService = _sheetRangeThemeService;
		this._sheetRangeThemeModel = _sheetRangeThemeModel;
		this._configService = _configService;
		_defineProperty(this, "_defaultThemeIndex", 0);
		_defineProperty(this, "_allThemes", []);
		this._initUserTableTheme();
		this.registerTableChangeEvent();
		this._initDefaultTableTheme();
	}
	registerTableChangeEvent() {
		this.disposeWithMe(this._tableManager.tableAdd$.subscribe((event) => {
			const { range, tableId, unitId, subUnitId, tableStyleId } = event;
			const table = this._tableManager.getTable(unitId, tableId);
			const _tableStyleId = tableStyleId || this._allThemes[this._defaultThemeIndex].name;
			table.setTableStyleId(_tableStyleId);
			this._sheetRangeThemeService.registerRangeThemeStyle(_tableStyleId, {
				unitId,
				subUnitId,
				range: { ...range }
			});
		}));
		this.disposeWithMe(this._tableManager.tableRangeChanged$.subscribe((event) => {
			const { range, oldRange, tableId, unitId, subUnitId } = event;
			const table = this._tableManager.getTable(unitId, tableId);
			let tableStyleId = table.getTableStyleId();
			if (!tableStyleId) {
				tableStyleId = this._allThemes[this._defaultThemeIndex].name;
				table.setTableStyleId(tableStyleId);
			}
			this._sheetRangeThemeService.removeRangeThemeRule(tableStyleId, {
				unitId,
				subUnitId,
				range: { ...oldRange }
			});
			this._sheetRangeThemeService.registerRangeThemeStyle(tableStyleId, {
				unitId,
				subUnitId,
				range: { ...range }
			});
		}));
		this.disposeWithMe(this._tableManager.tableThemeChanged$.subscribe((event) => {
			const { theme, oldTheme, tableId, unitId, subUnitId } = event;
			const range = this._tableManager.getTable(unitId, tableId).getRange();
			this._sheetRangeThemeService.removeRangeThemeRule(oldTheme, {
				unitId,
				subUnitId,
				range: { ...range }
			});
			this._sheetRangeThemeService.registerRangeThemeStyle(theme, {
				unitId,
				subUnitId,
				range: { ...range }
			});
		}));
		this.disposeWithMe(this._tableManager.tableDelete$.subscribe((event) => {
			const { range, unitId, subUnitId, tableStyleId = this._allThemes[this._defaultThemeIndex].name } = event;
			this._sheetRangeThemeService.removeRangeThemeRule(tableStyleId, {
				unitId,
				subUnitId,
				range: { ...range }
			});
		}));
	}
	_initUserTableTheme() {
		const tableConfig = this._configService.getConfig("sheets-table.config") || {};
		const defaultThemeIndex = tableConfig.defaultThemeIndex || 0;
		const userThemes = tableConfig.userThemes || [];
		this._defaultThemeIndex = defaultThemeIndex;
		this._allThemes = userThemes.concat(tableThemeConfig);
	}
	_initDefaultTableTheme() {
		for (let i = 0; i < this._allThemes.length; i++) {
			const { name, style } = this._allThemes[i];
			const rangeThemeStyle = new RangeThemeStyle(name, style);
			this._sheetRangeThemeModel.registerDefaultRangeTheme(rangeThemeStyle);
		}
	}
	dispose() {
		super.dispose();
		this._allThemes = [];
		this._defaultThemeIndex = 0;
	}
};
SheetsTableThemeController = __decorate([
	__decorateParam(0, Inject(TableManager)),
	__decorateParam(1, Inject(SheetRangeThemeService)),
	__decorateParam(2, Inject(SheetRangeThemeModel)),
	__decorateParam(3, IConfigService)
], SheetsTableThemeController);

//#endregion
//#region src/controllers/table-filter.controller.ts
let TableFilterController = class TableFilterController extends Disposable {
	constructor(_tableManager, _sheetInterceptorService, _univerInstanceService, _zebraCrossingCacheController) {
		super();
		this._tableManager = _tableManager;
		this._sheetInterceptorService = _sheetInterceptorService;
		this._univerInstanceService = _univerInstanceService;
		this._zebraCrossingCacheController = _zebraCrossingCacheController;
		_defineProperty(this, "_tableFilteredOutRows", /* @__PURE__ */ new Map());
		_defineProperty(this, "_subscription", null);
		this.registerFilterChangeEvent();
		this.initTableHiddenRowIntercept();
		this._initFilteredOutRows();
	}
	initTableHiddenRowIntercept() {
		this.disposeWithMe(this._sheetInterceptorService.intercept(INTERCEPTOR_POINT.ROW_FILTERED, {
			priority: 100,
			handler: (filtered, rowLocation, next) => {
				if (filtered) return true;
				const isTableFiltered = this._getTableFilteredOutRows(rowLocation.unitId, rowLocation.subUnitId).has(rowLocation.row);
				return isTableFiltered ? true : next(isTableFiltered);
			}
		}));
	}
	_initFilteredOutRows() {
		this._tableManager.tableInitStatus$.pipe(filter((initialized) => initialized), switchMap(() => {
			return this._univerInstanceService.getCurrentTypeOfUnit$(UniverInstanceType.UNIVER_SHEET);
		}), filter((workbook) => workbook !== null && workbook !== void 0), switchMap((workbook) => workbook.activeSheet$), filter((sheet) => sheet !== null && sheet !== void 0)).subscribe(() => {
			const target = getSheetCommandTarget(this._univerInstanceService);
			if (!target) return;
			const { unitId, subUnitId } = target;
			this._refreshTableFilteredOutRows(unitId, subUnitId);
		});
	}
	registerFilterChangeEvent() {
		this.disposeWithMe(this._tableManager.tableFilterChanged$.subscribe((event) => {
			var _this$_univerInstance;
			const { unitId, subUnitId, tableId } = event;
			const worksheet = (_this$_univerInstance = this._univerInstanceService.getUnit(unitId)) === null || _this$_univerInstance === void 0 ? void 0 : _this$_univerInstance.getSheetBySheetId(subUnitId);
			const table = this._tableManager.getTable(unitId, tableId);
			if (!worksheet || !table) return;
			table.getTableFilters().doFilter(worksheet, table.getTableFilterRange());
			this._refreshTableFilteredOutRows(unitId, subUnitId);
			this._zebraCrossingCacheController.updateZebraCrossingCache(unitId, subUnitId);
		}));
	}
	_refreshTableFilteredOutRows(unitId, subUnitId) {
		const filteredOutRows = /* @__PURE__ */ new Set();
		this._tableManager.getTablesBySubunitId(unitId, subUnitId).forEach((table) => {
			const tableFilteredRows = table.getTableFilters().getFilterOutRows();
			if (!tableFilteredRows) return;
			for (const row of tableFilteredRows) filteredOutRows.add(row);
		});
		this._tableFilteredOutRows.set(this._getSheetKey(unitId, subUnitId), filteredOutRows);
	}
	_getTableFilteredOutRows(unitId, subUnitId) {
		var _this$_tableFilteredO;
		return (_this$_tableFilteredO = this._tableFilteredOutRows.get(this._getSheetKey(unitId, subUnitId))) !== null && _this$_tableFilteredO !== void 0 ? _this$_tableFilteredO : /* @__PURE__ */ new Set();
	}
	_getSheetKey(unitId, subUnitId) {
		return `${unitId}|${subUnitId}`;
	}
	dispose() {
		var _this$_subscription;
		super.dispose();
		(_this$_subscription = this._subscription) === null || _this$_subscription === void 0 || _this$_subscription.unsubscribe();
	}
};
TableFilterController = __decorate([
	__decorateParam(0, Inject(TableManager)),
	__decorateParam(1, Inject(SheetInterceptorService)),
	__decorateParam(2, Inject(IUniverInstanceService)),
	__decorateParam(3, Inject(ZebraCrossingCacheController))
], TableFilterController);

//#endregion
//#region src/plugin.ts
let UniverSheetsTablePlugin = class UniverSheetsTablePlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _injector, _configService, _commandService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		this._commandService = _commandService;
		const { ...rest } = merge({}, defaultPluginConfig, this._config);
		this._configService.setConfig(SHEETS_TABLE_PLUGIN_CONFIG_KEY, rest);
		this._initRegisterCommand();
	}
	onStarting() {
		registerDependencies(this._injector, [
			[TableManager],
			[SheetsTableThemeController],
			[SheetsTableController],
			[SheetTableService],
			[TableFilterController],
			[SheetTableRangeController],
			[SheetTableRefRangeController],
			[SheetTableFormulaController]
		]);
	}
	onReady() {
		touchDependencies(this._injector, [
			[SheetTableFormulaController],
			[SheetTableRangeController],
			[SheetTableRefRangeController],
			[SheetsTableThemeController],
			[SheetsTableController],
			[SheetTableService],
			[TableFilterController]
		]);
		touchDependencies(this._injector, [[TableManager]]);
	}
	_initRegisterCommand() {
		[
			AddSheetTableCommand,
			AddSheetTableMutation,
			DeleteSheetTableCommand,
			DeleteSheetTableMutation,
			SetSheetTableFilterMutation,
			SetSheetTableFilterCommand,
			SetSheetTableCommand,
			SetSheetTableMutation,
			AddTableThemeCommand,
			RemoveTableThemeCommand,
			SheetTableInsertRowCommand,
			SheetTableInsertColCommand,
			SheetTableInsertRowAtCommand,
			SheetTableInsertColumnAtCommand,
			SheetTableRemoveRowCommand,
			SheetTableRemoveColCommand,
			SheetTableRemoveColumnAtCommand
		].forEach((m) => this._commandService.registerCommand(m));
	}
};
_defineProperty(UniverSheetsTablePlugin, "pluginName", PLUGIN_NAME);
_defineProperty(UniverSheetsTablePlugin, "packageName", name);
_defineProperty(UniverSheetsTablePlugin, "version", version);
_defineProperty(UniverSheetsTablePlugin, "type", UniverInstanceType.UNIVER_SHEET);
UniverSheetsTablePlugin = __decorate([
	__decorateParam(1, Inject(Injector)),
	__decorateParam(2, IConfigService),
	__decorateParam(3, Inject(ICommandService))
], UniverSheetsTablePlugin);

//#endregion
export { AddSheetTableCommand, AddSheetTableMutation, AddTableThemeCommand, DeleteSheetTableCommand, DeleteSheetTableMutation, RemoveTableThemeCommand, SHEET_TABLE_CUSTOM_THEME_PREFIX, SetSheetTableCommand, SetSheetTableFilterCommand, SetSheetTableFilterMutation, SetSheetTableMutation, SheetTableInsertColCommand, SheetTableInsertColumnAtCommand, SheetTableInsertRowAtCommand, SheetTableInsertRowCommand, SheetTableRemoveColCommand, SheetTableRemoveColumnAtCommand, SheetTableRemoveRowCommand, SheetTableService, SheetsTableButtonStateEnum, SheetsTableController, SheetsTableSortStateEnum, TABLE_FILTER_EMPTY_VALUE, TableColumnDataTypeEnum, TableColumnFilterTypeEnum, TableConditionTypeEnum, TableDateCompareTypeEnum, TableManager, TableNumberCompareTypeEnum, TableStringCompareTypeEnum, UniverSheetsTablePlugin, customEmptyThemeWithBorderStyle, getExistingNamesSet, isConditionFilter, isManualTableFilter, processStyleWithBorderStyle, validateSheetTableName };
