// @univerjs/find-replace/lib/index.js
import { CommandType, Disposable, DisposableCollection, EDITOR_ACTIVATED, FOCUSING_SHEET, ICommandService, IConfigService, IConfirmService, IContextService, IUniverInstanceService, Inject, Injector, LocaleService, Plugin, RxDisposable, UniverInstanceType, createIdentifier, merge, toDisposable } from "@univerjs/core";
import { Button, Checkbox, FormDualColumnLayout, FormLayout, Input, MessageType, Pager, Select } from "@univerjs/design";
import { ComponentManager, IDialogService, ILayoutService, IMenuManagerService, IMessageService, IShortcutService, KeyCode, MenuItemType, MetaKeys, RibbonDataGroup, getMenuHiddenObservable, useDebounceFn, useDependency, useObservable } from "@univerjs/ui";
import { RENDER_RAW_FORMULA_KEY } from "@univerjs/engine-render";
import { BehaviorSubject, Subject, combineLatest, debounceTime, fromEvent, map, takeUntil, throttleTime } from "rxjs";
import { SearchIcon } from "@univerjs/icons";
import { forwardRef, useCallback, useEffect, useImperativeHandle, useMemo, useRef, useState } from "react";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";

//#region src/services/context-keys.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const FIND_REPLACE_INPUT_FOCUS = "FIND_REPLACE_INPUT_FOCUS";
/**
* If find replace panel is focused.
*/
const FIND_REPLACE_DIALOG_FOCUS = "FIND_REPLACE_DIALOG_FOCUS";
/**
* If the find replace feature is activated and the replace is revealed.
*/
const FIND_REPLACE_REPLACE_REVEALED = "FIND_REPLACE_REPLACE_REVEALED";

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/services/find-replace.service.ts
var FindModel = class extends Disposable {};
const IFindReplaceService = createIdentifier("find-replace.service");
/**
*
* @param statusUpdate
*/
function shouldStateUpdateTriggerResearch(statusUpdate) {
	if (typeof statusUpdate.findString !== "undefined") return true;
	if (typeof statusUpdate.inputtingFindString !== "undefined") return true;
	if (typeof statusUpdate.findDirection !== "undefined") return true;
	if (typeof statusUpdate.matchesTheWholeCell !== "undefined") return true;
	if (typeof statusUpdate.caseSensitive !== "undefined") return true;
	if (typeof statusUpdate.findScope !== "undefined") return true;
	if (typeof statusUpdate.findBy !== "undefined") return true;
	return false;
}
let FindReplaceModel = class FindReplaceModel extends Disposable {
	get searched() {
		return this._findModels.length > 0;
	}
	constructor(_state, _providers, _univerInstanceService, _commandService) {
		super();
		this._state = _state;
		this._providers = _providers;
		this._univerInstanceService = _univerInstanceService;
		this._commandService = _commandService;
		_defineProperty(this, "currentMatch$", new BehaviorSubject(null));
		_defineProperty(this, "replaceables$", new BehaviorSubject([]));
		_defineProperty(this, "_findModels", []);
		_defineProperty(this, "_matchingModel", null);
		_defineProperty(this, "_matches", []);
		_defineProperty(this, "_currentSearchingDisposables", null);
		this.disposeWithMe(this._state.stateUpdates$.pipe(throttleTime(200, void 0, {
			leading: true,
			trailing: true
		})).subscribe(async (stateUpdate) => {
			const state = this._state.state;
			if (shouldStateUpdateTriggerResearch(stateUpdate)) {
				if (state.findString !== "" && !state.replaceRevealed) {
					await this._startSearching();
					this._state.changeState({ findCompleted: true });
				} else if (stateUpdate.replaceRevealed !== true) this._stopSearching();
			}
		}));
	}
	dispose() {
		super.dispose();
		this._stopSearching();
		this.currentMatch$.complete();
		this.replaceables$.complete();
		this._state.changeState({
			...createInitFindReplaceState(),
			revealed: false
		});
	}
	async start() {
		if (!this._state.findString) return { results: [] };
		const complete = await this._startSearching();
		this._state.changeState({ findCompleted: true });
		return complete;
	}
	focusSelection() {
		var _this$_matchingModel;
		(_this$_matchingModel = this._matchingModel) === null || _this$_matchingModel === void 0 || _this$_matchingModel.focusSelection();
	}
	/** Call this method to start a `searching`. */
	async _startSearching() {
		if (!this._state.findString) return { results: [] };
		const providers = Array.from(this._providers);
		const findModels = this._findModels = (await Promise.all(providers.map((provider) => provider.find({
			findString: this._state.findString,
			findDirection: this._state.findDirection,
			findScope: this._state.findScope,
			findBy: this._state.findBy,
			replaceRevealed: this._state.replaceRevealed,
			caseSensitive: this._state.caseSensitive,
			matchesTheWholeCell: this._state.matchesTheWholeCell
		})))).flat();
		this._subscribeToModelsChanges(findModels);
		const newMatches = this._matches = findModels.map((c) => c.getMatches()).flat();
		this.replaceables$.next(newMatches.filter((m) => m.replaceable));
		if (!newMatches.length) {
			this._state.changeState({
				matchesCount: 0,
				matchesPosition: 0
			});
			return { results: [] };
		}
		this._moveToInitialMatch(findModels);
		this._state.changeState({ matchesCount: newMatches.length });
		return { results: newMatches };
	}
	/** Terminate the current searching session, when searching string is empty. */
	_stopSearching() {
		var _this$_currentSearchi;
		this._providers.forEach((provider) => provider.terminate());
		this._findModels = [];
		this._matches = [];
		this._matchingModel = null;
		(_this$_currentSearchi = this._currentSearchingDisposables) === null || _this$_currentSearchi === void 0 || _this$_currentSearchi.dispose();
		this._currentSearchingDisposables = null;
		this.currentMatch$.next(null);
		this.replaceables$.next([]);
		this._state.changeState({
			findCompleted: false,
			matchesCount: 0,
			matchesPosition: 0
		});
	}
	_subscribeToModelsChanges(models) {
		const disposables = this._currentSearchingDisposables = new DisposableCollection();
		const matchesUpdateSubscription = combineLatest(models.map((model) => model.matchesUpdate$)).pipe(debounceTime(220)).subscribe(([ ...allMatches]) => {
			const newMatches = this._matches = allMatches.flat();
			if (newMatches.length) {
				this._moveToInitialMatch(this._findModels, true);
				this._state.changeState({ matchesCount: newMatches.length });
				this.replaceables$.next(newMatches.filter((m) => m.replaceable));
			} else {
				this._state.changeState({
					matchesCount: 0,
					matchesPosition: 0
				});
				this.replaceables$.next([]);
			}
		});
		models.forEach((model) => disposables.add(toDisposable(model.activelyChangingMatch$.subscribe((match) => {
			const index = this._matches.findIndex((m) => m === match);
			this._state.changeState({ matchesPosition: index + 1 });
		}))));
		disposables.add(toDisposable(matchesUpdateSubscription));
	}
	async replace() {
		if (!this._matchingModel) return false;
		return this._matchingModel.replace(this._state.replaceString);
	}
	async replaceAll() {
		const result = await Promise.all(this._findModels.map((m) => m.replaceAll(this._state.replaceString))).then((results) => results.reduce((acc, cur) => {
			acc.success += cur.success;
			acc.failure += cur.failure;
			return acc;
		}, {
			success: 0,
			failure: 0
		}));
		if (result.failure === 0) this._stopSearching();
		return result;
	}
	getCurrentMatch() {
		return this._state.matchesPosition > 0 ? this._matches[this._state.matchesPosition - 1] : null;
	}
	_markMatch(match) {
		const index = this._matches.findIndex((value) => value === match);
		this.currentMatch$.next(match);
		this._state.changeState({ matchesPosition: index + 1 });
	}
	moveToNextMatch() {
		if (!this._matchingModel) return;
		const loopInCurrentUnit = this._findModels.length === 1;
		const nextMatch = this._matchingModel.moveToNextMatch({ loop: loopInCurrentUnit });
		if (nextMatch) {
			this._markMatch(nextMatch);
			return nextMatch;
		} else {
			const currentModelIndex = this._findModels.findIndex((m) => m === this._matchingModel);
			return this._moveToNextUnitMatch(currentModelIndex);
		}
	}
	_moveToNextUnitMatch(startingIndex) {
		const l = this._findModels.length;
		for (let i = (startingIndex + 1) % l; i !== startingIndex;) {
			const nextPositionModel = this._findModels[i];
			const nextMatch = nextPositionModel.moveToNextMatch({ ignoreSelection: true });
			if (nextMatch) {
				this._matchingModel = nextPositionModel;
				this._markMatch(nextMatch);
				return nextMatch;
			}
			i = (i + 1) % l;
		}
		if (this._matchingModel) {
			const nextMatch = this._matchingModel.moveToNextMatch({ ignoreSelection: true });
			if (nextMatch) this._markMatch(nextMatch);
			return nextMatch;
		}
	}
	moveToPreviousMatch() {
		if (!this._matchingModel) return;
		const loopInCurrentUnit = this._findModels.length === 1;
		const nextMatch = this._matchingModel.moveToPreviousMatch({ loop: loopInCurrentUnit });
		if (nextMatch) {
			const index = this._matches.findIndex((value) => value === nextMatch);
			this.currentMatch$.next(nextMatch);
			this._state.changeState({ matchesPosition: index + 1 });
			return nextMatch;
		} else {
			const l = this._findModels.length;
			const currentModelIndex = this._findModels.findIndex((m) => m === this._matchingModel);
			for (let i = (currentModelIndex - 1 + l) % l; i !== currentModelIndex;) {
				const nextPositionModel = this._findModels[i];
				const nextMatch = nextPositionModel.moveToPreviousMatch({ ignoreSelection: true });
				if (nextMatch) {
					this._matchingModel = nextPositionModel;
					this._markMatch(nextMatch);
					return nextMatch;
				}
				i = (i - 1) % l;
			}
			const nextMatch = this._matchingModel.moveToPreviousMatch({ ignoreSelection: true });
			if (nextMatch) this._markMatch(nextMatch);
			return nextMatch;
		}
	}
	_moveToInitialMatch(findModels, noFocus = false) {
		var _this$_univerInstance;
		const focusedUnitId = (_this$_univerInstance = this._univerInstanceService.getFocusedUnit()) === null || _this$_univerInstance === void 0 ? void 0 : _this$_univerInstance.getUnitId();
		if (!focusedUnitId) return -1;
		const i = findModels.findIndex((model) => model.unitId === focusedUnitId);
		if (i !== -1) {
			this._matchingModel = findModels[i];
			const nextMatch = this._matchingModel.moveToNextMatch({
				stayIfOnMatch: true,
				noFocus
			});
			if (nextMatch) {
				this._markMatch(nextMatch);
				return i;
			}
		}
		this._moveToNextUnitMatch(i);
		return 0;
	}
};
FindReplaceModel = __decorate([__decorateParam(2, IUniverInstanceService), __decorateParam(3, ICommandService)], FindReplaceModel);
let FindDirection = /* @__PURE__ */ function(FindDirection) {
	/** Default. */
	FindDirection["ROW"] = "row";
	FindDirection["COLUMN"] = "column";
	return FindDirection;
}({});
let FindBy = /* @__PURE__ */ function(FindBy) {
	FindBy["VALUE"] = "value";
	FindBy["FORMULA"] = "formula";
	return FindBy;
}({});
let FindScope = /* @__PURE__ */ function(FindScope) {
	/** Default. */
	FindScope["SUBUNIT"] = "subunit";
	/** Find the scope in the current unit. */
	FindScope["UNIT"] = "unit";
	return FindScope;
}({});
function createInitFindReplaceState() {
	return {
		caseSensitive: false,
		findBy: "value",
		findCompleted: false,
		findDirection: "row",
		findScope: "subunit",
		findString: "",
		inputtingFindString: "",
		matchesCount: 0,
		matchesPosition: 0,
		matchesTheWholeCell: false,
		replaceRevealed: false,
		replaceString: "",
		revealed: true
	};
}
/**
* This class stores find replace options state. These state are stored
* here instead of the React component so we can change state from
* operations.
*/
var FindReplaceState = class {
	constructor() {
		_defineProperty(this, "_stateUpdates$", new Subject());
		_defineProperty(this, "stateUpdates$", this._stateUpdates$.asObservable());
		_defineProperty(this, "_state$", new BehaviorSubject(createInitFindReplaceState()));
		_defineProperty(this, "state$", this._state$.asObservable());
		_defineProperty(this, "_findString", "");
		_defineProperty(this, "_inputtingFindString", "");
		_defineProperty(this, "_replaceString", "");
		_defineProperty(this, "_revealed", false);
		_defineProperty(this, "_replaceRevealed", false);
		_defineProperty(this, "_matchesPosition", 0);
		_defineProperty(this, "_matchesCount", 0);
		_defineProperty(this, "_caseSensitive", true);
		_defineProperty(this, "_matchesTheWholeCell", false);
		_defineProperty(this, "_findDirection", "row");
		_defineProperty(this, "_findScope", "subunit");
		_defineProperty(this, "_findBy", "value");
		_defineProperty(this, "_findCompleted", false);
	}
	get state() {
		return this._state$.getValue();
	}
	get inputtingFindString() {
		return this._inputtingFindString;
	}
	get findString() {
		return this._findString;
	}
	get revealed() {
		return this._revealed;
	}
	get replaceRevealed() {
		return this._replaceRevealed;
	}
	get matchesPosition() {
		return this._matchesPosition;
	}
	get matchesCount() {
		return this._matchesCount;
	}
	get replaceString() {
		return this._replaceString;
	}
	get caseSensitive() {
		return this._caseSensitive;
	}
	get matchesTheWholeCell() {
		return this._matchesTheWholeCell;
	}
	get findDirection() {
		return this._findDirection;
	}
	get findScope() {
		return this._findScope;
	}
	get findBy() {
		return this._findBy;
	}
	get findCompleted() {
		return this._findCompleted;
	}
	changeState(changes) {
		let changed = false;
		const changedState = {};
		if (typeof changes.findString !== "undefined" && changes.findString !== this._findString) {
			this._findString = changes.findString;
			changedState.findString = this._findString;
			changed = true;
		}
		if (typeof changes.revealed !== "undefined" && changes.revealed !== this._revealed) {
			this._revealed = changes.revealed;
			changedState.revealed = changes.revealed;
			changed = true;
		}
		if (typeof changes.replaceRevealed !== "undefined" && changes.replaceRevealed !== this._replaceRevealed) {
			this._replaceRevealed = changes.replaceRevealed;
			changedState.replaceRevealed = changes.replaceRevealed;
			changed = true;
		}
		if (typeof changes.replaceString !== "undefined" && changes.replaceString !== this._replaceString) {
			this._replaceString = changes.replaceString;
			changedState.replaceString = changes.replaceString;
			changed = true;
		}
		if (typeof changes.matchesCount !== "undefined" && changes.matchesCount !== this._matchesCount) {
			this._matchesCount = changes.matchesCount;
			changedState.matchesCount = changes.matchesCount;
			changed = true;
		}
		if (typeof changes.matchesPosition !== "undefined" && changes.matchesPosition !== this._matchesPosition) {
			this._matchesPosition = changes.matchesPosition;
			changedState.matchesPosition = changes.matchesPosition;
			changed = true;
		}
		if (typeof changes.findBy !== "undefined" && changes.findBy !== this._findBy) {
			this._findBy = changes.findBy;
			changedState.findBy = changes.findBy;
			changed = true;
		}
		if (typeof changes.findScope !== "undefined" && changes.findScope !== this._findScope) {
			this._findScope = changes.findScope;
			changedState.findScope = changes.findScope;
			changed = true;
		}
		if (typeof changes.findDirection !== "undefined" && changes.findDirection !== this._findDirection) {
			this._findDirection = changes.findDirection;
			changedState.findDirection = changes.findDirection;
			changed = true;
		}
		if (typeof changes.caseSensitive !== "undefined" && changes.caseSensitive !== this._caseSensitive) {
			this._caseSensitive = changes.caseSensitive;
			changedState.caseSensitive = changes.caseSensitive;
			changed = true;
		}
		if (typeof changes.matchesTheWholeCell !== "undefined" && changes.matchesTheWholeCell !== this._matchesTheWholeCell) {
			this._matchesTheWholeCell = changes.matchesTheWholeCell;
			changedState.matchesTheWholeCell = changes.matchesTheWholeCell;
			changed = true;
		}
		if (typeof changes.inputtingFindString !== "undefined" && changes.inputtingFindString !== this._inputtingFindString) {
			this._inputtingFindString = changes.inputtingFindString;
			changedState.inputtingFindString = changes.inputtingFindString;
			changed = true;
		}
		if (typeof changes.findCompleted !== "undefined" && changes.findCompleted !== this._findCompleted) {
			this._findCompleted = changes.findCompleted;
			changedState.findCompleted = changes.findCompleted;
			changed = true;
		}
		if (changed) {
			this._state$.next({
				caseSensitive: this._caseSensitive,
				findBy: this._findBy,
				findCompleted: this._findCompleted,
				findDirection: this._findDirection,
				findScope: this._findScope,
				findString: this._findString,
				inputtingFindString: this._inputtingFindString,
				matchesCount: this._matchesCount,
				matchesPosition: this._matchesPosition,
				matchesTheWholeCell: this._matchesTheWholeCell,
				replaceRevealed: this._replaceRevealed,
				revealed: this._revealed
			});
			this._stateUpdates$.next(changedState);
		}
	}
};
let FindReplaceService = class FindReplaceService extends Disposable {
	get stateUpdates$() {
		return this._state.stateUpdates$;
	}
	get state$() {
		return this._state.state$;
	}
	get revealed() {
		return this._state.revealed;
	}
	get replaceRevealed() {
		return this._state.replaceRevealed;
	}
	constructor(_injector, _contextService) {
		super();
		this._injector = _injector;
		this._contextService = _contextService;
		_defineProperty(this, "_providers", /* @__PURE__ */ new Set());
		_defineProperty(this, "_state", new FindReplaceState());
		_defineProperty(this, "_model", void 0);
		_defineProperty(this, "_modelDisposables", null);
		_defineProperty(this, "_currentMatch$", new BehaviorSubject(null));
		_defineProperty(this, "currentMatch$", this._currentMatch$.asObservable());
		_defineProperty(this, "_replaceables$", new BehaviorSubject([]));
		_defineProperty(this, "replaceables$", this._replaceables$.asObservable());
		_defineProperty(this, "_focusSignal$", new Subject());
		_defineProperty(this, "focusSignal$", this._focusSignal$.asObservable());
	}
	dispose() {
		super.dispose();
		this._currentMatch$.next(null);
		this._currentMatch$.complete();
		this._replaceables$.next([]);
		this._replaceables$.complete();
		this._focusSignal$.complete();
	}
	getProviders() {
		return this._providers;
	}
	getCurrentMatch() {
		var _this$_model;
		return (_this$_model = this._model) === null || _this$_model === void 0 ? void 0 : _this$_model.getCurrentMatch();
	}
	getFindString() {
		return this._state.findString;
	}
	changeFindString(findString) {
		this._state.changeState({ findString });
	}
	focusFindInput() {
		this._focusSignal$.next();
	}
	changeInputtingFindString(value) {
		if (value) this._state.changeState({ inputtingFindString: value });
		else this._state.changeState({
			inputtingFindString: "",
			findString: ""
		});
	}
	changeReplaceString(replaceString) {
		this._state.changeState({ replaceString });
	}
	changeMatchesTheWholeCell(matchesTheWholeCell) {
		this._state.changeState({ matchesTheWholeCell });
	}
	changeCaseSensitive(caseSensitive) {
		this._state.changeState({ caseSensitive });
	}
	changeFindBy(findBy) {
		this._state.changeState({ findBy });
		this._toggleDisplayRawFormula(findBy === "formula");
	}
	changeFindScope(scope) {
		this._state.changeState({ findScope: scope });
	}
	changeFindDirection(direction) {
		this._state.changeState({ findDirection: direction });
	}
	moveToNextMatch() {
		if (!this._model) return;
		if (this._state.replaceRevealed && !this._model.searched) {
			this._state.changeState({ findString: this._state.inputtingFindString });
			this._model.start();
		} else this._model.moveToNextMatch();
		this._focusSignal$.next();
	}
	moveToPreviousMatch() {
		if (!this._model) return;
		if (this._state.replaceRevealed && !this._model.searched) {
			this._state.changeState({ findString: this._state.inputtingFindString });
			this._model.start();
		} else this._model.moveToPreviousMatch();
		this._focusSignal$.next();
	}
	async replace() {
		if (!this._model) return false;
		return this._model.replace();
	}
	async replaceAll() {
		if (!this._model) throw new Error("[FindReplaceService] replaceAll: model is not initialized!");
		return this._model.replaceAll();
	}
	revealReplace() {
		this._state.changeState({
			replaceRevealed: true,
			inputtingFindString: this._state.findString
		});
		this._toggleRevealReplace(true);
	}
	focusSelection() {
		var _this$_model2;
		(_this$_model2 = this._model) === null || _this$_model2 === void 0 || _this$_model2.focusSelection();
	}
	start(revealReplace = false) {
		if (this._providers.size === 0) return false;
		this._model = this._injector.createInstance(FindReplaceModel, this._state, this._providers);
		this._modelDisposables = new DisposableCollection();
		this._modelDisposables.add(toDisposable(this._model.currentMatch$.subscribe((match) => this._currentMatch$.next(match))));
		this._modelDisposables.add(toDisposable(this._model.replaceables$.subscribe((replaceables) => this._replaceables$.next(replaceables))));
		const newState = createInitFindReplaceState();
		if (revealReplace) newState.replaceRevealed = true;
		this._state.changeState(newState);
		this._toggleRevealReplace(revealReplace);
		return true;
	}
	find() {
		var _this$_model3;
		(_this$_model3 = this._model) === null || _this$_model3 === void 0 || _this$_model3.start();
	}
	terminate() {
		var _this$_model4, _this$_modelDisposabl;
		(_this$_model4 = this._model) === null || _this$_model4 === void 0 || _this$_model4.dispose();
		this._model = null;
		(_this$_modelDisposabl = this._modelDisposables) === null || _this$_modelDisposabl === void 0 || _this$_modelDisposabl.dispose();
		this._modelDisposables = null;
		this._toggleDisplayRawFormula(false);
		this._toggleRevealReplace(false);
	}
	registerFindReplaceProvider(provider) {
		this._providers.add(provider);
		return toDisposable(() => this._providers.delete(provider));
	}
	_toggleRevealReplace(revealReplace) {
		this._contextService.setContextValue(FIND_REPLACE_REPLACE_REVEALED, revealReplace);
	}
	_toggleDisplayRawFormula(force) {
		this._contextService.setContextValue(RENDER_RAW_FORMULA_KEY, force);
	}
};
FindReplaceService = __decorate([__decorateParam(0, Inject(Injector)), __decorateParam(1, IContextService)], FindReplaceService);

//#endregion
//#region src/commands/commands/replace.command.ts
const ReplaceCurrentMatchCommand = {
	id: "ui.command.replace-current-match",
	type: CommandType.COMMAND,
	handler: (accessor) => {
		return accessor.get(IFindReplaceService).replace();
	}
};
const CONFIRM_REPLACE_ALL_ID = "CONFIRM_REPLACE_ALL";
const ReplaceAllMatchesCommand = {
	id: "ui.command.replace-all-matches",
	type: CommandType.COMMAND,
	handler: async (accessor) => {
		const confirmService = accessor.get(IConfirmService);
		const localeService = accessor.get(LocaleService);
		const messageService = accessor.get(IMessageService);
		if (!await confirmService.confirm({
			id: CONFIRM_REPLACE_ALL_ID,
			title: { title: localeService.t("find-replace.replace.confirm.title") },
			cancelText: localeService.t("find-replace.button.cancel"),
			confirmText: localeService.t("find-replace.button.confirm")
		})) return false;
		const { success, failure } = await accessor.get(IFindReplaceService).replaceAll();
		if (failure > 0) {
			if (success === 0) messageService.show({
				type: MessageType.Error,
				content: localeService.t("find-replace.replace.all-failure")
			});
			else messageService.show({
				type: MessageType.Warning,
				content: localeService.t("find-replace.replace.partial-success", `${success}`, `${failure}`)
			});
			return false;
		}
		messageService.show({
			type: MessageType.Success,
			content: localeService.t("find-replace.replace.all-success", `${success}`)
		});
		return true;
	}
};

//#endregion
//#region src/commands/operations/find-replace.operation.ts
const OpenFindDialogOperation = {
	id: "ui.operation.open-find-dialog",
	type: CommandType.OPERATION,
	handler: (accessor) => {
		const findReplaceService = accessor.get(IFindReplaceService);
		if (!findReplaceService.revealed) findReplaceService.start();
		else findReplaceService.focusFindInput();
		return true;
	}
};
const OpenReplaceDialogOperation = {
	id: "ui.operation.open-replace-dialog",
	type: CommandType.OPERATION,
	handler: (accessor) => {
		const findReplaceService = accessor.get(IFindReplaceService);
		if (!findReplaceService.revealed) findReplaceService.start(true);
		else if (!findReplaceService.replaceRevealed) findReplaceService.revealReplace();
		else findReplaceService.focusFindInput();
		return true;
	}
};
const GoToNextMatchOperation = {
	type: CommandType.OPERATION,
	id: "ui.operation.go-to-next-match",
	handler: (accessor) => {
		accessor.get(IFindReplaceService).moveToNextMatch();
		return true;
	}
};
const GoToPreviousMatchOperation = {
	type: CommandType.OPERATION,
	id: "ui.operation.go-to-previous-match",
	handler: (accessor) => {
		accessor.get(IFindReplaceService).moveToPreviousMatch();
		return true;
	}
};
const FocusSelectionOperation = {
	type: CommandType.OPERATION,
	id: "ui.operation.focus-selection",
	handler: (accessor) => {
		accessor.get(IFindReplaceService).focusSelection();
		return true;
	}
};

//#endregion
//#region src/menu/find-replace.menu.ts
function FindReplaceMenuItemFactory(accessor) {
	const contextService = accessor.get(IContextService);
	return {
		id: OpenFindDialogOperation.id,
		icon: "SearchIcon",
		tooltip: "find-replace.toolbar",
		type: MenuItemType.BUTTON,
		hidden$: getMenuHiddenObservable(accessor, UniverInstanceType.UNIVER_SHEET),
		disabled$: combineLatest([contextService.subscribeContextValue$(EDITOR_ACTIVATED), contextService.subscribeContextValue$(FOCUSING_SHEET)]).pipe(map(([editorActivated, focusingSheet]) => editorActivated || !focusingSheet))
	};
}

//#endregion
//#region src/menu/schema.ts
const menuSchema = { [RibbonDataGroup.ORGANIZATION]: { [OpenFindDialogOperation.id]: {
	order: 2,
	menuItemFactory: FindReplaceMenuItemFactory
} } };

//#endregion
//#region src/views/dialog/SearchInput.tsx
function SearchInput(props) {
	const { findCompleted: findComplete, localeService, matchesCount, matchesPosition, initialFindString, findReplaceService, onChange, ...rest } = props;
	const [value, setValue] = useState(initialFindString);
	const text = findComplete && matchesCount === 0 ? localeService.t("find-replace.dialog.no-result") : matchesCount === 0 ? " " : void 0;
	function handleChangePosition(newIndex) {
		if (matchesPosition === matchesCount && newIndex === 1) findReplaceService.moveToNextMatch();
		else if (matchesPosition === 1 && newIndex === matchesCount) findReplaceService.moveToPreviousMatch();
		else if (newIndex < matchesPosition) findReplaceService.moveToPreviousMatch();
		else findReplaceService.moveToNextMatch();
	}
	return /* @__PURE__ */ jsx("div", {
		className: "univer-relative univer-flex univer-items-center univer-gap-2",
		onDrag: (e) => e.stopPropagation(),
		children: /* @__PURE__ */ jsx(Input, {
			"data-u-comp": "search-input",
			autoFocus: true,
			placeholder: localeService.t("find-replace.dialog.find-placeholder"),
			value,
			onChange: (value) => {
				setValue(value);
				onChange === null || onChange === void 0 || onChange(value);
			},
			slot: /* @__PURE__ */ jsx(Pager, {
				loop: true,
				text,
				value: matchesPosition,
				total: matchesCount,
				onChange: handleChangePosition
			}),
			...rest
		})
	});
}

//#endregion
//#region src/views/dialog/FindReplaceDialog.tsx
function useFindInputFocus(findReplaceService, ref) {
	const focus = useCallback(() => {
		var _document$querySelect;
		(_document$querySelect = document.querySelector(".univer-find-input input")) === null || _document$querySelect === void 0 || _document$querySelect.focus();
	}, []);
	const selectHasFocus = useCallback(() => {
		const allInputs = document.querySelectorAll("[data-u-comp=find-replace-dialog] [data-u-comp=search-input]");
		return Array.from(allInputs).some((input) => input === document.activeElement);
	}, []);
	useImperativeHandle(ref, () => ({
		focus,
		selectHasFocus
	}));
	useEffect(() => {
		const subscription = findReplaceService.focusSignal$.subscribe(() => focus());
		return () => subscription.unsubscribe();
	}, [findReplaceService, focus]);
	return {
		focus,
		selectHasFocus
	};
}
const FindDialog = forwardRef(function FindDialogImpl(_props, ref) {
	const localeService = useDependency(LocaleService);
	const findReplaceService = useDependency(IFindReplaceService);
	const commandService = useDependency(ICommandService);
	const { findCompleted, findString, matchesCount, matchesPosition } = useObservable(findReplaceService.state$, void 0, true);
	const revealReplace = useCallback(() => {
		commandService.executeCommand(OpenReplaceDialogOperation.id);
	}, [commandService]);
	const onFindStringChange = useDebounceFn((findString) => {
		return findReplaceService.changeFindString(findString);
	}, 500);
	useFindInputFocus(findReplaceService, ref);
	return /* @__PURE__ */ jsxs(Fragment, { children: [/* @__PURE__ */ jsx(SearchInput, {
		findCompleted,
		matchesCount,
		matchesPosition,
		findReplaceService,
		localeService,
		initialFindString: findString,
		onChange: onFindStringChange
	}), /* @__PURE__ */ jsx("div", {
		className: "univer-mt-4 univer-text-center",
		children: /* @__PURE__ */ jsx("a", {
			className: "hover:univer-text-primary-500/80 univer-cursor-pointer univer-text-sm univer-text-primary-500 univer-transition-colors",
			onClick: revealReplace,
			children: localeService.t("find-replace.dialog.advanced-finding")
		})
	})] });
});
const ReplaceDialog = forwardRef(function ReplaceDialogImpl(_props, ref) {
	const findReplaceService = useDependency(IFindReplaceService);
	const localeService = useDependency(LocaleService);
	const commandService = useDependency(ICommandService);
	const messageService = useDependency(IMessageService);
	const currentMatch = useObservable(findReplaceService.currentMatch$, void 0, true);
	const replaceables = useObservable(findReplaceService.replaceables$, void 0, true);
	const { matchesCount, matchesPosition, findString, inputtingFindString, replaceString, caseSensitive, matchesTheWholeCell, findDirection, findScope, findBy, findCompleted } = useObservable(findReplaceService.state$, void 0, true);
	const findDisabled = inputtingFindString.length === 0;
	const replaceDisabled = matchesCount === 0 || !(currentMatch === null || currentMatch === void 0 ? void 0 : currentMatch.replaceable);
	const replaceAllDisabled = replaceables.length === 0;
	const onFindStringChange = useCallback((newValue) => findReplaceService.changeInputtingFindString(newValue), [findReplaceService]);
	const onReplaceStringChange = useCallback((replaceString) => findReplaceService.changeReplaceString(replaceString), [findReplaceService]);
	const { focus } = useFindInputFocus(findReplaceService, ref);
	const onClickFindButton = useCallback(() => {
		if (findString === inputtingFindString) findReplaceService.moveToNextMatch();
		else {
			findReplaceService.changeFindString(inputtingFindString);
			findReplaceService.find();
		}
	}, [
		findString,
		inputtingFindString,
		findReplaceService
	]);
	const onClickReplaceButton = useCallback(() => commandService.executeCommand(ReplaceCurrentMatchCommand.id), [commandService]);
	const onClickReplaceAllButton = useCallback(async () => {
		await commandService.executeCommand(ReplaceAllMatchesCommand.id);
		focus();
	}, [commandService]);
	const onChangeFindDirection = useCallback((findDirection) => {
		findReplaceService.changeFindDirection(findDirection);
	}, [findReplaceService]);
	const onChangeFindScope = useCallback((findScope) => {
		findReplaceService.changeFindScope(findScope);
	}, [findReplaceService]);
	const onChangeFindBy = useCallback((findBy) => {
		findReplaceService.changeFindBy(findBy);
	}, [findReplaceService]);
	const findScopeOptions = useFindScopeOptions(localeService);
	const findDirectionOptions = useFindDirectionOptions(localeService);
	const findByOptions = useFindByOptions(localeService);
	useEffect(() => {
		if (findCompleted && matchesCount === 0) messageService.show({
			content: localeService.t("find-replace.dialog.no-match"),
			type: MessageType.Warning,
			duration: 5e3
		});
	}, [
		findCompleted,
		matchesCount,
		messageService,
		localeService
	]);
	return /* @__PURE__ */ jsxs("div", { children: [
		/* @__PURE__ */ jsx(FormLayout, {
			label: localeService.t("find-replace.dialog.find"),
			children: /* @__PURE__ */ jsx(SearchInput, {
				findCompleted,
				className: "univer-find-input",
				matchesCount,
				matchesPosition,
				findReplaceService,
				localeService,
				initialFindString: inputtingFindString,
				onChange: onFindStringChange
			})
		}),
		/* @__PURE__ */ jsx(FormLayout, {
			label: localeService.t("find-replace.dialog.replace"),
			children: /* @__PURE__ */ jsx(Input, {
				placeholder: localeService.t("find-replace.dialog.replace-placeholder"),
				value: replaceString,
				onChange: (value) => onReplaceStringChange(value)
			})
		}),
		/* @__PURE__ */ jsx(FormLayout, {
			label: localeService.t("find-replace.dialog.find-direction.title"),
			children: /* @__PURE__ */ jsx(Select, {
				value: findDirection,
				options: findDirectionOptions,
				onChange: onChangeFindDirection
			})
		}),
		/* @__PURE__ */ jsx(FormDualColumnLayout, { children: /* @__PURE__ */ jsxs(Fragment, { children: [/* @__PURE__ */ jsx(FormLayout, {
			label: localeService.t("find-replace.dialog.find-scope.title"),
			children: /* @__PURE__ */ jsx(Select, {
				value: findScope,
				options: findScopeOptions,
				onChange: onChangeFindScope
			})
		}), /* @__PURE__ */ jsx(FormLayout, {
			label: localeService.t("find-replace.dialog.find-by.title"),
			children: /* @__PURE__ */ jsx(Select, {
				value: findBy,
				options: findByOptions,
				onChange: onChangeFindBy
			})
		})] }) }),
		/* @__PURE__ */ jsx(FormDualColumnLayout, { children: /* @__PURE__ */ jsxs(Fragment, { children: [/* @__PURE__ */ jsx(FormLayout, { children: /* @__PURE__ */ jsx(Checkbox, {
			checked: caseSensitive,
			onChange: (checked) => {
				findReplaceService.changeCaseSensitive(checked);
			},
			children: localeService.t("find-replace.dialog.case-sensitive")
		}) }), /* @__PURE__ */ jsx(FormLayout, { children: /* @__PURE__ */ jsx(Checkbox, {
			checked: matchesTheWholeCell,
			onChange: (checked) => {
				findReplaceService.changeMatchesTheWholeCell(checked);
			},
			children: localeService.t("find-replace.dialog.match-the-whole-cell")
		}) })] }) }),
		/* @__PURE__ */ jsxs("div", {
			className: "univer-mt-6 univer-flex univer-justify-between",
			children: [/* @__PURE__ */ jsx(Button, {
				variant: "primary",
				onClick: onClickFindButton,
				disabled: findDisabled,
				children: localeService.t("find-replace.dialog.find")
			}), /* @__PURE__ */ jsxs("span", {
				className: "univer-inline-flex univer-gap-2",
				children: [/* @__PURE__ */ jsx(Button, {
					disabled: replaceDisabled,
					onClick: onClickReplaceButton,
					children: localeService.t("find-replace.dialog.replace")
				}), /* @__PURE__ */ jsx(Button, {
					disabled: replaceAllDisabled,
					onClick: onClickReplaceAllButton,
					children: localeService.t("find-replace.dialog.replace-all")
				})]
			})]
		})
	] });
});
function FindReplaceDialog() {
	const findReplaceService = useDependency(IFindReplaceService);
	const layoutService = useDependency(ILayoutService);
	const contextService = useDependency(IContextService);
	const state = useObservable(findReplaceService.state$, void 0, true);
	const dialogContainerRef = useRef(null);
	useEffect(() => {
		let disposable;
		if (dialogContainerRef.current) disposable = layoutService.registerContainerElement(dialogContainerRef.current);
		return () => disposable === null || disposable === void 0 ? void 0 : disposable.dispose();
	}, [layoutService]);
	const focusRef = useRef(null);
	const setDialogContainerFocus = useCallback((focused) => contextService.setContextValue(FIND_REPLACE_DIALOG_FOCUS, focused), [contextService]);
	const setDialogInputFocus = useCallback((focused) => contextService.setContextValue(FIND_REPLACE_INPUT_FOCUS, focused), [contextService]);
	useEffect(() => {
		var _focusRef$current;
		const focusSubscription = fromEvent(document, "focusin").subscribe((event) => {
			var _dialogContainerRef$c;
			if (event.target && ((_dialogContainerRef$c = dialogContainerRef.current) === null || _dialogContainerRef$c === void 0 ? void 0 : _dialogContainerRef$c.contains(event.target))) setDialogContainerFocus(true);
			else setDialogContainerFocus(false);
			if (!focusRef.current || !focusRef.current.selectHasFocus()) setDialogInputFocus(false);
			else setDialogInputFocus(true);
		});
		(_focusRef$current = focusRef.current) === null || _focusRef$current === void 0 || _focusRef$current.focus();
		setDialogContainerFocus(true);
		setDialogInputFocus(true);
		return () => {
			focusSubscription.unsubscribe();
			setDialogContainerFocus(false);
		};
	}, [setDialogContainerFocus, setDialogInputFocus]);
	return /* @__PURE__ */ jsx("div", {
		ref: dialogContainerRef,
		"data-u-comp": "find-replace-dialog",
		children: !state.replaceRevealed ? /* @__PURE__ */ jsx(FindDialog, { ref: focusRef }) : /* @__PURE__ */ jsx(ReplaceDialog, { ref: focusRef })
	});
}
function useFindScopeOptions(localeService) {
	return useMemo(() => {
		return [{
			label: localeService.t("find-replace.dialog.find-scope.current-sheet"),
			value: "subunit"
		}, {
			label: localeService.t("find-replace.dialog.find-scope.workbook"),
			value: "unit"
		}];
	}, [localeService.getCurrentLocale()]);
}
function useFindDirectionOptions(localeService) {
	return useMemo(() => {
		return [{
			label: localeService.t("find-replace.dialog.find-direction.row"),
			value: "row"
		}, {
			label: localeService.t("find-replace.dialog.find-direction.column"),
			value: "column"
		}];
	}, [localeService.getCurrentLocale()]);
}
function useFindByOptions(localeService) {
	return useMemo(() => {
		return [{
			label: localeService.t("find-replace.dialog.find-by.value"),
			value: "value"
		}, {
			label: localeService.t("find-replace.dialog.find-by.formula"),
			value: "formula"
		}];
	}, [localeService.getCurrentLocale()]);
}

//#endregion
//#region src/controllers/find-replace.shortcut.ts
function whenFindReplaceDialogFocused(contextService) {
	return contextService.getContextValue(FIND_REPLACE_DIALOG_FOCUS);
}
function whenReplaceRevealed(contextService) {
	return contextService.getContextValue(FIND_REPLACE_REPLACE_REVEALED);
}
function whenFindReplaceInputFocused(contextService) {
	return contextService.getContextValue(FIND_REPLACE_INPUT_FOCUS);
}
const FIND_REPLACE_SHORTCUT_GROUP = "7_find-replace-shortcuts";
function whenSheetFocused(contextService) {
	return contextService.getContextValue(FOCUSING_SHEET);
}
function whenEditorNotActivated(contextService) {
	return !contextService.getContextValue(EDITOR_ACTIVATED);
}
const OpenFindDialogShortcutItem = {
	id: OpenFindDialogOperation.id,
	description: "find-replace.shortcut.open-find-dialog",
	binding: KeyCode.F | MetaKeys.CTRL_COMMAND,
	group: FIND_REPLACE_SHORTCUT_GROUP,
	groupTitle: "find-replace.shortcut.panel",
	preconditions(contextService) {
		return !whenFindReplaceDialogFocused(contextService) && whenSheetFocused(contextService) && whenEditorNotActivated(contextService);
	}
};
const MacOpenFindDialogShortcutItem = {
	id: OpenFindDialogOperation.id,
	description: "find-replace.shortcut.open-find-dialog",
	binding: KeyCode.F | MetaKeys.CTRL_COMMAND,
	mac: KeyCode.F | MetaKeys.MAC_CTRL,
	preconditions(contextService) {
		return !whenFindReplaceDialogFocused(contextService) && whenSheetFocused(contextService) && whenEditorNotActivated(contextService);
	}
};
const OpenReplaceDialogShortcutItem = {
	id: OpenReplaceDialogOperation.id,
	description: "find-replace.shortcut.open-replace-dialog",
	binding: KeyCode.H | MetaKeys.CTRL_COMMAND,
	mac: KeyCode.H | MetaKeys.MAC_CTRL,
	group: FIND_REPLACE_SHORTCUT_GROUP,
	groupTitle: "find-replace.shortcut.panel",
	preconditions(contextService) {
		return whenSheetFocused(contextService) && whenEditorNotActivated(contextService) && (!whenFindReplaceDialogFocused(contextService) || !whenReplaceRevealed(contextService));
	}
};
const GoToNextFindMatchShortcutItem = {
	id: GoToNextMatchOperation.id,
	description: "find-replace.shortcut.go-to-next-match",
	binding: KeyCode.ENTER,
	group: FIND_REPLACE_SHORTCUT_GROUP,
	groupTitle: "find-replace.shortcut.panel",
	priority: 1e3,
	preconditions(contextService) {
		return whenFindReplaceInputFocused(contextService) && whenFindReplaceDialogFocused(contextService);
	}
};
const GoToPreviousFindMatchShortcutItem = {
	id: GoToPreviousMatchOperation.id,
	description: "find-replace.shortcut.go-to-previous-match",
	binding: KeyCode.ENTER | MetaKeys.SHIFT,
	group: FIND_REPLACE_SHORTCUT_GROUP,
	groupTitle: "find-replace.shortcut.panel",
	priority: 1e3,
	preconditions(contextService) {
		return whenFindReplaceInputFocused(contextService) && whenFindReplaceDialogFocused(contextService);
	}
};
const FocusSelectionShortcutItem = {
	id: FocusSelectionOperation.id,
	description: "find-replace.shortcut.focus-selection",
	binding: KeyCode.ESC,
	group: FIND_REPLACE_SHORTCUT_GROUP,
	groupTitle: "find-replace.shortcut.panel",
	priority: 1e3,
	preconditions(contextService) {
		return whenFindReplaceDialogFocused(contextService);
	}
};

//#endregion
//#region src/controllers/find-replace.controller.ts
const FIND_REPLACE_DIALOG_ID = "DESKTOP_FIND_REPLACE_DIALOG";
const FIND_REPLACE_PANEL_WIDTH = 350;
const FIND_REPLACE_PANEL_RIGHT_PADDING = 20;
const FIND_REPLACE_PANEL_TOP_PADDING = 64;
let FindReplaceController = class FindReplaceController extends RxDisposable {
	constructor(_univerInstanceService, _menuManagerService, _shortcutService, _commandService, _findReplaceService, _dialogService, _layoutService, _localeService, _componentManager) {
		super();
		this._univerInstanceService = _univerInstanceService;
		this._menuManagerService = _menuManagerService;
		this._shortcutService = _shortcutService;
		this._commandService = _commandService;
		this._findReplaceService = _findReplaceService;
		this._dialogService = _dialogService;
		this._layoutService = _layoutService;
		this._localeService = _localeService;
		this._componentManager = _componentManager;
		_defineProperty(this, "_closingListenerDisposable", void 0);
		this._initCommands();
		this._initUI();
		this._initShortcuts();
	}
	dispose() {
		var _this$_closingListene;
		super.dispose();
		(_this$_closingListene = this._closingListenerDisposable) === null || _this$_closingListene === void 0 || _this$_closingListene.dispose();
		this._closingListenerDisposable = null;
	}
	_initCommands() {
		[
			OpenFindDialogOperation,
			OpenReplaceDialogOperation,
			GoToNextMatchOperation,
			GoToPreviousMatchOperation,
			ReplaceAllMatchesCommand,
			ReplaceCurrentMatchCommand,
			FocusSelectionOperation
		].forEach((c) => {
			this.disposeWithMe(this._commandService.registerCommand(c));
		});
	}
	_initShortcuts() {
		[
			OpenReplaceDialogShortcutItem,
			OpenFindDialogShortcutItem,
			MacOpenFindDialogShortcutItem,
			GoToPreviousFindMatchShortcutItem,
			GoToNextFindMatchShortcutItem,
			FocusSelectionShortcutItem
		].forEach((s) => this.disposeWithMe(this._shortcutService.registerShortcut(s)));
	}
	_initUI() {
		[["FindReplaceDialog", FindReplaceDialog], ["SearchIcon", SearchIcon]].forEach(([key, comp]) => {
			this.disposeWithMe(this._componentManager.register(key, comp));
		});
		this._menuManagerService.mergeMenu(menuSchema);
		this._findReplaceService.stateUpdates$.pipe(takeUntil(this.dispose$)).subscribe((newState) => {
			if (newState.revealed === true) this._openPanel();
		});
	}
	_openPanel() {
		this._dialogService.open({
			id: FIND_REPLACE_DIALOG_ID,
			draggable: true,
			width: FIND_REPLACE_PANEL_WIDTH,
			title: { title: this._localeService.t("find-replace.dialog.title") },
			children: { label: "FindReplaceDialog" },
			destroyOnClose: true,
			mask: false,
			maskClosable: false,
			defaultPosition: getFindReplaceDialogDefaultPosition(),
			preservePositionOnDestroy: true,
			onClose: () => this.closePanel()
		});
		this._closingListenerDisposable = toDisposable(this._univerInstanceService.focused$.pipe(takeUntil(this.dispose$)).subscribe((focused) => {
			if (!focused || !this._univerInstanceService.getUniverSheetInstance(focused)) this.closePanel();
		}));
	}
	closePanel() {
		if (!this._closingListenerDisposable) return;
		this._closingListenerDisposable.dispose();
		this._closingListenerDisposable = null;
		this._dialogService.close(FIND_REPLACE_DIALOG_ID);
		this._findReplaceService.terminate();
		queueMicrotask(() => this._layoutService.focus());
	}
};
FindReplaceController = __decorate([
	__decorateParam(0, IUniverInstanceService),
	__decorateParam(1, IMenuManagerService),
	__decorateParam(2, IShortcutService),
	__decorateParam(3, ICommandService),
	__decorateParam(4, IFindReplaceService),
	__decorateParam(5, IDialogService),
	__decorateParam(6, ILayoutService),
	__decorateParam(7, Inject(LocaleService)),
	__decorateParam(8, Inject(ComponentManager))
], FindReplaceController);
function getFindReplaceDialogDefaultPosition() {
	const { innerWidth } = window;
	return {
		x: innerWidth - FIND_REPLACE_PANEL_WIDTH - FIND_REPLACE_PANEL_RIGHT_PADDING,
		y: FIND_REPLACE_PANEL_TOP_PADDING
	};
}

//#endregion
//#region package.json
var name = "@univerjs/find-replace";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
const FIND_REPLACE_PLUGIN_CONFIG_KEY = "find-replace.config";
const configSymbol = Symbol(FIND_REPLACE_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/plugin.ts
let UniverFindReplacePlugin = class UniverFindReplacePlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _injector, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		const { ...rest } = merge({}, defaultPluginConfig, this._config);
		this._configService.setConfig(FIND_REPLACE_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		[[FindReplaceController], [IFindReplaceService, { useClass: FindReplaceService }]].forEach((d) => this._injector.add(d));
	}
	onRendered() {
		this._injector.get(FindReplaceController);
	}
};
_defineProperty(UniverFindReplacePlugin, "pluginName", "UNIVER_FIND_REPLACE_PLUGIN");
_defineProperty(UniverFindReplacePlugin, "packageName", name);
_defineProperty(UniverFindReplacePlugin, "version", version);
UniverFindReplacePlugin = __decorate([__decorateParam(1, Inject(Injector)), __decorateParam(2, IConfigService)], UniverFindReplacePlugin);

//#endregion
export { FindBy, FindDirection, FindModel, FindReplaceController, FindReplaceModel, FindReplaceState, FindScope, GoToNextMatchOperation, GoToPreviousMatchOperation, IFindReplaceService, OpenFindDialogOperation, OpenReplaceDialogOperation, ReplaceAllMatchesCommand, ReplaceCurrentMatchCommand, UniverFindReplacePlugin, createInitFindReplaceState };
