// @univerjs/sheets-thread-comment/lib/es/index.js
import { DependentOn, Disposable, ICommandService, IUniverInstanceService, Inject, Injector, ObjectMatrix, Plugin, UniverInstanceType, generateRandomId, sequenceExecuteAsync, toDisposable, touchDependencies } from "@univerjs/core";
import { serializeRange, singleReferenceToGrid } from "@univerjs/engine-formula";
import { CopySheetCommand, RefRangeService, RemoveSheetCommand, SheetInterceptorService, SheetsSelectionsService, handleCommonRangeChangeWithEffectRefCommandsSkipNoInterests } from "@univerjs/sheets";
import { AddCommentMutation, DeleteCommentMutation, IThreadCommentDataSourceService, ThreadCommentModel, UniverThreadCommentPlugin, UpdateCommentRefMutation } from "@univerjs/thread-comment";
import { Subject } from "rxjs";

//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/models/sheets-thread-comment.model.ts
let SheetsThreadCommentModel = class SheetsThreadCommentModel extends Disposable {
	constructor(_threadCommentModel, _univerInstanceService) {
		super();
		this._threadCommentModel = _threadCommentModel;
		this._univerInstanceService = _univerInstanceService;
		_defineProperty(this, "_matrixMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_locationMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_commentUpdate$", new Subject());
		_defineProperty(this, "commentUpdate$", this._commentUpdate$.asObservable());
		this._init();
		this.disposeWithMe(() => {
			this._commentUpdate$.complete();
		});
	}
	_init() {
		this._initData();
		this._initUpdateTransform();
	}
	_ensureCommentMatrix(unitId, subUnitId) {
		let unitMap = this._matrixMap.get(unitId);
		if (!unitMap) {
			unitMap = /* @__PURE__ */ new Map();
			this._matrixMap.set(unitId, unitMap);
		}
		let subUnitMap = unitMap.get(subUnitId);
		if (!subUnitMap) {
			subUnitMap = new ObjectMatrix();
			unitMap.set(subUnitId, subUnitMap);
		}
		return subUnitMap;
	}
	_ensureCommentLocationMap(unitId, subUnitId) {
		let unitMap = this._locationMap.get(unitId);
		if (!unitMap) {
			unitMap = /* @__PURE__ */ new Map();
			this._locationMap.set(unitId, unitMap);
		}
		let subUnitMap = unitMap.get(subUnitId);
		if (!subUnitMap) {
			subUnitMap = /* @__PURE__ */ new Map();
			unitMap.set(subUnitId, subUnitMap);
		}
		return subUnitMap;
	}
	_addCommentToMatrix(matrix, row, column, commentId) {
		var _matrix$getValue;
		const current = (_matrix$getValue = matrix.getValue(row, column)) !== null && _matrix$getValue !== void 0 ? _matrix$getValue : /* @__PURE__ */ new Set();
		current.add(commentId);
		matrix.setValue(row, column, current);
	}
	_deleteCommentFromMatrix(matrix, row, column, commentId) {
		if (row >= 0 && column >= 0) {
			const current = matrix.getValue(row, column);
			if (current && current.has(commentId)) {
				current.delete(commentId);
				if (current.size === 0) matrix.realDeleteValue(row, column);
			}
		}
	}
	_ensure(unitId, subUnitId) {
		return {
			matrix: this._ensureCommentMatrix(unitId, subUnitId),
			locationMap: this._ensureCommentLocationMap(unitId, subUnitId)
		};
	}
	_initData() {
		const datas = this._threadCommentModel.getAll();
		for (const data of datas) for (const thread of data.threads) {
			const { unitId, subUnitId, root } = thread;
			this._addComment(unitId, subUnitId, root);
		}
	}
	_addComment(unitId, subUnitId, comment) {
		const location = singleReferenceToGrid(comment.ref);
		const parentId = comment.parentId;
		const { row, column } = location;
		const commentId = comment.id;
		const { matrix, locationMap } = this._ensure(unitId, subUnitId);
		if (!parentId && row >= 0 && column >= 0) {
			this._addCommentToMatrix(matrix, row, column, commentId);
			locationMap.set(commentId, {
				row,
				column
			});
		}
		if (!parentId) this._commentUpdate$.next({
			unitId,
			subUnitId,
			payload: comment,
			type: "add",
			isRoot: !parentId,
			...location
		});
	}
	_initUpdateTransform() {
		this.disposeWithMe(this._threadCommentModel.commentUpdate$.subscribe((update) => {
			const { unitId, subUnitId } = update;
			try {
				if (this._univerInstanceService.getUnitType(unitId) !== UniverInstanceType.UNIVER_SHEET) return;
			} catch (error) {}
			const { matrix, locationMap } = this._ensure(unitId, subUnitId);
			switch (update.type) {
				case "add":
					this._addComment(update.unitId, update.subUnitId, update.payload);
					break;
				case "delete": {
					const { isRoot, comment } = update.payload;
					if (isRoot) {
						const location = singleReferenceToGrid(comment.ref);
						const { row, column } = location;
						this._deleteCommentFromMatrix(matrix, row, column, comment.id);
						this._commentUpdate$.next({
							...update,
							...location
						});
					}
					break;
				}
				case "update": {
					const { commentId } = update.payload;
					const comment = this._threadCommentModel.getComment(unitId, subUnitId, commentId);
					if (!comment) return;
					const location = singleReferenceToGrid(comment.ref);
					this._commentUpdate$.next({
						...update,
						...location
					});
					break;
				}
				case "updateRef": {
					const location = singleReferenceToGrid(update.payload.ref);
					const { commentId } = update.payload;
					const currentLoc = locationMap.get(commentId);
					if (!currentLoc) return;
					const { row, column } = currentLoc;
					this._deleteCommentFromMatrix(matrix, row, column, commentId);
					locationMap.delete(commentId);
					if (location.row >= 0 && location.column >= 0) {
						this._addCommentToMatrix(matrix, location.row, location.column, commentId);
						locationMap.set(commentId, {
							row: location.row,
							column: location.column
						});
					}
					this._commentUpdate$.next({
						...update,
						...location
					});
					break;
				}
				case "resolve": {
					const { unitId, subUnitId, payload } = update;
					const { locationMap } = this._ensure(unitId, subUnitId);
					const location = locationMap.get(payload.commentId);
					if (location) this._commentUpdate$.next({
						...update,
						...location
					});
					break;
				}
				default: break;
			}
		}));
	}
	getByLocation(unitId, subUnitId, row, column) {
		var _activeComments$;
		return (_activeComments$ = this.getAllByLocation(unitId, subUnitId, row, column).filter((comment) => !comment.resolved)[0]) === null || _activeComments$ === void 0 ? void 0 : _activeComments$.id;
	}
	getAllByLocation(unitId, subUnitId, row, column) {
		const current = this._ensureCommentMatrix(unitId, subUnitId).getValue(row, column);
		if (!current) return [];
		const comments = [];
		for (const id of current) {
			const comment = this.getComment(unitId, subUnitId, id);
			if (comment) comments.push(comment);
		}
		return comments;
	}
	getComment(unitId, subUnitId, commentId) {
		return this._threadCommentModel.getComment(unitId, subUnitId, commentId);
	}
	getCommentWithChildren(unitId, subUnitId, row, column) {
		const commentId = this.getByLocation(unitId, subUnitId, row, column);
		if (!commentId) return;
		const comment = this.getComment(unitId, subUnitId, commentId);
		if (!comment) return;
		return this._threadCommentModel.getThread(unitId, subUnitId, comment.threadId);
	}
	showCommentMarker(unitId, subUnitId, row, column) {
		const commentId = this.getByLocation(unitId, subUnitId, row, column);
		if (!commentId) return false;
		const comment = this.getComment(unitId, subUnitId, commentId);
		if (comment && !comment.resolved) return true;
		return false;
	}
	getSubUnitAll(unitId, subUnitId) {
		return this._threadCommentModel.getUnit(unitId).filter((i) => i.subUnitId === subUnitId).map((i) => i.root);
	}
};
SheetsThreadCommentModel = __decorate([__decorateParam(0, Inject(ThreadCommentModel)), __decorateParam(1, IUniverInstanceService)], SheetsThreadCommentModel);

//#endregion
//#region src/controllers/sheets-thread-comment-ref-range.controller.ts
let SheetsThreadCommentRefRangeController = class SheetsThreadCommentRefRangeController extends Disposable {
	constructor(_refRangeService, _sheetsThreadCommentModel, _threadCommentModel, _selectionManagerService, _commandService) {
		super();
		this._refRangeService = _refRangeService;
		this._sheetsThreadCommentModel = _sheetsThreadCommentModel;
		this._threadCommentModel = _threadCommentModel;
		this._selectionManagerService = _selectionManagerService;
		this._commandService = _commandService;
		_defineProperty(this, "_disposableMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_watcherMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_handleRangeChange", (unitId, subUnitId, comment, resultRange, silent) => {
			const commentId = comment.id;
			const oldRange = {
				startColumn: comment.column,
				endColumn: comment.column,
				startRow: comment.row,
				endRow: comment.row
			};
			if (!resultRange) return {
				redos: [{
					id: DeleteCommentMutation.id,
					params: {
						unitId,
						subUnitId,
						commentId
					}
				}],
				undos: [{
					id: AddCommentMutation.id,
					params: {
						unitId,
						subUnitId,
						comment,
						sync: true
					}
				}]
			};
			return {
				redos: [{
					id: UpdateCommentRefMutation.id,
					params: {
						unitId,
						subUnitId,
						payload: {
							ref: serializeRange(resultRange),
							commentId
						},
						silent
					}
				}],
				undos: [{
					id: UpdateCommentRefMutation.id,
					params: {
						unitId,
						subUnitId,
						payload: {
							ref: serializeRange(oldRange),
							commentId
						},
						silent
					}
				}]
			};
		});
		this._initData();
		this._initRefRange();
	}
	_getIdWithUnitId(unitId, subUnitId, id) {
		return `${unitId}-${subUnitId}-${id}`;
	}
	_register(unitId, subUnitId, comment) {
		const commentId = comment.id;
		const oldRange = {
			startColumn: comment.column,
			endColumn: comment.column,
			startRow: comment.row,
			endRow: comment.row
		};
		this._disposableMap.set(this._getIdWithUnitId(unitId, subUnitId, commentId), this._refRangeService.registerRefRange(oldRange, (commandInfo) => {
			const resultRanges = handleCommonRangeChangeWithEffectRefCommandsSkipNoInterests(oldRange, commandInfo, { selectionManagerService: this._selectionManagerService });
			const resultRange = Array.isArray(resultRanges) ? resultRanges[0] : resultRanges;
			if (resultRange && resultRange.startColumn === oldRange.startColumn && resultRange.startRow === oldRange.startRow) return {
				undos: [],
				redos: []
			};
			return this._handleRangeChange(unitId, subUnitId, comment, resultRange, false);
		}, unitId, subUnitId));
	}
	_watch(unitId, subUnitId, comment) {
		const commentId = comment.id;
		const oldRange = {
			startColumn: comment.column,
			endColumn: comment.column,
			startRow: comment.row,
			endRow: comment.row
		};
		this._watcherMap.set(this._getIdWithUnitId(unitId, subUnitId, commentId), this._refRangeService.watchRange(unitId, subUnitId, oldRange, (before, after) => {
			const { redos } = this._handleRangeChange(unitId, subUnitId, comment, after, true);
			sequenceExecuteAsync(redos, this._commandService, { onlyLocal: true });
		}, true));
	}
	_unwatch(unitId, subUnitId, commentId) {
		var _this$_watcherMap$get;
		const id = this._getIdWithUnitId(unitId, subUnitId, commentId);
		(_this$_watcherMap$get = this._watcherMap.get(id)) === null || _this$_watcherMap$get === void 0 || _this$_watcherMap$get.dispose();
		this._watcherMap.delete(id);
	}
	_unregister(unitId, subUnitId, commentId) {
		var _this$_disposableMap$;
		const id = this._getIdWithUnitId(unitId, subUnitId, commentId);
		(_this$_disposableMap$ = this._disposableMap.get(id)) === null || _this$_disposableMap$ === void 0 || _this$_disposableMap$.dispose();
		this._disposableMap.delete(id);
	}
	_initData() {
		const datas = this._threadCommentModel.getAll();
		for (const data of datas) for (const thread of data.threads) {
			const { unitId, subUnitId, root } = thread;
			const pos = singleReferenceToGrid(root.ref);
			const sheetComment = {
				...root,
				...pos
			};
			this._register(unitId, subUnitId, sheetComment);
			this._watch(unitId, subUnitId, sheetComment);
		}
	}
	_initRefRange() {
		this.disposeWithMe(this._sheetsThreadCommentModel.commentUpdate$.subscribe((option) => {
			const { unitId, subUnitId } = option;
			switch (option.type) {
				case "add": {
					if (option.payload.parentId) return;
					const comment = {
						...option.payload,
						row: option.row,
						column: option.column
					};
					this._register(option.unitId, option.subUnitId, comment);
					this._watch(option.unitId, option.subUnitId, comment);
					break;
				}
				case "delete":
					this._unregister(unitId, subUnitId, option.payload.commentId);
					this._unwatch(unitId, subUnitId, option.payload.commentId);
					break;
				case "updateRef": {
					const comment = this._sheetsThreadCommentModel.getComment(unitId, subUnitId, option.payload.commentId);
					if (!comment) return;
					this._unregister(unitId, subUnitId, option.payload.commentId);
					const sheetComment = {
						...comment,
						row: option.row,
						column: option.column
					};
					if (!option.silent) {
						this._unwatch(unitId, subUnitId, option.payload.commentId);
						this._watch(unitId, subUnitId, sheetComment);
					}
					this._register(option.unitId, option.subUnitId, sheetComment);
					break;
				}
			}
		}));
		this.disposeWithMe(toDisposable(() => {
			this._disposableMap.forEach((item) => {
				item.dispose();
			});
			this._disposableMap.clear();
		}));
	}
};
SheetsThreadCommentRefRangeController = __decorate([
	__decorateParam(0, Inject(RefRangeService)),
	__decorateParam(1, Inject(SheetsThreadCommentModel)),
	__decorateParam(2, Inject(ThreadCommentModel)),
	__decorateParam(3, Inject(SheetsSelectionsService)),
	__decorateParam(4, ICommandService)
], SheetsThreadCommentRefRangeController);

//#endregion
//#region package.json
var name = "@univerjs/sheets-thread-comment";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const SHEETS_THREAD_COMMENT_PLUGIN_CONFIG_KEY = "sheets-thread-comment.config";
const configSymbol = Symbol(SHEETS_THREAD_COMMENT_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/controllers/sheets-thread-comment-resource.controller.ts
let SheetsThreadCommentResourceController = class SheetsThreadCommentResourceController extends Disposable {
	constructor(_univerInstanceService, _sheetInterceptorService, _threadCommentModel, _threadCommentDataSourceService) {
		super();
		this._univerInstanceService = _univerInstanceService;
		this._sheetInterceptorService = _sheetInterceptorService;
		this._threadCommentModel = _threadCommentModel;
		this._threadCommentDataSourceService = _threadCommentDataSourceService;
		this._initSheetChange();
	}
	_initSheetChange() {
		this.disposeWithMe(this._sheetInterceptorService.interceptCommand({ getMutations: (commandInfo) => {
			if (commandInfo.id === RemoveSheetCommand.id) {
				var _getActiveSheet;
				const params = commandInfo.params;
				const unitId = params.unitId || this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_SHEET).getUnitId();
				const subUnitId = params.subUnitId || ((_getActiveSheet = this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_SHEET).getActiveSheet()) === null || _getActiveSheet === void 0 ? void 0 : _getActiveSheet.getSheetId());
				if (!unitId || !subUnitId) return {
					redos: [],
					undos: []
				};
				const commentMap = this._threadCommentModel.ensureMap(unitId, subUnitId);
				const comments = Array.from(commentMap.values()).filter((comment) => !comment.parentId);
				const shouldSync = this._threadCommentDataSourceService.syncUpdateMutationToColla;
				const redos = [];
				const undos = [];
				comments.forEach(({ children, ...comment }) => {
					redos.push({
						id: DeleteCommentMutation.id,
						params: {
							unitId,
							subUnitId,
							commentId: comment.id
						}
					});
					undos.push({
						id: AddCommentMutation.id,
						params: {
							unitId,
							subUnitId,
							comment: {
								...comment,
								children: shouldSync ? children : void 0
							},
							sync: !shouldSync
						}
					});
				});
				return {
					redos,
					undos
				};
			} else if (commandInfo.id === CopySheetCommand.id) {
				const { unitId, subUnitId, targetSubUnitId } = commandInfo.params;
				if (!unitId || !subUnitId || !targetSubUnitId) return {
					redos: [],
					undos: []
				};
				const commentMap = this._threadCommentModel.ensureMap(unitId, subUnitId);
				const comments = Array.from(commentMap.values()).map((comment) => {
					return {
						...comment,
						subUnitId: targetSubUnitId,
						id: generateRandomId(),
						threadId: generateRandomId()
					};
				}).filter((comment) => !comment.parentId);
				const shouldSync = this._threadCommentDataSourceService.syncUpdateMutationToColla;
				const redos = [];
				const undos = [];
				comments.forEach(({ children, ...comment }) => {
					redos.push({
						id: AddCommentMutation.id,
						params: {
							unitId,
							subUnitId: targetSubUnitId,
							comment: {
								...comment,
								children: shouldSync ? children : void 0
							},
							sync: !shouldSync
						}
					});
					undos.push({
						id: DeleteCommentMutation.id,
						params: {
							unitId,
							subUnitId: targetSubUnitId,
							commentId: comment.id
						}
					});
				});
				return {
					redos,
					undos
				};
			}
			return {
				redos: [],
				undos: []
			};
		} }));
	}
};
SheetsThreadCommentResourceController = __decorate([
	__decorateParam(0, IUniverInstanceService),
	__decorateParam(1, Inject(SheetInterceptorService)),
	__decorateParam(2, Inject(ThreadCommentModel)),
	__decorateParam(3, IThreadCommentDataSourceService)
], SheetsThreadCommentResourceController);

//#endregion
//#region src/types/const.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const SHEET_THREAD_COMMENT_BASE = "SHEET_THREAD_COMMENT_BASE_PLUGIN";

//#endregion
//#region src/plugin.ts
let UniverSheetsThreadCommentPlugin = class UniverSheetsThreadCommentPlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _injector, _commandService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._commandService = _commandService;
	}
	onStarting() {
		[
			[SheetsThreadCommentModel],
			[SheetsThreadCommentRefRangeController],
			[SheetsThreadCommentResourceController]
		].forEach((dep) => {
			this._injector.add(dep);
		});
		touchDependencies(this._injector, [[SheetsThreadCommentRefRangeController], [SheetsThreadCommentResourceController]]);
	}
};
_defineProperty(UniverSheetsThreadCommentPlugin, "pluginName", SHEET_THREAD_COMMENT_BASE);
_defineProperty(UniverSheetsThreadCommentPlugin, "packageName", name);
_defineProperty(UniverSheetsThreadCommentPlugin, "version", version);
_defineProperty(UniverSheetsThreadCommentPlugin, "type", UniverInstanceType.UNIVER_SHEET);
UniverSheetsThreadCommentPlugin = __decorate([
	DependentOn(UniverThreadCommentPlugin),
	__decorateParam(1, Inject(Injector)),
	__decorateParam(2, Inject(ICommandService))
], UniverSheetsThreadCommentPlugin);

//#endregion
export { SheetsThreadCommentModel, SheetsThreadCommentRefRangeController, UniverSheetsThreadCommentPlugin };
