// @univerjs/thread-comment-ui/lib/index.js
import { BuildTextUtils, CommandType, CustomRangeType, DOCS_COMMENT_EDITOR_UNIT_ID_KEY, DOCS_NORMAL_EDITOR_UNIT_ID_KEY, DependentOn, Disposable, ICommandService, IConfigService, IUniverInstanceService, Inject, Injector, LocaleService, Plugin, Tools, UniverInstanceType, UserManagerService, generateRandomId, getBodySlice, merge, mergeOverrideWithDependencies } from "@univerjs/core";
import { ISidebarService, KeyCode, UI_PLUGIN_CONFIG_KEY, useConfigValue, useDependency, useObservable } from "@univerjs/ui";
import { BehaviorSubject, debounceTime, filter } from "rxjs";
import { AddCommentCommand, DeleteCommentCommand, DeleteCommentTreeCommand, ResolveCommentCommand, ThreadCommentModel, UniverThreadCommentPlugin, UpdateCommentCommand, getDT } from "@univerjs/thread-comment";
import { Button, Dropdown, Select, Tooltip, borderClassName, clsx, scrollbarClassName } from "@univerjs/design";
import { DeleteIcon, IncreaseIcon, MoreHorizontalIcon, ReplyToCommentIcon, ResolvedIcon, SolveIcon } from "@univerjs/icons";
import { forwardRef, useEffect, useImperativeHandle, useMemo, useRef, useState } from "react";
import { BreakLineCommand, IEditorService, RichTextEditor } from "@univerjs/docs-ui";
import { jsx, jsxs } from "react/jsx-runtime";

//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/services/thread-comment-panel.service.ts
let ThreadCommentPanelService = class ThreadCommentPanelService extends Disposable {
	constructor(_sidebarService, _univerInstanceService) {
		super();
		this._sidebarService = _sidebarService;
		this._univerInstanceService = _univerInstanceService;
		_defineProperty(this, "_panelVisible", false);
		_defineProperty(this, "_panelVisible$", new BehaviorSubject(false));
		_defineProperty(this, "_activeCommentId", void 0);
		_defineProperty(this, "_activeCommentId$", new BehaviorSubject(void 0));
		_defineProperty(this, "panelVisible$", this._panelVisible$.asObservable());
		_defineProperty(this, "activeCommentId$", this._activeCommentId$.asObservable());
		this._init();
		this.disposeWithMe(() => {
			this._activeCommentId$.complete();
			this._panelVisible$.complete();
		});
	}
	_init() {
		this.disposeWithMe(this._sidebarService.sidebarOptions$.subscribe((opt) => {
			if (!opt.visible) this.setPanelVisible(false);
		}));
		this.disposeWithMe(this._univerInstanceService.getCurrentTypeOfUnit$(UniverInstanceType.UNIVER_SHEET).pipe(filter((sheet) => !sheet)).subscribe(() => {
			this._sidebarService.close();
		}));
	}
	get panelVisible() {
		return this._panelVisible;
	}
	get activeCommentId() {
		return this._activeCommentId;
	}
	setPanelVisible(visible) {
		this._panelVisible = visible;
		this._panelVisible$.next(visible);
	}
	setActiveComment(commentInfo) {
		this._activeCommentId = commentInfo;
		this._activeCommentId$.next(commentInfo);
	}
};
ThreadCommentPanelService = __decorate([__decorateParam(0, Inject(ISidebarService)), __decorateParam(1, IUniverInstanceService)], ThreadCommentPanelService);

//#endregion
//#region src/commands/operations/comment.operations.ts
const SetActiveCommentOperation = {
	id: "thread-comment-ui.operation.set-active-comment",
	type: CommandType.OPERATION,
	handler(accessor, params) {
		accessor.get(ThreadCommentPanelService).setActiveComment(params);
		return true;
	}
};

//#endregion
//#region package.json
var name = "@univerjs/thread-comment-ui";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
const THREAD_COMMENT_UI_PLUGIN_CONFIG_KEY = "thread-comment-ui.config";
const configSymbol = Symbol(THREAD_COMMENT_UI_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/types/const.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const PLUGIN_NAME = "UNIVER_THREAD_COMMENT_UI_PLUGIN";

//#endregion
//#region src/plugin.ts
let UniverThreadCommentUIPlugin = class UniverThreadCommentUIPlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _injector, _commandService, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._commandService = _commandService;
		this._configService = _configService;
		const { menu, ...rest } = merge({}, defaultPluginConfig, this._config);
		if (menu) this._configService.setConfig("menu", menu, { merge: true });
		this._configService.setConfig(THREAD_COMMENT_UI_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		var _this$_config;
		mergeOverrideWithDependencies([[ThreadCommentPanelService]], (_this$_config = this._config) === null || _this$_config === void 0 ? void 0 : _this$_config.overrides).forEach((dep) => {
			this._injector.add(dep);
		});
		[SetActiveCommentOperation].forEach((command) => {
			this._commandService.registerCommand(command);
		});
	}
};
_defineProperty(UniverThreadCommentUIPlugin, "pluginName", PLUGIN_NAME);
_defineProperty(UniverThreadCommentUIPlugin, "packageName", name);
_defineProperty(UniverThreadCommentUIPlugin, "version", version);
_defineProperty(UniverThreadCommentUIPlugin, "type", UniverInstanceType.UNIVER_UNKNOWN);
UniverThreadCommentUIPlugin = __decorate([
	DependentOn(UniverThreadCommentPlugin),
	__decorateParam(1, Inject(Injector)),
	__decorateParam(2, ICommandService),
	__decorateParam(3, IConfigService)
], UniverThreadCommentUIPlugin);

//#endregion
//#region src/views/thread-comment-editor/util.ts
const transformDocument2TextNodesInParagraph = (doc) => {
	const { dataStream, customRanges } = doc;
	const end = dataStream.endsWith("\r\n") ? dataStream.length - 2 : dataStream.length;
	const textNodes = [];
	let lastIndex = 0;
	customRanges === null || customRanges === void 0 || customRanges.forEach((range) => {
		if (lastIndex < range.startIndex) textNodes.push({
			type: "text",
			content: dataStream.slice(lastIndex, range.startIndex)
		});
		textNodes.push({
			type: "mention",
			content: {
				label: dataStream.slice(range.startIndex, range.endIndex + 1),
				id: range.rangeId
			}
		});
		lastIndex = range.endIndex + 1;
	});
	textNodes.push({
		type: "text",
		content: dataStream.slice(lastIndex, end)
	});
	return textNodes;
};
const transformDocument2TextNodes = (doc) => {
	if (!doc) return [];
	const { paragraphs = [] } = doc;
	return paragraphs.map((paragraph, index) => {
		return transformDocument2TextNodesInParagraph(getBodySlice(doc, index === 0 ? 0 : paragraphs[index - 1].startIndex + 1, paragraph.startIndex));
	});
};
const transformTextNodes2Document = (nodes) => {
	let str = "";
	const customRanges = [];
	nodes.forEach((node) => {
		switch (node.type) {
			case "text":
				str += node.content;
				break;
			case "mention": {
				const start = str.length;
				str += node.content.label;
				const end = str.length - 1;
				customRanges.push({
					rangeId: node.content.id,
					rangeType: CustomRangeType.MENTION,
					startIndex: start,
					endIndex: end,
					properties: {},
					wholeEntity: true
				});
				break;
			}
			default: break;
		}
	});
	str += "\r\n";
	return {
		textRuns: [],
		paragraphs: [{
			startIndex: str.length - 2,
			paragraphStyle: {}
		}],
		sectionBreaks: [{ startIndex: str.length - 1 }],
		dataStream: str,
		customRanges
	};
};
function focusThreadCommentEditor(editorService, editorId, editor) {
	editorService.focus(editorId);
	editor === null || editor === void 0 || editor.focus();
}

//#endregion
//#region src/views/thread-comment-editor/index.tsx
function getSnapshot(body) {
	return {
		id: "d",
		body,
		documentStyle: {}
	};
}
const ThreadCommentEditor = forwardRef((props, ref) => {
	var _editor$current5;
	const { comment, onSave, id, onCancel, autoFocus, unitId, type, editorId } = props;
	const commandService = useDependency(ICommandService);
	const localeService = useDependency(LocaleService);
	const [editing, setEditing] = useState(false);
	const editorService = useDependency(IEditorService);
	const editor = useRef(null);
	const rootEditorId = type === UniverInstanceType.UNIVER_DOC ? DOCS_NORMAL_EDITOR_UNIT_ID_KEY : unitId;
	const [canSubmit, setCanSubmit] = useState(() => {
		var _editor$current$getDo, _editor$current;
		return BuildTextUtils.transform.getPlainText((_editor$current$getDo = (_editor$current = editor.current) === null || _editor$current === void 0 || (_editor$current = _editor$current.getDocumentData().body) === null || _editor$current === void 0 ? void 0 : _editor$current.dataStream) !== null && _editor$current$getDo !== void 0 ? _editor$current$getDo : "");
	});
	useEffect(() => {
		var _editor$current$getDo2, _editor$current2, _editor$current3;
		setCanSubmit(BuildTextUtils.transform.getPlainText((_editor$current$getDo2 = (_editor$current2 = editor.current) === null || _editor$current2 === void 0 || (_editor$current2 = _editor$current2.getDocumentData().body) === null || _editor$current2 === void 0 ? void 0 : _editor$current2.dataStream) !== null && _editor$current$getDo2 !== void 0 ? _editor$current$getDo2 : ""));
		const sub = (_editor$current3 = editor.current) === null || _editor$current3 === void 0 ? void 0 : _editor$current3.selectionChange$.subscribe(() => {
			var _editor$current$getDo3, _editor$current4;
			setCanSubmit(BuildTextUtils.transform.getPlainText((_editor$current$getDo3 = (_editor$current4 = editor.current) === null || _editor$current4 === void 0 || (_editor$current4 = _editor$current4.getDocumentData().body) === null || _editor$current4 === void 0 ? void 0 : _editor$current4.dataStream) !== null && _editor$current$getDo3 !== void 0 ? _editor$current$getDo3 : ""));
		});
		return () => sub === null || sub === void 0 ? void 0 : sub.unsubscribe();
	}, [(_editor$current5 = editor.current) === null || _editor$current5 === void 0 ? void 0 : _editor$current5.selectionChange$]);
	const keyboardEventConfig = useMemo(() => ({
		keyCodes: [{ keyCode: KeyCode.ENTER }],
		handler: (keyCode) => {
			if (keyCode === KeyCode.ENTER) commandService.executeCommand(BreakLineCommand.id);
		}
	}), [commandService]);
	useImperativeHandle(ref, () => ({ reply(text) {
		var _editor$current6;
		if (!editor.current) return;
		focusThreadCommentEditor(editorService, editorId, editor.current);
		const documentData = getSnapshot(text);
		(_editor$current6 = editor.current) === null || _editor$current6 === void 0 || _editor$current6.setDocumentData(documentData, [{
			startOffset: documentData.body.dataStream.length - 2,
			endOffset: documentData.body.dataStream.length - 2,
			collapsed: true
		}]);
		setCanSubmit(BuildTextUtils.transform.getPlainText(documentData.body.dataStream));
		setEditing(true);
	} }));
	const handleSave = () => {
		const currentEditor = editor.current;
		if (currentEditor) {
			const newText = Tools.deepClone(currentEditor.getDocumentData().body);
			currentEditor.blur();
			currentEditor.replaceText("", false);
			currentEditor.setSelectionRanges([], false);
			setCanSubmit("");
			setEditing(false);
			onSave === null || onSave === void 0 || onSave({
				...comment,
				text: newText
			});
		}
	};
	const handleEditorMouseDown = () => {
		focusThreadCommentEditor(editorService, editorId, editor.current);
		setEditing(true);
	};
	useEffect(() => {
		if (!autoFocus) return;
		const timer = setTimeout(() => {
			focusThreadCommentEditor(editorService, editorId, editor.current);
		});
		return () => clearTimeout(timer);
	}, [
		autoFocus,
		editorId,
		editorService
	]);
	return /* @__PURE__ */ jsxs("div", {
		onClick: (e) => e.preventDefault(),
		children: [/* @__PURE__ */ jsx("div", {
			onMouseDown: handleEditorMouseDown,
			children: /* @__PURE__ */ jsx(RichTextEditor, {
				className: "univer-w-full",
				editorRef: editor,
				editorId,
				autoFocus,
				keyboardEventConfig,
				placeholder: localeService.t("thread-comment-ui.editor.placeholder"),
				initialValue: (comment === null || comment === void 0 ? void 0 : comment.text) && getSnapshot(comment.text),
				onFocusChange: (isFocus) => isFocus && setEditing(isFocus),
				isSingle: false,
				maxHeight: 64,
				onClickOutside: () => {
					setTimeout(() => {
						editorService.focus(rootEditorId);
					}, 30);
				}
			})
		}), editing ? /* @__PURE__ */ jsxs("div", {
			className: "univer-mt-3 univer-flex univer-flex-row univer-justify-end univer-gap-2",
			children: [/* @__PURE__ */ jsx(Button, {
				type: "button",
				onClick: () => {
					const currentEditor = editor.current;
					currentEditor === null || currentEditor === void 0 || currentEditor.blur();
					currentEditor === null || currentEditor === void 0 || currentEditor.replaceText("", false);
					currentEditor === null || currentEditor === void 0 || currentEditor.setSelectionRanges([], false);
					setCanSubmit("");
					onCancel === null || onCancel === void 0 || onCancel();
					setEditing(false);
					commandService.executeCommand(SetActiveCommentOperation.id);
				},
				children: localeService.t("thread-comment-ui.editor.cancel")
			}), /* @__PURE__ */ jsx(Button, {
				type: "button",
				variant: "primary",
				disabled: !canSubmit,
				onClick: handleSave,
				children: localeService.t(id ? "thread-comment-ui.editor.save" : "thread-comment-ui.editor.reply")
			})]
		}) : null]
	});
});

//#endregion
//#region src/views/thread-comment-tree/util.ts
function getThreadCommentEditorId(params) {
	const { location, unitId, subUnitId, commentId, fallbackId } = params;
	return `${DOCS_COMMENT_EDITOR_UNIT_ID_KEY}_${location}_${unitId}_${subUnitId}_${commentId || fallbackId}`;
}

//#endregion
//#region src/views/thread-comment-tree/index.tsx
let ThreadCommentTreeLocation = /* @__PURE__ */ function(ThreadCommentTreeLocation) {
	ThreadCommentTreeLocation["CELL"] = "CELL";
	ThreadCommentTreeLocation["PANEL"] = "PANEL";
	return ThreadCommentTreeLocation;
}({});
const MOCK_ID = "__mock__";
const ThreadCommentItem = (props) => {
	const { item, unitId, subUnitId, editing, onEditingChange, onReply, resolved, isRoot, onClose, onDeleteComment, type, threadCommentEditorId } = props;
	const commandService = useDependency(ICommandService);
	const localeService = useDependency(LocaleService);
	const userManagerService = useDependency(UserManagerService);
	const user = userManagerService.getUser(item.personId);
	const currentUser = useObservable(userManagerService.currentUser$);
	const isCommentBySelf = (currentUser === null || currentUser === void 0 ? void 0 : currentUser.userID) === item.personId;
	const isMock = item.id === MOCK_ID;
	const [showReply, setShowReply] = useState(false);
	const uiConfig = useConfigValue(UI_PLUGIN_CONFIG_KEY);
	const avatarFallback = uiConfig === null || uiConfig === void 0 ? void 0 : uiConfig.avatarFallback;
	const handleDeleteItem = () => {
		if ((onDeleteComment === null || onDeleteComment === void 0 ? void 0 : onDeleteComment(item)) === false) return;
		commandService.executeCommand(isRoot ? DeleteCommentTreeCommand.id : DeleteCommentCommand.id, {
			unitId,
			subUnitId,
			commentId: item.id
		});
		if (isRoot) onClose === null || onClose === void 0 || onClose();
	};
	return /* @__PURE__ */ jsxs("div", {
		className: "univer-relative univer-mb-3 univer-pl-[30px]",
		onMouseLeave: () => setShowReply(false),
		onMouseEnter: () => setShowReply(true),
		children: [
			/* @__PURE__ */ jsx("div", {
				className: "univer-absolute univer-left-0 univer-top-0 univer-size-6 univer-rounded-full univer-bg-cover univer-bg-center univer-bg-no-repeat",
				style: { backgroundImage: `url(${(user === null || user === void 0 ? void 0 : user.avatar) || avatarFallback})` }
			}),
			user ? /* @__PURE__ */ jsxs("div", {
				className: "univer-mb-1 univer-flex univer-h-6 univer-items-center univer-justify-between",
				children: [/* @__PURE__ */ jsx("div", {
					className: "univer-text-sm univer-font-medium univer-leading-5",
					children: (user === null || user === void 0 ? void 0 : user.name) || " "
				}), /* @__PURE__ */ jsxs("div", { children: [isMock || resolved ? null : showReply && user ? /* @__PURE__ */ jsx("div", {
					className: "univer-ml-1 univer-inline-flex univer-size-6 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded-sm univer-text-base hover:univer-bg-gray-50 dark:hover:!univer-bg-gray-800",
					onClick: () => onReply(user),
					children: /* @__PURE__ */ jsx(ReplyToCommentIcon, {})
				}) : null, isCommentBySelf && !isMock && !resolved ? /* @__PURE__ */ jsx(Dropdown, {
					overlay: /* @__PURE__ */ jsx("div", {
						className: "univer-rounded-lg",
						children: /* @__PURE__ */ jsxs("ul", {
							className: "univer-m-0 univer-box-border univer-grid univer-list-none univer-p-1.5 univer-text-sm [&_a]:univer-block [&_a]:univer-cursor-pointer [&_a]:univer-rounded [&_a]:univer-px-2 [&_a]:univer-py-1.5 [&_a]:univer-transition-colors",
							children: [/* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx("a", {
								className: "hover:univer-bg-gray-200",
								onClick: () => onEditingChange === null || onEditingChange === void 0 ? void 0 : onEditingChange(true),
								children: localeService.t("thread-comment-ui.item.edit")
							}) }), /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx("a", {
								className: "hover:univer-bg-gray-200",
								onClick: handleDeleteItem,
								children: localeService.t("thread-comment-ui.item.delete")
							}) })]
						})
					}),
					children: /* @__PURE__ */ jsx("div", {
						className: "univer-ml-1 univer-inline-flex univer-size-6 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded-sm univer-text-base hover:univer-bg-gray-50 dark:hover:!univer-bg-gray-800",
						children: /* @__PURE__ */ jsx(MoreHorizontalIcon, {})
					})
				}) : null] })]
			}) : null,
			/* @__PURE__ */ jsx("time", {
				className: "univer-mb-1 univer-text-xs/normal univer-text-gray-600 dark:!univer-text-gray-200",
				children: item.dT
			}),
			editing ? /* @__PURE__ */ jsx(ThreadCommentEditor, {
				type,
				id: item.id,
				comment: item,
				onCancel: () => onEditingChange === null || onEditingChange === void 0 ? void 0 : onEditingChange(false),
				autoFocus: true,
				unitId,
				subUnitId,
				editorId: threadCommentEditorId,
				onSave: ({ text, attachments }) => {
					onEditingChange === null || onEditingChange === void 0 || onEditingChange(false);
					commandService.executeCommand(UpdateCommentCommand.id, {
						unitId,
						subUnitId,
						payload: {
							commentId: item.id,
							text,
							attachments
						}
					});
				}
			}) : /* @__PURE__ */ jsx("div", {
				className: "univer-text-sm univer-text-gray-900 dark:!univer-text-white",
				children: transformDocument2TextNodes(item.text).map((paragraph, i) => /* @__PURE__ */ jsx("div", {
					className: "univer-break-words",
					children: paragraph.map((item, i) => {
						switch (item.type) {
							case "mention": return /* @__PURE__ */ jsxs("a", {
								className: "univer-text-primary-600",
								children: [item.content.label, " "]
							}, i);
							default: return item.content;
						}
					})
				}, i))
			})
		]
	});
};
const ThreadCommentTree = (props) => {
	var _currentUser$userID, _comments$children, _comments$root$subUni;
	const { id, unitId, subUnitId, refStr, showEdit = true, onClick, showHighlight, onClose, getSubUnitName, location, autoFocus, onMouseEnter, onMouseLeave, onAddComment, onDeleteComment, onResolve, type, style, full } = props;
	const threadCommentModel = useDependency(ThreadCommentModel);
	const [isHover, setIsHover] = useState(false);
	const [editingId, setEditingId] = useState("");
	useObservable(useMemo(() => threadCommentModel.commentUpdate$.pipe(debounceTime(16)), [threadCommentModel]));
	const comments = id ? threadCommentModel.getCommentWithChildren(unitId, subUnitId, id) : null;
	const commandService = useDependency(ICommandService);
	const userManagerService = useDependency(UserManagerService);
	const resolved = comments === null || comments === void 0 ? void 0 : comments.root.resolved;
	const currentUser = useObservable(userManagerService.currentUser$);
	const editorRef = useRef(null);
	const fallbackEditorId = useMemo(() => generateRandomId(6), []);
	const renderComments = [...comments ? [comments.root] : [{
		id: MOCK_ID,
		text: { dataStream: "\n\r" },
		personId: (_currentUser$userID = currentUser === null || currentUser === void 0 ? void 0 : currentUser.userID) !== null && _currentUser$userID !== void 0 ? _currentUser$userID : "",
		ref: refStr !== null && refStr !== void 0 ? refStr : "",
		dT: "",
		unitId,
		subUnitId,
		threadId: ""
	}], ...(_comments$children = comments === null || comments === void 0 ? void 0 : comments.children) !== null && _comments$children !== void 0 ? _comments$children : []];
	const scroller = useRef(null);
	const handleResolve = (e) => {
		e.stopPropagation();
		if (!resolved) commandService.executeCommand(SetActiveCommentOperation.id);
		else commandService.executeCommand(SetActiveCommentOperation.id, {
			unitId,
			subUnitId,
			commentId: id
		});
		commandService.executeCommand(ResolveCommentCommand.id, {
			unitId,
			subUnitId,
			commentId: id,
			resolved: !resolved
		});
		onResolve === null || onResolve === void 0 || onResolve(!resolved);
	};
	const handleDeleteRoot = (e) => {
		e.stopPropagation();
		commandService.executeCommand(SetActiveCommentOperation.id);
		if ((comments === null || comments === void 0 ? void 0 : comments.root) && (onDeleteComment === null || onDeleteComment === void 0 ? void 0 : onDeleteComment(comments.root)) === false) return;
		commandService.executeCommand(DeleteCommentTreeCommand.id, {
			unitId,
			subUnitId,
			commentId: id
		});
		onClose === null || onClose === void 0 || onClose();
	};
	useEffect(() => {
		return onMouseLeave === null || onMouseLeave === void 0 ? void 0 : onMouseLeave();
	}, []);
	const subUnitName = getSubUnitName((_comments$root$subUni = comments === null || comments === void 0 ? void 0 : comments.root.subUnitId) !== null && _comments$root$subUni !== void 0 ? _comments$root$subUni : subUnitId);
	const editorVisible = showEdit && !editingId && !resolved;
	const title = `${refStr || (comments === null || comments === void 0 ? void 0 : comments.root.ref) || ""}${subUnitName ? " · " : ""}${subUnitName}`;
	const threadCommentEditorId = getThreadCommentEditorId({
		location,
		unitId,
		subUnitId,
		commentId: id,
		fallbackId: fallbackEditorId
	});
	return /* @__PURE__ */ jsxs("div", {
		id: `${location}-${unitId}-${subUnitId}-${id}`,
		className: clsx("univer-relative univer-box-border univer-rounded-md univer-bg-white univer-p-4 dark:!univer-bg-gray-900 dark:!univer-text-white", borderClassName, {
			"univer-w-[278px]": !full,
			"univer-w-full": full,
			"univer-shadow": !resolved && (showHighlight || isHover || location === "CELL")
		}),
		style,
		onClick,
		onMouseEnter: () => {
			onMouseEnter === null || onMouseEnter === void 0 || onMouseEnter();
			setIsHover(true);
		},
		onMouseLeave: () => {
			onMouseLeave === null || onMouseLeave === void 0 || onMouseLeave();
			setIsHover(false);
		},
		children: [
			!resolved && showHighlight && /* @__PURE__ */ jsx("div", { className: "univer-absolute univer-left-0 univer-right-0 univer-top-0 univer-h-1.5 univer-rounded-t-md univer-bg-yellow-400" }),
			/* @__PURE__ */ jsxs("div", {
				className: "univer-mb-4 univer-flex univer-flex-row univer-items-center univer-justify-between univer-text-sm univer-leading-5",
				children: [/* @__PURE__ */ jsxs("div", {
					className: "univer-flex univer-flex-1 univer-flex-row univer-items-center univer-overflow-hidden",
					children: [/* @__PURE__ */ jsx("div", { className: "univer-mr-2 univer-h-3.5 univer-w-[3px] univer-flex-shrink-0 univer-flex-grow-0 univer-rounded-sm univer-bg-yellow-500" }), /* @__PURE__ */ jsx(Tooltip, {
						showIfEllipsis: true,
						title,
						children: /* @__PURE__ */ jsx("span", {
							className: "univer-flex-1 univer-truncate",
							children: title
						})
					})]
				}), !!comments && /* @__PURE__ */ jsxs("div", {
					className: "univer-flex univer-flex-shrink-0 univer-flex-grow-0 univer-flex-row",
					children: [/* @__PURE__ */ jsx("div", {
						className: clsx("univer-ml-1 univer-inline-flex univer-size-6 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded-[3px] univer-text-base hover:univer-bg-gray-50 dark:hover:!univer-bg-gray-800", { "univer-text-green-500": resolved }),
						onClick: handleResolve,
						children: resolved ? /* @__PURE__ */ jsx(ResolvedIcon, {}) : /* @__PURE__ */ jsx(SolveIcon, {})
					}), (currentUser === null || currentUser === void 0 ? void 0 : currentUser.userID) === comments.root.personId ? /* @__PURE__ */ jsx("div", {
						className: "univer-ml-1 univer-inline-flex univer-size-6 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded-[3px] univer-text-base hover:univer-bg-gray-50 dark:hover:!univer-bg-gray-800",
						onClick: handleDeleteRoot,
						children: /* @__PURE__ */ jsx(DeleteIcon, {})
					}) : null]
				})]
			}),
			/* @__PURE__ */ jsx("div", {
				ref: scroller,
				className: clsx("univer-max-h-80 univer-overflow-y-auto univer-overflow-x-hidden", scrollbarClassName),
				children: renderComments.map((item) => /* @__PURE__ */ jsx(ThreadCommentItem, {
					unitId,
					subUnitId,
					item,
					isRoot: item.id === (comments === null || comments === void 0 ? void 0 : comments.root.id),
					editing: editingId === item.id,
					resolved: comments === null || comments === void 0 ? void 0 : comments.root.resolved,
					type,
					threadCommentEditorId,
					onClose,
					onEditingChange: (editing) => {
						if (editing) setEditingId(item.id);
						else setEditingId("");
					},
					onReply: (user) => {
						if (!user) return;
						requestAnimationFrame(() => {
							var _editorRef$current;
							(_editorRef$current = editorRef.current) === null || _editorRef$current === void 0 || _editorRef$current.reply(transformTextNodes2Document([{
								type: "mention",
								content: {
									id: user.userID,
									label: `@${user.name}`
								}
							}, {
								type: "text",
								content: " "
							}]));
						});
					},
					onAddComment,
					onDeleteComment
				}, item.id))
			}),
			editorVisible && /* @__PURE__ */ jsx("div", { children: /* @__PURE__ */ jsx(ThreadCommentEditor, {
				ref: editorRef,
				type,
				unitId,
				subUnitId,
				editorId: threadCommentEditorId,
				onSave: async ({ text, attachments }) => {
					var _currentUser$userID2, _comments$root$thread;
					const comment = {
						text,
						attachments,
						dT: getDT(),
						id: generateRandomId(),
						ref: refStr,
						personId: (_currentUser$userID2 = currentUser === null || currentUser === void 0 ? void 0 : currentUser.userID) !== null && _currentUser$userID2 !== void 0 ? _currentUser$userID2 : "",
						parentId: comments === null || comments === void 0 ? void 0 : comments.root.id,
						unitId,
						subUnitId,
						threadId: (_comments$root$thread = comments === null || comments === void 0 ? void 0 : comments.root.threadId) !== null && _comments$root$thread !== void 0 ? _comments$root$thread : ""
					};
					if ((onAddComment === null || onAddComment === void 0 ? void 0 : onAddComment(comment)) === false) return;
					await commandService.executeCommand(AddCommentCommand.id, {
						unitId,
						subUnitId,
						comment
					});
					if (scroller.current) scroller.current.scrollTop = scroller.current.scrollHeight;
				},
				autoFocus: autoFocus || !comments,
				onCancel: () => {
					if (!comments) onClose === null || onClose === void 0 || onClose();
				}
			}, `${autoFocus}`) })
		]
	});
};

//#endregion
//#region src/views/thread-comment-panel/util.ts
function getThreadCommentPanelItemKey(comment, index, section) {
	if (comment.id) return comment.id;
	return [
		"thread-comment-panel-temp",
		section,
		comment.unitId,
		comment.subUnitId,
		comment.threadId,
		comment.ref,
		comment.dT,
		index
	].join("-");
}

//#endregion
//#region src/views/thread-comment-panel/index.tsx
const ThreadCommentPanel = (props) => {
	const { unitId, subUnitId$, type, onAdd, getSubUnitName, onResolve, sortComments, onItemLeave, onItemEnter, disableAdd, tempComment, onAddComment, onDeleteComment, showComments } = props;
	const [unit, setUnit] = useState("all");
	const [status, setStatus] = useState("all");
	const localeService = useDependency(LocaleService);
	const userService = useDependency(UserManagerService);
	const threadCommentModel = useDependency(ThreadCommentModel);
	const [unitComments, setUnitComments] = useState(() => threadCommentModel.getUnit(unitId));
	const activeCommentId = useObservable(useDependency(ThreadCommentPanelService).activeCommentId$);
	const update = useObservable(threadCommentModel.commentUpdate$);
	const commandService = useDependency(ICommandService);
	const subUnitId = useObservable(subUnitId$);
	const shouldScroll = useRef(true);
	const location = "PANEL";
	const currentUser = useObservable(userService.currentUser$);
	const comments = useMemo(() => {
		var _unitComments$filter;
		const allComments = unit === "all" ? unitComments : (_unitComments$filter = unitComments.filter((i) => i.subUnitId === subUnitId)) !== null && _unitComments$filter !== void 0 ? _unitComments$filter : [];
		const sort = sortComments !== null && sortComments !== void 0 ? sortComments : ((a) => a);
		const res = allComments.map((i) => {
			var _i$children;
			return {
				...i.root,
				children: (_i$children = i.children) !== null && _i$children !== void 0 ? _i$children : [],
				users: i.relativeUsers
			};
		});
		if (showComments) {
			const map = /* @__PURE__ */ new Map();
			res.forEach((comment) => {
				map.set(comment.id, comment);
			});
			return [...showComments, ""].map((id) => map.get(id)).filter(Boolean);
		} else return sort(res);
	}, [
		showComments,
		unit,
		unitComments,
		sortComments,
		subUnitId
	]);
	const commentsSorted = useMemo(() => [...comments.filter((comment) => !comment.resolved), ...comments.filter((comment) => comment.resolved)], [comments]);
	const statuedComments = useMemo(() => {
		if (status === "resolved") return commentsSorted.filter((comment) => comment.resolved);
		if (status === "unsolved") return commentsSorted.filter((comment) => !comment.resolved);
		if (status === "concern_me") {
			if (!(currentUser === null || currentUser === void 0 ? void 0 : currentUser.userID)) return commentsSorted;
			return commentsSorted.filter((comment) => comment === null || comment === void 0 ? void 0 : comment.users.has(currentUser.userID));
		}
		return commentsSorted;
	}, [
		commentsSorted,
		currentUser === null || currentUser === void 0 ? void 0 : currentUser.userID,
		status
	]);
	const renderComments = tempComment ? [tempComment, ...statuedComments] : statuedComments;
	const unSolvedComments = renderComments.filter((comment) => !comment.resolved);
	const solvedComments = renderComments.filter((comment) => comment.resolved);
	const isFiltering = status !== "all" || unit !== "all";
	const onReset = () => {
		setStatus("all");
		setUnit("all");
	};
	useEffect(() => {
		if (unitId) setUnitComments(threadCommentModel.getUnit(unitId));
	}, [
		unitId,
		threadCommentModel,
		update
	]);
	useEffect(() => {
		var _document$getElementB;
		if (!activeCommentId) return;
		if (!shouldScroll.current) {
			shouldScroll.current = true;
			return;
		}
		const { unitId, subUnitId, commentId } = activeCommentId;
		const id = `${location}-${unitId}-${subUnitId}-${commentId}`;
		(_document$getElementB = document.getElementById(id)) === null || _document$getElementB === void 0 || _document$getElementB.scrollIntoView({ block: "center" });
	}, [activeCommentId]);
	const renderComment = (section) => (comment, index) => /* @__PURE__ */ jsx(ThreadCommentTree, {
		location,
		getSubUnitName,
		id: comment.id,
		unitId: comment.unitId,
		subUnitId: comment.subUnitId,
		refStr: comment.ref,
		type,
		showEdit: (activeCommentId === null || activeCommentId === void 0 ? void 0 : activeCommentId.commentId) === comment.id,
		showHighlight: (activeCommentId === null || activeCommentId === void 0 ? void 0 : activeCommentId.commentId) === comment.id,
		onClick: () => {
			shouldScroll.current = false;
			if (!comment.resolved) commandService.executeCommand(SetActiveCommentOperation.id, {
				unitId: comment.unitId,
				subUnitId: comment.subUnitId,
				commentId: comment.id,
				temp: false
			});
			else commandService.executeCommand(SetActiveCommentOperation.id);
		},
		onMouseEnter: () => onItemEnter === null || onItemEnter === void 0 ? void 0 : onItemEnter(comment),
		onMouseLeave: () => onItemLeave === null || onItemLeave === void 0 ? void 0 : onItemLeave(comment),
		onAddComment,
		onDeleteComment,
		onResolve: (resolved) => onResolve === null || onResolve === void 0 ? void 0 : onResolve(comment.id, resolved)
	}, getThreadCommentPanelItemKey(comment, index, section));
	return /* @__PURE__ */ jsxs("div", {
		className: "univer-flex univer-min-h-full univer-flex-col univer-pb-3",
		children: [/* @__PURE__ */ jsxs("div", {
			className: "univer-mt-3 univer-flex univer-flex-row univer-justify-between",
			children: [type === UniverInstanceType.UNIVER_SHEET ? /* @__PURE__ */ jsx(Select, {
				borderless: true,
				value: unit,
				options: [{
					value: "current",
					label: localeService.t("thread-comment-ui.filter.sheet.current")
				}, {
					value: "all",
					label: localeService.t("thread-comment-ui.filter.sheet.all")
				}],
				onChange: setUnit
			}) : null, /* @__PURE__ */ jsx(Select, {
				borderless: true,
				value: status,
				options: [
					{
						value: "all",
						label: localeService.t("thread-comment-ui.filter.status.all")
					},
					{
						value: "resolved",
						label: localeService.t("thread-comment-ui.filter.status.resolved")
					},
					{
						value: "unsolved",
						label: localeService.t("thread-comment-ui.filter.status.unsolved")
					},
					{
						value: "concern_me",
						label: localeService.t("thread-comment-ui.filter.status.concernMe")
					}
				],
				onChange: setStatus
			})]
		}), renderComments.length === 0 ? /* @__PURE__ */ jsxs("div", {
			className: "univer-flex univer-flex-1 univer-flex-col univer-items-center univer-justify-center univer-text-sm univer-text-gray-600 dark:!univer-text-gray-200",
			children: [localeService.t("thread-comment-ui.panel.empty"), isFiltering ? /* @__PURE__ */ jsx("div", {
				className: "univer-mt-2 univer-flex univer-flex-row",
				children: /* @__PURE__ */ jsx(Button, {
					onClick: onReset,
					children: localeService.t("thread-comment-ui.panel.reset")
				})
			}) : !disableAdd ? /* @__PURE__ */ jsx("div", {
				className: "univer-mt-2 univer-flex univer-flex-row",
				children: /* @__PURE__ */ jsxs(Button, {
					onClick: onAdd,
					children: [/* @__PURE__ */ jsx(IncreaseIcon, { className: "univer-mr-1.5" }), localeService.t("thread-comment-ui.panel.addComment")]
				})
			}) : null]
		}) : /* @__PURE__ */ jsxs("div", {
			className: "univer-mt-3 univer-flex univer-flex-col univer-gap-3",
			children: [
				unSolvedComments.map(renderComment("unsolved")),
				solvedComments.length > 0 && /* @__PURE__ */ jsx("div", {
					className: "univer-text-xs",
					children: localeService.t("thread-comment-ui.panel.solved")
				}),
				solvedComments.map(renderComment("solved"))
			]
		})]
	});
};

//#endregion
export { SetActiveCommentOperation, ThreadCommentPanel, ThreadCommentPanelService, ThreadCommentTree, ThreadCommentTreeLocation, UniverThreadCommentUIPlugin };
