// @univerjs/sheets-note/lib/cjs/facade.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
let _univerjs_core_facade = require("@univerjs/core/facade");
let _univerjs_sheets_note = require("@univerjs/sheets-note");
let _univerjs_sheets_facade = require("@univerjs/sheets/facade");
let _univerjs_core = require("@univerjs/core");
let _univerjs_sheets = require("@univerjs/sheets");

//#region src/facade/f-event.ts
/**
* @ignore
*/
var FSheetsNoteEventNameMixin = class extends _univerjs_core_facade.FEventName {
	get SheetNoteAdd() {
		return "SheetNoteAdd";
	}
	get SheetNoteDelete() {
		return "SheetNoteDelete";
	}
	get SheetNoteUpdate() {
		return "SheetNoteUpdate";
	}
	get SheetNoteShow() {
		return "SheetNoteShow";
	}
	get SheetNoteHide() {
		return "SheetNoteHide";
	}
	get BeforeSheetNoteAdd() {
		return "BeforeSheetNoteAdd";
	}
	get BeforeSheetNoteDelete() {
		return "BeforeSheetNoteDelete";
	}
	get BeforeSheetNoteUpdate() {
		return "BeforeSheetNoteUpdate";
	}
	get BeforeSheetNoteShow() {
		return "BeforeSheetNoteShow";
	}
	get BeforeSheetNoteHide() {
		return "BeforeSheetNoteHide";
	}
};
_univerjs_core_facade.FEventName.extend(FSheetsNoteEventNameMixin);

//#endregion
//#region src/facade/f-range.ts
var FRangeSheetsNoteMixin = class extends _univerjs_sheets_facade.FRange {
	createOrUpdateNote(note) {
		this._commandService.syncExecuteCommand(_univerjs_sheets_note.UpdateNoteMutation.id, {
			unitId: this.getUnitId(),
			sheetId: this.getSheetId(),
			row: this.getRow(),
			col: this.getColumn(),
			note
		});
		return this;
	}
	deleteNote() {
		this._commandService.syncExecuteCommand(_univerjs_sheets_note.RemoveNoteMutation.id, {
			unitId: this.getUnitId(),
			sheetId: this.getSheetId(),
			row: this.getRow(),
			col: this.getColumn()
		});
		return this;
	}
	getNote() {
		return this._injector.get(_univerjs_sheets_note.SheetsNoteModel).getNote(this.getUnitId(), this.getSheetId(), {
			row: this.getRow(),
			col: this.getColumn()
		});
	}
};
_univerjs_sheets_facade.FRange.extend(FRangeSheetsNoteMixin);

//#endregion
//#region src/facade/f-univer.ts
var FUniverSheetsNoteMixin = class extends _univerjs_core_facade.FUniver {
	_initialize(injector) {
		const commandService = injector.get(_univerjs_core.ICommandService);
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetNoteAdd, () => {
			return injector.get(_univerjs_sheets_note.SheetsNoteModel).change$.subscribe((change) => {
				if (change.type === "update" && !change.oldNote && change.newNote) {
					const { unitId, subUnitId, newNote } = change;
					const target = this.getSheetTarget(unitId, subUnitId);
					if (!target) return;
					const { workbook, worksheet } = target;
					const eventParams = {
						workbook,
						worksheet,
						row: newNote.row,
						col: newNote.col,
						note: newNote
					};
					this.fireEvent(this.Event.SheetNoteAdd, eventParams);
				}
			});
		}));
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetNoteDelete, () => {
			return injector.get(_univerjs_sheets_note.SheetsNoteModel).change$.subscribe((change) => {
				if (change.type === "update" && change.oldNote && !change.newNote) {
					const { unitId, subUnitId, oldNote } = change;
					const target = this.getSheetTarget(unitId, subUnitId);
					if (!target) return;
					const { workbook, worksheet } = target;
					const eventParams = {
						workbook,
						worksheet,
						row: oldNote.row,
						col: oldNote.col,
						oldNote
					};
					this.fireEvent(this.Event.SheetNoteDelete, eventParams);
				}
			});
		}));
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetNoteUpdate, () => {
			return injector.get(_univerjs_sheets_note.SheetsNoteModel).change$.subscribe((change) => {
				if (change.type === "update" && change.oldNote && change.newNote) {
					const { unitId, subUnitId, oldNote, newNote } = change;
					const target = this.getSheetTarget(unitId, subUnitId);
					if (!target) return;
					const { workbook, worksheet } = target;
					const eventParams = {
						workbook,
						worksheet,
						row: newNote.row,
						col: newNote.col,
						note: newNote,
						oldNote
					};
					this.fireEvent(this.Event.SheetNoteUpdate, eventParams);
				}
			});
		}));
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetNoteShow, () => {
			return injector.get(_univerjs_sheets_note.SheetsNoteModel).change$.subscribe((change) => {
				if (change.type === "update" && change.oldNote && change.newNote && !change.oldNote.show && change.newNote.show) {
					const { unitId, subUnitId, newNote } = change;
					const target = this.getSheetTarget(unitId, subUnitId);
					if (!target) return;
					const { workbook, worksheet } = target;
					const eventParams = {
						workbook,
						worksheet,
						row: newNote.row,
						col: newNote.col
					};
					this.fireEvent(this.Event.SheetNoteShow, eventParams);
				}
			});
		}));
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetNoteHide, () => {
			return injector.get(_univerjs_sheets_note.SheetsNoteModel).change$.subscribe((change) => {
				if (change.type === "update" && change.oldNote && change.newNote && change.oldNote.show && !change.newNote.show) {
					const { unitId, subUnitId, newNote } = change;
					const target = this.getSheetTarget(unitId, subUnitId);
					if (!target) return;
					const { workbook, worksheet } = target;
					const eventParams = {
						workbook,
						worksheet,
						row: newNote.row,
						col: newNote.col
					};
					this.fireEvent(this.Event.SheetNoteHide, eventParams);
				}
			});
		}));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetNoteAdd, () => commandService.beforeCommandExecuted((command) => {
			if (command.id === _univerjs_sheets_note.SheetUpdateNoteCommand.id) {
				const { unitId, sheetId, row, col, note } = command.params;
				const target = this.getSheetTarget(unitId, sheetId);
				if (!target) return;
				if (injector.get(_univerjs_sheets_note.SheetsNoteModel).getNote(unitId, sheetId, {
					noteId: note.id,
					row,
					col
				})) return;
				const { workbook, worksheet } = target;
				const eventParams = {
					workbook,
					worksheet,
					row,
					col,
					note
				};
				if (this.fireEvent(this.Event.BeforeSheetNoteAdd, eventParams)) throw new _univerjs_core.CanceledError();
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetNoteDelete, () => commandService.beforeCommandExecuted((command) => {
			if (command.id === _univerjs_sheets_note.SheetDeleteNoteCommand.id) {
				const { unitId, sheetId, row, col } = command.params;
				const target = this.getSheetTarget(unitId, sheetId);
				if (!target) return;
				if (row === void 0 || col === void 0) return;
				const oldNote = injector.get(_univerjs_sheets_note.SheetsNoteModel).getNote(unitId, sheetId, {
					row,
					col
				});
				if (!oldNote) return;
				const { workbook, worksheet } = target;
				const eventParams = {
					workbook,
					worksheet,
					row,
					col,
					oldNote
				};
				if (this.fireEvent(this.Event.BeforeSheetNoteDelete, eventParams)) throw new _univerjs_core.CanceledError();
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetNoteUpdate, () => commandService.beforeCommandExecuted((command) => {
			if (command.id === _univerjs_sheets_note.SheetUpdateNoteCommand.id) {
				const { unitId, sheetId, row, col, note } = command.params;
				const target = this.getSheetTarget(unitId, sheetId);
				if (!target) return;
				const oldNote = injector.get(_univerjs_sheets_note.SheetsNoteModel).getNote(unitId, sheetId, {
					row,
					col
				});
				if (!oldNote) return;
				const { workbook, worksheet } = target;
				const eventParams = {
					workbook,
					worksheet,
					row,
					col,
					note,
					oldNote
				};
				if (this.fireEvent(this.Event.BeforeSheetNoteUpdate, eventParams)) throw new _univerjs_core.CanceledError();
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetNoteShow, () => commandService.beforeCommandExecuted((command) => {
			if (command.id === _univerjs_sheets_note.SheetToggleNotePopupCommand.id) {
				const target = (0, _univerjs_sheets.getSheetCommandTarget)(injector.get(_univerjs_core.IUniverInstanceService));
				if (!target) return;
				const { unitId, subUnitId } = target;
				const workbook = this.getUniverSheet(unitId);
				if (!workbook) return;
				const worksheet = workbook.getSheetBySheetId(subUnitId);
				if (!worksheet) return;
				const selection = injector.get(_univerjs_sheets.SheetsSelectionsService).getCurrentLastSelection();
				if (!(selection === null || selection === void 0 ? void 0 : selection.primary)) return;
				const sheetsNoteModel = injector.get(_univerjs_sheets_note.SheetsNoteModel);
				const { actualColumn, actualRow } = selection.primary;
				const note = sheetsNoteModel.getNote(unitId, subUnitId, {
					row: actualRow,
					col: actualColumn
				});
				if (!note || note.show) return;
				const eventParams = {
					workbook,
					worksheet,
					row: actualRow,
					col: actualColumn
				};
				if (this.fireEvent(this.Event.BeforeSheetNoteShow, eventParams)) throw new _univerjs_core.CanceledError();
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetNoteHide, () => commandService.beforeCommandExecuted((command) => {
			if (command.id === _univerjs_sheets_note.SheetToggleNotePopupCommand.id) {
				const target = (0, _univerjs_sheets.getSheetCommandTarget)(injector.get(_univerjs_core.IUniverInstanceService));
				if (!target) return;
				const { unitId, subUnitId } = target;
				const workbook = this.getUniverSheet(unitId);
				if (!workbook) return;
				const worksheet = workbook.getSheetBySheetId(subUnitId);
				if (!worksheet) return;
				const selection = injector.get(_univerjs_sheets.SheetsSelectionsService).getCurrentLastSelection();
				if (!(selection === null || selection === void 0 ? void 0 : selection.primary)) return;
				const sheetsNoteModel = injector.get(_univerjs_sheets_note.SheetsNoteModel);
				const { actualColumn, actualRow } = selection.primary;
				const note = sheetsNoteModel.getNote(unitId, subUnitId, {
					row: actualRow,
					col: actualColumn
				});
				if (!note || !note.show) return;
				const eventParams = {
					workbook,
					worksheet,
					row: actualRow,
					col: actualColumn
				};
				if (this.fireEvent(this.Event.BeforeSheetNoteHide, eventParams)) throw new _univerjs_core.CanceledError();
			}
		})));
	}
};
_univerjs_core_facade.FUniver.extend(FUniverSheetsNoteMixin);

//#endregion
//#region src/facade/f-worksheet.ts
var FWorksheetNoteMixin = class extends _univerjs_sheets_facade.FWorksheet {
	getNotes() {
		const notes = this._injector.get(_univerjs_sheets_note.SheetsNoteModel).getSheetNotes(this.getWorkbook().getUnitId(), this.getSheetId());
		if (!notes) return [];
		return Array.from(notes.values()).map((note) => ({ ...note }));
	}
};
_univerjs_sheets_facade.FWorksheet.extend(FWorksheetNoteMixin);

//#endregion
exports.FRangeSheetsNoteMixin = FRangeSheetsNoteMixin;
exports.FSheetsNoteEventNameMixin = FSheetsNoteEventNameMixin;
exports.FUniverSheetsNoteMixin = FUniverSheetsNoteMixin;
exports.FWorksheetNoteMixin = FWorksheetNoteMixin;
