// @univerjs/ui/lib/facade.js
import { FBase, FEnum, FHooks, FUniver } from "@univerjs/core/facade";
import { IRenderManagerService } from "@univerjs/engine-render";
import { BuiltInUIPart, ComponentManager, CopyCommand, IDialogService, IFontService, IMenuManagerService, IMessageService, IShortcutService, ISidebarService, IUIPartsService, KeyCode, MenuItemType, MenuManagerPosition, PasteCommand, RibbonPosition, RibbonStartGroup, SheetPasteShortKeyCommandName, connectInjector } from "@univerjs/ui";
import { CommandType, ICommandService, IUniverInstanceService, Inject, Injector, UniverInstanceType, generateRandomId } from "@univerjs/core";

//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/facade/f-menu-builder.ts
/**
* @ignore
*/
var FMenuBase = class extends FBase {
	/**
	* Append the menu to any menu position on Univer UI.
	* @param {string | string[]} path - Some predefined path to append the menu. The paths can be an array,
	* or an array joined by `|` separator. Since lots of submenus reuse the same name,
	* you may need to specify their parent menus as well.
	*
	* @example
	* ```typescript
	* // This menu item will appear on every `contextMenu.others` section.
	* univerAPI.createMenu({
	*   id: 'custom-menu-id-1',
	*   title: 'Custom Menu 1',
	*   action: () => {
	*     console.log('Custom Menu 1 clicked');
	*   },
	* }).appendTo('contextMenu.others');
	*
	* // This menu item will only appear on the `contextMenu.others` section on the main area.
	* univerAPI.createMenu({
	*   id: 'custom-menu-id-2',
	*   title: 'Custom Menu 2',
	*   action: () => {
	*     console.log('Custom Menu 2 clicked');
	*   },
	* }).appendTo(['contextMenu.mainArea', 'contextMenu.others']);
	* ```
	*/
	appendTo(path) {
		const paths = typeof path === "string" ? path.split("|") : path;
		const len = paths.length;
		const menuConfig = {};
		let obj = menuConfig;
		const schema = this.__getSchema();
		paths.forEach((p, index) => {
			if (index === len - 1) obj[p] = schema;
			else obj[p] = {};
			obj = obj[p];
		});
		this._menuManagerService.mergeMenu(menuConfig);
	}
};
let FMenu = class FMenu extends FMenuBase {
	constructor(_item, _injector, _commandService, _menuManagerService) {
		super();
		this._item = _item;
		this._injector = _injector;
		this._commandService = _commandService;
		this._menuManagerService = _menuManagerService;
		_defineProperty(this, "_commandToRegister", /* @__PURE__ */ new Map());
		_defineProperty(this, "_buildingSchema", void 0);
		const commandId = typeof _item.action === "string" ? _item.action : generateRandomId(12);
		if (commandId !== _item.action) this._commandToRegister.set(commandId, _item.action);
		this._buildingSchema = { menuItemFactory: () => ({
			id: _item.id,
			type: MenuItemType.BUTTON,
			icon: _item.icon,
			title: _item.title,
			tooltip: _item.tooltip,
			commandId
		}) };
		if (typeof _item.order !== "undefined") this._buildingSchema.order = _item.order;
	}
	/**
	* @ignore
	*/
	__getSchema() {
		this._commandToRegister.forEach((command, id) => {
			if (!this._commandService.hasCommand(id)) this._commandService.registerCommand({
				id,
				type: CommandType.COMMAND,
				handler: command
			});
		});
		return { [this._item.id]: this._buildingSchema };
	}
};
_defineProperty(FMenu, "RibbonStartGroup", RibbonStartGroup);
_defineProperty(FMenu, "RibbonPosition", RibbonPosition);
_defineProperty(FMenu, "MenuManagerPosition", MenuManagerPosition);
FMenu = __decorate([
	__decorateParam(1, Inject(Injector)),
	__decorateParam(2, ICommandService),
	__decorateParam(3, IMenuManagerService)
], FMenu);
let FSubmenu = class FSubmenu extends FMenuBase {
	constructor(_item, _injector, _menuManagerService) {
		super();
		this._item = _item;
		this._injector = _injector;
		this._menuManagerService = _menuManagerService;
		_defineProperty(this, "_menuByGroups", []);
		_defineProperty(this, "_submenus", []);
		_defineProperty(this, "_buildingSchema", void 0);
		this._buildingSchema = { menuItemFactory: () => ({
			id: _item.id,
			type: MenuItemType.SUBITEMS,
			icon: _item.icon,
			title: _item.title,
			tooltip: _item.tooltip
		}) };
		if (typeof _item.order !== "undefined") this._buildingSchema.order = _item.order;
	}
	/**
	* Add a menu to the submenu. It can be a {@link FMenu} or a {@link FSubmenu}.
	* @param {FMenu | FSubmenu} submenu - Menu to add to the submenu.
	* @returns {FSubmenu} The FSubmenu itself for chaining calls.
	* @example
	* ```typescript
	* // Create two leaf menus.
	* const menu1 = univerAPI.createMenu({
	*   id: 'submenu-nested-1',
	*   title: 'Item 1',
	*   action: () => {
	*     console.log('Item 1 clicked');
	*   }
	* });
	* const menu2 = univerAPI.createMenu({
	*   id: 'submenu-nested-2',
	*   title: 'Item 2',
	*   action: () => {
	*     console.log('Item 2 clicked');
	*   }
	* });
	*
	* // Add the leaf menus to a submenu.
	* const submenu = univerAPI.createSubmenu({ id: 'submenu-nested', title: 'Nested Submenu' })
	*   .addSubmenu(menu1)
	*   .addSeparator()
	*   .addSubmenu(menu2);
	*
	* // Create a root submenu append to the `contextMenu.others` section.
	* univerAPI.createSubmenu({ id: 'custom-submenu', title: 'Custom Submenu' })
	*   .addSubmenu(submenu)
	*   .appendTo('contextMenu.others');
	* ```
	*/
	addSubmenu(submenu) {
		this._submenus.push(submenu);
		return this;
	}
	/**
	* Add a separator to the submenu.
	* @returns {FSubmenu} The FSubmenu itself for chaining calls.
	* @example
	* ```typescript
	* // Create two leaf menus.
	* const menu1 = univerAPI.createMenu({
	*   id: 'submenu-nested-1',
	*   title: 'Item 1',
	*   action: () => {
	*     console.log('Item 1 clicked');
	*   }
	* });
	* const menu2 = univerAPI.createMenu({
	*   id: 'submenu-nested-2',
	*   title: 'Item 2',
	*   action: () => {
	*     console.log('Item 2 clicked');
	*   }
	* });
	*
	* // Add the leaf menus to a submenu and add a separator between them.
	* // Append the submenu to the `contextMenu.others` section.
	* univerAPI.createSubmenu({ id: 'submenu-nested', title: 'Nested Submenu' })
	*   .addSubmenu(menu1)
	*   .addSeparator()
	*   .addSubmenu(menu2)
	*   .appendTo('contextMenu.others');
	* ```
	*/
	addSeparator() {
		this._menuByGroups.push(this._submenus);
		this._submenus = [];
		return this;
	}
	/**
	* @ignore
	*/
	__getSchema() {
		const schema = {};
		this.addSeparator();
		this._menuByGroups.forEach((group, index) => {
			const groupSchema = {};
			group.forEach((menu) => {
				Object.assign(groupSchema, menu.__getSchema());
			});
			schema[`${this._item.id}-group-${index}`] = groupSchema;
		});
		return { [this._item.id]: Object.assign(this._buildingSchema, schema) };
	}
};
FSubmenu = __decorate([__decorateParam(1, Inject(Injector)), __decorateParam(2, IMenuManagerService)], FSubmenu);

//#endregion
//#region src/facade/f-shortcut.ts
let FShortcut = class FShortcut extends FBase {
	constructor(_injector, _renderManagerService, _univerInstanceService, _shortcutService) {
		super();
		this._injector = _injector;
		this._renderManagerService = _renderManagerService;
		this._univerInstanceService = _univerInstanceService;
		this._shortcutService = _shortcutService;
		_defineProperty(this, "_forceDisableDisposable", null);
	}
	/**
	* Enable shortcuts of Univer.
	* @returns {FShortcut} The Facade API instance itself for chaining.
	*
	* @example
	* ```typescript
	* fShortcut.enableShortcut(); // Use the FShortcut instance used by disableShortcut before, do not create a new instance
	* ```
	*/
	enableShortcut() {
		var _this$_forceDisableDi;
		(_this$_forceDisableDi = this._forceDisableDisposable) === null || _this$_forceDisableDi === void 0 || _this$_forceDisableDi.dispose();
		this._forceDisableDisposable = null;
		return this;
	}
	/**
	* Disable shortcuts of Univer.
	* @returns {FShortcut} The Facade API instance itself for chaining.
	*
	* @example
	* ```typescript
	* const fShortcut = univerAPI.getShortcut();
	* fShortcut.disableShortcut();
	* ```
	*/
	disableShortcut() {
		if (!this._forceDisableDisposable) this._forceDisableDisposable = this._shortcutService.forceDisable();
		return this;
	}
	/**
	* Trigger shortcut of Univer by a KeyboardEvent and return the matched shortcut item.
	* @param {KeyboardEvent} e - The KeyboardEvent to trigger.
	* @returns {IShortcutItem<object> | undefined} The matched shortcut item.
	*
	* @example
	* ```typescript
	* // Assum the current sheet is empty sheet.
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const fRange = fWorksheet.getRange('A1');
	*
	* // Set A1 cell active and set value to 'Hello Univer'.
	* fRange.activate();
	* fRange.setValue('Hello Univer');
	* console.log(fRange.getCellStyle().bold); // false
	*
	* // Set A1 cell bold by shortcut.
	* const fShortcut = univerAPI.getShortcut();
	* const pseudoEvent = new KeyboardEvent('keydown', {
	*   key: 'b',
	*   ctrlKey: true,
	*   keyCode: univerAPI.Enum.KeyCode.B
	* });
	* const ifShortcutItem = fShortcut.triggerShortcut(pseudoEvent);
	* if (ifShortcutItem) {
	*   const commandId = ifShortcutItem.id;
	*   console.log(fRange.getCellStyle().bold); // true
	* }
	* ```
	*/
	triggerShortcut(e) {
		const workbook = this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_SHEET);
		if (!workbook) return;
		const renderUnit = this._renderManagerService.getRenderById(workbook.getUnitId());
		if (!renderUnit) return;
		renderUnit.engine.getCanvasElement().dispatchEvent(e);
		return this._shortcutService.dispatch(e);
	}
	/**
	* Dispatch a KeyboardEvent to the shortcut service and return the matched shortcut item.
	* @param {KeyboardEvent} e - The KeyboardEvent to dispatch.
	* @returns {IShortcutItem<object> | undefined} The matched shortcut item.
	*
	* @example
	* ```typescript
	* const fShortcut = univerAPI.getShortcut();
	* const pseudoEvent = new KeyboardEvent('keydown', { key: 's', ctrlKey: true });
	* const ifShortcutItem = fShortcut.dispatchShortcutEvent(pseudoEvent);
	* if (ifShortcutItem) {
	*   const commandId = ifShortcutItem.id;
	*   // Do something with the commandId.
	* }
	* ```
	*/
	dispatchShortcutEvent(e) {
		return this._shortcutService.dispatch(e);
	}
};
FShortcut = __decorate([
	__decorateParam(0, Inject(Injector)),
	__decorateParam(1, Inject(IRenderManagerService)),
	__decorateParam(2, IUniverInstanceService),
	__decorateParam(3, IShortcutService)
], FShortcut);

//#endregion
//#region src/facade/f-univer.ts
/**
* @ignore
*/
var FUniverUIMixin = class extends FUniver {
	getURL() {
		return new URL(window.location.href);
	}
	getShortcut() {
		return this._injector.createInstance(FShortcut);
	}
	copy() {
		return this._commandService.executeCommand(CopyCommand.id);
	}
	paste() {
		return this._commandService.executeCommand(PasteCommand.id);
	}
	createMenu(menuItem) {
		return this._injector.createInstance(FMenu, menuItem);
	}
	createSubmenu(submenuItem) {
		return this._injector.createInstance(FSubmenu, submenuItem);
	}
	openSiderbar(params) {
		return this._injector.get(ISidebarService).open(params);
	}
	openSidebar(params) {
		return this.openSiderbar(params);
	}
	openDialog(dialog) {
		const disposable = this._injector.get(IDialogService).open({
			...dialog,
			onClose: () => {
				disposable.dispose();
			}
		});
		return disposable;
	}
	getComponentManager() {
		return this._injector.get(ComponentManager);
	}
	showMessage(options) {
		this._injector.get(IMessageService).show(options);
		return this;
	}
	setUIVisible(ui, visible) {
		this._injector.get(IUIPartsService).setUIVisible(ui, visible);
		return this;
	}
	isUIVisible(ui) {
		return this._injector.get(IUIPartsService).isUIVisible(ui);
	}
	registerUIPart(key, component) {
		return this._injector.get(IUIPartsService).registerComponent(key, () => connectInjector(component, this._injector));
	}
	registerComponent(name, component, options) {
		const componentManager = this._injector.get(ComponentManager);
		return this.disposeWithMe(componentManager.register(name, component, options));
	}
	setCurrent(unitId) {
		if (!this._injector.get(IRenderManagerService).getRenderById(unitId)) throw new Error("Unit not found");
		this._univerInstanceService.setCurrentUnitForType(unitId);
	}
	addFonts(fonts) {
		const fontService = this._injector.get(IFontService);
		fonts.forEach((font) => {
			fontService.addFont(font);
		});
	}
};
FUniver.extend(FUniverUIMixin);

//#endregion
//#region src/facade/f-hooks.ts
/**
* @ignore
*/
var FHooksSheetsMixin = class extends FHooks {
	onBeforeCopy(callback) {
		return this._injector.get(ICommandService).beforeCommandExecuted((command) => {
			if (command.id === CopyCommand.id) callback();
		});
	}
	onCopy(callback) {
		return this._injector.get(ICommandService).onCommandExecuted((command) => {
			if (command.id === CopyCommand.id) callback();
		});
	}
	onBeforePaste(callback) {
		return this._injector.get(ICommandService).beforeCommandExecuted((command) => {
			if (command.id === PasteCommand.id) callback();
		});
	}
	onPaste(callback) {
		return this._injector.get(ICommandService).onCommandExecuted((command) => {
			if (command.id === PasteCommand.id || command.id === SheetPasteShortKeyCommandName) callback();
		});
	}
};
FHooks.extend(FHooksSheetsMixin);

//#endregion
//#region src/facade/f-enum.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
/**
* @ignore
*/
var FUIEnumMixin = class extends FEnum {
	get BuiltInUIPart() {
		return BuiltInUIPart;
	}
	get KeyCode() {
		return KeyCode;
	}
};
FEnum.extend(FUIEnumMixin);

//#endregion
export {  };
