// @univerjs/thread-comment/lib/cjs/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
let _univerjs_core = require("@univerjs/core");
let rxjs = require("rxjs");

//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region src/services/tc-datasource.service.ts
/**
* Preserve for import async comment system
*/
var ThreadCommentDataSourceService = class extends _univerjs_core.Disposable {
	set dataSource(dataSource) {
		this._dataSource = dataSource;
	}
	get dataSource() {
		return this._dataSource;
	}
	constructor() {
		super();
		_defineProperty(this, "_dataSource", null);
		_defineProperty(this, "syncUpdateMutationToColla", true);
	}
	async getThreadComment(unitId, subUnitId, threadId) {
		if (this._dataSource) return (await this._dataSource.listComments(unitId, subUnitId, [threadId]))[0];
		return null;
	}
	async addComment(comment) {
		const savedComment = this._dataSource ? await this._dataSource.addComment(comment) : comment;
		return {
			...savedComment,
			threadId: savedComment.threadId || savedComment.id
		};
	}
	async updateComment(comment) {
		if (this._dataSource) return this._dataSource.updateComment(comment);
		return true;
	}
	async resolveComment(comment) {
		if (this._dataSource) return this._dataSource.resolveComment(comment);
		return true;
	}
	async deleteComment(unitId, subUnitId, threadId, commentId) {
		if (this._dataSource) return this._dataSource.deleteComment(unitId, subUnitId, threadId, commentId);
		return true;
	}
	async listThreadComments(unitId, subUnitId, threadIds) {
		if (this.dataSource) return this.dataSource.listComments(unitId, subUnitId, threadIds);
		return false;
	}
	saveToSnapshot(unitComments, unitId) {
		if (this._dataSource) {
			const map = {};
			Object.keys(unitComments).forEach((subUnitId) => {
				map[subUnitId] = unitComments[subUnitId].map(this.dataSource.saveCommentToSnapshot);
			});
			return map;
		}
		return unitComments;
	}
};
const IThreadCommentDataSourceService = (0, _univerjs_core.createIdentifier)("univer.thread-comment.data-source-service");

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/models/thread-comment.model.ts
let ThreadCommentModel = class ThreadCommentModel extends _univerjs_core.Disposable {
	constructor(_dataSourceService, _lifecycleService) {
		super();
		this._dataSourceService = _dataSourceService;
		this._lifecycleService = _lifecycleService;
		_defineProperty(this, "_commentsMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_threadMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_commentUpdate$", new rxjs.Subject());
		_defineProperty(this, "commentUpdate$", this._commentUpdate$.asObservable());
		_defineProperty(this, "_tasks", []);
		this.disposeWithMe(() => {
			this._commentUpdate$.complete();
		});
		this.disposeWithMe(this._lifecycleService.lifecycle$.subscribe((stage) => {
			const taskMap = /* @__PURE__ */ new Map();
			if (stage === _univerjs_core.LifecycleStages.Rendered) {
				this._tasks.forEach(({ unitId, subUnitId, threadIds }) => {
					let unitMap = taskMap.get(unitId);
					if (!unitMap) {
						unitMap = /* @__PURE__ */ new Map();
						taskMap.set(unitId, unitMap);
					}
					let subUnitMap = unitMap.get(subUnitId);
					if (!subUnitMap) {
						subUnitMap = /* @__PURE__ */ new Set();
						unitMap.set(subUnitId, subUnitMap);
					}
					for (const threadId of threadIds) subUnitMap.add(threadId);
				});
				this._tasks = [];
				taskMap.forEach((subUnitMap, unitId) => {
					subUnitMap.forEach((threadIds, subUnitId) => {
						this.syncThreadComments(unitId, subUnitId, Array.from(threadIds));
					});
				});
			}
		}));
	}
	_ensureCommentMap(unitId, subUnitId) {
		let unitMap = this._commentsMap.get(unitId);
		if (!unitMap) {
			unitMap = /* @__PURE__ */ new Map();
			this._commentsMap.set(unitId, unitMap);
		}
		let subUnitMap = unitMap.get(subUnitId);
		if (!subUnitMap) {
			subUnitMap = /* @__PURE__ */ new Map();
			unitMap.set(subUnitId, subUnitMap);
		}
		return subUnitMap;
	}
	ensureMap(unitId, subUnitId) {
		return this._ensureCommentMap(unitId, subUnitId);
	}
	_ensureThreadMap(unitId, subUnitId) {
		let unitMap = this._threadMap.get(unitId);
		if (!unitMap) {
			unitMap = /* @__PURE__ */ new Map();
			this._threadMap.set(unitId, unitMap);
		}
		let subUnitMap = unitMap.get(subUnitId);
		if (!subUnitMap) {
			subUnitMap = /* @__PURE__ */ new Map();
			unitMap.set(subUnitId, subUnitMap);
		}
		return subUnitMap;
	}
	_replaceComment(unitId, subUnitId, comment) {
		const commentMap = this._ensureCommentMap(unitId, subUnitId);
		const currentComment = commentMap.get(comment.id);
		if (!currentComment) {
			this.addComment(unitId, subUnitId, comment);
			return;
		}
		const { children, ...rest } = comment;
		const newComment = {
			...rest,
			ref: currentComment.ref
		};
		commentMap.set(comment.id, newComment);
		children === null || children === void 0 || children.forEach((child) => {
			commentMap.set(child.id, {
				...child,
				ref: ""
			});
		});
		this._commentUpdate$.next({
			unitId,
			subUnitId,
			type: "syncUpdate",
			payload: newComment
		});
		if (Boolean(comment.resolved) !== Boolean(currentComment.resolved)) this._commentUpdate$.next({
			unitId,
			subUnitId,
			type: "resolve",
			payload: {
				commentId: comment.id,
				resolved: Boolean(comment.resolved)
			}
		});
	}
	async syncThreadComments(unitId, subUnitId, threadIds) {
		if (this._lifecycleService.stage < _univerjs_core.LifecycleStages.Rendered) {
			this._tasks.push({
				unitId,
				subUnitId,
				threadIds
			});
			return;
		}
		const threadMap = this._ensureThreadMap(unitId, subUnitId);
		const commentMap = this._ensureCommentMap(unitId, subUnitId);
		const comments = await this._dataSourceService.listThreadComments(unitId, subUnitId, threadIds);
		if (!comments) return;
		const deleteThreads = new Set(threadIds);
		comments.forEach((comment) => {
			if (!deleteThreads.has(comment.threadId)) return;
			this._replaceComment(unitId, subUnitId, comment);
			deleteThreads.delete(comment.threadId);
		});
		deleteThreads.forEach((id) => {
			threadMap.delete(id);
			commentMap.forEach((comment, commentId) => {
				if (comment.threadId === id) commentMap.delete(commentId);
			});
		});
	}
	addComment(unitId, subUnitId, origin, shouldSync) {
		const commentMap = this._ensureCommentMap(unitId, subUnitId);
		const { parentId, children = [], ...rest } = origin;
		const comment = {
			...rest,
			parentId: parentId === origin.id ? void 0 : parentId
		};
		if (!comment.threadId) comment.threadId = comment.parentId || comment.id;
		const addCommentItem = (item) => {
			commentMap.set(item.id, item);
			this._commentUpdate$.next({
				unitId,
				subUnitId,
				type: "add",
				payload: item,
				isRoot: !item.parentId
			});
		};
		addCommentItem(comment);
		const threadMap = this._ensureThreadMap(unitId, subUnitId);
		if (!comment.parentId) {
			threadMap.set(comment.threadId, comment);
			for (const child of children) addCommentItem(child);
		}
		if (shouldSync) this.syncThreadComments(unitId, subUnitId, [comment.threadId]);
		return true;
	}
	updateComment(unitId, subUnitId, payload, silent) {
		const oldComment = this._ensureCommentMap(unitId, subUnitId).get(payload.commentId);
		if (!oldComment) return true;
		oldComment.updated = true;
		oldComment.text = payload.text;
		oldComment.attachments = payload.attachments;
		oldComment.updateT = payload.updateT;
		this._commentUpdate$.next({
			unitId,
			subUnitId,
			type: "update",
			payload,
			silent
		});
		return true;
	}
	updateCommentRef(unitId, subUnitId, payload, silent) {
		const oldComment = this._ensureCommentMap(unitId, subUnitId).get(payload.commentId);
		if (!oldComment) return false;
		oldComment.ref = payload.ref;
		this._commentUpdate$.next({
			unitId,
			subUnitId,
			type: "updateRef",
			payload,
			silent,
			threadId: oldComment.threadId
		});
		return true;
	}
	resolveComment(unitId, subUnitId, commentId, resolved) {
		const oldComment = this._ensureCommentMap(unitId, subUnitId).get(commentId);
		if (!oldComment) return false;
		oldComment.resolved = resolved;
		this._commentUpdate$.next({
			unitId,
			subUnitId,
			type: "resolve",
			payload: {
				commentId,
				resolved
			}
		});
		return true;
	}
	getComment(unitId, subUnitId, commentId) {
		return this._ensureCommentMap(unitId, subUnitId).get(commentId);
	}
	getRootComment(unitId, subUnitId, threadId) {
		return this._ensureThreadMap(unitId, subUnitId).get(threadId);
	}
	getThread(unitId, subUnitId, threadId) {
		const commentMap = this._ensureCommentMap(unitId, subUnitId);
		const comments = Array.from(commentMap.values()).filter((comment) => comment.threadId === threadId);
		let root;
		const children = [];
		const relativeUsers = /* @__PURE__ */ new Set();
		for (const comment of comments) {
			if (!comment.parentId) root = comment;
			else children.push(comment);
			relativeUsers.add(comment.personId);
		}
		if (!root) return;
		return {
			root,
			children,
			relativeUsers,
			unitId,
			subUnitId,
			threadId
		};
	}
	getCommentWithChildren(unitId, subUnitId, commentId) {
		const comment = this.getComment(unitId, subUnitId, commentId);
		if (!comment) return;
		return this.getThread(unitId, subUnitId, comment.threadId);
	}
	_deleteComment(unitId, subUnitId, commentId) {
		const commentMap = this._ensureCommentMap(unitId, subUnitId);
		const current = commentMap.get(commentId);
		if (!current) return;
		commentMap.delete(commentId);
		this._commentUpdate$.next({
			unitId,
			subUnitId,
			type: "delete",
			payload: {
				commentId,
				isRoot: !current.parentId,
				comment: current
			}
		});
	}
	deleteThread(unitId, subUnitId, threadId) {
		this._ensureThreadMap(unitId, subUnitId).delete(threadId);
		this._ensureCommentMap(unitId, subUnitId).forEach((comment) => {
			if (comment.threadId === threadId) this._deleteComment(unitId, subUnitId, comment.id);
		});
	}
	deleteComment(unitId, subUnitId, commentId) {
		const current = this._ensureCommentMap(unitId, subUnitId).get(commentId);
		if (!current) return true;
		if (current.parentId) this._deleteComment(unitId, subUnitId, commentId);
		else this.deleteThread(unitId, subUnitId, current.threadId);
		return true;
	}
	deleteUnit(unitId) {
		const unitMap = this._commentsMap.get(unitId);
		if (!unitMap) return;
		unitMap.forEach((subUnitMap, subUnitId) => {
			subUnitMap.forEach((comment) => {
				this.deleteComment(unitId, subUnitId, comment.id);
			});
		});
	}
	getUnit(unitId) {
		const unitMap = this._threadMap.get(unitId);
		if (!unitMap) return [];
		const threads = [];
		unitMap.forEach((subUnitSet, subUnitId) => {
			subUnitSet.forEach((threadComment, threadId) => {
				const thread = this.getThread(unitId, subUnitId, threadId);
				if (thread) threads.push(thread);
			});
		});
		return threads;
	}
	getAll() {
		const all = [];
		this._commentsMap.forEach((unitMap, unitId) => {
			all.push({
				unitId,
				threads: this.getUnit(unitId)
			});
		});
		return all;
	}
};
ThreadCommentModel = __decorate([__decorateParam(0, (0, _univerjs_core.Inject)(IThreadCommentDataSourceService)), __decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.LifecycleService))], ThreadCommentModel);

//#endregion
//#region src/commands/mutations/comment.mutation.ts
const AddCommentMutation = {
	id: "thread-comment.mutation.add-comment",
	type: _univerjs_core.CommandType.MUTATION,
	handler(accessor, params, options) {
		if (!params) return false;
		const threadCommentModel = accessor.get(ThreadCommentModel);
		const { unitId, subUnitId, comment, sync } = params;
		const shouldSync = sync || (options === null || options === void 0 ? void 0 : options.fromChangeset) && !comment.parentId;
		return threadCommentModel.addComment(unitId, subUnitId, comment, shouldSync);
	}
};
const UpdateCommentMutation = {
	id: "thread-comment.mutation.update-comment",
	type: _univerjs_core.CommandType.MUTATION,
	handler(accessor, params) {
		if (!params) return false;
		const threadCommentModel = accessor.get(ThreadCommentModel);
		const { unitId, subUnitId, payload, silent } = params;
		return threadCommentModel.updateComment(unitId, subUnitId, payload, silent);
	}
};
const UpdateCommentRefMutation = {
	id: "thread-comment.mutation.update-comment-ref",
	type: _univerjs_core.CommandType.MUTATION,
	handler(accessor, params) {
		if (!params) return false;
		const threadCommentModel = accessor.get(ThreadCommentModel);
		const { unitId, subUnitId, payload, silent } = params;
		return threadCommentModel.updateCommentRef(unitId, subUnitId, payload, silent);
	}
};
const ResolveCommentMutation = {
	id: "thread-comment.mutation.resolve-comment",
	type: _univerjs_core.CommandType.MUTATION,
	handler(accessor, params) {
		if (!params) return false;
		const threadCommentModel = accessor.get(ThreadCommentModel);
		const { unitId, subUnitId, resolved, commentId } = params;
		return threadCommentModel.resolveComment(unitId, subUnitId, commentId, resolved);
	}
};
const DeleteCommentMutation = {
	id: "thread-comment.mutation.delete-comment",
	type: _univerjs_core.CommandType.MUTATION,
	handler(accessor, params) {
		if (!params) return false;
		const threadCommentModel = accessor.get(ThreadCommentModel);
		const { unitId, subUnitId, commentId } = params;
		return threadCommentModel.deleteComment(unitId, subUnitId, commentId);
	}
};

//#endregion
//#region src/commands/commands/comment.command.ts
const AddCommentCommand = {
	id: "thread-comment.command.add-comment",
	type: _univerjs_core.CommandType.COMMAND,
	async handler(accessor, params) {
		if (!params) return false;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const dataSourceService = accessor.get(IThreadCommentDataSourceService);
		const { comment: originComment } = params;
		const comment = await dataSourceService.addComment(originComment);
		const syncUpdateMutationToColla = dataSourceService.syncUpdateMutationToColla;
		const isRoot = !originComment.parentId;
		const redo = {
			id: AddCommentMutation.id,
			params: {
				...params,
				comment
			}
		};
		if (isRoot) return await commandService.executeCommand(redo.id, redo.params);
		return commandService.executeCommand(redo.id, redo.params, { onlyLocal: !syncUpdateMutationToColla });
	}
};
const UpdateCommentCommand = {
	id: "thread-comment.command.update-comment",
	type: _univerjs_core.CommandType.COMMAND,
	async handler(accessor, params) {
		if (!params) return false;
		const { unitId, subUnitId, payload } = params;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const threadCommentModel = accessor.get(ThreadCommentModel);
		const dataSourceService = accessor.get(IThreadCommentDataSourceService);
		const syncUpdateMutationToColla = dataSourceService.syncUpdateMutationToColla;
		const current = threadCommentModel.getComment(unitId, subUnitId, payload.commentId);
		if (!current) return false;
		const { children, ...currentComment } = current;
		if (!await dataSourceService.updateComment({
			...currentComment,
			...payload
		})) return false;
		const redo = {
			id: UpdateCommentMutation.id,
			params
		};
		commandService.executeCommand(redo.id, redo.params, { onlyLocal: !syncUpdateMutationToColla });
		return true;
	}
};
const ResolveCommentCommand = {
	id: "thread-comment.command.resolve-comment",
	type: _univerjs_core.CommandType.COMMAND,
	async handler(accessor, params) {
		if (!params) return false;
		const { unitId, subUnitId, resolved, commentId } = params;
		const dataSourceService = accessor.get(IThreadCommentDataSourceService);
		const currentComment = accessor.get(ThreadCommentModel).getComment(unitId, subUnitId, commentId);
		const syncUpdateMutationToColla = dataSourceService.syncUpdateMutationToColla;
		if (!currentComment) return false;
		if (!await dataSourceService.resolveComment({
			...currentComment,
			resolved
		})) return false;
		return accessor.get(_univerjs_core.ICommandService).executeCommand(ResolveCommentMutation.id, params, { onlyLocal: !syncUpdateMutationToColla });
	}
};
/**
* Delete Reply
*/
const DeleteCommentCommand = {
	id: "thread-comment.command.delete-comment",
	type: _univerjs_core.CommandType.COMMAND,
	async handler(accessor, params) {
		if (!params) return false;
		const threadCommentModel = accessor.get(ThreadCommentModel);
		const dataSourceService = accessor.get(IThreadCommentDataSourceService);
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const { unitId, subUnitId, commentId } = params;
		const syncUpdateMutationToColla = dataSourceService.syncUpdateMutationToColla;
		const comment = threadCommentModel.getComment(unitId, subUnitId, commentId);
		if (!comment) return false;
		if (!await dataSourceService.deleteComment(unitId, subUnitId, comment.threadId, commentId)) return false;
		const redo = {
			id: DeleteCommentMutation.id,
			params
		};
		return commandService.executeCommand(redo.id, redo.params, { onlyLocal: !syncUpdateMutationToColla });
	}
};
const DeleteCommentTreeCommand = {
	id: "thread-comment.command.delete-comment-tree",
	type: _univerjs_core.CommandType.COMMAND,
	async handler(accessor, params) {
		if (!params) return false;
		const threadCommentModel = accessor.get(ThreadCommentModel);
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const dataSourceService = accessor.get(IThreadCommentDataSourceService);
		const { unitId, subUnitId, commentId } = params;
		const commentWithChildren = threadCommentModel.getCommentWithChildren(unitId, subUnitId, commentId);
		if (!commentWithChildren) return false;
		if (!await dataSourceService.deleteComment(unitId, subUnitId, commentWithChildren.root.threadId, commentId)) return false;
		return await commandService.executeCommand(DeleteCommentMutation.id, {
			unitId,
			subUnitId,
			commentId: commentWithChildren.root.id
		});
	}
};

//#endregion
//#region src/common/utils.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
function getDT(date) {
	return (0, _univerjs_core.dateKit)(date).format("YYYY/MM/DD HH:mm");
}

//#endregion
//#region src/types/const/index.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const TC_PLUGIN_NAME = "UNIVER_THREAD_COMMENT_PLUGIN";

//#endregion
//#region src/controllers/tc-resource.controller.ts
const SHEET_UNIVER_THREAD_COMMENT_PLUGIN = `SHEET_${TC_PLUGIN_NAME}`;
let ThreadCommentResourceController = class ThreadCommentResourceController extends _univerjs_core.Disposable {
	constructor(_resourceManagerService, _threadCommentModel, _threadCommentDataSourceService) {
		super();
		this._resourceManagerService = _resourceManagerService;
		this._threadCommentModel = _threadCommentModel;
		this._threadCommentDataSourceService = _threadCommentDataSourceService;
		this._initSnapshot();
	}
	_initSnapshot() {
		const toJson = (unitID) => {
			const map = this._threadCommentModel.getUnit(unitID);
			const resultMap = {};
			if (map) {
				map.forEach((info) => {
					var _resultMap$info$subUn;
					const subUnitComments = (_resultMap$info$subUn = resultMap[info.subUnitId]) !== null && _resultMap$info$subUn !== void 0 ? _resultMap$info$subUn : [];
					subUnitComments.push({
						...info.root,
						children: info.children
					});
					resultMap[info.subUnitId] = subUnitComments;
				});
				return JSON.stringify(this._threadCommentDataSourceService.saveToSnapshot(resultMap, unitID));
			}
			return "";
		};
		const parseJson = (json) => {
			if (!json) return {};
			try {
				return JSON.parse(json);
			} catch (err) {
				return {};
			}
		};
		this.disposeWithMe(this._resourceManagerService.registerPluginResource({
			pluginName: SHEET_UNIVER_THREAD_COMMENT_PLUGIN,
			businesses: [_univerjs_core.UniverInstanceType.UNIVER_SHEET, _univerjs_core.UniverInstanceType.UNIVER_DOC],
			toJson: (unitID) => toJson(unitID),
			parseJson: (json) => parseJson(json),
			onUnLoad: (unitID) => {
				this._threadCommentModel.deleteUnit(unitID);
			},
			onLoad: async (unitID, value) => {
				Object.keys(value).forEach((subunitId) => {
					const commentList = value[subunitId];
					commentList.forEach((comment) => {
						this._threadCommentModel.addComment(unitID, subunitId, comment);
					});
					this._threadCommentModel.syncThreadComments(unitID, subunitId, commentList.map((i) => i.threadId));
				});
			}
		}));
	}
};
ThreadCommentResourceController = __decorate([
	__decorateParam(0, _univerjs_core.IResourceManagerService),
	__decorateParam(1, (0, _univerjs_core.Inject)(ThreadCommentModel)),
	__decorateParam(2, IThreadCommentDataSourceService)
], ThreadCommentResourceController);

//#endregion
//#region package.json
var name = "@univerjs/thread-comment";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
const THREAD_COMMENT_PLUGIN_CONFIG_KEY = "thread-comment.config";
const configSymbol = Symbol(THREAD_COMMENT_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/plugin.ts
let UniverThreadCommentPlugin = class UniverThreadCommentPlugin extends _univerjs_core.Plugin {
	constructor(_config = defaultPluginConfig, _injector, _commandService, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._commandService = _commandService;
		this._configService = _configService;
		const { ...rest } = (0, _univerjs_core.merge)({}, defaultPluginConfig, this._config);
		this._configService.setConfig(THREAD_COMMENT_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		var _this$_config;
		(0, _univerjs_core.mergeOverrideWithDependencies)([
			[IThreadCommentDataSourceService, { useClass: ThreadCommentDataSourceService }],
			[ThreadCommentModel],
			[ThreadCommentResourceController]
		], (_this$_config = this._config) === null || _this$_config === void 0 ? void 0 : _this$_config.overrides).forEach((d) => {
			this._injector.add(d);
		});
		[
			AddCommentCommand,
			UpdateCommentCommand,
			DeleteCommentCommand,
			ResolveCommentCommand,
			DeleteCommentTreeCommand,
			AddCommentMutation,
			UpdateCommentMutation,
			UpdateCommentRefMutation,
			DeleteCommentMutation,
			ResolveCommentMutation
		].forEach((command) => {
			this._commandService.registerCommand(command);
		});
		this._injector.get(ThreadCommentResourceController);
	}
};
_defineProperty(UniverThreadCommentPlugin, "pluginName", TC_PLUGIN_NAME);
_defineProperty(UniverThreadCommentPlugin, "packageName", name);
_defineProperty(UniverThreadCommentPlugin, "version", version);
_defineProperty(UniverThreadCommentPlugin, "type", _univerjs_core.UniverInstanceType.UNIVER_UNKNOWN);
UniverThreadCommentPlugin = __decorate([
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector)),
	__decorateParam(2, _univerjs_core.ICommandService),
	__decorateParam(3, _univerjs_core.IConfigService)
], UniverThreadCommentPlugin);

//#endregion
exports.AddCommentCommand = AddCommentCommand;
exports.AddCommentMutation = AddCommentMutation;
exports.DeleteCommentCommand = DeleteCommentCommand;
exports.DeleteCommentMutation = DeleteCommentMutation;
exports.DeleteCommentTreeCommand = DeleteCommentTreeCommand;
exports.IThreadCommentDataSourceService = IThreadCommentDataSourceService;
exports.ResolveCommentCommand = ResolveCommentCommand;
exports.ResolveCommentMutation = ResolveCommentMutation;
exports.SHEET_UNIVER_THREAD_COMMENT_PLUGIN = SHEET_UNIVER_THREAD_COMMENT_PLUGIN;
exports.TC_PLUGIN_NAME = TC_PLUGIN_NAME;
exports.ThreadCommentDataSourceService = ThreadCommentDataSourceService;
Object.defineProperty(exports, 'ThreadCommentModel', {
  enumerable: true,
  get: function () {
    return ThreadCommentModel;
  }
});
Object.defineProperty(exports, 'ThreadCommentResourceController', {
  enumerable: true,
  get: function () {
    return ThreadCommentResourceController;
  }
});
Object.defineProperty(exports, 'UniverThreadCommentPlugin', {
  enumerable: true,
  get: function () {
    return UniverThreadCommentPlugin;
  }
});
exports.UpdateCommentCommand = UpdateCommentCommand;
exports.UpdateCommentMutation = UpdateCommentMutation;
exports.UpdateCommentRefMutation = UpdateCommentRefMutation;
exports.getDT = getDT;
