// @univerjs/preset-sheets-collaboration/lib/es/index.js
import { UniverCollaborationPlugin as s } from "@univerjs-pro/collaboration";
export * from "@univerjs-pro/collaboration";
import { UniverCollaborationClientPlugin as u } from "@univerjs-pro/collaboration-client";
export * from "@univerjs-pro/collaboration-client";
import { BrowserCollaborationSocketService as p, UniverCollaborationClientUIPlugin as c } from "@univerjs-pro/collaboration-client-ui";
export * from "@univerjs-pro/collaboration-client-ui";
import { UniverEditHistoryLoaderPlugin as m } from "@univerjs-pro/edit-history-loader";
export * from "@univerjs-pro/edit-history-loader";
import "@univerjs-pro/collaboration-client/lib/facade";
import "@univerjs-pro/collaboration-client-ui/lib/facade";
export * from "@univerjs-pro/edit-history-viewer";
function v(o) {
  const e = new URL(o, window.location.origin);
  switch (e.protocol) {
    case "https:":
      e.protocol = "wss:";
      break;
    case "http:":
      e.protocol = "ws:";
      break;
  }
  return e.toString();
}
function w(o = {}) {
  const {
    universerEndpoint: e,
    univerContainerId: i = "app",
    enableOfflineEditing: t = !0,
    enableSingleActiveInstanceLock: n = !0,
    enableFrontendLog: l = !1,
    historyWorkerURL: a
  } = o, r = e != null ? e : `${window.location.protocol}//${window.location.host}`;
  return {
    plugins: [
      s,
      [u, {
        socketService: p,
        enableOfflineEditing: t,
        enableSingleActiveInstanceLock: n,
        enableAuthServer: !0,
        authzUrl: `${r}/universer-api/authz`,
        snapshotServerUrl: `${r}/universer-api/snapshot`,
        collabSubmitChangesetUrl: `${r}/universer-api/comb`,
        collabWebSocketUrl: v(`${r}/universer-api/comb/connect`),
        loginUrlKey: `${r}/universer-api/oidc/authpage`,
        uploadFileServerUrl: `${r}/universer-api/stream/file/upload`,
        signUrlServerUrl: `${r}/universer-api/file/{fileID}/sign-url`,
        downloadEndpointUrl: `${r}/`,
        wsSessionTicketUrl: `${r}/universer-api/user/session-ticket`,
        startFormulaLimitUrl: `${r}/universer-api/license/formula/limit/start`,
        getFormulaLimitStatusUrl: `${r}/universer-api/license/formula/limit/status`,
        releaseFormulaLimitUrl: `${r}/universer-api/license/formula/limit/done`,
        sendChangesetTimeout: 200
      }],
      [c, {
        enableFrontendLog: l
      }],
      [m, {
        univerContainerId: i,
        historyListServerUrl: `${r}/universer-api/history`,
        workerURL: a
      }]
    ]
  };
}
export {
  w as UniverSheetsCollaborationPreset
};
