// @univerjs/data-validation/lib/cjs/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
let _univerjs_core = require("@univerjs/core");
let rxjs = require("rxjs");

//#region src/common/util.ts
function getRuleSetting(rule) {
	return {
		type: rule.type,
		operator: rule.operator,
		formula1: rule.formula1,
		formula2: rule.formula2,
		allowBlank: rule.allowBlank
	};
}
function getRuleOptions(rule) {
	return {
		error: rule.error,
		errorStyle: rule.errorStyle,
		errorTitle: rule.errorTitle,
		imeMode: rule.imeMode,
		prompt: rule.prompt,
		promptTitle: rule.promptTitle,
		showDropDown: rule.showDropDown,
		showErrorMessage: rule.showErrorMessage,
		showInputMessage: rule.showInputMessage,
		renderMode: rule.renderMode,
		bizInfo: rule.bizInfo
	};
}

//#endregion
//#region src/types/enum/update-rule-type.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
let UpdateRuleType = /* @__PURE__ */ function(UpdateRuleType) {
	UpdateRuleType[UpdateRuleType["SETTING"] = 0] = "SETTING";
	UpdateRuleType[UpdateRuleType["RANGE"] = 1] = "RANGE";
	UpdateRuleType[UpdateRuleType["OPTIONS"] = 2] = "OPTIONS";
	UpdateRuleType[UpdateRuleType["ALL"] = 3] = "ALL";
	return UpdateRuleType;
}({});

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/models/data-validation-model.ts
let DataValidationModel = class DataValidationModel extends _univerjs_core.Disposable {
	constructor(_logService) {
		super();
		this._logService = _logService;
		_defineProperty(this, "_model", /* @__PURE__ */ new Map());
		_defineProperty(this, "_ruleChange$", new rxjs.Subject());
		_defineProperty(this, "ruleChange$", this._ruleChange$.asObservable());
		_defineProperty(this, "ruleChangeDebounce$", this.ruleChange$.pipe((0, rxjs.debounceTime)(20)));
		this.disposeWithMe({ dispose: () => {
			this._ruleChange$.complete();
		} });
	}
	_ensureMap(unitId, subUnitId) {
		if (!this._model.has(unitId)) this._model.set(unitId, /* @__PURE__ */ new Map());
		const unitMap = this._model.get(unitId);
		if (unitMap.has(subUnitId)) return unitMap.get(subUnitId);
		const map = {
			map: /* @__PURE__ */ new Map(),
			list: []
		};
		unitMap.set(subUnitId, map);
		return map;
	}
	_addSubUnitRule(subUnit, rule, index) {
		const { map: dataValidationMap, list: dataValidations } = subUnit;
		const rules = (Array.isArray(rule) ? rule : [rule]).filter((item) => !dataValidationMap.has(item.uid));
		if (typeof index === "number" && index < dataValidations.length) dataValidations.splice(index, 0, ...rules);
		else dataValidations.push(...rules);
		rules.forEach((item) => {
			dataValidationMap.set(item.uid, item);
		});
	}
	_removeSubUnitRule(subUnit, ruleId) {
		const { map: dataValidationMap, list: dataValidations } = subUnit;
		const index = dataValidations.findIndex((item) => item.uid === ruleId);
		if (index > -1) {
			dataValidations.splice(index, 1);
			dataValidationMap.delete(ruleId);
		}
	}
	_updateSubUnitRule(subUnit, ruleId, payload) {
		const { map: dataValidationMap, list: dataValidations } = subUnit;
		const oldRule = dataValidationMap.get(ruleId);
		const index = dataValidations.findIndex((rule) => ruleId === rule.uid);
		if (!oldRule) throw new Error(`Data validation rule is not found, ruleId: ${ruleId}.`);
		const rule = { ...oldRule };
		switch (payload.type) {
			case 1:
				rule.ranges = payload.payload;
				break;
			case 0:
				Object.assign(rule, getRuleSetting(payload.payload));
				break;
			case 2:
				Object.assign(rule, getRuleOptions(payload.payload));
				break;
			case 3:
				Object.assign(rule, payload.payload);
				break;
			default: break;
		}
		dataValidations[index] = rule;
		dataValidationMap.set(ruleId, rule);
		return rule;
	}
	_addRuleSideEffect(unitId, subUnitId, rule, source) {
		if (this._ensureMap(unitId, subUnitId).map.get(rule.uid)) return;
		return {
			rule,
			type: "add",
			unitId,
			subUnitId,
			source
		};
	}
	addRule(unitId, subUnitId, rule, source, index) {
		try {
			const subUnitMap = this._ensureMap(unitId, subUnitId);
			const effects = (Array.isArray(rule) ? rule : [rule]).map((item) => this._addRuleSideEffect(unitId, subUnitId, item, source));
			this._addSubUnitRule(subUnitMap, rule, index);
			effects.forEach((effect) => {
				if (effect) this._ruleChange$.next(effect);
			});
		} catch (error) {
			this._logService.error(error);
		}
	}
	updateRule(unitId, subUnitId, ruleId, payload, source) {
		try {
			const subUnitMap = this._ensureMap(unitId, subUnitId);
			const oldRule = _univerjs_core.Tools.deepClone(subUnitMap.map.get(ruleId));
			if (!oldRule) throw new Error(`Data validation rule is not found, ruleId: ${ruleId}.`);
			const rule = this._updateSubUnitRule(subUnitMap, ruleId, payload);
			this._ruleChange$.next({
				rule,
				type: "update",
				unitId,
				subUnitId,
				source,
				updatePayload: payload,
				oldRule
			});
		} catch (error) {
			this._logService.error(error);
		}
	}
	removeRule(unitId, subUnitId, ruleId, source) {
		try {
			const map = this._ensureMap(unitId, subUnitId);
			const oldRule = map.map.get(ruleId);
			if (oldRule) {
				this._removeSubUnitRule(map, ruleId);
				this._ruleChange$.next({
					rule: oldRule,
					type: "remove",
					unitId,
					subUnitId,
					source
				});
			}
		} catch (error) {
			this._logService.error(error);
		}
	}
	getRuleById(unitId, subUnitId, ruleId) {
		return this._ensureMap(unitId, subUnitId).map.get(ruleId);
	}
	getRuleIndex(unitId, subUnitId, ruleId) {
		return this._ensureMap(unitId, subUnitId).list.findIndex((rule) => rule.uid === ruleId);
	}
	getRules(unitId, subUnitId) {
		return [...this._ensureMap(unitId, subUnitId).list];
	}
	getUnitRules(unitId) {
		const unitMap = this._model.get(unitId);
		if (!unitMap) return [];
		const res = [];
		unitMap.forEach((manager, subUnitId) => {
			res.push([subUnitId, manager.list]);
		});
		return res;
	}
	deleteUnitRules(unitId) {
		this._model.delete(unitId);
	}
	getSubUnitIds(unitId) {
		var _this$_model$get$keys, _this$_model$get;
		return Array.from((_this$_model$get$keys = (_this$_model$get = this._model.get(unitId)) === null || _this$_model$get === void 0 ? void 0 : _this$_model$get.keys()) !== null && _this$_model$get$keys !== void 0 ? _this$_model$get$keys : []);
	}
	getAll() {
		return Array.from(this._model.keys()).map((unitId) => [unitId, this.getUnitRules(unitId)]);
	}
};
DataValidationModel = __decorate([__decorateParam(0, _univerjs_core.ILogService)], DataValidationModel);

//#endregion
//#region src/commands/mutations/data-validation.mutation.ts
const AddDataValidationMutation = {
	type: _univerjs_core.CommandType.MUTATION,
	id: "data-validation.mutation.addRule",
	handler(accessor, params) {
		if (!params) return false;
		const { unitId, subUnitId, rule, index, source = "command" } = params;
		accessor.get(DataValidationModel).addRule(unitId, subUnitId, rule, source, index);
		return true;
	}
};
const RemoveDataValidationMutation = {
	type: _univerjs_core.CommandType.MUTATION,
	id: "data-validation.mutation.removeRule",
	handler(accessor, params) {
		if (!params) return false;
		const { unitId, subUnitId, ruleId, source = "command" } = params;
		const dataValidationModel = accessor.get(DataValidationModel);
		if (Array.isArray(ruleId)) ruleId.forEach((item) => {
			dataValidationModel.removeRule(unitId, subUnitId, item, source);
		});
		else dataValidationModel.removeRule(unitId, subUnitId, ruleId, source);
		return true;
	}
};
const UpdateDataValidationMutation = {
	type: _univerjs_core.CommandType.MUTATION,
	id: "data-validation.mutation.updateRule",
	handler(accessor, params) {
		if (!params) return false;
		const { unitId, subUnitId, ruleId, payload, source = "command" } = params;
		accessor.get(DataValidationModel).updateRule(unitId, subUnitId, ruleId, payload, source);
		return true;
	}
};

//#endregion
//#region src/controllers/dv-resource.controller.ts
const DATA_VALIDATION_PLUGIN_NAME = "SHEET_DATA_VALIDATION_PLUGIN";
let DataValidationResourceController = class DataValidationResourceController extends _univerjs_core.Disposable {
	constructor(_resourceManagerService, _univerInstanceService, _dataValidationModel) {
		super();
		this._resourceManagerService = _resourceManagerService;
		this._univerInstanceService = _univerInstanceService;
		this._dataValidationModel = _dataValidationModel;
		this._initSnapshot();
	}
	_initSnapshot() {
		const toJson = (unitID) => {
			const map = this._dataValidationModel.getUnitRules(unitID);
			const resultMap = {};
			if (map) {
				map.forEach(([key, v]) => {
					resultMap[key] = v;
				});
				return JSON.stringify(resultMap);
			}
			return "";
		};
		const parseJson = (json) => {
			if (!json) return {};
			try {
				return JSON.parse(json);
			} catch (err) {
				return {};
			}
		};
		this.disposeWithMe(this._resourceManagerService.registerPluginResource({
			pluginName: DATA_VALIDATION_PLUGIN_NAME,
			businesses: [_univerjs_core.UniverInstanceType.UNIVER_SHEET],
			toJson: (unitID) => toJson(unitID),
			parseJson: (json) => parseJson(json),
			onUnLoad: (unitID) => {
				this._dataValidationModel.deleteUnitRules(unitID);
			},
			onLoad: (unitID, value) => {
				Object.keys(value).forEach((subunitId) => {
					value[subunitId].forEach((rule) => {
						this._dataValidationModel.addRule(unitID, subunitId, rule, "patched");
					});
				});
			}
		}));
	}
};
DataValidationResourceController = __decorate([
	__decorateParam(0, _univerjs_core.IResourceManagerService),
	__decorateParam(1, _univerjs_core.IUniverInstanceService),
	__decorateParam(2, (0, _univerjs_core.Inject)(DataValidationModel))
], DataValidationResourceController);

//#endregion
//#region package.json
var name = "@univerjs/data-validation";
var version = "0.25.1";

//#endregion
//#region src/services/data-validator-registry.service.ts
let DataValidatorRegistryScope = /* @__PURE__ */ function(DataValidatorRegistryScope) {
	DataValidatorRegistryScope["SHEET"] = "sheet";
	return DataValidatorRegistryScope;
}({});
/**
* Register data validator
*/
var DataValidatorRegistryService = class {
	constructor() {
		_defineProperty(this, "_validatorByScopes", /* @__PURE__ */ new Map());
		_defineProperty(this, "_validatorMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_validatorsChange$", new rxjs.BehaviorSubject(void 0));
		_defineProperty(this, "validatorsChange$", this._validatorsChange$.asObservable());
	}
	_addValidatorToScope(validator, scope) {
		if (!this._validatorByScopes.has(scope)) this._validatorByScopes.set(scope, []);
		const validators = this._validatorByScopes.get(scope);
		if (validators.findIndex((m) => m.id === validator.id) > -1) throw new Error(`Validator item with the same id ${validator.id} has already been added!`);
		validators.push(validator);
	}
	_removeValidatorFromScope(validator, scope) {
		const validators = this._validatorByScopes.get(scope);
		if (!validators) return;
		const index = validators.findIndex((v) => v.id === validator.id);
		if (index > -1) validators.splice(index, 1);
	}
	register(validator) {
		this._validatorMap.set(validator.id, validator);
		if (Array.isArray(validator.scopes)) validator.scopes.forEach((scope) => {
			this._addValidatorToScope(validator, scope);
		});
		else this._addValidatorToScope(validator, validator.scopes);
		this._validatorsChange$.next();
		return (0, _univerjs_core.toDisposable)(() => {
			this._validatorMap.delete(validator.id);
			if (Array.isArray(validator.scopes)) validator.scopes.forEach((scope) => {
				this._removeValidatorFromScope(validator, scope);
			});
			else this._removeValidatorFromScope(validator, validator.scopes);
			this._validatorsChange$.next();
		});
	}
	getValidatorItem(id) {
		return this._validatorMap.get(id);
	}
	getValidatorsByScope(scope) {
		return this._validatorByScopes.get(scope);
	}
};

//#endregion
//#region src/commands/commands/data-validation.command.ts
/**
* @deprecated `AddDataValidationCommand` is deprecated, please use `AddSheetDataValidationCommand` in `@univerjs/sheets-data-validation` instead!
*/
const AddDataValidationCommand = {
	type: _univerjs_core.CommandType.COMMAND,
	id: "data-validation.command.addRule",
	async handler(accessor, params) {
		accessor.get(_univerjs_core.ILogService).error("[Deprecated]: `AddDataValidationCommand` is deprecated, please use `AddSheetDataValidationCommand` in `@univerjs/sheets-data-validation` instead!");
		if (!params) return false;
		const { rule, unitId, subUnitId } = params;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const undoRedoService = accessor.get(_univerjs_core.IUndoRedoService);
		const mutationParams = {
			...params,
			rule: {
				...params.rule,
				ranges: [params.rule.range]
			}
		};
		const redoMutations = [{
			id: AddDataValidationMutation.id,
			params: mutationParams
		}];
		const undoMutations = [{
			id: RemoveDataValidationMutation.id,
			params: {
				unitId,
				subUnitId,
				ruleId: rule.uid
			}
		}];
		undoRedoService.pushUndoRedo({
			unitID: unitId,
			redoMutations,
			undoMutations
		});
		await commandService.executeCommand(AddDataValidationMutation.id, mutationParams);
		return true;
	}
};
/**
* @deprecated `RemoveDataValidationCommand` is deprecated, please use `RemoveSheetDataValidationCommand` in `@univerjs/sheets-data-validation` instead!
*/
const RemoveDataValidationCommand = {
	type: _univerjs_core.CommandType.COMMAND,
	id: "data-validation.command.removeRule",
	handler(accessor, params) {
		accessor.get(_univerjs_core.ILogService).error("[Deprecated]: `RemoveDataValidationCommand` is deprecated, please use `RemoveSheetDataValidationCommand` in `@univerjs/sheets-data-validation` instead!");
		if (!params) return false;
		const { unitId, subUnitId, ruleId } = params;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const undoRedoService = accessor.get(_univerjs_core.IUndoRedoService);
		const dataValidationModel = accessor.get(DataValidationModel);
		const redoMutations = [{
			id: RemoveDataValidationMutation.id,
			params
		}];
		const undoMutations = [{
			id: AddDataValidationMutation.id,
			params: {
				unitId,
				subUnitId,
				rule: { ...dataValidationModel.getRuleById(unitId, subUnitId, ruleId) },
				index: dataValidationModel.getRuleIndex(unitId, subUnitId, ruleId)
			}
		}];
		undoRedoService.pushUndoRedo({
			undoMutations,
			redoMutations,
			unitID: params.unitId
		});
		commandService.executeCommand(RemoveDataValidationMutation.id, params);
		return true;
	}
};
/**
* @deprecated `UpdateDataValidationOptionsCommand` is deprecated, please use `UpdateSheetDataValidationOptionsCommand` in `@univerjs/sheets-data-validation` instead!
*/
const UpdateDataValidationOptionsCommand = {
	type: _univerjs_core.CommandType.COMMAND,
	id: "data-validation.command.updateDataValidationSetting",
	handler(accessor, params) {
		accessor.get(_univerjs_core.ILogService).warn("[Deprecated]: `UpdateDataValidationOptionsCommand` is deprecated, please use `UpdateSheetDataValidationOptionsCommand` in `@univerjs/sheets-data-validation` instead!");
		if (!params) return false;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const redoUndoService = accessor.get(_univerjs_core.IUndoRedoService);
		const dataValidationModel = accessor.get(DataValidationModel);
		const { unitId, subUnitId, ruleId, options } = params;
		const rule = dataValidationModel.getRuleById(unitId, subUnitId, ruleId);
		if (!rule) return false;
		const mutationParams = {
			unitId,
			subUnitId,
			ruleId,
			payload: {
				type: 2,
				payload: options
			}
		};
		const redoMutations = [{
			id: UpdateDataValidationMutation.id,
			params: mutationParams
		}];
		const undoMutationParams = {
			unitId,
			subUnitId,
			ruleId,
			payload: {
				type: 2,
				payload: getRuleOptions(rule)
			}
		};
		const undoMutations = [{
			id: UpdateDataValidationMutation.id,
			params: undoMutationParams
		}];
		redoUndoService.pushUndoRedo({
			unitID: unitId,
			redoMutations,
			undoMutations
		});
		commandService.executeCommand(UpdateDataValidationMutation.id, mutationParams);
		return true;
	}
};
/**
* @deprecated `UpdateDataValidationSettingCommand` is deprecated, please use `UpdateSheetDataValidationSettingCommand` in `@univerjs/sheets-data-validation` instead!
*/
const UpdateDataValidationSettingCommand = {
	type: _univerjs_core.CommandType.COMMAND,
	id: "data-validation.command.updateDataValidationOptions",
	handler(accessor, params) {
		accessor.get(_univerjs_core.ILogService).error("[Deprecated]: `UpdateDataValidationSettingCommand` is deprecated, please use `UpdateSheetDataValidationSettingCommand` in `@univerjs/sheets-data-validation` instead!");
		if (!params) return false;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const redoUndoService = accessor.get(_univerjs_core.IUndoRedoService);
		const dataValidationModel = accessor.get(DataValidationModel);
		const dataValidatorRegistryService = accessor.get(DataValidatorRegistryService);
		const { unitId, subUnitId, ruleId, setting } = params;
		const validator = dataValidatorRegistryService.getValidatorItem(setting.type);
		if (!validator) return false;
		const rule = dataValidationModel.getRuleById(unitId, subUnitId, ruleId);
		if (!rule) return false;
		const newRule = {
			...rule,
			...setting
		};
		if (!validator.validatorFormula(newRule, unitId, subUnitId).success) return false;
		const mutationParams = {
			unitId,
			subUnitId,
			ruleId,
			payload: {
				type: 0,
				payload: {
					...setting,
					...validator.normalizeFormula(newRule, unitId, subUnitId)
				}
			}
		};
		const redoMutations = [{
			id: UpdateDataValidationMutation.id,
			params: mutationParams
		}];
		const undoMutationParams = {
			unitId,
			subUnitId,
			ruleId,
			payload: {
				type: 0,
				payload: getRuleSetting(rule)
			}
		};
		const undoMutations = [{
			id: UpdateDataValidationMutation.id,
			params: undoMutationParams
		}];
		redoUndoService.pushUndoRedo({
			unitID: unitId,
			redoMutations,
			undoMutations
		});
		commandService.executeCommand(UpdateDataValidationMutation.id, mutationParams);
		return true;
	}
};
/**
* @deprecated `RemoveAllDataValidationCommand` is deprecated, please use `RemoveSheetAllDataValidationCommand` in `@univerjs/sheets-data-validation` instead!
*/
const RemoveAllDataValidationCommand = {
	type: _univerjs_core.CommandType.COMMAND,
	id: "data-validation.command.removeAll",
	handler(accessor, params) {
		accessor.get(_univerjs_core.ILogService).error("[Deprecated]: `RemoveAllDataValidationCommand` is deprecated, please use `RemoveSheetAllDataValidationCommand` in `@univerjs/sheets-data-validation` instead!");
		if (!params) return false;
		const { unitId, subUnitId } = params;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const dataValidationModel = accessor.get(DataValidationModel);
		const undoRedoService = accessor.get(_univerjs_core.IUndoRedoService);
		const currentRules = [...dataValidationModel.getRules(unitId, subUnitId)];
		const redoParams = {
			unitId,
			subUnitId,
			ruleId: currentRules.map((rule) => rule.uid)
		};
		const redoMutations = [{
			id: RemoveDataValidationMutation.id,
			params: redoParams
		}];
		const undoMutations = [{
			id: AddDataValidationMutation.id,
			params: {
				unitId,
				subUnitId,
				rule: currentRules
			}
		}];
		undoRedoService.pushUndoRedo({
			redoMutations,
			undoMutations,
			unitID: unitId
		});
		commandService.executeCommand(RemoveDataValidationMutation.id, redoParams);
		return true;
	}
};

//#endregion
//#region src/config/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const DATA_VALIDATION_PLUGIN_CONFIG_KEY = "data-validation.config";
const configSymbol = Symbol(DATA_VALIDATION_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/plugin.ts
let UniverDataValidationPlugin = class UniverDataValidationPlugin extends _univerjs_core.Plugin {
	constructor(_config = defaultPluginConfig, _injector, _commandService, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._commandService = _commandService;
		this._configService = _configService;
		const { ...rest } = (0, _univerjs_core.merge)({}, defaultPluginConfig, this._config);
		this._configService.setConfig(DATA_VALIDATION_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		[
			[DataValidationModel],
			[DataValidatorRegistryService],
			[DataValidationResourceController]
		].forEach((d) => this._injector.add(d));
		[
			AddDataValidationCommand,
			RemoveAllDataValidationCommand,
			UpdateDataValidationOptionsCommand,
			UpdateDataValidationSettingCommand,
			RemoveDataValidationCommand,
			AddDataValidationMutation,
			UpdateDataValidationMutation,
			RemoveDataValidationMutation
		].forEach((command) => {
			this._commandService.registerCommand(command);
		});
	}
	onReady() {
		this._injector.get(DataValidationResourceController);
	}
};
_defineProperty(UniverDataValidationPlugin, "pluginName", "UNIVER_DATA_VALIDATION_PLUGIN");
_defineProperty(UniverDataValidationPlugin, "packageName", name);
_defineProperty(UniverDataValidationPlugin, "version", version);
_defineProperty(UniverDataValidationPlugin, "type", _univerjs_core.UniverInstanceType.UNIVER_SHEET);
UniverDataValidationPlugin = __decorate([
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector)),
	__decorateParam(2, _univerjs_core.ICommandService),
	__decorateParam(3, _univerjs_core.IConfigService)
], UniverDataValidationPlugin);

//#endregion
//#region src/types/const/two-formula-operators.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const TWO_FORMULA_OPERATOR_COUNT = [_univerjs_core.DataValidationOperator.BETWEEN, _univerjs_core.DataValidationOperator.NOT_BETWEEN];

//#endregion
//#region src/types/const/operator-text-map.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const OperatorTextMap = {
	[_univerjs_core.DataValidationOperator.BETWEEN]: "data-validation.operators.between",
	[_univerjs_core.DataValidationOperator.EQUAL]: "data-validation.operators.equal",
	[_univerjs_core.DataValidationOperator.GREATER_THAN]: "data-validation.operators.greaterThan",
	[_univerjs_core.DataValidationOperator.GREATER_THAN_OR_EQUAL]: "data-validation.operators.greaterThanOrEqual",
	[_univerjs_core.DataValidationOperator.LESS_THAN]: "data-validation.operators.lessThan",
	[_univerjs_core.DataValidationOperator.LESS_THAN_OR_EQUAL]: "data-validation.operators.lessThanOrEqual",
	[_univerjs_core.DataValidationOperator.NOT_BETWEEN]: "data-validation.operators.notBetween",
	[_univerjs_core.DataValidationOperator.NOT_EQUAL]: "data-validation.operators.notEqual"
};
const OperatorTitleMap = {
	[_univerjs_core.DataValidationOperator.BETWEEN]: "data-validation.ruleName.between",
	[_univerjs_core.DataValidationOperator.EQUAL]: "data-validation.ruleName.equal",
	[_univerjs_core.DataValidationOperator.GREATER_THAN]: "data-validation.ruleName.greaterThan",
	[_univerjs_core.DataValidationOperator.GREATER_THAN_OR_EQUAL]: "data-validation.ruleName.greaterThanOrEqual",
	[_univerjs_core.DataValidationOperator.LESS_THAN]: "data-validation.ruleName.lessThan",
	[_univerjs_core.DataValidationOperator.LESS_THAN_OR_EQUAL]: "data-validation.ruleName.lessThanOrEqual",
	[_univerjs_core.DataValidationOperator.NOT_BETWEEN]: "data-validation.ruleName.notBetween",
	[_univerjs_core.DataValidationOperator.NOT_EQUAL]: "data-validation.ruleName.notEqual",
	NONE: "data-validation.ruleName.legal"
};
const OperatorErrorTitleMap = {
	[_univerjs_core.DataValidationOperator.BETWEEN]: "data-validation.errorMsg.between",
	[_univerjs_core.DataValidationOperator.EQUAL]: "data-validation.errorMsg.equal",
	[_univerjs_core.DataValidationOperator.GREATER_THAN]: "data-validation.errorMsg.greaterThan",
	[_univerjs_core.DataValidationOperator.GREATER_THAN_OR_EQUAL]: "data-validation.errorMsg.greaterThanOrEqual",
	[_univerjs_core.DataValidationOperator.LESS_THAN]: "data-validation.errorMsg.lessThan",
	[_univerjs_core.DataValidationOperator.LESS_THAN_OR_EQUAL]: "data-validation.errorMsg.lessThanOrEqual",
	[_univerjs_core.DataValidationOperator.NOT_BETWEEN]: "data-validation.errorMsg.notBetween",
	[_univerjs_core.DataValidationOperator.NOT_EQUAL]: "data-validation.errorMsg.notEqual",
	NONE: "data-validation.errorMsg.legal"
};

//#endregion
//#region src/validators/base-data-validator.ts
const FORMULA1 = "{FORMULA1}";
const FORMULA2 = "{FORMULA2}";
const TYPE = "{TYPE}";
const operatorNameMap = {
	[_univerjs_core.DataValidationOperator.BETWEEN]: "data-validation.operators.between",
	[_univerjs_core.DataValidationOperator.EQUAL]: "data-validation.operators.equal",
	[_univerjs_core.DataValidationOperator.GREATER_THAN]: "data-validation.operators.greaterThan",
	[_univerjs_core.DataValidationOperator.GREATER_THAN_OR_EQUAL]: "data-validation.operators.greaterThanOrEqual",
	[_univerjs_core.DataValidationOperator.LESS_THAN]: "data-validation.operators.lessThan",
	[_univerjs_core.DataValidationOperator.LESS_THAN_OR_EQUAL]: "data-validation.operators.lessThanOrEqual",
	[_univerjs_core.DataValidationOperator.NOT_BETWEEN]: "data-validation.operators.notBetween",
	[_univerjs_core.DataValidationOperator.NOT_EQUAL]: "data-validation.operators.notEqual"
};
let DataValidatorDropdownType = /* @__PURE__ */ function(DataValidatorDropdownType) {
	DataValidatorDropdownType["DATE"] = "date";
	DataValidatorDropdownType["TIME"] = "time";
	DataValidatorDropdownType["DATETIME"] = "datetime";
	DataValidatorDropdownType["LIST"] = "list";
	DataValidatorDropdownType["MULTIPLE_LIST"] = "multipleList";
	DataValidatorDropdownType["COLOR"] = "color";
	DataValidatorDropdownType["CASCADE"] = "cascade";
	return DataValidatorDropdownType;
}({});
let BaseDataValidator = class BaseDataValidator {
	constructor(localeService, injector) {
		this.localeService = localeService;
		this.injector = injector;
		_defineProperty(this, "offsetFormulaByRange", true);
		_defineProperty(this, "formulaInput", void 0);
		_defineProperty(this, "canvasRender", null);
		_defineProperty(this, "dropdownType", void 0);
		_defineProperty(this, "optionsInput", void 0);
		_defineProperty(this, "skipDefaultFontRender", void 0);
	}
	get operatorNames() {
		return this.operators.map((operator) => this.localeService.t(operatorNameMap[operator]));
	}
	get titleStr() {
		return this.localeService.t(this.title);
	}
	generateRuleName(rule) {
		var _rule$formula, _rule$formula2;
		if (!rule.operator) return this.localeService.t(OperatorTitleMap.NONE).replace(TYPE, this.titleStr);
		const ruleName = this.localeService.t(OperatorTitleMap[rule.operator]).replace(FORMULA1, (_rule$formula = rule.formula1) !== null && _rule$formula !== void 0 ? _rule$formula : "").replace(FORMULA2, (_rule$formula2 = rule.formula2) !== null && _rule$formula2 !== void 0 ? _rule$formula2 : "");
		return `${this.titleStr} ${ruleName}`;
	}
	generateRuleErrorMessage(rule, position) {
		var _rule$formula3, _rule$formula4;
		if (!rule.operator) return this.localeService.t(OperatorErrorTitleMap.NONE).replace(TYPE, this.titleStr);
		return `${this.localeService.t(OperatorErrorTitleMap[rule.operator]).replace(FORMULA1, (_rule$formula3 = rule.formula1) !== null && _rule$formula3 !== void 0 ? _rule$formula3 : "").replace(FORMULA2, (_rule$formula4 = rule.formula2) !== null && _rule$formula4 !== void 0 ? _rule$formula4 : "")}`;
	}
	getExtraStyle(rule, value, ctx, row, column) {}
	getRuleFinalError(rule, position) {
		if (rule.showErrorMessage && rule.error) return rule.error;
		return this.generateRuleErrorMessage(rule, position);
	}
	isEmptyCellValue(cellValue) {
		if (cellValue === "" || cellValue === void 0 || cellValue === null) return true;
		return false;
	}
	normalizeFormula(rule, unitId, subUnitId) {
		return {
			formula1: rule.formula1,
			formula2: rule.formula2
		};
	}
	async isValidType(cellInfo, formula, rule) {
		return true;
	}
	transform(cellInfo, formula, rule) {
		return cellInfo;
	}
	async validatorIsEqual(cellInfo, formula, rule) {
		const { formula1 } = formula;
		const { value: cellValue } = cellInfo;
		if (Number.isNaN(formula1)) return true;
		return cellValue === formula1;
	}
	async validatorIsNotEqual(cellInfo, formula, _rule) {
		const { formula1 } = formula;
		if (Number.isNaN(formula1)) return true;
		return cellInfo.value !== formula1;
	}
	async validatorIsBetween(cellInfo, formula, _rule) {
		const { formula1, formula2 } = formula;
		if (Number.isNaN(formula1) || Number.isNaN(formula2)) return true;
		const start = Math.min(formula1, formula2);
		const end = Math.max(formula1, formula2);
		return cellInfo.value >= start && cellInfo.value <= end;
	}
	async validatorIsNotBetween(cellInfo, formula, _rule) {
		const { formula1, formula2 } = formula;
		if (Number.isNaN(formula1) || Number.isNaN(formula2)) return true;
		const start = Math.min(formula1, formula2);
		const end = Math.max(formula1, formula2);
		return cellInfo.value < start || cellInfo.value > end;
	}
	async validatorIsGreaterThan(cellInfo, formula, _rule) {
		const { formula1 } = formula;
		if (Number.isNaN(formula1)) return true;
		return cellInfo.value > formula1;
	}
	async validatorIsGreaterThanOrEqual(cellInfo, formula, _rule) {
		const { formula1 } = formula;
		if (Number.isNaN(formula1)) return true;
		return cellInfo.value >= formula1;
	}
	async validatorIsLessThan(cellInfo, formula, _rule) {
		const { formula1 } = formula;
		if (Number.isNaN(formula1)) return true;
		return cellInfo.value < formula1;
	}
	async validatorIsLessThanOrEqual(cellInfo, formula, _rule) {
		const { formula1 } = formula;
		if (Number.isNaN(formula1)) return true;
		return cellInfo.value <= formula1;
	}
	async validator(cellInfo, rule) {
		const { value: cellValue, unitId, subUnitId } = cellInfo;
		const isEmpty = this.isEmptyCellValue(cellValue);
		const { allowBlank = true, operator } = rule;
		if (isEmpty) return allowBlank;
		const formulaInfo = await this.parseFormula(rule, unitId, subUnitId, cellInfo.row, cellInfo.column);
		if (!formulaInfo.isFormulaValid) return false;
		if (!await this.isValidType(cellInfo, formulaInfo, rule)) return false;
		if (!operator) return true;
		const transformedCell = this.transform(cellInfo, formulaInfo, rule);
		switch (operator) {
			case _univerjs_core.DataValidationOperator.BETWEEN: return this.validatorIsBetween(transformedCell, formulaInfo, rule);
			case _univerjs_core.DataValidationOperator.EQUAL: return this.validatorIsEqual(transformedCell, formulaInfo, rule);
			case _univerjs_core.DataValidationOperator.GREATER_THAN: return this.validatorIsGreaterThan(transformedCell, formulaInfo, rule);
			case _univerjs_core.DataValidationOperator.GREATER_THAN_OR_EQUAL: return this.validatorIsGreaterThanOrEqual(transformedCell, formulaInfo, rule);
			case _univerjs_core.DataValidationOperator.LESS_THAN: return this.validatorIsLessThan(transformedCell, formulaInfo, rule);
			case _univerjs_core.DataValidationOperator.LESS_THAN_OR_EQUAL: return this.validatorIsLessThanOrEqual(transformedCell, formulaInfo, rule);
			case _univerjs_core.DataValidationOperator.NOT_BETWEEN: return this.validatorIsNotBetween(transformedCell, formulaInfo, rule);
			case _univerjs_core.DataValidationOperator.NOT_EQUAL: return this.validatorIsNotEqual(transformedCell, formulaInfo, rule);
			default: throw new Error("Unknown operator.");
		}
	}
};
BaseDataValidator = __decorate([__decorateParam(0, (0, _univerjs_core.Inject)(_univerjs_core.LocaleService)), __decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector))], BaseDataValidator);

//#endregion
exports.AddDataValidationMutation = AddDataValidationMutation;
Object.defineProperty(exports, 'BaseDataValidator', {
  enumerable: true,
  get: function () {
    return BaseDataValidator;
  }
});
Object.defineProperty(exports, 'DataValidationModel', {
  enumerable: true,
  get: function () {
    return DataValidationModel;
  }
});
Object.defineProperty(exports, 'DataValidationResourceController', {
  enumerable: true,
  get: function () {
    return DataValidationResourceController;
  }
});
exports.DataValidatorDropdownType = DataValidatorDropdownType;
exports.DataValidatorRegistryScope = DataValidatorRegistryScope;
exports.DataValidatorRegistryService = DataValidatorRegistryService;
exports.FORMULA1 = FORMULA1;
exports.FORMULA2 = FORMULA2;
exports.RemoveDataValidationMutation = RemoveDataValidationMutation;
exports.TWO_FORMULA_OPERATOR_COUNT = TWO_FORMULA_OPERATOR_COUNT;
exports.TYPE = TYPE;
Object.defineProperty(exports, 'UniverDataValidationPlugin', {
  enumerable: true,
  get: function () {
    return UniverDataValidationPlugin;
  }
});
exports.UpdateDataValidationMutation = UpdateDataValidationMutation;
exports.UpdateRuleType = UpdateRuleType;
exports.getRuleOptions = getRuleOptions;
exports.getRuleSetting = getRuleSetting;
