// @univerjs/docs-drawing-ui/lib/cjs/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
let _univerjs_core = require("@univerjs/core");
let _univerjs_docs_drawing = require("@univerjs/docs-drawing");
let _univerjs_docs = require("@univerjs/docs");
let _univerjs_docs_ui = require("@univerjs/docs-ui");
let _univerjs_engine_render = require("@univerjs/engine-render");
let _univerjs_design = require("@univerjs/design");
let _univerjs_drawing = require("@univerjs/drawing");
let _univerjs_ui = require("@univerjs/ui");
let rxjs = require("rxjs");
let _univerjs_drawing_ui = require("@univerjs/drawing-ui");
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");

//#region src/commands/commands/remove-doc-drawing.command.ts
/**
* The command to remove new sheet image
*/
const RemoveDocDrawingCommand = {
	id: "doc.command.remove-doc-image",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		var _docSelectionRenderSe, _documentDataModel$ge, _documentDataModel$ge2;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const univerInstanceService = accessor.get(_univerjs_core.IUniverInstanceService);
		const renderManagerService = accessor.get(_univerjs_engine_render.IRenderManagerService);
		const documentDataModel = univerInstanceService.getCurrentUniverDocInstance();
		if (params == null || documentDataModel == null) return false;
		const docSelectionRenderService = renderManagerService.getRenderById(params.unitId).with(_univerjs_docs_ui.DocSelectionRenderService);
		const { drawings: removeDrawings } = params;
		const segmentId = (_docSelectionRenderSe = docSelectionRenderService.getSegment()) !== null && _docSelectionRenderSe !== void 0 ? _docSelectionRenderSe : "";
		const textX = new _univerjs_core.TextX();
		const jsonX = _univerjs_core.JSONX.getInstance();
		const customBlocks = (_documentDataModel$ge = (_documentDataModel$ge2 = documentDataModel.getSelfOrHeaderFooterModel(segmentId).getBody()) === null || _documentDataModel$ge2 === void 0 ? void 0 : _documentDataModel$ge2.customBlocks) !== null && _documentDataModel$ge !== void 0 ? _documentDataModel$ge : [];
		const removeCustomBlocks = removeDrawings.map((drawing) => customBlocks.find((customBlock) => customBlock.blockId === drawing.drawingId)).filter((block) => !!block).sort((a, b) => a.startIndex > b.startIndex ? 1 : -1);
		const unitId = removeDrawings[0].unitId;
		const memoryCursor = new _univerjs_core.MemoryCursor();
		memoryCursor.reset();
		const cursorIndex = removeCustomBlocks[0].startIndex;
		const textRanges = [{
			startOffset: cursorIndex,
			endOffset: cursorIndex
		}];
		const doMutation = {
			id: _univerjs_docs.RichTextEditingMutation.id,
			params: {
				unitId,
				actions: [],
				textRanges
			}
		};
		const rawActions = [];
		for (const block of removeCustomBlocks) {
			const { startIndex } = block;
			if (startIndex > memoryCursor.cursor) textX.push({
				t: _univerjs_core.TextXActionType.RETAIN,
				len: startIndex - memoryCursor.cursor
			});
			textX.push({
				t: _univerjs_core.TextXActionType.DELETE,
				len: 1
			});
			memoryCursor.moveCursorTo(startIndex + 1);
		}
		const path = (0, _univerjs_docs_ui.getRichTextEditPath)(documentDataModel, segmentId);
		rawActions.push(jsonX.editOp(textX.serialize(), path));
		for (const block of removeCustomBlocks) {
			var _documentDataModel$ge3;
			const { blockId } = block;
			const drawing = ((_documentDataModel$ge3 = documentDataModel.getDrawings()) !== null && _documentDataModel$ge3 !== void 0 ? _documentDataModel$ge3 : {})[blockId];
			const drawingIndex = documentDataModel.getDrawingsOrder().indexOf(blockId);
			const removeDrawingAction = jsonX.removeOp(["drawings", blockId], drawing);
			const removeDrawingOrderAction = jsonX.removeOp(["drawingsOrder", drawingIndex], blockId);
			rawActions.push(removeDrawingAction);
			rawActions.push(removeDrawingOrderAction);
		}
		doMutation.params.actions = rawActions.reduce((acc, cur) => {
			return _univerjs_core.JSONX.compose(acc, cur);
		}, null);
		const result = commandService.syncExecuteCommand(doMutation.id, doMutation.params);
		return Boolean(result);
	}
};

//#endregion
//#region src/commands/commands/delete-doc-drawing.command.ts
const DeleteDocDrawingsCommand = {
	id: "doc.command.delete-drawing",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor) => {
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const drawings = accessor.get(_univerjs_docs_drawing.IDocDrawingService).getFocusDrawings();
		if (drawings.length === 0) return false;
		const { unitId } = drawings[0];
		const newDrawings = drawings.map((drawing) => {
			const { unitId, subUnitId, drawingId, drawingType } = drawing;
			return {
				unitId,
				subUnitId,
				drawingId,
				drawingType
			};
		});
		return commandService.executeCommand(RemoveDocDrawingCommand.id, {
			unitId,
			drawings: newDrawings
		});
	}
};

//#endregion
//#region src/commands/commands/group-doc-drawing.command.ts
/**
* The command to insert new defined name
*/
const GroupDocDrawingCommand = {
	id: "doc.command.group-doc-image",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		return false;
	}
};

//#endregion
//#region src/commands/commands/insert-doc-drawing.command.ts
/**
* The command to insert new drawings
*/
const InsertDocDrawingCommand = {
	id: "doc.command.insert-doc-image",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		var _ref, _contentInsertRange$s, _documentDataModel$ge, _documentDataModel$ge2;
		if (params == null) return false;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const docSelectionManagerService = accessor.get(_univerjs_docs.DocSelectionManagerService);
		const univerInstanceService = accessor.get(_univerjs_core.IUniverInstanceService);
		const activeTextRange = docSelectionManagerService.getActiveTextRange();
		const documentDataModel = univerInstanceService.getCurrentUniverDocInstance();
		if (documentDataModel == null) return false;
		const unitId = documentDataModel.getUnitId();
		const contentInsertRange = getContentInsertRange(accessor, unitId);
		const targetTextRange = contentInsertRange ? {
			...activeTextRange,
			startOffset: contentInsertRange.startOffset,
			endOffset: contentInsertRange.endOffset,
			collapsed: contentInsertRange.startOffset === contentInsertRange.endOffset,
			segmentId: (_ref = (_contentInsertRange$s = contentInsertRange.segmentId) !== null && _contentInsertRange$s !== void 0 ? _contentInsertRange$s : activeTextRange === null || activeTextRange === void 0 ? void 0 : activeTextRange.segmentId) !== null && _ref !== void 0 ? _ref : ""
		} : activeTextRange;
		if (targetTextRange == null) return false;
		const { drawings } = params;
		const { collapsed, startOffset, segmentId = "" } = targetTextRange;
		const body = documentDataModel.getSelfOrHeaderFooterModel(segmentId).getBody();
		if (body == null) return false;
		const textX = new _univerjs_core.TextX();
		const jsonX = _univerjs_core.JSONX.getInstance();
		const rawActions = [];
		const drawingOrderLength = (_documentDataModel$ge = (_documentDataModel$ge2 = documentDataModel.getSnapshot().drawingsOrder) === null || _documentDataModel$ge2 === void 0 ? void 0 : _documentDataModel$ge2.length) !== null && _documentDataModel$ge !== void 0 ? _documentDataModel$ge : 0;
		let removeDrawingLen = 0;
		if (collapsed) {
			if (startOffset > 0) textX.push({
				t: _univerjs_core.TextXActionType.RETAIN,
				len: startOffset
			});
		} else {
			var _documentDataModel$ge3, _documentDataModel$ge4;
			const dos = _univerjs_core.BuildTextUtils.selection.delete([targetTextRange], body, 0, null, false);
			textX.push(...dos);
			const removedCustomBlockIds = (0, _univerjs_docs_ui.getCustomBlockIdsInSelections)(body, [targetTextRange]);
			const drawings = (_documentDataModel$ge3 = documentDataModel.getDrawings()) !== null && _documentDataModel$ge3 !== void 0 ? _documentDataModel$ge3 : {};
			const drawingOrder = (_documentDataModel$ge4 = documentDataModel.getDrawingsOrder()) !== null && _documentDataModel$ge4 !== void 0 ? _documentDataModel$ge4 : [];
			const sortedRemovedCustomBlockIds = removedCustomBlockIds.sort((a, b) => {
				if (drawingOrder.indexOf(a) > drawingOrder.indexOf(b)) return -1;
				else if (drawingOrder.indexOf(a) < drawingOrder.indexOf(b)) return 1;
				return 0;
			});
			if (sortedRemovedCustomBlockIds.length > 0) for (const blockId of sortedRemovedCustomBlockIds) {
				const drawing = drawings[blockId];
				const drawingIndex = drawingOrder.indexOf(blockId);
				if (drawing == null || drawingIndex < 0) continue;
				const removeDrawingAction = jsonX.removeOp(["drawings", blockId], drawing);
				const removeDrawingOrderAction = jsonX.removeOp(["drawingsOrder", drawingIndex], blockId);
				rawActions.push(removeDrawingAction);
				rawActions.push(removeDrawingOrderAction);
				removeDrawingLen++;
			}
		}
		textX.push({
			t: _univerjs_core.TextXActionType.INSERT,
			body: {
				dataStream: "\b".repeat(drawings.length),
				customBlocks: drawings.map((drawing, i) => ({
					startIndex: i,
					blockId: drawing.drawingId
				}))
			},
			len: drawings.length
		});
		const path = (0, _univerjs_docs_ui.getRichTextEditPath)(documentDataModel, segmentId);
		const placeHolderAction = jsonX.editOp(textX.serialize(), path);
		rawActions.push(placeHolderAction);
		for (const drawing of drawings) {
			const { drawingId } = drawing;
			const addDrawingAction = jsonX.insertOp(["drawings", drawingId], drawing);
			const addDrawingOrderAction = jsonX.insertOp(["drawingsOrder", drawingOrderLength - removeDrawingLen], drawingId);
			rawActions.push(addDrawingAction);
			rawActions.push(addDrawingOrderAction);
		}
		const doMutation = {
			id: _univerjs_docs.RichTextEditingMutation.id,
			params: {
				unitId,
				actions: [],
				textRanges: []
			}
		};
		doMutation.params.actions = rawActions.reduce((acc, cur) => {
			return _univerjs_core.JSONX.compose(acc, cur);
		}, null);
		const result = commandService.syncExecuteCommand(doMutation.id, doMutation.params);
		return Boolean(result);
	}
};
function getContentInsertRange(accessor, unitId) {
	try {
		const range = accessor.get(_univerjs_docs_ui.DocContentInsertService).consumeInsertRange(unitId);
		if (range == null) return null;
		return {
			startOffset: range.startOffset,
			endOffset: range.endOffset,
			collapsed: range.startOffset === range.endOffset,
			segmentId: range.segmentId
		};
	} catch {
		return null;
	}
}

//#endregion
//#region src/commands/commands/set-drawing-arrange.command.ts
/**
* The command to arrange drawings.
*/
const SetDocDrawingArrangeCommand = {
	id: "doc.command.set-drawing-arrange",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const docDrawingService = accessor.get(_univerjs_docs_drawing.IDocDrawingService);
		if (params == null) return false;
		const { unitId, subUnitId, drawingIds, arrangeType } = params;
		const drawingOrderMapParam = {
			unitId,
			subUnitId,
			drawingIds
		};
		let jsonOp;
		if (arrangeType === _univerjs_core.ArrangeTypeEnum.forward) jsonOp = docDrawingService.getForwardDrawingsOp(drawingOrderMapParam);
		else if (arrangeType === _univerjs_core.ArrangeTypeEnum.backward) jsonOp = docDrawingService.getBackwardDrawingOp(drawingOrderMapParam);
		else if (arrangeType === _univerjs_core.ArrangeTypeEnum.front) jsonOp = docDrawingService.getFrontDrawingsOp(drawingOrderMapParam);
		else if (arrangeType === _univerjs_core.ArrangeTypeEnum.back) jsonOp = docDrawingService.getBackDrawingsOp(drawingOrderMapParam);
		if (jsonOp == null) return false;
		const { redo } = jsonOp;
		if (redo == null) return false;
		const rawActions = [];
		let redoCopy = _univerjs_core.Tools.deepClone(redo);
		redoCopy = redoCopy.slice(3);
		redoCopy.unshift("drawingsOrder");
		rawActions.push(redoCopy);
		const doMutation = {
			id: _univerjs_docs.RichTextEditingMutation.id,
			params: {
				unitId,
				actions: [],
				textRanges: null
			}
		};
		doMutation.params.actions = rawActions.reduce((acc, cur) => {
			return _univerjs_core.JSONX.compose(acc, cur);
		}, null);
		const result = commandService.syncExecuteCommand(doMutation.id, doMutation.params);
		return Boolean(result);
	}
};

//#endregion
//#region src/commands/commands/ungroup-doc-drawing.command.ts
/**
* The command to insert new defined name
*/
const UngroupDocDrawingCommand = {
	id: "doc.command.ungroup-doc-image",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		const docDrawingService = accessor.get(_univerjs_docs_drawing.IDocDrawingService);
		if (!params) return false;
		const unitIds = [];
		params.forEach(({ parent, children }) => {
			unitIds.push(parent.unitId);
			children.forEach((child) => {
				unitIds.push(child.unitId);
			});
		});
		const { unitId, subUnitId, undo, redo, objects } = docDrawingService.getUngroupDrawingOp(params);
		return false;
	}
};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region src/services/doc-refresh-drawings.service.ts
var DocRefreshDrawingsService = class {
	constructor() {
		_defineProperty(this, "_refreshDrawings$", new rxjs.BehaviorSubject(null));
		_defineProperty(this, "refreshDrawings$", this._refreshDrawings$.asObservable());
	}
	refreshDrawings(skeleton) {
		this._refreshDrawings$.next(skeleton);
	}
};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/controllers/render-controllers/doc-drawing-update.render-controller.ts
let DocDrawingUpdateRenderController = class DocDrawingUpdateRenderController extends _univerjs_core.Disposable {
	constructor(_context, _commandService, _docSelectionManagerService, _renderManagerSrv, _imageIoService, _docDrawingService, _drawingManagerService, _contextService, _messageService, _localeService, _docSelectionRenderService, _docRefreshDrawingsService, _fileOpenerService) {
		super();
		this._context = _context;
		this._commandService = _commandService;
		this._docSelectionManagerService = _docSelectionManagerService;
		this._renderManagerSrv = _renderManagerSrv;
		this._imageIoService = _imageIoService;
		this._docDrawingService = _docDrawingService;
		this._drawingManagerService = _drawingManagerService;
		this._contextService = _contextService;
		this._messageService = _messageService;
		this._localeService = _localeService;
		this._docSelectionRenderService = _docSelectionRenderService;
		this._docRefreshDrawingsService = _docRefreshDrawingsService;
		this._fileOpenerService = _fileOpenerService;
		this._updateOrderListener();
		this._groupDrawingListener();
		this._focusDrawingListener();
		this._transformDrawingListener();
		this._editAreaChangeListener();
	}
	dispose() {
		super.dispose();
		delete this._context;
	}
	async insertDocImage() {
		const insertPosition = this._getCurrentImageInsertPosition();
		const files = await this._fileOpenerService.openFile({
			multiple: true,
			accept: _univerjs_drawing.DRAWING_IMAGE_ALLOW_IMAGE_LIST.map((image) => `.${image.replace("image/", "")}`).join(",")
		});
		const fileLength = files.length;
		if (fileLength > _univerjs_drawing.DRAWING_IMAGE_COUNT_LIMIT) {
			this._messageService.show({
				type: _univerjs_design.MessageType.Error,
				content: this._localeService.t("docs-drawing-ui.update-status.exceedMaxCount", String(_univerjs_drawing.DRAWING_IMAGE_COUNT_LIMIT))
			});
			return false;
		} else if (fileLength === 0) return false;
		await this._insertFloatImages(files, insertPosition);
		return true;
	}
	async _insertFloatImages(files, insertPosition) {
		let imageParams = [];
		try {
			imageParams = await Promise.all(files.map((file) => this._imageIoService.saveImage(file)));
		} catch (error) {
			const type = error.message;
			let content = "";
			switch (type) {
				case _univerjs_drawing.ImageUploadStatusType.ERROR_EXCEED_SIZE:
					content = this._localeService.t("docs-drawing-ui.update-status.exceedMaxSize", String((0, _univerjs_drawing.getDrawingImageAllowSize)() / (1024 * 1024)));
					break;
				case _univerjs_drawing.ImageUploadStatusType.ERROR_IMAGE_TYPE:
					content = this._localeService.t("docs-drawing-ui.update-status.invalidImageType");
					break;
				case _univerjs_drawing.ImageUploadStatusType.ERROR_IMAGE:
					content = this._localeService.t("docs-drawing-ui.update-status.invalidImage");
					break;
				default: break;
			}
			this._messageService.show({
				type: _univerjs_design.MessageType.Error,
				content
			});
		}
		if (imageParams.length === 0) return;
		const { unitId } = this._context;
		const docDrawingParams = [];
		for (const imageParam of imageParams) {
			if (imageParam == null) continue;
			const { imageId, imageSourceType, source, base64Cache } = imageParam;
			const { width, height, image } = await (0, _univerjs_drawing.getImageSize)(base64Cache || "");
			this._imageIoService.addImageSourceCache(imageId, imageSourceType, image);
			let scale = 1;
			if (width > _univerjs_drawing.DRAWING_IMAGE_WIDTH_LIMIT || height > _univerjs_drawing.DRAWING_IMAGE_HEIGHT_LIMIT) {
				const scaleWidth = _univerjs_drawing.DRAWING_IMAGE_WIDTH_LIMIT / width;
				const scaleHeight = _univerjs_drawing.DRAWING_IMAGE_HEIGHT_LIMIT / height;
				scale = Math.min(scaleWidth, scaleHeight);
			}
			const imagePosition = insertPosition !== null && insertPosition !== void 0 ? insertPosition : this._getCurrentImageInsertPosition();
			const docTransform = this._getImagePosition(width * scale, height * scale, imagePosition);
			if (docTransform == null) return;
			const transform = (0, _univerjs_docs_ui.docDrawingPositionToTransform)(docTransform);
			if (transform != null && imagePosition != null) transform.top = imagePosition.top;
			const docDrawingParam = {
				unitId,
				subUnitId: unitId,
				drawingId: imageId,
				drawingType: _univerjs_core.DrawingTypeEnum.DRAWING_IMAGE,
				imageSourceType,
				source,
				transform,
				docTransform,
				behindDoc: _univerjs_core.BooleanNumber.FALSE,
				title: "",
				description: "",
				layoutType: _univerjs_core.PositionedObjectLayoutType.INLINE,
				wrapText: _univerjs_core.WrapTextType.BOTH_SIDES,
				distB: 0,
				distL: 0,
				distR: 0,
				distT: 0
			};
			if (this._isInsertInHeaderFooter()) {
				docDrawingParam.isMultiTransform = _univerjs_core.BooleanNumber.TRUE;
				docDrawingParam.transforms = docDrawingParam.transform ? [docDrawingParam.transform] : null;
			}
			docDrawingParams.push(docDrawingParam);
		}
		this._commandService.executeCommand(InsertDocDrawingCommand.id, {
			unitId,
			drawings: docDrawingParams
		});
	}
	_isInsertInHeaderFooter() {
		var _this$_renderManagerS;
		const { unitId } = this._context;
		const viewModel = (_this$_renderManagerS = this._renderManagerSrv.getRenderById(unitId)) === null || _this$_renderManagerS === void 0 ? void 0 : _this$_renderManagerS.with(_univerjs_docs.DocSkeletonManagerService).getViewModel();
		const editArea = viewModel === null || viewModel === void 0 ? void 0 : viewModel.getEditArea();
		return editArea === _univerjs_engine_render.DocumentEditArea.HEADER || editArea === _univerjs_engine_render.DocumentEditArea.FOOTER;
	}
	_getImagePosition(imageWidth, imageHeight, insertPosition) {
		var _ref;
		const position = (_ref = insertPosition !== null && insertPosition !== void 0 ? insertPosition : this._getCurrentImageInsertPosition()) !== null && _ref !== void 0 ? _ref : {
			left: 0,
			top: 0
		};
		return {
			size: {
				width: imageWidth,
				height: imageHeight
			},
			positionH: {
				relativeFrom: _univerjs_core.ObjectRelativeFromH.PAGE,
				posOffset: position.left
			},
			positionV: {
				relativeFrom: _univerjs_core.ObjectRelativeFromV.PARAGRAPH,
				posOffset: 0
			},
			angle: 0
		};
	}
	_getCurrentImageInsertPosition() {
		var _this$_docSelectionRe;
		const position = (_this$_docSelectionRe = this._docSelectionRenderService.getActiveTextRange()) === null || _this$_docSelectionRe === void 0 ? void 0 : _this$_docSelectionRe.getAbsolutePosition();
		if (position == null) return null;
		return {
			left: position.left,
			top: position.top
		};
	}
	_updateOrderListener() {
		this.disposeWithMe(this._drawingManagerService.featurePluginOrderUpdate$.subscribe((params) => {
			const { unitId, subUnitId, drawingIds, arrangeType } = params;
			this._commandService.executeCommand(SetDocDrawingArrangeCommand.id, {
				unitId,
				subUnitId,
				drawingIds,
				arrangeType
			});
		}));
	}
	_groupDrawingListener() {
		this.disposeWithMe(this._drawingManagerService.featurePluginGroupUpdate$.subscribe((params) => {
			this._commandService.executeCommand(GroupDocDrawingCommand.id, params);
		}));
		this.disposeWithMe(this._drawingManagerService.featurePluginUngroupUpdate$.subscribe((params) => {
			this._commandService.executeCommand(UngroupDocDrawingCommand.id, params);
		}));
	}
	_getCurrentSceneAndTransformer() {
		const { scene, mainComponent } = this._context;
		if (scene == null || mainComponent == null) return;
		const transformer = scene.getTransformerByCreate();
		const { docsLeft, docsTop } = mainComponent.getOffsetConfig();
		return {
			scene,
			transformer,
			docsLeft,
			docsTop
		};
	}
	_transformDrawingListener() {
		const res = this._getCurrentSceneAndTransformer();
		if (res && res.transformer) this.disposeWithMe(res.transformer.changeEnd$.pipe((0, rxjs.debounceTime)(30)).subscribe((params) => {
			this._docSelectionManagerService.refreshSelection();
		}));
		else throw new Error("transformer is not init");
	}
	_focusDrawingListener() {
		this.disposeWithMe(this._drawingManagerService.focus$.subscribe((params) => {
			var _this$_getCurrentScen;
			const { transformer, docsLeft, docsTop } = (_this$_getCurrentScen = this._getCurrentSceneAndTransformer()) !== null && _this$_getCurrentScen !== void 0 ? _this$_getCurrentScen : {};
			if (params == null || params.length === 0) {
				this._contextService.setContextValue(_univerjs_core.FOCUSING_COMMON_DRAWINGS, false);
				this._docDrawingService.focusDrawing([]);
				if (transformer) transformer.resetProps({
					zeroTop: 0,
					zeroLeft: 0
				});
			} else {
				this._contextService.setContextValue(_univerjs_core.FOCUSING_COMMON_DRAWINGS, true);
				this._docDrawingService.focusDrawing(params);
				this._setDrawingSelections(params);
				const prevSegmentId = this._docSelectionRenderService.getSegment();
				const segmentId = this._findSegmentIdByDrawingId(params[0].drawingId);
				if (prevSegmentId !== segmentId) this._docSelectionRenderService.setSegment(segmentId);
				if (transformer) transformer.resetProps({
					zeroTop: docsTop,
					zeroLeft: docsLeft
				});
			}
		}));
	}
	_findSegmentIdByDrawingId(drawingId) {
		var _body$customBlocks;
		const { unit: DocDataModel } = this._context;
		const { body, headers = {}, footers = {} } = DocDataModel.getSnapshot();
		if (((_body$customBlocks = body === null || body === void 0 ? void 0 : body.customBlocks) !== null && _body$customBlocks !== void 0 ? _body$customBlocks : []).some((b) => b.blockId === drawingId)) return "";
		for (const headerId of Object.keys(headers)) {
			var _headers$headerId$bod;
			if ((_headers$headerId$bod = headers[headerId].body.customBlocks) === null || _headers$headerId$bod === void 0 ? void 0 : _headers$headerId$bod.some((b) => b.blockId === drawingId)) return headerId;
		}
		for (const footerId of Object.keys(footers)) {
			var _footers$footerId$bod;
			if ((_footers$footerId$bod = footers[footerId].body.customBlocks) === null || _footers$footerId$bod === void 0 ? void 0 : _footers$footerId$bod.some((b) => b.blockId === drawingId)) return footerId;
		}
		return "";
	}
	_updateDrawingsEditStatus() {
		var _this$_renderManagerS2;
		if (!this._context) return;
		const { unit: docDataModel, scene, unitId } = this._context;
		const viewModel = (_this$_renderManagerS2 = this._renderManagerSrv.getRenderById(unitId)) === null || _this$_renderManagerS2 === void 0 ? void 0 : _this$_renderManagerS2.with(_univerjs_docs.DocSkeletonManagerService).getViewModel();
		if (viewModel == null || docDataModel == null) return;
		const { drawings = {} } = docDataModel.getSnapshot();
		const isEditBody = viewModel.getEditArea() === _univerjs_engine_render.DocumentEditArea.BODY;
		for (const key of Object.keys(drawings)) {
			const drawing = drawings[key];
			const objectKey = (0, _univerjs_drawing.getDrawingShapeKeyByDrawingSearch)({
				unitId,
				drawingId: drawing.drawingId,
				subUnitId: unitId
			});
			const drawingShapes = scene.fuzzyMathObjects(objectKey, true);
			if (drawingShapes.length) for (const shape of drawingShapes) {
				scene.detachTransformerFrom(shape);
				try {
					shape.setOpacity(.5);
				} catch (e) {}
				if (isEditBody && drawing.isMultiTransform !== _univerjs_core.BooleanNumber.TRUE || !isEditBody && drawing.isMultiTransform === _univerjs_core.BooleanNumber.TRUE) {
					if (drawing.allowTransform !== false) scene.attachTransformerTo(shape);
					try {
						shape.setOpacity(1);
					} catch (e) {}
				}
			}
		}
	}
	_editAreaChangeListener() {
		var _this$_renderManagerS3;
		const { unitId } = this._context;
		const viewModel = (_this$_renderManagerS3 = this._renderManagerSrv.getRenderById(unitId)) === null || _this$_renderManagerS3 === void 0 ? void 0 : _this$_renderManagerS3.with(_univerjs_docs.DocSkeletonManagerService).getViewModel();
		if (viewModel == null) return;
		this._updateDrawingsEditStatus();
		this.disposeWithMe(viewModel.editAreaChange$.subscribe(() => {
			this._updateDrawingsEditStatus();
		}));
		this.disposeWithMe(this._docRefreshDrawingsService.refreshDrawings$.subscribe((skeleton) => {
			if (skeleton == null) return;
			queueMicrotask(() => {
				this._updateDrawingsEditStatus();
			});
		}));
		this.disposeWithMe(this._commandService.onCommandExecuted(async (command) => {
			if (command.id === _univerjs_docs.RichTextEditingMutation.id) queueMicrotask(() => {
				this._updateDrawingsEditStatus();
			});
		}));
	}
	_setDrawingSelections(params) {
		var _unit$getSnapshot$bod, _unit$getSnapshot$bod2;
		const { unit } = this._context;
		const customBlocks = (_unit$getSnapshot$bod = (_unit$getSnapshot$bod2 = unit.getSnapshot().body) === null || _unit$getSnapshot$bod2 === void 0 ? void 0 : _unit$getSnapshot$bod2.customBlocks) !== null && _unit$getSnapshot$bod !== void 0 ? _unit$getSnapshot$bod : [];
		const ranges = params.map((item) => {
			const id = item.drawingId;
			const block = customBlocks.find((b) => b.blockId === id);
			if (block) return block.startIndex;
			return null;
		}).filter((e) => e !== null).map((offset) => ({
			startOffset: offset,
			endOffset: offset + 1
		}));
		this._docSelectionManagerService.replaceDocRanges(ranges);
	}
};
DocDrawingUpdateRenderController = __decorate([
	__decorateParam(1, _univerjs_core.ICommandService),
	__decorateParam(2, (0, _univerjs_core.Inject)(_univerjs_docs.DocSelectionManagerService)),
	__decorateParam(3, _univerjs_engine_render.IRenderManagerService),
	__decorateParam(4, _univerjs_drawing.IImageIoService),
	__decorateParam(5, _univerjs_docs_drawing.IDocDrawingService),
	__decorateParam(6, _univerjs_drawing.IDrawingManagerService),
	__decorateParam(7, _univerjs_core.IContextService),
	__decorateParam(8, _univerjs_ui.IMessageService),
	__decorateParam(9, (0, _univerjs_core.Inject)(_univerjs_core.LocaleService)),
	__decorateParam(10, (0, _univerjs_core.Inject)(_univerjs_docs_ui.DocSelectionRenderService)),
	__decorateParam(11, (0, _univerjs_core.Inject)(DocRefreshDrawingsService)),
	__decorateParam(12, _univerjs_ui.ILocalFileService)
], DocDrawingUpdateRenderController);

//#endregion
//#region src/commands/commands/insert-image.command.ts
const InsertDocImageCommand = {
	id: "doc.command.insert-float-image",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor) => {
		var _getCurrentTypeOfRend, _getCurrentTypeOfRend2;
		const univerInstanceService = accessor.get(_univerjs_core.IUniverInstanceService);
		const renderManagerService = accessor.get(_univerjs_engine_render.IRenderManagerService);
		return (_getCurrentTypeOfRend = (_getCurrentTypeOfRend2 = (0, _univerjs_engine_render.getCurrentTypeOfRenderer)(_univerjs_core.UniverInstanceType.UNIVER_DOC, univerInstanceService, renderManagerService)) === null || _getCurrentTypeOfRend2 === void 0 ? void 0 : _getCurrentTypeOfRend2.with(DocDrawingUpdateRenderController).insertDocImage()) !== null && _getCurrentTypeOfRend !== void 0 ? _getCurrentTypeOfRend : false;
	}
};

//#endregion
//#region src/commands/commands/update-doc-drawing.command.ts
const WRAPPING_STYLE_TO_LAYOUT_TYPE = {
	["inline"]: _univerjs_core.PositionedObjectLayoutType.INLINE,
	["wrapSquare"]: _univerjs_core.PositionedObjectLayoutType.WRAP_SQUARE,
	["wrapTopAndBottom"]: _univerjs_core.PositionedObjectLayoutType.WRAP_TOP_AND_BOTTOM,
	["inFrontOfText"]: _univerjs_core.PositionedObjectLayoutType.WRAP_NONE,
	["behindText"]: _univerjs_core.PositionedObjectLayoutType.WRAP_NONE
};
function findDrawingAnchorInPage(page, drawingId, pageMarginTop, pageMarginLeft) {
	const skeDrawing = page.skeDrawings.get(drawingId);
	if (skeDrawing) return {
		skeDrawing,
		pageMarginTop,
		pageMarginLeft
	};
	for (const table of page.skeTables.values()) for (const row of table.rows) for (const cell of row.cells) {
		const cellAnchor = findDrawingAnchorInPage(cell, drawingId, cell.marginTop, cell.marginLeft);
		if (cellAnchor) return cellAnchor;
	}
	return null;
}
function getDeleteAndInsertCustomBlockActions(segmentId, oldSegmentId, segmentPage, offset, drawingId, documentDataModel, docSelectionRenderManager) {
	var _oldBody$customBlocks;
	const textX = new _univerjs_core.TextX();
	const jsonX = _univerjs_core.JSONX.getInstance();
	const rawActions = [];
	const oldBody = documentDataModel.getSelfOrHeaderFooterModel(oldSegmentId).getBody();
	const body = documentDataModel.getSelfOrHeaderFooterModel(segmentId).getBody();
	if (oldBody == null || body == null) return;
	const oldOffset = (_oldBody$customBlocks = oldBody.customBlocks) === null || _oldBody$customBlocks === void 0 || (_oldBody$customBlocks = _oldBody$customBlocks.find((block) => block.blockId === drawingId)) === null || _oldBody$customBlocks === void 0 ? void 0 : _oldBody$customBlocks.startIndex;
	if (oldOffset == null) return;
	offset = Math.min(body.dataStream.length - 2, offset);
	if (segmentId === oldSegmentId) {
		if (offset < oldOffset) {
			if (offset > 0) textX.push({
				t: _univerjs_core.TextXActionType.RETAIN,
				len: offset
			});
			textX.push({
				t: _univerjs_core.TextXActionType.INSERT,
				body: {
					dataStream: "\b",
					customBlocks: [{
						startIndex: 0,
						blockId: drawingId
					}]
				},
				len: 1
			});
			textX.push({
				t: _univerjs_core.TextXActionType.RETAIN,
				len: oldOffset - offset
			});
			textX.push({
				t: _univerjs_core.TextXActionType.DELETE,
				len: 1
			});
		} else {
			if (oldOffset > 0) textX.push({
				t: _univerjs_core.TextXActionType.RETAIN,
				len: oldOffset
			});
			textX.push({
				t: _univerjs_core.TextXActionType.DELETE,
				len: 1
			});
			if (offset - oldOffset - 1 > 0) textX.push({
				t: _univerjs_core.TextXActionType.RETAIN,
				len: offset - oldOffset - 1
			});
			textX.push({
				t: _univerjs_core.TextXActionType.INSERT,
				body: {
					dataStream: "\b",
					customBlocks: [{
						startIndex: 0,
						blockId: drawingId
					}]
				},
				len: 1
			});
		}
		if (offset !== oldOffset) {
			const path = (0, _univerjs_docs_ui.getRichTextEditPath)(documentDataModel, oldSegmentId);
			const action = jsonX.editOp(textX.serialize(), path);
			rawActions.push(action);
		}
	} else {
		if (oldOffset > 0) textX.push({
			t: _univerjs_core.TextXActionType.RETAIN,
			len: oldOffset
		});
		textX.push({
			t: _univerjs_core.TextXActionType.DELETE,
			len: 1
		});
		let path = (0, _univerjs_docs_ui.getRichTextEditPath)(documentDataModel, oldSegmentId);
		let action = jsonX.editOp(textX.serialize(), path);
		rawActions.push(action);
		textX.empty();
		if (offset > 0) textX.push({
			t: _univerjs_core.TextXActionType.RETAIN,
			len: offset
		});
		textX.push({
			t: _univerjs_core.TextXActionType.INSERT,
			body: {
				dataStream: "\b",
				customBlocks: [{
					startIndex: 0,
					blockId: drawingId
				}]
			},
			len: 1
		});
		path = (0, _univerjs_docs_ui.getRichTextEditPath)(documentDataModel, segmentId);
		action = jsonX.editOp(textX.serialize(), path);
		rawActions.push(action);
		docSelectionRenderManager.setSegment(segmentId);
		docSelectionRenderManager.setSegmentPage(segmentPage);
	}
	return rawActions;
}
/**
* The command to update drawing wrapping style.
*/
const UpdateDocDrawingWrappingStyleCommand = {
	id: "doc.command.update-doc-drawing-wrapping-style",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		if (params == null) return false;
		const { drawings, wrappingStyle, unitId } = params;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const univerInstanceService = accessor.get(_univerjs_core.IUniverInstanceService);
		const renderObject = accessor.get(_univerjs_engine_render.IRenderManagerService).getRenderById(unitId);
		const skeletonData = renderObject === null || renderObject === void 0 ? void 0 : renderObject.with(_univerjs_docs.DocSkeletonManagerService).getSkeleton().getSkeletonData();
		const viewModel = renderObject === null || renderObject === void 0 ? void 0 : renderObject.with(_univerjs_docs.DocSkeletonManagerService).getViewModel();
		const scene = renderObject === null || renderObject === void 0 ? void 0 : renderObject.scene;
		const documentDataModel = univerInstanceService.getCurrentUniverDocInstance();
		if (documentDataModel == null || skeletonData == null || scene == null || viewModel == null) return false;
		const editArea = viewModel.getEditArea();
		const transformer = scene.getTransformerByCreate();
		const { pages, skeHeaders, skeFooters } = skeletonData;
		const jsonX = _univerjs_core.JSONX.getInstance();
		const rawActions = [];
		const { drawings: oldDrawings = {} } = documentDataModel.getSnapshot();
		for (const drawing of drawings) {
			const { drawingId } = drawing;
			const oldLayoutType = oldDrawings[drawingId].layoutType;
			const newLayoutType = WRAPPING_STYLE_TO_LAYOUT_TYPE[wrappingStyle];
			if (oldLayoutType !== newLayoutType) {
				const updateLayoutTypeAction = jsonX.replaceOp([
					"drawings",
					drawingId,
					"layoutType"
				], oldLayoutType, newLayoutType);
				rawActions.push(updateLayoutTypeAction);
			}
			if (wrappingStyle === "behindText" || wrappingStyle === "inFrontOfText") {
				const oldBehindDoc = oldDrawings[drawingId].behindDoc;
				const newBehindDoc = wrappingStyle === "behindText" ? _univerjs_core.BooleanNumber.TRUE : _univerjs_core.BooleanNumber.FALSE;
				if (oldBehindDoc !== newBehindDoc) {
					const updateBehindDocAction = jsonX.replaceOp([
						"drawings",
						drawingId,
						"behindDoc"
					], oldBehindDoc, newBehindDoc);
					rawActions.push(updateBehindDocAction);
				}
			}
			if (wrappingStyle === "inline") continue;
			let drawingAnchor = null;
			for (const page of pages) {
				const { headerId, footerId, marginTop, marginLeft, marginBottom, pageWidth, pageHeight } = page;
				switch (editArea) {
					case _univerjs_engine_render.DocumentEditArea.HEADER: {
						var _skeHeaders$get;
						const headerSke = (_skeHeaders$get = skeHeaders.get(headerId)) === null || _skeHeaders$get === void 0 ? void 0 : _skeHeaders$get.get(pageWidth);
						if (headerSke != null) drawingAnchor = findDrawingAnchorInPage(headerSke, drawingId, headerSke.marginTop, marginLeft);
						break;
					}
					case _univerjs_engine_render.DocumentEditArea.FOOTER: {
						var _skeFooters$get;
						const footerSke = (_skeFooters$get = skeFooters.get(footerId)) === null || _skeFooters$get === void 0 ? void 0 : _skeFooters$get.get(pageWidth);
						if (footerSke != null) drawingAnchor = findDrawingAnchorInPage(footerSke, drawingId, pageHeight - marginBottom + footerSke.marginTop, marginLeft);
						break;
					}
					case _univerjs_engine_render.DocumentEditArea.BODY:
						drawingAnchor = findDrawingAnchorInPage(page, drawingId, marginTop, marginLeft);
						break;
				}
				if (drawingAnchor != null) break;
			}
			if (drawingAnchor != null) {
				const { skeDrawing, pageMarginTop, pageMarginLeft } = drawingAnchor;
				const { aTop, aLeft } = skeDrawing;
				const oldPositionH = oldDrawings[drawingId].docTransform.positionH;
				let posOffsetH = aLeft;
				if (oldPositionH.relativeFrom === _univerjs_core.ObjectRelativeFromH.MARGIN) posOffsetH -= pageMarginLeft;
				else if (oldPositionH.relativeFrom === _univerjs_core.ObjectRelativeFromH.COLUMN) posOffsetH -= skeDrawing.columnLeft;
				const newPositionH = {
					relativeFrom: oldPositionH.relativeFrom,
					posOffset: posOffsetH
				};
				if (oldPositionH.posOffset !== newPositionH.posOffset) {
					const action = jsonX.replaceOp([
						"drawings",
						drawingId,
						"docTransform",
						"positionH"
					], oldPositionH, newPositionH);
					rawActions.push(action);
				}
				const oldPositionV = oldDrawings[drawingId].docTransform.positionV;
				let posOffsetV = aTop;
				if (oldPositionV.relativeFrom === _univerjs_core.ObjectRelativeFromV.PAGE) posOffsetV += pageMarginTop;
				else if (oldPositionV.relativeFrom === _univerjs_core.ObjectRelativeFromV.LINE) posOffsetV -= skeDrawing.lineTop;
				else if (oldPositionV.relativeFrom === _univerjs_core.ObjectRelativeFromV.PARAGRAPH) posOffsetV -= skeDrawing.blockAnchorTop;
				const newPositionV = {
					relativeFrom: oldPositionV.relativeFrom,
					posOffset: posOffsetV
				};
				if (oldPositionV.posOffset !== newPositionV.posOffset) {
					const action = jsonX.replaceOp([
						"drawings",
						drawingId,
						"docTransform",
						"positionV"
					], oldPositionV, newPositionV);
					rawActions.push(action);
				}
			}
		}
		const doMutation = {
			id: _univerjs_docs.RichTextEditingMutation.id,
			params: {
				unitId,
				actions: [],
				textRanges: null
			}
		};
		doMutation.params.actions = rawActions.reduce((acc, cur) => {
			return _univerjs_core.JSONX.compose(acc, cur);
		}, null);
		const result = commandService.syncExecuteCommand(doMutation.id, doMutation.params);
		transformer.refreshControls();
		return Boolean(result);
	}
};
/**
* The command to update drawing wrap text.
*/
const UpdateDocDrawingDistanceCommand = {
	id: "doc.command.update-doc-drawing-distance",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		if (params == null) return false;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const documentDataModel = accessor.get(_univerjs_core.IUniverInstanceService).getCurrentUniverDocInstance();
		if (documentDataModel == null) return false;
		const { drawings, dist, unitId } = params;
		const jsonX = _univerjs_core.JSONX.getInstance();
		const rawActions = [];
		const { drawings: oldDrawings = {} } = documentDataModel.getSnapshot();
		for (const drawing of drawings) {
			const { drawingId } = drawing;
			for (const [key, value] of Object.entries(dist)) {
				const oldValue = oldDrawings[drawingId][key];
				if (oldValue !== value) {
					const action = jsonX.replaceOp([
						"drawings",
						drawingId,
						key
					], oldValue, value);
					rawActions.push(action);
				}
			}
		}
		const doMutation = {
			id: _univerjs_docs.RichTextEditingMutation.id,
			params: {
				unitId,
				actions: [],
				textRanges: null
			}
		};
		doMutation.params.actions = rawActions.reduce((acc, cur) => {
			return _univerjs_core.JSONX.compose(acc, cur);
		}, null);
		const result = commandService.syncExecuteCommand(doMutation.id, doMutation.params);
		return Boolean(result);
	}
};
/**
* The command to update drawing wrap text.
*/
const UpdateDocDrawingWrapTextCommand = {
	id: "doc.command.update-doc-drawing-wrap-text",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		if (params == null) return false;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const documentDataModel = accessor.get(_univerjs_core.IUniverInstanceService).getCurrentUniverDocInstance();
		if (documentDataModel == null) return false;
		const { drawings, wrapText, unitId } = params;
		const jsonX = _univerjs_core.JSONX.getInstance();
		const rawActions = [];
		const { drawings: oldDrawings = {} } = documentDataModel.getSnapshot();
		for (const drawing of drawings) {
			const { drawingId } = drawing;
			const oldWrapText = oldDrawings[drawingId].wrapText;
			if (oldWrapText !== wrapText) {
				const action = jsonX.replaceOp([
					"drawings",
					drawingId,
					"wrapText"
				], oldWrapText, wrapText);
				rawActions.push(action);
			}
		}
		const doMutation = {
			id: _univerjs_docs.RichTextEditingMutation.id,
			params: {
				unitId,
				actions: [],
				textRanges: null
			}
		};
		doMutation.params.actions = rawActions.reduce((acc, cur) => {
			return _univerjs_core.JSONX.compose(acc, cur);
		}, null);
		const result = commandService.syncExecuteCommand(doMutation.id, doMutation.params);
		return Boolean(result);
	}
};
/**
* The command to update drawing position.
*/
const UpdateDrawingDocTransformCommand = {
	id: "doc.command.update-drawing-doc-transform",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		if (params == null) return false;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const univerInstanceService = accessor.get(_univerjs_core.IUniverInstanceService);
		const renderObject = accessor.get(_univerjs_engine_render.IRenderManagerService).getRenderById(params.unitId);
		const scene = renderObject === null || renderObject === void 0 ? void 0 : renderObject.scene;
		if (scene == null) return false;
		const transformer = scene.getTransformerByCreate();
		const documentDataModel = univerInstanceService.getCurrentUniverDocInstance();
		if (documentDataModel == null) return false;
		const { drawings, unitId } = params;
		const jsonX = _univerjs_core.JSONX.getInstance();
		const rawActions = [];
		const { drawings: oldDrawings = {} } = documentDataModel.getSnapshot();
		for (const drawing of drawings) {
			const { drawingId, key, value } = drawing;
			const oldValue = oldDrawings[drawingId].docTransform[key];
			if (!_univerjs_core.Tools.diffValue(oldValue, value)) {
				const action = jsonX.replaceOp([
					"drawings",
					drawingId,
					"docTransform",
					key
				], oldValue, value);
				rawActions.push(action);
			}
		}
		const doMutation = {
			id: _univerjs_docs.RichTextEditingMutation.id,
			params: {
				unitId,
				actions: [],
				textRanges: null,
				debounce: true
			}
		};
		doMutation.params.actions = rawActions.reduce((acc, cur) => {
			return _univerjs_core.JSONX.compose(acc, cur);
		}, null);
		const result = commandService.syncExecuteCommand(doMutation.id, doMutation.params);
		transformer.refreshControls();
		return Boolean(result);
	}
};
/**
* The command to move inline drawing.
*/
const IMoveInlineDrawingCommand = {
	id: "doc.command.move-inline-drawing",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		var _renderManagerService, _docSelectionRenderSe;
		if (params == null) return false;
		const renderManagerService = accessor.get(_univerjs_engine_render.IRenderManagerService);
		const docSelectionRenderService = (_renderManagerService = renderManagerService.getRenderById(params.unitId)) === null || _renderManagerService === void 0 ? void 0 : _renderManagerService.with(_univerjs_docs_ui.DocSelectionRenderService);
		const docRefreshDrawingsService = accessor.get(DocRefreshDrawingsService);
		const renderObject = renderManagerService.getRenderById(params.unitId);
		const scene = renderObject === null || renderObject === void 0 ? void 0 : renderObject.scene;
		const skeleton = renderObject === null || renderObject === void 0 ? void 0 : renderObject.with(_univerjs_docs.DocSkeletonManagerService).getSkeleton();
		if (scene == null || docSelectionRenderService == null) return false;
		const transformer = scene.getTransformerByCreate();
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const documentDataModel = accessor.get(_univerjs_core.IUniverInstanceService).getCurrentUniverDocInstance();
		if (documentDataModel == null) return false;
		const { drawing, unitId, offset, segmentId: newSegmentId, segmentPage, needRefreshDrawings } = params;
		if (needRefreshDrawings) {
			docRefreshDrawingsService.refreshDrawings(skeleton);
			transformer.refreshControls();
			return true;
		}
		const rawActions = [];
		const { drawingId } = drawing;
		const actions = getDeleteAndInsertCustomBlockActions(newSegmentId, (_docSelectionRenderSe = docSelectionRenderService.getSegment()) !== null && _docSelectionRenderSe !== void 0 ? _docSelectionRenderSe : "", segmentPage, offset, drawingId, documentDataModel, docSelectionRenderService);
		if (actions == null || actions.length === 0) {
			docRefreshDrawingsService.refreshDrawings(skeleton);
			transformer.refreshControls();
			return false;
		}
		rawActions.push(...actions);
		const doMutation = {
			id: _univerjs_docs.RichTextEditingMutation.id,
			params: {
				unitId,
				actions: [],
				textRanges: null
			}
		};
		doMutation.params.actions = rawActions.reduce((acc, cur) => {
			return _univerjs_core.JSONX.compose(acc, cur);
		}, null);
		const result = commandService.syncExecuteCommand(doMutation.id, doMutation.params);
		transformer.refreshControls();
		return Boolean(result);
	}
};
/**
* The command to transform non-inline drawing.
*/
const ITransformNonInlineDrawingCommand = {
	id: "doc.command.transform-non-inline-drawing",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		var _renderManagerService2, _docSelectionRenderSe2;
		if (params == null) return false;
		const renderManagerService = accessor.get(_univerjs_engine_render.IRenderManagerService);
		const docSelectionRenderService = (_renderManagerService2 = renderManagerService.getRenderById(params.unitId)) === null || _renderManagerService2 === void 0 ? void 0 : _renderManagerService2.with(_univerjs_docs_ui.DocSelectionRenderService);
		const renderObject = renderManagerService.getRenderById(params.unitId);
		const scene = renderObject === null || renderObject === void 0 ? void 0 : renderObject.scene;
		if (scene == null || docSelectionRenderService == null) return false;
		const transformer = scene.getTransformerByCreate();
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const documentDataModel = accessor.get(_univerjs_core.IUniverInstanceService).getCurrentUniverDocInstance();
		if (documentDataModel == null) return false;
		const { drawing, unitId, offset, docTransform, segmentId: newSegmentId, segmentPage } = params;
		const rawActions = [];
		const { drawingId } = drawing;
		const actions = getDeleteAndInsertCustomBlockActions(newSegmentId, (_docSelectionRenderSe2 = docSelectionRenderService.getSegment()) !== null && _docSelectionRenderSe2 !== void 0 ? _docSelectionRenderSe2 : "", segmentPage, offset, drawingId, documentDataModel, docSelectionRenderService);
		if (actions == null) return false;
		if (actions.length > 0) rawActions.push(...actions);
		const jsonX = _univerjs_core.JSONX.getInstance();
		const { drawings: oldDrawings = {} } = documentDataModel.getSnapshot();
		const { positionH: oldPositionH, positionV: oldPositionV, size: oldSize, angle: oldAngle } = oldDrawings[drawingId].docTransform;
		if (!_univerjs_core.Tools.diffValue(oldPositionH, docTransform.positionH)) {
			const updateAction = jsonX.replaceOp([
				"drawings",
				drawingId,
				"docTransform",
				"positionH"
			], oldPositionH, docTransform.positionH);
			rawActions.push(updateAction);
		}
		if (!_univerjs_core.Tools.diffValue(oldPositionV, docTransform.positionV)) {
			const updateAction = jsonX.replaceOp([
				"drawings",
				drawingId,
				"docTransform",
				"positionV"
			], oldPositionV, docTransform.positionV);
			rawActions.push(updateAction);
		}
		if (!_univerjs_core.Tools.diffValue(oldSize, docTransform.size)) {
			const updateAction = jsonX.replaceOp([
				"drawings",
				drawingId,
				"docTransform",
				"size"
			], oldSize, docTransform.size);
			rawActions.push(updateAction);
		}
		if (!_univerjs_core.Tools.diffValue(oldAngle, docTransform.angle)) {
			const updateAction = jsonX.replaceOp([
				"drawings",
				drawingId,
				"docTransform",
				"angle"
			], oldAngle, docTransform.angle);
			rawActions.push(updateAction);
		}
		const doMutation = {
			id: _univerjs_docs.RichTextEditingMutation.id,
			params: {
				unitId,
				actions: [],
				textRanges: null,
				debounce: true
			}
		};
		doMutation.params.actions = rawActions.reduce((acc, cur) => {
			return _univerjs_core.JSONX.compose(acc, cur);
		}, null);
		const result = commandService.syncExecuteCommand(doMutation.id, doMutation.params);
		transformer.refreshControls();
		return Boolean(result);
	}
};

//#endregion
//#region src/commands/commands/move-drawings.command.ts
const MoveDocDrawingsCommand = {
	id: "doc.command.move-drawing",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const docDrawingService = accessor.get(_univerjs_docs_drawing.IDocDrawingService);
		const univerInstanceService = accessor.get(_univerjs_core.IUniverInstanceService);
		const renderManagerService = accessor.get(_univerjs_engine_render.IRenderManagerService);
		const { direction } = params;
		const drawings = docDrawingService.getFocusDrawings();
		if (drawings.length === 0) return false;
		const unitId = drawings[0].unitId;
		const renderObject = renderManagerService.getRenderById(unitId);
		const scene = renderObject === null || renderObject === void 0 ? void 0 : renderObject.scene;
		if (scene == null) return false;
		const transformer = scene.getTransformerByCreate();
		const documentDataModel = univerInstanceService.getUniverDocInstance(unitId);
		const newDrawings = drawings.map((drawing) => {
			var _documentDataModel$ge;
			const { drawingId } = drawing;
			const drawingData = documentDataModel === null || documentDataModel === void 0 || (_documentDataModel$ge = documentDataModel.getSnapshot().drawings) === null || _documentDataModel$ge === void 0 ? void 0 : _documentDataModel$ge[drawingId];
			if (drawingData == null || drawingData.layoutType === _univerjs_core.PositionedObjectLayoutType.INLINE) return null;
			const { positionH, positionV } = drawingData.docTransform;
			const newPositionH = { ...positionH };
			const newPositionV = { ...positionV };
			if (direction === _univerjs_core.Direction.UP) {
				var _newPositionV$posOffs;
				newPositionV.posOffset = ((_newPositionV$posOffs = newPositionV.posOffset) !== null && _newPositionV$posOffs !== void 0 ? _newPositionV$posOffs : 0) - 2;
			} else if (direction === _univerjs_core.Direction.DOWN) {
				var _newPositionV$posOffs2;
				newPositionV.posOffset = ((_newPositionV$posOffs2 = newPositionV.posOffset) !== null && _newPositionV$posOffs2 !== void 0 ? _newPositionV$posOffs2 : 0) + 2;
			} else if (direction === _univerjs_core.Direction.LEFT) {
				var _newPositionH$posOffs;
				newPositionH.posOffset = ((_newPositionH$posOffs = newPositionH.posOffset) !== null && _newPositionH$posOffs !== void 0 ? _newPositionH$posOffs : 0) - 2;
			} else if (direction === _univerjs_core.Direction.RIGHT) {
				var _newPositionH$posOffs2;
				newPositionH.posOffset = ((_newPositionH$posOffs2 = newPositionH.posOffset) !== null && _newPositionH$posOffs2 !== void 0 ? _newPositionH$posOffs2 : 0) + 2;
			}
			return {
				drawingId,
				key: direction === _univerjs_core.Direction.UP || direction === _univerjs_core.Direction.DOWN ? "positionV" : "positionH",
				value: direction === _univerjs_core.Direction.UP || direction === _univerjs_core.Direction.DOWN ? newPositionV : newPositionH
			};
		}).filter((drawing) => drawing != null);
		if (newDrawings.length === 0) return false;
		const result = commandService.syncExecuteCommand(UpdateDrawingDocTransformCommand.id, {
			unitId,
			subUnitId: unitId,
			drawings: newDrawings
		});
		transformer.refreshControls();
		return Boolean(result);
	}
};

//#endregion
//#region src/commands/operations/clear-drawing-transformer.operation.ts
const ClearDocDrawingTransformerOperation = {
	id: "doc.operation.clear-drawing-transformer",
	type: _univerjs_core.CommandType.MUTATION,
	handler: (accessor, params) => {
		const renderManagerService = accessor.get(_univerjs_engine_render.IRenderManagerService);
		params.forEach((unitId) => {
			var _renderManagerService;
			(_renderManagerService = renderManagerService.getRenderById(unitId)) === null || _renderManagerService === void 0 || (_renderManagerService = _renderManagerService.scene.getTransformer()) === null || _renderManagerService === void 0 || _renderManagerService.debounceRefreshControls();
		});
		return true;
	}
};

//#endregion
//#region src/views/doc-image-panel/component-name.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const COMPONENT_DOC_DRAWING_PANEL = "COMPONENT_DOC_DRAWING_PANEL";

//#endregion
//#region src/commands/operations/open-drawing-panel.operation.ts
const SidebarDocDrawingOperation = {
	id: "sidebar.operation.doc-image",
	type: _univerjs_core.CommandType.COMMAND,
	handler: async (accessor, params) => {
		const sidebarService = accessor.get(_univerjs_ui.ISidebarService);
		const localeService = accessor.get(_univerjs_core.LocaleService);
		const drawingManagerService = accessor.get(_univerjs_drawing.IDrawingManagerService);
		switch (params.value) {
			case "open":
				sidebarService.open({
					header: { title: localeService.t("docs-drawing-ui.panel.title") },
					children: { label: COMPONENT_DOC_DRAWING_PANEL },
					onClose: () => {
						drawingManagerService.focusDrawing(null);
					},
					width: 360
				});
				break;
			default:
				sidebarService.close();
				break;
		}
		return true;
	}
};

//#endregion
//#region src/commands/operations/edit-doc-drawing.operation.ts
const EditDocDrawingOperation = {
	id: "doc.operation.edit-doc-image",
	type: _univerjs_core.CommandType.OPERATION,
	handler: (accessor, params) => {
		const drawingManagerService = accessor.get(_univerjs_drawing.IDrawingManagerService);
		const commandService = accessor.get(_univerjs_core.ICommandService);
		if (params == null) return false;
		drawingManagerService.focusDrawing([params]);
		commandService.executeCommand(SidebarDocDrawingOperation.id, { value: "open" });
		return true;
	}
};

//#endregion
//#region src/controllers/doc-float-dom.controller.ts
function calcDocFloatDomPositionByRect(rect, scene, opacity = 1, angle = 0) {
	const { top, left, bottom, right } = rect;
	const width = right - left;
	const height = bottom - top;
	const { viewportScrollX, viewportScrollY } = scene.getViewport(_univerjs_docs_ui.VIEWPORT_KEY.VIEW_MAIN);
	const { scaleX, scaleY } = scene.getAncestorScale();
	return {
		startX: (left - viewportScrollX) * scaleX,
		startY: (top - viewportScrollY) * scaleY,
		endX: (left + width - viewportScrollX) * scaleX,
		endY: (top + height - viewportScrollY) * scaleY,
		width: width * scaleX,
		height: height * scaleY,
		rotate: angle,
		absolute: {
			left: false,
			top: false
		},
		opacity: opacity !== null && opacity !== void 0 ? opacity : 1
	};
}
function calcDocFloatDomPosition(object, renderUnit) {
	const { top, left, width, height, angle, opacity } = object;
	return calcDocFloatDomPositionByRect({
		top,
		left,
		bottom: top + height,
		right: left + width
	}, renderUnit.scene, opacity, angle);
}
let DocFloatDomController = class DocFloatDomController extends _univerjs_core.Disposable {
	constructor(_renderManagerService, _drawingManagerService, _drawingRenderService, _canvasFloatDomService, _univerInstanceService, _commandService) {
		super();
		this._renderManagerService = _renderManagerService;
		this._drawingManagerService = _drawingManagerService;
		this._drawingRenderService = _drawingRenderService;
		this._canvasFloatDomService = _canvasFloatDomService;
		this._univerInstanceService = _univerInstanceService;
		this._commandService = _commandService;
		_defineProperty(this, "_domLayerInfoMap", /* @__PURE__ */ new Map());
		this._initialize();
	}
	dispose() {
		super.dispose();
	}
	_initialize() {
		this._drawingAddRemoveListener();
		this._initScrollAndZoomEvent();
	}
	_getSceneAndTransformerByDrawingSearch(unitId) {
		if (unitId == null) return;
		const renderObject = this._renderManagerService.getRenderById(unitId);
		if (renderObject == null) return null;
		const scene = renderObject.scene;
		return {
			scene,
			transformer: scene.getTransformerByCreate(),
			renderUnit: renderObject,
			canvas: renderObject.engine.getCanvasElement()
		};
	}
	_drawingAddRemoveListener() {
		this.disposeWithMe(this._drawingManagerService.add$.subscribe((params) => {
			this._insertRects(params);
		}));
		this.disposeWithMe(this._drawingManagerService.remove$.subscribe((params) => {
			params.forEach((param) => {
				this._removeDom(param.drawingId);
			});
		}));
	}
	_insertRects(params) {
		params.forEach(async (param) => {
			const { unitId } = param;
			if (!this._univerInstanceService.getUnit(unitId, _univerjs_core.UniverInstanceType.UNIVER_DOC)) return;
			const renderObject = this._getSceneAndTransformerByDrawingSearch(unitId);
			if (renderObject == null) return;
			const rectParam = this._drawingManagerService.getDrawingByParam(param);
			if (rectParam == null) return;
			const rects = await this._drawingRenderService.renderFloatDom(rectParam, renderObject.scene);
			if (rects == null || rects.length === 0) return;
			for (const rect of rects) {
				this._addHoverForRect(rect);
				const disposableCollection = new _univerjs_core.DisposableCollection();
				const position$ = new rxjs.BehaviorSubject(calcDocFloatDomPosition(rect, renderObject.renderUnit));
				const canvas = renderObject.canvas;
				const data = rectParam.data;
				const info = {
					dispose: disposableCollection,
					rect,
					position$,
					unitId
				};
				this._canvasFloatDomService.addFloatDom({
					position$,
					id: rectParam.drawingId,
					componentKey: rectParam.componentKey,
					onPointerDown: (evt) => {
						canvas.dispatchEvent(new PointerEvent(evt.type, evt));
					},
					onPointerMove: (evt) => {
						canvas.dispatchEvent(new PointerEvent(evt.type, evt));
					},
					onPointerUp: (evt) => {
						canvas.dispatchEvent(new PointerEvent(evt.type, evt));
					},
					onWheel: (evt) => {
						canvas.dispatchEvent(new WheelEvent(evt.type, evt));
					},
					data,
					unitId
				});
				const listener = rect.onTransformChange$.subscribeEvent(() => {
					const newPosition = calcDocFloatDomPosition(rect, renderObject.renderUnit);
					position$.next(newPosition);
				});
				disposableCollection.add(() => {
					this._canvasFloatDomService.removeFloatDom(rectParam.drawingId);
				});
				listener && disposableCollection.add(listener);
				this._domLayerInfoMap.set(rectParam.drawingId, info);
			}
		});
	}
	_addHoverForRect(o) {
		this.disposeWithMe((0, _univerjs_core.toDisposable)(o.onPointerEnter$.subscribeEvent(() => {
			o.cursor = _univerjs_engine_render.CURSOR_TYPE.GRAB;
		})));
		this.disposeWithMe((0, _univerjs_core.toDisposable)(o.onPointerLeave$.subscribeEvent(() => {
			o.cursor = _univerjs_engine_render.CURSOR_TYPE.DEFAULT;
		})));
	}
	_removeDom(id) {
		const info = this._domLayerInfoMap.get(id);
		if (!info) return;
		const { unitId } = info;
		this._domLayerInfoMap.delete(id);
		info.dispose.dispose();
		const renderObject = this._getSceneAndTransformerByDrawingSearch(unitId);
		if (renderObject) renderObject.scene.removeObject(info.rect);
	}
	_initScrollAndZoomEvent() {
		const updateDoc = (unitId) => {
			const renderObject = this._getSceneAndTransformerByDrawingSearch(unitId);
			if (!renderObject) return;
			this._domLayerInfoMap.forEach((floatDomInfo) => {
				if (floatDomInfo.unitId !== unitId) return;
				const position = calcDocFloatDomPosition(floatDomInfo.rect, renderObject.renderUnit);
				floatDomInfo.position$.next(position);
			});
		};
		this.disposeWithMe(this._univerInstanceService.getCurrentTypeOfUnit$(_univerjs_core.UniverInstanceType.UNIVER_DOC).pipe((0, rxjs.map)((documentDataModel) => {
			if (!documentDataModel) return null;
			const unitId = documentDataModel.getUnitId();
			const render = this._renderManagerService.getRenderById(unitId);
			return render ? {
				render,
				unitId
			} : null;
		}), (0, rxjs.switchMap)((render) => render ? (0, _univerjs_core.fromEventSubject)(render.render.scene.getViewport(_univerjs_docs_ui.VIEWPORT_KEY.VIEW_MAIN).onScrollAfter$).pipe((0, rxjs.map)(() => ({ unitId: render.unitId }))) : (0, rxjs.of)(null))).subscribe((value) => {
			if (!value) return;
			const { unitId } = value;
			updateDoc(unitId);
		}));
		this.disposeWithMe(this._commandService.onCommandExecuted((commandInfo) => {
			if (commandInfo.id === _univerjs_docs_ui.SetDocZoomRatioOperation.id) {
				const { unitId } = commandInfo.params;
				updateDoc(unitId);
			}
		}));
	}
	insertFloatDom(floatDom, opts) {
		var _skeleton$getSkeleton, _opts$width, _opts$drawingId;
		const currentDoc = this._univerInstanceService.getCurrentUnitOfType(_univerjs_core.UniverInstanceType.UNIVER_DOC);
		if (!currentDoc) return false;
		const render = this._getSceneAndTransformerByDrawingSearch(currentDoc.getUnitId());
		if (!render) return false;
		const page = (_skeleton$getSkeleton = render.renderUnit.with(_univerjs_docs.DocSkeletonManagerService).getSkeleton().getSkeletonData()) === null || _skeleton$getSkeleton === void 0 ? void 0 : _skeleton$getSkeleton.pages[0];
		if (!page) return false;
		const { pageWidth, marginLeft, marginRight } = page;
		const width = pageWidth - marginLeft - marginRight;
		const docTransform = {
			size: {
				width: (_opts$width = opts.width) !== null && _opts$width !== void 0 ? _opts$width : width,
				height: opts.height
			},
			positionH: {
				relativeFrom: _univerjs_core.ObjectRelativeFromH.PAGE,
				posOffset: 0
			},
			positionV: {
				relativeFrom: _univerjs_core.ObjectRelativeFromV.PAGE,
				posOffset: 0
			},
			angle: 0
		};
		const drawingId = (_opts$drawingId = opts.drawingId) !== null && _opts$drawingId !== void 0 ? _opts$drawingId : (0, _univerjs_core.generateRandomId)();
		const params = {
			unitId: currentDoc.getUnitId(),
			drawings: [{
				drawingId,
				drawingType: _univerjs_core.DrawingTypeEnum.DRAWING_DOM,
				subUnitId: currentDoc.getUnitId(),
				unitId: currentDoc.getUnitId(),
				...floatDom,
				title: "",
				description: "",
				docTransform,
				layoutType: _univerjs_core.PositionedObjectLayoutType.INLINE,
				transform: (0, _univerjs_docs_ui.docDrawingPositionToTransform)(docTransform)
			}]
		};
		this._commandService.syncExecuteCommand(InsertDocDrawingCommand.id, params);
		return drawingId;
	}
};
DocFloatDomController = __decorate([
	__decorateParam(0, _univerjs_engine_render.IRenderManagerService),
	__decorateParam(1, _univerjs_drawing.IDrawingManagerService),
	__decorateParam(2, (0, _univerjs_core.Inject)(_univerjs_drawing_ui.DrawingRenderService)),
	__decorateParam(3, (0, _univerjs_core.Inject)(_univerjs_ui.CanvasFloatDomService)),
	__decorateParam(4, _univerjs_core.IUniverInstanceService),
	__decorateParam(5, _univerjs_core.ICommandService)
], DocFloatDomController);

//#endregion
//#region src/menu/image.menu.ts
const DOCS_IMAGE_MENU_ID = "doc.menu.image";
const IMAGE_MENU_UPLOAD_FLOAT_ID = InsertDocImageCommand.id;
const getDisableWhenSelectionInTableObservable = (accessor) => {
	const docSelectionManagerService = accessor.get(_univerjs_docs.DocSelectionManagerService);
	const univerInstanceService = accessor.get(_univerjs_core.IUniverInstanceService);
	return new rxjs.Observable((subscriber) => {
		const observable = docSelectionManagerService.textSelection$.subscribe(() => {
			const activeRange = docSelectionManagerService.getActiveTextRange();
			if (activeRange) {
				var _docDataModel$getSelf;
				const { segmentId, startOffset, endOffset } = activeRange;
				const docDataModel = univerInstanceService.getCurrentUniverDocInstance();
				const tables = docDataModel === null || docDataModel === void 0 || (_docDataModel$getSelf = docDataModel.getSelfOrHeaderFooterModel(segmentId).getBody()) === null || _docDataModel$getSelf === void 0 ? void 0 : _docDataModel$getSelf.tables;
				if (tables && tables.length) {
					if (tables.some((table) => {
						const { startIndex, endIndex } = table;
						return startOffset >= startIndex && startOffset < endIndex || endOffset >= startIndex && endOffset < endIndex;
					})) {
						subscriber.next(true);
						return;
					}
				}
			} else {
				subscriber.next(true);
				return;
			}
			subscriber.next(false);
		});
		return () => observable.unsubscribe();
	});
};
function ImageMenuFactory(accessor) {
	return {
		id: DOCS_IMAGE_MENU_ID,
		type: _univerjs_ui.MenuItemType.SUBITEMS,
		icon: "AddImageIcon",
		tooltip: "docs-drawing-ui.title",
		disabled$: getDisableWhenSelectionInTableObservable(accessor),
		hidden$: (0, _univerjs_ui.getMenuHiddenObservable)(accessor, _univerjs_core.UniverInstanceType.UNIVER_DOC, void 0, _univerjs_core.DOCS_ZEN_EDITOR_UNIT_ID_KEY)
	};
}
function UploadFloatImageMenuFactory(_accessor) {
	return {
		id: IMAGE_MENU_UPLOAD_FLOAT_ID,
		title: "docs-drawing-ui.upload.float",
		type: _univerjs_ui.MenuItemType.BUTTON,
		icon: "AddImageIcon",
		hidden$: (0, _univerjs_ui.getMenuHiddenObservable)(_accessor, _univerjs_core.UniverInstanceType.UNIVER_DOC, void 0, _univerjs_core.DOCS_ZEN_EDITOR_UNIT_ID_KEY)
	};
}

//#endregion
//#region package.json
var name = "@univerjs/docs-drawing-ui";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const DOCS_DRAWING_UI_PLUGIN_CONFIG_KEY = "docs-drawing-ui.config";
const configSymbol = Symbol(DOCS_DRAWING_UI_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/controllers/doc-drawing-notification.controller.ts
function getAddOrRemoveDrawings(actions) {
	if (_univerjs_core.JSONX.isNoop(actions) || !Array.isArray(actions)) return null;
	const drawingsOp = actions.find((action) => Array.isArray(action) && (action === null || action === void 0 ? void 0 : action[0]) === "drawings");
	if (drawingsOp == null || !Array.isArray(drawingsOp) || drawingsOp.length < 3) return null;
	if (typeof drawingsOp[1] === "string" && typeof drawingsOp[2] !== "object") return null;
	if (Array.isArray(drawingsOp[1]) && typeof drawingsOp[1][1] !== "object") return null;
	const drawings = [];
	if (Array.isArray(drawingsOp === null || drawingsOp === void 0 ? void 0 : drawingsOp[1])) {
		for (const op of drawingsOp) if (Array.isArray(op)) {
			var _ref, _ref2;
			drawings.push({
				type: (op === null || op === void 0 || (_ref = op[1]) === null || _ref === void 0 ? void 0 : _ref.i) ? "add" : "remove",
				drawingId: op === null || op === void 0 ? void 0 : op[0],
				drawing: op === null || op === void 0 || (_ref2 = op[1]) === null || _ref2 === void 0 ? void 0 : _ref2.i
			});
		}
	} else {
		var _drawingsOp$, _drawingsOp$2;
		drawings.push({
			type: ((_drawingsOp$ = drawingsOp[2]) === null || _drawingsOp$ === void 0 ? void 0 : _drawingsOp$.i) ? "add" : "remove",
			drawingId: drawingsOp[1],
			drawing: (_drawingsOp$2 = drawingsOp[2]) === null || _drawingsOp$2 === void 0 ? void 0 : _drawingsOp$2.i
		});
	}
	return drawings;
}
function getReOrderedDrawings(actions) {
	if (!Array.isArray(actions) || actions.length < 3 || actions[0] !== "drawingsOrder") return [];
	const drawingIndexes = [];
	for (let i = 1; i < actions.length; i++) {
		const action = actions[i];
		if (Array.isArray(action) && typeof action[0] === "number" && typeof action[1] === "object") drawingIndexes.push(action[0]);
		else {
			drawingIndexes.length = 0;
			break;
		}
	}
	return drawingIndexes;
}
let DocDrawingAddRemoveController = class DocDrawingAddRemoveController extends _univerjs_core.Disposable {
	constructor(_univerInstanceService, _commandService, _drawingManagerService, _docDrawingService, _renderManagerService) {
		super();
		this._univerInstanceService = _univerInstanceService;
		this._commandService = _commandService;
		this._drawingManagerService = _drawingManagerService;
		this._docDrawingService = _docDrawingService;
		this._renderManagerService = _renderManagerService;
		this._initialize();
	}
	_initialize() {
		this._commandExecutedListener();
	}
	_commandExecutedListener() {
		this.disposeWithMe(this._commandService.beforeCommandExecuted((command) => {
			if (command.id !== _univerjs_docs.RichTextEditingMutation.id) return;
			const { unitId, actions, isSync, syncer } = command.params;
			const addOrRemoveDrawings = getAddOrRemoveDrawings(actions);
			if (addOrRemoveDrawings != null) for (const { type, drawingId, drawing } of addOrRemoveDrawings) {
				if (isSync && (drawing === null || drawing === void 0 ? void 0 : drawing.unitId) === syncer) continue;
				if (type === "add") this._addDrawings(unitId, [drawing]);
				else this._removeDrawings(unitId, [drawingId]);
			}
		}));
		this.disposeWithMe(this._commandService.onCommandExecuted((command) => {
			if (command.id !== _univerjs_docs.RichTextEditingMutation.id) return;
			const { unitId, actions } = command.params;
			if (getReOrderedDrawings(actions).length > 0) this._updateDrawingsOrder(unitId);
		}));
		this.disposeWithMe(this._commandService.onCommandExecuted((command) => {
			var _this$_univerInstance;
			if (command.id !== _univerjs_core.UndoCommand.id && command.id !== _univerjs_core.RedoCommand.id) return;
			const unitId = (_this$_univerInstance = this._univerInstanceService.getCurrentUniverDocInstance()) === null || _this$_univerInstance === void 0 ? void 0 : _this$_univerInstance.getUnitId();
			const focusedDrawings = this._drawingManagerService.getFocusDrawings();
			if (unitId == null || focusedDrawings.length === 0) return;
			const renderObject = this._renderManagerService.getRenderById(unitId);
			const scene = renderObject === null || renderObject === void 0 ? void 0 : renderObject.scene;
			if (scene == null) return false;
			scene.getTransformerByCreate().refreshControls();
		}));
	}
	_addDrawings(unitId, drawings) {
		const drawingManagerService = this._drawingManagerService;
		const docDrawingService = this._docDrawingService;
		const { subUnitId, redo: op, objects } = this._docDrawingService.getBatchAddOp(drawings);
		drawingManagerService.applyJson1(unitId, subUnitId, op);
		docDrawingService.applyJson1(unitId, subUnitId, op);
		drawingManagerService.addNotification(objects);
		docDrawingService.addNotification(objects);
	}
	_removeDrawings(unitId, drawingIds) {
		const drawingManagerService = this._drawingManagerService;
		const docDrawingService = this._docDrawingService;
		const { subUnitId, redo: op, objects } = this._docDrawingService.getBatchRemoveOp(drawingIds.map((drawingId) => {
			return {
				unitId,
				subUnitId: unitId,
				drawingId
			};
		}));
		drawingManagerService.applyJson1(unitId, subUnitId, op);
		docDrawingService.applyJson1(unitId, subUnitId, op);
		drawingManagerService.removeNotification(objects);
		docDrawingService.removeNotification(objects);
	}
	_updateDrawingsOrder(unitId) {
		const documentDataModel = this._univerInstanceService.getUniverDocInstance(unitId);
		if (documentDataModel == null) return;
		const drawingsOrder = documentDataModel.getSnapshot().drawingsOrder;
		if (drawingsOrder == null) return;
		const drawingManagerService = this._drawingManagerService;
		const docDrawingService = this._docDrawingService;
		drawingManagerService.setDrawingOrder(unitId, unitId, drawingsOrder);
		docDrawingService.setDrawingOrder(unitId, unitId, drawingsOrder);
		const objects = {
			unitId,
			subUnitId: unitId,
			drawingIds: drawingsOrder
		};
		drawingManagerService.orderNotification(objects);
		docDrawingService.orderNotification(objects);
	}
};
DocDrawingAddRemoveController = __decorate([
	__decorateParam(0, _univerjs_core.IUniverInstanceService),
	__decorateParam(1, _univerjs_core.ICommandService),
	__decorateParam(2, _univerjs_drawing.IDrawingManagerService),
	__decorateParam(3, _univerjs_docs_drawing.IDocDrawingService),
	__decorateParam(4, _univerjs_engine_render.IRenderManagerService)
], DocDrawingAddRemoveController);

//#endregion
//#region src/views/printing-float-dom/index.tsx
const DocPrintingFloatDom = (props) => {
	const { floatDomInfos, scene, offset, bound } = props;
	const width = bound.right - bound.left;
	const height = bound.bottom - bound.top;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "univer-absolute univer-left-0 univer-top-0",
		children: (0, react.useMemo)(() => floatDomInfos.map((info) => {
			const { width = 0, height = 0, left = 0, top = 0 } = info.transform;
			const domPos = calcDocFloatDomPositionByRect({
				left,
				right: left + width,
				top,
				bottom: top + height
			}, scene);
			const floatDom = {
				position$: new rxjs.BehaviorSubject(domPos),
				position: domPos,
				id: info.drawingId,
				componentKey: info.componentKey,
				onPointerMove: () => {},
				onPointerDown: () => {},
				onPointerUp: () => {},
				onWheel: () => {},
				unitId: info.unitId,
				data: info.data
			};
			return [info.drawingId, floatDom];
		}).filter(([_, floatDom]) => !(floatDom.position.endX < 0 || floatDom.position.endY < 0 || floatDom.position.startX > width || floatDom.position.startY > height)), [
			floatDomInfos,
			scene,
			offset,
			width,
			height
		]).map(([id, floatDom]) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_ui.PrintFloatDomSingle, {
			layer: floatDom,
			id,
			position: floatDom.position
		}, id))
	});
};

//#endregion
//#region src/controllers/doc-drawing-printing.controller.tsx
let DocDrawingPrintingController = class DocDrawingPrintingController extends _univerjs_core.Disposable {
	constructor(_docPrintInterceptorService, _drawingRenderService, _drawingManagerService, _componetManager, _injector) {
		super();
		this._docPrintInterceptorService = _docPrintInterceptorService;
		this._drawingRenderService = _drawingRenderService;
		this._drawingManagerService = _drawingManagerService;
		this._componetManager = _componetManager;
		this._injector = _injector;
		this._initPrinting();
		this._initPrintingDom();
	}
	_initPrinting() {
		this.disposeWithMe(this._docPrintInterceptorService.interceptor.intercept(this._docPrintInterceptorService.interceptor.getInterceptPoints().PRINTING_COMPONENT_COLLECT, { handler: (_param, pos, next) => {
			const { unitId, scene } = pos;
			const unitData = this._drawingManagerService.getDrawingDataForUnit(unitId);
			const subUnitData = unitData === null || unitData === void 0 ? void 0 : unitData[unitId];
			if (subUnitData) subUnitData.order.forEach((id) => {
				const drawing = subUnitData.data[id];
				if (drawing.drawingType !== _univerjs_core.DrawingTypeEnum.DRAWING_CHART && drawing.drawingType !== _univerjs_core.DrawingTypeEnum.DRAWING_DOM) this._drawingRenderService.renderDrawing(drawing, scene);
			});
			return next();
		} }));
	}
	_initPrintingDom() {
		this.disposeWithMe(this._docPrintInterceptorService.interceptor.intercept(this._docPrintInterceptorService.interceptor.getInterceptPoints().PRINTING_DOM_COLLECT, { handler: (disposableCollection, pos, next) => {
			const { unitId } = pos;
			const unitData = this._drawingManagerService.getDrawingDataForUnit(unitId);
			const subUnitData = unitData === null || unitData === void 0 ? void 0 : unitData[unitId];
			if (subUnitData) {
				const floatDomInfos = subUnitData.order.map((id) => {
					const drawing = subUnitData.data[id];
					if (drawing.drawingType === _univerjs_core.DrawingTypeEnum.DRAWING_CHART) return {
						...drawing,
						componentKey: this._componetManager.get(_univerjs_core.DOC_DRAWING_PRINTING_COMPONENT_KEY)
					};
					if (drawing.drawingType === _univerjs_core.DrawingTypeEnum.DRAWING_DOM) {
						const printingComponentKey = this._docPrintInterceptorService.getPrintComponent(drawing.componentKey);
						return {
							...drawing,
							componentKey: this._componetManager.get(printingComponentKey || drawing.componentKey)
						};
					}
					return null;
				}).filter(Boolean);
				(0, _univerjs_design.render)(/* @__PURE__ */ (0, react_jsx_runtime.jsx)((0, _univerjs_ui.connectInjector)(DocPrintingFloatDom, this._injector), {
					unitId,
					floatDomInfos,
					scene: pos.scene,
					skeleton: pos.skeleton,
					offset: pos.offset,
					bound: pos.bound
				}), pos.root);
				disposableCollection === null || disposableCollection === void 0 || disposableCollection.add(() => {
					(0, _univerjs_design.unmount)(pos.root);
				});
				return next(disposableCollection);
			}
		} }));
	}
};
DocDrawingPrintingController = __decorate([
	__decorateParam(0, (0, _univerjs_core.Inject)(_univerjs_docs_ui.DocPrintInterceptorService)),
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_drawing_ui.DrawingRenderService)),
	__decorateParam(2, _univerjs_drawing.IDrawingManagerService),
	__decorateParam(3, (0, _univerjs_core.Inject)(_univerjs_ui.ComponentManager)),
	__decorateParam(4, (0, _univerjs_core.Inject)(_univerjs_core.Injector))
], DocDrawingPrintingController);

//#endregion
//#region src/controllers/render-controllers/doc-drawing-transform-update.controller.ts
function getDocsTableCellDrawingOffset(unitId, table, row, cell) {
	const sourceTableId = (0, _univerjs_engine_render.getTableIdAndSliceIndex)(table.tableId).tableId;
	const viewport = (0, _univerjs_engine_render.getDocsTableRenderViewport)(unitId, sourceTableId);
	const scrollLeft = viewport && viewport.contentWidth > viewport.viewportWidth ? viewport.scrollLeft : 0;
	return {
		left: table.left + cell.left - scrollLeft + cell.marginLeft,
		top: table.top + row.top + cell.marginTop
	};
}
let DocDrawingTransformUpdateController = class DocDrawingTransformUpdateController extends _univerjs_core.Disposable {
	constructor(_context, _docSkeletonManagerService, _commandService, _editorService, _drawingManagerService, _docRefreshDrawingsService, _univerInstanceService, _lifecycleService) {
		super();
		this._context = _context;
		this._docSkeletonManagerService = _docSkeletonManagerService;
		this._commandService = _commandService;
		this._editorService = _editorService;
		this._drawingManagerService = _drawingManagerService;
		this._docRefreshDrawingsService = _docRefreshDrawingsService;
		this._univerInstanceService = _univerInstanceService;
		this._lifecycleService = _lifecycleService;
		_defineProperty(this, "_liquid", new _univerjs_engine_render.Liquid());
		this._initialize();
		this._commandExecutedListener();
	}
	_initialize() {
		this._initialRenderRefresh();
		this._drawingInitializeListener();
		this._initResize();
	}
	_initialRenderRefresh() {
		this.disposeWithMe(this._docSkeletonManagerService.currentSkeleton$.subscribe((documentSkeleton) => {
			if (documentSkeleton == null) return;
			this._refreshDrawing(documentSkeleton);
		}));
		this.disposeWithMe(this._docRefreshDrawingsService.refreshDrawings$.subscribe((skeleton) => {
			if (skeleton == null) return;
			this._refreshDrawing(skeleton);
		}));
	}
	_commandExecutedListener() {
		const updateCommandList = [_univerjs_docs.RichTextEditingMutation.id, _univerjs_docs_ui.SetDocZoomRatioOperation.id];
		this.disposeWithMe(this._commandService.onCommandExecuted((command) => {
			if (updateCommandList.includes(command.id)) {
				const { unitId: commandUnitId } = command.params;
				const { unitId, mainComponent } = this._context;
				if (commandUnitId !== unitId) return;
				const skeleton = this._docSkeletonManagerService.getSkeleton();
				if (skeleton == null) return;
				if (this._editorService.isEditor(unitId) && unitId !== _univerjs_core.DOCS_ZEN_EDITOR_UNIT_ID_KEY) {
					mainComponent === null || mainComponent === void 0 || mainComponent.makeDirty();
					return;
				}
				this._refreshDrawing(skeleton);
			}
		}));
	}
	_initResize() {
		this.disposeWithMe((0, _univerjs_core.fromEventSubject)(this._context.engine.onTransformChange$).pipe((0, rxjs.filter)((evt) => evt.type === _univerjs_engine_render.TRANSFORM_CHANGE_OBSERVABLE_TYPE.resize), (0, rxjs.debounceTime)(16)).subscribe(() => {
			var _scene$getTransformer;
			const skeleton = this._docSkeletonManagerService.getSkeleton();
			const { scene } = this._context;
			(_scene$getTransformer = scene.getTransformer()) === null || _scene$getTransformer === void 0 || _scene$getTransformer.refreshControls();
			this._refreshDrawing(skeleton);
		}));
	}
	_refreshDrawing(skeleton) {
		const skeletonData = skeleton === null || skeleton === void 0 ? void 0 : skeleton.getSkeletonData();
		const { mainComponent, unitId } = this._context;
		const documentComponent = mainComponent;
		if (!skeletonData) return;
		const { left: docsLeft, top: docsTop, pageLayoutType, pageMarginLeft, pageMarginTop } = documentComponent;
		const { pages, skeHeaders, skeFooters } = skeletonData;
		const updateDrawingMap = {};
		this._liquid.reset();
		/**
		* TODO: @DR-Univer We should not refresh all floating elements, but instead make a diff.
		*/
		for (let i = 0, len = pages.length; i < len; i++) {
			const page = pages[i];
			const { headerId, footerId, pageWidth } = page;
			if (headerId) {
				var _skeHeaders$get;
				const headerPage = (_skeHeaders$get = skeHeaders.get(headerId)) === null || _skeHeaders$get === void 0 ? void 0 : _skeHeaders$get.get(pageWidth);
				if (headerPage) {
					this._calculateDrawingPosition(unitId, headerPage, docsLeft, docsTop, updateDrawingMap, headerPage.marginTop, page.marginLeft);
					this._calculateTableCellDrawingPositions(unitId, headerPage, docsLeft, docsTop, updateDrawingMap, headerPage.marginTop, page.marginLeft);
				}
			}
			if (footerId) {
				var _skeFooters$get;
				const footerPage = (_skeFooters$get = skeFooters.get(footerId)) === null || _skeFooters$get === void 0 ? void 0 : _skeFooters$get.get(pageWidth);
				if (footerPage) {
					const footerTop = page.pageHeight - page.marginBottom + footerPage.marginTop;
					this._calculateDrawingPosition(unitId, footerPage, docsLeft, docsTop, updateDrawingMap, footerTop, page.marginLeft);
					this._calculateTableCellDrawingPositions(unitId, footerPage, docsLeft, docsTop, updateDrawingMap, footerTop, page.marginLeft);
				}
			}
			this._calculateDrawingPosition(unitId, page, docsLeft, docsTop, updateDrawingMap, page.marginTop, page.marginLeft);
			this._calculateTableCellDrawingPositions(unitId, page, docsLeft, docsTop, updateDrawingMap, page.marginTop, page.marginLeft);
			this._liquid.translatePage(page, pageLayoutType, pageMarginLeft, pageMarginTop);
		}
		const updateDrawings = Object.values(updateDrawingMap);
		const nonMultiDrawings = updateDrawings.filter((drawing) => !drawing.isMultiTransform);
		const multiDrawings = updateDrawings.filter((drawing) => drawing.isMultiTransform);
		if (nonMultiDrawings.length > 0) this._drawingManagerService.refreshTransform(nonMultiDrawings);
		this._handleMultiDrawingsTransform(multiDrawings);
	}
	_handleMultiDrawingsTransform(multiDrawings) {
		const { scene, unitId } = this._context;
		const transformer = scene.getTransformerByCreate();
		multiDrawings.forEach((updateParam) => {
			const param = this._drawingManagerService.getDrawingByParam(updateParam);
			if (param == null) return;
			param.transform = updateParam.transform;
			param.transforms = updateParam.transforms;
			param.isMultiTransform = updateParam.isMultiTransform;
		});
		const selectedObjectKeys = [...transformer.getSelectedObjectMap().keys()];
		const allMultiDrawings = Object.values(this._drawingManagerService.getDrawingData(unitId, unitId)).filter((drawing) => drawing.isMultiTransform === _univerjs_core.BooleanNumber.TRUE);
		this._drawingManagerService.removeNotification(allMultiDrawings);
		if (multiDrawings.length > 0) this._drawingManagerService.addNotification(multiDrawings);
		for (const key of selectedObjectKeys) {
			const drawingShape = scene.getObject(key);
			if (drawingShape) transformer.setSelectedControl(drawingShape);
		}
	}
	_calculateDrawingPosition(unitId, page, docsLeft, docsTop, updateDrawingMap, marginTop, marginLeft) {
		const { skeDrawings } = page;
		this._liquid.translatePagePadding({
			marginTop,
			marginLeft
		});
		skeDrawings.forEach((drawing) => {
			const { aLeft, aTop, height, width, angle, drawingId, drawingOrigin } = drawing;
			const behindText = drawingOrigin.layoutType === _univerjs_core.PositionedObjectLayoutType.WRAP_NONE && drawingOrigin.behindDoc === _univerjs_core.BooleanNumber.TRUE;
			const { isMultiTransform = _univerjs_core.BooleanNumber.FALSE } = drawingOrigin;
			const transform = {
				left: aLeft + docsLeft + this._liquid.x,
				top: aTop + docsTop + this._liquid.y,
				width,
				height,
				angle
			};
			if (updateDrawingMap[drawingId] == null) updateDrawingMap[drawingId] = {
				unitId,
				subUnitId: unitId,
				drawingId,
				behindText,
				transform,
				transforms: [transform],
				isMultiTransform
			};
			else if (isMultiTransform === _univerjs_core.BooleanNumber.TRUE) updateDrawingMap[drawingId].transforms.push(transform);
		});
		this._liquid.restorePagePadding({
			marginTop,
			marginLeft
		});
	}
	_calculateTableCellDrawingPositions(unitId, page, docsLeft, docsTop, updateDrawingMap, baseMarginTop, baseMarginLeft) {
		var _page$skeTables;
		(_page$skeTables = page.skeTables) === null || _page$skeTables === void 0 || _page$skeTables.forEach((table) => {
			table.rows.forEach((row) => {
				row.cells.forEach((cell) => {
					const cellOffset = getDocsTableCellDrawingOffset(unitId, table, row, cell);
					const marginTop = baseMarginTop + cellOffset.top;
					const marginLeft = baseMarginLeft + cellOffset.left;
					this._calculateDrawingPosition(unitId, cell, docsLeft, docsTop, updateDrawingMap, marginTop, marginLeft);
					this._calculateTableCellDrawingPositions(unitId, cell, docsLeft, docsTop, updateDrawingMap, marginTop, marginLeft);
				});
			});
		});
	}
	_drawingInitializeListener() {
		const init = () => {
			const skeleton = this._docSkeletonManagerService.getSkeleton();
			if (skeleton == null) return;
			this._refreshDrawing(skeleton);
			this._drawingManagerService.initializeNotification(this._context.unitId);
		};
		if (this._lifecycleService.stage >= _univerjs_core.LifecycleStages.Rendered) if (this._docSkeletonManagerService.getSkeleton()) init();
		else setTimeout(init, 500);
		else this.disposeWithMe(this._lifecycleService.lifecycle$.pipe((0, rxjs.filter)((stage) => stage === _univerjs_core.LifecycleStages.Rendered)).subscribe(init));
	}
};
DocDrawingTransformUpdateController = __decorate([
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_docs.DocSkeletonManagerService)),
	__decorateParam(2, _univerjs_core.ICommandService),
	__decorateParam(3, _univerjs_docs_ui.IEditorService),
	__decorateParam(4, _univerjs_drawing.IDrawingManagerService),
	__decorateParam(5, (0, _univerjs_core.Inject)(DocRefreshDrawingsService)),
	__decorateParam(6, _univerjs_core.IUniverInstanceService),
	__decorateParam(7, (0, _univerjs_core.Inject)(_univerjs_core.LifecycleService))
], DocDrawingTransformUpdateController);

//#endregion
//#region src/controllers/doc-drawing-transformer-update.controller.ts
const INLINE_DRAWING_ANCHOR_KEY_PREFIX = "__InlineDrawingAnchor__";
function getDocsTableCellAnchorContext(unitId, cell) {
	var _row$cells;
	const row = cell.parent;
	const table = row === null || row === void 0 ? void 0 : row.parent;
	const hostPage = table === null || table === void 0 ? void 0 : table.parent;
	if (!row || !table || !hostPage || !((_row$cells = row.cells) === null || _row$cells === void 0 ? void 0 : _row$cells.includes(cell))) return null;
	return {
		cell,
		hostPage,
		offset: getDocsTableCellDrawingOffset(unitId, table, row, cell),
		row,
		table
	};
}
let DocDrawingTransformerController = class DocDrawingTransformerController extends _univerjs_core.Disposable {
	constructor(_commandService, _univerInstanceService, _drawingManagerService, _renderManagerService) {
		super();
		this._commandService = _commandService;
		this._univerInstanceService = _univerInstanceService;
		this._drawingManagerService = _drawingManagerService;
		this._renderManagerService = _renderManagerService;
		_defineProperty(this, "_liquid", new _univerjs_engine_render.Liquid());
		_defineProperty(this, "_listenerOnImageMap", /* @__PURE__ */ new Set());
		_defineProperty(this, "_transformerCache", /* @__PURE__ */ new Map());
		_defineProperty(this, "_anchorShape", void 0);
		this._init();
	}
	_init() {
		this._listenDrawingFocus();
	}
	_listenDrawingFocus() {
		this.disposeWithMe(this._drawingManagerService.add$.subscribe((drawingParams) => {
			if (drawingParams.length === 0) return;
			for (const drawingParam of drawingParams) {
				const { unitId } = drawingParam;
				if (!this._listenerOnImageMap.has(unitId)) {
					this._listenTransformerChange(unitId);
					this._listenerOnImageMap.add(unitId);
				}
			}
		}));
	}
	_listenTransformerChange(unitId) {
		var _this$_getSceneAndTra;
		const transformer = (_this$_getSceneAndTra = this._getSceneAndTransformerByDrawingSearch(unitId)) === null || _this$_getSceneAndTra === void 0 ? void 0 : _this$_getSceneAndTra.transformer;
		if (transformer == null) return;
		this.disposeWithMe((0, _univerjs_core.toDisposable)(transformer.changeStart$.subscribe((state) => {
			this._transformerCache.clear();
			const { objects } = state;
			for (const object of objects.values()) {
				var _documentDataModel$ge;
				const { oKey, width, height, left, top, angle } = object;
				const drawing = this._drawingManagerService.getDrawingOKey(oKey);
				if (drawing == null) continue;
				const documentDataModel = this._univerInstanceService.getUniverDocInstance(drawing.unitId);
				const drawingData = documentDataModel === null || documentDataModel === void 0 || (_documentDataModel$ge = documentDataModel.getSnapshot().drawings) === null || _documentDataModel$ge === void 0 ? void 0 : _documentDataModel$ge[drawing.drawingId];
				if ((drawingData === null || drawingData === void 0 ? void 0 : drawingData.layoutType) === _univerjs_core.PositionedObjectLayoutType.INLINE) try {
					object.setOpacity(.2);
				} catch (e) {}
				if (drawingData != null) this._transformerCache.set(drawing.drawingId, {
					drawing: drawingData,
					top,
					left,
					width,
					height,
					angle
				});
			}
		})));
		const throttleMultipleDrawingUpdate = (0, _univerjs_core.throttle)(this._updateMultipleDrawingDocTransform.bind(this), 50);
		(0, _univerjs_core.throttle)(this._nonInlineDrawingTransform.bind(this), 50);
		this.disposeWithMe((0, _univerjs_core.toDisposable)(transformer.changing$.subscribe((state) => {
			const { objects, offsetX, offsetY } = state;
			if (objects.size > 1) throttleMultipleDrawingUpdate(objects);
			else if (objects.size === 1) {
				const drawingCache = this._transformerCache.values().next().value;
				const { width, height, top, left, angle } = objects.values().next().value;
				if (drawingCache && width === drawingCache.width && height === drawingCache.height && top === drawingCache.top && left === drawingCache.left && angle === drawingCache.angle) return;
				if (drawingCache && drawingCache.drawing.layoutType !== _univerjs_core.PositionedObjectLayoutType.INLINE) {}
				if (drawingCache && drawingCache.drawing.layoutType === _univerjs_core.PositionedObjectLayoutType.INLINE && offsetX != null && offsetY != null) this._updateInlineDrawingAnchor(drawingCache.drawing, offsetX, offsetY);
			}
		})));
		this.disposeWithMe((0, _univerjs_core.toDisposable)(transformer.changeEnd$.subscribe((state) => {
			const { objects, offsetX, offsetY } = state;
			for (const object of objects.values()) {
				const drawing = this._drawingManagerService.getDrawingOKey(object.oKey);
				if (drawing == null) continue;
				const drawingCache = this._transformerCache.get(drawing === null || drawing === void 0 ? void 0 : drawing.drawingId);
				if ((drawingCache === null || drawingCache === void 0 ? void 0 : drawingCache.drawing.layoutType) === _univerjs_core.PositionedObjectLayoutType.INLINE) try {
					object.setOpacity(1);
				} catch (e) {}
			}
			if (this._anchorShape) this._anchorShape.hide();
			if (objects.size > 1) this._updateMultipleDrawingDocTransform(objects);
			else if (objects.size === 1) {
				const drawingCache = this._transformerCache.values().next().value;
				const object = objects.values().next().value;
				const { width, height, top, left, angle } = object;
				if (drawingCache && width === drawingCache.width && height === drawingCache.height && top === drawingCache.top && left === drawingCache.left && angle === drawingCache.angle) return;
				if (drawingCache && drawingCache.drawing.layoutType === _univerjs_core.PositionedObjectLayoutType.INLINE) {
					if (width !== drawingCache.width || height !== drawingCache.height || angle !== drawingCache.angle) this._updateDrawingSize(drawingCache, object);
					else if (offsetX != null && offsetY != null) this._moveInlineDrawing(drawingCache.drawing, offsetX, offsetY);
				} else if (drawingCache) this._nonInlineDrawingTransform(drawingCache.drawing, object);
			}
			this._transformerCache.clear();
		})));
	}
	_updateMultipleDrawingDocTransform(objects) {
		if (objects.size < 1) return;
		const drawings = [];
		let unitId;
		let subUnitId;
		for (const object of objects.values()) {
			const { oKey, left, top, angle } = object;
			let { width, height } = object;
			const drawing = this._drawingManagerService.getDrawingOKey(oKey);
			if (drawing == null) continue;
			if (unitId == null) unitId = drawing.unitId;
			if (subUnitId == null) subUnitId = drawing.subUnitId;
			const drawingCache = this._transformerCache.get(drawing.drawingId);
			if (drawingCache == null) continue;
			const { drawing: drawingData, top: oldTop, left: oldLeft, width: oldWidth, height: oldHeight, angle: oldAngle } = drawingCache;
			const { width: maxWidth, height: maxHeight } = this._getPageContentSize(drawingData);
			width = Math.min(width, maxWidth);
			height = Math.min(height, maxHeight);
			if (oldWidth !== width || oldHeight !== height) drawings.push({
				drawingId: drawing.drawingId,
				key: "size",
				value: {
					width,
					height
				}
			});
			if (oldAngle !== angle) drawings.push({
				drawingId: drawing.drawingId,
				key: "angle",
				value: angle
			});
			if (oldTop !== top || oldLeft !== left) {
				const verticalDelta = top - oldTop;
				const horizontalDelta = left - oldLeft;
				if (verticalDelta !== 0) drawings.push({
					drawingId: drawing.drawingId,
					key: "positionV",
					value: {
						relativeFrom: drawingData.docTransform.positionV.relativeFrom,
						posOffset: drawingData.docTransform.positionV.posOffset + verticalDelta
					}
				});
				if (horizontalDelta !== 0) drawings.push({
					drawingId: drawing.drawingId,
					key: "positionH",
					value: {
						relativeFrom: drawingData.docTransform.positionH.relativeFrom,
						posOffset: drawingData.docTransform.positionH.posOffset + horizontalDelta
					}
				});
			}
		}
		if (drawings.length > 0 && unitId && subUnitId) this._commandService.executeCommand(UpdateDrawingDocTransformCommand.id, {
			unitId,
			subUnitId,
			drawings
		});
	}
	_updateDrawingAnchor(objects) {
		if (this._transformerCache.size !== 1) return;
		const drawingCache = this._transformerCache.values().next().value;
		const object = objects.values().next().value;
		this._getDrawingAnchor(drawingCache.drawing, object);
	}
	_updateInlineDrawingAnchor(drawing, offsetX, offsetY) {
		var _this$_getInlineDrawi;
		if (this._transformerCache.size !== 1) return;
		const { contentBoxPointGroup } = (_this$_getInlineDrawi = this._getInlineDrawingAnchor(drawing, offsetX, offsetY)) !== null && _this$_getInlineDrawi !== void 0 ? _this$_getInlineDrawi : {};
		if (contentBoxPointGroup == null) return;
		this._createOrUpdateInlineAnchor(drawing.unitId, contentBoxPointGroup);
	}
	_getInlineDrawingAnchor(drawing, offsetX, offsetY) {
		var _this$_renderManagerS, _getOneTextSelectionR;
		const currentRender = this._renderManagerService.getRenderById(drawing.unitId);
		const skeleton = currentRender === null || currentRender === void 0 ? void 0 : currentRender.with(_univerjs_docs.DocSkeletonManagerService).getSkeleton();
		if (currentRender == null) return;
		const { mainComponent, scene } = currentRender;
		const documentComponent = mainComponent;
		const activeViewport = scene.getViewports()[0];
		const { pageLayoutType = _univerjs_engine_render.PageLayoutType.VERTICAL, pageMarginLeft, pageMarginTop } = documentComponent.getOffsetConfig();
		let glyphAnchor = null;
		let isBack = false;
		let segmentPageIndex = -1;
		let segmentId = "";
		const HALF = .5;
		const coord = this._getTransformCoordForDocumentOffset(documentComponent, activeViewport, offsetX, offsetY);
		if (coord == null) return;
		const docSelectionRenderService = (_this$_renderManagerS = this._renderManagerService.getRenderById(drawing.unitId)) === null || _this$_renderManagerS === void 0 ? void 0 : _this$_renderManagerS.with(_univerjs_docs_ui.DocSelectionRenderService);
		if (docSelectionRenderService == null) return;
		const nodeInfo = skeleton === null || skeleton === void 0 ? void 0 : skeleton.findNodeByCoord(coord, pageLayoutType, pageMarginLeft, pageMarginTop, {
			strict: false,
			segmentId: docSelectionRenderService.getSegment(),
			segmentPage: docSelectionRenderService.getSegmentPage()
		});
		if (nodeInfo) {
			const { node, ratioX, segmentPage, segmentId: nodeSegmentId } = nodeInfo;
			isBack = ratioX < HALF;
			glyphAnchor = node;
			segmentPageIndex = segmentPage;
			segmentId = nodeSegmentId;
		}
		if (glyphAnchor == null) return;
		const nodePosition = skeleton === null || skeleton === void 0 ? void 0 : skeleton.findPositionByGlyph(glyphAnchor, segmentPageIndex);
		const docObject = this._getDocObject();
		if (nodePosition == null || skeleton == null || docObject == null) return;
		const positionWithIsBack = {
			...nodePosition,
			isBack
		};
		const { cursorList, contentBoxPointGroup } = new _univerjs_docs_ui.NodePositionConvertToCursor(docObject.document.getOffsetConfig(), skeleton).getRangePointData(positionWithIsBack, positionWithIsBack);
		const { startOffset } = (_getOneTextSelectionR = (0, _univerjs_docs_ui.getOneTextSelectionRange)(cursorList)) !== null && _getOneTextSelectionR !== void 0 ? _getOneTextSelectionR : {};
		if (startOffset == null) return;
		return {
			offset: startOffset,
			contentBoxPointGroup,
			segmentId,
			segmentPage: segmentPageIndex
		};
	}
	_getDrawingAnchor(drawing, object) {
		var _this$_renderManagerS2, _glyphAnchor$parent, _column$lines$find, _column$parent, _tableCellContext$hos, _getOneTextSelectionR2;
		const currentRender = this._renderManagerService.getRenderById(drawing.unitId);
		const skeleton = currentRender === null || currentRender === void 0 ? void 0 : currentRender.with(_univerjs_docs.DocSkeletonManagerService).getSkeleton();
		const skeletonData = skeleton === null || skeleton === void 0 ? void 0 : skeleton.getSkeletonData();
		if (skeletonData == null || currentRender == null) return;
		const { pages, skeHeaders, skeFooters } = skeletonData;
		const { mainComponent, scene } = currentRender;
		const documentComponent = mainComponent;
		const activeViewport = scene.getViewports()[0];
		const { pageLayoutType = _univerjs_engine_render.PageLayoutType.VERTICAL, pageMarginLeft, pageMarginTop, docsLeft, docsTop } = documentComponent.getOffsetConfig();
		const { left, top, angle } = object;
		let { width, height } = object;
		const { positionV, positionH } = drawing.docTransform;
		const { width: maxWidth, height: maxHeight } = this._getPageContentSize(drawing);
		width = Math.min(width, maxWidth);
		height = Math.min(height, maxHeight);
		let glyphAnchor = null;
		let segmentId = "";
		let segmentPage = -1;
		const isBack = false;
		const docTransform = {
			...drawing.docTransform,
			size: {
				width,
				height
			},
			angle
		};
		const { x, y } = scene.getViewportScrollXY(activeViewport);
		const coord = this._getTransformCoordForDocumentOffset(documentComponent, activeViewport, left - x, top - y);
		if (coord == null) return;
		const docSelectionRenderService = (_this$_renderManagerS2 = this._renderManagerService.getRenderById(drawing.unitId)) === null || _this$_renderManagerS2 === void 0 ? void 0 : _this$_renderManagerS2.with(_univerjs_docs_ui.DocSelectionRenderService);
		if (docSelectionRenderService == null) return;
		const nodeInfo = skeleton === null || skeleton === void 0 ? void 0 : skeleton.findNodeByCoord(coord, pageLayoutType, pageMarginLeft, pageMarginTop, {
			strict: false,
			segmentId: docSelectionRenderService.getSegment(),
			segmentPage: docSelectionRenderService.getSegmentPage()
		});
		if (nodeInfo) {
			const { node, segmentPage: segmentPageIndex, segmentId: nodeSegmentId } = nodeInfo;
			glyphAnchor = node;
			segmentPage = segmentPageIndex;
			segmentId = nodeSegmentId;
		}
		if (glyphAnchor == null) return;
		const line = (_glyphAnchor$parent = glyphAnchor.parent) === null || _glyphAnchor$parent === void 0 ? void 0 : _glyphAnchor$parent.parent;
		const column = line === null || line === void 0 ? void 0 : line.parent;
		const paragraphStartLine = (_column$lines$find = column === null || column === void 0 ? void 0 : column.lines.find((l) => l.paragraphIndex === (line === null || line === void 0 ? void 0 : line.paragraphIndex) && l.paragraphStart)) !== null && _column$lines$find !== void 0 ? _column$lines$find : column === null || column === void 0 ? void 0 : column.lines[0];
		const page = column === null || column === void 0 || (_column$parent = column.parent) === null || _column$parent === void 0 ? void 0 : _column$parent.parent;
		if (line == null || column == null || paragraphStartLine == null || page == null) return;
		this._liquid.reset();
		const tableCellContext = page.type === _univerjs_engine_render.DocumentSkeletonPageType.CELL ? getDocsTableCellAnchorContext(drawing.unitId, page) : null;
		const anchorPage = (_tableCellContext$hos = tableCellContext === null || tableCellContext === void 0 ? void 0 : tableCellContext.hostPage) !== null && _tableCellContext$hos !== void 0 ? _tableCellContext$hos : page;
		const pageType = anchorPage.type;
		for (const p of pages) {
			const { headerId, footerId, pageHeight, pageWidth, marginLeft, marginBottom } = p;
			const pIndex = pages.indexOf(p);
			if (segmentPage > -1 && pIndex === segmentPage) {
				switch (pageType) {
					case _univerjs_engine_render.DocumentSkeletonPageType.HEADER: {
						var _skeHeaders$get;
						const headerSke = (_skeHeaders$get = skeHeaders.get(headerId)) === null || _skeHeaders$get === void 0 ? void 0 : _skeHeaders$get.get(pageWidth);
						if (headerSke) this._liquid.translatePagePadding({
							marginTop: headerSke.marginTop,
							marginLeft
						});
						else throw new Error("header skeleton not found");
						break;
					}
					case _univerjs_engine_render.DocumentSkeletonPageType.FOOTER: {
						var _skeFooters$get;
						const footerSke = (_skeFooters$get = skeFooters.get(footerId)) === null || _skeFooters$get === void 0 ? void 0 : _skeFooters$get.get(pageWidth);
						if (footerSke) this._liquid.translatePagePadding({
							marginTop: pageHeight - marginBottom + footerSke.marginTop,
							marginLeft
						});
						else throw new Error("footer skeleton not found");
						break;
					}
					default:
						this._liquid.translatePagePadding(p);
						break;
				}
				break;
			}
			this._liquid.translatePagePadding(p);
			if (p === anchorPage) break;
			this._liquid.restorePagePadding(p);
			this._liquid.translatePage(p, pageLayoutType, pageMarginLeft, pageMarginTop);
		}
		if (tableCellContext) this._liquid.translate(tableCellContext.offset.left, tableCellContext.offset.top);
		if (positionV.relativeFrom === _univerjs_core.ObjectRelativeFromV.LINE) glyphAnchor = line.divides[0].glyphGroup[0];
		else {
			var _paragraphStartLine$d, _paragraphStartLine$d2;
			glyphAnchor = (_paragraphStartLine$d = (_paragraphStartLine$d2 = paragraphStartLine.divides) === null || _paragraphStartLine$d2 === void 0 || (_paragraphStartLine$d2 = _paragraphStartLine$d2[0]) === null || _paragraphStartLine$d2 === void 0 || (_paragraphStartLine$d2 = _paragraphStartLine$d2.glyphGroup) === null || _paragraphStartLine$d2 === void 0 ? void 0 : _paragraphStartLine$d2[0]) !== null && _paragraphStartLine$d !== void 0 ? _paragraphStartLine$d : glyphAnchor;
		}
		docTransform.positionH = {
			relativeFrom: positionH.relativeFrom,
			posOffset: left - this._liquid.x - docsLeft
		};
		switch (positionH.relativeFrom) {
			case _univerjs_core.ObjectRelativeFromH.MARGIN:
				docTransform.positionH.posOffset = left - this._liquid.x - docsLeft - page.marginLeft;
				break;
			case _univerjs_core.ObjectRelativeFromH.COLUMN:
				docTransform.positionH.posOffset = left - this._liquid.x - docsLeft - column.left;
				break;
		}
		docTransform.positionV = {
			relativeFrom: positionV.relativeFrom,
			posOffset: top - this._liquid.y - docsTop
		};
		switch (positionV.relativeFrom) {
			case _univerjs_core.ObjectRelativeFromV.PAGE:
				docTransform.positionV.posOffset = top - this._liquid.y - docsTop - page.marginTop;
				break;
			case _univerjs_core.ObjectRelativeFromV.LINE:
				docTransform.positionV.posOffset = top - this._liquid.y - docsTop - line.top;
				break;
			case _univerjs_core.ObjectRelativeFromV.PARAGRAPH:
				docTransform.positionV.posOffset = top - this._liquid.y - docsTop - paragraphStartLine.top;
				break;
		}
		if (glyphAnchor == null) return;
		const nodePosition = skeleton === null || skeleton === void 0 ? void 0 : skeleton.findPositionByGlyph(glyphAnchor, segmentPage);
		const docObject = this._getDocObject();
		if (nodePosition == null || skeleton == null || docObject == null) return;
		const positionWithIsBack = {
			...nodePosition,
			isBack
		};
		const { cursorList } = new _univerjs_docs_ui.NodePositionConvertToCursor(docObject.document.getOffsetConfig(), skeleton).getRangePointData(positionWithIsBack, positionWithIsBack);
		const { startOffset } = (_getOneTextSelectionR2 = (0, _univerjs_docs_ui.getOneTextSelectionRange)(cursorList)) !== null && _getOneTextSelectionR2 !== void 0 ? _getOneTextSelectionR2 : {};
		if (startOffset == null) return;
		return {
			offset: startOffset,
			docTransform,
			segmentId,
			segmentPage
		};
	}
	_updateDrawingSize(drawingCache, object) {
		const drawings = [];
		const { drawing, width: oldWidth, height: oldHeight, angle: oldAngle } = drawingCache;
		const { unitId, subUnitId } = drawing;
		let { width, height, angle } = object;
		const { width: maxWidth, height: maxHeight } = this._getPageContentSize(drawing);
		width = Math.min(maxWidth, width);
		height = Math.min(maxHeight, height);
		if (width !== oldWidth || height !== oldHeight) drawings.push({
			drawingId: drawing.drawingId,
			key: "size",
			value: {
				width,
				height
			}
		});
		if (angle !== oldAngle) drawings.push({
			drawingId: drawing.drawingId,
			key: "angle",
			value: angle
		});
		if (drawings.length > 0 && unitId && subUnitId) this._commandService.executeCommand(UpdateDrawingDocTransformCommand.id, {
			unitId,
			subUnitId,
			drawings
		});
	}
	_moveInlineDrawing(drawing, offsetX, offsetY) {
		const anchor = this._getInlineDrawingAnchor(drawing, offsetX, offsetY);
		const { offset, segmentId, segmentPage } = anchor !== null && anchor !== void 0 ? anchor : {};
		return this._commandService.executeCommand(IMoveInlineDrawingCommand.id, {
			unitId: drawing.unitId,
			subUnitId: drawing.unitId,
			drawing,
			offset,
			segmentId,
			segmentPage,
			needRefreshDrawings: offset == null
		});
	}
	_limitDrawingInPage(drawing, object) {
		const currentRender = this._renderManagerService.getRenderById(drawing.unitId);
		const { left, top, width, height, angle } = object;
		const skeleton = currentRender === null || currentRender === void 0 ? void 0 : currentRender.with(_univerjs_docs.DocSkeletonManagerService).getSkeleton();
		const skeletonData = skeleton === null || skeleton === void 0 ? void 0 : skeleton.getSkeletonData();
		const { pages } = skeletonData !== null && skeletonData !== void 0 ? skeletonData : {};
		if (skeletonData == null || currentRender == null || pages == null) return {
			left,
			top,
			width,
			height,
			angle
		};
		const { mainComponent } = currentRender;
		const { top: docsTop, pageLayoutType, pageMarginLeft, pageMarginTop } = mainComponent;
		let newTop = top;
		this._liquid.reset();
		for (const page of pages) {
			const { marginBottom, pageHeight } = page;
			const nextPage = pages[pages.indexOf(page) + 1];
			if (nextPage == null) continue;
			if (_univerjs_core.Tools.hasIntersectionBetweenTwoRanges(top, top + height, this._liquid.y + docsTop + pageHeight - marginBottom, this._liquid.y + docsTop + pageHeight + pageMarginTop + nextPage.marginTop)) if (top + height / 2 < this._liquid.y + docsTop + pageHeight + pageMarginTop / 2) newTop = Math.min(top, this._liquid.y + docsTop + pageHeight - marginBottom - height);
			else newTop = Math.max(top, this._liquid.y + docsTop + pageHeight + pageMarginTop + nextPage.marginTop);
			this._liquid.translatePage(page, pageLayoutType, pageMarginLeft, pageMarginTop);
		}
		return {
			left,
			top: newTop,
			width,
			height,
			angle
		};
	}
	_nonInlineDrawingTransform(drawing, object, isMoving = false) {
		const objectPosition = drawing.isMultiTransform === _univerjs_core.BooleanNumber.TRUE ? object : this._limitDrawingInPage(drawing, object);
		if (isMoving && objectPosition.top !== object.top) return;
		const anchor = this._getDrawingAnchor(drawing, objectPosition);
		const { offset, docTransform, segmentId, segmentPage } = anchor !== null && anchor !== void 0 ? anchor : {};
		if (offset == null || docTransform == null) return this._updateMultipleDrawingDocTransform(new Map([[drawing.drawingId, object]]));
		return this._commandService.executeCommand(ITransformNonInlineDrawingCommand.id, {
			unitId: drawing.unitId,
			subUnitId: drawing.unitId,
			drawing,
			offset,
			docTransform,
			segmentId,
			segmentPage
		});
	}
	_getSceneAndTransformerByDrawingSearch(unitId) {
		if (unitId == null) return;
		const renderObject = this._renderManagerService.getRenderById(unitId);
		const scene = renderObject === null || renderObject === void 0 ? void 0 : renderObject.scene;
		if (scene == null) return;
		return {
			scene,
			transformer: scene.getTransformerByCreate()
		};
	}
	_getTransformCoordForDocumentOffset(document, viewport, evtOffsetX, evtOffsetY) {
		const { documentTransform } = document.getOffsetConfig();
		const originCoord = viewport.transformVector2SceneCoord(_univerjs_engine_render.Vector2.FromArray([evtOffsetX, evtOffsetY]));
		if (!originCoord) return;
		return documentTransform.clone().invert().applyPoint(originCoord);
	}
	_createOrUpdateInlineAnchor(unitId, pointsGroup) {
		const currentRender = this._renderManagerService.getRenderById(unitId);
		if (currentRender == null) return;
		const { mainComponent, scene } = currentRender;
		const { docsLeft, docsTop } = mainComponent.getOffsetConfig();
		const { left: boundingLeft, top: boundingTop, height } = (0, _univerjs_docs_ui.getAnchorBounding)(pointsGroup);
		const left = boundingLeft + docsLeft;
		const top = boundingTop + docsTop;
		if (this._anchorShape) {
			this._anchorShape.transformByState({
				left,
				top,
				height
			});
			this._anchorShape.show();
			return;
		}
		const anchor = new _univerjs_engine_render.Rect(INLINE_DRAWING_ANCHOR_KEY_PREFIX + (0, _univerjs_core.generateRandomId)(6), {
			left,
			top,
			height,
			strokeWidth: 2,
			stroke: (0, _univerjs_engine_render.getColor)(_univerjs_core.COLORS.darkgray, 1),
			evented: false
		});
		this._anchorShape = anchor;
		scene.addObject(anchor, _univerjs_docs_ui.TEXT_RANGE_LAYER_INDEX);
	}
	_getDocObject() {
		return (0, _univerjs_docs_ui.getDocObject)(this._univerInstanceService, this._renderManagerService);
	}
	_getPageContentSize(drawing) {
		const currentRender = this._renderManagerService.getRenderById(drawing.unitId);
		const skeleton = currentRender === null || currentRender === void 0 ? void 0 : currentRender.with(_univerjs_docs.DocSkeletonManagerService).getSkeleton();
		const MAX_WIDTH = 500;
		const MAX_HEIGHT = 500;
		const skeletonData = skeleton === null || skeleton === void 0 ? void 0 : skeleton.getSkeletonData();
		if (skeletonData == null || currentRender == null) return {
			width: MAX_WIDTH,
			height: MAX_HEIGHT
		};
		const { pages } = skeletonData;
		let page = null;
		for (const p of pages) {
			const { skeDrawings } = p;
			if (skeDrawings.has(drawing.drawingId)) {
				page = p;
				break;
			}
		}
		if (page) {
			const { pageWidth, pageHeight, marginLeft, marginBottom, marginRight, marginTop } = page;
			return {
				width: Math.max(MAX_WIDTH, pageWidth - marginLeft - marginRight),
				height: Math.max(MAX_HEIGHT, pageHeight - marginTop - marginBottom)
			};
		} else return {
			width: MAX_WIDTH,
			height: MAX_HEIGHT
		};
	}
};
DocDrawingTransformerController = __decorate([
	__decorateParam(0, _univerjs_core.ICommandService),
	__decorateParam(1, _univerjs_core.IUniverInstanceService),
	__decorateParam(2, _univerjs_drawing.IDrawingManagerService),
	__decorateParam(3, _univerjs_engine_render.IRenderManagerService)
], DocDrawingTransformerController);

//#endregion
//#region src/menu/schema.ts
const menuSchema = {
	[_univerjs_ui.RibbonInsertGroup.MEDIA]: { [DOCS_IMAGE_MENU_ID]: {
		order: 0,
		menuItemFactory: ImageMenuFactory,
		[IMAGE_MENU_UPLOAD_FLOAT_ID]: {
			order: 0,
			menuItemFactory: UploadFloatImageMenuFactory
		}
	} },
	[_univerjs_ui.ContextMenuPosition.PARAGRAPH]: {
		[_univerjs_ui.ContextMenuGroup.LAYOUT]: { [_univerjs_docs_ui.INSERT_BELLOW_MENU_ID]: { [IMAGE_MENU_UPLOAD_FLOAT_ID]: {
			order: 5,
			menuItemFactory: UploadFloatImageMenuFactory
		} } },
		[_univerjs_docs_ui.EMPTY_PARAGRAPH_MENU_ID]: { [_univerjs_ui.ContextMenuGroup.LAYOUT]: { [IMAGE_MENU_UPLOAD_FLOAT_ID]: {
			order: 5,
			menuItemFactory: UploadFloatImageMenuFactory
		} } },
		[_univerjs_docs_ui.DOC_CONTENT_INSERT_MENU_ID]: { [_univerjs_ui.ContextMenuGroup.LAYOUT]: { [IMAGE_MENU_UPLOAD_FLOAT_ID]: {
			order: 5,
			menuItemFactory: UploadFloatImageMenuFactory
		} } }
	}
};

//#endregion
//#region src/views/doc-image-panel/DocDrawingPosition.tsx
const MIN_OFFSET = -1e3;
const MAX_OFFSET = 1e3;
const DocDrawingPosition = (props) => {
	const commandService = (0, _univerjs_ui.useDependency)(_univerjs_core.ICommandService);
	const localeService = (0, _univerjs_ui.useDependency)(_univerjs_core.LocaleService);
	const drawingManagerService = (0, _univerjs_ui.useDependency)(_univerjs_drawing.IDrawingManagerService);
	const renderManagerService = (0, _univerjs_ui.useDependency)(_univerjs_engine_render.IRenderManagerService);
	const univerInstanceService = (0, _univerjs_ui.useDependency)(_univerjs_core.IUniverInstanceService);
	const { drawings } = props;
	const drawingParam = drawings[0];
	if (drawingParam == null) return;
	const { unitId } = drawingParam;
	const documentDataModel = univerInstanceService.getUniverDocInstance(unitId);
	const documentFlavor = documentDataModel === null || documentDataModel === void 0 ? void 0 : documentDataModel.getSnapshot().documentStyle.documentFlavor;
	const renderObject = renderManagerService.getRenderById(unitId);
	const scene = renderObject === null || renderObject === void 0 ? void 0 : renderObject.scene;
	if (scene == null) return;
	const transformer = scene.getTransformerByCreate();
	const HORIZONTAL_RELATIVE_FROM = [
		{
			label: localeService.t("docs-drawing-ui.image-position.column"),
			value: String(_univerjs_core.ObjectRelativeFromH.COLUMN)
		},
		{
			label: localeService.t("docs-drawing-ui.image-position.page"),
			value: String(_univerjs_core.ObjectRelativeFromH.PAGE)
		},
		{
			label: localeService.t("docs-drawing-ui.image-position.margin"),
			value: String(_univerjs_core.ObjectRelativeFromH.MARGIN)
		}
	];
	const VERTICAL_RELATIVE_FROM = [
		{
			label: localeService.t("docs-drawing-ui.image-position.line"),
			value: String(_univerjs_core.ObjectRelativeFromV.LINE),
			disabled: documentFlavor === _univerjs_core.DocumentFlavor.MODERN
		},
		{
			label: localeService.t("docs-drawing-ui.image-position.page"),
			value: String(_univerjs_core.ObjectRelativeFromV.PAGE),
			disabled: documentFlavor === _univerjs_core.DocumentFlavor.MODERN
		},
		{
			label: localeService.t("docs-drawing-ui.image-position.margin"),
			value: String(_univerjs_core.ObjectRelativeFromV.MARGIN),
			disabled: documentFlavor === _univerjs_core.DocumentFlavor.MODERN
		},
		{
			label: localeService.t("docs-drawing-ui.image-position.paragraph"),
			value: String(_univerjs_core.ObjectRelativeFromV.PARAGRAPH)
		}
	];
	const [disabled, setDisabled] = (0, react.useState)(true);
	const [hPosition, setHPosition] = (0, react.useState)({
		relativeFrom: _univerjs_core.ObjectRelativeFromH.PAGE,
		posOffset: 0
	});
	const [vPosition, setVPosition] = (0, react.useState)({
		relativeFrom: _univerjs_core.ObjectRelativeFromV.PAGE,
		posOffset: 0
	});
	const [followTextMove, setFollowTextMove] = (0, react.useState)(true);
	const [showPanel, setShowPanel] = (0, react.useState)(true);
	function handlePositionChange(direction, value) {
		var _renderManagerService;
		if (direction === "positionH") setHPosition(value);
		else setVPosition(value);
		const focusDrawings = drawingManagerService.getFocusDrawings();
		if (focusDrawings.length === 0) return;
		const drawings = focusDrawings.map((drawing) => {
			return {
				unitId: drawing.unitId,
				subUnitId: drawing.subUnitId,
				drawingId: drawing.drawingId
			};
		});
		commandService.executeCommand(UpdateDrawingDocTransformCommand.id, {
			unitId: focusDrawings[0].unitId,
			subUnitId: focusDrawings[0].unitId,
			drawings: drawings.map((drawing) => ({
				drawingId: drawing.drawingId,
				key: direction,
				value
			}))
		});
		const docSelectionRenderService = (_renderManagerService = renderManagerService.getRenderById(unitId)) === null || _renderManagerService === void 0 ? void 0 : _renderManagerService.with(_univerjs_docs_ui.DocSelectionRenderService);
		if (docSelectionRenderService) docSelectionRenderService.blur();
		transformer.refreshControls();
	}
	function handleHorizontalRelativeFromChange(value) {
		var _renderManagerService2;
		const prevRelativeFrom = hPosition.relativeFrom;
		const prevPosOffset = hPosition.posOffset;
		const relativeFrom = Number(value);
		if (prevRelativeFrom === relativeFrom) return;
		const focusDrawings = drawingManagerService.getFocusDrawings();
		if (focusDrawings.length === 0) return;
		const drawingId = focusDrawings[0].drawingId;
		const unitId = focusDrawings[0].unitId;
		let drawing = null;
		let pageMarginLeft = 0;
		const skeleton = (_renderManagerService2 = renderManagerService.getRenderById(unitId)) === null || _renderManagerService2 === void 0 ? void 0 : _renderManagerService2.with(_univerjs_docs.DocSkeletonManagerService).getSkeleton();
		const skeletonData = skeleton === null || skeleton === void 0 ? void 0 : skeleton.getSkeletonData();
		if (skeletonData == null) return;
		const { pages, skeHeaders, skeFooters } = skeletonData;
		for (const page of pages) {
			var _skeHeaders$get, _skeFooters$get;
			const { marginLeft, skeDrawings, headerId, footerId, pageWidth } = page;
			if (skeDrawings.has(drawingId)) {
				drawing = skeDrawings.get(drawingId);
				pageMarginLeft = marginLeft;
				break;
			}
			const headerPage = (_skeHeaders$get = skeHeaders.get(headerId)) === null || _skeHeaders$get === void 0 ? void 0 : _skeHeaders$get.get(pageWidth);
			if (headerPage === null || headerPage === void 0 ? void 0 : headerPage.skeDrawings.has(drawingId)) {
				drawing = headerPage === null || headerPage === void 0 ? void 0 : headerPage.skeDrawings.get(drawingId);
				pageMarginLeft = marginLeft;
				break;
			}
			const footerPage = (_skeFooters$get = skeFooters.get(footerId)) === null || _skeFooters$get === void 0 ? void 0 : _skeFooters$get.get(pageWidth);
			if (footerPage === null || footerPage === void 0 ? void 0 : footerPage.skeDrawings.has(drawingId)) {
				drawing = footerPage === null || footerPage === void 0 ? void 0 : footerPage.skeDrawings.get(drawingId);
				pageMarginLeft = marginLeft;
				break;
			}
		}
		if (drawing == null) return;
		let delta = 0;
		if (prevRelativeFrom === _univerjs_core.ObjectRelativeFromH.COLUMN) delta -= drawing.columnLeft;
		else if (prevRelativeFrom === _univerjs_core.ObjectRelativeFromH.MARGIN) delta -= pageMarginLeft;
		if (relativeFrom === _univerjs_core.ObjectRelativeFromH.COLUMN) delta += drawing.columnLeft;
		else if (relativeFrom === _univerjs_core.ObjectRelativeFromH.MARGIN) delta += pageMarginLeft;
		else if (relativeFrom === _univerjs_core.ObjectRelativeFromH.PAGE) {}
		handlePositionChange("positionH", {
			relativeFrom,
			posOffset: (prevPosOffset !== null && prevPosOffset !== void 0 ? prevPosOffset : 0) - delta
		});
	}
	function handleVerticalRelativeFromChange(value) {
		var _renderManagerService3, _renderManagerService4, _documentDataModel$ge, _glyph$parent, _column$parent;
		const prevRelativeFrom = vPosition.relativeFrom;
		const prevPosOffset = vPosition.posOffset;
		const relativeFrom = Number(value);
		if (prevRelativeFrom === relativeFrom) return;
		const focusDrawings = drawingManagerService.getFocusDrawings();
		if (focusDrawings.length === 0) return;
		const { drawingId, unitId } = focusDrawings[0];
		const documentDataModel = univerInstanceService.getUniverDocInstance(unitId);
		const skeleton = (_renderManagerService3 = renderManagerService.getRenderById(unitId)) === null || _renderManagerService3 === void 0 ? void 0 : _renderManagerService3.with(_univerjs_docs.DocSkeletonManagerService).getSkeleton();
		const docSelectionRenderService = (_renderManagerService4 = renderManagerService.getRenderById(unitId)) === null || _renderManagerService4 === void 0 ? void 0 : _renderManagerService4.with(_univerjs_docs_ui.DocSelectionRenderService);
		const segmentId = docSelectionRenderService === null || docSelectionRenderService === void 0 ? void 0 : docSelectionRenderService.getSegment();
		const segmentPage = docSelectionRenderService === null || docSelectionRenderService === void 0 ? void 0 : docSelectionRenderService.getSegmentPage();
		const drawing = documentDataModel === null || documentDataModel === void 0 || (_documentDataModel$ge = documentDataModel.getSelfOrHeaderFooterModel(segmentId).getBody()) === null || _documentDataModel$ge === void 0 || (_documentDataModel$ge = _documentDataModel$ge.customBlocks) === null || _documentDataModel$ge === void 0 ? void 0 : _documentDataModel$ge.find((c) => c.blockId === drawingId);
		if (drawing == null || skeleton == null || docSelectionRenderService == null) return;
		const { startIndex } = drawing;
		const glyph = skeleton.findNodeByCharIndex(startIndex, segmentId, segmentPage);
		const line = glyph === null || glyph === void 0 || (_glyph$parent = glyph.parent) === null || _glyph$parent === void 0 ? void 0 : _glyph$parent.parent;
		const column = line === null || line === void 0 ? void 0 : line.parent;
		const paragraphStartLine = column === null || column === void 0 ? void 0 : column.lines.find((l) => l.paragraphIndex === (line === null || line === void 0 ? void 0 : line.paragraphIndex) && l.paragraphStart);
		const page = column === null || column === void 0 || (_column$parent = column.parent) === null || _column$parent === void 0 ? void 0 : _column$parent.parent;
		if (glyph == null || line == null || paragraphStartLine == null || column == null || page == null) return;
		let delta = 0;
		if (prevRelativeFrom === _univerjs_core.ObjectRelativeFromV.PARAGRAPH) delta -= paragraphStartLine.top;
		else if (prevRelativeFrom === _univerjs_core.ObjectRelativeFromV.LINE) delta -= line.top;
		else if (prevRelativeFrom === _univerjs_core.ObjectRelativeFromV.PAGE) delta += page.marginTop;
		if (relativeFrom === _univerjs_core.ObjectRelativeFromV.PARAGRAPH) delta += paragraphStartLine.top;
		else if (relativeFrom === _univerjs_core.ObjectRelativeFromV.LINE) delta += line.top;
		else if (relativeFrom === _univerjs_core.ObjectRelativeFromV.PAGE) delta -= page.marginTop;
		handlePositionChange("positionV", {
			relativeFrom,
			posOffset: (prevPosOffset !== null && prevPosOffset !== void 0 ? prevPosOffset : 0) - delta
		});
	}
	function updateState(drawingParam) {
		var _snapshot$drawings;
		const snapshot = documentDataModel === null || documentDataModel === void 0 ? void 0 : documentDataModel.getSnapshot();
		const drawing = snapshot === null || snapshot === void 0 || (_snapshot$drawings = snapshot.drawings) === null || _snapshot$drawings === void 0 ? void 0 : _snapshot$drawings[drawingParam.drawingId];
		if (drawing == null) return;
		const { layoutType } = drawing;
		const { positionH, positionV } = drawing.docTransform;
		setHPosition(positionH);
		setVPosition(positionV);
		setDisabled(layoutType === _univerjs_core.PositionedObjectLayoutType.INLINE);
		setFollowTextMove(positionV.relativeFrom === _univerjs_core.ObjectRelativeFromV.PARAGRAPH || positionV.relativeFrom === _univerjs_core.ObjectRelativeFromV.LINE);
	}
	function updateFocusDrawingState() {
		const focusDrawings = drawingManagerService.getFocusDrawings();
		if (focusDrawings.length === 0) return;
		updateState(focusDrawings[0]);
	}
	function handleFollowTextMoveCheck(val) {
		setFollowTextMove(val);
		handleVerticalRelativeFromChange(val ? String(_univerjs_core.ObjectRelativeFromV.PARAGRAPH) : String(_univerjs_core.ObjectRelativeFromV.PAGE));
	}
	(0, react.useEffect)(() => {
		updateFocusDrawingState();
		const subscription = drawingManagerService.focus$.subscribe((drawingParams) => {
			if (drawingParams.length === 0) {
				setShowPanel(false);
				return;
			}
			setShowPanel(true);
			updateState(drawingParams[0]);
		});
		const mutationListener = commandService.onCommandExecuted(async (command) => {
			if (command.id === _univerjs_docs.RichTextEditingMutation.id) updateFocusDrawingState();
		});
		return () => {
			subscription.unsubscribe();
			mutationListener.dispose();
		};
	}, []);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, _univerjs_design.clsx)("univer-grid univer-gap-2 univer-py-2 univer-text-gray-400", { "univer-hidden": !showPanel }),
		children: [
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("header", {
				className: "univer-text-gray-600 dark:!univer-text-gray-200",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { children: localeService.t("docs-drawing-ui.image-position.title") })
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "univer-text-gray-600 dark:!univer-text-gray-200",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { children: localeService.t("docs-drawing-ui.image-position.horizontal") })
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: "univer-grid univer-grid-cols-2 univer-gap-2 [&>div]:univer-grid [&>div]:univer-gap-2",
				children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { children: localeService.t("docs-drawing-ui.image-position.absolutePosition") }), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.InputNumber, {
					min: MIN_OFFSET,
					max: MAX_OFFSET,
					precision: 1,
					disabled,
					value: hPosition.posOffset,
					onChange: (val) => {
						handlePositionChange("positionH", {
							relativeFrom: hPosition.relativeFrom,
							posOffset: val
						});
					}
				})] }), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { children: localeService.t("docs-drawing-ui.image-position.toTheRightOf") }), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Select, {
					value: String(hPosition.relativeFrom),
					disabled,
					options: HORIZONTAL_RELATIVE_FROM,
					onChange: handleHorizontalRelativeFromChange
				})] })]
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "univer-text-gray-600 dark:!univer-text-gray-200",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { children: localeService.t("docs-drawing-ui.image-position.vertical") })
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: "univer-grid univer-grid-cols-2 univer-gap-2 [&>div]:univer-grid [&>div]:univer-gap-2",
				children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { children: localeService.t("docs-drawing-ui.image-position.absolutePosition") }), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.InputNumber, {
					min: MIN_OFFSET,
					max: MAX_OFFSET,
					precision: 1,
					disabled,
					value: vPosition.posOffset,
					onChange: (val) => {
						handlePositionChange("positionV", {
							relativeFrom: vPosition.relativeFrom,
							posOffset: val
						});
					}
				})] }), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { children: localeService.t("docs-drawing-ui.image-position.bellow") }), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Select, {
					disabled,
					value: String(vPosition.relativeFrom),
					options: VERTICAL_RELATIVE_FROM,
					onChange: handleVerticalRelativeFromChange
				})] })]
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "univer-text-gray-600 dark:!univer-text-gray-200",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { children: localeService.t("docs-drawing-ui.image-position.options") })
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Checkbox, {
				disabled,
				checked: followTextMove,
				onChange: handleFollowTextMoveCheck,
				children: localeService.t("docs-drawing-ui.image-position.moveObjectWithText")
			}) })
		]
	});
};

//#endregion
//#region src/views/doc-image-panel/DocDrawingTextWrap.tsx
const MIN_MARGIN = 0;
const MAX_MARGIN = 100;
const DocDrawingTextWrap = (props) => {
	const commandService = (0, _univerjs_ui.useDependency)(_univerjs_core.ICommandService);
	const localeService = (0, _univerjs_ui.useDependency)(_univerjs_core.LocaleService);
	const drawingManagerService = (0, _univerjs_ui.useDependency)(_univerjs_drawing.IDrawingManagerService);
	const renderManagerService = (0, _univerjs_ui.useDependency)(_univerjs_engine_render.IRenderManagerService);
	const univerInstanceService = (0, _univerjs_ui.useDependency)(_univerjs_core.IUniverInstanceService);
	const { drawings } = props;
	const drawingParam = drawings[0];
	if (drawingParam == null) return null;
	const { unitId } = drawingParam;
	const documentDataModel = univerInstanceService.getUniverDocInstance(unitId);
	const renderObject = renderManagerService.getRenderById(unitId);
	if ((renderObject === null || renderObject === void 0 ? void 0 : renderObject.scene) == null) return null;
	const [disableWrapText, setDisableWrapText] = (0, react.useState)(true);
	const [disableDistTB, setDisableDistTB] = (0, react.useState)(true);
	const [disableDistLR, setDisableDistLR] = (0, react.useState)(true);
	const [wrappingStyle, setWrappingStyle] = (0, react.useState)("inline");
	const [wrapText, setWrapText] = (0, react.useState)("");
	const [distToText, setDistToText] = (0, react.useState)({
		distT: 0,
		distL: 0,
		distB: 0,
		distR: 0
	});
	const [showPanel, setShowPanel] = (0, react.useState)(true);
	function handleWrappingStyleChange(value) {
		setWrappingStyle(value);
		const focusDrawings = drawingManagerService.getFocusDrawings();
		if (focusDrawings.length === 0) return;
		const { unitId, subUnitId } = focusDrawings[0];
		const drawings = focusDrawings.map(({ unitId, subUnitId, drawingId }) => ({
			unitId,
			subUnitId,
			drawingId
		}));
		commandService.executeCommand(UpdateDocDrawingWrappingStyleCommand.id, {
			unitId,
			subUnitId,
			drawings,
			wrappingStyle: value
		});
	}
	function handleWrapTextChange(value) {
		setWrapText(value);
		const focusDrawings = drawingManagerService.getFocusDrawings();
		if (focusDrawings.length === 0) return;
		const drawings = focusDrawings.map((drawing) => {
			return {
				unitId: drawing.unitId,
				subUnitId: drawing.subUnitId,
				drawingId: drawing.drawingId
			};
		});
		commandService.executeCommand(UpdateDocDrawingWrapTextCommand.id, {
			unitId: focusDrawings[0].unitId,
			subUnitId: focusDrawings[0].unitId,
			drawings,
			wrapText: value
		});
	}
	function handleDistToTextChange(value, direction) {
		if (value == null) return;
		setDistToText({
			...distToText,
			[direction]: value
		});
		const focusDrawings = drawingManagerService.getFocusDrawings();
		if (focusDrawings.length === 0) return;
		const drawings = focusDrawings.map((drawing) => {
			return {
				unitId: drawing.unitId,
				subUnitId: drawing.subUnitId,
				drawingId: drawing.drawingId
			};
		});
		commandService.executeCommand(UpdateDocDrawingDistanceCommand.id, {
			unitId: focusDrawings[0].unitId,
			subUnitId: focusDrawings[0].unitId,
			drawings,
			dist: { [direction]: value }
		});
	}
	function updateFocusDrawingState() {
		const focusDrawings = drawingManagerService.getFocusDrawings();
		if (focusDrawings.length === 0) return;
		updateState(focusDrawings[0]);
	}
	function updateState(drawingParam) {
		var _documentDataModel$ge;
		const drawing = documentDataModel === null || documentDataModel === void 0 || (_documentDataModel$ge = documentDataModel.getSnapshot()) === null || _documentDataModel$ge === void 0 || (_documentDataModel$ge = _documentDataModel$ge.drawings) === null || _documentDataModel$ge === void 0 ? void 0 : _documentDataModel$ge[drawingParam.drawingId];
		if (drawing == null) return;
		const { distT = 0, distL = 0, distB = 0, distR = 0, layoutType = _univerjs_core.PositionedObjectLayoutType.INLINE, behindDoc = _univerjs_core.BooleanNumber.FALSE, wrapText = _univerjs_core.WrapTextType.BOTH_SIDES } = drawing;
		setDistToText({
			distT,
			distL,
			distB,
			distR
		});
		setWrapText(wrapText);
		setDisableWrapText(layoutType !== _univerjs_core.PositionedObjectLayoutType.WRAP_SQUARE);
		if (layoutType === _univerjs_core.PositionedObjectLayoutType.WRAP_NONE || layoutType === _univerjs_core.PositionedObjectLayoutType.INLINE) setDisableDistTB(true);
		else setDisableDistTB(false);
		if (layoutType === _univerjs_core.PositionedObjectLayoutType.WRAP_NONE || layoutType === _univerjs_core.PositionedObjectLayoutType.INLINE || layoutType === _univerjs_core.PositionedObjectLayoutType.WRAP_TOP_AND_BOTTOM) setDisableDistLR(true);
		else setDisableDistLR(false);
		if (layoutType === _univerjs_core.PositionedObjectLayoutType.WRAP_NONE) if (behindDoc === _univerjs_core.BooleanNumber.TRUE) setWrappingStyle("behindText");
		else setWrappingStyle("inFrontOfText");
		else switch (layoutType) {
			case _univerjs_core.PositionedObjectLayoutType.INLINE:
				setWrappingStyle("inline");
				break;
			case _univerjs_core.PositionedObjectLayoutType.WRAP_SQUARE:
				setWrappingStyle("wrapSquare");
				break;
			case _univerjs_core.PositionedObjectLayoutType.WRAP_TOP_AND_BOTTOM:
				setWrappingStyle("wrapTopAndBottom");
				break;
			default: throw new Error(`Unsupported layout type: ${layoutType}`);
		}
	}
	(0, react.useEffect)(() => {
		updateFocusDrawingState();
		const subscription = drawingManagerService.focus$.subscribe((drawingParams) => {
			if (drawingParams.length === 0) {
				setShowPanel(false);
				return;
			}
			setShowPanel(true);
			updateState(drawingParams[0]);
		});
		const mutationListener = commandService.onCommandExecuted(async (command) => {
			if (command.id === _univerjs_docs.RichTextEditingMutation.id) updateFocusDrawingState();
		});
		return () => {
			subscription.unsubscribe();
			mutationListener.dispose();
		};
	}, []);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, _univerjs_design.clsx)("univer-grid univer-gap-2 univer-py-2 univer-text-gray-400", { "univer-hidden": !showPanel }),
		children: [
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("header", {
				className: "univer-text-gray-600 dark:!univer-text-gray-200",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { children: localeService.t("docs-drawing-ui.image-text-wrap.title") })
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "univer-text-gray-600 dark:!univer-text-gray-200",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { children: localeService.t("docs-drawing-ui.image-text-wrap.wrappingStyle") })
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(_univerjs_design.RadioGroup, {
				value: wrappingStyle,
				onChange: handleWrappingStyleChange,
				direction: "vertical",
				children: [
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Radio, {
						value: "inline",
						children: localeService.t("docs-drawing-ui.image-text-wrap.inline")
					}),
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Radio, {
						value: "wrapSquare",
						children: localeService.t("docs-drawing-ui.image-text-wrap.square")
					}),
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Radio, {
						value: "wrapTopAndBottom",
						children: localeService.t("docs-drawing-ui.image-text-wrap.topAndBottom")
					}),
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Radio, {
						value: "behindText",
						children: localeService.t("docs-drawing-ui.image-text-wrap.behindText")
					}),
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Radio, {
						value: "inFrontOfText",
						children: localeService.t("docs-drawing-ui.image-text-wrap.inFrontText")
					})
				]
			}) }),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "univer-text-gray-600 dark:!univer-text-gray-200",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { children: localeService.t("docs-drawing-ui.image-text-wrap.wrapText") })
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(_univerjs_design.RadioGroup, {
				disabled: disableWrapText,
				value: wrapText,
				onChange: handleWrapTextChange,
				direction: "horizontal",
				children: [
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Radio, {
						value: _univerjs_core.WrapTextType.BOTH_SIDES,
						children: localeService.t("docs-drawing-ui.image-text-wrap.bothSide")
					}),
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Radio, {
						value: _univerjs_core.WrapTextType.LEFT,
						children: localeService.t("docs-drawing-ui.image-text-wrap.leftOnly")
					}),
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Radio, {
						value: _univerjs_core.WrapTextType.RIGHT,
						children: localeService.t("docs-drawing-ui.image-text-wrap.rightOnly")
					})
				]
			}) }),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "univer-text-gray-600 dark:!univer-text-gray-200",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { children: localeService.t("docs-drawing-ui.image-text-wrap.distanceFromText") })
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: "univer-grid univer-grid-cols-2 univer-gap-2 [&>div]:univer-grid [&>div]:univer-gap-2",
				children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { children: localeService.t("docs-drawing-ui.image-text-wrap.top") }), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.InputNumber, {
					min: MIN_MARGIN,
					max: MAX_MARGIN,
					disabled: disableDistTB,
					precision: 1,
					value: distToText.distT,
					onChange: (val) => {
						handleDistToTextChange(val, "distT");
					}
				})] }), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { children: localeService.t("docs-drawing-ui.image-text-wrap.left") }), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.InputNumber, {
					min: MIN_MARGIN,
					max: MAX_MARGIN,
					disabled: disableDistLR,
					precision: 1,
					value: distToText.distL,
					onChange: (val) => {
						handleDistToTextChange(val, "distL");
					}
				})] })]
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: "univer-grid univer-grid-cols-2 univer-gap-2 [&>div]:univer-grid [&>div]:univer-gap-2",
				children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { children: localeService.t("docs-drawing-ui.image-text-wrap.bottom") }), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.InputNumber, {
					min: MIN_MARGIN,
					max: MAX_MARGIN,
					disabled: disableDistTB,
					precision: 1,
					value: distToText.distB,
					onChange: (val) => {
						handleDistToTextChange(val, "distB");
					}
				})] }), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { children: localeService.t("docs-drawing-ui.image-text-wrap.right") }), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.InputNumber, {
					min: MIN_MARGIN,
					max: MAX_MARGIN,
					disabled: disableDistLR,
					precision: 1,
					value: distToText.distR,
					onChange: (val) => {
						handleDistToTextChange(val, "distR");
					}
				})] })]
			})
		]
	});
};

//#endregion
//#region src/views/doc-image-panel/DocDrawingPanel.tsx
const DocDrawingPanel = () => {
	const drawingManagerService = (0, _univerjs_ui.useDependency)(_univerjs_drawing.IDrawingManagerService);
	const [drawings, setDrawings] = (0, react.useState)(drawingManagerService.getFocusDrawings());
	(0, react.useEffect)(() => {
		const focusDispose = drawingManagerService.focus$.subscribe((drawings) => {
			setDrawings(drawings);
		});
		return () => {
			focusDispose.unsubscribe();
		};
	}, []);
	return !!(drawings === null || drawings === void 0 ? void 0 : drawings.length) && /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: "univer-text-sm",
		children: [
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_drawing_ui.DrawingCommonPanel, {
				drawings,
				hasAlign: false,
				hasCropper: true,
				hasGroup: false,
				hasTransform: false
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)(DocDrawingTextWrap, { drawings }),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)(DocDrawingPosition, { drawings })
		]
	});
};

//#endregion
//#region src/controllers/shortcuts/drawing.shortcut.ts
function whenDocDrawingFocused(contextService) {
	return contextService.getContextValue(_univerjs_core.FOCUSING_DOC) && contextService.getContextValue(_univerjs_core.FOCUSING_UNIVER_EDITOR) && contextService.getContextValue(_univerjs_core.FOCUSING_COMMON_DRAWINGS);
}
const MoveDrawingDownShortcutItem = {
	id: MoveDocDrawingsCommand.id,
	description: "docs-drawing-ui.shortcut.drawing-move-down",
	group: "4_drawing-view",
	groupTitle: "docs-drawing-ui.shortcut.drawing-view",
	binding: _univerjs_ui.KeyCode.ARROW_DOWN,
	priority: 100,
	preconditions: whenDocDrawingFocused,
	staticParameters: { direction: _univerjs_core.Direction.DOWN }
};
const MoveDrawingUpShortcutItem = {
	id: MoveDocDrawingsCommand.id,
	description: "docs-drawing-ui.shortcut.drawing-move-up",
	group: "4_drawing-view",
	groupTitle: "docs-drawing-ui.shortcut.drawing-view",
	binding: _univerjs_ui.KeyCode.ARROW_UP,
	priority: 100,
	preconditions: whenDocDrawingFocused,
	staticParameters: { direction: _univerjs_core.Direction.UP }
};
const MoveDrawingLeftShortcutItem = {
	id: MoveDocDrawingsCommand.id,
	description: "docs-drawing-ui.shortcut.drawing-move-left",
	group: "4_drawing-view",
	groupTitle: "docs-drawing-ui.shortcut.drawing-view",
	binding: _univerjs_ui.KeyCode.ARROW_LEFT,
	priority: 100,
	preconditions: whenDocDrawingFocused,
	staticParameters: { direction: _univerjs_core.Direction.LEFT }
};
const MoveDrawingRightShortcutItem = {
	id: MoveDocDrawingsCommand.id,
	description: "docs-drawing-ui.shortcut.drawing-move-right",
	group: "4_drawing-view",
	groupTitle: "docs-drawing-ui.shortcut.drawing-view",
	binding: _univerjs_ui.KeyCode.ARROW_RIGHT,
	priority: 100,
	preconditions: whenDocDrawingFocused,
	staticParameters: { direction: _univerjs_core.Direction.RIGHT }
};
const DeleteDrawingsShortcutItem = {
	id: DeleteDocDrawingsCommand.id,
	description: "docs-drawing-ui.shortcut.drawing-delete",
	group: "4_drawing-view",
	groupTitle: "docs-drawing-ui.shortcut.drawing-view",
	preconditions: whenDocDrawingFocused,
	binding: _univerjs_ui.KeyCode.DELETE,
	mac: _univerjs_ui.KeyCode.BACKSPACE
};

//#endregion
//#region src/controllers/doc-drawing.controller.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
let DocDrawingUIController = class DocDrawingUIController extends _univerjs_core.Disposable {
	constructor(_componentManager, _menuManagerService, _commandService, _shortcutService) {
		super();
		this._componentManager = _componentManager;
		this._menuManagerService = _menuManagerService;
		this._commandService = _commandService;
		this._shortcutService = _shortcutService;
		this._init();
	}
	_initCustomComponents() {
		const componentManager = this._componentManager;
		this.disposeWithMe(componentManager.register(COMPONENT_DOC_DRAWING_PANEL, DocDrawingPanel));
	}
	_initMenus() {
		this._menuManagerService.mergeMenu(menuSchema);
	}
	_initCommands() {
		[
			InsertDocImageCommand,
			InsertDocDrawingCommand,
			UpdateDocDrawingWrappingStyleCommand,
			UpdateDocDrawingDistanceCommand,
			UpdateDocDrawingWrapTextCommand,
			UpdateDrawingDocTransformCommand,
			IMoveInlineDrawingCommand,
			ITransformNonInlineDrawingCommand,
			RemoveDocDrawingCommand,
			SidebarDocDrawingOperation,
			ClearDocDrawingTransformerOperation,
			EditDocDrawingOperation,
			GroupDocDrawingCommand,
			UngroupDocDrawingCommand,
			MoveDocDrawingsCommand,
			DeleteDocDrawingsCommand,
			SetDocDrawingArrangeCommand
		].forEach((command) => this.disposeWithMe(this._commandService.registerCommand(command)));
	}
	_initShortcuts() {
		[
			MoveDrawingDownShortcutItem,
			MoveDrawingUpShortcutItem,
			MoveDrawingLeftShortcutItem,
			MoveDrawingRightShortcutItem,
			DeleteDrawingsShortcutItem
		].forEach((item) => {
			this.disposeWithMe(this._shortcutService.registerShortcut(item));
		});
	}
	_init() {
		this._initCommands();
		this._initCustomComponents();
		this._initMenus();
		this._initShortcuts();
	}
};
DocDrawingUIController = __decorate([
	__decorateParam(0, (0, _univerjs_core.Inject)(_univerjs_ui.ComponentManager)),
	__decorateParam(1, _univerjs_ui.IMenuManagerService),
	__decorateParam(2, _univerjs_core.ICommandService),
	__decorateParam(3, _univerjs_ui.IShortcutService)
], DocDrawingUIController);

//#endregion
//#region src/menu/drawing-popup-menu.controller.ts
let DocDrawingPopupMenuController = class DocDrawingPopupMenuController extends _univerjs_core.RxDisposable {
	constructor(_drawingManagerService, _canvasPopManagerService, _renderManagerService, _univerInstanceService, _contextService, _commandService) {
		super();
		this._drawingManagerService = _drawingManagerService;
		this._canvasPopManagerService = _canvasPopManagerService;
		this._renderManagerService = _renderManagerService;
		this._univerInstanceService = _univerInstanceService;
		this._contextService = _contextService;
		this._commandService = _commandService;
		_defineProperty(this, "_initImagePopupMenu", /* @__PURE__ */ new Set());
		_defineProperty(this, "_disposePopups", []);
		_defineProperty(this, "_isDrawingPanelOpen", false);
		this._init();
	}
	_init() {
		this.disposeWithMe(this._commandService.onCommandExecuted((command) => {
			if (command.id === EditDocDrawingOperation.id) {
				this._isDrawingPanelOpen = true;
				this._clearPopups();
			}
			if (command.id === SidebarDocDrawingOperation.id) {
				const params = command.params;
				this._isDrawingPanelOpen = (params === null || params === void 0 ? void 0 : params.value) === "open";
				if (this._isDrawingPanelOpen) this._clearPopups();
			}
		}));
		this.disposeWithMe(this._drawingManagerService.focus$.subscribe((params) => {
			if (params.length === 0) this._isDrawingPanelOpen = false;
		}));
		this.disposeWithMe(this._univerInstanceService.getCurrentTypeOfUnit$(_univerjs_core.UniverInstanceType.UNIVER_DOC).pipe((0, rxjs.takeUntil)(this.dispose$)).subscribe((documentDataModel) => this._create(documentDataModel)));
		this.disposeWithMe(this._univerInstanceService.getTypeOfUnitDisposed$(_univerjs_core.UniverInstanceType.UNIVER_DOC).pipe((0, rxjs.takeUntil)(this.dispose$)).subscribe((documentDataModel) => this._dispose(documentDataModel)));
		this._univerInstanceService.getAllUnitsForType(_univerjs_core.UniverInstanceType.UNIVER_DOC).forEach((documentDataModel) => this._create(documentDataModel));
	}
	_dispose(documentDataModel) {
		const unitId = documentDataModel.getUnitId();
		this._clearPopups();
		this._renderManagerService.removeRender(unitId);
	}
	_clearPopups() {
		this._disposePopups.forEach((dispose) => dispose.dispose());
		this._disposePopups.length = 0;
	}
	_create(documentDataModel) {
		if (!documentDataModel) return;
		const unitId = documentDataModel.getUnitId();
		if ((0, _univerjs_core.isInternalEditorID)(unitId)) return;
		if (this._renderManagerService.has(unitId) && !this._initImagePopupMenu.has(unitId)) {
			this._popupMenuListener(unitId);
			this._initImagePopupMenu.add(unitId);
		}
	}
	_hasCropObject(scene) {
		const objects = scene.getAllObjects();
		for (const object of objects) if (object instanceof _univerjs_drawing_ui.ImageCropperObject) return true;
		return false;
	}
	_popupMenuListener(unitId) {
		var _this$_renderManagerS;
		const scene = (_this$_renderManagerS = this._renderManagerService.getRenderById(unitId)) === null || _this$_renderManagerS === void 0 ? void 0 : _this$_renderManagerS.scene;
		if (!scene) return;
		const transformer = scene.getTransformerByCreate();
		if (!transformer) return;
		const disposePopups = this._disposePopups;
		this.disposeWithMe(transformer.createControl$.subscribe(() => {
			if (this._hasCropObject(scene)) return;
			const selectedObjects = transformer.getSelectedObjectMap();
			disposePopups.forEach((dispose) => dispose.dispose());
			disposePopups.length = 0;
			if (this._isDrawingPanelOpen) return;
			if (selectedObjects.size > 1) return;
			const object = selectedObjects.values().next().value;
			if (!object) return;
			const oKey = object.oKey;
			const drawingParam = this._drawingManagerService.getDrawingOKey(oKey);
			if (!drawingParam || drawingParam.drawingType === _univerjs_core.DrawingTypeEnum.DRAWING_DOM || drawingParam.drawingType === _univerjs_core.DrawingTypeEnum.DRAWING_SHAPE) return;
			const { unitId, subUnitId, drawingId, drawingType } = drawingParam;
			const isImage = drawingType === _univerjs_core.DrawingTypeEnum.DRAWING_IMAGE;
			const popup = this._canvasPopManagerService.attachPopupToObject(object, {
				componentKey: _univerjs_drawing_ui.COMPONENT_IMAGE_POPUP_MENU,
				direction: isImage ? "top-center" : "horizontal",
				offset: isImage ? [0, 8] : [2, 0],
				extraProps: {
					menuItems: this._getImageMenuItems(unitId, subUnitId, drawingId, drawingType),
					variant: isImage ? "doc-floating-toolbar" : void 0,
					unitId,
					subUnitId,
					drawingId
				}
			}, unitId);
			disposePopups.push(this.disposeWithMe(popup));
			if (this._drawingManagerService.getFocusDrawings().find((drawing) => drawing.unitId === unitId && drawing.subUnitId === subUnitId && drawing.drawingId === drawingId)) return;
			this._drawingManagerService.focusDrawing([{
				unitId,
				subUnitId,
				drawingId
			}]);
		}));
		this.disposeWithMe(transformer.clearControl$.subscribe(() => {
			disposePopups.forEach((dispose) => dispose.dispose());
			disposePopups.length = 0;
			this._contextService.setContextValue(_univerjs_core.FOCUSING_COMMON_DRAWINGS, false);
			this._drawingManagerService.focusDrawing(null);
		}));
		this.disposeWithMe(transformer.changing$.subscribe(() => {
			disposePopups.forEach((dispose) => dispose.dispose());
			disposePopups.length = 0;
		}));
		this.disposeWithMe(transformer.changeStart$.subscribe(() => {
			disposePopups.forEach((dispose) => dispose.dispose());
			disposePopups.length = 0;
		}));
	}
	_getImageMenuItems(unitId, subUnitId, drawingId, drawingType) {
		return [
			{
				label: "docs-drawing-ui.image-popup.edit",
				index: 0,
				commandId: EditDocDrawingOperation.id,
				commandParams: {
					unitId,
					subUnitId,
					drawingId
				},
				disable: drawingType === _univerjs_core.DrawingTypeEnum.DRAWING_DOM
			},
			{
				label: "docs-drawing-ui.image-popup.delete",
				index: 1,
				commandId: RemoveDocDrawingCommand.id,
				commandParams: {
					unitId,
					drawings: [{
						unitId,
						subUnitId,
						drawingId
					}]
				},
				disable: false
			},
			{
				label: "docs-drawing-ui.image-popup.crop",
				index: 2,
				commandId: _univerjs_drawing_ui.OpenImageCropOperation.id,
				commandParams: {
					unitId,
					subUnitId,
					drawingId
				},
				disable: drawingType === _univerjs_core.DrawingTypeEnum.DRAWING_DOM
			},
			{
				label: "docs-drawing-ui.image-popup.reset",
				index: 3,
				commandId: _univerjs_drawing_ui.ImageResetSizeOperation.id,
				commandParams: [{
					unitId,
					subUnitId,
					drawingId
				}],
				disable: true
			}
		];
	}
};
DocDrawingPopupMenuController = __decorate([
	__decorateParam(0, _univerjs_drawing.IDrawingManagerService),
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_docs_ui.DocCanvasPopManagerService)),
	__decorateParam(2, _univerjs_engine_render.IRenderManagerService),
	__decorateParam(3, _univerjs_core.IUniverInstanceService),
	__decorateParam(4, _univerjs_core.IContextService),
	__decorateParam(5, _univerjs_core.ICommandService)
], DocDrawingPopupMenuController);

//#endregion
//#region src/plugin.ts
let UniverDocsDrawingUIPlugin = class UniverDocsDrawingUIPlugin extends _univerjs_core.Plugin {
	constructor(_config = defaultPluginConfig, _injector, _renderManagerSrv, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._renderManagerSrv = _renderManagerSrv;
		this._configService = _configService;
		const { ...rest } = (0, _univerjs_core.merge)({}, defaultPluginConfig, this._config);
		this._configService.setConfig(DOCS_DRAWING_UI_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		[
			[DocDrawingUIController],
			[DocDrawingPopupMenuController],
			[DocDrawingTransformerController],
			[DocDrawingAddRemoveController],
			[DocRefreshDrawingsService],
			[DocFloatDomController],
			[DocDrawingPrintingController]
		].forEach((dependency) => this._injector.add(dependency));
	}
	onReady() {
		[[DocDrawingUpdateRenderController], [DocDrawingTransformUpdateController]].forEach((m) => this._renderManagerSrv.registerRenderModule(_univerjs_core.UniverInstanceType.UNIVER_DOC, m));
		this._injector.get(DocDrawingAddRemoveController);
		this._injector.get(DocDrawingUIController);
		this._injector.get(DocDrawingTransformerController);
		this._injector.get(DocDrawingPrintingController);
	}
	onRendered() {
		this._injector.get(DocDrawingPopupMenuController);
		this._injector.get(DocFloatDomController);
	}
};
_defineProperty(UniverDocsDrawingUIPlugin, "type", _univerjs_core.UniverInstanceType.UNIVER_DOC);
_defineProperty(UniverDocsDrawingUIPlugin, "pluginName", "DOC_DRAWING_UI_PLUGIN");
_defineProperty(UniverDocsDrawingUIPlugin, "packageName", name);
_defineProperty(UniverDocsDrawingUIPlugin, "version", version);
UniverDocsDrawingUIPlugin = __decorate([
	(0, _univerjs_core.DependentOn)(_univerjs_drawing_ui.UniverDrawingUIPlugin, _univerjs_drawing.UniverDrawingPlugin, _univerjs_docs_drawing.UniverDocsDrawingPlugin, _univerjs_ui.UniverUIPlugin),
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector)),
	__decorateParam(2, _univerjs_engine_render.IRenderManagerService),
	__decorateParam(3, _univerjs_core.IConfigService)
], UniverDocsDrawingUIPlugin);

//#endregion
exports.ClearDocDrawingTransformerOperation = ClearDocDrawingTransformerOperation;
exports.DOCS_IMAGE_MENU_ID = DOCS_IMAGE_MENU_ID;
exports.DeleteDocDrawingsCommand = DeleteDocDrawingsCommand;
exports.DocDrawingPosition = DocDrawingPosition;
exports.DocDrawingTextWrap = DocDrawingTextWrap;
Object.defineProperty(exports, 'DocFloatDomController', {
  enumerable: true,
  get: function () {
    return DocFloatDomController;
  }
});
exports.EditDocDrawingOperation = EditDocDrawingOperation;
exports.GroupDocDrawingCommand = GroupDocDrawingCommand;
exports.InsertDocDrawingCommand = InsertDocDrawingCommand;
exports.InsertDocImageCommand = InsertDocImageCommand;
exports.MoveDocDrawingsCommand = MoveDocDrawingsCommand;
exports.RemoveDocDrawingCommand = RemoveDocDrawingCommand;
exports.SetDocDrawingArrangeCommand = SetDocDrawingArrangeCommand;
exports.SidebarDocDrawingOperation = SidebarDocDrawingOperation;
exports.UngroupDocDrawingCommand = UngroupDocDrawingCommand;
Object.defineProperty(exports, 'UniverDocsDrawingUIPlugin', {
  enumerable: true,
  get: function () {
    return UniverDocsDrawingUIPlugin;
  }
});
