// @univerjs/sheets-find-replace/lib/facade.js
import { Disposable, IUniverInstanceService, Inject, Injector } from "@univerjs/core";
import { FindBy, FindReplaceModel, FindReplaceState, IFindReplaceService, createInitFindReplaceState } from "@univerjs/find-replace";
import { FRange } from "@univerjs/sheets/facade";
import { filter, firstValueFrom } from "rxjs";
import { FUniver } from "@univerjs/core/facade";

//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/facade/f-text-finder.ts
let FTextFinder = class FTextFinder extends Disposable {
	constructor(_initialState, _injector, _univerInstanceService, _findReplaceService) {
		super();
		this._injector = _injector;
		this._univerInstanceService = _univerInstanceService;
		this._findReplaceService = _findReplaceService;
		_defineProperty(this, "_state", new FindReplaceState());
		_defineProperty(this, "_model", void 0);
		_defineProperty(this, "_complete", void 0);
		const providers = this._findReplaceService.getProviders();
		this._model = this._injector.createInstance(FindReplaceModel, this._state, providers);
		const newState = {
			...createInitFindReplaceState(),
			..._initialState
		};
		this._state.changeState(newState);
	}
	findAll() {
		if (!this._state.findCompleted || !this._complete) return [];
		return this._complete.results.map((result) => {
			return this._findMatchToFRange(result);
		});
	}
	findNext() {
		var _this$_model;
		if (!this._state.findCompleted || !this._complete) return null;
		const match = (_this$_model = this._model) === null || _this$_model === void 0 ? void 0 : _this$_model.moveToNextMatch();
		if (!match) return null;
		return this._findMatchToFRange(match);
	}
	findPrevious() {
		var _this$_model2;
		const match = (_this$_model2 = this._model) === null || _this$_model2 === void 0 ? void 0 : _this$_model2.moveToPreviousMatch();
		if (!match) return null;
		return this._findMatchToFRange(match);
	}
	getCurrentMatch() {
		var _this$_model3;
		if (!this._state.findCompleted || !this._complete) throw new Error("Find operation is not completed.");
		const match = (_this$_model3 = this._model) === null || _this$_model3 === void 0 ? void 0 : _this$_model3.currentMatch$.value;
		if (!match) return null;
		return this._findMatchToFRange(match);
	}
	async matchCaseAsync(matchCase) {
		this._state.changeState({
			caseSensitive: matchCase,
			findCompleted: false
		});
		await firstValueFrom(this._state.stateUpdates$.pipe(filter((state) => state.findCompleted === true)));
		await this.ensureCompleteAsync();
		return this;
	}
	async matchEntireCellAsync(matchEntireCell) {
		this._state.changeState({
			matchesTheWholeCell: matchEntireCell,
			findCompleted: false
		});
		await firstValueFrom(this._state.stateUpdates$.pipe(filter((state) => state.findCompleted === true)));
		await this.ensureCompleteAsync();
		return this;
	}
	async matchFormulaTextAsync(matchFormulaText) {
		this._state.changeState({
			findBy: matchFormulaText ? FindBy.FORMULA : FindBy.VALUE,
			findCompleted: false
		});
		await firstValueFrom(this._state.stateUpdates$.pipe(filter((state) => state.findCompleted === true)));
		await this.ensureCompleteAsync();
		return this;
	}
	async replaceAllWithAsync(replaceText) {
		var _await$this$_model$re, _await$this$_model$re2, _this$_model4;
		await this._state.changeState({
			replaceRevealed: true,
			replaceString: replaceText
		});
		const res = (_await$this$_model$re = (_await$this$_model$re2 = await ((_this$_model4 = this._model) === null || _this$_model4 === void 0 ? void 0 : _this$_model4.replaceAll())) === null || _await$this$_model$re2 === void 0 ? void 0 : _await$this$_model$re2.success) !== null && _await$this$_model$re !== void 0 ? _await$this$_model$re : 0;
		this._state.changeState({ replaceRevealed: false });
		return res;
	}
	async replaceWithAsync(replaceText) {
		var _this$_model5;
		await this._state.changeState({
			replaceRevealed: true,
			replaceString: replaceText
		});
		await ((_this$_model5 = this._model) === null || _this$_model5 === void 0 ? void 0 : _this$_model5.replace());
		this._state.changeState({ replaceRevealed: false });
		return true;
	}
	async ensureCompleteAsync() {
		var _this$_model6;
		this._complete = await ((_this$_model6 = this._model) === null || _this$_model6 === void 0 ? void 0 : _this$_model6.start());
	}
	_findMatchToFRange(match) {
		const { unitId } = match;
		const { subUnitId, range } = match.range;
		const workbook = this._univerInstanceService.getUnit(unitId);
		const worksheet = workbook.getSheetBySheetId(subUnitId);
		return this._injector.createInstance(FRange, workbook, worksheet, range);
	}
};
FTextFinder = __decorate([
	__decorateParam(1, Inject(Injector)),
	__decorateParam(2, IUniverInstanceService),
	__decorateParam(3, IFindReplaceService)
], FTextFinder);

//#endregion
//#region src/facade/f-univer.ts
var FUniverSheetsFindReplaceMixin = class extends FUniver {
	async createTextFinderAsync(text) {
		const state = { findString: text };
		const textFinder = this._injector.createInstance(FTextFinder, state);
		await textFinder.ensureCompleteAsync();
		return textFinder;
	}
};
FUniver.extend(FUniverSheetsFindReplaceMixin);

//#endregion
export {  };
