// @univerjs/sheets-find-replace/lib/index.js
import { ColorKit, CommandType, DependentOn, Disposable, EDITOR_ACTIVATED, ICommandService, IConfigService, IContextService, IUndoRedoService, IUniverInstanceService, Inject, Injector, ObjectMatrix, Plugin, Rectangle, ThemeService, Tools, UniverInstanceType, escapeRegExp, fromCallback, groupBy, merge, replaceInDocumentBody, rotate } from "@univerjs/core";
import { SelectRangeCommand, SetRangeValuesCommand, SetSelectionsOperation, SetWorksheetActivateCommand, SetWorksheetActiveOperation, SheetsSelectionsService, UniverSheetsPlugin } from "@univerjs/sheets";
import { IRenderManagerService, RENDER_RAW_FORMULA_KEY, Rect, Shape } from "@univerjs/engine-render";
import { FindBy, FindDirection, FindModel, FindReplaceController, FindScope, IFindReplaceService, UniverFindReplacePlugin } from "@univerjs/find-replace";
import { ScrollToCellCommand, SheetSkeletonManagerService, getCoordByCell, getSheetObject } from "@univerjs/sheets-ui";
import { Subject, debounceTime, filter, merge as merge$1, skip, throttleTime } from "rxjs";

//#region src/commands/commands/sheet-replace.command.ts
/**
* This command is used for the SheetFindReplaceController to deal with replacing, including undo redo.
*
*/
const SheetReplaceCommand = {
	id: "sheet.command.replace",
	type: CommandType.COMMAND,
	handler: async (accessor, params) => {
		const undoRedoService = accessor.get(IUndoRedoService);
		const commandService = accessor.get(ICommandService);
		const { unitId, replacements } = params;
		const disposeBatchingHandler = undoRedoService.__tempBatchingUndoRedo(unitId);
		const results = await Promise.all(replacements.map((replacement) => commandService.executeCommand(SetRangeValuesCommand.id, {
			unitId,
			subUnitId: replacement.subUnitId,
			value: replacement.value
		})));
		disposeBatchingHandler.dispose();
		return getReplaceAllResult(results, replacements);
	}
};
function getReplaceAllResult(results, replacements) {
	let success = 0;
	let failure = 0;
	results.forEach((r, index) => {
		const count = replacements[index].count;
		if (r) success += count;
		else failure += count;
	});
	return {
		success,
		failure
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region src/views/shapes/find-replace-highlight.shape.ts
var SheetFindReplaceHighlightShape = class extends Shape {
	constructor(key, props) {
		super(key, props);
		_defineProperty(this, "_activated", false);
		_defineProperty(this, "_inHiddenRange", false);
		_defineProperty(this, "_color", void 0);
		if (props) this.setShapeProps(props);
	}
	setShapeProps(props) {
		this._activated = !!props.activated;
		if (typeof props.inHiddenRange !== "undefined") this._inHiddenRange = props.inHiddenRange;
		if (typeof props.color !== "undefined") this._color = props.color;
		this.transformByState({
			width: props.width,
			height: props.height
		});
	}
	_draw(ctx) {
		const activated = this._activated;
		const color = `rgba(${this._color.r}, ${this._color.g}, ${this._color.b}, 0.35)`;
		const borderColor = `rgb(${this._color.r}, ${this._color.g}, ${this._color.b})`;
		Rect.drawWith(ctx, {
			width: this.width,
			height: this.height,
			fill: color,
			stroke: activated ? borderColor : void 0,
			strokeWidth: activated ? 2 : 0,
			evented: false
		});
	}
};

//#endregion
//#region src/controllers/utils.ts
function isSamePosition(range1, range2) {
	return range1.startRow === range2.startRow && range1.startColumn === range2.startColumn;
}
/**
* Tell if `range2` is after (or the same as) `range1` with row direction is at priority.
* @param range1
* @param range2
* @returns
*/
function isBehindPositionWithRowPriority(range1, range2) {
	return range1.startRow < range2.startRow || range1.startRow === range2.startRow && range1.startColumn <= range2.startColumn;
}
/**
* Tell if `range2` is after (or the same as) `range1` with column direction is at priority.
* @param range1
* @param range2
* @returns
*/
function isBehindPositionWithColumnPriority(range1, range2) {
	return range1.startColumn < range2.startColumn || range1.startColumn === range2.startColumn && range1.startRow <= range2.startRow;
}
/**
* Tell if `range2` is before (or the same as) `range1` with column direction is at priority.
* @param range1
* @param range2
* @returns
*/
function isBeforePositionWithRowPriority(range1, range2) {
	return range1.startRow > range2.startRow || range1.startRow === range2.startRow && range1.startColumn >= range2.startColumn;
}
function isBeforePositionWithColumnPriority(range1, range2) {
	return range1.startColumn > range2.startColumn || range1.startColumn === range2.startColumn && range1.startRow >= range2.startRow;
}
function isSelectionSingleCell(selection, worksheet) {
	const { range } = selection;
	const { startRow, startColumn } = range;
	const hasMergedCell = worksheet.getMergedCell(startRow, startColumn);
	if (hasMergedCell) return Rectangle.equals(range, hasMergedCell);
	else return range.endRow === range.startRow && range.endColumn === range.startColumn;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/controllers/sheet-find-replace.controller.ts
let SheetsFindReplaceController = class SheetsFindReplaceController extends Disposable {
	constructor(_injector, _findReplaceController, _contextService, _findReplaceService, _commandService) {
		super();
		this._injector = _injector;
		this._findReplaceController = _findReplaceController;
		this._contextService = _contextService;
		this._findReplaceService = _findReplaceService;
		this._commandService = _commandService;
		_defineProperty(this, "_provider", void 0);
		this._init();
		this._initCommands();
	}
	dispose() {
		super.dispose();
		this._findReplaceController.closePanel();
		this._provider.dispose();
	}
	_init() {
		const provider = this._injector.createInstance(SheetsFindReplaceProvider);
		this._provider = provider;
		this.disposeWithMe(this._findReplaceService.registerFindReplaceProvider(provider));
		this.disposeWithMe(this._contextService.subscribeContextValue$(EDITOR_ACTIVATED).pipe(filter((v) => !!v)).subscribe(() => this._findReplaceController.closePanel()));
	}
	_initCommands() {
		[SheetReplaceCommand].forEach((command) => this.disposeWithMe(this._commandService.registerCommand(command)));
	}
};
SheetsFindReplaceController = __decorate([
	__decorateParam(0, Inject(Injector)),
	__decorateParam(1, Inject(FindReplaceController)),
	__decorateParam(2, IContextService),
	__decorateParam(3, IFindReplaceService),
	__decorateParam(4, ICommandService)
], SheetsFindReplaceController);
const SHEETS_FIND_REPLACE_PROVIDER_NAME = "sheets-find-replace-provider";
const FIND_REPLACE_Z_INDEX = 1e4;
let SheetFindModel = class SheetFindModel extends FindModel {
	get _matchesCount() {
		return this._matches.length;
	}
	get unitId() {
		return this._workbook.getUnitId();
	}
	get matchesCount() {
		return this._matchesCount;
	}
	get matchesPosition() {
		return this._matchesPosition;
	}
	get currentMatch() {
		return this._matchesPosition > 0 ? this._matches[this._matchesPosition - 1] : null;
	}
	constructor(_workbook, _sheetSkeletonManagerService, _univerInstanceService, _renderManagerService, _commandService, _contextService, _themeService, _selectionManagerService) {
		super();
		this._workbook = _workbook;
		this._sheetSkeletonManagerService = _sheetSkeletonManagerService;
		this._univerInstanceService = _univerInstanceService;
		this._renderManagerService = _renderManagerService;
		this._commandService = _commandService;
		this._contextService = _contextService;
		this._themeService = _themeService;
		_defineProperty(this, "_matchesUpdate$", new Subject());
		_defineProperty(this, "matchesUpdate$", this._matchesUpdate$.asObservable());
		_defineProperty(this, "_activelyChangingMatch$", new Subject());
		_defineProperty(this, "activelyChangingMatch$", this._activelyChangingMatch$.asObservable());
		_defineProperty(this, "_matchesByWorksheet", /* @__PURE__ */ new Map());
		_defineProperty(this, "_matches", []);
		_defineProperty(this, "_matchesPosition", 0);
		_defineProperty(this, "_activeHighlightIndex", -1);
		_defineProperty(this, "_highlightShapes", []);
		_defineProperty(this, "_currentHighlightShape", null);
		_defineProperty(this, "_query", null);
		_defineProperty(this, "_workbookSelections", void 0);
		this._workbookSelections = _selectionManagerService.getWorkbookSelections(this.unitId);
	}
	dispose() {
		super.dispose();
		this._disposeHighlights();
		this._toggleDisplayRawFormula(false);
	}
	getMatches() {
		return this._matches;
	}
	start(query) {
		this._query = query;
		if (query.findBy === FindBy.FORMULA) this._toggleDisplayRawFormula(true);
		else this._toggleDisplayRawFormula(false);
		switch (query.findScope) {
			case FindScope.UNIT:
				this.findInWorkbook(query);
				break;
			case FindScope.SUBUNIT:
			default:
				this.findInActiveWorksheet(query);
				break;
		}
	}
	focusSelection() {
		const currentMatch = this.currentMatch;
		if (!currentMatch) return;
		this._commandService.executeCommand(SelectRangeCommand.id, {
			unitId: currentMatch.unitId,
			subUnit: currentMatch.range.subUnitId,
			range: currentMatch.range.range
		});
	}
	_toggleDisplayRawFormula(force) {
		this._contextService.setContextValue(RENDER_RAW_FORMULA_KEY, force);
	}
	/**
	* Find all matches in the current workbook no matter which worksheet is activated.
	* @param query the query object
	* @returns the query complete event
	*/
	findInWorkbook(query) {
		const unitId = this._workbook.getUnitId();
		let complete;
		let firstSearch = true;
		const findInWorkbook = () => {
			const allCompletes = this._workbook.getSheets().filter((worksheet) => !worksheet.isSheetHidden()).map((worksheet) => {
				const complete = this._findInWorksheet(worksheet, query, unitId);
				const sheetId = worksheet.getSheetId();
				const { results } = complete;
				if (results.length) this._matchesByWorksheet.set(sheetId, complete.results);
				else this._matchesByWorksheet.delete(sheetId);
				return complete;
			});
			this._matches = allCompletes.map((c) => c.results).flat();
			this._updateFindHighlight();
			if (firstSearch) {
				complete = { results: this._matches };
				firstSearch = false;
			} else this._matchesUpdate$.next(this._matches);
		};
		this.disposeWithMe(this._sheetSkeletonManagerService.currentSkeleton$.subscribe(() => {
			this._updateFindHighlight();
			this._updateCurrentHighlightShape(this._activeHighlightIndex);
		}));
		this.disposeWithMe(fromCallback(this._commandService.onCommandExecuted.bind(this._commandService)).pipe(filter(([command, options]) => command.id === SetWorksheetActiveOperation.id && !(options === null || options === void 0 ? void 0 : options.fromFindReplace))).subscribe(() => {
			const activeSheet = this._workbook.getActiveSheet();
			if (!activeSheet) return;
			const activeSheetId = activeSheet.getSheetId();
			if (!this._matchesByWorksheet.has(activeSheetId)) return;
			this._findNextMatchOnActiveSheetChange(activeSheet);
		}));
		this.disposeWithMe(fromCallback(this._commandService.onCommandExecuted.bind(this._commandService)).pipe(filter(([command]) => command.type === CommandType.MUTATION && command.params.unitId === this._workbook.getUnitId()), throttleTime(600, void 0, {
			leading: false,
			trailing: true
		})).subscribe(() => findInWorkbook()));
		findInWorkbook();
		return complete;
	}
	/**
	* This method is used in `findInWorkbook`. When the active sheet changes, this method helps to find the next match
	* in the new worksheet.
	*/
	_findNextMatchOnActiveSheetChange(activeSheet) {
		let match;
		let index;
		let globalIndex = 0;
		const matchesByWorksheet = this._matchesByWorksheet.get(activeSheet.getSheetId());
		const selections = this._workbookSelections.getCurrentSelections();
		if (!(selections === null || selections === void 0 ? void 0 : selections.length)) {
			match = matchesByWorksheet[0];
			index = 0;
			globalIndex = this._matches.findIndex((m) => m === match);
		} else {
			[match, globalIndex] = this._findNextMatchByRange(matchesByWorksheet, selections[0].range);
			index = matchesByWorksheet.findIndex((m) => m === match);
		}
		this._matchesPosition = globalIndex + 1;
		this._activelyChangingMatch$.next(match);
		this._activeHighlightIndex = index;
		this._updateFindHighlight();
		this._updateCurrentHighlightShape(index);
	}
	/**
	* Find all matches (only) in the currently activated worksheet.
	* @param query the query object
	* @returns the query complete event
	*/
	findInActiveWorksheet(query) {
		const unitId = this._workbook.getUnitId();
		const checkShouldFindInSelections = () => {
			var _currentSelections$so;
			const currentWorksheet = this._workbook.getActiveSheet();
			if (!currentWorksheet) return false;
			const currentSelections = this._workbookSelections.getCurrentSelections();
			return (_currentSelections$so = currentSelections === null || currentSelections === void 0 ? void 0 : currentSelections.some((selection) => !isSelectionSingleCell(selection, currentWorksheet))) !== null && _currentSelections$so !== void 0 ? _currentSelections$so : false;
		};
		let complete;
		let firstSearch = true;
		let findBySelections = false;
		const performFindInWorksheet = () => {
			const currentWorksheet = this._workbook.getActiveSheet();
			if (!currentWorksheet) return { results: [] };
			const lastMatch = this.currentMatch;
			findBySelections = checkShouldFindInSelections();
			const currentSelections = this._workbookSelections.getCurrentSelections();
			const newComplete = findBySelections ? this._findInSelections(currentWorksheet, currentSelections, query, unitId) : this._findInWorksheet(currentWorksheet, query, unitId);
			this._matches = newComplete.results;
			this._matchesPosition = this._tryRestoreLastMatchesPosition(lastMatch, this._matches);
			if (firstSearch) {
				complete = newComplete;
				firstSearch = false;
			} else this._matchesUpdate$.next(this._matches);
			this._updateFindHighlight();
			return newComplete;
		};
		this.disposeWithMe(this._sheetSkeletonManagerService.currentSkeleton$.subscribe(() => this._updateFindHighlight()));
		this.disposeWithMe(merge$1(fromCallback(this._commandService.onCommandExecuted.bind(this._commandService)).pipe(filter(([command]) => {
			if (command.type === CommandType.MUTATION && command.params.unitId === this._workbook.getUnitId()) return true;
			if (command.id === SetSelectionsOperation.id && command.params.unitId === unitId) {
				const shouldFindBySelections = checkShouldFindInSelections();
				if (shouldFindBySelections === false && findBySelections === false) return false;
				findBySelections = shouldFindBySelections;
				return true;
			}
			return false;
		})), this._workbook.activeSheet$.pipe(skip(1))).pipe(debounceTime(200)).subscribe(() => performFindInWorksheet()));
		performFindInWorksheet();
		return complete;
	}
	_findInRange(worksheet, query, range, unitId, dedupeFn) {
		const results = [];
		const subUnitId = worksheet.getSheetId();
		const iter = (query.findDirection === FindDirection.COLUMN ? worksheet.iterateByColumn : worksheet.iterateByRow).bind(worksheet)(range);
		for (const value of iter) {
			const { row, col, colSpan, rowSpan, value: cellData } = value;
			if ((dedupeFn === null || dedupeFn === void 0 ? void 0 : dedupeFn(row, col)) || !cellData) continue;
			if (worksheet.getRowFiltered(row)) continue;
			const { hit, replaceable, isFormula } = hitCell(worksheet, row, col, query, cellData);
			if (hit) {
				const result = {
					provider: SHEETS_FIND_REPLACE_PROVIDER_NAME,
					unitId,
					replaceable,
					isFormula,
					range: {
						subUnitId,
						range: {
							startRow: row,
							startColumn: col,
							endColumn: col + (colSpan !== null && colSpan !== void 0 ? colSpan : 1) - 1,
							endRow: row + (rowSpan !== null && rowSpan !== void 0 ? rowSpan : 1) - 1
						}
					}
				};
				results.push(result);
			}
		}
		return { results };
	}
	_findInSelections(worksheet, selections, query, unitId) {
		const { findDirection } = query;
		const sortFn = findDirection === FindDirection.ROW ? isBehindPositionWithRowPriority : isBehindPositionWithColumnPriority;
		const dedupeSet = /* @__PURE__ */ new Set();
		return { results: selections.map((selection) => this._findInRange(worksheet, query, selection.range, unitId, (row, col) => {
			const key = `${row}-${col}`;
			if (dedupeSet.has(key)) return true;
			dedupeSet.add(key);
			return false;
		}).results).flat().sort((a, b) => sortFn(a.range.range, b.range.range) ? -1 : 1) };
	}
	/** Find matches in a given worksheet. */
	_findInWorksheet(worksheet, query, unitId) {
		const rowCount = worksheet.getRowCount();
		const colCount = worksheet.getColumnCount();
		const range = {
			startRow: 0,
			startColumn: 0,
			endRow: rowCount - 1,
			endColumn: colCount - 1
		};
		return this._findInRange(worksheet, query, range, unitId);
	}
	_disposeHighlights() {
		var _this$_currentHighlig;
		this._highlightShapes.forEach((shape) => {
			var _shape$getScene;
			(_shape$getScene = shape.getScene()) === null || _shape$getScene === void 0 || _shape$getScene.makeDirty();
			shape.dispose();
		});
		this._highlightShapes = [];
		(_this$_currentHighlig = this._currentHighlightShape) === null || _this$_currentHighlig === void 0 || _this$_currentHighlig.dispose();
		this._currentHighlightShape = null;
	}
	_updateFindHighlight() {
		this._disposeHighlights();
		const skeleton = this._sheetSkeletonManagerService.getCurrentSkeleton();
		if (!skeleton) return;
		const unitId = this._workbook.getUnitId();
		const currentRender = this._renderManagerService.getRenderById(unitId);
		if (currentRender == null) return;
		const { scene } = currentRender;
		const matches = this._matches;
		const color = new ColorKit(this._themeService.getColorFromTheme("yellow.400")).toRgb();
		const worksheet = this._workbook.getActiveSheet();
		if (!worksheet) return;
		const activeSheetId = worksheet.getSheetId();
		const highlightShapes = matches.filter((match) => match.range.subUnitId === activeSheetId).map((find, index) => {
			const { startColumn, startRow, endColumn, endRow } = find.range.range;
			const startPosition = getCoordByCell(startRow, startColumn, scene, skeleton);
			const endPosition = getCoordByCell(endRow, endColumn, scene, skeleton);
			const { startX, startY } = startPosition;
			const { endX, endY } = endPosition;
			let isAllRowHidden = true;
			for (let row = startRow; row <= endRow; row++) if (worksheet.getRowRawVisible(row)) {
				isAllRowHidden = false;
				break;
			}
			let isAllColHidden = true;
			for (let col = startColumn; col <= endColumn; col++) if (worksheet.getColVisible(col)) {
				isAllColHidden = false;
				break;
			}
			const inHiddenRange = isAllRowHidden || isAllColHidden;
			const props = {
				left: startX,
				top: startY,
				color,
				width: isAllColHidden ? 2 : endX - startX,
				height: isAllRowHidden ? 2 : endY - startY,
				evented: false,
				inHiddenRange,
				zIndex: FIND_REPLACE_Z_INDEX
			};
			return new SheetFindReplaceHighlightShape(`find-highlight-${index}`, props);
		});
		scene.addObjects(highlightShapes);
		this._highlightShapes = highlightShapes;
		scene.makeDirty();
	}
	_updateCurrentHighlightShape(matchIndex) {
		var _this$_currentHighlig2;
		(_this$_currentHighlig2 = this._currentHighlightShape) === null || _this$_currentHighlig2 === void 0 || _this$_currentHighlig2.setShapeProps({ activated: false });
		this._currentHighlightShape = null;
		if (matchIndex !== void 0) {
			const shape = this._highlightShapes[matchIndex];
			if (!shape) return;
			this._currentHighlightShape = shape;
			shape.setShapeProps({ activated: true });
		}
	}
	_getSheetObject() {
		return getSheetObject(this._univerInstanceService, this._renderManagerService);
	}
	async _focusMatch(match) {
		var _this$_workbook$getAc;
		const { subUnitId, range } = match.range;
		if (subUnitId !== ((_this$_workbook$getAc = this._workbook.getActiveSheet()) === null || _this$_workbook$getAc === void 0 ? void 0 : _this$_workbook$getAc.getSheetId())) {
			const unitId = this._workbook.getUnitId();
			await this._commandService.executeCommand(SetWorksheetActivateCommand.id, {
				unitId,
				subUnitId
			}, { fromFindReplace: true });
		}
		this._commandService.executeCommand(ScrollToCellCommand.id, { range }, { fromFindReplace: true });
	}
	_tryRestoreLastMatchesPosition(lastMatch, newMatches) {
		if (!lastMatch) return 0;
		const { subUnitId: lastSubUnitId } = lastMatch.range;
		const { startColumn: lastStartColumn, startRow: lastStartRow } = lastMatch.range.range;
		const index = newMatches.findIndex((match) => {
			if (lastSubUnitId !== match.range.subUnitId) return false;
			const { startColumn, startRow } = match.range.range;
			return startColumn === lastStartColumn && startRow === lastStartRow;
		});
		return index > -1 ? index + 1 : 0;
	}
	moveToNextMatch(params) {
		var _params$loop, _params$stayIfOnMatch, _params$noFocus, _params$ignoreSelecti;
		if (!this._matches.length) return null;
		const loop = (_params$loop = params === null || params === void 0 ? void 0 : params.loop) !== null && _params$loop !== void 0 ? _params$loop : false;
		const stayIfOnMatch = (_params$stayIfOnMatch = params === null || params === void 0 ? void 0 : params.stayIfOnMatch) !== null && _params$stayIfOnMatch !== void 0 ? _params$stayIfOnMatch : false;
		const noFocus = (_params$noFocus = params === null || params === void 0 ? void 0 : params.noFocus) !== null && _params$noFocus !== void 0 ? _params$noFocus : false;
		const ignoreSelection = (_params$ignoreSelecti = params === null || params === void 0 ? void 0 : params.ignoreSelection) !== null && _params$ignoreSelecti !== void 0 ? _params$ignoreSelecti : false;
		const matchToMove = this._findNextMatch(loop, stayIfOnMatch, ignoreSelection);
		if (matchToMove) {
			const [match, index] = matchToMove;
			this._matchesPosition = index + 1;
			if (this._query.findScope === FindScope.UNIT) this._activeHighlightIndex = this._matchesByWorksheet.get(match.range.subUnitId).findIndex((m) => m === match);
			else this._activeHighlightIndex = index;
			if (!noFocus) this._focusMatch(match);
			this._updateCurrentHighlightShape(this._activeHighlightIndex);
			return match;
		}
		this._matchesPosition = 0;
		this._updateCurrentHighlightShape();
		return null;
	}
	moveToPreviousMatch(params) {
		var _params$loop2, _params$stayIfOnMatch2, _params$noFocus2, _params$ignoreSelecti2;
		if (!this._matches.length) return null;
		const loop = (_params$loop2 = params === null || params === void 0 ? void 0 : params.loop) !== null && _params$loop2 !== void 0 ? _params$loop2 : false;
		const stayIfOnMatch = (_params$stayIfOnMatch2 = params === null || params === void 0 ? void 0 : params.stayIfOnMatch) !== null && _params$stayIfOnMatch2 !== void 0 ? _params$stayIfOnMatch2 : false;
		const noFocus = (_params$noFocus2 = params === null || params === void 0 ? void 0 : params.noFocus) !== null && _params$noFocus2 !== void 0 ? _params$noFocus2 : false;
		const ignoreSelection = (_params$ignoreSelecti2 = params === null || params === void 0 ? void 0 : params.ignoreSelection) !== null && _params$ignoreSelecti2 !== void 0 ? _params$ignoreSelecti2 : false;
		const matchToMove = this._findPreviousMatch(loop, stayIfOnMatch, ignoreSelection);
		if (matchToMove) {
			const [match, index] = matchToMove;
			this._matchesPosition = index + 1;
			if (this._query.findScope === FindScope.UNIT) this._activeHighlightIndex = this._matchesByWorksheet.get(match.range.subUnitId).findIndex((m) => m === match);
			else this._activeHighlightIndex = index;
			if (!noFocus) this._focusMatch(match);
			this._updateCurrentHighlightShape(this._activeHighlightIndex);
			return match;
		}
		this._matchesPosition = 0;
		this._updateCurrentHighlightShape();
		return null;
	}
	_findPreviousMatch(loop = false, stayIfOnMatch = false, ignoreSelection = false) {
		var _this$_workbook$getAc2;
		if (this.currentMatch) {
			const currentMatchIndex = this._matches.findIndex((match) => match === this.currentMatch);
			if (stayIfOnMatch) return [this.currentMatch, currentMatchIndex];
			const nextMatchIndex = currentMatchIndex - 1;
			if (!loop && nextMatchIndex < 0) return null;
			const length = this._matches.length;
			const modded = (nextMatchIndex + length) % length;
			return [this._matches[modded], modded];
		}
		const lastSelection = this._workbookSelections.getCurrentLastSelection();
		if (ignoreSelection || !lastSelection) {
			const lastIndex = this._matches.length - 1;
			return [this._matches[lastIndex], lastIndex];
		}
		if (this._query.findScope !== FindScope.UNIT) return this._findPreviousMatchByRange(this._matches, lastSelection.range);
		const currentSheetId = (_this$_workbook$getAc2 = this._workbook.getActiveSheet()) === null || _this$_workbook$getAc2 === void 0 ? void 0 : _this$_workbook$getAc2.getSheetId();
		if (!currentSheetId) return null;
		const worksheetThatHasMatch = this._findPreviousWorksheetThatHasAMatch(currentSheetId, loop);
		if (!worksheetThatHasMatch) return null;
		return this._findPreviousMatchByRange(this._matchesByWorksheet.get(worksheetThatHasMatch), lastSelection.range);
	}
	_findNextMatch(loop = false, stayIfOnMatch = false, ignoreSelection = false) {
		var _this$_workbook$getAc3;
		if (this.currentMatch) {
			const currentMatchIndex = this._matches.findIndex((match) => match === this.currentMatch);
			if (stayIfOnMatch) return [this.currentMatch, currentMatchIndex];
			const nextMatchIndex = currentMatchIndex + 1;
			const length = this._matches.length;
			if (!loop && nextMatchIndex >= length) return null;
			const modded = nextMatchIndex % length;
			return [this._matches[modded], modded];
		}
		const last = this._workbookSelections.getCurrentLastSelection();
		if (ignoreSelection || !last) return [this._matches[0], 0];
		if (this._query.findScope !== FindScope.UNIT) return this._findNextMatchByRange(this._matches, last.range, stayIfOnMatch);
		const currentSheetId = (_this$_workbook$getAc3 = this._workbook.getActiveSheet()) === null || _this$_workbook$getAc3 === void 0 ? void 0 : _this$_workbook$getAc3.getSheetId();
		if (!currentSheetId) return null;
		const worksheetThatHasMatch = this._findNextWorksheetThatHasAMatch(currentSheetId, loop);
		if (!worksheetThatHasMatch) return null;
		return this._findNextMatchByRange(this._matchesByWorksheet.get(worksheetThatHasMatch), last.range);
	}
	_findPreviousWorksheetThatHasAMatch(currentWorksheet, loop = false) {
		const rawWorksheetsInOrder = this._workbook.getSheetOrders();
		const currentSheetIndex = rawWorksheetsInOrder.findIndex((sheet) => sheet === currentWorksheet);
		const first = (loop ? rotate(rawWorksheetsInOrder, currentSheetIndex + 1) : rawWorksheetsInOrder.slice(0, currentSheetIndex + 1)).findLast((worksheet) => this._matchesByWorksheet.has(worksheet));
		return first !== null && first !== void 0 ? first : null;
	}
	_findNextWorksheetThatHasAMatch(currentWorksheet, loop = false) {
		const rawWorksheetsInOrder = this._workbook.getSheetOrders();
		const currentSheetIndex = rawWorksheetsInOrder.findIndex((sheet) => sheet === currentWorksheet);
		const first = (loop ? rotate(rawWorksheetsInOrder, currentSheetIndex) : rawWorksheetsInOrder.slice(currentSheetIndex)).find((worksheet) => this._matchesByWorksheet.has(worksheet));
		return first !== null && first !== void 0 ? first : null;
	}
	_findNextMatchByRange(matches, range, stayIfOnMatch = false) {
		const findByRow = this._query.findDirection === FindDirection.ROW;
		let index = matches.findIndex((match) => {
			const matchRange = match.range.range;
			if (!(findByRow ? isBehindPositionWithRowPriority(range, matchRange) : isBehindPositionWithColumnPriority(range, matchRange))) return false;
			const isSame = isSamePosition(range, matchRange);
			return stayIfOnMatch ? isSame : !isSame;
		});
		if (index === -1) index = matches.length - 1;
		const match = matches[index];
		return [match, this._matches.findIndex((m) => m === match)];
	}
	_findPreviousMatchByRange(matches, range, stayIfOnMatch = false) {
		const findByRow = this._query.findDirection === FindDirection.ROW;
		let index = this._matches.findLastIndex((match) => {
			const matchRange = match.range.range;
			if (!(findByRow ? isBeforePositionWithRowPriority(range, matchRange) : isBeforePositionWithColumnPriority(range, matchRange))) return false;
			const isSame = isSamePosition(range, matchRange);
			return stayIfOnMatch ? isSame : !isSame;
		});
		if (index === -1) index = 0;
		const match = matches[index];
		return [match, this._matches.findIndex((m) => m === match)];
	}
	async replace(replaceString) {
		if (this._matchesCount === 0 || !this.currentMatch || !this._query || !this.currentMatch.replaceable) return false;
		const range = this.currentMatch.range;
		const targetWorksheet = this._workbook.getSheetBySheetId(this.currentMatch.range.subUnitId);
		const newContent = this._getReplacedCellData(this.currentMatch, targetWorksheet, this._query.findBy === FindBy.FORMULA, this._query.findString, replaceString, this._query.caseSensitive ? "g" : "ig");
		const params = {
			unitId: this.currentMatch.unitId,
			subUnitId: range.subUnitId,
			value: { [range.range.startRow]: { [range.range.startColumn]: newContent } }
		};
		return this._commandService.executeCommand(SetRangeValuesCommand.id, params);
	}
	async replaceAll(replaceString) {
		if (this._matchesCount === 0 || !this._query) return {
			success: 0,
			failure: 0
		};
		const unitId = this._workbook.getUnitId();
		const { findString, caseSensitive, findBy } = this._query;
		const shouldReplaceFormula = findBy === FindBy.FORMULA;
		const replaceFlag = caseSensitive ? "g" : "ig";
		const replacements = [];
		groupBy(this._matches.filter((m) => m.replaceable), (match) => match.range.subUnitId).forEach((matches, subUnitId) => {
			const matrix = new ObjectMatrix();
			const worksheet = this._workbook.getSheetBySheetId(subUnitId);
			matches.forEach((match) => {
				const { startColumn, startRow } = match.range.range;
				const newCellData = this._getReplacedCellData(match, worksheet, shouldReplaceFormula, findString, replaceString, replaceFlag);
				if (newCellData) matrix.setValue(startRow, startColumn, newCellData);
			});
			replacements.push({
				count: matches.length,
				subUnitId,
				value: matrix.getMatrix()
			});
		});
		if (!replacements) return {
			success: 0,
			failure: 0
		};
		return this._commandService.executeCommand(SheetReplaceCommand.id, {
			unitId,
			replacements
		});
	}
	_getReplacedCellData(match, worksheet, shouldReplaceFormula, findString, replaceString, replaceFlag) {
		var _currentContent$p;
		const { startRow, startColumn } = match.range.range;
		const currentContent = worksheet.getCellRaw(startRow, startColumn);
		if (match.isFormula) {
			if (!shouldReplaceFormula) return null;
			return {
				f: currentContent.f.replace(new RegExp(escapeRegExp(findString), replaceFlag), replaceString),
				v: null
			};
		}
		if (!!((_currentContent$p = currentContent.p) === null || _currentContent$p === void 0 ? void 0 : _currentContent$p.body)) {
			const clonedRichText = Tools.deepClone(currentContent.p);
			replaceInDocumentBody(clonedRichText.body, findString, replaceString, this._query.caseSensitive);
			return { p: clonedRichText };
		}
		return { v: currentContent.v.toString().replace(new RegExp(escapeRegExp(findString), replaceFlag), replaceString) };
	}
};
SheetFindModel = __decorate([
	__decorateParam(2, IUniverInstanceService),
	__decorateParam(3, IRenderManagerService),
	__decorateParam(4, ICommandService),
	__decorateParam(5, IContextService),
	__decorateParam(6, Inject(ThemeService)),
	__decorateParam(7, Inject(SheetsSelectionsService))
], SheetFindModel);
/**
* This module is responsible for searching and replacing in the sheets.
* It also adds the search results to the search view by highlighting them.
*/
let SheetsFindReplaceProvider = class SheetsFindReplaceProvider extends Disposable {
	constructor(_univerInstanceService, _renderManagerService, _injector) {
		super();
		this._univerInstanceService = _univerInstanceService;
		this._renderManagerService = _renderManagerService;
		this._injector = _injector;
		_defineProperty(this, "_findModelsByUnitId", /* @__PURE__ */ new Map());
	}
	async find(query) {
		this._terminate();
		const workbook = this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_SHEET);
		if (!workbook) return [];
		const parsedQuery = this._preprocessQuery(query);
		const skeletonManagerService = this._renderManagerService.getRenderById(workbook.getUnitId()).with(SheetSkeletonManagerService);
		const sheetFind = this._injector.createInstance(SheetFindModel, workbook, skeletonManagerService);
		this._findModelsByUnitId.set(workbook.getUnitId(), sheetFind);
		sheetFind.start(parsedQuery);
		return [sheetFind];
	}
	terminate() {
		this._terminate();
	}
	_terminate() {
		this._findModelsByUnitId.forEach((model) => model.dispose());
		this._findModelsByUnitId.clear();
	}
	/**
	* Parsed the query object before do actual searching in favor of performance.
	* @param query the raw query object
	* @returns the parsed query object
	*/
	_preprocessQuery(query) {
		let findString = query.caseSensitive ? query.findString : query.findString.toLowerCase();
		findString = findString.trim();
		return {
			...query,
			findString
		};
	}
};
SheetsFindReplaceProvider = __decorate([
	__decorateParam(0, IUniverInstanceService),
	__decorateParam(1, IRenderManagerService),
	__decorateParam(2, Inject(Injector))
], SheetsFindReplaceProvider);
const VALUE_PASSING_OBJECT = {
	hit: false,
	replaceable: false,
	isFormula: false,
	rawData: null
};
/**
* This function determines if a cell's content matches what is searched for.
* @param worksheet worksheet the Worksheet to search
* @param row the row index
* @param col the column index
* @param query the parsed query object
* @returns if the cell is hit, replaceable and is a formula
*/
function hitCell(worksheet, row, col, query, cellData) {
	const { findBy } = query;
	const findByFormula = findBy === FindBy.FORMULA;
	const rawData = worksheet.getCellRaw(row, col);
	VALUE_PASSING_OBJECT.rawData = rawData;
	if (!!(rawData === null || rawData === void 0 ? void 0 : rawData.f)) {
		VALUE_PASSING_OBJECT.isFormula = true;
		if (findByFormula) {
			if (matchCellData({ v: rawData.f }, query)) {
				VALUE_PASSING_OBJECT.hit = true;
				VALUE_PASSING_OBJECT.replaceable = true;
				return VALUE_PASSING_OBJECT;
			}
			VALUE_PASSING_OBJECT.hit = false;
			VALUE_PASSING_OBJECT.replaceable = false;
			return VALUE_PASSING_OBJECT;
		}
		VALUE_PASSING_OBJECT.replaceable = false;
		if (matchCellData(cellData, query)) VALUE_PASSING_OBJECT.hit = true;
		else VALUE_PASSING_OBJECT.hit = false;
		return VALUE_PASSING_OBJECT;
	}
	VALUE_PASSING_OBJECT.isFormula = false;
	if (!matchCellData(cellData, query)) {
		VALUE_PASSING_OBJECT.hit = false;
		VALUE_PASSING_OBJECT.replaceable = false;
	} else if (!rawData) {
		VALUE_PASSING_OBJECT.hit = true;
		VALUE_PASSING_OBJECT.replaceable = false;
	} else {
		VALUE_PASSING_OBJECT.hit = true;
		VALUE_PASSING_OBJECT.replaceable = true;
	}
	return VALUE_PASSING_OBJECT;
}
function matchCellData(cellData, query) {
	let value = extractPureValue(cellData);
	if (!value) return false;
	if (query.matchesTheWholeCell) {
		value = trimLeadingTrailingWhitespace(value);
		return query.caseSensitive ? value === query.findString : value.toLowerCase() === query.findString;
	}
	return query.caseSensitive ? value.indexOf(query.findString) > -1 : value.toLowerCase().indexOf(query.findString) > -1;
}
function extractPureValue(cell) {
	var _cell$p$body$dataStre, _cell$p;
	const rawValue = (_cell$p$body$dataStre = cell === null || cell === void 0 || (_cell$p = cell.p) === null || _cell$p === void 0 || (_cell$p = _cell$p.body) === null || _cell$p === void 0 ? void 0 : _cell$p.dataStream) !== null && _cell$p$body$dataStre !== void 0 ? _cell$p$body$dataStre : cell === null || cell === void 0 ? void 0 : cell.v;
	if (typeof rawValue === "number") return `${rawValue}`;
	if (typeof rawValue === "boolean") return rawValue ? "1" : "0";
	return rawValue;
}
/**
* Trim only white spaces but not line breaks from the start and end of a string.
* @param value the string to be trimmed
*/
function trimLeadingTrailingWhitespace(value) {
	return value.replace(/^ +/g, "").replace(/ +$/g, "");
}

//#endregion
//#region package.json
var name = "@univerjs/sheets-find-replace";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const SHEETS_FIND_REPLACE_PLUGIN_CONFIG_KEY = "sheets-find-replace.config";
const configSymbol = Symbol(SHEETS_FIND_REPLACE_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/plugin.ts
let UniverSheetsFindReplacePlugin = class UniverSheetsFindReplacePlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _injector, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		const { ...rest } = merge({}, defaultPluginConfig, this._config);
		this._configService.setConfig(SHEETS_FIND_REPLACE_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		[[SheetsFindReplaceController]].forEach((d) => this._injector.add(d));
	}
	onSteady() {
		this._injector.get(SheetsFindReplaceController);
	}
};
_defineProperty(UniverSheetsFindReplacePlugin, "pluginName", "SHEET_FIND_REPLACE_PLUGIN");
_defineProperty(UniverSheetsFindReplacePlugin, "packageName", name);
_defineProperty(UniverSheetsFindReplacePlugin, "version", version);
_defineProperty(UniverSheetsFindReplacePlugin, "type", UniverInstanceType.UNIVER_SHEET);
UniverSheetsFindReplacePlugin = __decorate([
	DependentOn(UniverSheetsPlugin, UniverSheetsPlugin, UniverFindReplacePlugin),
	__decorateParam(1, Inject(Injector)),
	__decorateParam(2, IConfigService)
], UniverSheetsFindReplacePlugin);

//#endregion
export { SheetReplaceCommand, SheetsFindReplaceController, UniverSheetsFindReplacePlugin };
