// @univerjs/sheets-filter-ui/lib/index.js
import { BooleanNumber, ColorKit, CommandType, DependentOn, Disposable, ICommandService, IConfigService, IContextService, ILogService, IUniverInstanceService, Inject, Injector, InterceptorEffectEnum, LocaleService, Optional, Plugin, Quantity, RxDisposable, ThemeService, Tools, UniverInstanceType, VerticalAlign, createIdentifier, extractPureTextFromCell, fromCallback, isNumeric, merge, numfmt, registerDependencies, touchDependencies } from "@univerjs/core";
import { ClearSheetsFilterCriteriaCommand, CustomFilterOperator, FILTER_MUTATIONS, FilterBy, ReCalcSheetsFilterCommand, ReCalcSheetsFilterMutation, RemoveSheetsFilterMutation, SetSheetsFilterCriteriaCommand, SetSheetsFilterCriteriaMutation, SetSheetsFilterRangeMutation, SheetsFilterService, SheetsFilterSyncController, SmartToggleSheetsFilterCommand, UniverSheetsFilterPlugin } from "@univerjs/sheets-filter";
import { IEditorBridgeService, ISheetSelectionRenderService, SelectionControl, SetCellEditVisibleOperation, SheetCanvasPopManagerService, SheetSkeletonManagerService, SheetsRenderService, SheetsUIPart, getCoordByCell, getCurrentRangeDisable$, getObservableWithExclusiveRange$, whenSheetEditorFocused } from "@univerjs/sheets-ui";
import { ComponentContainer, ComponentManager, ILayoutService, IMenuManagerService, IMessageService, IShortcutService, KeyCode, MenuItemType, MetaKeys, RibbonDataGroup, getMenuHiddenObservable, useComponentsOfPart, useDependency, useObservable } from "@univerjs/ui";
import { COLOR_BLACK_RGB, IRenderManagerService, Rect, Shape } from "@univerjs/engine-render";
import { INTERCEPTOR_POINT, RangeProtectionPermissionViewPoint, RefRangeService, SetRangeValuesMutation, SheetInterceptorService, SheetPermissionCheckController, SheetsSelectionsService, WorksheetFilterPermission, WorksheetViewPermission, attachSelectionWithCoord, expandToContinuousRange, getSheetCommandTarget } from "@univerjs/sheets";
import { BehaviorSubject, ReplaySubject, Subject, combineLatest, distinctUntilChanged, filter, map, merge as merge$1, of, shareReplay, startWith, switchMap, takeUntil, throttleTime } from "rxjs";
import { IRPCChannelService, fromModule, toModule } from "@univerjs/rpc";
import { Button, Checkbox, Input, MessageType, Radio, RadioGroup, Segmented, Select, Switch, Tooltip, Tree, borderClassName, clsx } from "@univerjs/design";
import { AIcon, BanIcon, FilterIcon, InfoIcon, SuccessIcon } from "@univerjs/icons";
import { useCallback, useMemo } from "react";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";

//#region src/models/conditions.ts
let FilterConditionItems;
(function(_FilterConditionItems) {
	const NONE = _FilterConditionItems.NONE = {
		label: "sheets-filter-ui.conditions.none",
		operator: "none",
		order: 1,
		numOfParameters: 0,
		getDefaultFormParams: () => {
			throw new Error("[FilterConditionItems.NONE]: should not have initial form params!");
		},
		testMappingParams: (params) => {
			return params.operator1 === "none";
		},
		mapToFilterColumn: () => null,
		testMappingFilterColumn: (filterColumn) => {
			if (!filterColumn.customFilters && !filterColumn.filters) return {};
			return false;
		}
	};
	const ALL_CONDITIONS = _FilterConditionItems.ALL_CONDITIONS = [
		NONE,
		_FilterConditionItems.EMPTY = {
			label: "sheets-filter-ui.conditions.empty",
			operator: "empty",
			order: 1,
			numOfParameters: 0,
			getDefaultFormParams: () => {
				throw new Error("[FilterConditionItems.EMPTY]: should not have initial form params!");
			},
			testMappingParams: ({ operator1 }) => operator1 === "empty",
			mapToFilterColumn: () => ({ customFilters: { customFilters: [{ val: "" }] } }),
			testMappingFilterColumn: (filterColumn) => {
				var _filterColumn$customF;
				if (((_filterColumn$customF = filterColumn.customFilters) === null || _filterColumn$customF === void 0 ? void 0 : _filterColumn$customF.customFilters.length) !== 1) return false;
				const firstCustomFilter = filterColumn.customFilters.customFilters[0];
				if (!(firstCustomFilter.val === "" && firstCustomFilter.operator === void 0)) return false;
				return { operator1: "empty" };
			}
		},
		_FilterConditionItems.NOT_EMPTY = {
			label: "sheets-filter-ui.conditions.not-empty",
			operator: "notEmpty",
			order: 1,
			numOfParameters: 0,
			getDefaultFormParams: () => {
				throw new Error("[FilterConditionItems.NOT_EMPTY]: should not have initial form params!");
			},
			testMappingParams: ({ operator1 }) => operator1 === "notEmpty",
			mapToFilterColumn: () => ({ customFilters: { customFilters: [{
				val: "",
				operator: CustomFilterOperator.NOT_EQUALS
			}] } }),
			testMappingFilterColumn: (filterColumn) => {
				var _filterColumn$customF2;
				if (((_filterColumn$customF2 = filterColumn.customFilters) === null || _filterColumn$customF2 === void 0 ? void 0 : _filterColumn$customF2.customFilters.length) !== 1) return false;
				const firstCustomFilter = filterColumn.customFilters.customFilters[0];
				if (!(firstCustomFilter.val === " " && firstCustomFilter.operator === CustomFilterOperator.NOT_EQUALS)) return false;
				return { operator1: "notEmpty" };
			}
		},
		_FilterConditionItems.TEXT_CONTAINS = {
			label: "sheets-filter-ui.conditions.text-contains",
			operator: "contains",
			order: 0,
			numOfParameters: 1,
			getDefaultFormParams: () => ({
				operator1: "contains",
				val1: ""
			}),
			testMappingParams: (params) => {
				const [op] = getOnlyOperatorAndVal(params);
				return op === "contains";
			},
			mapToFilterColumn: (mapParams) => {
				const { val1 } = mapParams;
				if (val1 === "") return null;
				return { customFilters: { customFilters: [{ val: `*${val1}*` }] } };
			},
			testMappingFilterColumn: (filterColumn) => {
				var _filterColumn$customF3;
				if (((_filterColumn$customF3 = filterColumn.customFilters) === null || _filterColumn$customF3 === void 0 ? void 0 : _filterColumn$customF3.customFilters.length) !== 1) return false;
				const firstCustomFilter = filterColumn.customFilters.customFilters[0];
				const valAsString = firstCustomFilter.val.toString();
				if (!firstCustomFilter.operator && valAsString.startsWith("*") && valAsString.endsWith("*")) return {
					operator1: "contains",
					val1: valAsString.slice(1, -1)
				};
				return false;
			}
		},
		_FilterConditionItems.DOES_NOT_CONTAIN = {
			label: "sheets-filter-ui.conditions.does-not-contain",
			operator: "doesNotContain",
			order: 0,
			numOfParameters: 1,
			getDefaultFormParams: () => ({
				operator1: "doesNotContain",
				val1: ""
			}),
			mapToFilterColumn: (mapParams) => ({ customFilters: { customFilters: [{
				val: `*${mapParams.val1}*`,
				operator: CustomFilterOperator.NOT_EQUALS
			}] } }),
			testMappingParams: (params) => {
				const [op] = getOnlyOperatorAndVal(params);
				return op === "doesNotContain";
			},
			testMappingFilterColumn: (filterColumn) => {
				var _filterColumn$customF4;
				if (((_filterColumn$customF4 = filterColumn.customFilters) === null || _filterColumn$customF4 === void 0 ? void 0 : _filterColumn$customF4.customFilters.length) !== 1) return false;
				const firstCustomFilter = filterColumn.customFilters.customFilters[0];
				const valAsString = firstCustomFilter.val.toString();
				if (firstCustomFilter.operator === CustomFilterOperator.NOT_EQUALS && valAsString.startsWith("*") && valAsString.endsWith("*")) return {
					operator1: "doesNotContain",
					val1: valAsString.slice(1, -1)
				};
				return false;
			}
		},
		_FilterConditionItems.STARTS_WITH = {
			label: "sheets-filter-ui.conditions.starts-with",
			operator: "startsWith",
			order: 0,
			numOfParameters: 1,
			getDefaultFormParams: () => ({
				operator1: "startsWith",
				val1: ""
			}),
			mapToFilterColumn: (mapParams) => ({ customFilters: { customFilters: [{ val: `${mapParams.val1}*` }] } }),
			testMappingParams: (params) => {
				const [op] = getOnlyOperatorAndVal(params);
				return op === "startsWith";
			},
			testMappingFilterColumn: (filterColumn) => {
				var _filterColumn$customF5;
				if (((_filterColumn$customF5 = filterColumn.customFilters) === null || _filterColumn$customF5 === void 0 ? void 0 : _filterColumn$customF5.customFilters.length) !== 1) return false;
				const firstCustomFilter = filterColumn.customFilters.customFilters[0];
				const valAsString = firstCustomFilter.val.toString();
				if (!firstCustomFilter.operator && valAsString.endsWith("*") && !valAsString.startsWith("*")) return {
					operator1: "startsWith",
					val1: valAsString.slice(0, -1)
				};
				return false;
			}
		},
		_FilterConditionItems.ENDS_WITH = {
			label: "sheets-filter-ui.conditions.ends-with",
			operator: "endsWith",
			order: 0,
			numOfParameters: 1,
			getDefaultFormParams: () => ({
				operator1: "endsWith",
				val1: ""
			}),
			mapToFilterColumn: (mapParams) => ({ customFilters: { customFilters: [{ val: `*${mapParams.val1}` }] } }),
			testMappingParams: (params) => {
				const [op] = getOnlyOperatorAndVal(params);
				return op === "endsWith";
			},
			testMappingFilterColumn: (filterColumn) => {
				var _filterColumn$customF6;
				if (((_filterColumn$customF6 = filterColumn.customFilters) === null || _filterColumn$customF6 === void 0 ? void 0 : _filterColumn$customF6.customFilters.length) !== 1) return false;
				const firstCustomFilter = filterColumn.customFilters.customFilters[0];
				const valAsString = firstCustomFilter.val.toString();
				if (!firstCustomFilter.operator && valAsString.startsWith("*") && !valAsString.endsWith("*")) return {
					operator1: "endsWith",
					val1: valAsString.slice(1)
				};
				return false;
			}
		},
		_FilterConditionItems.EQUALS = {
			label: "sheets-filter-ui.conditions.equals",
			operator: "equals",
			order: 0,
			numOfParameters: 1,
			getDefaultFormParams: () => ({
				operator1: "equals",
				val1: ""
			}),
			testMappingParams: (params) => {
				const [op] = getOnlyOperatorAndVal(params);
				return op === "equals";
			},
			mapToFilterColumn: (mapParams) => {
				const { val1 } = mapParams;
				if (val1 === "") return null;
				return { customFilters: { customFilters: [{ val: val1 }] } };
			},
			testMappingFilterColumn: (filterColumn) => {
				var _filterColumn$filters, _filterColumn$customF7;
				if (((_filterColumn$filters = filterColumn.filters) === null || _filterColumn$filters === void 0 || (_filterColumn$filters = _filterColumn$filters.filters) === null || _filterColumn$filters === void 0 ? void 0 : _filterColumn$filters.length) === 1) return {
					operator1: "equals",
					val1: ""
				};
				if (((_filterColumn$customF7 = filterColumn.customFilters) === null || _filterColumn$customF7 === void 0 ? void 0 : _filterColumn$customF7.customFilters.length) === 1 && !filterColumn.customFilters.customFilters[0].operator) return {
					operator1: "equals",
					val1: filterColumn.customFilters.customFilters[0].val.toString()
				};
				return false;
			}
		},
		_FilterConditionItems.GREATER_THAN = {
			label: "sheets-filter-ui.conditions.greater-than",
			operator: CustomFilterOperator.GREATER_THAN,
			numOfParameters: 1,
			order: 0,
			getDefaultFormParams: () => ({
				operator1: CustomFilterOperator.GREATER_THAN,
				val1: ""
			}),
			mapToFilterColumn: (mapParams) => ({ customFilters: { customFilters: [{
				val: mapParams.val1,
				operator: CustomFilterOperator.GREATER_THAN
			}] } }),
			testMappingParams: (params) => {
				const [op] = getOnlyOperatorAndVal(params);
				return op === CustomFilterOperator.GREATER_THAN;
			},
			testMappingFilterColumn: (filterColumn) => {
				var _filterColumn$customF8;
				if (((_filterColumn$customF8 = filterColumn.customFilters) === null || _filterColumn$customF8 === void 0 ? void 0 : _filterColumn$customF8.customFilters.length) !== 1) return false;
				const firstCustomFilter = filterColumn.customFilters.customFilters[0];
				if (firstCustomFilter.operator !== CustomFilterOperator.GREATER_THAN) return false;
				return {
					operator1: CustomFilterOperator.GREATER_THAN,
					val1: firstCustomFilter.val.toString()
				};
			}
		},
		_FilterConditionItems.GREATER_THAN_OR_EQUAL = {
			label: "sheets-filter-ui.conditions.greater-than-or-equal",
			operator: CustomFilterOperator.GREATER_THAN_OR_EQUAL,
			numOfParameters: 1,
			order: 0,
			getDefaultFormParams: () => ({
				operator1: CustomFilterOperator.GREATER_THAN_OR_EQUAL,
				val1: ""
			}),
			testMappingParams: (params) => {
				const [op] = getOnlyOperatorAndVal(params);
				return op === CustomFilterOperator.GREATER_THAN_OR_EQUAL;
			},
			mapToFilterColumn: (mapParams) => ({ customFilters: { customFilters: [{
				val: mapParams.val1,
				operator: CustomFilterOperator.GREATER_THAN_OR_EQUAL
			}] } }),
			testMappingFilterColumn: (filterColumn) => {
				var _filterColumn$customF9;
				if (((_filterColumn$customF9 = filterColumn.customFilters) === null || _filterColumn$customF9 === void 0 ? void 0 : _filterColumn$customF9.customFilters.length) !== 1) return false;
				const firstCustomFilter = filterColumn.customFilters.customFilters[0];
				if (firstCustomFilter.operator !== CustomFilterOperator.GREATER_THAN_OR_EQUAL) return false;
				return {
					operator1: CustomFilterOperator.GREATER_THAN_OR_EQUAL,
					val1: firstCustomFilter.val.toString()
				};
			}
		},
		_FilterConditionItems.LESS_THAN = {
			label: "sheets-filter-ui.conditions.less-than",
			operator: CustomFilterOperator.LESS_THAN,
			numOfParameters: 1,
			order: 0,
			getDefaultFormParams: () => ({
				operator1: CustomFilterOperator.LESS_THAN,
				val1: ""
			}),
			testMappingParams: (params) => {
				const [op] = getOnlyOperatorAndVal(params);
				return op === CustomFilterOperator.LESS_THAN;
			},
			mapToFilterColumn: (mapParams) => ({ customFilters: { customFilters: [{
				val: mapParams.val1,
				operator: CustomFilterOperator.LESS_THAN
			}] } }),
			testMappingFilterColumn: (filterColumn) => {
				var _filterColumn$customF10;
				if (((_filterColumn$customF10 = filterColumn.customFilters) === null || _filterColumn$customF10 === void 0 ? void 0 : _filterColumn$customF10.customFilters.length) !== 1) return false;
				const firstCustomFilter = filterColumn.customFilters.customFilters[0];
				if (firstCustomFilter.operator !== CustomFilterOperator.LESS_THAN) return false;
				return {
					operator1: CustomFilterOperator.LESS_THAN,
					val1: firstCustomFilter.val.toString()
				};
			}
		},
		_FilterConditionItems.LESS_THAN_OR_EQUAL = {
			label: "sheets-filter-ui.conditions.less-than-or-equal",
			operator: CustomFilterOperator.LESS_THAN_OR_EQUAL,
			numOfParameters: 1,
			order: 0,
			getDefaultFormParams: () => ({
				operator1: CustomFilterOperator.LESS_THAN_OR_EQUAL,
				val1: ""
			}),
			testMappingParams: (params) => {
				const [op] = getOnlyOperatorAndVal(params);
				return op === CustomFilterOperator.LESS_THAN_OR_EQUAL;
			},
			mapToFilterColumn: (mapParams) => ({ customFilters: { customFilters: [{
				val: mapParams.val1,
				operator: CustomFilterOperator.LESS_THAN_OR_EQUAL
			}] } }),
			testMappingFilterColumn: (filterColumn) => {
				var _filterColumn$customF11;
				if (((_filterColumn$customF11 = filterColumn.customFilters) === null || _filterColumn$customF11 === void 0 ? void 0 : _filterColumn$customF11.customFilters.length) !== 1) return false;
				const firstCustomFilter = filterColumn.customFilters.customFilters[0];
				if (firstCustomFilter.operator !== CustomFilterOperator.LESS_THAN_OR_EQUAL) return false;
				return {
					operator1: CustomFilterOperator.LESS_THAN_OR_EQUAL,
					val1: firstCustomFilter.val.toString()
				};
			}
		},
		_FilterConditionItems.EQUAL = {
			label: "sheets-filter-ui.conditions.equal",
			operator: CustomFilterOperator.EQUAL,
			numOfParameters: 1,
			order: 0,
			getDefaultFormParams: () => ({
				operator1: CustomFilterOperator.EQUAL,
				val1: ""
			}),
			testMappingParams: (params) => {
				const [op] = getOnlyOperatorAndVal(params);
				return op === CustomFilterOperator.EQUAL;
			},
			mapToFilterColumn: (mapParams) => ({ customFilters: { customFilters: [{
				val: mapParams.val1,
				operator: CustomFilterOperator.EQUAL
			}] } }),
			testMappingFilterColumn: (filterColumn) => {
				var _filterColumn$customF12;
				if (((_filterColumn$customF12 = filterColumn.customFilters) === null || _filterColumn$customF12 === void 0 ? void 0 : _filterColumn$customF12.customFilters.length) !== 1) return false;
				const firstCustomFilter = filterColumn.customFilters.customFilters[0];
				if (firstCustomFilter.operator !== CustomFilterOperator.EQUAL) return false;
				return {
					operator1: CustomFilterOperator.EQUAL,
					val1: firstCustomFilter.val.toString()
				};
			}
		},
		_FilterConditionItems.NOT_EQUAL = {
			label: "sheets-filter-ui.conditions.not-equal",
			operator: CustomFilterOperator.NOT_EQUALS,
			numOfParameters: 1,
			order: 0,
			getDefaultFormParams: () => ({
				operator1: CustomFilterOperator.NOT_EQUALS,
				val1: ""
			}),
			testMappingParams: (params) => {
				const [op] = getOnlyOperatorAndVal(params);
				return op === CustomFilterOperator.NOT_EQUALS;
			},
			mapToFilterColumn: (mapParams) => ({ customFilters: { customFilters: [{
				val: mapParams.val1,
				operator: CustomFilterOperator.NOT_EQUALS
			}] } }),
			testMappingFilterColumn: (filterColumn) => {
				var _filterColumn$customF13;
				if (((_filterColumn$customF13 = filterColumn.customFilters) === null || _filterColumn$customF13 === void 0 ? void 0 : _filterColumn$customF13.customFilters.length) !== 1) return false;
				const firstCustomFilter = filterColumn.customFilters.customFilters[0];
				if (firstCustomFilter.operator !== CustomFilterOperator.NOT_EQUALS) return false;
				return {
					operator1: CustomFilterOperator.NOT_EQUALS,
					val1: firstCustomFilter.val.toString()
				};
			}
		},
		_FilterConditionItems.BETWEEN = {
			label: "sheets-filter-ui.conditions.between",
			operator: "between",
			order: 1,
			numOfParameters: 2,
			getDefaultFormParams: () => ({
				and: true,
				operator1: CustomFilterOperator.GREATER_THAN_OR_EQUAL,
				val1: "",
				operator2: CustomFilterOperator.LESS_THAN_OR_EQUAL,
				val2: ""
			}),
			testMappingParams: (params) => {
				const { and, operator1, operator2 } = params;
				if (!and) return false;
				const operators = [operator1, operator2];
				return operators.includes(CustomFilterOperator.GREATER_THAN_OR_EQUAL) && operators.includes(CustomFilterOperator.LESS_THAN_OR_EQUAL);
			},
			mapToFilterColumn: (mapParams) => {
				const { val1, val2, operator1 } = mapParams;
				const operator1IsGreater = operator1 === CustomFilterOperator.GREATER_THAN_OR_EQUAL;
				return { customFilters: {
					and: BooleanNumber.TRUE,
					customFilters: [{
						val: operator1IsGreater ? val1 : val2,
						operator: CustomFilterOperator.GREATER_THAN_OR_EQUAL
					}, {
						val: operator1IsGreater ? val2 : val1,
						operator: CustomFilterOperator.LESS_THAN_OR_EQUAL
					}]
				} };
			},
			testMappingFilterColumn: (filterColumn) => {
				var _filterColumn$customF14;
				if (((_filterColumn$customF14 = filterColumn.customFilters) === null || _filterColumn$customF14 === void 0 ? void 0 : _filterColumn$customF14.customFilters.length) !== 2) return false;
				const [firstCustomFilter, secondCustomFilter] = filterColumn.customFilters.customFilters;
				if (firstCustomFilter.operator === CustomFilterOperator.GREATER_THAN_OR_EQUAL && secondCustomFilter.operator === CustomFilterOperator.LESS_THAN_OR_EQUAL && filterColumn.customFilters.and) return {
					and: true,
					operator1: CustomFilterOperator.GREATER_THAN_OR_EQUAL,
					val1: firstCustomFilter.val.toString(),
					operator2: CustomFilterOperator.LESS_THAN_OR_EQUAL,
					val2: secondCustomFilter.val.toString()
				};
				if (secondCustomFilter.operator === CustomFilterOperator.GREATER_THAN_OR_EQUAL && firstCustomFilter.operator === CustomFilterOperator.LESS_THAN_OR_EQUAL && filterColumn.customFilters.and) return {
					and: true,
					operator1: CustomFilterOperator.GREATER_THAN_OR_EQUAL,
					val1: secondCustomFilter.val.toString(),
					operator2: CustomFilterOperator.LESS_THAN_OR_EQUAL,
					val2: firstCustomFilter.val.toLocaleString()
				};
				return false;
			}
		},
		_FilterConditionItems.NOT_BETWEEN = {
			label: "sheets-filter-ui.conditions.not-between",
			operator: "notBetween",
			order: 1,
			numOfParameters: 2,
			getDefaultFormParams: () => ({
				operator1: CustomFilterOperator.LESS_THAN,
				val1: "",
				operator2: CustomFilterOperator.GREATER_THAN,
				val2: ""
			}),
			testMappingParams: (params) => {
				const { and, operator1, operator2 } = params;
				if (and) return false;
				const operators = [operator1, operator2];
				return operators.includes(CustomFilterOperator.GREATER_THAN) && operators.includes(CustomFilterOperator.LESS_THAN);
			},
			mapToFilterColumn: (mapParams) => {
				const { val1, val2, operator1 } = mapParams;
				const operator1IsGreater = operator1 === CustomFilterOperator.GREATER_THAN;
				return { customFilters: { customFilters: [{
					val: operator1IsGreater ? val1 : val2,
					operator: CustomFilterOperator.GREATER_THAN
				}, {
					val: operator1IsGreater ? val2 : val1,
					operator: CustomFilterOperator.LESS_THAN
				}] } };
			},
			testMappingFilterColumn: (filterColumn) => {
				var _filterColumn$customF15;
				if (((_filterColumn$customF15 = filterColumn.customFilters) === null || _filterColumn$customF15 === void 0 ? void 0 : _filterColumn$customF15.customFilters.length) !== 2) return false;
				const [firstCustomFilter, secondCustomFilter] = filterColumn.customFilters.customFilters;
				if (firstCustomFilter.operator === CustomFilterOperator.LESS_THAN && secondCustomFilter.operator === CustomFilterOperator.GREATER_THAN && !filterColumn.customFilters.and) return {
					operator1: CustomFilterOperator.LESS_THAN,
					val1: firstCustomFilter.val.toString(),
					operator2: CustomFilterOperator.GREATER_THAN,
					val2: secondCustomFilter.val.toString()
				};
				if (secondCustomFilter.operator === CustomFilterOperator.LESS_THAN && firstCustomFilter.operator === CustomFilterOperator.GREATER_THAN && !filterColumn.customFilters.and) return {
					operator1: CustomFilterOperator.GREATER_THAN,
					val1: secondCustomFilter.val.toString(),
					operator2: CustomFilterOperator.LESS_THAN,
					val2: firstCustomFilter.val.toLocaleString()
				};
				return false;
			}
		},
		_FilterConditionItems.CUSTOM = {
			label: "sheets-filter-ui.conditions.custom",
			operator: "custom",
			order: 1,
			numOfParameters: 2,
			getDefaultFormParams: () => {
				return {
					operator1: "none",
					val1: "",
					operator2: "none",
					val2: ""
				};
			},
			testMappingParams: () => true,
			mapToFilterColumn: (mapParams) => {
				const { and, val1, val2, operator1, operator2 } = mapParams;
				function mapOperator(operator, val) {
					for (const condition of ALL_CONDITIONS) if (condition.operator === operator) return condition.mapToFilterColumn({
						val1: val,
						operator1: operator
					});
				}
				const operator1IsNone = !operator1 || operator1 === FilterConditionItems.NONE.operator;
				const operator2IsNone = !operator2 || operator2 === FilterConditionItems.NONE.operator;
				if (operator1IsNone && operator2IsNone) return NONE.mapToFilterColumn({});
				if (operator1IsNone) return mapOperator(operator2, val2);
				if (operator2IsNone) return mapOperator(operator1, val1);
				const mappedCustomFilter1 = mapOperator(operator1, val1);
				const mappedCustomFilter2 = mapOperator(operator2, val2);
				const customFilters = { customFilters: [mappedCustomFilter1.customFilters.customFilters[0], mappedCustomFilter2.customFilters.customFilters[0]] };
				if (and) customFilters.and = BooleanNumber.TRUE;
				return { customFilters };
			},
			testMappingFilterColumn: (filterColumn) => {
				var _filterColumn$customF16;
				if (((_filterColumn$customF16 = filterColumn.customFilters) === null || _filterColumn$customF16 === void 0 ? void 0 : _filterColumn$customF16.customFilters.length) !== 2) return false;
				const params = filterColumn.customFilters.customFilters.map((customFilter) => {
					return testMappingFilterColumn({ customFilters: { customFilters: [customFilter] } });
				});
				const result = {
					operator1: params[0][0].operator,
					val1: params[0][1].val1,
					operator2: params[1][0].operator,
					val2: params[1][1].val1
				};
				if (filterColumn.customFilters.and) result.and = true;
				return result;
			}
		}
	];
	function getItemByOperator(operator) {
		const item = ALL_CONDITIONS.find((condition) => condition.operator === operator);
		if (!item) throw new Error(`[SheetsFilter]: no condition item found for operator: ${operator}`);
		return item;
	}
	_FilterConditionItems.getItemByOperator = getItemByOperator;
	function testMappingParams(mapParams, numOfParameters) {
		for (const condition of ALL_CONDITIONS.filter((condition) => condition.numOfParameters === numOfParameters)) if (condition.numOfParameters !== 0 && condition.testMappingParams(mapParams)) return condition;
		for (const condition of ALL_CONDITIONS) if (condition.testMappingParams(mapParams)) return condition;
		throw new Error("[SheetsFilter]: no condition item can be mapped from the filter map params!");
	}
	_FilterConditionItems.testMappingParams = testMappingParams;
	function getInitialFormParams(operator) {
		const condition = ALL_CONDITIONS.find((condition) => condition.operator === operator);
		if ((condition === null || condition === void 0 ? void 0 : condition.numOfParameters) === 0) return { operator1: condition.operator };
		return condition.getDefaultFormParams();
	}
	_FilterConditionItems.getInitialFormParams = getInitialFormParams;
	function mapToFilterColumn(condition, mapParams) {
		return condition.mapToFilterColumn(mapParams);
	}
	_FilterConditionItems.mapToFilterColumn = mapToFilterColumn;
	function testMappingFilterColumn(filterColumn) {
		if (!filterColumn) return [NONE, {}];
		for (const condition of ALL_CONDITIONS) {
			const mapParams = condition.testMappingFilterColumn(filterColumn);
			if (mapParams) return [condition, mapParams];
		}
		return [NONE, {}];
	}
	_FilterConditionItems.testMappingFilterColumn = testMappingFilterColumn;
})(FilterConditionItems || (FilterConditionItems = {}));
function getOnlyOperatorAndVal(mapParams) {
	const { operator1, operator2, val1, val2 } = mapParams;
	if (operator1 && operator2) throw new Error("Both operator1 and operator2 are set!");
	if (!operator1 && !operator2) throw new Error("Neither operator1 and operator2 and both not set!");
	return operator1 ? [operator1, val1] : [operator2, val2];
}

//#endregion
//#region src/models/utils.ts
function statisticFilterByValueItems(items) {
	const checkedItems = [];
	const uncheckedItems = [];
	let checked = 0;
	let unchecked = 0;
	function traverse(node) {
		if (node.leaf) if (node.checked) {
			checkedItems.push(node);
			checked += node.count;
		} else {
			uncheckedItems.push(node);
			unchecked += node.count;
		}
		if (node.children) node.children.forEach(traverse);
	}
	items.forEach(traverse);
	return {
		checkedItems,
		uncheckedItems,
		checked,
		unchecked
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/worker/generate-filter-values.service.ts
const SHEETS_GENERATE_FILTER_VALUES_SERVICE_NAME = "sheets-filter.generate-filter-values.service";
const ISheetsGenerateFilterValuesService = createIdentifier(SHEETS_GENERATE_FILTER_VALUES_SERVICE_NAME);
let SheetsGenerateFilterValuesService = class SheetsGenerateFilterValuesService extends Disposable {
	constructor(_localeService, _univerInstanceService, _logService) {
		super();
		this._localeService = _localeService;
		this._univerInstanceService = _univerInstanceService;
		this._logService = _logService;
	}
	async getFilterValues(params) {
		var _this$_univerInstance;
		const { unitId, subUnitId, filteredOutRowsByOtherColumns, filterColumn, filters, blankChecked, iterateRange, alreadyChecked } = params;
		const workbook = this._univerInstanceService.getUnit(unitId);
		const worksheet = (_this$_univerInstance = this._univerInstanceService.getUnit(unitId)) === null || _this$_univerInstance === void 0 ? void 0 : _this$_univerInstance.getSheetBySheetId(subUnitId);
		if (!workbook || !worksheet) return [];
		this._logService.debug("[SheetsGenerateFilterValuesService]", "getFilterValues for", {
			unitId,
			subUnitId
		});
		return getFilterTreeByValueItems(filters, this._localeService, iterateRange, worksheet, new Set(filteredOutRowsByOtherColumns), filterColumn, new Set(alreadyChecked.map(String)), blankChecked, workbook.getStyles());
	}
};
SheetsGenerateFilterValuesService = __decorate([
	__decorateParam(0, Inject(LocaleService)),
	__decorateParam(1, IUniverInstanceService),
	__decorateParam(2, ILogService)
], SheetsGenerateFilterValuesService);
function getFilterTreeByValueItems(filters, localeService, iterateRange, worksheet, filteredOutRowsByOtherColumns, filterColumn, alreadyChecked, blankChecked, styles) {
	var _filterColumn$filtere;
	const items = /* @__PURE__ */ new Map();
	const treeMap = /* @__PURE__ */ new Map();
	const DefaultPattern = "yyyy-mm-dd";
	const EmptyKey = "empty";
	const isNeedClearCheckedStatus = !filters && ((filterColumn === null || filterColumn === void 0 ? void 0 : filterColumn.filterBy) === FilterBy.COLORS || (filterColumn === null || filterColumn === void 0 ? void 0 : filterColumn.filterBy) === FilterBy.CONDITIONS) && ((_filterColumn$filtere = filterColumn.filteredOutRows) === null || _filterColumn$filtere === void 0 ? void 0 : _filterColumn$filtere.size);
	let emptyCount = 0;
	for (const cell of worksheet.iterateByColumn(iterateRange, false, false)) {
		const { row, rowSpan = 1 } = cell;
		let rowIndex = 0;
		while (rowIndex < rowSpan) {
			var _cell$value, _styles$get, _cell$value2;
			const targetRow = row + rowIndex;
			if (filteredOutRowsByOtherColumns.has(targetRow)) {
				rowIndex++;
				continue;
			}
			const value = (cell === null || cell === void 0 ? void 0 : cell.value) ? extractPureTextFromCell(cell.value) : "";
			if (!value) {
				emptyCount += 1;
				rowIndex += rowSpan;
				continue;
			}
			const fmtStr = ((_cell$value = cell.value) === null || _cell$value === void 0 ? void 0 : _cell$value.v) && !cell.value.p ? (_styles$get = styles.get((_cell$value2 = cell.value) === null || _cell$value2 === void 0 ? void 0 : _cell$value2.s)) === null || _styles$get === void 0 || (_styles$get = _styles$get.n) === null || _styles$get === void 0 ? void 0 : _styles$get.pattern : "";
			const isDateValue = fmtStr && numfmt.getFormatInfo(fmtStr).isDate;
			let isIncludeDatePart = false;
			if (isDateValue) {
				const { year, month, day } = numfmt.getFormatDateInfo(fmtStr);
				isIncludeDatePart = year || month || day;
			}
			if (fmtStr && isDateValue && isIncludeDatePart) {
				var _worksheet$getCellRaw, _yearItem$children, _monthItem$children;
				const originValue = (_worksheet$getCellRaw = worksheet.getCellRaw(cell.row, cell.col)) === null || _worksheet$getCellRaw === void 0 ? void 0 : _worksheet$getCellRaw.v;
				if (!originValue) {
					rowIndex++;
					continue;
				}
				const [year, month, day] = numfmt.format(DefaultPattern, Number(originValue)).split("-").map(Number);
				let yearItem = items.get(`${year}`);
				if (!yearItem) {
					yearItem = {
						title: `${year}`,
						key: `${year}`,
						children: [],
						count: 0,
						leaf: false,
						checked: false
					};
					items.set(`${year}`, yearItem);
					treeMap.set(`${year}`, [`${year}`]);
				}
				let monthItem = (_yearItem$children = yearItem.children) === null || _yearItem$children === void 0 ? void 0 : _yearItem$children.find((item) => item.key === `${year}-${month}`);
				if (!monthItem) {
					var _yearItem$children2;
					monthItem = {
						title: localeService.t(`sheets-filter-ui.date.${month}`),
						key: `${year}-${month}`,
						children: [],
						count: 0,
						leaf: false,
						checked: false
					};
					(_yearItem$children2 = yearItem.children) === null || _yearItem$children2 === void 0 || _yearItem$children2.push(monthItem);
					treeMap.set(`${year}-${month}`, [`${year}`, `${year}-${month}`]);
				}
				const dayItem = monthItem === null || monthItem === void 0 || (_monthItem$children = monthItem.children) === null || _monthItem$children === void 0 ? void 0 : _monthItem$children.find((item) => item.key === `${year}-${month}-${day}`);
				if (!dayItem) {
					var _monthItem$children2;
					(_monthItem$children2 = monthItem.children) === null || _monthItem$children2 === void 0 || _monthItem$children2.push({
						title: `${day}`,
						key: `${year}-${month}-${day}`,
						count: 1,
						originValues: new Set([value]),
						leaf: true,
						checked: isNeedClearCheckedStatus ? false : alreadyChecked.size ? alreadyChecked.has(value) : !blankChecked
					});
					monthItem.count++;
					yearItem.count++;
					treeMap.set(`${year}-${month}-${day}`, [
						`${year}`,
						`${year}-${month}`,
						`${year}-${month}-${day}`
					]);
				} else {
					dayItem.originValues.add(value);
					dayItem.count++;
					monthItem.count++;
					yearItem.count++;
				}
			} else {
				const key = value;
				let item = items.get(key);
				if (!item) {
					item = {
						title: value,
						leaf: true,
						checked: isNeedClearCheckedStatus ? false : alreadyChecked.size ? alreadyChecked.has(value) : !blankChecked,
						key,
						count: 1
					};
					items.set(key, item);
					treeMap.set(key, [key]);
				} else item.count++;
			}
			rowIndex++;
		}
	}
	const initialBlankChecked = isNeedClearCheckedStatus ? false : filters ? blankChecked : true;
	if (emptyCount > 0) {
		const item = {
			title: localeService.t("sheets-filter-ui.panel.empty"),
			count: emptyCount,
			leaf: true,
			checked: initialBlankChecked,
			key: EmptyKey
		};
		items.set("empty", item);
		treeMap.set("empty", [EmptyKey]);
	}
	return {
		filterTreeItems: generateFilterTreeBySort(Array.from(items.values())),
		filterTreeMapCache: treeMap
	};
}
function generateFilterTreeBySort(tree) {
	return Array.from(tree).sort((a, b) => {
		if (a.children && !b.children) return -1;
		if (!a.children && b.children) return 1;
		return compare(a.title, b.title);
	}).map((yearItem) => {
		if (yearItem.children) yearItem.children.sort((a, b) => {
			return Number.parseInt(a.key.split("-")[1], 10) - Number.parseInt(b.key.split("-")[1], 10);
		}).forEach((monthItem) => {
			if (monthItem.children) monthItem.children.sort((a, b) => {
				return Number.parseInt(a.key.split("-")[2], 10) - Number.parseInt(b.key.split("-")[2], 10);
			});
		});
		return yearItem;
	});
}
function compare(strA, strB) {
	const aIsNumeric = isNumeric(strA);
	const bIsNumeric = isNumeric(strB);
	if (aIsNumeric && bIsNumeric) return Number.parseFloat(strA) - Number.parseFloat(strB);
	else if (aIsNumeric && !bIsNumeric) return -1;
	else if (!aIsNumeric && bIsNumeric) return 1;
	else return strA.localeCompare(strB);
}

//#endregion
//#region src/services/util.ts
function findObjectByKey(data, targetKey) {
	for (const node of data) {
		if (node.key === targetKey) return node;
		if (node.children) {
			const result = findObjectByKey(node.children, targetKey);
			if (result) return result;
		}
	}
	return null;
}
function areAllLeafNodesChecked(node) {
	if (node.leaf) return node.checked;
	return node.children ? node.children.every((child) => areAllLeafNodesChecked(child)) : true;
}
function updateLeafNodesCheckedStatus(node, status) {
	if (node.leaf) if (status !== void 0) node.checked = status;
	else node.checked = !node.checked;
	if (node.children) node.children.forEach((child) => updateLeafNodesCheckedStatus(child, status));
}
function searchTree(items, searchKeywords) {
	const result = [];
	items.forEach((item) => {
		const originMatches = item.originValues ? searchKeywords.some((keyword) => Array.from(item.originValues).some((value) => value.toLowerCase().includes(keyword.toLowerCase()))) : false;
		const titleMatches = !originMatches && searchKeywords.some((keyword) => item.title.toLowerCase().includes(keyword.toLowerCase()));
		if (originMatches || titleMatches) result.push({ ...item });
		else if (item.children) {
			const filteredChildren = searchTree(item.children, searchKeywords);
			if (filteredChildren.length > 0) {
				const aggregatedCount = filteredChildren.reduce((sum, child) => sum + child.count, 0);
				result.push({
					...item,
					count: aggregatedCount,
					children: filteredChildren
				});
			}
		}
	});
	return result;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region src/services/sheets-filter-panel.service.ts
var _ByConditionsModel, _ByValuesModel, _ByColorsModel;
const ISheetsFilterPanelService = createIdentifier("sheets-filter-ui.sheets-filter-panel.service");
let SheetsFilterPanelService = class SheetsFilterPanelService extends Disposable {
	get filterBy() {
		return this._filterBy$.getValue();
	}
	get filterByModel() {
		return this._filterByModel;
	}
	set filterByModel(model) {
		this._filterByModel = model;
		this._filterByModel$.next(model);
	}
	get filterModel() {
		return this._filterModel;
	}
	get col() {
		return this._col$.getValue();
	}
	constructor(_injector, _refRangeService) {
		super();
		this._injector = _injector;
		this._refRangeService = _refRangeService;
		_defineProperty(this, "_filterBy$", new BehaviorSubject(FilterBy.VALUES));
		_defineProperty(this, "filterBy$", this._filterBy$.asObservable());
		_defineProperty(this, "_filterByModel$", new ReplaySubject(1));
		_defineProperty(this, "filterByModel$", this._filterByModel$.asObservable());
		_defineProperty(this, "_filterByModel", null);
		_defineProperty(this, "_hasCriteria$", new BehaviorSubject(false));
		_defineProperty(this, "hasCriteria$", this._hasCriteria$.asObservable());
		_defineProperty(this, "_filterModel", null);
		_defineProperty(this, "_col$", new BehaviorSubject(-1));
		_defineProperty(this, "col$", this._col$.asObservable());
		_defineProperty(this, "_filterHeaderListener", null);
	}
	dispose() {
		this._filterBy$.complete();
		this._filterByModel$.complete();
		this._hasCriteria$.complete();
	}
	setupCol(filterModel, col) {
		this.terminate();
		this._filterModel = filterModel;
		this._col$.next(col);
		const filterColumn = filterModel.getFilterColumn(col);
		if (filterColumn) {
			const info = filterColumn.getColumnData();
			if (info.customFilters) {
				this._hasCriteria$.next(true);
				this._setupByConditions(filterModel, col);
				return;
			}
			if (info.colorFilters) {
				this._hasCriteria$.next(true);
				this._setupByColors(filterModel, col);
				return;
			}
			if (info.filters) {
				this._hasCriteria$.next(true);
				this._setupByValues(filterModel, col);
				return;
			}
			this._hasCriteria$.next(false);
			this._setupByValues(filterModel, col);
			return;
		}
		this._hasCriteria$.next(false);
		this._setupByValues(filterModel, col);
	}
	changeFilterBy(filterBy) {
		if (!this._filterModel || this.col === -1) return false;
		switch (filterBy) {
			case FilterBy.VALUES:
				this._setupByValues(this._filterModel, this.col);
				break;
			case FilterBy.COLORS:
				this._setupByColors(this._filterModel, this.col);
				break;
			case FilterBy.CONDITIONS:
				this._setupByConditions(this._filterModel, this.col);
				break;
		}
		return true;
	}
	terminate() {
		this._filterModel = null;
		this._col$.next(-1);
		this._disposeFilterHeaderChangeListener();
		return true;
	}
	_disposeFilterHeaderChangeListener() {
		var _this$_filterHeaderLi;
		(_this$_filterHeaderLi = this._filterHeaderListener) === null || _this$_filterHeaderLi === void 0 || _this$_filterHeaderLi.dispose();
		this._filterHeaderListener = null;
	}
	_listenToFilterHeaderChange(filterModel, col) {
		this._disposeFilterHeaderChangeListener();
		const unitId = filterModel.unitId;
		const subUnitId = filterModel.subUnitId;
		const filterRange = filterModel.getRange();
		const columnHeaderRange = {
			startColumn: col,
			startRow: filterRange.startRow,
			endRow: filterRange.startRow,
			endColumn: col
		};
		this._filterHeaderListener = this._refRangeService.watchRange(unitId, subUnitId, columnHeaderRange, (before, after) => {
			if (!after) this.terminate();
			else {
				const offset = after.startColumn - before.startColumn;
				if (offset !== 0) this._filterByModel.deltaCol(offset);
			}
		});
	}
	async _setupByValues(filterModel, col) {
		this._disposePreviousModel();
		const range = filterModel.getRange();
		if (range.startRow === range.endRow) return false;
		const filterByModel = await ByValuesModel.fromFilterColumn(this._injector, filterModel, col);
		this.filterByModel = filterByModel;
		this._filterBy$.next(FilterBy.VALUES);
		this._listenToFilterHeaderChange(filterModel, col);
		return true;
	}
	async _setupByColors(filterModel, col) {
		this._disposePreviousModel();
		const range = filterModel.getRange();
		if (range.startRow === range.endRow) return false;
		const filterByModel = await ByColorsModel.fromFilterColumn(this._injector, filterModel, col);
		this.filterByModel = filterByModel;
		this._filterBy$.next(FilterBy.COLORS);
		this._listenToFilterHeaderChange(filterModel, col);
		return true;
	}
	_setupByConditions(filterModel, col) {
		this._disposePreviousModel();
		const range = filterModel.getRange();
		if (range.startRow === range.endRow) return false;
		const model = ByConditionsModel.fromFilterColumn(this._injector, filterModel, col, filterModel.getFilterColumn(col));
		this.filterByModel = model;
		this._filterBy$.next(FilterBy.CONDITIONS);
		this._listenToFilterHeaderChange(filterModel, col);
		return true;
	}
	_disposePreviousModel() {
		var _this$_filterByModel;
		(_this$_filterByModel = this._filterByModel) === null || _this$_filterByModel === void 0 || _this$_filterByModel.dispose();
		this.filterByModel = null;
	}
};
SheetsFilterPanelService = __decorate([__decorateParam(0, Inject(Injector)), __decorateParam(1, Inject(RefRangeService))], SheetsFilterPanelService);
let ByConditionsModel = _ByConditionsModel = class ByConditionsModel extends Disposable {
	/**
	* Create a model with targeting filter column. If there is not a filter column, the model would be created with
	* default values.
	*
	* @param injector
	* @param filterModel
	* @param col
	* @param filterColumn
	*
	* @returns the model to control the panel's state
	*/
	static fromFilterColumn(injector, filterModel, col, filterColumn) {
		const [conditionItem, conditionParams] = FilterConditionItems.testMappingFilterColumn(filterColumn === null || filterColumn === void 0 ? void 0 : filterColumn.getColumnData());
		return injector.createInstance(_ByConditionsModel, filterModel, col, conditionItem, conditionParams);
	}
	get conditionItem() {
		return this._conditionItem$.getValue();
	}
	get filterConditionFormParams() {
		return this._filterConditionFormParams$.getValue();
	}
	constructor(_filterModel, col, conditionItem, conditionParams, _commandService) {
		super();
		this._filterModel = _filterModel;
		this.col = col;
		this._commandService = _commandService;
		_defineProperty(this, "canApply$", of(true));
		_defineProperty(this, "_conditionItem$", void 0);
		_defineProperty(this, "conditionItem$", void 0);
		_defineProperty(this, "_filterConditionFormParams$", void 0);
		_defineProperty(this, "filterConditionFormParams$", void 0);
		this._conditionItem$ = new BehaviorSubject(conditionItem);
		this.conditionItem$ = this._conditionItem$.asObservable();
		this._filterConditionFormParams$ = new BehaviorSubject(conditionParams);
		this.filterConditionFormParams$ = this._filterConditionFormParams$.asObservable();
	}
	dispose() {
		super.dispose();
		this._conditionItem$.complete();
		this._filterConditionFormParams$.complete();
	}
	deltaCol(offset) {
		this.col += offset;
	}
	clear() {
		if (this._disposed) return Promise.resolve(false);
		return this._commandService.executeCommand(SetSheetsFilterCriteriaCommand.id, {
			unitId: this._filterModel.unitId,
			subUnitId: this._filterModel.subUnitId,
			col: this.col,
			criteria: null
		});
	}
	/**
	* Apply the filter condition to the target filter column.
	*/
	async apply() {
		if (this._disposed) return false;
		const filterColumn = FilterConditionItems.mapToFilterColumn(this.conditionItem, this.filterConditionFormParams);
		return this._commandService.executeCommand(SetSheetsFilterCriteriaCommand.id, {
			unitId: this._filterModel.unitId,
			subUnitId: this._filterModel.subUnitId,
			col: this.col,
			criteria: filterColumn
		});
	}
	/**
	* This method would be called when user changes the primary condition. The model would load the corresponding
	* `IFilterConditionFormParams` and load default condition form params.
	*/
	onPrimaryConditionChange(operator) {
		const conditionItem = FilterConditionItems.ALL_CONDITIONS.find((item) => item.operator === operator);
		if (!conditionItem) throw new Error(`[ByConditionsModel]: condition item not found for operator: ${operator}!`);
		this._conditionItem$.next(conditionItem);
		this._filterConditionFormParams$.next(FilterConditionItems.getInitialFormParams(operator));
	}
	/**
	* This method would be called when user changes the primary conditions, the input values or "AND" "OR" ratio.
	* If the primary conditions or the ratio is changed, the method would load the corresponding `IFilterCondition`.
	*
	* When the panel call this method, it only has to pass the changed keys.
	*
	* @param params
	*/
	onConditionFormChange(params) {
		const newParams = {
			...this.filterConditionFormParams,
			...params
		};
		if (newParams.and !== true) delete newParams.and;
		if (typeof params.and !== "undefined" || typeof params.operator1 !== "undefined" || typeof params.operator2 !== "undefined") {
			const conditionItem = FilterConditionItems.testMappingParams(newParams, this.conditionItem.numOfParameters);
			this._conditionItem$.next(conditionItem);
		}
		this._filterConditionFormParams$.next(newParams);
	}
};
ByConditionsModel = _ByConditionsModel = __decorate([__decorateParam(4, ICommandService)], ByConditionsModel);
let ByValuesModel = _ByValuesModel = class ByValuesModel extends Disposable {
	/**
	* Create a model with targeting filter column. If there is not a filter column, the model would be created with
	* default values.
	*
	* @param injector
	* @param filterModel
	* @param col
	*
	* @returns the model to control the panel's state
	*/
	static async fromFilterColumn(injector, filterModel, col) {
		const univerInstanceService = injector.get(IUniverInstanceService);
		const localeService = injector.get(LocaleService);
		const generateFilterValuesService = injector.get(ISheetsGenerateFilterValuesService, Quantity.OPTIONAL);
		const { unitId, subUnitId } = filterModel;
		const workbook = univerInstanceService.getUniverSheetInstance(unitId);
		if (!workbook) throw new Error(`[ByValuesModel]: Workbook not found for filter model with unitId: ${unitId}!`);
		const worksheet = workbook === null || workbook === void 0 ? void 0 : workbook.getSheetBySheetId(subUnitId);
		if (!worksheet) throw new Error(`[ByValuesModel]: Worksheet not found for filter model with unitId: ${unitId} and subUnitId: ${subUnitId}!`);
		const range = filterModel.getRange();
		const column = col;
		const filterColumn = filterModel.getFilterColumn(col);
		const filters = filterColumn === null || filterColumn === void 0 ? void 0 : filterColumn.getColumnData().filters;
		const alreadyChecked = new Set(filters === null || filters === void 0 ? void 0 : filters.filters);
		const blankChecked = !!(filters && filters.blank);
		const filteredOutRowsByOtherColumns = filterModel.getFilteredOutRowsExceptCol(col);
		const iterateRange = {
			...range,
			startRow: range.startRow + 1,
			startColumn: column,
			endColumn: column
		};
		let items;
		let cache;
		if (generateFilterValuesService) {
			const res = await generateFilterValuesService.getFilterValues({
				unitId,
				subUnitId,
				filteredOutRowsByOtherColumns: Array.from(filteredOutRowsByOtherColumns),
				filterColumn,
				filters: !!filters,
				blankChecked,
				iterateRange,
				alreadyChecked: Array.from(alreadyChecked)
			});
			items = res.filterTreeItems;
			cache = res.filterTreeMapCache;
		} else {
			const res = getFilterTreeByValueItems(!!filters, localeService, iterateRange, worksheet, filteredOutRowsByOtherColumns, filterColumn, alreadyChecked, blankChecked, workbook.getStyles());
			items = res.filterTreeItems;
			cache = res.filterTreeMapCache;
		}
		return injector.createInstance(_ByValuesModel, filterModel, col, items, cache);
	}
	get rawFilterItems() {
		return this._rawFilterItems$.getValue();
	}
	get filterItems() {
		return this._filterItems;
	}
	get treeMapCache() {
		return this._treeMapCache;
	}
	constructor(_filterModel, col, items, cache, _commandService) {
		super();
		this._filterModel = _filterModel;
		this.col = col;
		this._commandService = _commandService;
		_defineProperty(this, "_rawFilterItems$", void 0);
		_defineProperty(this, "rawFilterItems$", void 0);
		_defineProperty(this, "filterItems$", void 0);
		_defineProperty(this, "_filterItems", []);
		_defineProperty(this, "_treeMapCache", void 0);
		_defineProperty(this, "canApply$", void 0);
		_defineProperty(this, "_manuallyUpdateFilterItems$", void 0);
		_defineProperty(this, "_searchString$", void 0);
		_defineProperty(this, "searchString$", void 0);
		this._treeMapCache = cache;
		this._searchString$ = new BehaviorSubject("");
		this.searchString$ = this._searchString$.asObservable();
		this._rawFilterItems$ = new BehaviorSubject(items);
		this.rawFilterItems$ = this._rawFilterItems$.asObservable();
		this._manuallyUpdateFilterItems$ = new Subject();
		this.filterItems$ = merge$1(combineLatest([this._searchString$.pipe(throttleTime(500, void 0, {
			leading: true,
			trailing: true
		}), startWith(void 0)), this._rawFilterItems$]).pipe(map(([searchString, items]) => {
			if (!searchString) return items;
			return searchTree(items, searchString.toLowerCase().split(/\s+/).filter((s) => !!s));
		})), this._manuallyUpdateFilterItems$).pipe(shareReplay({
			bufferSize: 1,
			refCount: true
		}));
		this.canApply$ = this.filterItems$.pipe(map((items) => {
			return statisticFilterByValueItems(items).checked > 0;
		}));
		this.disposeWithMe(this.filterItems$.subscribe((items) => this._filterItems = items));
	}
	dispose() {
		this._rawFilterItems$.complete();
		this._searchString$.complete();
	}
	deltaCol(offset) {
		this.col += offset;
	}
	setSearchString(str) {
		this._searchString$.next(str);
	}
	onCheckAllToggled(checked) {
		const items = Tools.deepClone(this._filterItems);
		items.forEach((item) => updateLeafNodesCheckedStatus(item, checked));
		this._manuallyUpdateFilterItems(items);
	}
	/**
	* Toggle a filter item.
	*/
	onFilterCheckToggled(item) {
		const items = Tools.deepClone(this._filterItems);
		const changedItem = findObjectByKey(items, item.key);
		if (!changedItem) return;
		updateLeafNodesCheckedStatus(changedItem, !areAllLeafNodesChecked(changedItem));
		this._manuallyUpdateFilterItems(items);
	}
	onFilterOnly(itemKeys) {
		const items = Tools.deepClone(this._filterItems);
		items.forEach((item) => updateLeafNodesCheckedStatus(item, false));
		itemKeys.forEach((key) => {
			const changedItem = findObjectByKey(items, key);
			if (changedItem) updateLeafNodesCheckedStatus(changedItem, true);
		});
		this._manuallyUpdateFilterItems(items);
	}
	_manuallyUpdateFilterItems(items) {
		this._manuallyUpdateFilterItems$.next(items);
	}
	clear() {
		if (this._disposed) return Promise.resolve(false);
		return this._commandService.executeCommand(SetSheetsFilterCriteriaCommand.id, {
			unitId: this._filterModel.unitId,
			subUnitId: this._filterModel.subUnitId,
			col: this.col,
			criteria: null
		});
	}
	/**
	* Apply the filter condition to the target filter column.
	*/
	async apply() {
		if (this._disposed) return false;
		const statistics = statisticFilterByValueItems(this._filterItems);
		const { checked, checkedItems } = statistics;
		const rawFilterItems = this.rawFilterItems;
		let rawFilterCount = 0;
		for (const item of rawFilterItems) rawFilterCount += item.count;
		const noChecked = checked === 0;
		const allChecked = statistics.checked === rawFilterCount;
		const criteria = { colId: this.col };
		if (noChecked) throw new Error("[ByValuesModel]: no checked items!");
		else if (allChecked) return this._commandService.executeCommand(SetSheetsFilterCriteriaCommand.id, {
			unitId: this._filterModel.unitId,
			subUnitId: this._filterModel.subUnitId,
			col: this.col,
			criteria: null
		});
		else {
			criteria.filters = {};
			const nonEmptyItems = checkedItems.filter((item) => item.key !== "empty");
			if (nonEmptyItems.length > 0) criteria.filters = { filters: nonEmptyItems.flatMap((item) => item.originValues ? Array.from(item.originValues) : [item.title]) };
			if (nonEmptyItems.length !== checkedItems.length) criteria.filters.blank = true;
		}
		return this._commandService.executeCommand(SetSheetsFilterCriteriaCommand.id, {
			unitId: this._filterModel.unitId,
			subUnitId: this._filterModel.subUnitId,
			col: this.col,
			criteria
		});
	}
};
ByValuesModel = _ByValuesModel = __decorate([__decorateParam(4, ICommandService)], ByValuesModel);
let ByColorsModel = _ByColorsModel = class ByColorsModel extends Disposable {
	/**
	* Create a model with targeting filter column. If there is not a filter column, the model would be created with
	* default values.
	*
	* @param injector
	* @param filterModel
	* @param col
	*
	* @returns the model to control the panel's state
	*/
	static async fromFilterColumn(injector, filterModel, col) {
		var _filterModel$getFilte, _colorFilters$cellFil, _colorFilters$cellTex;
		const univerInstanceService = injector.get(IUniverInstanceService);
		const { unitId, subUnitId } = filterModel;
		const workbook = univerInstanceService.getUniverSheetInstance(unitId);
		if (!workbook) throw new Error(`[ByColorsModel]: Workbook not found for filter model with unitId: ${unitId}!`);
		const worksheet = workbook === null || workbook === void 0 ? void 0 : workbook.getSheetBySheetId(subUnitId);
		if (!worksheet) throw new Error(`[ByColorsModel]: Worksheet not found for filter model with unitId: ${unitId} and subUnitId: ${subUnitId}!`);
		const range = filterModel.getRange();
		const column = col;
		const colorFilters = (_filterModel$getFilte = filterModel.getFilterColumn(col)) === null || _filterModel$getFilte === void 0 ? void 0 : _filterModel$getFilte.getColumnData().colorFilters;
		const filteredOutRowsByOtherColumns = filterModel.getFilteredOutRowsExceptCol(col);
		const iterateRange = {
			...range,
			startRow: range.startRow + 1,
			startColumn: column,
			endColumn: column
		};
		const cellFillColors = /* @__PURE__ */ new Map();
		const cellFillColorsChecked = new Set((_colorFilters$cellFil = colorFilters === null || colorFilters === void 0 ? void 0 : colorFilters.cellFillColors) !== null && _colorFilters$cellFil !== void 0 ? _colorFilters$cellFil : []);
		const cellTextColors = /* @__PURE__ */ new Map();
		const cellTextColorsChecked = new Set((_colorFilters$cellTex = colorFilters === null || colorFilters === void 0 ? void 0 : colorFilters.cellTextColors) !== null && _colorFilters$cellTex !== void 0 ? _colorFilters$cellTex : []);
		for (const cell of worksheet.iterateByColumn(iterateRange, false, true)) {
			const { row, col, value } = cell;
			if (filteredOutRowsByOtherColumns.has(row)) continue;
			const style = worksheet.getComposedCellStyleByCellData(row, col, value);
			if (style.bg && style.bg.rgb) {
				const bg = new ColorKit(style.bg.rgb).toRgbString();
				if (!cellFillColors.has(bg)) cellFillColors.set(bg, {
					color: bg,
					checked: cellFillColorsChecked.has(bg)
				});
			} else cellFillColors.set("default-fill-color", {
				color: null,
				checked: cellFillColorsChecked.has(null)
			});
			if (style.cl && style.cl.rgb) {
				const cl = new ColorKit(style.cl.rgb).toRgbString();
				if (!cellTextColors.has(cl)) cellTextColors.set(cl, {
					color: cl,
					checked: cellTextColorsChecked.has(cl)
				});
			} else cellTextColors.set("default-font-color", {
				color: COLOR_BLACK_RGB,
				checked: cellTextColorsChecked.has(COLOR_BLACK_RGB)
			});
		}
		return injector.createInstance(_ByColorsModel, filterModel, col, cellFillColors, cellTextColors);
	}
	get cellFillColors() {
		return this._cellFillColors$.getValue();
	}
	get cellTextColors() {
		return this._cellTextColors$.getValue();
	}
	constructor(_filterModel, col, cellFillColors, cellTextColors, _commandService) {
		super();
		this._filterModel = _filterModel;
		this.col = col;
		this._commandService = _commandService;
		_defineProperty(this, "canApply$", of(true));
		_defineProperty(this, "_cellFillColors$", void 0);
		_defineProperty(this, "cellFillColors$", void 0);
		_defineProperty(this, "_cellTextColors$", void 0);
		_defineProperty(this, "cellTextColors$", void 0);
		this._cellFillColors$ = new BehaviorSubject(Array.from(cellFillColors.values()));
		this.cellFillColors$ = this._cellFillColors$.asObservable();
		this._cellTextColors$ = new BehaviorSubject(Array.from(cellTextColors.values()));
		this.cellTextColors$ = this._cellTextColors$.asObservable();
	}
	dispose() {
		super.dispose();
		this._cellFillColors$.complete();
	}
	deltaCol(offset) {
		this.col += offset;
	}
	clear() {
		if (this._disposed) return Promise.resolve(false);
		return this._commandService.executeCommand(SetSheetsFilterCriteriaCommand.id, {
			unitId: this._filterModel.unitId,
			subUnitId: this._filterModel.subUnitId,
			col: this.col,
			criteria: null
		});
	}
	onFilterCheckToggled(item, isFillColor = true) {
		const colors = isFillColor ? this.cellFillColors : this.cellTextColors;
		const items = [];
		let found = false;
		for (let i = 0; i < colors.length; i++) {
			const colorItem = colors[i];
			if (colorItem.color === item.color) {
				found = true;
				items.push({
					color: colorItem.color,
					checked: !colorItem.checked
				});
				continue;
			}
			items.push({
				color: colorItem.color,
				checked: colorItem.checked
			});
		}
		if (!found) return;
		this._resetColorsCheckedStatus(!isFillColor);
		if (isFillColor) this._cellFillColors$.next([...items]);
		else this._cellTextColors$.next([...items]);
	}
	_resetColorsCheckedStatus(isFillColor = true) {
		const colors = isFillColor ? this.cellFillColors : this.cellTextColors;
		const items = [];
		for (let i = 0; i < colors.length; i++) items.push({
			color: colors[i].color,
			checked: false
		});
		if (isFillColor) this._cellFillColors$.next([...items]);
		else this._cellTextColors$.next([...items]);
	}
	/**
	* Apply the filter condition to the target filter column.
	*/
	async apply() {
		if (this._disposed) return false;
		const cellFillColorsChecked = this.cellFillColors.filter((item) => item.checked).map((item) => item.color);
		const cellTextColorsChecked = this.cellTextColors.filter((item) => item.checked).map((item) => item.color);
		if (cellFillColorsChecked.length === 0 && cellTextColorsChecked.length === 0) return this._commandService.executeCommand(SetSheetsFilterCriteriaCommand.id, {
			unitId: this._filterModel.unitId,
			subUnitId: this._filterModel.subUnitId,
			col: this.col,
			criteria: null
		});
		const criteria = { colId: this.col };
		if (cellFillColorsChecked.length > 0) criteria.colorFilters = { cellFillColors: cellFillColorsChecked };
		else if (cellTextColorsChecked.length > 0) criteria.colorFilters = { cellTextColors: cellTextColorsChecked };
		return this._commandService.executeCommand(SetSheetsFilterCriteriaCommand.id, {
			unitId: this._filterModel.unitId,
			subUnitId: this._filterModel.subUnitId,
			col: this.col,
			criteria
		});
	}
};
ByColorsModel = _ByColorsModel = __decorate([__decorateParam(4, ICommandService)], ByColorsModel);

//#endregion
//#region src/commands/operations/sheets-filter.operation.ts
const FILTER_PANEL_OPENED_KEY = "FILTER_PANEL_OPENED";
/**
* The operation to open the filter panel and prepare for changing the filter conditions on a given column.
*/
const OpenFilterPanelOperation = {
	id: "sheet.operation.open-filter-panel",
	type: CommandType.OPERATION,
	handler: (accessor, params) => {
		const contextService = accessor.get(IContextService);
		const sheetsFilterService = accessor.get(SheetsFilterService);
		const sheetsFilterPanelService = accessor.get(SheetsFilterPanelService);
		const commandService = accessor.get(ICommandService);
		const editorBridgeService = accessor.has(IEditorBridgeService) ? accessor.get(IEditorBridgeService) : null;
		if (editorBridgeService === null || editorBridgeService === void 0 ? void 0 : editorBridgeService.isVisible().visible) commandService.syncExecuteCommand(SetCellEditVisibleOperation.id, { visible: false });
		const { unitId, subUnitId, col } = params;
		const filterModel = sheetsFilterService.getFilterModel(unitId, subUnitId);
		if (!filterModel) return false;
		sheetsFilterPanelService.setupCol(filterModel, col);
		if (!contextService.getContextValue("FILTER_PANEL_OPENED")) contextService.setContextValue(FILTER_PANEL_OPENED_KEY, true);
		return true;
	}
};
const CloseFilterPanelOperation = {
	id: "sheet.operation.close-filter-panel",
	type: CommandType.OPERATION,
	handler: (accessor) => {
		const contextService = accessor.get(IContextService);
		const sheetsFilterPanelService = accessor.get(SheetsFilterPanelService);
		const layoutService = accessor.get(ILayoutService, Quantity.OPTIONAL);
		if (contextService.getContextValue("FILTER_PANEL_OPENED")) {
			contextService.setContextValue(FILTER_PANEL_OPENED_KEY, false);
			layoutService === null || layoutService === void 0 || layoutService.focus();
			return sheetsFilterPanelService.terminate();
		}
		return false;
	}
};
const ChangeFilterByOperation = {
	id: "sheet.operation.apply-filter",
	type: CommandType.OPERATION,
	handler: (accessor, params) => {
		const { filterBy } = params;
		return accessor.get(SheetsFilterPanelService).changeFilterBy(filterBy);
	}
};

//#endregion
//#region package.json
var name = "@univerjs/sheets-filter-ui";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
const SHEETS_FILTER_UI_PLUGIN_CONFIG_KEY = "sheets-filter-ui.config";
const configSymbol = Symbol(SHEETS_FILTER_UI_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/controllers/sheets-filter-permission.controller.ts
let SheetsFilterPermissionController = class SheetsFilterPermissionController extends Disposable {
	constructor(_sheetsFilterService, _localeService, _commandService, _sheetPermissionCheckPermission, _injector, _sheetsSelectionService) {
		super();
		this._sheetsFilterService = _sheetsFilterService;
		this._localeService = _localeService;
		this._commandService = _commandService;
		this._sheetPermissionCheckPermission = _sheetPermissionCheckPermission;
		this._injector = _injector;
		this._sheetsSelectionService = _sheetsSelectionService;
		this._commandExecutedListener();
	}
	_commandExecutedListener() {
		this.disposeWithMe(this._commandService.beforeCommandExecuted((command) => {
			if (command.id === SmartToggleSheetsFilterCommand.id) {
				var _this$_sheetsFilterSe;
				const target = getSheetCommandTarget(this._injector.get(IUniverInstanceService));
				if (!target) return;
				const { unitId, subUnitId, worksheet } = target;
				const filterRange = (_this$_sheetsFilterSe = this._sheetsFilterService.getFilterModel(unitId, subUnitId)) === null || _this$_sheetsFilterSe === void 0 ? void 0 : _this$_sheetsFilterSe.getRange();
				let permission;
				if (filterRange) permission = this._sheetPermissionCheckPermission.permissionCheckWithRanges({
					rangeTypes: [RangeProtectionPermissionViewPoint],
					worksheetTypes: [WorksheetFilterPermission, WorksheetViewPermission]
				}, [filterRange], unitId, subUnitId);
				else {
					var _this$_sheetsSelectio;
					const range = (_this$_sheetsSelectio = this._sheetsSelectionService.getCurrentLastSelection()) === null || _this$_sheetsSelectio === void 0 ? void 0 : _this$_sheetsSelectio.range;
					if (range) {
						let newRange = { ...range };
						newRange = range.startColumn === range.endColumn && range.startRow === range.endRow ? expandToContinuousRange(newRange, {
							left: true,
							right: true,
							up: true,
							down: true
						}, worksheet) : newRange;
						permission = this._sheetPermissionCheckPermission.permissionCheckWithRanges({
							rangeTypes: [RangeProtectionPermissionViewPoint],
							worksheetTypes: [WorksheetViewPermission, WorksheetFilterPermission]
						}, [newRange], unitId, subUnitId);
					} else permission = this._sheetPermissionCheckPermission.permissionCheckWithoutRange({
						rangeTypes: [RangeProtectionPermissionViewPoint],
						worksheetTypes: [WorksheetViewPermission, WorksheetFilterPermission]
					});
				}
				if (!permission) this._sheetPermissionCheckPermission.blockExecuteWithoutPermission(this._localeService.t("sheets-filter-ui.permission.filterErr"));
			}
			if (command.id === OpenFilterPanelOperation.id) {
				var _this$_sheetsFilterSe2;
				const params = command.params;
				const { unitId, subUnitId } = params;
				const filterRange = (_this$_sheetsFilterSe2 = this._sheetsFilterService.getFilterModel(unitId, subUnitId)) === null || _this$_sheetsFilterSe2 === void 0 ? void 0 : _this$_sheetsFilterSe2.getRange();
				const colRange = Tools.deepClone(filterRange);
				if (colRange) {
					colRange.startColumn = params.col;
					colRange.endColumn = params.col;
					if (!this._sheetPermissionCheckPermission.permissionCheckWithRanges({
						rangeTypes: [RangeProtectionPermissionViewPoint],
						worksheetTypes: [WorksheetFilterPermission, WorksheetViewPermission]
					}, [colRange], unitId, subUnitId)) this._sheetPermissionCheckPermission.blockExecuteWithoutPermission(this._localeService.t("sheets-filter-ui.permission.filterErr"));
				}
			}
		}));
	}
};
SheetsFilterPermissionController = __decorate([
	__decorateParam(0, Inject(SheetsFilterService)),
	__decorateParam(1, Inject(LocaleService)),
	__decorateParam(2, ICommandService),
	__decorateParam(3, Inject(SheetPermissionCheckController)),
	__decorateParam(4, Inject(Injector)),
	__decorateParam(5, Inject(SheetsSelectionsService))
], SheetsFilterPermissionController);

//#endregion
//#region src/views/widgets/drawings.ts
const BUTTON_VIEWPORT = 16;
const FILTER_BUTTON_EMPTY = new Path2D("M3.30363 3C2.79117 3 2.51457 3.60097 2.84788 3.99024L6.8 8.60593V12.5662C6.8 12.7184 6.8864 12.8575 7.02289 12.9249L8.76717 13.7863C8.96655 13.8847 9.2 13.7396 9.2 13.5173V8.60593L13.1521 3.99024C13.4854 3.60097 13.2088 3 12.6964 3H3.30363Z");
var FilterButton = class {
	static drawNoCriteria(ctx, size, fgColor, bgColor) {
		ctx.save();
		Rect.drawWith(ctx, {
			radius: 2,
			width: BUTTON_VIEWPORT,
			height: BUTTON_VIEWPORT,
			fill: bgColor
		});
		ctx.lineCap = "square";
		ctx.strokeStyle = fgColor;
		ctx.scale(size / BUTTON_VIEWPORT, size / BUTTON_VIEWPORT);
		ctx.beginPath();
		ctx.lineWidth = 1;
		ctx.lineCap = "round";
		ctx.moveTo(3, 4);
		ctx.lineTo(13, 4);
		ctx.moveTo(4.5, 8);
		ctx.lineTo(11.5, 8);
		ctx.moveTo(6, 12);
		ctx.lineTo(10, 12);
		ctx.stroke();
		ctx.restore();
	}
	static drawHasCriteria(ctx, size, fgColor, bgColor) {
		ctx.save();
		Rect.drawWith(ctx, {
			radius: 2,
			width: BUTTON_VIEWPORT,
			height: BUTTON_VIEWPORT,
			fill: bgColor
		});
		ctx.scale(size / BUTTON_VIEWPORT, size / BUTTON_VIEWPORT);
		ctx.fillStyle = fgColor;
		ctx.fill(FILTER_BUTTON_EMPTY);
		ctx.restore();
	}
};

//#endregion
//#region src/views/widgets/filter-button.shape.ts
const FILTER_ICON_SIZE = 16;
let SheetsFilterButtonShape = class SheetsFilterButtonShape extends Shape {
	constructor(key, props, _contextService, _commandService, _themeService) {
		super(key, props);
		this._contextService = _contextService;
		this._commandService = _commandService;
		this._themeService = _themeService;
		_defineProperty(this, "_cellWidth", 0);
		_defineProperty(this, "_cellHeight", 0);
		_defineProperty(this, "_filterParams", void 0);
		_defineProperty(this, "_hovered", false);
		this.setShapeProps(props);
		this.onPointerDown$.subscribeEvent((evt) => this.onPointerDown(evt));
		this.onPointerEnter$.subscribeEvent(() => this.onPointerEnter());
		this.onPointerLeave$.subscribeEvent(() => this.onPointerLeave());
	}
	setShapeProps(props) {
		if (typeof props.cellHeight !== "undefined") this._cellHeight = props.cellHeight;
		if (typeof props.cellWidth !== "undefined") this._cellWidth = props.cellWidth;
		if (typeof props.filterParams !== "undefined") this._filterParams = props.filterParams;
		this.transformByState({
			width: props.width,
			height: props.height
		});
	}
	_draw(ctx) {
		const cellHeight = this._cellHeight;
		const cellWidth = this._cellWidth;
		const left = 16 - cellWidth;
		const top = 16 - cellHeight;
		ctx.save();
		const cellRegion = new Path2D();
		cellRegion.rect(left, top, cellWidth, cellHeight);
		ctx.clip(cellRegion);
		const { hasCriteria } = this._filterParams;
		const fgColor = this._themeService.getColorFromTheme("primary.600");
		const bgColor = this._hovered ? this._themeService.getColorFromTheme("gray.50") : "rgba(255, 255, 255, 1.0)";
		if (hasCriteria) FilterButton.drawHasCriteria(ctx, 16, fgColor, bgColor);
		else FilterButton.drawNoCriteria(ctx, 16, fgColor, bgColor);
		ctx.restore();
	}
	onPointerDown(evt) {
		if (evt.button === 2) return;
		const { col, unitId, subUnitId } = this._filterParams;
		if (this._contextService.getContextValue("FILTER_PANEL_OPENED") || !this._commandService.hasCommand(OpenFilterPanelOperation.id)) return;
		setTimeout(() => {
			this._commandService.executeCommand(OpenFilterPanelOperation.id, {
				unitId,
				subUnitId,
				col
			});
		}, 200);
	}
	onPointerEnter() {
		this._hovered = true;
		this.makeDirty(true);
	}
	onPointerLeave() {
		this._hovered = false;
		this.makeDirty(true);
	}
};
SheetsFilterButtonShape = __decorate([
	__decorateParam(2, IContextService),
	__decorateParam(3, ICommandService),
	__decorateParam(4, Inject(ThemeService))
], SheetsFilterButtonShape);

//#endregion
//#region src/views/render-modules/sheets-filter.render-controller.ts
const DEFAULT_Z_INDEX = 1e3;
const SHEETS_FILTER_BUTTON_Z_INDEX = 5e3;
function computeIconTop(startY, endY, cellHeight, verticalAlign) {
	switch (verticalAlign) {
		case VerticalAlign.TOP: return startY + 1;
		case VerticalAlign.MIDDLE: return startY + Math.max(0, (cellHeight - 16) / 2);
		case VerticalAlign.BOTTOM:
		default: return endY - 16 - 1;
	}
}
let SheetsFilterRenderController = class SheetsFilterRenderController extends RxDisposable {
	constructor(_context, _injector, _sheetSkeletonManagerService, _sheetsFilterService, _themeService, _sheetInterceptorService, _commandService, _selectionRenderService) {
		super();
		this._context = _context;
		this._injector = _injector;
		this._sheetSkeletonManagerService = _sheetSkeletonManagerService;
		this._sheetsFilterService = _sheetsFilterService;
		this._themeService = _themeService;
		this._sheetInterceptorService = _sheetInterceptorService;
		this._commandService = _commandService;
		this._selectionRenderService = _selectionRenderService;
		_defineProperty(this, "_currentRenderParams", null);
		_defineProperty(this, "_filterRangeShape", null);
		_defineProperty(this, "_buttonRenderDisposable", null);
		_defineProperty(this, "_filterButtonShapes", []);
		this._initRenderer();
	}
	dispose() {
		super.dispose();
		this._disposeRendering();
	}
	_initRenderer() {
		this.disposeWithMe(this._themeService.currentTheme$.subscribe(() => {
			this._refreshRendering(this._currentRenderParams);
		}));
		this._sheetSkeletonManagerService.currentSkeleton$.pipe(switchMap((skeletonParams) => {
			var _workbook$getActiveSh, _this$_sheetsFilterSe;
			if (!skeletonParams) return of(null);
			const { unit: workbook, unitId } = this._context;
			const worksheetId = ((_workbook$getActiveSh = workbook.getActiveSheet()) === null || _workbook$getActiveSh === void 0 ? void 0 : _workbook$getActiveSh.getSheetId()) || "";
			const filterModel = (_this$_sheetsFilterSe = this._sheetsFilterService.getFilterModel(unitId, worksheetId)) !== null && _this$_sheetsFilterSe !== void 0 ? _this$_sheetsFilterSe : void 0;
			const getParams = () => ({
				unitId,
				worksheetId,
				filterModel,
				range: filterModel === null || filterModel === void 0 ? void 0 : filterModel.getRange(),
				skeleton: skeletonParams.skeleton
			});
			return fromCallback(this._commandService.onCommandExecuted.bind(this._commandService)).pipe(filter(([command]) => {
				var _command$params;
				return command.type === CommandType.MUTATION && ((_command$params = command.params) === null || _command$params === void 0 ? void 0 : _command$params.unitId) === workbook.getUnitId() && (FILTER_MUTATIONS.has(command.id) || command.id === SetRangeValuesMutation.id);
			}), throttleTime(20, void 0, {
				leading: false,
				trailing: true
			}), map(getParams), startWith(getParams()));
		}), takeUntil(this.dispose$)).subscribe((renderParams) => {
			this._currentRenderParams = renderParams;
			this._refreshRendering(renderParams);
		});
	}
	_refreshRendering(renderParams) {
		this._disposeRendering();
		if (!renderParams || !renderParams.range) return;
		this._renderRange(renderParams.range, renderParams.skeleton);
		this._renderButtons(renderParams);
	}
	_renderRange(range, skeleton) {
		const { scene } = this._context;
		const { rowHeaderWidth, columnHeaderHeight } = skeleton;
		const filterRangeShape = this._filterRangeShape = new SelectionControl(scene, DEFAULT_Z_INDEX, this._themeService, {
			rowHeaderWidth,
			columnHeaderHeight,
			enableAutoFill: false,
			highlightHeader: false
		});
		const selectionWithCoord = attachSelectionWithCoord({
			range,
			primary: null,
			style: { fill: "rgba(0, 0, 0, 0.0)" }
		}, skeleton);
		filterRangeShape.updateRangeBySelectionWithCoord(selectionWithCoord);
		filterRangeShape.setEvent(false);
		scene.makeDirty(true);
	}
	_renderButtons(params) {
		const { range, filterModel, unitId, skeleton, worksheetId } = params;
		const { unit: workbook, scene } = this._context;
		const worksheet = workbook.getSheetBySheetId(worksheetId);
		if (!worksheet) return;
		this._interceptCellContent(unitId, worksheetId, params.range);
		const { startColumn, endColumn, startRow } = range;
		for (let col = startColumn; col <= endColumn; col++) {
			const key = `sheets-filter-button-${col}`;
			const startPosition = getCoordByCell(startRow, col, scene, skeleton);
			const cellStyle = worksheet.getComposedCellStyle(startRow, col);
			const verticalAlign = (cellStyle === null || cellStyle === void 0 ? void 0 : cellStyle.vt) || VerticalAlign.BOTTOM;
			const { startX, startY, endX, endY } = startPosition;
			const cellWidth = endX - startX;
			const cellHeight = endY - startY;
			if (cellHeight <= 1 || cellWidth <= 1) continue;
			const hasCriteria = !!filterModel.getFilterColumn(col);
			const props = {
				left: endX - 16 - 1,
				top: computeIconTop(startY, endY, cellHeight, verticalAlign),
				height: 16,
				width: 16,
				zIndex: SHEETS_FILTER_BUTTON_Z_INDEX,
				cellHeight,
				cellWidth,
				filterParams: {
					unitId,
					subUnitId: worksheetId,
					col,
					hasCriteria
				}
			};
			const buttonShape = this._injector.createInstance(SheetsFilterButtonShape, key, props);
			this._filterButtonShapes.push(buttonShape);
		}
		scene.addObjects(this._filterButtonShapes);
		scene.makeDirty();
	}
	_interceptCellContent(workbookId, worksheetId, range) {
		const { startRow, startColumn, endColumn } = range;
		this._buttonRenderDisposable = this._sheetInterceptorService.intercept(INTERCEPTOR_POINT.CELL_CONTENT, {
			effect: InterceptorEffectEnum.Style,
			handler: (cell, pos, next) => {
				const { row, col, unitId, subUnitId } = pos;
				if (unitId !== workbookId || subUnitId !== worksheetId || row !== startRow || col < startColumn || col > endColumn) return next(cell);
				if (!cell || cell === pos.rawData) cell = { ...pos.rawData };
				cell.fontRenderExtension = {
					...cell === null || cell === void 0 ? void 0 : cell.fontRenderExtension,
					rightOffset: 16
				};
				return next(cell);
			},
			priority: 10
		});
	}
	_disposeRendering() {
		var _this$_filterRangeSha, _this$_buttonRenderDi;
		(_this$_filterRangeSha = this._filterRangeShape) === null || _this$_filterRangeSha === void 0 || _this$_filterRangeSha.dispose();
		this._filterButtonShapes.forEach((s) => s.dispose());
		(_this$_buttonRenderDi = this._buttonRenderDisposable) === null || _this$_buttonRenderDi === void 0 || _this$_buttonRenderDi.dispose();
		this._filterRangeShape = null;
		this._buttonRenderDisposable = null;
		this._filterButtonShapes = [];
	}
};
SheetsFilterRenderController = __decorate([
	__decorateParam(1, Inject(Injector)),
	__decorateParam(2, Inject(SheetSkeletonManagerService)),
	__decorateParam(3, Inject(SheetsFilterService)),
	__decorateParam(4, Inject(ThemeService)),
	__decorateParam(5, Inject(SheetInterceptorService)),
	__decorateParam(6, ICommandService),
	__decorateParam(7, ISheetSelectionRenderService)
], SheetsFilterRenderController);

//#endregion
//#region src/controllers/sheets-filter-ui-mobile.controller.ts
let SheetsFilterUIMobileController = class SheetsFilterUIMobileController extends RxDisposable {
	constructor(_renderManagerService, _sheetsRenderService) {
		super();
		this._renderManagerService = _renderManagerService;
		this._sheetsRenderService = _sheetsRenderService;
		[
			SetSheetsFilterRangeMutation,
			SetSheetsFilterCriteriaMutation,
			RemoveSheetsFilterMutation,
			ReCalcSheetsFilterMutation
		].forEach((m) => this.disposeWithMe(this._sheetsRenderService.registerSkeletonChangingMutations(m.id)));
		this.disposeWithMe(this._renderManagerService.registerRenderModule(UniverInstanceType.UNIVER_SHEET, [SheetsFilterRenderController]));
	}
};
SheetsFilterUIMobileController = __decorate([__decorateParam(0, IRenderManagerService), __decorateParam(1, Inject(SheetsRenderService))], SheetsFilterUIMobileController);

//#endregion
//#region src/mobile-plugin.ts
let UniverSheetsFilterMobileUIPlugin = class UniverSheetsFilterMobileUIPlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _injector, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		const { menu, ...rest } = merge({}, defaultPluginConfig, this._config);
		if (menu) this._configService.setConfig("menu", menu, { merge: true });
		this._configService.setConfig(SHEETS_FILTER_UI_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		[[SheetsFilterPermissionController], [SheetsFilterUIMobileController]].forEach((d) => this._injector.add(d));
	}
	onReady() {
		this._injector.get(SheetsFilterPermissionController);
	}
	onRendered() {
		this._injector.get(SheetsFilterUIMobileController);
	}
};
_defineProperty(UniverSheetsFilterMobileUIPlugin, "type", UniverInstanceType.UNIVER_SHEET);
_defineProperty(UniverSheetsFilterMobileUIPlugin, "pluginName", "SHEET_FILTER_UI_PLUGIN");
_defineProperty(UniverSheetsFilterMobileUIPlugin, "packageName", name);
_defineProperty(UniverSheetsFilterMobileUIPlugin, "version", version);
UniverSheetsFilterMobileUIPlugin = __decorate([
	DependentOn(UniverSheetsFilterPlugin),
	__decorateParam(1, Inject(Injector)),
	__decorateParam(2, IConfigService)
], UniverSheetsFilterMobileUIPlugin);

//#endregion
//#region src/menu/sheets-filter.menu.ts
function SmartToggleFilterMenuItemFactory(accessor) {
	const sheetsFilterService = accessor.get(SheetsFilterService);
	return {
		id: SmartToggleSheetsFilterCommand.id,
		type: MenuItemType.BUTTON_SELECTOR,
		icon: "FilterIcon",
		tooltip: "sheets-filter-ui.toolbar.smart-toggle-filter-tooltip",
		hidden$: getMenuHiddenObservable(accessor, UniverInstanceType.UNIVER_SHEET),
		activated$: sheetsFilterService.activeFilterModel$.pipe(map((model) => !!model)),
		disabled$: getObservableWithExclusiveRange$(accessor, getCurrentRangeDisable$(accessor, {
			worksheetTypes: [WorksheetFilterPermission, WorksheetViewPermission],
			rangeTypes: [RangeProtectionPermissionViewPoint]
		}))
	};
}
function ClearFilterCriteriaMenuItemFactory(accessor) {
	const sheetsFilterService = accessor.get(SheetsFilterService);
	return {
		id: ClearSheetsFilterCriteriaCommand.id,
		type: MenuItemType.BUTTON,
		title: "sheets-filter-ui.toolbar.clear-filter-criteria",
		hidden$: getMenuHiddenObservable(accessor, UniverInstanceType.UNIVER_SHEET),
		disabled$: sheetsFilterService.activeFilterModel$.pipe(switchMap((model) => {
			var _model$hasCriteria$$p;
			return (_model$hasCriteria$$p = model === null || model === void 0 ? void 0 : model.hasCriteria$.pipe(map((m) => !m))) !== null && _model$hasCriteria$$p !== void 0 ? _model$hasCriteria$$p : of(true);
		}))
	};
}
function ReCalcFilterMenuItemFactory(accessor) {
	const sheetsFilterService = accessor.get(SheetsFilterService);
	return {
		id: ReCalcSheetsFilterCommand.id,
		type: MenuItemType.BUTTON,
		title: "sheets-filter-ui.toolbar.re-calc-filter-conditions",
		hidden$: getMenuHiddenObservable(accessor, UniverInstanceType.UNIVER_SHEET),
		disabled$: sheetsFilterService.activeFilterModel$.pipe(switchMap((model) => {
			var _model$hasCriteria$$p2;
			return (_model$hasCriteria$$p2 = model === null || model === void 0 ? void 0 : model.hasCriteria$.pipe(map((m) => !m))) !== null && _model$hasCriteria$$p2 !== void 0 ? _model$hasCriteria$$p2 : of(true);
		}))
	};
}

//#endregion
//#region src/menu/schema.ts
const menuSchema = { [RibbonDataGroup.ORGANIZATION]: { [SmartToggleSheetsFilterCommand.id]: {
	order: 2,
	menuItemFactory: SmartToggleFilterMenuItemFactory,
	[ClearSheetsFilterCriteriaCommand.id]: {
		order: 0,
		menuItemFactory: ClearFilterCriteriaMenuItemFactory
	},
	[ReCalcSheetsFilterCommand.id]: {
		order: 1,
		menuItemFactory: ReCalcFilterMenuItemFactory
	}
} } };

//#endregion
//#region src/views/components/SheetsFilterByColorsPanel.tsx
/**
* Filter by colors.
*/
function FilterByColor(props) {
	const { model } = props;
	const localeService = useDependency(LocaleService);
	const cellFillColors = useObservable(model.cellFillColors$, [], true);
	const cellTextColors = useObservable(model.cellTextColors$, [], true);
	const handleSelectCellFillColor = useCallback((color) => {
		model.onFilterCheckToggled(color);
	}, [model]);
	const handleSelectCellTextColor = useCallback((color) => {
		model.onFilterCheckToggled(color, false);
	}, [model]);
	return /* @__PURE__ */ jsx("div", {
		"data-u-comp": "sheets-filter-panel-colors-container",
		className: "univer-flex univer-h-full univer-min-h-[300px] univer-flex-col",
		children: /* @__PURE__ */ jsxs("div", {
			"data-u-comp": "sheets-filter-panel",
			className: clsx("univer-mt-2 univer-box-border univer-flex univer-h-[300px] univer-flex-grow univer-flex-col univer-gap-4 univer-overflow-auto univer-rounded-md univer-px-2 univer-py-2.5", borderClassName),
			children: [
				cellFillColors.length > 1 && /* @__PURE__ */ jsxs("div", { children: [/* @__PURE__ */ jsx("div", {
					className: "univer-mb-2 univer-text-sm univer-text-gray-900 dark:!univer-text-white",
					children: localeService.t("sheets-filter-ui.panel.filter-by-cell-fill-color")
				}), /* @__PURE__ */ jsx("div", {
					className: "univer-grid univer-grid-cols-8 univer-items-center univer-justify-start univer-gap-2",
					children: cellFillColors.map((color, index) => /* @__PURE__ */ jsxs("div", {
						className: "univer-relative univer-size-6",
						onClick: () => handleSelectCellFillColor(color),
						children: [!color.color ? /* @__PURE__ */ jsx(BanIcon, { className: "univer-size-6 univer-cursor-pointer univer-rounded-full hover:univer-ring-2 hover:univer-ring-offset-2 hover:univer-ring-offset-white" }) : /* @__PURE__ */ jsx("button", {
							type: "button",
							className: clsx("univer-box-border univer-size-6 univer-cursor-pointer univer-rounded-full univer-border univer-border-solid univer-border-transparent univer-bg-gray-300 univer-transition-shadow hover:univer-ring-2 hover:univer-ring-offset-2 hover:univer-ring-offset-white"),
							style: { backgroundColor: color.color }
						}), color.checked && /* @__PURE__ */ jsx(CheckedIcon, {})]
					}, `sheets-filter-cell-fill-color-${index}`))
				})] }),
				cellTextColors.length > 1 && /* @__PURE__ */ jsxs("div", { children: [/* @__PURE__ */ jsx("div", {
					className: "univer-mb-2 univer-text-sm univer-text-gray-900 dark:!univer-text-white",
					children: localeService.t("sheets-filter-ui.panel.filter-by-cell-text-color")
				}), /* @__PURE__ */ jsx("div", {
					className: "univer-grid univer-grid-cols-8 univer-items-center univer-justify-start univer-gap-2",
					children: cellTextColors.map((color, index) => /* @__PURE__ */ jsxs("div", {
						className: "univer-relative univer-size-6",
						onClick: () => handleSelectCellTextColor(color),
						children: [/* @__PURE__ */ jsx("div", {
							className: "univer-box-border univer-flex univer-size-full univer-cursor-pointer univer-items-center univer-justify-center univer-rounded-full univer-border univer-border-solid univer-border-[rgba(13,13,13,0.06)] univer-p-0.5 hover:univer-ring-2 hover:univer-ring-offset-2 hover:univer-ring-offset-white dark:!univer-border-[rgba(255,255,255,0.06)]",
							children: /* @__PURE__ */ jsx(AIcon, { style: { color: color.color } })
						}), color.checked && /* @__PURE__ */ jsx(CheckedIcon, {})]
					}, `sheets-filter-cell-text-color-${index}`))
				})] }),
				cellFillColors.length <= 1 && cellTextColors.length <= 1 && /* @__PURE__ */ jsx("div", {
					className: "univer-flex univer-size-full univer-items-center univer-justify-center univer-text-sm univer-text-gray-900 dark:!univer-text-gray-200",
					children: localeService.t("sheets-filter-ui.panel.filter-by-color-none")
				})
			]
		})
	});
}
function CheckedIcon() {
	return /* @__PURE__ */ jsx("div", {
		className: "univer-absolute -univer-bottom-0.5 -univer-right-0.5 univer-flex univer-size-3 univer-cursor-pointer univer-items-center univer-justify-center univer-rounded-full univer-bg-white",
		children: /* @__PURE__ */ jsx(SuccessIcon, { className: "univer-size-full univer-font-bold univer-text-[#418F1F]" })
	});
}

//#endregion
//#region src/views/components/SheetsFilterByConditionsPanel.tsx
/**
* Filter by conditions.
*/
function FilterByCondition(props) {
	var _formParams$val, _formParams$val2;
	const { model } = props;
	const localeService = useDependency(LocaleService);
	const condition = useObservable(model.conditionItem$, void 0);
	const formParams = useObservable(model.filterConditionFormParams$, void 0);
	const radioValue = (formParams === null || formParams === void 0 ? void 0 : formParams.and) ? "AND" : "OR";
	const onRadioChange = useCallback((key) => {
		model.onConditionFormChange({ and: key === "AND" });
	}, [model]);
	const primaryOptions = usePrimaryOptions(localeService);
	const onPrimaryConditionChange = useCallback((value) => {
		model.onPrimaryConditionChange(value);
	}, [model]);
	const secondaryOptions = useSecondaryOptions(localeService);
	const onFormParamsChange = useCallback((diffParams) => {
		model.onConditionFormChange(diffParams);
	}, [model]);
	const placeholder = localeService.t("sheets-filter-ui.panel.input-values-placeholder");
	function renderSecondaryCondition(operator, val, name) {
		const shouldRenderInput = FilterConditionItems.getItemByOperator(operator).numOfParameters === 1;
		return /* @__PURE__ */ jsxs(Fragment, { children: [
			name === "operator2" && /* @__PURE__ */ jsxs(RadioGroup, {
				value: radioValue,
				onChange: onRadioChange,
				children: [/* @__PURE__ */ jsx(Radio, {
					value: "AND",
					children: localeService.t("sheets-filter-ui.panel.and")
				}), /* @__PURE__ */ jsx(Radio, {
					value: "OR",
					children: localeService.t("sheets-filter-ui.panel.or")
				})]
			}),
			/* @__PURE__ */ jsx(Select, {
				value: operator,
				options: secondaryOptions,
				onChange: (operator) => onFormParamsChange({ [name]: operator })
			}),
			shouldRenderInput && /* @__PURE__ */ jsx("div", { children: /* @__PURE__ */ jsx(Input, {
				className: "univer-mt-2",
				value: val,
				placeholder,
				onChange: (value) => onFormParamsChange({ [name === "operator1" ? "val1" : "val2"]: value })
			}) })
		] });
	}
	return /* @__PURE__ */ jsx("div", {
		"data-u-comp": "sheets-filter-panel-conditions-container",
		className: "univer-flex univer-h-full univer-min-h-[300px] univer-flex-col",
		children: condition && formParams && /* @__PURE__ */ jsxs(Fragment, { children: [/* @__PURE__ */ jsx(Select, {
			value: condition.operator,
			options: primaryOptions,
			onChange: onPrimaryConditionChange
		}), FilterConditionItems.getItemByOperator(condition.operator).numOfParameters !== 0 ? /* @__PURE__ */ jsxs("div", {
			"data-u-comp": "sheets-filter-panel-conditions-container-inner",
			className: clsx("univer-mt-2 univer-flex-grow univer-overflow-hidden univer-rounded-md univer-p-2", borderClassName),
			children: [
				condition.numOfParameters >= 1 && renderSecondaryCondition(formParams.operator1, (_formParams$val = formParams.val1) !== null && _formParams$val !== void 0 ? _formParams$val : "", "operator1"),
				condition.numOfParameters >= 2 && renderSecondaryCondition(formParams.operator2, (_formParams$val2 = formParams.val2) !== null && _formParams$val2 !== void 0 ? _formParams$val2 : "", "operator2"),
				/* @__PURE__ */ jsxs("div", {
					"data-u-comp": "sheets-filter-panel-conditions-desc",
					className: "univer-mt-2 univer-text-xs univer-text-gray-500",
					children: [
						localeService.t("sheets-filter-ui.panel.?"),
						/* @__PURE__ */ jsx("br", {}),
						localeService.t("sheets-filter-ui.panel.*")
					]
				})
			]
		}) : null] })
	});
}
function usePrimaryOptions(localeService) {
	return useMemo(() => [
		{ options: [{
			label: localeService.t(FilterConditionItems.NONE.label),
			value: FilterConditionItems.NONE.operator
		}] },
		{ options: [{
			label: localeService.t(FilterConditionItems.EMPTY.label),
			value: FilterConditionItems.EMPTY.operator
		}, {
			label: localeService.t(FilterConditionItems.NOT_EMPTY.label),
			value: FilterConditionItems.NOT_EMPTY.operator
		}] },
		{ options: [
			{
				label: localeService.t(FilterConditionItems.TEXT_CONTAINS.label),
				value: FilterConditionItems.TEXT_CONTAINS.operator
			},
			{
				label: localeService.t(FilterConditionItems.DOES_NOT_CONTAIN.label),
				value: FilterConditionItems.DOES_NOT_CONTAIN.operator
			},
			{
				label: localeService.t(FilterConditionItems.STARTS_WITH.label),
				value: FilterConditionItems.STARTS_WITH.operator
			},
			{
				label: localeService.t(FilterConditionItems.ENDS_WITH.label),
				value: FilterConditionItems.ENDS_WITH.operator
			},
			{
				label: localeService.t(FilterConditionItems.EQUALS.label),
				value: FilterConditionItems.EQUALS.operator
			}
		] },
		{ options: [
			{
				label: localeService.t(FilterConditionItems.GREATER_THAN.label),
				value: FilterConditionItems.GREATER_THAN.operator
			},
			{
				label: localeService.t(FilterConditionItems.GREATER_THAN_OR_EQUAL.label),
				value: FilterConditionItems.GREATER_THAN_OR_EQUAL.operator
			},
			{
				label: localeService.t(FilterConditionItems.LESS_THAN.label),
				value: FilterConditionItems.LESS_THAN.operator
			},
			{
				label: localeService.t(FilterConditionItems.LESS_THAN_OR_EQUAL.label),
				value: FilterConditionItems.LESS_THAN_OR_EQUAL.operator
			},
			{
				label: localeService.t(FilterConditionItems.EQUAL.label),
				value: FilterConditionItems.EQUAL.operator
			},
			{
				label: localeService.t(FilterConditionItems.NOT_EQUAL.label),
				value: FilterConditionItems.NOT_EQUAL.operator
			},
			{
				label: localeService.t(FilterConditionItems.BETWEEN.label),
				value: FilterConditionItems.BETWEEN.operator
			},
			{
				label: localeService.t(FilterConditionItems.NOT_BETWEEN.label),
				value: FilterConditionItems.NOT_BETWEEN.operator
			}
		] },
		{ options: [{
			label: localeService.t(FilterConditionItems.CUSTOM.label),
			value: FilterConditionItems.CUSTOM.operator
		}] }
	], [localeService.getCurrentLocale(), localeService]);
}
function useSecondaryOptions(localeService) {
	return useMemo(() => FilterConditionItems.ALL_CONDITIONS.filter((c) => c.numOfParameters !== 2).map((c) => ({
		label: localeService.t(c.label),
		value: c.operator
	})), [localeService.getCurrentLocale(), localeService]);
}

//#endregion
//#region src/views/components/SheetsFilterByValuesPanel.tsx
/**
* Filter by values.
*/
function FilterByValue(props) {
	const { model } = props;
	const localeService = useDependency(LocaleService);
	const searchText = useObservable(model.searchString$, "", true);
	const items = useObservable(model.filterItems$, void 0, true);
	const filterOnly = localeService.t("sheets-filter-ui.panel.filter-only");
	const stat = statisticFilterByValueItems(items);
	const allChecked = stat.checked > 0 && stat.unchecked === 0;
	const indeterminate = stat.checked > 0 && stat.unchecked > 0;
	const treeMap = model.treeMapCache;
	const onCheckAllToggled = useCallback(() => {
		model.onCheckAllToggled(!allChecked);
	}, [model, allChecked]);
	const onSearchValueChange = useCallback((str) => {
		model.setSearchString(str);
	}, [model]);
	function extractCheckedKeys(items) {
		let checkedKeys = [];
		items.forEach((item) => {
			if (item.checked) checkedKeys.push(item.key);
			if (item.children) checkedKeys = checkedKeys.concat(extractCheckedKeys(item.children));
		});
		return checkedKeys;
	}
	return /* @__PURE__ */ jsxs("div", {
		"data-u-comp": "sheets-filter-panel-values-container",
		className: "univer-flex univer-h-full univer-min-h-[300px] univer-flex-col",
		children: [/* @__PURE__ */ jsx(Input, {
			autoFocus: true,
			value: searchText,
			placeholder: localeService.t("sheets-filter-ui.panel.search-placeholder"),
			onChange: onSearchValueChange
		}), /* @__PURE__ */ jsxs("div", {
			"data-u-comp": "sheets-filter-panel",
			className: clsx("univer-mt-2 univer-box-border univer-flex univer-flex-grow univer-flex-col univer-overflow-hidden univer-rounded-md univer-px-2 univer-py-2.5", borderClassName),
			children: [/* @__PURE__ */ jsx("div", {
				"data-u-comp": "sheets-filter-panel-values-item",
				className: "univer-box-border univer-h-8 univer-w-full univer-py-0.5",
				children: /* @__PURE__ */ jsxs("div", {
					"data-u-comp": "sheets-filter-panel-values-item-inner",
					className: "univer-box-border univer-flex univer-h-7 univer-items-center univer-rounded-md univer-py-0 univer-pl-5 univer-pr-0.5 univer-text-sm",
					children: [
						/* @__PURE__ */ jsx(Checkbox, {
							indeterminate,
							disabled: items.length === 0,
							checked: allChecked,
							onChange: onCheckAllToggled
						}),
						/* @__PURE__ */ jsx("span", {
							"data-u-comp": "sheets-filter-panel-values-item-text",
							className: "univer-mx-1 univer-inline-block univer-flex-shrink univer-truncate univer-text-gray-900 dark:!univer-text-white",
							children: `${localeService.t("sheets-filter-ui.panel.select-all")}`
						}),
						/* @__PURE__ */ jsx("span", {
							"data-u-comp": "sheets-filter-panel-values-item-count",
							className: "univer-text-gray-400 dark:!univer-text-gray-500",
							children: `(${stat.checked}/${stat.checked + stat.unchecked})`
						})
					]
				})
			}), /* @__PURE__ */ jsx("div", {
				"data-u-comp": "sheets-filter-panel-values-virtual",
				className: "univer-flex-grow",
				children: /* @__PURE__ */ jsx(Tree, {
					data: items,
					defaultExpandAll: false,
					valueGroup: extractCheckedKeys(items),
					onChange: (node) => {
						model.onFilterCheckToggled(node);
					},
					defaultCache: treeMap,
					itemHeight: 28,
					treeNodeClassName: `
                          univer-pr-2 univer-border-box univer-rounded-md
                          [&:hover_a]:univer-inline-block
                          hover:univer-bg-gray-50 univer-h-full
                          univer-text-gray-900 dark:hover:!univer-bg-gray-900
                          dark:!univer-text-white
                        `,
					attachRender: (item) => /* @__PURE__ */ jsxs("div", {
						className: "univer-ml-1 univer-flex univer-h-5 univer-flex-1 univer-cursor-pointer univer-items-center univer-justify-between univer-text-sm univer-text-primary-500",
						children: [/* @__PURE__ */ jsx("span", {
							"data-u-comp": "sheets-filter-panel-values-item-count",
							className: "univer-text-gray-400 dark:!univer-text-gray-500",
							children: `(${item.count})`
						}), /* @__PURE__ */ jsx("a", {
							className: "univer-box-border univer-hidden univer-h-4 univer-whitespace-nowrap univer-px-1.5",
							onClick: () => {
								const filterValues = [];
								if (item.children) item.children.forEach((child) => {
									if (child.children) child.children.forEach((subChild) => {
										filterValues.push(subChild.key);
									});
									else filterValues.push(child.key);
								});
								else filterValues.push(item.key);
								model.onFilterOnly(filterValues);
							},
							children: filterOnly
						})]
					})
				})
			})]
		})]
	});
}

//#endregion
//#region src/views/components/SheetsFilterSyncSwitch.tsx
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
function FilterSyncSwitch() {
	const sheetsFilterSyncController = useDependency(SheetsFilterSyncController);
	if (!useObservable(sheetsFilterSyncController.visible$, void 0, true)) return null;
	const localeService = useDependency(LocaleService);
	const messageService = useDependency(IMessageService);
	const enabled = useObservable(sheetsFilterSyncController.enabled$, void 0, true);
	return /* @__PURE__ */ jsxs("div", {
		className: "univer-mt-2 univer-flex univer-items-center univer-justify-between univer-text-sm univer-text-gray-900 dark:!univer-text-gray-200",
		children: [/* @__PURE__ */ jsxs("div", {
			className: "univer-flex univer-items-center univer-gap-1",
			children: [/* @__PURE__ */ jsx("span", { children: localeService.t("sheets-filter-ui.sync.title") }), /* @__PURE__ */ jsx(Tooltip, {
				title: enabled ? localeService.t("sheets-filter-ui.sync.statusTips.off") : localeService.t("sheets-filter-ui.sync.statusTips.on"),
				asChild: true,
				children: /* @__PURE__ */ jsx(InfoIcon, { className: "univer-block" })
			})]
		}), /* @__PURE__ */ jsx(Switch, {
			defaultChecked: enabled,
			onChange: (checked) => {
				const message = checked ? localeService.t("sheets-filter-ui.sync.switchTips.on") : localeService.t("sheets-filter-ui.sync.switchTips.off");
				sheetsFilterSyncController.setEnabled(checked);
				messageService.show({
					content: message,
					type: MessageType.Success,
					duration: 2e3
				});
			}
		})]
	});
}

//#endregion
//#region src/views/components/SheetsFilterPanel.tsx
/**
* This Filter Panel component is used to filter the data in the sheet.
*
* @returns React element
*/
function FilterPanel() {
	var _filterService$active;
	const sheetsFilterPanelService = useDependency(SheetsFilterPanelService);
	const localeService = useDependency(LocaleService);
	const commandService = useDependency(ICommandService);
	const filterBy = useObservable(sheetsFilterPanelService.filterBy$, void 0, true);
	const filterByModel = useObservable(sheetsFilterPanelService.filterByModel$, void 0, false);
	const canApply = useObservable(() => (filterByModel === null || filterByModel === void 0 ? void 0 : filterByModel.canApply$) || of(false), void 0, false, [filterByModel]);
	const items = useFilterByOptions(localeService);
	const clearFilterDisabled = !useObservable(sheetsFilterPanelService.hasCriteria$);
	const onFilterByTypeChange = useCallback((value) => {
		commandService.executeCommand(ChangeFilterByOperation.id, { filterBy: value });
	}, [commandService]);
	const onClearCriteria = useCallback(async () => {
		await (filterByModel === null || filterByModel === void 0 ? void 0 : filterByModel.clear());
		commandService.executeCommand(CloseFilterPanelOperation.id);
	}, [filterByModel, commandService]);
	const onCancel = useCallback(() => {
		commandService.executeCommand(CloseFilterPanelOperation.id);
	}, [commandService]);
	const onApply = useCallback(async () => {
		await (filterByModel === null || filterByModel === void 0 ? void 0 : filterByModel.apply());
		commandService.executeCommand(CloseFilterPanelOperation.id);
	}, [filterByModel, commandService]);
	const range = (_filterService$active = useDependency(SheetsFilterService).activeFilterModel) === null || _filterService$active === void 0 ? void 0 : _filterService$active.getRange();
	const colIndex = sheetsFilterPanelService.col;
	return /* @__PURE__ */ jsxs("div", {
		"data-u-comp": "sheets-filter-panel",
		className: "univer-box-border univer-flex univer-max-h-[500px] univer-w-[400px] univer-flex-col univer-rounded-lg univer-bg-white univer-p-4 univer-shadow-lg dark:!univer-border-gray-600 dark:!univer-bg-gray-700",
		children: [
			/* @__PURE__ */ jsx(ComponentContainer, {
				components: useComponentsOfPart(SheetsUIPart.FILTER_PANEL_EMBED_POINT),
				sharedProps: {
					range,
					colIndex,
					onClose: onCancel
				}
			}),
			/* @__PURE__ */ jsx("div", {
				className: "univer-mb-1 univer-flex-shrink-0 univer-flex-grow-0",
				children: /* @__PURE__ */ jsx(Segmented, {
					value: filterBy,
					items,
					onChange: (value) => onFilterByTypeChange(value)
				})
			}),
			filterByModel ? /* @__PURE__ */ jsx("div", {
				"data-u-comp": "sheets-filter-panel-content",
				className: "univer-flex-shrink univer-flex-grow univer-pt-2",
				children: filterBy === FilterBy.VALUES ? /* @__PURE__ */ jsx(FilterByValue, { model: filterByModel }) : filterBy === FilterBy.COLORS ? /* @__PURE__ */ jsx(FilterByColor, { model: filterByModel }) : /* @__PURE__ */ jsx(FilterByCondition, { model: filterByModel })
			}) : /* @__PURE__ */ jsx("div", { className: "univer-flex-1" }),
			/* @__PURE__ */ jsx(FilterSyncSwitch, {}),
			/* @__PURE__ */ jsxs("div", {
				"data-u-comp": "sheets-filter-panel-footer",
				className: "univer-mt-4 univer-inline-flex univer-flex-shrink-0 univer-flex-grow-0 univer-flex-nowrap univer-justify-between univer-overflow-hidden",
				children: [/* @__PURE__ */ jsx(Button, {
					variant: "link",
					onClick: onClearCriteria,
					disabled: clearFilterDisabled,
					children: localeService.t("sheets-filter-ui.panel.clear-filter")
				}), /* @__PURE__ */ jsxs("span", {
					className: "univer-flex univer-gap-2",
					children: [/* @__PURE__ */ jsx(Button, {
						variant: "default",
						onClick: onCancel,
						children: localeService.t("sheets-filter-ui.panel.cancel")
					}), /* @__PURE__ */ jsx(Button, {
						disabled: !canApply,
						variant: "primary",
						onClick: onApply,
						children: localeService.t("sheets-filter-ui.panel.confirm")
					})]
				})]
			})
		]
	});
}
function useFilterByOptions(localeService) {
	return useMemo(() => [
		{
			label: localeService.t("sheets-filter-ui.panel.by-values"),
			value: FilterBy.VALUES
		},
		{
			label: localeService.t("sheets-filter-ui.panel.by-colors"),
			value: FilterBy.COLORS
		},
		{
			label: localeService.t("sheets-filter-ui.panel.by-conditions"),
			value: FilterBy.CONDITIONS
		}
	], [localeService.getCurrentLocale(), localeService]);
}

//#endregion
//#region src/controllers/sheets-filter.shortcut.ts
const SmartToggleFilterShortcut = {
	id: SmartToggleSheetsFilterCommand.id,
	binding: KeyCode.L | MetaKeys.CTRL_COMMAND | MetaKeys.SHIFT,
	description: "sheets-filter-ui.shortcut.smart-toggle-filter",
	preconditions: whenSheetEditorFocused,
	group: "4_sheet-edit",
	groupTitle: "sheets-ui.shortcut.sheet-edit"
};

//#endregion
//#region src/controllers/sheets-filter-ui-desktop.controller.ts
const FILTER_PANEL_POPUP_KEY = "FILTER_PANEL_POPUP";
let SheetsFilterUIDesktopController = class SheetsFilterUIDesktopController extends SheetsFilterUIMobileController {
	constructor(_injector, _componentManager, _sheetsFilterPanelService, _sheetCanvasPopupService, _sheetsFilterService, _localeService, _shortcutService, _commandService, _menuManagerService, _contextService, _messageService, sheetsRenderService, renderManagerService) {
		super(renderManagerService, sheetsRenderService);
		this._injector = _injector;
		this._componentManager = _componentManager;
		this._sheetsFilterPanelService = _sheetsFilterPanelService;
		this._sheetCanvasPopupService = _sheetCanvasPopupService;
		this._sheetsFilterService = _sheetsFilterService;
		this._localeService = _localeService;
		this._shortcutService = _shortcutService;
		this._commandService = _commandService;
		this._menuManagerService = _menuManagerService;
		this._contextService = _contextService;
		this._messageService = _messageService;
		_defineProperty(this, "_popupDisposable", void 0);
		this._initCommands();
		this._initShortcuts();
		this._initMenuItems();
		this._initUI();
	}
	dispose() {
		super.dispose();
		this._closeFilterPopup();
	}
	_initShortcuts() {
		[SmartToggleFilterShortcut].forEach((shortcut) => {
			this.disposeWithMe(this._shortcutService.registerShortcut(shortcut));
		});
	}
	_initCommands() {
		[
			SmartToggleSheetsFilterCommand,
			ChangeFilterByOperation,
			OpenFilterPanelOperation,
			CloseFilterPanelOperation
		].forEach((c) => {
			this.disposeWithMe(this._commandService.registerCommand(c));
		});
	}
	_initMenuItems() {
		this._menuManagerService.mergeMenu(menuSchema);
	}
	_initUI() {
		[[FILTER_PANEL_POPUP_KEY, FilterPanel], ["FilterIcon", FilterIcon]].forEach(([key, comp]) => {
			this.disposeWithMe(this._componentManager.register(key, comp));
		});
		this.disposeWithMe(this._contextService.subscribeContextValue$(FILTER_PANEL_OPENED_KEY).pipe(distinctUntilChanged()).subscribe((open) => {
			if (open) this._openFilterPopup();
			else this._closeFilterPopup();
		}));
		this.disposeWithMe(this._sheetsFilterService.errorMsg$.subscribe((content) => {
			if (content) this._messageService.show({
				type: MessageType.Error,
				content: this._localeService.t(content)
			});
		}));
	}
	_openFilterPopup() {
		const currentFilterModel = this._sheetsFilterPanelService.filterModel;
		if (!currentFilterModel) throw new Error("[SheetsFilterUIController]: no filter model when opening filter popup!");
		const range = currentFilterModel.getRange();
		const col = this._sheetsFilterPanelService.col;
		const { startRow } = range;
		this._popupDisposable = this._sheetCanvasPopupService.attachPopupToCell(startRow, col, {
			componentKey: FILTER_PANEL_POPUP_KEY,
			direction: "horizontal",
			onClickOutside: () => this._commandService.syncExecuteCommand(CloseFilterPanelOperation.id),
			offset: [5, 0]
		});
	}
	_closeFilterPopup() {
		var _this$_popupDisposabl;
		(_this$_popupDisposabl = this._popupDisposable) === null || _this$_popupDisposabl === void 0 || _this$_popupDisposabl.dispose();
		this._popupDisposable = null;
	}
};
SheetsFilterUIDesktopController = __decorate([
	__decorateParam(0, Inject(Injector)),
	__decorateParam(1, Inject(ComponentManager)),
	__decorateParam(2, Inject(SheetsFilterPanelService)),
	__decorateParam(3, Inject(SheetCanvasPopManagerService)),
	__decorateParam(4, Inject(SheetsFilterService)),
	__decorateParam(5, Inject(LocaleService)),
	__decorateParam(6, IShortcutService),
	__decorateParam(7, ICommandService),
	__decorateParam(8, IMenuManagerService),
	__decorateParam(9, IContextService),
	__decorateParam(10, IMessageService),
	__decorateParam(11, Inject(SheetsRenderService)),
	__decorateParam(12, IRenderManagerService)
], SheetsFilterUIDesktopController);

//#endregion
//#region src/plugin.ts
let UniverSheetsFilterUIPlugin = class UniverSheetsFilterUIPlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _injector, _configService, _rpcChannelService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		this._rpcChannelService = _rpcChannelService;
		const { menu, ...rest } = merge({}, defaultPluginConfig, this._config);
		if (menu) this._configService.setConfig("menu", menu, { merge: true });
		this._configService.setConfig(SHEETS_FILTER_UI_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		registerDependencies(this._injector, [
			[SheetsFilterPanelService],
			[SheetsFilterPermissionController],
			[SheetsFilterUIDesktopController]
		]);
		if (this._config.useRemoteFilterValuesGenerator && this._rpcChannelService) this._injector.add([ISheetsGenerateFilterValuesService, { useFactory: () => toModule(this._rpcChannelService.requestChannel(SHEETS_GENERATE_FILTER_VALUES_SERVICE_NAME)) }]);
	}
	onReady() {
		touchDependencies(this._injector, [[SheetsFilterPermissionController]]);
	}
	onRendered() {
		touchDependencies(this._injector, [[SheetsFilterUIDesktopController]]);
	}
};
_defineProperty(UniverSheetsFilterUIPlugin, "type", UniverInstanceType.UNIVER_SHEET);
_defineProperty(UniverSheetsFilterUIPlugin, "pluginName", "SHEET_FILTER_UI_PLUGIN");
_defineProperty(UniverSheetsFilterUIPlugin, "packageName", name);
_defineProperty(UniverSheetsFilterUIPlugin, "version", version);
UniverSheetsFilterUIPlugin = __decorate([
	DependentOn(UniverSheetsFilterPlugin),
	__decorateParam(1, Inject(Injector)),
	__decorateParam(2, IConfigService),
	__decorateParam(3, Optional(IRPCChannelService))
], UniverSheetsFilterUIPlugin);

//#endregion
//#region src/worker/plugin.ts
let UniverSheetsFilterUIWorkerPlugin = class UniverSheetsFilterUIWorkerPlugin extends Plugin {
	constructor(_config, _injector, _rpcChannelService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._rpcChannelService = _rpcChannelService;
	}
	onStarting() {
		[[ISheetsGenerateFilterValuesService, { useClass: SheetsGenerateFilterValuesService }]].forEach((d) => this._injector.add(d));
	}
	onReady() {
		this._rpcChannelService.registerChannel(SHEETS_GENERATE_FILTER_VALUES_SERVICE_NAME, fromModule(this._injector.get(ISheetsGenerateFilterValuesService)));
	}
};
_defineProperty(UniverSheetsFilterUIWorkerPlugin, "type", UniverInstanceType.UNIVER_SHEET);
_defineProperty(UniverSheetsFilterUIWorkerPlugin, "pluginName", "SHEET_FILTER_UI_WORKER_PLUGIN");
_defineProperty(UniverSheetsFilterUIWorkerPlugin, "packageName", name);
_defineProperty(UniverSheetsFilterUIWorkerPlugin, "version", version);
UniverSheetsFilterUIWorkerPlugin = __decorate([__decorateParam(1, Inject(Injector)), __decorateParam(2, IRPCChannelService)], UniverSheetsFilterUIWorkerPlugin);

//#endregion
export { ChangeFilterByOperation, CloseFilterPanelOperation, OpenFilterPanelOperation, UniverSheetsFilterMobileUIPlugin, UniverSheetsFilterUIPlugin, UniverSheetsFilterUIWorkerPlugin };
