// @univerjs/sheets-sort-ui/lib/index.js
import { CommandType, DependentOn, Disposable, ICommandService, IConfigService, IConfirmService, IUniverInstanceService, Inject, Injector, LocaleService, LocaleType, Plugin, RxDisposable, Tools, UniverInstanceType, merge, throttle } from "@univerjs/core";
import { RangeProtectionPermissionEditPoint, SetSelectionsOperation, SheetsSelectionsService, WorkbookEditablePermission, WorksheetEditPermission, WorksheetSortPermission, expandToContinuousRange, getPrimaryForRange, getSheetCommandTarget } from "@univerjs/sheets";
import { SheetsSortService, SortRangeCommand, SortType, UniverSheetsSortPlugin } from "@univerjs/sheets-sort";
import { BehaviorSubject, takeUntil } from "rxjs";
import { Button, ButtonGroup, Checkbox, DraggableList, Dropdown, Radio, RadioGroup, clsx, scrollbarClassName } from "@univerjs/design";
import { ComponentManager, ContextMenuGroup, ContextMenuPosition, IDialogService, ILayoutService, IMenuManagerService, IUIPartsService, MenuItemType, RibbonDataGroup, connectInjector, getMenuHiddenObservable, useDependency, useObservable } from "@univerjs/ui";
import { useCallback, useEffect, useRef, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
import { serializeRange } from "@univerjs/engine-formula";
import { AscendingIcon, CheckMarkIcon, CustomSortIcon, DeleteEmptyIcon, DescendingIcon, ExpandAscendingIcon, ExpandDescendingIcon, IncreaseIcon, MoreDownIcon, SequenceIcon } from "@univerjs/icons";
import { SheetsRenderService, SheetsUIPart, getCurrentExclusiveRangeInterest$, getCurrentRangeDisable$ } from "@univerjs/sheets-ui";

//#region src/views/ExtendConfirm.tsx
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const ExtendConfirm = (props) => {
	const [extend, setExtend] = useState("0");
	const localeService = useDependency(LocaleService);
	return /* @__PURE__ */ jsxs("div", {
		className: "univer-text-sm",
		children: [/* @__PURE__ */ jsx("div", {
			className: "extend-confirm-desc",
			children: localeService.t("sheets-sort-ui.dialog.sort-reminder-desc")
		}), /* @__PURE__ */ jsxs(RadioGroup, {
			className: "univer-mt-4",
			value: extend,
			direction: "vertical",
			onChange: (value) => {
				setExtend(value);
				props.onChange(value);
			},
			children: [/* @__PURE__ */ jsx(Radio, {
				value: "0",
				children: localeService.t("sheets-sort-ui.dialog.sort-reminder-no")
			}), /* @__PURE__ */ jsx(Radio, {
				value: "1",
				children: localeService.t("sheets-sort-ui.dialog.sort-reminder-ext")
			})]
		})]
	});
};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/services/sheets-sort-ui.service.tsx
const SORT_ERROR_MESSAGE = {
	MERGE_ERROR: "sheets-sort-ui.error.merge-size",
	EMPTY_ERROR: "sheets-sort-ui.error.empty",
	SINGLE_ERROR: "sheets-sort-ui.error.single",
	FORMULA_ARRAY: "sheets-sort-ui.error.formula-array"
};
let SheetsSortUIService = class SheetsSortUIService extends Disposable {
	constructor(_univerInstanceService, _confirmService, _selectionManagerService, _sheetsSortService, _localeService, _commandService) {
		super();
		this._univerInstanceService = _univerInstanceService;
		this._confirmService = _confirmService;
		this._selectionManagerService = _selectionManagerService;
		this._sheetsSortService = _sheetsSortService;
		this._localeService = _localeService;
		this._commandService = _commandService;
		_defineProperty(this, "_customSortState$", new BehaviorSubject(null));
		_defineProperty(this, "customSortState$", this._customSortState$.asObservable());
	}
	async triggerSortDirectly(asc, extend, sheetRangeLocation) {
		const location = sheetRangeLocation || await this._detectSortLocation(extend);
		if (!location) return false;
		if (!this._check(location)) return false;
		const sortOption = {
			orderRules: [{
				type: asc ? SortType.ASC : SortType.DESC,
				colIndex: location.colIndex
			}],
			range: location.range
		};
		this._sheetsSortService.applySort(sortOption, location.unitId, location.subUnitId);
		return true;
	}
	async triggerSortCustomize() {
		const location = await this._detectSortLocation();
		if (!location) return false;
		if (!this._check(location)) return false;
		this.showCustomSortPanel(location);
		return true;
	}
	customSortState() {
		return this._customSortState$.getValue();
	}
	getTitles(hasTitle) {
		var _this$customSortState, _this$_univerInstance;
		const location = (_this$customSortState = this.customSortState()) === null || _this$customSortState === void 0 ? void 0 : _this$customSortState.location;
		if (!location) return [];
		const { unitId, subUnitId, range } = location;
		const worksheet = (_this$_univerInstance = this._univerInstanceService.getUnit(unitId)) === null || _this$_univerInstance === void 0 ? void 0 : _this$_univerInstance.getSheetBySheetId(subUnitId);
		if (!worksheet) return [];
		const colTranslator = colIndexTranslator(this._localeService);
		return Array.from({ length: range.endColumn - range.startColumn + 1 }, (_, i) => {
			var _worksheet$getCell;
			const cellValue = (_worksheet$getCell = worksheet.getCell(range.startRow, i + range.startColumn)) === null || _worksheet$getCell === void 0 ? void 0 : _worksheet$getCell.v;
			return {
				index: i + range.startColumn,
				label: hasTitle ? `${cellValue !== null && cellValue !== void 0 ? cellValue : colTranslator(i + range.startColumn)}` : colTranslator(i + range.startColumn)
			};
		});
	}
	setSelection(unitId, subUnitId, range) {
		var _this$_univerInstance2;
		const worksheet = (_this$_univerInstance2 = this._univerInstanceService.getUnit(unitId)) === null || _this$_univerInstance2 === void 0 ? void 0 : _this$_univerInstance2.getSheetBySheetId(subUnitId);
		if (!worksheet) return;
		const setSelectionsOperationParams = {
			unitId,
			subUnitId,
			selections: [{
				range,
				primary: getPrimaryForRange(range, worksheet),
				style: null
			}]
		};
		this._commandService.executeCommand(SetSelectionsOperation.id, setSelectionsOperationParams);
	}
	async showCheckError(content) {
		return await this._confirmService.confirm({
			id: "sort-range-check-error",
			title: { title: this._localeService.t("sheets-sort-ui.info.tooltip") },
			children: { title: /* @__PURE__ */ jsx("div", { children: this._localeService.t(content) }) },
			cancelText: this._localeService.t("sheets-sort-ui.dialog.cancel"),
			confirmText: this._localeService.t("sheets-sort-ui.dialog.confirm")
		});
	}
	async showExtendConfirm() {
		let shouldExtend = false;
		if (await this._confirmService.confirm({
			id: "extend-sort-range-dialog",
			title: { title: this._localeService.t("sheets-sort-ui.dialog.sort-reminder") },
			children: { title: /* @__PURE__ */ jsx(ExtendConfirm, { onChange: (value) => {
				shouldExtend = value === "1";
			} }) },
			width: 400,
			cancelText: this._localeService.t("sheets-sort-ui.dialog.cancel"),
			confirmText: this._localeService.t("sheets-sort-ui.dialog.confirm")
		})) return shouldExtend ? "extend" : "keep";
		return "cancel";
	}
	showCustomSortPanel(location) {
		this._customSortState$.next({
			location,
			show: true
		});
	}
	closeCustomSortPanel() {
		this._customSortState$.next({ show: false });
	}
	_check(location) {
		if (!this._sheetsSortService.singleCheck(location)) {
			this.showCheckError(SORT_ERROR_MESSAGE.SINGLE_ERROR);
			return false;
		}
		if (!this._sheetsSortService.mergeCheck(location)) {
			this.showCheckError(SORT_ERROR_MESSAGE.MERGE_ERROR);
			return false;
		}
		if (!this._sheetsSortService.formulaCheck(location)) {
			this.showCheckError(SORT_ERROR_MESSAGE.FORMULA_ARRAY);
			return false;
		}
		if (!this._sheetsSortService.emptyCheck(location)) {
			this.showCheckError(SORT_ERROR_MESSAGE.EMPTY_ERROR);
			return false;
		}
		return true;
	}
	async _detectSortLocation(extend) {
		const workbook = this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_SHEET);
		const worksheet = workbook.getActiveSheet();
		const unitId = workbook.getUnitId();
		const subUnitId = worksheet.getSheetId();
		const selection = this._selectionManagerService.getCurrentLastSelection();
		if (!selection) return null;
		let range;
		if (extend === true) {
			range = expandToContinuousRange(selection.range, {
				up: true,
				down: true,
				left: true,
				right: true
			}, worksheet);
			this.setSelection(unitId, subUnitId, range);
		} else if (extend === false) range = selection.range;
		else {
			const confirmRes = await this.showExtendConfirm();
			if (confirmRes === "cancel") return null;
			if (confirmRes === "keep") range = selection.range;
			else {
				range = expandToContinuousRange(selection.range, {
					up: true,
					down: true,
					left: true,
					right: true
				}, worksheet);
				this.setSelection(unitId, subUnitId, range);
			}
		}
		return {
			range,
			unitId,
			subUnitId,
			colIndex: selection.primary.actualColumn
		};
	}
};
SheetsSortUIService = __decorate([
	__decorateParam(0, IUniverInstanceService),
	__decorateParam(1, IConfirmService),
	__decorateParam(2, Inject(SheetsSelectionsService)),
	__decorateParam(3, Inject(SheetsSortService)),
	__decorateParam(4, Inject(LocaleService)),
	__decorateParam(5, ICommandService)
], SheetsSortUIService);
function colIndexTranslator(localeService) {
	return (colIndex) => {
		const colName = Tools.chatAtABC(colIndex);
		switch (localeService.getCurrentLocale()) {
			case LocaleType.ZH_CN: return `"${colName}"列`;
			case LocaleType.EN_US: return `Column "${colName}"`;
			default: return `Column "${colName}"`;
		}
	};
}

//#endregion
//#region src/commands/commands/sheets-sort.command.ts
const SortRangeAscCommand = {
	id: "sheet.command.sort-range-asc",
	type: CommandType.COMMAND,
	handler: async (accessor) => {
		return await accessor.get(SheetsSortUIService).triggerSortDirectly(true, false);
	}
};
const SortRangeAscExtCommand = {
	id: "sheet.command.sort-range-asc-ext",
	type: CommandType.COMMAND,
	handler: async (accessor) => {
		return await accessor.get(SheetsSortUIService).triggerSortDirectly(true, true);
	}
};
const SortRangeDescCommand = {
	id: "sheet.command.sort-range-desc",
	type: CommandType.COMMAND,
	handler: async (accessor) => {
		return await accessor.get(SheetsSortUIService).triggerSortDirectly(false, false);
	}
};
const SortRangeDescExtCommand = {
	id: "sheet.command.sort-range-desc-ext",
	type: CommandType.COMMAND,
	handler: async (accessor) => {
		return await accessor.get(SheetsSortUIService).triggerSortDirectly(false, true);
	}
};
const SortRangeCustomCommand = {
	id: "sheet.command.sort-range-custom",
	type: CommandType.COMMAND,
	handler: async (accessor) => {
		return await accessor.get(SheetsSortUIService).triggerSortCustomize();
	}
};
const SortRangeAscInCtxMenuCommand = {
	id: "sheet.command.sort-range-asc-ctx",
	type: CommandType.COMMAND,
	handler: async (accessor) => {
		return await accessor.get(SheetsSortUIService).triggerSortDirectly(true, false);
	}
};
const SortRangeAscExtInCtxMenuCommand = {
	id: "sheet.command.sort-range-asc-ext-ctx",
	type: CommandType.COMMAND,
	handler: async (accessor) => {
		return await accessor.get(SheetsSortUIService).triggerSortDirectly(true, true);
	}
};
const SortRangeDescInCtxMenuCommand = {
	id: "sheet.command.sort-range-desc-ctx",
	type: CommandType.COMMAND,
	handler: async (accessor) => {
		return await accessor.get(SheetsSortUIService).triggerSortDirectly(false, false);
	}
};
const SortRangeDescExtInCtxMenuCommand = {
	id: "sheet.command.sort-range-desc-ext-ctx",
	type: CommandType.COMMAND,
	handler: async (accessor) => {
		return await accessor.get(SheetsSortUIService).triggerSortDirectly(false, true);
	}
};
const SortRangeCustomInCtxMenuCommand = {
	id: "sheet.command.sort-range-custom-ctx",
	type: CommandType.COMMAND,
	handler: async (accessor) => {
		return await accessor.get(SheetsSortUIService).triggerSortCustomize();
	}
};

//#endregion
//#region package.json
var name = "@univerjs/sheets-sort-ui";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
const SHEETS_SORT_UI_PLUGIN_CONFIG_KEY = "sheets-sort-ui.config";
const configSymbol = Symbol(SHEETS_SORT_UI_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/menu/sheets-sort.menu.ts
const SHEETS_SORT_MENU_ID = "sheet.menu.sheets-sort";
const SHEETS_SORT_CTX_MENU_ID = "sheet.menu.sheets-sort-ctx";
const SHEETS_SORT_ASC_ICON = "AscendingIcon";
const SHEETS_SORT_ASC_EXT_ICON = "ExpandAscendingIcon";
const SHEETS_SORT_DESC_ICON = "DescendingIcon";
const SHEETS_SORT_DESC_EXT_ICON = "ExpandDescendingIcon";
const SHEETS_SORT_CUSTOM_ICON = "CustomSortIcon";
function sortRangeMenuFactory(accessor) {
	return {
		id: SHEETS_SORT_MENU_ID,
		type: MenuItemType.SUBITEMS,
		icon: SHEETS_SORT_ASC_ICON,
		tooltip: "sheets-sort-ui.general.sort",
		hidden$: getMenuHiddenObservable(accessor, UniverInstanceType.UNIVER_SHEET),
		disabled$: getCurrentRangeDisable$(accessor, {
			workbookTypes: [WorkbookEditablePermission],
			worksheetTypes: [WorksheetSortPermission, WorksheetEditPermission],
			rangeTypes: [RangeProtectionPermissionEditPoint]
		})
	};
}
function sortRangeAscMenuFactory(_accessor) {
	return {
		id: SortRangeAscCommand.id,
		icon: SHEETS_SORT_ASC_ICON,
		title: "sheets-sort-ui.general.sort-asc-cur",
		type: MenuItemType.BUTTON,
		hidden$: getCurrentExclusiveRangeInterest$(_accessor)
	};
}
function sortRangeAscExtMenuFactory(_accessor) {
	return {
		id: SortRangeAscExtCommand.id,
		title: "sheets-sort-ui.general.sort-asc-ext",
		icon: SHEETS_SORT_ASC_EXT_ICON,
		type: MenuItemType.BUTTON
	};
}
function sortRangeDescMenuFactory(_accessor) {
	return {
		id: SortRangeDescCommand.id,
		title: "sheets-sort-ui.general.sort-desc-cur",
		icon: SHEETS_SORT_DESC_ICON,
		type: MenuItemType.BUTTON
	};
}
function sortRangeDescExtMenuFactory(_accessor) {
	return {
		id: SortRangeDescExtCommand.id,
		title: "sheets-sort-ui.general.sort-desc-ext",
		icon: SHEETS_SORT_DESC_EXT_ICON,
		type: MenuItemType.BUTTON
	};
}
function sortRangeCustomMenuFactory(_accessor) {
	return {
		id: SortRangeCustomCommand.id,
		title: "sheets-sort-ui.general.sort-custom",
		type: MenuItemType.BUTTON,
		icon: SHEETS_SORT_CUSTOM_ICON
	};
}
function sortRangeCtxMenuFactory(accessor) {
	return {
		id: SHEETS_SORT_CTX_MENU_ID,
		title: "sheets-sort-ui.general.sort",
		type: MenuItemType.SUBITEMS,
		icon: SHEETS_SORT_ASC_ICON,
		hidden$: getMenuHiddenObservable(accessor, UniverInstanceType.UNIVER_SHEET),
		disabled$: getCurrentRangeDisable$(accessor, {
			workbookTypes: [WorkbookEditablePermission],
			worksheetTypes: [WorksheetSortPermission, WorksheetEditPermission],
			rangeTypes: [RangeProtectionPermissionEditPoint]
		})
	};
}
function sortRangeAscCtxMenuFactory(_accessor) {
	return {
		id: SortRangeAscInCtxMenuCommand.id,
		title: "sheets-sort-ui.general.sort-asc-cur",
		type: MenuItemType.BUTTON,
		icon: SHEETS_SORT_ASC_ICON,
		disabled$: getCurrentRangeDisable$(_accessor, {
			workbookTypes: [WorkbookEditablePermission],
			worksheetTypes: [WorksheetSortPermission, WorksheetEditPermission],
			rangeTypes: [RangeProtectionPermissionEditPoint]
		})
	};
}
function sortRangeAscExtCtxMenuFactory(_accessor) {
	return {
		id: SortRangeAscExtInCtxMenuCommand.id,
		title: "sheets-sort-ui.general.sort-asc-ext",
		type: MenuItemType.BUTTON,
		icon: SHEETS_SORT_ASC_EXT_ICON,
		disabled$: getCurrentRangeDisable$(_accessor, {
			workbookTypes: [WorkbookEditablePermission],
			worksheetTypes: [WorksheetSortPermission, WorksheetEditPermission],
			rangeTypes: [RangeProtectionPermissionEditPoint]
		})
	};
}
function sortRangeDescCtxMenuFactory(_accessor) {
	return {
		id: SortRangeDescInCtxMenuCommand.id,
		title: "sheets-sort-ui.general.sort-desc-cur",
		type: MenuItemType.BUTTON,
		icon: SHEETS_SORT_DESC_ICON,
		disabled$: getCurrentRangeDisable$(_accessor, {
			workbookTypes: [WorkbookEditablePermission],
			worksheetTypes: [WorksheetSortPermission, WorksheetEditPermission],
			rangeTypes: [RangeProtectionPermissionEditPoint]
		})
	};
}
function sortRangeDescExtCtxMenuFactory(_accessor) {
	return {
		id: SortRangeDescExtInCtxMenuCommand.id,
		title: "sheets-sort-ui.general.sort-desc-ext",
		type: MenuItemType.BUTTON,
		icon: SHEETS_SORT_DESC_EXT_ICON,
		disabled$: getCurrentRangeDisable$(_accessor, {
			workbookTypes: [WorkbookEditablePermission],
			worksheetTypes: [WorksheetSortPermission, WorksheetEditPermission],
			rangeTypes: [RangeProtectionPermissionEditPoint]
		})
	};
}
function sortRangeCustomCtxMenuFactory(_accessor) {
	return {
		id: SortRangeCustomInCtxMenuCommand.id,
		title: "sheets-sort-ui.general.sort-custom",
		type: MenuItemType.BUTTON,
		icon: SHEETS_SORT_CUSTOM_ICON,
		disabled$: getCurrentRangeDisable$(_accessor, {
			workbookTypes: [WorkbookEditablePermission],
			worksheetTypes: [WorksheetSortPermission, WorksheetEditPermission],
			rangeTypes: [RangeProtectionPermissionEditPoint]
		})
	};
}

//#endregion
//#region src/menu/schema.ts
const menuSchema = {
	[RibbonDataGroup.ORGANIZATION]: { [SHEETS_SORT_MENU_ID]: {
		order: 3,
		menuItemFactory: sortRangeMenuFactory,
		[SortRangeAscCommand.id]: {
			order: 0,
			menuItemFactory: sortRangeAscMenuFactory
		},
		[SortRangeAscExtCommand.id]: {
			order: 1,
			menuItemFactory: sortRangeAscExtMenuFactory
		},
		[SortRangeDescCommand.id]: {
			order: 2,
			menuItemFactory: sortRangeDescMenuFactory
		},
		[SortRangeDescExtCommand.id]: {
			order: 3,
			menuItemFactory: sortRangeDescExtMenuFactory
		},
		[SortRangeCustomCommand.id]: {
			order: 4,
			menuItemFactory: sortRangeCustomMenuFactory
		}
	} },
	[ContextMenuPosition.MAIN_AREA]: { [ContextMenuGroup.DATA]: { [SHEETS_SORT_CTX_MENU_ID]: {
		order: 0,
		menuItemFactory: sortRangeCtxMenuFactory,
		[SortRangeAscInCtxMenuCommand.id]: {
			order: 0,
			menuItemFactory: sortRangeAscCtxMenuFactory
		},
		[SortRangeAscExtInCtxMenuCommand.id]: {
			order: 1,
			menuItemFactory: sortRangeAscExtCtxMenuFactory
		},
		[SortRangeDescInCtxMenuCommand.id]: {
			order: 2,
			menuItemFactory: sortRangeDescCtxMenuFactory
		},
		[SortRangeDescExtInCtxMenuCommand.id]: {
			order: 3,
			menuItemFactory: sortRangeDescExtCtxMenuFactory
		},
		[SortRangeCustomInCtxMenuCommand.id]: {
			order: 4,
			menuItemFactory: sortRangeCustomCtxMenuFactory
		}
	} } }
};

//#endregion
//#region src/views/CustomSortPanel.tsx
function CustomSortPanel() {
	const state = useObservable(useDependency(SheetsSortUIService).customSortState$);
	if (!state || !state.location) return null;
	return /* @__PURE__ */ jsx(CustomSortPanelImpl, { state });
}
function CustomSortPanelImpl({ state }) {
	const sheetsSortService = useDependency(SheetsSortService);
	const localeService = useDependency(LocaleService);
	const sheetsSortUIService = useDependency(SheetsSortUIService);
	const [hasTitle, setHasTitle] = useState(false);
	const [scrollPosition, setScrollPosition] = useState(0);
	const listEndRef = useRef(null);
	const { range, unitId, subUnitId } = state.location;
	const titles = sheetsSortUIService.getTitles(hasTitle);
	const [list, setList] = useState([{
		type: SortType.ASC,
		colIndex: range.startColumn
	}]);
	const onItemChange = useCallback((index, value) => {
		const newList = [...list];
		if (value === null) newList.splice(index, 1);
		else newList[index] = value;
		setList(newList);
	}, [list]);
	const newItem = useCallback(throttle(() => {
		const newList = [...list];
		const nextColIndex = findNextColIndex(range, list);
		if (nextColIndex !== null) {
			newList.push({
				type: SortType.ASC,
				colIndex: nextColIndex
			});
			setList(newList);
		}
	}, 200), [list, range]);
	const apply = useCallback((orderRules, hasTitle) => {
		sheetsSortService.applySort({
			range,
			orderRules,
			hasTitle
		});
		sheetsSortUIService.closeCustomSortPanel();
	}, [
		sheetsSortService,
		sheetsSortUIService,
		range
	]);
	const cancel = useCallback(() => {
		sheetsSortUIService.closeCustomSortPanel();
	}, [sheetsSortUIService]);
	const setTitle = useCallback((value) => {
		setHasTitle(value);
		if (value) sheetsSortUIService.setSelection(unitId, subUnitId, {
			...range,
			startRow: range.startRow + 1
		});
		else sheetsSortUIService.setSelection(unitId, subUnitId, range);
	}, [
		sheetsSortUIService,
		range,
		subUnitId,
		unitId
	]);
	useEffect(() => {
		if (listEndRef.current && list.length > 5) listEndRef.current.scrollTop = listEndRef.current.scrollHeight;
	}, [list]);
	const canNew = list.length < titles.length;
	const dragList = list.map((item) => ({
		...item,
		id: `${item.colIndex}`
	}));
	return /* @__PURE__ */ jsxs("div", { children: [/* @__PURE__ */ jsxs("div", {
		onMouseDown: (e) => {
			e.stopPropagation();
		},
		children: [/* @__PURE__ */ jsxs("div", {
			className: "univer-mb-2 univer-flex univer-items-center univer-justify-between",
			children: [/* @__PURE__ */ jsx(Checkbox, {
				checked: hasTitle,
				onChange: (value) => setTitle(!!value),
				children: localeService.t("sheets-sort-ui.dialog.first-row-check")
			}), canNew ? /* @__PURE__ */ jsxs("div", {
				className: "univer-flex univer-cursor-pointer univer-select-none univer-items-center univer-text-base",
				onClick: newItem,
				children: [/* @__PURE__ */ jsx(IncreaseIcon, {}), /* @__PURE__ */ jsx("span", {
					className: "univer-ml-1.5",
					children: localeService.t("sheets-sort-ui.dialog.add-condition")
				})]
			}) : /* @__PURE__ */ jsxs("div", {
				className: "univer-flex univer-cursor-pointer univer-select-none univer-items-center univer-text-base univer-text-primary-500 disabled:univer-cursor-not-allowed disabled:univer-divide-opacity-30 disabled:univer-text-gray-800",
				children: [/* @__PURE__ */ jsx(IncreaseIcon, {}), /* @__PURE__ */ jsx("span", {
					className: "univer-ml-1.5 univer-text-xs",
					children: localeService.t("sheets-sort-ui.dialog.add-condition")
				})]
			})]
		}), /* @__PURE__ */ jsx("div", {
			ref: listEndRef,
			className: clsx("univer-max-h-[310px] univer-overflow-y-auto univer-overflow-x-hidden", scrollbarClassName),
			onScroll: (e) => {
				const position = e.currentTarget.scrollTop;
				setScrollPosition(position);
			},
			children: /* @__PURE__ */ jsx(DraggableList, {
				list: dragList,
				onListChange: setList,
				idKey: "id",
				draggableHandle: "[data-u-comp=sort-panel-item-handler]",
				itemRender: (item) => /* @__PURE__ */ jsx(SortOptionItem, {
					titles,
					list: dragList,
					item,
					onChange: (value, index) => onItemChange(index, value),
					scrollPosition
				}),
				rowHeight: 32,
				margin: [0, 12]
			})
		})]
	}), /* @__PURE__ */ jsxs("div", {
		className: "univer-mt-5 univer-flex univer-justify-end",
		children: [/* @__PURE__ */ jsx(Button, {
			className: "univer-ml-3",
			onClick: () => cancel(),
			children: localeService.t("sheets-sort-ui.dialog.cancel")
		}), /* @__PURE__ */ jsx(Button, {
			className: "univer-ml-3",
			variant: "primary",
			onClick: () => apply(list, hasTitle),
			children: localeService.t("sheets-sort-ui.dialog.confirm")
		})]
	})] });
}
function SortOptionItem(props) {
	var _titles$find;
	const { list, item, titles, onChange, scrollPosition } = props;
	const localeService = useDependency(LocaleService);
	const availableMenu = titles.filter((title) => !list.some((item) => item.colIndex === title.index) || title.index === item.colIndex);
	const currentIndex = list.findIndex((listItem) => listItem.colIndex === item.colIndex);
	const handleChangeColIndex = useCallback((menuItem) => {
		onChange({
			...item,
			colIndex: menuItem.index
		}, currentIndex);
		setVisible(false);
	}, [
		currentIndex,
		item,
		onChange
	]);
	const [visible, setVisible] = useState(false);
	const onVisibleChange = (visible) => {
		setVisible(visible);
	};
	useEffect(() => {
		setVisible(false);
	}, [scrollPosition]);
	const showDelete = list.length > 1;
	const itemLabel = (_titles$find = titles.find((title) => title.index === item.colIndex)) === null || _titles$find === void 0 ? void 0 : _titles$find.label;
	return /* @__PURE__ */ jsxs("div", {
		className: "univer-grid univer-grid-flow-col univer-grid-cols-2 univer-items-center univer-gap-2",
		children: [/* @__PURE__ */ jsxs("div", {
			className: "univer-flex univer-items-center",
			children: [/* @__PURE__ */ jsx("div", {
				"data-u-comp": "sort-panel-item-handler",
				className: "univer-flex univer-cursor-pointer univer-items-center univer-justify-center univer-text-base univer-text-gray-700",
				children: /* @__PURE__ */ jsx(SequenceIcon, {})
			}), /* @__PURE__ */ jsx(Dropdown, {
				overlay: /* @__PURE__ */ jsx("ul", {
					className: clsx("univer-my-0 univer-box-border univer-grid univer-max-h-[310px] univer-w-[--radix-popper-anchor-width] univer-items-center univer-gap-1 univer-overflow-y-auto univer-overflow-x-hidden univer-rounded-lg univer-border univer-bg-white univer-p-1 univer-text-base univer-shadow-lg", scrollbarClassName),
					children: availableMenu.map((menuItem) => /* @__PURE__ */ jsxs("li", {
						onClick: () => handleChangeColIndex(menuItem),
						className: "univer-relative univer-box-border univer-flex univer-h-7 univer-cursor-pointer univer-list-none univer-items-center univer-justify-between univer-rounded univer-px-2 univer-text-sm univer-transition-all hover:univer-bg-gray-100 dark:hover:!univer-bg-gray-700",
						children: [/* @__PURE__ */ jsx("span", {
							className: "univer-max-w-[220px] univer-truncate",
							children: menuItem.label
						}), /* @__PURE__ */ jsx("span", { children: menuItem.index === item.colIndex && /* @__PURE__ */ jsx(CheckMarkIcon, {}) })]
					}, menuItem.index))
				}),
				open: visible,
				onOpenChange: onVisibleChange,
				children: /* @__PURE__ */ jsxs("div", {
					className: clsx("univer-ml-2 univer-flex univer-w-full univer-items-center univer-justify-between univer-overflow-hidden univer-rounded-md univer-py-1.5 univer-text-sm univer-text-gray-900 dark:!univer-text-white"),
					children: [/* @__PURE__ */ jsx("span", {
						className: "univer-max-w-[220px] univer-truncate",
						children: itemLabel
					}), /* @__PURE__ */ jsx(MoreDownIcon, {})]
				})
			})]
		}), /* @__PURE__ */ jsxs("div", {
			className: "univer-flex univer-items-center univer-justify-end univer-gap-2",
			children: [/* @__PURE__ */ jsxs(RadioGroup, {
				value: item.type,
				onChange: (value) => {
					onChange({
						...item,
						type: value
					}, currentIndex);
				},
				children: [/* @__PURE__ */ jsx(Radio, {
					value: SortType.ASC,
					children: localeService.t("sheets-sort-ui.general.sort-asc")
				}), /* @__PURE__ */ jsx(Radio, {
					value: SortType.DESC,
					children: localeService.t("sheets-sort-ui.general.sort-desc")
				})]
			}), /* @__PURE__ */ jsx("a", {
				className: "univer-flex univer-cursor-pointer univer-items-center univer-text-sm univer-transition-colors hover:univer-text-red-500",
				onClick: () => onChange(null, currentIndex),
				children: !showDelete && /* @__PURE__ */ jsx(DeleteEmptyIcon, {})
			})]
		})]
	});
}
function findNextColIndex(range, list) {
	const { startColumn, endColumn } = range;
	const used = new Set(list.map((item) => item === null || item === void 0 ? void 0 : item.colIndex));
	for (let i = startColumn; i <= endColumn; i++) if (!used.has(i)) return i;
	return null;
}

//#endregion
//#region src/views/EmbedSortBtn.tsx
function EmbedSortBtn(props) {
	const { range, colIndex, onClose } = props;
	const sheetsSortUIService = useDependency(SheetsSortUIService);
	const univerInstanceService = useDependency(IUniverInstanceService);
	const localeService = useDependency(LocaleService);
	const apply = useCallback((asc) => {
		const { unitId, subUnitId } = getSheetCommandTarget(univerInstanceService) || {};
		if (range && unitId && subUnitId) {
			const noTitleRange = {
				...range,
				startRow: range.startRow + 1
			};
			sheetsSortUIService.triggerSortDirectly(asc, false, {
				unitId,
				subUnitId,
				range: noTitleRange,
				colIndex
			});
		} else throw new Error(`Cannot find the target to sort. unitId: ${unitId}, subUnitId: ${subUnitId}, range: ${range}, colIndex: ${colIndex}`);
		onClose();
	}, [
		range,
		colIndex,
		sheetsSortUIService,
		univerInstanceService,
		onClose
	]);
	return /* @__PURE__ */ jsxs(ButtonGroup, {
		className: "univer-mb-3 univer-w-full univer-grid-cols-2",
		children: [/* @__PURE__ */ jsxs(Button, {
			onClick: () => apply(true),
			children: [/* @__PURE__ */ jsx(AscendingIcon, {}), localeService.t("sheets-sort-ui.general.sort-asc")]
		}), /* @__PURE__ */ jsxs(Button, {
			onClick: () => apply(false),
			children: [/* @__PURE__ */ jsx(DescendingIcon, {}), localeService.t("sheets-sort-ui.general.sort-desc")]
		})]
	});
}

//#endregion
//#region src/controllers/sheets-sort-ui.controller.ts
const CUSTOM_SORT_DIALOG_ID = "custom-sort-dialog";
const CUSTOM_SORT_PANEL_WIDTH = 560;
const CUSTOM_SORT_PANEL_TOP = 128;
let SheetsSortUIController = class SheetsSortUIController extends RxDisposable {
	constructor(_commandService, _menuManagerService, _dialogService, _layoutService, _uiPartsService, _sheetRenderService, _localeService, _sheetsSortUIService, _injector, _componentManager) {
		super();
		this._commandService = _commandService;
		this._menuManagerService = _menuManagerService;
		this._dialogService = _dialogService;
		this._layoutService = _layoutService;
		this._uiPartsService = _uiPartsService;
		this._sheetRenderService = _sheetRenderService;
		this._localeService = _localeService;
		this._sheetsSortUIService = _sheetsSortUIService;
		this._injector = _injector;
		this._componentManager = _componentManager;
		this._initCommands();
		this._initMenu();
		this._initUI();
	}
	_initMenu() {
		this._menuManagerService.mergeMenu(menuSchema);
	}
	_initCommands() {
		[
			SortRangeAscCommand,
			SortRangeAscExtCommand,
			SortRangeDescCommand,
			SortRangeDescExtCommand,
			SortRangeCustomCommand,
			SortRangeAscInCtxMenuCommand,
			SortRangeAscExtInCtxMenuCommand,
			SortRangeDescInCtxMenuCommand,
			SortRangeDescExtInCtxMenuCommand,
			SortRangeCustomInCtxMenuCommand
		].forEach((command) => this.disposeWithMe(this._commandService.registerCommand(command)));
		this.disposeWithMe(this._sheetRenderService.registerSkeletonChangingMutations(SortRangeCommand.id));
	}
	_initUI() {
		this.disposeWithMe(this._uiPartsService.registerComponent(SheetsUIPart.FILTER_PANEL_EMBED_POINT, () => connectInjector(EmbedSortBtn, this._injector)));
		[
			["CustomSortPanel", CustomSortPanel],
			[SHEETS_SORT_ASC_ICON, AscendingIcon],
			[SHEETS_SORT_ASC_EXT_ICON, ExpandAscendingIcon],
			[SHEETS_SORT_DESC_ICON, DescendingIcon],
			[SHEETS_SORT_DESC_EXT_ICON, ExpandDescendingIcon],
			[SHEETS_SORT_CUSTOM_ICON, CustomSortIcon]
		].forEach(([key, comp]) => {
			this.disposeWithMe(this._componentManager.register(key, comp));
		});
		this._sheetsSortUIService.customSortState$.pipe(takeUntil(this.dispose$)).subscribe((newState) => {
			if (newState && newState.show && newState.location) this._openCustomSortPanel(newState.location);
			else if (newState && !(newState === null || newState === void 0 ? void 0 : newState.show)) this._closePanel();
		});
	}
	_openCustomSortPanel(location) {
		this._dialogService.open({
			id: CUSTOM_SORT_DIALOG_ID,
			draggable: true,
			width: CUSTOM_SORT_PANEL_WIDTH,
			title: { title: `${this._localeService.t("sheets-sort-ui.general.sort-custom")}: ${serializeRange(location.range)}` },
			children: { label: "CustomSortPanel" },
			destroyOnClose: true,
			defaultPosition: getCustomSortDialogDefaultPosition(),
			preservePositionOnDestroy: false,
			onClose: () => this._closePanel(),
			mask: true
		});
	}
	_closePanel() {
		this._dialogService.close(CUSTOM_SORT_DIALOG_ID);
		queueMicrotask(() => this._layoutService.focus());
	}
};
SheetsSortUIController = __decorate([
	__decorateParam(0, ICommandService),
	__decorateParam(1, IMenuManagerService),
	__decorateParam(2, IDialogService),
	__decorateParam(3, ILayoutService),
	__decorateParam(4, IUIPartsService),
	__decorateParam(5, Inject(SheetsRenderService)),
	__decorateParam(6, Inject(LocaleService)),
	__decorateParam(7, Inject(SheetsSortUIService)),
	__decorateParam(8, Inject(Injector)),
	__decorateParam(9, Inject(ComponentManager))
], SheetsSortUIController);
function getCustomSortDialogDefaultPosition() {
	return {
		x: (window.innerWidth - CUSTOM_SORT_PANEL_WIDTH) / 2,
		y: CUSTOM_SORT_PANEL_TOP
	};
}

//#endregion
//#region src/plugin.ts
let UniverSheetsSortUIPlugin = class UniverSheetsSortUIPlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _injector, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		const { ...rest } = merge({}, defaultPluginConfig, this._config);
		this._configService.setConfig(SHEETS_SORT_UI_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		[[SheetsSortUIService], [SheetsSortUIController]].forEach((d) => this._injector.add(d));
	}
	onRendered() {
		this._injector.get(SheetsSortUIController);
	}
};
_defineProperty(UniverSheetsSortUIPlugin, "type", UniverInstanceType.UNIVER_SHEET);
_defineProperty(UniverSheetsSortUIPlugin, "pluginName", "SHEET_SORT_UI_PLUGIN");
_defineProperty(UniverSheetsSortUIPlugin, "packageName", name);
_defineProperty(UniverSheetsSortUIPlugin, "version", version);
UniverSheetsSortUIPlugin = __decorate([
	DependentOn(UniverSheetsSortPlugin),
	__decorateParam(1, Inject(Injector)),
	__decorateParam(2, IConfigService)
], UniverSheetsSortUIPlugin);

//#endregion
export { SortRangeAscCommand, SortRangeAscExtCommand, SortRangeAscExtInCtxMenuCommand, SortRangeAscInCtxMenuCommand, SortRangeCustomCommand, SortRangeCustomInCtxMenuCommand, SortRangeDescCommand, SortRangeDescExtCommand, SortRangeDescExtInCtxMenuCommand, SortRangeDescInCtxMenuCommand, UniverSheetsSortUIPlugin };
