// @univerjs/preset-docs-core/lib/es/index.js
import { UniverDocsPlugin as u } from "@univerjs/docs";
export * from "@univerjs/docs";
import { UniverDocsUIPlugin as l } from "@univerjs/docs-ui";
export * from "@univerjs/docs-ui";
import { UniverFormulaEnginePlugin as g } from "@univerjs/engine-formula";
export * from "@univerjs/engine-formula";
import { UniverRenderEnginePlugin as x } from "@univerjs/engine-render";
export * from "@univerjs/engine-render";
import { UniverNetworkPlugin as U } from "@univerjs/network";
export * from "@univerjs/network";
import { UniverUIPlugin as c } from "@univerjs/ui";
export * from "@univerjs/ui";
import "@univerjs/network/lib/facade";
import "@univerjs/docs-ui/lib/facade";
export * from "@univerjs/design";
function F(r = {}) {
  const {
    container: o = "app",
    header: e,
    footer: n,
    toolbar: i,
    ribbonType: t,
    menu: m,
    contextMenu: p,
    disableAutoFocus: f
  } = r;
  return {
    plugins: [
      U,
      u,
      x,
      [c, {
        container: o,
        header: e,
        footer: n,
        toolbar: i,
        ribbonType: t,
        menu: m,
        contextMenu: p,
        disableAutoFocus: f
      }],
      l,
      g
    ]
  };
}
export {
  F as UniverDocsCorePreset
};
