// @univerjs/sheets-hyper-link/lib/cjs/facade.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
let _univerjs_engine_formula = require("@univerjs/engine-formula");
let _univerjs_sheets_hyper_link = require("@univerjs/sheets-hyper-link");
let _univerjs_sheets_facade = require("@univerjs/sheets/facade");
let _univerjs_core = require("@univerjs/core");
let _univerjs_core_facade = require("@univerjs/core/facade");

//#region src/facade/f-workbook.ts
var FWorkbookHyperlinkMixin = class extends _univerjs_sheets_facade.FWorkbook {
	getUrlOfDefineName(name) {
		const definedNameService = this._injector.get(_univerjs_engine_formula.IDefinedNamesService);
		const parserService = this._injector.get(_univerjs_sheets_hyper_link.SheetsHyperLinkParserService);
		const definedName = definedNameService.getValueByName(this._workbook.getUnitId(), name);
		if (!definedName) throw new Error(`Defined name "${name}" does not exist.`);
		if (definedName.formulaOrRefString.startsWith(_univerjs_engine_formula.operatorToken.EQUALS)) throw new Error(`Defined name "${name}" is not a reference range, cannot be converted to hyperlink.`);
		return parserService.buildHyperLink(definedName.id);
	}
	parseSheetHyperlink(hyperlink) {
		return this._injector.get(_univerjs_sheets_hyper_link.SheetsHyperLinkParserService).parseHyperLink(hyperlink);
	}
};
_univerjs_sheets_facade.FWorkbook.extend(FWorkbookHyperlinkMixin);

//#endregion
//#region src/facade/f-worksheet.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
var FWorksheetHyperlinkMixin = class extends _univerjs_sheets_facade.FWorksheet {
	getUrl() {
		return this._injector.get(_univerjs_sheets_hyper_link.SheetsHyperLinkParserService).buildHyperLink(null, this._worksheet.getSheetId());
	}
};
_univerjs_sheets_facade.FWorksheet.extend(FWorksheetHyperlinkMixin);

//#endregion
//#region src/facade/f-range.ts
var FRangeSheetsHyperlinkMixin = class extends _univerjs_sheets_facade.FRange {
	setHyperLink(url, label) {
		return this._commandService.executeCommand(_univerjs_sheets_hyper_link.AddHyperLinkCommand.id, {
			unitId: this.getUnitId(),
			subUnitId: this._worksheet.getSheetId(),
			link: {
				id: (0, _univerjs_core.generateRandomId)(),
				row: this._range.startRow,
				column: this._range.startColumn,
				payload: url,
				display: label
			}
		});
	}
	getHyperLinks() {
		const hyperlinks = [];
		this.forEach((row, column, cell) => {
			var _cell$p$body$customRa, _cell$p$body;
			if (!cell.p) return;
			const ranges = (_cell$p$body$customRa = (_cell$p$body = cell.p.body) === null || _cell$p$body === void 0 || (_cell$p$body = _cell$p$body.customRanges) === null || _cell$p$body === void 0 ? void 0 : _cell$p$body.filter((range) => range.rangeType === _univerjs_core.CustomRangeType.HYPERLINK)) !== null && _cell$p$body$customRa !== void 0 ? _cell$p$body$customRa : [];
			if (ranges.length > 0) {
				var _cell$p, _properties$url, _dataStream$slice$rep;
				const dataStream = (_cell$p = cell.p) === null || _cell$p === void 0 || (_cell$p = _cell$p.body) === null || _cell$p === void 0 ? void 0 : _cell$p.dataStream;
				const { rangeId, properties, startIndex, endIndex } = ranges[0];
				const url = (_properties$url = properties === null || properties === void 0 ? void 0 : properties.url) !== null && _properties$url !== void 0 ? _properties$url : "";
				const label = (_dataStream$slice$rep = dataStream === null || dataStream === void 0 ? void 0 : dataStream.slice(startIndex, endIndex + 1).replaceAll(_univerjs_core.DataStreamTreeTokenType.CUSTOM_RANGE_START, "").replaceAll(_univerjs_core.DataStreamTreeTokenType.CUSTOM_RANGE_END, "")) !== null && _dataStream$slice$rep !== void 0 ? _dataStream$slice$rep : "";
				hyperlinks.push({
					id: `${rangeId}`,
					row,
					column,
					url,
					label
				});
			}
		});
		return hyperlinks;
	}
	updateHyperLink(url, label) {
		const hyperlink = this.getHyperLinks().find((link) => link.row === this._range.startRow && link.column === this._range.startColumn);
		if (!hyperlink) return Promise.reject(/* @__PURE__ */ new Error("No hyperlink found in the top left cell of the range"));
		const { id, row, column } = hyperlink;
		return this._commandService.executeCommand(_univerjs_sheets_hyper_link.UpdateHyperLinkCommand.id, {
			unitId: this.getUnitId(),
			subUnitId: this._worksheet.getSheetId(),
			id,
			row,
			column,
			payload: {
				payload: url,
				display: label
			}
		});
	}
	cancelHyperLink(hyperlink) {
		if (hyperlink) {
			const { id, row, column } = hyperlink;
			return this._commandService.syncExecuteCommand(_univerjs_sheets_hyper_link.CancelHyperLinkCommand.id, {
				unitId: this.getUnitId(),
				subUnitId: this._worksheet.getSheetId(),
				id,
				row,
				column
			});
		} else {
			const hyperlinks = this.getHyperLinks();
			if (hyperlinks.length === 0) return true;
			return hyperlinks.every((link) => {
				const { id, row, column } = link;
				return this._commandService.syncExecuteCommand(_univerjs_sheets_hyper_link.CancelHyperLinkCommand.id, {
					unitId: this.getUnitId(),
					subUnitId: this._worksheet.getSheetId(),
					id,
					row,
					column
				});
			});
		}
	}
	getUrl() {
		return this._injector.get(_univerjs_sheets_hyper_link.SheetsHyperLinkParserService).buildHyperLink(this.getRange(), this.getSheetId());
	}
};
_univerjs_sheets_facade.FRange.extend(FRangeSheetsHyperlinkMixin);

//#endregion
//#region src/facade/f-event.ts
var FSheetsHyperlinkEventNameMixin = class extends _univerjs_core_facade.FEventName {
	get BeforeSheetLinkAdd() {
		return "BeforeSheetLinkAdd";
	}
	get BeforeSheetLinkCancel() {
		return "BeforeSheetLinkCancel";
	}
	get BeforeSheetLinkUpdate() {
		return "BeforeSheetLinkUpdate";
	}
};
_univerjs_core_facade.FEventName.extend(FSheetsHyperlinkEventNameMixin);

//#endregion
//#region src/facade/f-univer.ts
var FUniverSheetsHyperlinkMixin = class extends _univerjs_core_facade.FUniver {
	/**
	* @ignore
	*/
	_initialize(injector) {
		const commandService = injector.get(_univerjs_core.ICommandService);
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetLinkAdd, () => commandService.beforeCommandExecuted((commandInfo) => {
			if (commandInfo.id !== _univerjs_sheets_hyper_link.AddHyperLinkCommand.id) return;
			const eventTarget = this.getCommandSheetTarget(commandInfo);
			if (!eventTarget) return;
			const params = commandInfo.params;
			const eventParams = {
				workbook: eventTarget.workbook,
				worksheet: eventTarget.worksheet,
				row: params.link.row,
				col: params.link.column,
				link: params.link
			};
			this.fireEvent(this.Event.BeforeSheetLinkAdd, eventParams);
			if (eventParams.cancel) throw new _univerjs_core.CanceledError();
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetLinkUpdate, () => commandService.beforeCommandExecuted((commandInfo) => {
			if (commandInfo.id !== _univerjs_sheets_hyper_link.UpdateHyperLinkCommand.id) return;
			const eventTarget = this.getCommandSheetTarget(commandInfo);
			if (!eventTarget) return;
			const params = commandInfo.params;
			const eventParams = {
				workbook: eventTarget.workbook,
				worksheet: eventTarget.worksheet,
				row: params.row,
				column: params.column,
				id: params.id,
				payload: params.payload
			};
			this.fireEvent(this.Event.BeforeSheetLinkUpdate, eventParams);
			if (eventParams.cancel) throw new _univerjs_core.CanceledError();
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetLinkCancel, () => commandService.beforeCommandExecuted((commandInfo) => {
			if (commandInfo.id !== _univerjs_sheets_hyper_link.CancelHyperLinkCommand.id) return;
			const eventTarget = this.getCommandSheetTarget(commandInfo);
			if (!eventTarget) return;
			const params = commandInfo.params;
			const eventParams = {
				workbook: eventTarget.workbook,
				worksheet: eventTarget.worksheet,
				row: params.row,
				column: params.column,
				id: params.id
			};
			this.fireEvent(this.Event.BeforeSheetLinkCancel, eventParams);
			if (eventParams.cancel) throw new _univerjs_core.CanceledError();
		})));
	}
};
_univerjs_core_facade.FUniver.extend(FUniverSheetsHyperlinkMixin);

//#endregion
exports.FWorkbookHyperlinkMixin = FWorkbookHyperlinkMixin;
