// @univerjs/sheets-filter/lib/cjs/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
let _univerjs_core = require("@univerjs/core");
let _univerjs_sheets = require("@univerjs/sheets");
let rxjs = require("rxjs");
let _univerjs_engine_render = require("@univerjs/engine-render");
let _univerjs_rpc = require("@univerjs/rpc");
let _univerjs_engine_formula = require("@univerjs/engine-formula");

//#region src/common/const.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const SetSheetsFilterRangeMutationId = "sheet.mutation.set-filter-range";
const SetSheetsFilterCriteriaMutationId = "sheet.mutation.set-filter-criteria";
const RemoveSheetsFilterMutationId = "sheet.mutation.remove-filter";
const ReCalcSheetsFilterMutationId = "sheet.mutation.re-calc-filter";
const FILTER_MUTATIONS = new Set([
	SetSheetsFilterRangeMutationId,
	SetSheetsFilterCriteriaMutationId,
	RemoveSheetsFilterMutationId,
	ReCalcSheetsFilterMutationId
]);

//#endregion
//#region src/models/types.ts
/**
* The filter types.
*/
let FilterBy = /* @__PURE__ */ function(FilterBy) {
	FilterBy[FilterBy["VALUES"] = 0] = "VALUES";
	FilterBy[FilterBy["COLORS"] = 1] = "COLORS";
	FilterBy[FilterBy["CONDITIONS"] = 2] = "CONDITIONS";
	return FilterBy;
}({});
/**
* Basic custom filter operators.
*
* @internal
* doesNotContain, isBlank, isNotBlank are not defined in OOXML. They are represented by regex-like values.
*/
let CustomFilterOperator = /* @__PURE__ */ function(CustomFilterOperator) {
	/** "EQUAL" operator. */
	CustomFilterOperator["EQUAL"] = "equal";
	/** "GREATER_THAN" operator. */
	CustomFilterOperator["GREATER_THAN"] = "greaterThan";
	/** "GREATER_THAN_OR_EQUAL" operator. */
	CustomFilterOperator["GREATER_THAN_OR_EQUAL"] = "greaterThanOrEqual";
	/** "LESS_THAN" operator. */
	CustomFilterOperator["LESS_THAN"] = "lessThan";
	/** "LESS_THAN_OR_EQUAL" operator. */
	CustomFilterOperator["LESS_THAN_OR_EQUAL"] = "lessThanOrEqual";
	/** "NOT_EQUALS" operator. */
	CustomFilterOperator["NOT_EQUALS"] = "notEqual";
	return CustomFilterOperator;
}({});

//#endregion
//#region src/models/custom-filters.ts
const greaterThan = {
	operator: "greaterThan",
	fn: (value, compare) => {
		if (!ensureNumber(value)) return false;
		return value > compare;
	}
};
const greaterThanOrEqualTo = {
	operator: "greaterThanOrEqual",
	fn: (value, compare) => {
		if (!ensureNumber(value)) return false;
		return value >= compare;
	}
};
const lessThan = {
	operator: "lessThan",
	fn: (value, compare) => {
		if (!ensureNumber(value)) return false;
		return value < compare;
	}
};
const lessThanOrEqualTo = {
	operator: "lessThanOrEqual",
	fn: (value, compare) => {
		if (!ensureNumber(value)) return false;
		return value <= compare;
	}
};
const equals = {
	operator: "equal",
	fn: (value, compare) => {
		if (!ensureNumber(value)) return false;
		return value === compare;
	}
};
const notEquals = {
	operator: "notEqual",
	fn: (value, compare) => {
		if (typeof compare === "string") {
			if (compare === " ") {
				if (value !== void 0 && value !== null) return true;
				return false;
			}
			const ensuredString = ensureString(value);
			if (ensuredString && isWildCardString(compare)) return !(0, _univerjs_core.createREGEXFromWildChar)(compare).test(ensuredString);
			return ensuredString !== compare;
		}
		if (!ensureNumber(value)) return true;
		return value !== compare;
	}
};
const CustomFilterFnRegistry = /* @__PURE__ */ new Map([]);
[
	greaterThan,
	greaterThanOrEqualTo,
	lessThan,
	lessThanOrEqualTo,
	equals,
	notEquals
].forEach((fn) => {
	CustomFilterFnRegistry.set(fn.operator, fn);
});
function isNumericFilterFn(operator) {
	return !!operator;
}
/** This operators matches texts. */
const textMatch = { fn: (value, compare) => {
	const ensured = ensureString(value);
	if (ensured === null) {
		if (compare === "") return true;
		return false;
	}
	return (0, _univerjs_core.createREGEXFromWildChar)(compare).test(ensured);
} };
function getCustomFilterFn(operator) {
	if (!operator) return textMatch;
	return CustomFilterFnRegistry.get(operator);
}
function ensureNumber(value) {
	return typeof value === "number";
}
function ensureNumeric(value) {
	if (typeof value === "number") return true;
	if (typeof value === "string" && (0, _univerjs_core.isNumeric)(value)) return true;
	return false;
}
function ensureString(value) {
	if (typeof value === "boolean" || value == null) return null;
	return typeof value === "string" ? value : value.toString();
}
function isWildCardString(str) {
	if (typeof str === "number") return false;
	return str.indexOf("*") !== -1 || str.indexOf("?") !== -1;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region src/models/filter-model.ts
const EMPTY = () => /* @__PURE__ */ new Set();
/**
* This is the in-memory model of filter.
*/
var FilterModel = class FilterModel extends _univerjs_core.Disposable {
	get filteredOutRows() {
		return this._filteredOutRows$.getValue();
	}
	set filteredOutRows(rows) {
		this._alreadyFilteredOutRows = rows;
		this._filteredOutRows$.next(rows);
	}
	constructor(unitId, subUnitId, _worksheet) {
		super();
		this.unitId = unitId;
		this.subUnitId = subUnitId;
		this._worksheet = _worksheet;
		_defineProperty(this, "_filteredOutRows$", new rxjs.BehaviorSubject(EMPTY()));
		_defineProperty(this, "filteredOutRows$", this._filteredOutRows$.asObservable());
		_defineProperty(this, "_hasCriteria$", new rxjs.BehaviorSubject(false));
		_defineProperty(this, "hasCriteria$", this._hasCriteria$.asObservable());
		_defineProperty(this, "_filterColumnByIndex", /* @__PURE__ */ new Map());
		_defineProperty(this, "_alreadyFilteredOutRows", EMPTY());
		_defineProperty(this, "_range", void 0);
	}
	dispose() {
		super.dispose();
		this._filteredOutRows$.complete();
		this._hasCriteria$.complete();
		this._worksheet = null;
	}
	/**
	* Serialize this filter model to the JSON format representation.
	*/
	serialize() {
		const result = {
			ref: _univerjs_core.Rectangle.clone(this._range),
			filterColumns: this._getAllFilterColumns(true).sort(([offset1], [offset2]) => offset1 - offset2).map(([_, filterColumn]) => filterColumn.serialize())
		};
		if (this._alreadyFilteredOutRows) result.cachedFilteredOut = Array.from(this._alreadyFilteredOutRows).sort();
		return result;
	}
	/**
	* Deserialize auto filter info to construct a `FilterModel` object.
	* @param unitId workbook id
	* @param subUnitId worksheet id
	* @param worksheet the Worksheet object
	* @param autoFilter auto filter data
	*/
	static deserialize(unitId, subUnitId, worksheet, autoFilter) {
		const filterModel = new FilterModel(unitId, subUnitId, worksheet);
		filterModel._dump(autoFilter);
		return filterModel;
	}
	_dump(autoFilter) {
		var _autoFilter$filterCol;
		this.setRange(autoFilter.ref);
		(_autoFilter$filterCol = autoFilter.filterColumns) === null || _autoFilter$filterCol === void 0 || _autoFilter$filterCol.filter((filterColumn) => {
			if (!filterColumn.filters && !filterColumn.colorFilters && !filterColumn.customFilters) return false;
			return true;
		}).forEach((filterColumn) => this._setCriteriaWithoutReCalc(filterColumn.colId, filterColumn));
		if (autoFilter.cachedFilteredOut) {
			this._alreadyFilteredOutRows = new Set(autoFilter.cachedFilteredOut);
			this._emit();
		} else if (autoFilter.filterColumns && autoFilter.filterColumns.length > 0) {
			this._reCalcAllColumns();
			this._emit();
		}
		this._emitHasCriteria();
	}
	isRowFiltered(row) {
		return this._alreadyFilteredOutRows.has(row);
	}
	getRange() {
		if (!this._range) throw new Error("[FilterModel] could not get range before a range is set!");
		return this._range;
	}
	/**
	* Get filtered out rows except the specific column. This method is considered as "pure". In
	* another word it would not change `filteredOutRows` on `FilterModel` nor `FilterColumn`.
	* @param col
	*/
	getFilteredOutRowsExceptCol(col) {
		return this._getAllFilterColumns(true).filter(([colOffset]) => colOffset !== col).reduce((acc, [, filterColumn]) => {
			const newResult = filterColumn.calc({ getAlreadyFilteredOutRows: () => acc });
			if (newResult) return (0, _univerjs_core.mergeSets)(acc, newResult);
			return acc;
		}, /* @__PURE__ */ new Set());
	}
	/**
	* Set range of the filter model, this would remove some `IFilterColumn`
	* if the new range not overlaps the old range.
	*/
	setRange(range) {
		this._range = range;
		this._getAllFilterColumns(true).forEach(([col, filterColumn]) => {
			filterColumn.setRangeAndColumn({
				startRow: range.startRow,
				endRow: range.endRow,
				startColumn: col,
				endColumn: col
			}, col);
		});
	}
	/**
	* Set or remove filter criteria on a specific row.
	*/
	setCriteria(col, criteria, reCalc = false) {
		if (!this._range) throw new Error("[FilterModel] could not set criteria before a range is set!");
		if (!criteria) {
			this._removeCriteria(col);
			this._rebuildAlreadyFilteredOutRowsWithCache();
			if (reCalc) this._reCalcAllColumns();
			this._emit();
			this._emitHasCriteria();
			return;
		}
		this._setCriteriaWithoutReCalc(col, criteria);
		if (reCalc) {
			this._rebuildAlreadyFilteredOutRowsWithCache();
			this._getAllFilterColumns().forEach((filterColumn) => filterColumn.__clearCache());
			this._reCalcWithNoCacheColumns();
			this._emit();
			this._emitHasCriteria();
		}
	}
	getAllFilterColumns() {
		return this._getAllFilterColumns(true);
	}
	getFilterColumn(index) {
		var _this$_filterColumnBy;
		return (_this$_filterColumnBy = this._filterColumnByIndex.get(index)) !== null && _this$_filterColumnBy !== void 0 ? _this$_filterColumnBy : null;
	}
	reCalc() {
		this._reCalcAllColumns();
		this._emit();
	}
	_getAllFilterColumns(withCol = false) {
		const columns = Array.from(this._filterColumnByIndex.entries());
		if (withCol) return columns;
		return columns.map(([_, filterColumn]) => filterColumn);
	}
	_reCalcAllColumns() {
		this._alreadyFilteredOutRows = EMPTY();
		this._getAllFilterColumns().forEach((filterColumn) => filterColumn.__clearCache());
		this._reCalcWithNoCacheColumns();
	}
	_setCriteriaWithoutReCalc(col, criteria) {
		const range = this._range;
		if (!range) throw new Error("[FilterModel] could not set criteria before a range is set!");
		const { startColumn, endColumn } = range;
		if (col > endColumn || col < startColumn) throw new Error(`[FilterModel] could not set criteria on column ${col} which is out of range!`);
		let filterColumn;
		if (this._filterColumnByIndex.has(col)) filterColumn = this._filterColumnByIndex.get(col);
		else {
			filterColumn = new FilterColumn(this.unitId, this.subUnitId, this._worksheet, criteria, { getAlreadyFilteredOutRows: () => this._alreadyFilteredOutRows });
			filterColumn.setRangeAndColumn(range, col);
			this._filterColumnByIndex.set(col, filterColumn);
		}
		filterColumn.setCriteria(criteria);
	}
	_removeCriteria(col) {
		const filterColumn = this._filterColumnByIndex.get(col);
		if (filterColumn) {
			filterColumn.dispose();
			this._filterColumnByIndex.delete(col);
		}
	}
	_emit() {
		this._filteredOutRows$.next(this._alreadyFilteredOutRows);
	}
	_emitHasCriteria() {
		this._hasCriteria$.next(this._filterColumnByIndex.size > 0);
	}
	_rebuildAlreadyFilteredOutRowsWithCache() {
		const newFilteredOutRows = this._getAllFilterColumns().filter((filterColumn) => filterColumn.hasCache()).reduce((acc, filterColumn) => {
			return (0, _univerjs_core.mergeSets)(acc, filterColumn.filteredOutRows);
		}, /* @__PURE__ */ new Set());
		this._alreadyFilteredOutRows = newFilteredOutRows;
	}
	_reCalcWithNoCacheColumns() {
		const noCacheFilteredOutRows = this._getAllFilterColumns().filter((filterColumn) => !filterColumn.hasCache());
		for (const filterColumn of noCacheFilteredOutRows) {
			const filteredRows = filterColumn.reCalc();
			if (filteredRows) this._alreadyFilteredOutRows = (0, _univerjs_core.mergeSets)(this._alreadyFilteredOutRows, filteredRows);
		}
	}
};
/**
* This is the filter criteria on a specific column.
*/
var FilterColumn = class extends _univerjs_core.Disposable {
	get filteredOutRows() {
		return this._filteredOutRows;
	}
	get filterBy() {
		return this._filterBy;
	}
	constructor(unitId, subUnitId, _worksheet, _criteria, _filterColumnContext) {
		super();
		this.unitId = unitId;
		this.subUnitId = subUnitId;
		this._worksheet = _worksheet;
		this._criteria = _criteria;
		this._filterColumnContext = _filterColumnContext;
		_defineProperty(this, "_filteredOutRows", null);
		_defineProperty(this, "_filterFn", null);
		_defineProperty(this, "_range", null);
		_defineProperty(this, "_column", 0);
		_defineProperty(this, "_filterBy", 0);
	}
	dispose() {
		super.dispose();
		this._filteredOutRows = null;
	}
	/**
	* @internal
	*/
	__clearCache() {
		this._filteredOutRows = null;
	}
	serialize() {
		if (!this._criteria) throw new Error("[FilterColumn]: could not serialize without a filter column!");
		return _univerjs_core.Tools.deepClone({
			...this._criteria,
			colId: this._column
		});
	}
	hasCache() {
		return this._filteredOutRows !== null;
	}
	setRangeAndColumn(range, column) {
		this._range = range;
		this._column = column;
	}
	setCriteria(criteria) {
		this._criteria = criteria;
		this._generateFilterFn();
		this._filteredOutRows = null;
	}
	getColumnData() {
		return _univerjs_core.Tools.deepClone(this._criteria);
	}
	/**
	* Trigger new calculation on this `FilterModel` instance.
	*
	* @external DO NOT EVER call this method from `FilterColumn` itself. The whole process heavily relies on
	* `filteredOutByOthers`, and it is more comprehensible if we let `FilterModel` take full control over the process.
	*/
	reCalc() {
		this._filteredOutRows = this.calc(this._filterColumnContext);
		return this._filteredOutRows;
	}
	calc(context) {
		if (!this._filterFn) throw new Error("[FilterColumn] cannot calculate without a filter fn!");
		if (!this._range) throw new Error("[FilterColumn] cannot calculate without a range!");
		if (typeof this._column !== "number") throw new TypeError("[FilterColumn] cannot calculate without a column offset!");
		const column = this._column;
		const iterateRange = {
			startColumn: column,
			endColumn: column,
			startRow: this._range.startRow + 1,
			endRow: this._range.endRow
		};
		const filteredOutRows = /* @__PURE__ */ new Set();
		const filteredOutByOthers = context.getAlreadyFilteredOutRows();
		for (const range of this._worksheet.iterateByColumn(iterateRange, false, false)) {
			const { row, rowSpan, col } = range;
			if (filteredOutByOthers.has(row) && (!rowSpan || rowSpan === 1)) continue;
			if (!(this._filterBy === 0 ? this._filterFn((0, _univerjs_core.extractPureTextFromCell)(this._worksheet.getCell(row, col))) : this._filterBy === 1 ? this._filterFn(this._worksheet.getComposedCellStyle(row, col)) : this._filterFn(getFilterValueForConditionalFiltering(this._worksheet, row, col)))) {
				filteredOutRows.add(row);
				if (rowSpan) for (let i = 1; i < rowSpan; i++) filteredOutRows.add(row + i);
			}
		}
		return filteredOutRows;
	}
	_generateFilterFn() {
		if (!this._criteria) return;
		this._filterFn = generateFilterFn(this._criteria);
		this._filterBy = this._criteria.filters ? 0 : this._criteria.colorFilters ? 1 : 2;
	}
};
/**
* This functions take a `IFilterColumn` as input and return a function that can be used to filter rows.
* @param column
* @returns the filter function that takes the cell's value and return a boolean.
*/
function generateFilterFn(column) {
	if (column.filters) return filterByValuesFnFactory(column.filters);
	if (column.colorFilters) return filterByColorsFnFactory(column.colorFilters);
	if (column.customFilters) return customFilterFnFactory(column.customFilters);
	throw new Error("[FilterModel]: other types of filters are not supported yet.");
}
function filterByValuesFnFactory(values) {
	const includeBlank = !!values.blank;
	const valuesSet = new Set(values.filters);
	return (value) => {
		if (value === void 0 || value === "") return includeBlank;
		return valuesSet.has(typeof value === "string" ? value : `${value}`);
	};
}
function filterByColorsFnFactory(colorFilters) {
	if (colorFilters.cellFillColors) {
		const fillColorsSet = new Set(colorFilters.cellFillColors);
		return (cellStyle) => {
			var _cellStyle$bg;
			if (!cellStyle || !((_cellStyle$bg = cellStyle.bg) === null || _cellStyle$bg === void 0 ? void 0 : _cellStyle$bg.rgb)) {
				if (fillColorsSet.has(null)) return true;
				return false;
			}
			const bg = new _univerjs_core.ColorKit(cellStyle.bg.rgb).toRgbString();
			return fillColorsSet.has(bg);
		};
	}
	if (colorFilters.cellTextColors) {
		const textColorsSet = new Set(colorFilters.cellTextColors);
		return (cellStyle) => {
			var _cellStyle$cl;
			if (!cellStyle || !((_cellStyle$cl = cellStyle.cl) === null || _cellStyle$cl === void 0 ? void 0 : _cellStyle$cl.rgb)) {
				if (textColorsSet.has(_univerjs_engine_render.COLOR_BLACK_RGB)) return true;
				return false;
			}
			const cl = new _univerjs_core.ColorKit(cellStyle.cl.rgb).toRgbString();
			return textColorsSet.has(cl);
		};
	}
	throw new Error("[FilterModel]: color filters are not supported yet.");
}
function customFilterFnFactory(customFilters) {
	const customFilterFns = customFilters.customFilters.map((filter) => generateCustomFilterFn(filter));
	if (isCompoundCustomFilter(customFilterFns)) {
		if (customFilters.and) return AND(customFilterFns);
		return OR(customFilterFns);
	}
	return customFilterFns[0];
}
function AND(filterFns) {
	const [fn1, fn2] = filterFns;
	return (value) => fn1(value) && fn2(value);
}
function OR(filterFns) {
	const [fn1, fn2] = filterFns;
	return (value) => fn1(value) || fn2(value);
}
function isCompoundCustomFilter(filter) {
	return filter.length === 2;
}
function generateCustomFilterFn(filter) {
	const compare = filter.val;
	if (filter.operator === "notEqual") {
		if (!ensureNumeric(compare)) return (value) => notEquals.fn(value, compare);
	}
	if (isNumericFilterFn(filter.operator)) {
		if (!ensureNumeric(compare)) return () => false;
		const customFilterFn = getCustomFilterFn(filter.operator);
		const ensuredNumber = Number(compare);
		return (value) => customFilterFn.fn(value, ensuredNumber);
	}
	const customFilterFn = getCustomFilterFn(filter.operator);
	return (value) => customFilterFn.fn(value, compare);
}
function getFilterValueForConditionalFiltering(worksheet, row, col) {
	const interceptedCell = worksheet.getCell(row, col);
	if (!interceptedCell) return null;
	const rawCell = worksheet.getCellRaw(row, col);
	if (interceptedCell && !rawCell) return extractFilterValueFromCell(interceptedCell);
	if (!rawCell) return null;
	if (interceptedCell.t === _univerjs_core.CellValueType.NUMBER && typeof interceptedCell.v === "string") return rawCell.v;
	if (interceptedCell.t === _univerjs_core.CellValueType.NUMBER) return Number(rawCell.v);
	return extractFilterValueFromCell(rawCell);
}
function extractFilterValueFromCell(cell) {
	var _cell$p;
	const richTextValue = (_cell$p = cell.p) === null || _cell$p === void 0 || (_cell$p = _cell$p.body) === null || _cell$p === void 0 ? void 0 : _cell$p.dataStream;
	if (richTextValue) return richTextValue.trimEnd();
	const rawValue = cell.v;
	if (typeof rawValue === "string") {
		if (cell.t === _univerjs_core.CellValueType.BOOLEAN) return rawValue.toUpperCase();
		return rawValue;
	}
	if (typeof rawValue === "number") {
		if (cell.t === _univerjs_core.CellValueType.BOOLEAN) return rawValue ? "TRUE" : "FALSE";
		return rawValue;
	}
	if (typeof rawValue === "boolean") return rawValue ? "TRUE" : "FALSE";
	return "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/services/sheet-filter.service.ts
const SHEET_FILTER_SNAPSHOT_ID = "SHEET_FILTER_PLUGIN";
let SheetsFilterService = class SheetsFilterService extends _univerjs_core.Disposable {
	/** The current Workbook's active Worksheet's filter model (if there is one). */
	get activeFilterModel() {
		return this._activeFilterModel$.getValue();
	}
	constructor(_resourcesManagerService, _univerInstanceService, _commandService) {
		super();
		this._resourcesManagerService = _resourcesManagerService;
		this._univerInstanceService = _univerInstanceService;
		this._commandService = _commandService;
		_defineProperty(this, "_filterModels", /* @__PURE__ */ new Map());
		_defineProperty(this, "_loadedUnitId$", new rxjs.BehaviorSubject(null));
		_defineProperty(this, "loadedUnitId$", this._loadedUnitId$.asObservable());
		_defineProperty(this, "_errorMsg$", new rxjs.BehaviorSubject(null));
		_defineProperty(this, "errorMsg$", this._errorMsg$.asObservable());
		_defineProperty(this, "_activeFilterModel$", new rxjs.BehaviorSubject(null));
		_defineProperty(this, "activeFilterModel$", this._activeFilterModel$.asObservable());
		this._initModel();
		this._initActiveFilterModel();
	}
	/**
	*
	* @param unitId
	* @param subUnitId
	*/
	ensureFilterModel(unitId, subUnitId) {
		const already = this.getFilterModel(unitId, subUnitId);
		if (already) return already;
		const workbook = this._univerInstanceService.getUniverSheetInstance(unitId);
		if (!workbook) throw new Error(`[SheetsFilterService]: could not create "FilterModel" on a non-existing workbook ${unitId}!`);
		const worksheet = workbook.getSheetBySheetId(subUnitId);
		if (!worksheet) throw new Error(`[SheetsFilterService]: could not create "FilterModel" on a non-existing worksheet ${subUnitId}!`);
		const filterModel = new FilterModel(unitId, subUnitId, worksheet);
		this._cacheFilterModel(unitId, subUnitId, filterModel);
		return filterModel;
	}
	getFilterModel(unitId, subUnitId) {
		var _this$_filterModels$g, _this$_filterModels$g2;
		return (_this$_filterModels$g = (_this$_filterModels$g2 = this._filterModels.get(unitId)) === null || _this$_filterModels$g2 === void 0 ? void 0 : _this$_filterModels$g2.get(subUnitId)) !== null && _this$_filterModels$g !== void 0 ? _this$_filterModels$g : null;
	}
	removeFilterModel(unitId, subUnitId) {
		const already = this.getFilterModel(unitId, subUnitId);
		if (already) {
			already.dispose();
			this._filterModels.get(unitId).delete(subUnitId);
			return true;
		}
		return false;
	}
	setFilterErrorMsg(content) {
		this._errorMsg$.next(content);
	}
	_updateActiveFilterModel() {
		let workbook;
		try {
			workbook = this._univerInstanceService.getCurrentUnitOfType(_univerjs_core.UniverInstanceType.UNIVER_SHEET);
			if (!workbook) {
				this._activeFilterModel$.next(null);
				return;
			}
		} catch (err) {
			console.error("[SheetsFilterService]: could not get active workbook!", err);
			return;
		}
		const activeSheet = workbook.getActiveSheet(true);
		if (!activeSheet) {
			this._activeFilterModel$.next(null);
			return;
		}
		const unitId = activeSheet.getUnitId();
		const subUnitId = activeSheet.getSheetId();
		const filterModel = this.getFilterModel(unitId, subUnitId);
		this._activeFilterModel$.next(filterModel);
	}
	_initActiveFilterModel() {
		this.disposeWithMe((0, rxjs.merge)((0, _univerjs_core.fromCallback)(this._commandService.onCommandExecuted.bind(this._commandService)).pipe((0, rxjs.filter)(([command]) => command.type === _univerjs_core.CommandType.MUTATION && FILTER_MUTATIONS.has(command.id))), this._univerInstanceService.getCurrentTypeOfUnit$(_univerjs_core.UniverInstanceType.UNIVER_SHEET).pipe((0, rxjs.switchMap)((workbook) => {
			var _workbook$activeSheet;
			return (_workbook$activeSheet = workbook === null || workbook === void 0 ? void 0 : workbook.activeSheet$) !== null && _workbook$activeSheet !== void 0 ? _workbook$activeSheet : (0, rxjs.of)(null);
		}))).subscribe(() => this._updateActiveFilterModel()));
	}
	_serializeAutoFiltersForUnit(unitId) {
		const allFilterModels = this._filterModels.get(unitId);
		if (!allFilterModels) return "{}";
		const json = {};
		allFilterModels.forEach((model, worksheetId) => {
			json[worksheetId] = model.serialize();
		});
		return JSON.stringify(json);
	}
	_deserializeAutoFiltersForUnit(unitId, json) {
		const workbook = this._univerInstanceService.getUniverSheetInstance(unitId);
		Object.keys(json).forEach((worksheetId) => {
			const autoFilter = json[worksheetId];
			const filterModel = FilterModel.deserialize(unitId, worksheetId, workbook.getSheetBySheetId(worksheetId), autoFilter);
			this._cacheFilterModel(unitId, worksheetId, filterModel);
		});
	}
	dispose() {
		super.dispose();
		this._loadedUnitId$.complete();
		this._errorMsg$.complete();
		this._activeFilterModel$.complete();
		this._filterModels.forEach((allFilterModels) => {
			allFilterModels.forEach((model) => model.dispose());
			allFilterModels.clear();
		});
		this._filterModels.clear();
	}
	_initModel() {
		this._resourcesManagerService.registerPluginResource({
			pluginName: SHEET_FILTER_SNAPSHOT_ID,
			businesses: [_univerjs_core.UniverInstanceType.UNIVER_SHEET],
			toJson: (id) => this._serializeAutoFiltersForUnit(id),
			parseJson: (json) => JSON.parse(json),
			onLoad: (unitId, value) => {
				this._deserializeAutoFiltersForUnit(unitId, value);
				this._loadedUnitId$.next(unitId);
				this._updateActiveFilterModel();
			},
			onUnLoad: (unitId) => {
				const allFilterModels = this._filterModels.get(unitId);
				if (allFilterModels) {
					allFilterModels.forEach((model) => model.dispose());
					this._filterModels.delete(unitId);
				}
			}
		});
	}
	_cacheFilterModel(unitId, subUnitId, filterModel) {
		if (!this._filterModels.has(unitId)) this._filterModels.set(unitId, /* @__PURE__ */ new Map());
		this._filterModels.get(unitId).set(subUnitId, filterModel);
	}
};
SheetsFilterService = __decorate([
	__decorateParam(0, _univerjs_core.IResourceManagerService),
	__decorateParam(1, _univerjs_core.IUniverInstanceService),
	__decorateParam(2, _univerjs_core.ICommandService)
], SheetsFilterService);

//#endregion
//#region src/commands/mutations/sheets-filter.mutation.ts
/**
* A {@link CommandType.MUTATION} to set filter range in a {@link Worksheet}. If no {@link FilterModel} exists,
* a new `FilterModel` will be created.
*
* Since there could only be a filter on a worksheet, when you want to update the range, you
* don't necessarily need to remove the filter first, you can just execute this mutation.
*/
const SetSheetsFilterRangeMutation = {
	id: SetSheetsFilterRangeMutationId,
	type: _univerjs_core.CommandType.MUTATION,
	handler: (accessor, params) => {
		const { subUnitId, unitId, range } = params;
		accessor.get(SheetsFilterService).ensureFilterModel(unitId, subUnitId).setRange(range);
		return true;
	}
};
/**
* A {@link CommandType.MUTATION} to set filter criteria of a given column of a {@link FilterModel}.
*/
const SetSheetsFilterCriteriaMutation = {
	id: SetSheetsFilterCriteriaMutationId,
	type: _univerjs_core.CommandType.MUTATION,
	handler: (accessor, params) => {
		const { subUnitId, unitId, criteria, col, reCalc = true } = params;
		const filterModel = accessor.get(SheetsFilterService).getFilterModel(unitId, subUnitId);
		if (!filterModel) return false;
		filterModel.setCriteria(col, criteria, reCalc);
		return true;
	}
};
/**
* A {@link CommandType.MUTATION} to remove a {@link FilterModel} in a {@link Worksheet}.
*/
const RemoveSheetsFilterMutation = {
	id: RemoveSheetsFilterMutationId,
	type: _univerjs_core.CommandType.MUTATION,
	handler: (accessor, params) => {
		const { unitId, subUnitId } = params;
		return accessor.get(SheetsFilterService).removeFilterModel(unitId, subUnitId);
	}
};
/**
* A {@link CommandType.MUTATION} to re-calculate a {@link FilterModel}.
*/
const ReCalcSheetsFilterMutation = {
	id: ReCalcSheetsFilterMutationId,
	type: _univerjs_core.CommandType.MUTATION,
	handler: (accessor, params) => {
		const { unitId, subUnitId } = params;
		const filterModel = accessor.get(SheetsFilterService).getFilterModel(unitId, subUnitId);
		if (!filterModel) return false;
		filterModel.reCalc();
		return true;
	}
};

//#endregion
//#region src/commands/commands/sheets-filter.command.ts
/**
* A {@link CommandType.COMMAND} to set filter range in a Worksheet. Its params {@link ISetSheetFilterRangeCommandParams}
* is required. If the {@link FilterModel} does not exist, it will be created.
*/
const SetSheetFilterRangeCommand = {
	id: "sheet.command.set-filter-range",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		const sheetsFilterService = accessor.get(SheetsFilterService);
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const undoRedoService = accessor.get(_univerjs_core.IUndoRedoService);
		const instanceSrv = accessor.get(_univerjs_core.IUniverInstanceService);
		const { unitId, subUnitId, range } = params;
		if (!(0, _univerjs_sheets.getSheetCommandTarget)(instanceSrv, params)) return false;
		if (sheetsFilterService.getFilterModel(unitId, subUnitId)) return false;
		if (range.endRow === range.startRow) {
			const errorService = accessor.get(_univerjs_core.ErrorService);
			const localeService = accessor.get(_univerjs_core.LocaleService);
			errorService.emit(localeService.t("sheets-filter.command.not-valid-filter-range"));
			return false;
		}
		const redoMutation = {
			id: SetSheetsFilterRangeMutation.id,
			params: {
				unitId,
				subUnitId,
				range
			}
		};
		const result = commandService.syncExecuteCommand(redoMutation.id, redoMutation.params);
		if (result) undoRedoService.pushUndoRedo({
			unitID: unitId,
			undoMutations: [{
				id: RemoveSheetsFilterMutation.id,
				params: {
					unitId,
					subUnitId
				}
			}],
			redoMutations: [redoMutation]
		});
		return result;
	}
};
/**
* A {@link CommandType.COMMAND} to remove filter in a Worksheet. Its params {@link ISheetCommandSharedParams} is
* required. If the {@link FilterModel} does not exist, it will fail to execute.
*/
const RemoveSheetFilterCommand = {
	id: "sheet.command.remove-sheet-filter",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		const target = (0, _univerjs_sheets.getSheetCommandTarget)(accessor.get(_univerjs_core.IUniverInstanceService), params);
		if (!target) return false;
		const sheetsFilterService = accessor.get(SheetsFilterService);
		const { unitId, subUnitId } = target;
		const filterModel = sheetsFilterService.getFilterModel(unitId, subUnitId);
		if (!filterModel) return false;
		const filterRange = filterModel.getRange();
		if (!filterRange) return false;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const undoRedoService = accessor.get(_univerjs_core.IUndoRedoService);
		const undoMutations = destructFilterModel(unitId, subUnitId, filterModel.serialize());
		const result = commandService.syncExecuteCommand(RemoveSheetsFilterMutation.id, {
			unitId,
			subUnitId
		});
		if (result) {
			undoRedoService.pushUndoRedo({
				unitID: unitId,
				undoMutations,
				redoMutations: [{
					id: RemoveSheetsFilterMutation.id,
					params: {
						unitId,
						subUnitId
					}
				}]
			});
			commandService.executeCommand(_univerjs_sheets.MarkDirtyFilterChangeMutation.id, {
				unitId,
				subUnitId,
				filterRange
			});
		}
		return result;
	}
};
/**
* A {@link CommandType.COMMAND} to toggle filter in the current {@link Worksheet}.
*/
const SmartToggleSheetsFilterCommand = {
	id: "sheet.command.smart-toggle-filter",
	type: _univerjs_core.CommandType.COMMAND,
	handler: async (accessor) => {
		const univerInstanceService = accessor.get(_univerjs_core.IUniverInstanceService);
		const sheetsFilterService = accessor.get(SheetsFilterService);
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const currentWorkbook = univerInstanceService.getCurrentUnitOfType(_univerjs_core.UniverInstanceType.UNIVER_SHEET);
		const currentWorksheet = currentWorkbook === null || currentWorkbook === void 0 ? void 0 : currentWorkbook.getActiveSheet();
		if (!currentWorksheet || !currentWorkbook) return false;
		const unitId = currentWorkbook.getUnitId();
		const subUnitId = currentWorksheet.getSheetId();
		if (sheetsFilterService.getFilterModel(unitId, subUnitId)) return commandService.executeCommand(RemoveSheetFilterCommand.id, {
			unitId,
			subUnitId
		});
		const lastSelection = accessor.get(_univerjs_sheets.SheetsSelectionsService).getCurrentLastSelection();
		if (!lastSelection) return false;
		const startRange = lastSelection.range;
		const targetFilterRange = (0, _univerjs_sheets.isSingleCellSelection)(lastSelection) ? (0, _univerjs_sheets.expandToContinuousRange)(startRange, {
			left: true,
			right: true,
			up: true,
			down: true
		}, currentWorksheet) : startRange.startRow === startRange.endRow ? (0, _univerjs_sheets.expandToContinuousRange)(startRange, { down: true }, currentWorksheet) : startRange;
		return commandService.executeCommand(SetSheetFilterRangeCommand.id, {
			unitId,
			subUnitId,
			range: targetFilterRange
		});
	}
};
/**
* A {@link CommandType.COMMAND} to set filter criteria to a column in the targeting {@link FilterModel}. Its params
* {@link ISetSheetsFilterCriteriaCommandParams} is required.
*/
const SetSheetsFilterCriteriaCommand = {
	id: "sheet.command.set-filter-criteria",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		const target = (0, _univerjs_sheets.getSheetCommandTarget)(accessor.get(_univerjs_core.IUniverInstanceService), params);
		if (!target) return false;
		const sheetsFilterService = accessor.get(SheetsFilterService);
		const { unitId, subUnitId } = target;
		const filterModel = sheetsFilterService.getFilterModel(unitId, subUnitId);
		if (!filterModel) return false;
		const { col, criteria } = params;
		const filterRange = filterModel.getRange();
		if (!filterRange || col < filterRange.startColumn || col > filterRange.endColumn) return false;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const undoRedoService = accessor.get(_univerjs_core.IUndoRedoService);
		const undoMutation = destructFilterColumn(unitId, subUnitId, col, filterModel.getFilterColumn(col));
		const redoMutation = {
			id: SetSheetsFilterCriteriaMutation.id,
			params: {
				unitId,
				subUnitId,
				col,
				criteria
			}
		};
		const result = commandService.syncExecuteCommand(redoMutation.id, redoMutation.params);
		if (result) {
			undoRedoService.pushUndoRedo({
				unitID: unitId,
				undoMutations: [undoMutation],
				redoMutations: [redoMutation]
			});
			commandService.executeCommand(_univerjs_sheets.MarkDirtyFilterChangeMutation.id, {
				unitId,
				subUnitId,
				filterRange
			});
		}
		return result;
	}
};
/**
* A {@link CommandType.COMMAND} to clear all filter criteria in the targeting {@link FilterModel}. Its params
* {@link ISheetCommandSharedParams} is required.
*/
const ClearSheetsFilterCriteriaCommand = {
	id: "sheet.command.clear-filter-criteria",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		const target = (0, _univerjs_sheets.getSheetCommandTarget)(accessor.get(_univerjs_core.IUniverInstanceService), params);
		if (!target) return false;
		const sheetsFilterService = accessor.get(SheetsFilterService);
		const { unitId, subUnitId } = target;
		const filterModel = sheetsFilterService.getFilterModel(unitId, subUnitId);
		if (!filterModel) return false;
		const filterRange = filterModel.getRange();
		if (!filterRange) return false;
		const undoRedoService = accessor.get(_univerjs_core.IUndoRedoService);
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const autoFilter = filterModel.serialize();
		const undoMutations = destructFilterCriteria(unitId, subUnitId, autoFilter);
		const redoMutations = generateRemoveCriteriaMutations(unitId, subUnitId, autoFilter);
		if ((0, _univerjs_core.sequenceExecute)(redoMutations, commandService).result) {
			undoRedoService.pushUndoRedo({
				unitID: unitId,
				undoMutations,
				redoMutations
			});
			commandService.executeCommand(_univerjs_sheets.MarkDirtyFilterChangeMutation.id, {
				unitId,
				subUnitId,
				filterRange
			});
			return true;
		}
		return false;
	}
};
/**
* A {@link CommandType.COMMAND} forcing the currently active {@link FilterModel} to re-calculate all filter criteria.
* Its params {@link ISheetCommandSharedParams} is required.
*/
const ReCalcSheetsFilterCommand = {
	id: "sheet.command.re-calc-filter",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		const sheetsFilterService = accessor.get(SheetsFilterService);
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const commandTarget = (0, _univerjs_sheets.getSheetCommandTarget)(accessor.get(_univerjs_core.IUniverInstanceService), params);
		if (!commandTarget) return false;
		const { unitId, subUnitId } = commandTarget;
		if (!sheetsFilterService.getFilterModel(commandTarget.unitId, commandTarget.subUnitId)) return false;
		return commandService.executeCommand(ReCalcSheetsFilterMutation.id, {
			unitId,
			subUnitId
		});
	}
};
/**
* Destruct a `FilterModel` to a list of mutations.
* @param {string} unitId - the unit id of the Workbook
* @param {string} subUnitId - the sub unit id of the Worksheet
* @param {IAutoFilter} autoFilter - the to be destructed FilterModel
* @returns a list of mutations those can be used to reconstruct the FilterModel
*/
function destructFilterModel(unitId, subUnitId, autoFilter) {
	const mutations = [];
	const setFilterMutation = {
		id: SetSheetsFilterRangeMutation.id,
		params: {
			unitId,
			subUnitId,
			range: autoFilter.ref
		}
	};
	mutations.push(setFilterMutation);
	destructFilterCriteria(unitId, subUnitId, autoFilter).forEach((m) => mutations.push(m));
	return mutations;
}
/**
* Transform a {@link FilterModel} to a list of mutations to set the filter criteria.
* @param unitId - the unit id of the {@link Workbook}
* @param subUnitId - the sub unit id of the {@link Worksheet}
* @param autoFilter - the to be destructed {@link FilterModel}
* @returns {IMutationInfo<ISetSheetsFilterCriteriaMutationParams>} a list of mutations those can be used to
* reconstruct the {@link FilterModel}
*/
function destructFilterCriteria(unitId, subUnitId, autoFilter) {
	var _autoFilter$filterCol;
	const mutations = [];
	(_autoFilter$filterCol = autoFilter.filterColumns) === null || _autoFilter$filterCol === void 0 || _autoFilter$filterCol.forEach((filterColumn) => {
		const setFilterCriteriaMutation = {
			id: SetSheetsFilterCriteriaMutation.id,
			params: {
				unitId,
				subUnitId,
				col: filterColumn.colId,
				criteria: filterColumn
			}
		};
		mutations.push(setFilterCriteriaMutation);
	});
	return mutations;
}
/** Generate mutations to remove all criteria on a `FilterModel` */
function generateRemoveCriteriaMutations(unitId, subUnitId, autoFilter) {
	var _autoFilter$filterCol2;
	const mutations = [];
	(_autoFilter$filterCol2 = autoFilter.filterColumns) === null || _autoFilter$filterCol2 === void 0 || _autoFilter$filterCol2.forEach((filterColumn) => {
		const removeFilterCriteriaMutation = {
			id: SetSheetsFilterCriteriaMutation.id,
			params: {
				unitId,
				subUnitId,
				col: filterColumn.colId,
				criteria: null
			}
		};
		mutations.push(removeFilterCriteriaMutation);
	});
	return mutations;
}
/**
* Prepare the undo mutation, it should rollback to the old criteria if there's already a `FilterColumn`,
* or remove the filter criteria when there is no `FilterColumn`.
* @param {string} unitId
* @param {string} subUnitId
* @param {number} colId
* @param {Nullable<FilterColumn>} filterColumn
* @returns {IMutationInfo<ISetSheetsFilterCriteriaMutationParams>} the undo mutation
*/
function destructFilterColumn(unitId, subUnitId, colId, filterColumn) {
	if (!filterColumn) return {
		id: SetSheetsFilterCriteriaMutation.id,
		params: {
			unitId,
			subUnitId,
			col: colId,
			criteria: null
		}
	};
	const serialize = filterColumn.serialize();
	return {
		id: SetSheetsFilterCriteriaMutation.id,
		params: {
			unitId,
			subUnitId,
			col: colId,
			criteria: serialize
		}
	};
}

//#endregion
//#region src/config/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const SHEETS_FILTER_PLUGIN_CONFIG_KEY = "sheets-filter.config";
const configSymbol = Symbol(SHEETS_FILTER_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/utils.ts
function objectsShaker(target, isEqual) {
	for (let i = 0; i < target.length; i++) {
		let cur = i;
		if (target[i]) {
			for (let j = i + 1; j < target.length; j++) if (target[cur] && target[j] && isEqual(target[cur], target[j])) {
				target[cur] = null;
				cur = j;
			}
		}
	}
	return target.filter((o) => o !== null);
}
function mergeSetFilterCriteria(mutations) {
	return objectsShaker(mutations, (o1, o2) => o1.id === SetSheetsFilterCriteriaMutation.id && o2.id === SetSheetsFilterCriteriaMutation.id && o1.params.unitId === o2.params.unitId && o1.params.subUnitId === o2.params.subUnitId && o1.params.col === o2.params.col);
}

//#endregion
//#region src/controllers/sheets-filter.controller.ts
let SheetsFilterController = class SheetsFilterController extends _univerjs_core.Disposable {
	constructor(_commandService, _sheetInterceptorService, _sheetsFilterService, _univerInstanceService, _refRangeService, _dataSyncPrimaryController, _zebraCrossingCacheController) {
		super();
		this._commandService = _commandService;
		this._sheetInterceptorService = _sheetInterceptorService;
		this._sheetsFilterService = _sheetsFilterService;
		this._univerInstanceService = _univerInstanceService;
		this._refRangeService = _refRangeService;
		this._dataSyncPrimaryController = _dataSyncPrimaryController;
		this._zebraCrossingCacheController = _zebraCrossingCacheController;
		_defineProperty(this, "_disposableCollection", new _univerjs_core.DisposableCollection());
		this._initCommands();
		this._initRowFilteredInterceptor();
		this._initInterceptors();
		this._commandExecutedListener();
		this._initErrorHandling();
		this._initZebraCrossingCacheListener();
	}
	_initZebraCrossingCacheListener() {
		this.disposeWithMe(this._sheetsFilterService.activeFilterModel$.subscribe((filterModel) => {
			if (!filterModel) return;
			this.disposeWithMe(filterModel.filteredOutRows$.subscribe(() => {
				this._zebraCrossingCacheController.updateZebraCrossingCache(filterModel.unitId, filterModel.subUnitId);
			}));
		}));
	}
	_initCommands() {
		[
			SetSheetFilterRangeCommand,
			RemoveSheetFilterCommand,
			SetSheetsFilterCriteriaCommand,
			ClearSheetsFilterCriteriaCommand,
			ReCalcSheetsFilterCommand
		].forEach((command) => {
			this.disposeWithMe(this._commandService.registerCommand(command));
		});
		[
			SetSheetsFilterCriteriaMutation,
			SetSheetsFilterRangeMutation,
			ReCalcSheetsFilterMutation,
			RemoveSheetsFilterMutation
		].forEach((command) => {
			var _this$_dataSyncPrimar;
			this.disposeWithMe(this._commandService.registerCommand(command));
			(_this$_dataSyncPrimar = this._dataSyncPrimaryController) === null || _this$_dataSyncPrimar === void 0 || _this$_dataSyncPrimar.registerSyncingMutations(command);
		});
	}
	_initInterceptors() {
		this.disposeWithMe(this._sheetInterceptorService.interceptCommand({ getMutations: (command) => this._getUpdateFilter(command) }));
		this.disposeWithMe(this._commandService.onCommandExecuted((commandInfo) => {
			if (commandInfo.id === _univerjs_sheets.SetWorksheetActiveOperation.id) {
				const params = commandInfo.params;
				const sheetId = params.subUnitId;
				const unitId = params.unitId;
				if (!sheetId || !unitId) return;
				this._registerRefRange(unitId, sheetId);
			}
			if (commandInfo.id === SetSheetsFilterRangeMutation.id) {
				const params = commandInfo.params;
				const sheetId = params.subUnitId;
				const unitId = params.unitId;
				if (!sheetId || !unitId) return;
				this._registerRefRange(params.unitId, params.subUnitId);
			}
		}));
		this.disposeWithMe(this._sheetsFilterService.loadedUnitId$.subscribe((unitId) => {
			if (unitId) {
				const workbook = this._univerInstanceService.getUniverSheetInstance(unitId);
				const sheet = workbook === null || workbook === void 0 ? void 0 : workbook.getActiveSheet();
				if (sheet) this._registerRefRange(unitId, sheet.getSheetId());
			}
		}));
	}
	_registerRefRange(unitId, subUnitId) {
		var _this$_sheetsFilterSe;
		this._disposableCollection.dispose();
		const workbook = this._univerInstanceService.getUniverSheetInstance(unitId);
		const workSheet = workbook === null || workbook === void 0 ? void 0 : workbook.getSheetBySheetId(subUnitId);
		if (!workbook || !workSheet) return;
		const range = (_this$_sheetsFilterSe = this._sheetsFilterService.getFilterModel(unitId, subUnitId)) === null || _this$_sheetsFilterSe === void 0 ? void 0 : _this$_sheetsFilterSe.getRange();
		const handler = (config) => {
			switch (config.id) {
				case _univerjs_sheets.InsertRowCommand.id: {
					const params = config.params;
					const _unitId = params.unitId || unitId;
					const _subUnitId = params.subUnitId || subUnitId;
					return this._handleInsertRowCommand(params, _unitId, _subUnitId);
				}
				case _univerjs_sheets.InsertColCommand.id: {
					const params = config.params;
					const _unitId = params.unitId || unitId;
					const _subUnitId = params.subUnitId || subUnitId;
					return this.handleInsertColCommand(params.range, _unitId, _subUnitId);
				}
				case _univerjs_sheets.RemoveColCommand.id: {
					const params = config.params;
					return this.handleRemoveColCommand(params.range, unitId, subUnitId);
				}
				case _univerjs_sheets.RemoveRowCommand.id: {
					const params = config.params;
					return this._handleRemoveRowCommand(params, unitId, subUnitId);
				}
				case _univerjs_sheets.EffectRefRangId.MoveColsCommandId: {
					const params = config.params;
					return this.handleMoveColsCommand({
						fromRange: params.fromRange,
						toRange: params.toRange
					}, unitId, subUnitId);
				}
				case _univerjs_sheets.EffectRefRangId.MoveRowsCommandId: {
					const params = config.params;
					return this._handleMoveRowsCommand(params, unitId, subUnitId);
				}
				case _univerjs_sheets.MoveRangeCommand.id: {
					const params = config.params;
					return this._handleMoveRangeCommand(params, unitId, subUnitId);
				}
			}
			return {
				redos: [],
				undos: []
			};
		};
		if (range) this._disposableCollection.add(this._refRangeService.registerRefRange(range, handler, unitId, subUnitId));
	}
	_getUpdateFilter(command) {
		const { id } = command;
		switch (id) {
			case _univerjs_sheets.RemoveSheetCommand.id: {
				const params = command.params;
				return this._handleRemoveSheetCommand(params, params.unitId, params.subUnitId);
			}
			case _univerjs_sheets.CopySheetCommand.id: {
				const { targetSubUnitId, unitId, subUnitId } = command.params;
				if (!unitId || !subUnitId || !targetSubUnitId) return this._handleNull();
				return this._handleCopySheetCommand(unitId, subUnitId, targetSubUnitId);
			}
		}
		return {
			redos: [],
			undos: []
		};
	}
	handleInsertColCommand(range, unitId, subUnitId) {
		var _filterModel$getRange;
		const filterModel = this._sheetsFilterService.getFilterModel(unitId, subUnitId);
		const filterRange = (_filterModel$getRange = filterModel === null || filterModel === void 0 ? void 0 : filterModel.getRange()) !== null && _filterModel$getRange !== void 0 ? _filterModel$getRange : null;
		if (!filterModel || !filterRange) return this._handleNull();
		const { startColumn, endColumn } = filterRange;
		const { startColumn: insertStartColumn, endColumn: insertEndColumn } = range;
		const count = insertEndColumn - insertStartColumn + 1;
		if (insertEndColumn > endColumn) return this._handleNull();
		const redos = [];
		const undos = [];
		const anchor = insertStartColumn;
		const setFilterRangeMutationParams = {
			unitId,
			subUnitId,
			range: {
				...filterRange,
				startColumn: insertStartColumn <= startColumn ? startColumn + count : startColumn,
				endColumn: endColumn + count
			}
		};
		const undoSetFilterRangeMutationParams = {
			unitId,
			subUnitId,
			range: filterRange
		};
		redos.push({
			id: SetSheetsFilterRangeMutation.id,
			params: setFilterRangeMutationParams
		});
		undos.push({
			id: SetSheetsFilterRangeMutation.id,
			params: undoSetFilterRangeMutationParams
		});
		const effected = filterModel.getAllFilterColumns().filter((column) => column[0] >= anchor);
		if (effected.length !== 0) {
			const { newRange, oldRange } = this._moveCriteria(unitId, subUnitId, effected, count);
			redos.push(...oldRange.redos, ...newRange.redos);
			undos.push(...newRange.undos, ...oldRange.undos);
		}
		return {
			redos: mergeSetFilterCriteria(redos),
			undos: mergeSetFilterCriteria(undos)
		};
	}
	_handleInsertRowCommand(config, unitId, subUnitId) {
		var _filterModel$getRange2;
		const filterModel = this._sheetsFilterService.getFilterModel(unitId, subUnitId);
		const filterRange = (_filterModel$getRange2 = filterModel === null || filterModel === void 0 ? void 0 : filterModel.getRange()) !== null && _filterModel$getRange2 !== void 0 ? _filterModel$getRange2 : null;
		if (!filterModel || !filterRange) return this._handleNull();
		const { startRow, endRow } = filterRange;
		const { startRow: insertStartRow, endRow: insertEndRow } = config.range;
		const rowCount = insertEndRow - insertStartRow + 1;
		if (insertEndRow > endRow) return this._handleNull();
		const redos = [];
		const undos = [];
		const setFilterRangeParams = {
			unitId,
			subUnitId,
			range: {
				...filterRange,
				startRow: insertStartRow <= startRow ? startRow + rowCount : startRow,
				endRow: endRow + rowCount
			}
		};
		const undoSetFilterRangeMutationParams = {
			unitId,
			subUnitId,
			range: filterRange
		};
		redos.push({
			id: SetSheetsFilterRangeMutation.id,
			params: setFilterRangeParams
		});
		undos.push({
			id: SetSheetsFilterRangeMutation.id,
			params: undoSetFilterRangeMutationParams
		});
		return {
			redos: mergeSetFilterCriteria(redos),
			undos: mergeSetFilterCriteria(undos)
		};
	}
	handleRemoveColCommand(range, unitId, subUnitId) {
		var _filterModel$getRange3;
		const filterModel = this._sheetsFilterService.getFilterModel(unitId, subUnitId);
		const filterRange = (_filterModel$getRange3 = filterModel === null || filterModel === void 0 ? void 0 : filterModel.getRange()) !== null && _filterModel$getRange3 !== void 0 ? _filterModel$getRange3 : null;
		if (!filterModel || !filterRange) return this._handleNull();
		const { startColumn, endColumn } = filterRange;
		const { startColumn: removeStartColumn, endColumn: removeEndColumn } = range;
		if (removeStartColumn > endColumn) return this._handleNull();
		const redos = [];
		const undos = [];
		const rangeRemoveCount = removeEndColumn < startColumn ? 0 : Math.min(removeEndColumn, endColumn) - Math.max(removeStartColumn, startColumn) + 1;
		const removeCount = removeEndColumn - removeStartColumn + 1;
		const filterColumn = filterModel.getAllFilterColumns();
		filterColumn.forEach((column) => {
			const [col, filter] = column;
			if (col <= removeEndColumn && col >= removeStartColumn) {
				redos.push({
					id: SetSheetsFilterCriteriaMutation.id,
					params: {
						unitId,
						subUnitId,
						col,
						criteria: null
					}
				});
				undos.push({
					id: SetSheetsFilterCriteriaMutation.id,
					params: {
						unitId,
						subUnitId,
						col,
						criteria: {
							...filter.serialize(),
							colId: col
						}
					}
				});
			}
		});
		const shifted = filterColumn.filter((column) => {
			const [col, _] = column;
			return col > removeEndColumn;
		});
		let newRangeCriteria = {
			undos: [],
			redos: []
		};
		if (shifted.length > 0) {
			const { oldRange, newRange } = this._moveCriteria(unitId, subUnitId, shifted, -removeCount);
			newRangeCriteria = newRange;
			redos.push(...oldRange.redos);
			undos.unshift(...oldRange.undos);
		}
		if (rangeRemoveCount === endColumn - startColumn + 1) {
			const removeFilterRangeMutationParams = {
				unitId,
				subUnitId
			};
			redos.push({
				id: RemoveSheetsFilterMutation.id,
				params: removeFilterRangeMutationParams
			});
			undos.unshift({
				id: SetSheetsFilterRangeMutation.id,
				params: {
					range: filterRange,
					unitId,
					subUnitId
				}
			});
		} else {
			const newStartColumn = startColumn <= removeStartColumn ? startColumn : rangeRemoveCount === 0 ? startColumn - removeCount : removeStartColumn;
			const newEndColumn = startColumn <= removeStartColumn ? endColumn - rangeRemoveCount : endColumn - removeCount;
			const setFilterRangeMutationParams = {
				unitId,
				subUnitId,
				range: {
					...filterRange,
					startColumn: newStartColumn,
					endColumn: newEndColumn
				}
			};
			redos.push({
				id: SetSheetsFilterRangeMutation.id,
				params: setFilterRangeMutationParams
			});
			undos.unshift({
				id: SetSheetsFilterRangeMutation.id,
				params: {
					range: filterRange,
					unitId,
					subUnitId
				}
			});
			redos.push(...newRangeCriteria.redos);
			undos.unshift(...newRangeCriteria.undos);
		}
		return {
			undos,
			redos
		};
	}
	_handleRemoveRowCommand(config, unitId, subUnitId) {
		const filterModel = this._sheetsFilterService.getFilterModel(unitId, subUnitId);
		if (!filterModel) return this._handleNull();
		const filterRange = filterModel.getRange();
		const { startRow, endRow } = filterRange;
		const { startRow: removeStartRow, endRow: removeEndRow } = config.range;
		if (removeStartRow > endRow) return this._handleNull();
		if (removeEndRow < startRow) return {
			undos: [{
				id: SetSheetsFilterRangeMutation.id,
				params: {
					range: filterRange,
					unitId,
					subUnitId
				}
			}],
			redos: [{
				id: SetSheetsFilterRangeMutation.id,
				params: {
					range: {
						...filterRange,
						startRow: startRow - (removeEndRow - removeStartRow + 1),
						endRow: endRow - (removeEndRow - removeStartRow + 1)
					},
					unitId,
					subUnitId
				}
			}]
		};
		const redos = [];
		const undos = [];
		const filterColumn = filterModel.getAllFilterColumns();
		const filterHeaderIsRemoved = startRow <= removeEndRow && startRow >= removeStartRow;
		undos.push({
			id: SetSheetsFilterRangeMutation.id,
			params: {
				range: filterRange,
				unitId,
				subUnitId
			}
		});
		const count = Math.min(removeEndRow, endRow) - Math.max(removeStartRow, startRow) + 1;
		if (count === endRow - startRow + 1 || filterHeaderIsRemoved) {
			const removeFilterRangeMutationParams = {
				unitId,
				subUnitId
			};
			redos.push({
				id: RemoveSheetsFilterMutation.id,
				params: removeFilterRangeMutationParams
			});
			filterColumn.forEach((column) => {
				const [offset, filter] = column;
				const setCriteriaMutationParams = {
					unitId,
					subUnitId,
					col: offset,
					criteria: {
						...filter.serialize(),
						colId: offset
					}
				};
				undos.push({
					id: SetSheetsFilterCriteriaMutation.id,
					params: setCriteriaMutationParams
				});
			});
		} else {
			var _this$_univerInstance;
			const worksheet = (_this$_univerInstance = this._univerInstanceService.getUniverSheetInstance(unitId)) === null || _this$_univerInstance === void 0 ? void 0 : _this$_univerInstance.getSheetBySheetId(subUnitId);
			if (!worksheet) return this._handleNull();
			const hiddenRowCount = this._getFilteredRowCount(worksheet, removeStartRow, removeEndRow);
			const afterStartRow = Math.min(startRow, removeStartRow);
			const afterEndRow = afterStartRow + (endRow - startRow) - count + hiddenRowCount;
			const setFilterRangeMutationParams = {
				unitId,
				subUnitId,
				range: {
					...filterRange,
					startRow: afterStartRow,
					endRow: afterEndRow
				}
			};
			redos.push({
				id: SetSheetsFilterRangeMutation.id,
				params: setFilterRangeMutationParams
			});
		}
		return {
			undos: mergeSetFilterCriteria(undos),
			redos: mergeSetFilterCriteria(redos)
		};
	}
	_getFilteredRowCount(worksheet, startRow, endRow) {
		let count = 0;
		for (let row = startRow; row <= endRow; row++) if (worksheet.getRowFiltered(row)) count++;
		return count;
	}
	handleMoveColsCommand({ fromRange, toRange }, unitId, subUnitId) {
		var _filterModel$getRange4;
		const filterModel = this._sheetsFilterService.getFilterModel(unitId, subUnitId);
		const filterRange = (_filterModel$getRange4 = filterModel === null || filterModel === void 0 ? void 0 : filterModel.getRange()) !== null && _filterModel$getRange4 !== void 0 ? _filterModel$getRange4 : null;
		if (!filterModel || !filterRange) return this._handleNull();
		const { startColumn, endColumn } = filterRange;
		if (fromRange.endColumn < startColumn && toRange.startColumn <= startColumn || fromRange.startColumn > endColumn && toRange.endColumn > endColumn) return this._handleNull();
		const redos = [];
		const undos = [];
		const filterCol = {};
		for (let col = startColumn; col <= endColumn; col++) filterCol[col] = {
			colIndex: col,
			filter: filterModel.getFilterColumn(col)
		};
		(0, _univerjs_core.moveMatrixArray)(fromRange.startColumn, fromRange.endColumn - fromRange.startColumn + 1, toRange.startColumn, filterCol);
		let startBorder = filterRange.startColumn;
		let endBorder = filterRange.endColumn;
		if (startColumn >= fromRange.startColumn && startColumn <= fromRange.endColumn && toRange.startColumn > fromRange.startColumn && fromRange.endColumn < endColumn) startBorder = fromRange.endColumn + 1;
		if (endColumn >= fromRange.startColumn && endColumn <= fromRange.endColumn && toRange.startColumn < fromRange.startColumn && fromRange.startColumn > startColumn) endBorder = fromRange.startColumn - 1;
		const numberCols = Object.keys(filterCol).map((col) => Number(col));
		const newEnd = numberCols.find((col) => filterCol[col].colIndex === endBorder);
		const newStart = numberCols.find((col) => filterCol[col].colIndex === startBorder);
		numberCols.forEach((col) => {
			const { colIndex: oldColIndex, filter } = filterCol[col];
			const newColIndex = col;
			if (filter) {
				var _filterCol$oldColInde;
				if (newColIndex >= newStart && newColIndex <= newEnd) {
					var _filterModel$getFilte;
					const setCriteriaMutationParams = {
						unitId,
						subUnitId,
						col: newColIndex,
						criteria: {
							...filter.serialize(),
							colId: newColIndex
						}
					};
					const undoSetCriteriaMutationParams = {
						unitId,
						subUnitId,
						col: newColIndex,
						criteria: filterModel.getFilterColumn(newColIndex) ? {
							...(_filterModel$getFilte = filterModel.getFilterColumn(newColIndex)) === null || _filterModel$getFilte === void 0 ? void 0 : _filterModel$getFilte.serialize(),
							colId: newColIndex
						} : null
					};
					redos.push({
						id: SetSheetsFilterCriteriaMutation.id,
						params: setCriteriaMutationParams
					});
					undos.push({
						id: SetSheetsFilterCriteriaMutation.id,
						params: undoSetCriteriaMutationParams
					});
				}
				if (!((_filterCol$oldColInde = filterCol[oldColIndex]) === null || _filterCol$oldColInde === void 0 ? void 0 : _filterCol$oldColInde.filter)) {
					const setCriteriaMutationParams = {
						unitId,
						subUnitId,
						col: oldColIndex,
						criteria: null
					};
					redos.push({
						id: SetSheetsFilterCriteriaMutation.id,
						params: setCriteriaMutationParams
					});
					undos.push({
						id: SetSheetsFilterCriteriaMutation.id,
						params: {
							unitId,
							subUnitId,
							col: oldColIndex,
							criteria: {
								...filter.serialize(),
								colId: oldColIndex
							}
						}
					});
				}
			}
		});
		if (startColumn !== newStart || endColumn !== newEnd) {
			const setFilterRangeMutationParams = {
				unitId,
				subUnitId,
				range: {
					...filterRange,
					startColumn: newStart,
					endColumn: newEnd
				}
			};
			redos.unshift({
				id: SetSheetsFilterRangeMutation.id,
				params: setFilterRangeMutationParams
			});
			undos.unshift({
				id: SetSheetsFilterRangeMutation.id,
				params: {
					range: filterRange,
					unitId,
					subUnitId
				}
			});
		}
		return {
			undos,
			redos
		};
	}
	_handleMoveRowsCommand(config, unitId, subUnitId) {
		var _filterModel$getRange5;
		const filterModel = this._sheetsFilterService.getFilterModel(unitId, subUnitId);
		const filterRange = (_filterModel$getRange5 = filterModel === null || filterModel === void 0 ? void 0 : filterModel.getRange()) !== null && _filterModel$getRange5 !== void 0 ? _filterModel$getRange5 : null;
		if (!filterModel || !filterRange) return this._handleNull();
		const { startRow, endRow } = filterRange;
		const { fromRange, toRange } = config;
		if (fromRange.endRow < startRow && toRange.startRow <= startRow || fromRange.startRow > endRow && toRange.endRow > endRow) return this._handleNull();
		const redos = [];
		const undos = [];
		const filterRow = {};
		for (let row = startRow; row <= endRow; row++) filterRow[row] = { oldIndex: row };
		const startBorder = startRow;
		let endBorder = endRow;
		if (endRow >= fromRange.startRow && endRow <= fromRange.endRow && toRange.startRow < fromRange.startRow && fromRange.startRow > startRow) endBorder = fromRange.startRow - 1;
		(0, _univerjs_core.moveMatrixArray)(fromRange.startRow, fromRange.endRow - fromRange.startRow + 1, toRange.startRow, filterRow);
		const numberRows = Object.keys(filterRow).map((row) => Number(row));
		const newEnd = numberRows.find((row) => filterRow[row].oldIndex === endBorder);
		const newStart = numberRows.find((row) => filterRow[row].oldIndex === startBorder);
		if (startRow !== newStart || endRow !== newEnd) {
			const setFilterRangeMutationParams = {
				unitId,
				subUnitId,
				range: {
					...filterRange,
					startRow: newStart,
					endRow: newEnd
				}
			};
			redos.push({
				id: SetSheetsFilterRangeMutation.id,
				params: setFilterRangeMutationParams
			}, {
				id: ReCalcSheetsFilterMutation.id,
				params: {
					unitId,
					subUnitId
				}
			});
			undos.push({
				id: SetSheetsFilterRangeMutation.id,
				params: {
					range: filterRange,
					unitId,
					subUnitId
				}
			}, {
				id: ReCalcSheetsFilterMutation.id,
				params: {
					unitId,
					subUnitId
				}
			});
		}
		return {
			redos,
			undos
		};
	}
	_handleMoveRangeCommand(config, unitId, subUnitId) {
		const { fromRange, toRange } = config;
		const filterModel = this._sheetsFilterService.getFilterModel(unitId, subUnitId);
		if (!filterModel) return this._handleNull();
		const filterRange = filterModel.getRange();
		if (!filterRange) return this._handleNull();
		const redos = [];
		const undos = [];
		if (_univerjs_core.Rectangle.contains(fromRange, filterRange)) {
			const rowOffset = filterRange.startRow - fromRange.startRow;
			const colOffset = filterRange.startColumn - fromRange.startColumn;
			const newFilterRange = {
				startRow: toRange.startRow + rowOffset,
				startColumn: toRange.startColumn + colOffset,
				endRow: toRange.startRow + rowOffset + (filterRange.endRow - filterRange.startRow),
				endColumn: toRange.startColumn + colOffset + (filterRange.endColumn - filterRange.startColumn)
			};
			const removeFilter = {
				id: RemoveSheetsFilterMutation.id,
				params: {
					unitId,
					subUnitId
				}
			};
			const setNewFilterRange = {
				id: SetSheetsFilterRangeMutation.id,
				params: {
					unitId,
					subUnitId,
					range: newFilterRange
				}
			};
			const setOldFilterRange = {
				id: SetSheetsFilterRangeMutation.id,
				params: {
					unitId,
					subUnitId,
					range: filterRange
				}
			};
			redos.push(removeFilter, setNewFilterRange);
			undos.push(removeFilter, setOldFilterRange);
			const filterColumn = filterModel.getAllFilterColumns();
			const moveColDelta = toRange.startColumn - fromRange.startColumn;
			filterColumn.forEach((column) => {
				const [col, criteria] = column;
				if (criteria) {
					redos.push({
						id: SetSheetsFilterCriteriaMutation.id,
						params: {
							unitId,
							subUnitId,
							col: col + moveColDelta,
							criteria: {
								...criteria.serialize(),
								colId: col + moveColDelta
							}
						}
					});
					undos.push({
						id: SetSheetsFilterCriteriaMutation.id,
						params: {
							unitId,
							subUnitId,
							col,
							criteria: {
								...criteria.serialize(),
								colId: col
							}
						}
					});
				}
			});
		} else if (_univerjs_core.Rectangle.intersects(toRange, filterRange)) {
			const newFilterRange = {
				...filterRange,
				endRow: Math.max(filterRange.endRow, toRange.endRow)
			};
			redos.push({
				id: SetSheetsFilterRangeMutation.id,
				params: {
					unitId,
					subUnitId,
					range: newFilterRange
				}
			});
			undos.push({
				id: SetSheetsFilterRangeMutation.id,
				params: {
					unitId,
					subUnitId,
					range: filterRange
				}
			});
		}
		return {
			redos,
			undos
		};
	}
	_handleRemoveSheetCommand(config, unitId, subUnitId) {
		const filterModel = this._sheetsFilterService.getFilterModel(unitId, subUnitId);
		if (!filterModel) return this._handleNull();
		const filterRange = filterModel.getRange();
		if (!filterRange) return this._handleNull();
		const redos = [];
		const undos = [];
		filterModel.getAllFilterColumns().forEach(([col, filter]) => {
			undos.push({
				id: SetSheetsFilterCriteriaMutation.id,
				params: {
					unitId,
					subUnitId,
					col,
					criteria: {
						...filter.serialize(),
						colId: col
					}
				}
			});
		});
		redos.push({
			id: RemoveSheetsFilterMutation.id,
			params: {
				unitId,
				subUnitId,
				range: filterRange
			}
		});
		undos.unshift({
			id: SetSheetsFilterRangeMutation.id,
			params: {
				range: filterRange,
				unitId,
				subUnitId
			}
		});
		return {
			undos,
			redos
		};
	}
	_handleCopySheetCommand(unitId, subUnitId, targetSubUnitId) {
		const filterModel = this._sheetsFilterService.getFilterModel(unitId, subUnitId);
		if (!filterModel) return this._handleNull();
		const filterRange = filterModel.getRange();
		if (!filterRange) return this._handleNull();
		const redos = [];
		const undos = [];
		const preUndos = [];
		const preRedos = [];
		filterModel.getAllFilterColumns().forEach(([col, filter]) => {
			redos.push({
				id: SetSheetsFilterCriteriaMutation.id,
				params: {
					unitId,
					subUnitId: targetSubUnitId,
					col,
					criteria: {
						...filter.serialize(),
						colId: col
					}
				}
			});
			preUndos.push({
				id: SetSheetsFilterCriteriaMutation.id,
				params: {
					unitId,
					subUnitId: targetSubUnitId,
					col,
					criteria: null
				}
			});
		});
		preUndos.push({
			id: RemoveSheetsFilterMutation.id,
			params: {
				unitId,
				subUnitId: targetSubUnitId,
				range: filterRange
			}
		});
		redos.unshift({
			id: SetSheetsFilterRangeMutation.id,
			params: {
				range: filterRange,
				unitId,
				subUnitId: targetSubUnitId
			}
		});
		return {
			undos,
			redos,
			preUndos,
			preRedos
		};
	}
	_handleNull() {
		return {
			redos: [],
			undos: []
		};
	}
	_initRowFilteredInterceptor() {
		this.disposeWithMe(this._sheetInterceptorService.intercept(_univerjs_sheets.INTERCEPTOR_POINT.ROW_FILTERED, { handler: (filtered, rowLocation) => {
			var _this$_sheetsFilterSe2, _this$_sheetsFilterSe3;
			if (filtered) return true;
			return (_this$_sheetsFilterSe2 = (_this$_sheetsFilterSe3 = this._sheetsFilterService.getFilterModel(rowLocation.unitId, rowLocation.subUnitId)) === null || _this$_sheetsFilterSe3 === void 0 ? void 0 : _this$_sheetsFilterSe3.isRowFiltered(rowLocation.row)) !== null && _this$_sheetsFilterSe2 !== void 0 ? _this$_sheetsFilterSe2 : false;
		} }));
	}
	_moveCriteria(unitId, subUnitId, target, step) {
		const defaultSetCriteriaMutationParams = {
			unitId,
			subUnitId,
			criteria: null,
			col: -1
		};
		const oldUndos = [];
		const oldRedos = [];
		const newUndos = [];
		const newRedos = [];
		target.forEach((column) => {
			const [offset, filter] = column;
			oldRedos.push({
				id: SetSheetsFilterCriteriaMutation.id,
				params: {
					...defaultSetCriteriaMutationParams,
					col: offset
				}
			});
			oldUndos.push({
				id: SetSheetsFilterCriteriaMutation.id,
				params: {
					...defaultSetCriteriaMutationParams,
					col: offset,
					criteria: {
						...filter.serialize(),
						colId: offset
					}
				}
			});
		});
		target.forEach((column) => {
			const [offset, filter] = column;
			newRedos.push({
				id: SetSheetsFilterCriteriaMutation.id,
				params: {
					...defaultSetCriteriaMutationParams,
					col: offset + step,
					criteria: {
						...filter.serialize(),
						colId: offset + step
					}
				}
			});
			newUndos.push({
				id: SetSheetsFilterCriteriaMutation.id,
				params: {
					...defaultSetCriteriaMutationParams,
					col: offset + step,
					criteria: null
				}
			});
		});
		return {
			newRange: {
				redos: newRedos,
				undos: newUndos
			},
			oldRange: {
				redos: oldRedos,
				undos: oldUndos
			}
		};
	}
	_commandExecutedListener() {
		this.disposeWithMe(this._commandService.onCommandExecuted((command, options) => {
			const { unitId, subUnitId } = command.params || {};
			const filterModel = this._sheetsFilterService.getFilterModel(unitId, subUnitId);
			if (!filterModel) return;
			const filteredOutRows = Array.from(filterModel.filteredOutRows).sort((a, b) => a - b);
			const newFilteredOutRows = [];
			let changed = false;
			if (command.id === _univerjs_sheets.RemoveRowMutation.id) {
				const { startRow, endRow } = command.params.range;
				const filterOutInRemove = filteredOutRows.filter((row) => row >= startRow && row <= endRow);
				filteredOutRows.forEach((row) => {
					if (row < startRow) newFilteredOutRows.push(row);
					else {
						changed = true;
						if (row <= endRow) {
							const newIndex = Math.max(startRow, newFilteredOutRows.length ? newFilteredOutRows[newFilteredOutRows.length - 1] + 1 : startRow);
							newFilteredOutRows.push(newIndex);
						} else newFilteredOutRows.push(row - (endRow - startRow + 1 - filterOutInRemove.length));
					}
				});
			}
			if (command.id === _univerjs_sheets.InsertRowMutation.id) {
				const { startRow, endRow } = command.params.range;
				filteredOutRows.forEach((row) => {
					if (row >= startRow) {
						changed = true;
						newFilteredOutRows.push(row + (endRow - startRow + 1));
					} else newFilteredOutRows.push(row);
				});
			}
			if (changed) filterModel.filteredOutRows = new Set(newFilteredOutRows);
			if (command.id === _univerjs_sheets.SetRangeValuesMutation.id && !(options === null || options === void 0 ? void 0 : options.onlyLocal)) {
				const extendRegion = this._getExtendRegion(unitId, subUnitId);
				if (extendRegion) {
					const cellValue = command.params.cellValue;
					if (cellValue) for (let col = extendRegion.startColumn; col <= extendRegion.endColumn; col++) {
						var _cellValue$extendRegi;
						const cell = cellValue === null || cellValue === void 0 || (_cellValue$extendRegi = cellValue[extendRegion.startRow]) === null || _cellValue$extendRegi === void 0 ? void 0 : _cellValue$extendRegi[col];
						if (cell && this._cellHasValue(cell)) {
							var _this$_univerInstance2;
							const worksheet = (_this$_univerInstance2 = this._univerInstanceService.getUnit(unitId)) === null || _this$_univerInstance2 === void 0 ? void 0 : _this$_univerInstance2.getSheetBySheetId(subUnitId);
							if (worksheet) {
								const extendedRange = (0, _univerjs_sheets.expandToContinuousRange)(extendRegion, { down: true }, worksheet);
								const filterModel = this._sheetsFilterService.getFilterModel(unitId, subUnitId);
								const filterRange = filterModel.getRange();
								filterModel.setRange({
									...filterRange,
									endRow: extendedRange.endRow
								});
								this._registerRefRange(unitId, subUnitId);
							}
						}
					}
				}
			}
		}));
	}
	_getExtendRegion(unitId, subUnitId) {
		var _this$_univerInstance3;
		const filterModel = this._sheetsFilterService.getFilterModel(unitId, subUnitId);
		if (!filterModel) return null;
		const worksheet = (_this$_univerInstance3 = this._univerInstanceService.getUnit(unitId)) === null || _this$_univerInstance3 === void 0 ? void 0 : _this$_univerInstance3.getSheetBySheetId(subUnitId);
		if (!worksheet) return null;
		const filterRange = filterModel.getRange();
		if (!filterRange) return null;
		const maxRowIndex = worksheet.getRowCount() - 1;
		const rowManager = worksheet.getRowManager();
		for (let row = filterRange.endRow + 1; row <= maxRowIndex; row++) if (rowManager.getRowRawVisible(row)) return {
			startRow: row,
			endRow: row,
			startColumn: filterRange.startColumn,
			endColumn: filterRange.endColumn
		};
		return null;
	}
	_initErrorHandling() {
		this.disposeWithMe(this._commandService.beforeCommandExecuted((command) => {
			const params = command.params;
			const target = (0, _univerjs_sheets.getSheetCommandTarget)(this._univerInstanceService, params);
			if (!target) return;
			const { subUnitId, unitId } = target;
			const filterModel = this._sheetsFilterService.getFilterModel(unitId, subUnitId);
			if (!filterModel) return;
			const filterRange = filterModel.getRange();
			if (command.id === _univerjs_sheets.MoveRowsCommand.id && params.fromRange.startRow <= filterRange.startRow && params.fromRange.endRow < filterRange.endRow && params.fromRange.endRow >= filterRange.startRow) {
				this._sheetsFilterService.setFilterErrorMsg("sheets-filter.msg.filter-header-forbidden");
				throw new Error("[SheetsFilterController]: Cannot move header row of filter");
			}
		}));
	}
	_cellHasValue(cell) {
		const values = Object.values(cell);
		if (values.length === 0 || values.every((v) => v == null)) return false;
		return true;
	}
};
SheetsFilterController = __decorate([
	__decorateParam(0, _univerjs_core.ICommandService),
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_sheets.SheetInterceptorService)),
	__decorateParam(2, (0, _univerjs_core.Inject)(SheetsFilterService)),
	__decorateParam(3, _univerjs_core.IUniverInstanceService),
	__decorateParam(4, (0, _univerjs_core.Inject)(_univerjs_sheets.RefRangeService)),
	__decorateParam(5, (0, _univerjs_core.Optional)(_univerjs_rpc.DataSyncPrimaryController)),
	__decorateParam(6, (0, _univerjs_core.Inject)(_univerjs_sheets.ZebraCrossingCacheController))
], SheetsFilterController);

//#endregion
//#region src/controllers/sheets-filter-sync.controller.ts
const sheetsFilterOnlyLocalMutationIds = [SetSheetsFilterCriteriaMutation.id, ReCalcSheetsFilterMutation.id];
const effectedByOnlyLocalMutationIds = [
	_univerjs_sheets.InsertColMutation.id,
	_univerjs_sheets.RemoveColMutation.id,
	_univerjs_sheets.MoveColsMutation.id
];
let SheetsFilterSyncController = class SheetsFilterSyncController extends _univerjs_core.Disposable {
	get visible() {
		return this._visible$.getValue();
	}
	get enabled() {
		return this._enabled$.getValue();
	}
	constructor(_sheetsFilterController, _commandService, _configService) {
		super();
		this._sheetsFilterController = _sheetsFilterController;
		this._commandService = _commandService;
		this._configService = _configService;
		_defineProperty(this, "_d", new _univerjs_core.DisposableCollection());
		_defineProperty(this, "_visible$", new rxjs.BehaviorSubject(false));
		_defineProperty(this, "visible$", this._visible$.asObservable());
		_defineProperty(this, "_enabled$", new rxjs.BehaviorSubject(true));
		_defineProperty(this, "enabled$", this._enabled$.asObservable());
		const config = this._configService.getConfig(SHEETS_FILTER_PLUGIN_CONFIG_KEY);
		if (config === null || config === void 0 ? void 0 : config.enableSyncSwitch) {
			this._visible$.next(true);
			if (typeof config.enableSyncSwitch === "object") {
				var _config$enableSyncSwi;
				this.setEnabled((_config$enableSyncSwi = config.enableSyncSwitch.defaultValue) !== null && _config$enableSyncSwi !== void 0 ? _config$enableSyncSwi : true);
			}
		}
	}
	setEnabled(enabled) {
		this._enabled$.next(enabled);
		if (enabled) this._d.dispose();
		else this._initOnlyLocalListener();
	}
	_initOnlyLocalListener() {
		this._d.add(this._commandService.beforeCommandExecuted((commandInfo, options) => {
			if (sheetsFilterOnlyLocalMutationIds.includes(commandInfo.id)) {
				if (!options) options = {};
				options.onlyLocal = true;
			}
		}));
		this._d.add(this._commandService.onCommandExecuted((commandInfo, options) => {
			if (effectedByOnlyLocalMutationIds.includes(commandInfo.id) && (options === null || options === void 0 ? void 0 : options.fromCollab)) {
				if (commandInfo.id === _univerjs_sheets.InsertColMutation.id) {
					const { range, unitId, subUnitId } = commandInfo.params;
					const { redos } = this._sheetsFilterController.handleInsertColCommand(range, unitId, subUnitId);
					(0, _univerjs_core.sequenceExecute)(redos, this._commandService, options);
				} else if (commandInfo.id === _univerjs_sheets.RemoveColMutation.id) {
					const { range, unitId, subUnitId } = commandInfo.params;
					const { redos } = this._sheetsFilterController.handleRemoveColCommand(range, unitId, subUnitId);
					(0, _univerjs_core.sequenceExecute)(redos, this._commandService, options);
				} else if (commandInfo.id === _univerjs_sheets.MoveColsMutation.id) {
					const { sourceRange: fromRange, targetRange: toRange, unitId, subUnitId } = commandInfo.params;
					const { redos } = this._sheetsFilterController.handleMoveColsCommand({
						fromRange,
						toRange
					}, unitId, subUnitId);
					(0, _univerjs_core.sequenceExecute)(redos, this._commandService, options);
				}
			}
		}));
	}
};
SheetsFilterSyncController = __decorate([
	__decorateParam(0, (0, _univerjs_core.Inject)(SheetsFilterController)),
	__decorateParam(1, _univerjs_core.ICommandService),
	__decorateParam(2, _univerjs_core.IConfigService)
], SheetsFilterSyncController);

//#endregion
//#region package.json
var name = "@univerjs/sheets-filter";
var version = "0.25.1";

//#endregion
//#region src/services/sheet-filter-formula.service.ts
let SheetsFilterFormulaService = class SheetsFilterFormulaService extends _univerjs_core.Disposable {
	constructor(_activeDirtyManagerService, _sheetRowFilteredService, _sheetsFilterService, _univerInstanceService) {
		super();
		this._activeDirtyManagerService = _activeDirtyManagerService;
		this._sheetRowFilteredService = _sheetRowFilteredService;
		this._sheetsFilterService = _sheetsFilterService;
		this._univerInstanceService = _univerInstanceService;
		this._initFormulaDirtyRange();
		this._registerSheetRowFiltered();
	}
	_initFormulaDirtyRange() {
		FILTER_MUTATIONS.forEach((commandId) => {
			this._activeDirtyManagerService.register(commandId, {
				commandId,
				getDirtyData: (commandInfo) => {
					const { unitId, subUnitId } = commandInfo.params;
					return {
						dirtyRanges: this._getHideRowMutation(unitId, subUnitId),
						clearDependencyTreeCache: { [unitId]: { [subUnitId]: "1" } }
					};
				}
			});
		});
	}
	_getHideRowMutation(unitId, subUnitId) {
		var _this$_sheetsFilterSe, _this$_univerInstance;
		const range = (_this$_sheetsFilterSe = this._sheetsFilterService.getFilterModel(unitId, subUnitId)) === null || _this$_sheetsFilterSe === void 0 ? void 0 : _this$_sheetsFilterSe.getRange();
		const sheet = (_this$_univerInstance = this._univerInstanceService.getUnit(unitId)) === null || _this$_univerInstance === void 0 ? void 0 : _this$_univerInstance.getSheetBySheetId(subUnitId);
		if (range == null || sheet == null) return [];
		const { startRow, endRow } = range;
		return [{
			unitId,
			sheetId: subUnitId,
			range: {
				startRow,
				startColumn: 0,
				endRow,
				endColumn: sheet.getColumnCount() - 1
			}
		}];
	}
	_registerSheetRowFiltered() {
		this._sheetRowFilteredService.register((unitId, subUnitId, row) => {
			var _this$_sheetsFilterSe2, _this$_sheetsFilterSe3;
			return (_this$_sheetsFilterSe2 = (_this$_sheetsFilterSe3 = this._sheetsFilterService.getFilterModel(unitId, subUnitId)) === null || _this$_sheetsFilterSe3 === void 0 ? void 0 : _this$_sheetsFilterSe3.isRowFiltered(row)) !== null && _this$_sheetsFilterSe2 !== void 0 ? _this$_sheetsFilterSe2 : false;
		});
	}
};
SheetsFilterFormulaService = __decorate([
	__decorateParam(0, (0, _univerjs_core.Inject)(_univerjs_engine_formula.IActiveDirtyManagerService)),
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_engine_formula.ISheetRowFilteredService)),
	__decorateParam(2, (0, _univerjs_core.Inject)(SheetsFilterService)),
	__decorateParam(3, _univerjs_core.IUniverInstanceService)
], SheetsFilterFormulaService);

//#endregion
//#region src/plugin.ts
let UniverSheetsFilterPlugin = class UniverSheetsFilterPlugin extends _univerjs_core.Plugin {
	constructor(_config = defaultPluginConfig, _injector, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		const { ...rest } = (0, _univerjs_core.merge)({}, defaultPluginConfig, this._config);
		this._configService.setConfig(SHEETS_FILTER_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		[
			[SheetsFilterFormulaService],
			[SheetsFilterService],
			[SheetsFilterController],
			[SheetsFilterSyncController]
		].forEach((d) => this._injector.add(d));
	}
	onReady() {
		(0, _univerjs_core.touchDependencies)(this._injector, [
			[SheetsFilterFormulaService],
			[SheetsFilterController],
			[SheetsFilterSyncController]
		]);
	}
};
_defineProperty(UniverSheetsFilterPlugin, "type", _univerjs_core.UniverInstanceType.UNIVER_SHEET);
_defineProperty(UniverSheetsFilterPlugin, "pluginName", SHEET_FILTER_SNAPSHOT_ID);
_defineProperty(UniverSheetsFilterPlugin, "packageName", name);
_defineProperty(UniverSheetsFilterPlugin, "version", version);
UniverSheetsFilterPlugin = __decorate([__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector)), __decorateParam(2, _univerjs_core.IConfigService)], UniverSheetsFilterPlugin);

//#endregion
exports.ClearSheetsFilterCriteriaCommand = ClearSheetsFilterCriteriaCommand;
exports.CustomFilterOperator = CustomFilterOperator;
exports.FILTER_MUTATIONS = FILTER_MUTATIONS;
exports.FilterBy = FilterBy;
exports.FilterColumn = FilterColumn;
exports.FilterModel = FilterModel;
exports.ReCalcSheetsFilterCommand = ReCalcSheetsFilterCommand;
exports.ReCalcSheetsFilterMutation = ReCalcSheetsFilterMutation;
exports.RemoveSheetFilterCommand = RemoveSheetFilterCommand;
exports.RemoveSheetsFilterMutation = RemoveSheetsFilterMutation;
exports.SHEET_FILTER_SNAPSHOT_ID = SHEET_FILTER_SNAPSHOT_ID;
exports.SetSheetFilterRangeCommand = SetSheetFilterRangeCommand;
exports.SetSheetsFilterCriteriaCommand = SetSheetsFilterCriteriaCommand;
exports.SetSheetsFilterCriteriaMutation = SetSheetsFilterCriteriaMutation;
exports.SetSheetsFilterRangeMutation = SetSheetsFilterRangeMutation;
Object.defineProperty(exports, 'SheetsFilterService', {
  enumerable: true,
  get: function () {
    return SheetsFilterService;
  }
});
Object.defineProperty(exports, 'SheetsFilterSyncController', {
  enumerable: true,
  get: function () {
    return SheetsFilterSyncController;
  }
});
exports.SmartToggleSheetsFilterCommand = SmartToggleSheetsFilterCommand;
Object.defineProperty(exports, 'UniverSheetsFilterPlugin', {
  enumerable: true,
  get: function () {
    return UniverSheetsFilterPlugin;
  }
});
exports.equals = equals;
exports.getCustomFilterFn = getCustomFilterFn;
exports.greaterThan = greaterThan;
exports.greaterThanOrEqualTo = greaterThanOrEqualTo;
exports.lessThan = lessThan;
exports.lessThanOrEqualTo = lessThanOrEqualTo;
exports.notEquals = notEquals;
