// @univerjs/sheets-sort/lib/cjs/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
let _univerjs_core = require("@univerjs/core");
let _univerjs_sheets = require("@univerjs/sheets");
let _univerjs_engine_formula = require("@univerjs/engine-formula");

//#region src/services/interface.ts
let SortType = /* @__PURE__ */ function(SortType) {
	SortType["DESC"] = "desc";
	SortType["ASC"] = "asc";
	return SortType;
}({});

//#endregion
//#region src/controllers/utils.ts
const removeStringSymbol = (str) => {
	return str.replace(/-/gi, "").replace(/'/gi, "");
};
const compareNull = (a1, a2) => {
	const isA1Null = a1 === null || a1 === "";
	const isA2Null = a2 === null || a2 === "";
	if (isA1Null && isA2Null) return 0;
	if (isA1Null) return 1;
	if (isA2Null) return -1;
	return null;
};
const compareNumber = (a1, a2, type) => {
	const isA1Num = typeof a1 === "number";
	const isA2Num = typeof a2 === "number";
	if (isA1Num && isA2Num) {
		if (a1 < a2) return type === "asc" ? -1 : 1;
		if (a1 > a2) return type === "asc" ? 1 : -1;
		return 0;
	}
	if (isA1Num) return type === "asc" ? 1 : -1;
	if (isA2Num) return type === "asc" ? -1 : 1;
	return null;
};
const compareString = (a1, a2, type) => {
	const isA1Str = typeof a1 === "string";
	const isA2Str = typeof a2 === "string";
	if (isA1Str) a1 = removeStringSymbol(a1.toLocaleLowerCase());
	if (isA2Str) a2 = removeStringSymbol(a2.toLocaleLowerCase());
	if (!isA1Str && !isA2Str) return null;
	if (isA1Str && isA2Str) {
		const a1AsString = a1;
		const a2AsString = a2;
		if (a1AsString < a2AsString) return type === "asc" ? -1 : 1;
		if (a1AsString > a2AsString) return type === "asc" ? 1 : -1;
		return 0;
	}
	if (isA1Str) return type === "asc" ? 1 : -1;
	if (isA2Str) return type === "asc" ? -1 : 1;
	return null;
};
const isNullValue = (cell) => {
	if (!cell) return true;
	if (Object.keys(cell).length === 0) return true;
	if ((cell === null || cell === void 0 ? void 0 : cell.v) == null && (cell === null || cell === void 0 ? void 0 : cell.p) == null) return true;
	return false;
};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/services/sheets-sort.service.ts
let SheetsSortService = class SheetsSortService extends _univerjs_core.Disposable {
	constructor(_univerInstanceService, _commandService, _formulaDataModel) {
		super();
		this._univerInstanceService = _univerInstanceService;
		this._commandService = _commandService;
		this._formulaDataModel = _formulaDataModel;
		_defineProperty(this, "_compareFns", []);
	}
	mergeCheck(location) {
		var _this$_univerInstance;
		const { unitId, subUnitId, range } = location;
		const sheet = (_this$_univerInstance = this._univerInstanceService.getUnit(unitId)) === null || _this$_univerInstance === void 0 ? void 0 : _this$_univerInstance.getSheetBySheetId(subUnitId);
		if (!sheet) return false;
		const mergeDataInRange = sheet.getMergeData().filter((merge) => _univerjs_core.Rectangle.contains(range, merge));
		if (mergeDataInRange.length === 0) return true;
		return isRangeDividedEqually(range, mergeDataInRange);
	}
	emptyCheck(location) {
		var _this$_univerInstance2;
		const { unitId, subUnitId, range } = location;
		const sheet = (_this$_univerInstance2 = this._univerInstanceService.getUnit(unitId)) === null || _this$_univerInstance2 === void 0 ? void 0 : _this$_univerInstance2.getSheetBySheetId(subUnitId);
		if (!sheet) return false;
		for (let row = range.startRow; row <= range.endRow; row++) for (let col = range.startColumn; col <= range.endColumn; col++) if (!isNullValue(sheet.getCellRaw(row, col))) return true;
		return false;
	}
	singleCheck(location) {
		if (location.range.startRow === location.range.endRow) return false;
		return true;
	}
	formulaCheck(location) {
		var _this$_formulaDataMod;
		const { unitId, subUnitId, range } = location;
		const arrayFormulaRange = (_this$_formulaDataMod = this._formulaDataModel.getArrayFormulaRange()) === null || _this$_formulaDataMod === void 0 || (_this$_formulaDataMod = _this$_formulaDataMod[unitId]) === null || _this$_formulaDataMod === void 0 ? void 0 : _this$_formulaDataMod[subUnitId];
		for (const row in arrayFormulaRange) {
			const rowData = arrayFormulaRange[Number(row)];
			for (const col in rowData) {
				const arrayFormula = rowData[Number(col)];
				if (arrayFormula && _univerjs_core.Rectangle.intersects(range, arrayFormula)) return false;
			}
		}
		return true;
	}
	registerCompareFn(fn) {
		this._compareFns.unshift(fn);
	}
	getAllCompareFns() {
		return this._compareFns;
	}
	applySort(sortOption, unitId, subUnitId) {
		var _sortOption$hasTitle;
		const { unitId: _unitId, subUnitId: _subUnitId } = (0, _univerjs_sheets.getSheetCommandTarget)(this._univerInstanceService) || {};
		this._commandService.executeCommand(SortRangeCommand.id, {
			orderRules: sortOption.orderRules,
			range: sortOption.range,
			hasTitle: (_sortOption$hasTitle = sortOption.hasTitle) !== null && _sortOption$hasTitle !== void 0 ? _sortOption$hasTitle : false,
			unitId: unitId || _unitId,
			subUnitId: subUnitId || _subUnitId
		});
	}
};
SheetsSortService = __decorate([
	__decorateParam(0, _univerjs_core.IUniverInstanceService),
	__decorateParam(1, _univerjs_core.ICommandService),
	__decorateParam(2, (0, _univerjs_core.Inject)(_univerjs_engine_formula.FormulaDataModel))
], SheetsSortService);
function isRangeDividedEqually(range, merges) {
	const rangeRows = range.endRow - range.startRow + 1;
	const rangeCols = range.endColumn - range.startColumn + 1;
	let mergeRows = null;
	let mergeCols = null;
	const totalArea = rangeRows * rangeCols;
	let totalMergeArea = 0;
	for (const merge of merges) if (merge.startRow >= range.startRow && merge.endRow <= range.endRow && merge.startColumn >= range.startColumn && merge.endColumn <= range.endColumn) {
		const currentMergeRows = merge.endRow - merge.startRow + 1;
		const currentMergeCols = merge.endColumn - merge.startColumn + 1;
		if (mergeRows === null && mergeCols === null) {
			mergeRows = currentMergeRows;
			mergeCols = currentMergeCols;
		} else if (currentMergeRows !== mergeRows || currentMergeCols !== mergeCols) return false;
		totalMergeArea += currentMergeRows * currentMergeCols;
	}
	return totalMergeArea === totalArea;
}

//#endregion
//#region src/commands/commands/sheets-sort.command.ts
const SortRangeCommand = {
	id: "sheet.command.sort-range",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		const { range, orderRules, hasTitle, unitId, subUnitId } = params;
		const sortService = accessor.get(SheetsSortService);
		const { worksheet } = (0, _univerjs_sheets.getSheetCommandTarget)(accessor.get(_univerjs_core.IUniverInstanceService), params) || {};
		if (!worksheet) return false;
		const mergeDataInRange = worksheet.getMergeData().filter((mergeData) => {
			return _univerjs_core.Rectangle.contains(range, mergeData);
		});
		const mergeMainRowIndexes = mergeDataInRange.map((mergeData) => {
			return mergeData.startRow;
		});
		const { startRow: rangeStartRow, endRow } = range;
		const startRow = hasTitle ? rangeStartRow + 1 : rangeStartRow;
		const toReorder = [];
		const oldOrder = [];
		for (let rowIndex = startRow; rowIndex <= endRow; rowIndex++) {
			if (worksheet.getRowFiltered(rowIndex)) continue;
			if (worksheet.getRowRawVisible(rowIndex) === false) continue;
			if (mergeDataInRange.length && !mergeMainRowIndexes.includes(rowIndex)) continue;
			toReorder.push({
				index: rowIndex,
				value: getRowCellData(worksheet, rowIndex, orderRules)
			});
			oldOrder.push(rowIndex);
		}
		const compareFns = sortService.getAllCompareFns();
		toReorder.sort(reorderFnGenerator(orderRules, combineCompareFnsAsOne(compareFns)));
		const order = {};
		toReorder.forEach(({ index, value }, oldIndex) => {
			order[oldOrder[oldIndex]] = index;
		});
		const reorderRangeCommand = {
			id: _univerjs_sheets.ReorderRangeCommand.id,
			params: {
				unitId,
				subUnitId,
				range,
				order
			}
		};
		const commandService = accessor.get(_univerjs_core.ICommandService);
		return (0, _univerjs_core.sequenceExecute)([reorderRangeCommand], commandService).result;
	}
};
function getRowCellData(worksheet, rowIndex, orderRules) {
	const result = [];
	orderRules.forEach(({ colIndex }) => {
		result.push(worksheet.getCellRaw(rowIndex, colIndex));
	});
	return result;
}
function combineCompareFnsAsOne(compareFns) {
	return (type, a, b) => {
		for (let i = 0; i < compareFns.length; i++) {
			const res = compareFns[i](type, a, b);
			if (res != null) return res;
		}
		return 0;
	};
}
function reorderFnGenerator(orderRules, valueCompare) {
	return function(a, b) {
		let ret = null;
		for (let index = 0; index < orderRules.length; index++) {
			const aCellData = a.value[index];
			const bCellData = b.value[index];
			ret = valueCompare(orderRules[index].type, aCellData, bCellData);
			if (ret !== 0 && ret !== null && ret !== void 0) return ret;
		}
		return 0;
	};
}

//#endregion
//#region package.json
var name = "@univerjs/sheets-sort";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const SHEETS_SORT_PLUGIN_CONFIG_KEY = "sheets-sort.config";
const configSymbol = Symbol(SHEETS_SORT_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/controllers/sheets-sort.controller.ts
let SheetsSortController = class SheetsSortController extends _univerjs_core.Disposable {
	constructor(_commandService, _sortService) {
		super();
		this._commandService = _commandService;
		this._sortService = _sortService;
		this._initCommands();
		this._registerCompareFns();
	}
	_initCommands() {
		[SortRangeCommand].forEach((command) => this.disposeWithMe(this._commandService.registerCommand(command)));
	}
	_registerCompareFns() {
		const commonFn = (type, a, b) => {
			const valueA = this._getCommonValue(a);
			const valueB = this._getCommonValue(b);
			const compareTypeFns = [
				compareNull,
				compareString,
				compareNumber
			];
			for (let i = 0; i < compareTypeFns.length; i++) {
				const res = compareTypeFns[i](valueA, valueB, type);
				if (res !== null) return res;
			}
			return null;
		};
		this._sortService.registerCompareFn(commonFn);
	}
	_getCommonValue(a) {
		var _a$p;
		if (isNullValue(a)) return null;
		const richTextValue = a === null || a === void 0 || (_a$p = a.p) === null || _a$p === void 0 || (_a$p = _a$p.body) === null || _a$p === void 0 ? void 0 : _a$p.dataStream;
		if (richTextValue) return richTextValue;
		if ((a === null || a === void 0 ? void 0 : a.t) === _univerjs_core.CellValueType.NUMBER) return Number.parseFloat(`${a.v}`);
		if ((a === null || a === void 0 ? void 0 : a.t) === _univerjs_core.CellValueType.STRING) {
			if (typeof a.v === "number") return a.v;
			return `${a.v}`;
		}
		if ((a === null || a === void 0 ? void 0 : a.t) === _univerjs_core.CellValueType.BOOLEAN) return `${a.v}`;
		if ((a === null || a === void 0 ? void 0 : a.t) === _univerjs_core.CellValueType.FORCE_STRING) return Number.parseFloat(`${a.v}`);
		return `${a === null || a === void 0 ? void 0 : a.v}`;
	}
};
SheetsSortController = __decorate([__decorateParam(0, _univerjs_core.ICommandService), __decorateParam(1, (0, _univerjs_core.Inject)(SheetsSortService))], SheetsSortController);

//#endregion
//#region src/plugin.ts
let UniverSheetsSortPlugin = class UniverSheetsSortPlugin extends _univerjs_core.Plugin {
	constructor(_config = defaultPluginConfig, _injector, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		const { ...rest } = (0, _univerjs_core.merge)({}, defaultPluginConfig, this._config);
		this._configService.setConfig(SHEETS_SORT_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		[[SheetsSortController], [SheetsSortService]].forEach((d) => this._injector.add(d));
	}
	onReady() {
		this._injector.get(SheetsSortController);
	}
};
_defineProperty(UniverSheetsSortPlugin, "type", _univerjs_core.UniverInstanceType.UNIVER_SHEET);
_defineProperty(UniverSheetsSortPlugin, "pluginName", "SHEET_SORT_PLUGIN");
_defineProperty(UniverSheetsSortPlugin, "packageName", name);
_defineProperty(UniverSheetsSortPlugin, "version", version);
UniverSheetsSortPlugin = __decorate([
	(0, _univerjs_core.DependentOn)(_univerjs_sheets.UniverSheetsPlugin, _univerjs_engine_formula.UniverFormulaEnginePlugin),
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector)),
	__decorateParam(2, _univerjs_core.IConfigService)
], UniverSheetsSortPlugin);

//#endregion
Object.defineProperty(exports, 'SheetsSortService', {
  enumerable: true,
  get: function () {
    return SheetsSortService;
  }
});
exports.SortRangeCommand = SortRangeCommand;
exports.SortType = SortType;
Object.defineProperty(exports, 'UniverSheetsSortPlugin', {
  enumerable: true,
  get: function () {
    return UniverSheetsSortPlugin;
  }
});
