// @univerjs/sheets-sort/lib/facade.js
import { ICommandService, RANGE_TYPE } from "@univerjs/core";
import { FEventName, FUniver } from "@univerjs/core/facade";
import { SortRangeCommand, SortType } from "@univerjs/sheets-sort";
import { FRange, FWorksheet } from "@univerjs/sheets/facade";

//#region src/facade/f-univer.ts
var FUniverSheetsSortMixin = class extends FUniver {
	/**
	* @ignore
	*/
	_initialize(injector) {
		const commandService = injector.get(ICommandService);
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetBeforeRangeSort, () => commandService.beforeCommandExecuted((commandInfo) => {
			if (commandInfo.id !== SortRangeCommand.id) return;
			this._beforeRangeSort(commandInfo);
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetRangeSorted, () => commandService.onCommandExecuted((commandInfo) => {
			if (commandInfo.id !== SortRangeCommand.id) return;
			this._onRangeSorted(commandInfo);
		})));
	}
	_beforeRangeSort(commandInfo) {
		const params = commandInfo.params;
		const fWorkbook = this.getUniverSheet(params.unitId);
		const fWorksheet = fWorkbook.getSheetBySheetId(params.subUnitId);
		const { startColumn, endColumn, startRow, endRow } = params.range;
		const eventParams = {
			workbook: fWorkbook,
			worksheet: fWorksheet,
			range: fWorksheet.getRange(startRow, startColumn, endRow - startRow + 1, endColumn - startColumn + 1),
			sortColumn: params.orderRules.map((rule) => ({
				column: rule.colIndex - startColumn,
				ascending: rule.type === SortType.ASC
			}))
		};
		this.fireEvent(this.Event.SheetBeforeRangeSort, eventParams);
		if (eventParams.cancel) throw new Error("SortRangeCommand canceled.");
	}
	_onRangeSorted(commandInfo) {
		const params = commandInfo.params;
		const fWorkbook = this.getUniverSheet(params.unitId);
		const fWorksheet = fWorkbook.getSheetBySheetId(params.subUnitId);
		const { startColumn, endColumn, startRow, endRow } = params.range;
		const eventParams = {
			workbook: fWorkbook,
			worksheet: fWorksheet,
			range: fWorksheet.getRange(startRow, startColumn, endRow - startRow + 1, endColumn - startColumn + 1),
			sortColumn: params.orderRules.map((rule) => ({
				column: rule.colIndex - startColumn,
				ascending: rule.type === SortType.ASC
			}))
		};
		this.fireEvent(this.Event.SheetRangeSorted, eventParams);
		if (eventParams.cancel) throw new Error("SortRangeCommand canceled.");
	}
};
FUniver.extend(FUniverSheetsSortMixin);

//#endregion
//#region src/facade/f-range.ts
var FRangeSheetsSortMixin = class extends FRange {
	sort(column) {
		const columnBase = this._range.startColumn;
		const orderRules = (Array.isArray(column) ? column : [column]).map((c) => {
			if (typeof c === "number") return {
				colIndex: c + columnBase,
				type: SortType.ASC
			};
			return {
				colIndex: c.column + columnBase,
				type: c.ascending ? SortType.ASC : SortType.DESC
			};
		});
		this._commandService.syncExecuteCommand(SortRangeCommand.id, {
			orderRules,
			range: this._range,
			hasTitle: false,
			unitId: this._workbook.getUnitId(),
			subUnitId: this._worksheet.getSheetId()
		});
		return this;
	}
};
FRange.extend(FRangeSheetsSortMixin);

//#endregion
//#region src/facade/f-worksheet.ts
var FWorksheetSortMixin = class extends FWorksheet {
	sort(colIndex, asc = true) {
		const orderRules = [{
			colIndex,
			type: asc ? SortType.ASC : SortType.DESC
		}];
		const range = {
			startRow: 0,
			startColumn: 0,
			endRow: this._worksheet.getRowCount() - 1,
			endColumn: this._worksheet.getColumnCount() - 1,
			rangeType: RANGE_TYPE.ALL
		};
		this._commandService.syncExecuteCommand(SortRangeCommand.id, {
			orderRules,
			range,
			hasTitle: false,
			unitId: this._workbook.getUnitId(),
			subUnitId: this._worksheet.getSheetId()
		});
		return this;
	}
};
FWorksheet.extend(FWorksheetSortMixin);

//#endregion
//#region src/facade/f-event.ts
var FSheetsSortEventNameMixin = class {
	get SheetRangeSorted() {
		return "SheetRangeSorted";
	}
	get SheetBeforeRangeSort() {
		return "SheetBeforeRangeSort";
	}
};
FEventName.extend(FSheetsSortEventNameMixin);

//#endregion
export {  };
