// @univerjs/sheets-data-validation/lib/cjs/facade.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
let _univerjs_core = require("@univerjs/core");
let _univerjs_sheets_data_validation = require("@univerjs/sheets-data-validation");
let _univerjs_sheets_facade = require("@univerjs/sheets/facade");
let _univerjs_data_validation = require("@univerjs/data-validation");
let _univerjs_engine_formula = require("@univerjs/engine-formula");
let _univerjs_sheets = require("@univerjs/sheets");
let _univerjs_core_facade = require("@univerjs/core/facade");

//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region src/facade/f-data-validation-builder.ts
/**
* Builder for data validation rules. use {@link FUniver} `univerAPI.newDataValidation()` to create a new builder.
* @example
* ```typescript
* // Set the data validation for cell A1 to require a value from B1:B10
* const fWorkbook = univerAPI.getActiveWorkbook();
* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
* if (!fWorksheet) return;
* const fRange = fWorksheet.getRange('B1:B2');
* fRange.setValues([
*   ['Yes'],
*   ['No']
* ]);
*
* const rule = univerAPI.newDataValidation()
*   .requireValueInRange(fRange)
*   .setOptions({
*     allowBlank: false,
*     showErrorMessage: true,
*     error: 'Please enter a value from the list'
*   })
*   .build();
* const cell = fWorksheet.getRange('A1');
* cell.setDataValidation(rule);
* ```
* @hideconstructor
*/
var FDataValidationBuilder = class FDataValidationBuilder {
	constructor(rule) {
		_defineProperty(this, "_rule", void 0);
		this._rule = rule !== null && rule !== void 0 ? rule : {
			uid: (0, _univerjs_core.generateRandomId)(),
			ranges: void 0,
			type: _univerjs_core.DataValidationType.CUSTOM
		};
	}
	/**
	* Builds an FDataValidation instance based on the _rule property of the current class
	* @returns {FDataValidation} A new instance of the FDataValidation class
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new data validation rule that requires a number between 1 and 10 for the range A1:B10
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireNumberBetween(1, 10)
	*   .setOptions({
	*     allowBlank: true,
	*     showErrorMessage: true,
	*     error: 'Please enter a number between 1 and 10'
	*   })
	*   .build();
	* fRange.setDataValidation(rule);
	* ```
	*/
	build() {
		return new FDataValidation(this._rule);
	}
	/**
	* Creates a duplicate of the current DataValidationBuilder object
	* @returns {FDataValidationBuilder} A new instance of the DataValidationBuilder class
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new data validation rule that requires a number between 1 and 10 for the range A1:B10
	* const fRange = fWorksheet.getRange('A1:B10');
	* const builder = univerAPI.newDataValidation()
	*   .requireNumberBetween(1, 10)
	*   .setOptions({
	*     allowBlank: true,
	*     showErrorMessage: true,
	*     error: 'Please enter a number between 1 and 10'
	*   });
	* fRange.setDataValidation(builder.build());
	*
	* // Copy the builder applied to the new range F1:G10
	* const newRange = fWorksheet.getRange('F1:G10');
	* const copyBuilder = builder.copy();
	* newRange.setDataValidation(copyBuilder.build());
	* ```
	*/
	copy() {
		return new FDataValidationBuilder({
			...this._rule,
			uid: (0, _univerjs_core.generateRandomId)()
		});
	}
	/**
	* Determines whether invalid data is allowed
	* @returns {boolean} True if invalid data is allowed, False otherwise
	* @example
	* ```typescript
	* const builder = univerAPI.newDataValidation().requireNumberBetween(1, 10);
	* console.log(builder.getAllowInvalid());
	* ```
	*/
	getAllowInvalid() {
		return this._rule.errorStyle !== _univerjs_core.DataValidationErrorStyle.STOP;
	}
	/**
	* Gets the data validation type of the rule
	* @returns {DataValidationType | string} The data validation type
	* @example
	* ```typescript
	* const builder = univerAPI.newDataValidation();
	* console.log(builder.getCriteriaType()); // custom
	*
	* builder.requireNumberBetween(1, 10);
	* console.log(builder.getCriteriaType()); // decimal
	*
	* builder.requireValueInList(['Yes', 'No']);
	* console.log(builder.getCriteriaType()); // list
	* ```
	*/
	getCriteriaType() {
		return this._rule.type;
	}
	/**
	* Gets the values used for criteria evaluation
	* @returns {[string | undefined, string | undefined, string | undefined]} An array containing the operator, formula1, and formula2 values
	* @example
	* ```typescript
	* const builder = univerAPI.newDataValidation().requireNumberBetween(1, 10);
	* const [operator, formula1, formula2] = builder.getCriteriaValues();
	* console.log(operator, formula1, formula2); // between 1 10
	*
	* builder.requireValueInList(['Yes', 'No']);
	* console.log(builder.getCriteriaValues()); // undefined Yes,No undefined
	* ```
	*/
	getCriteriaValues() {
		return [
			this._rule.operator,
			this._rule.formula1,
			this._rule.formula2
		];
	}
	/**
	* Gets the help text information, which is used to provide users with guidance and support
	* @returns {string | undefined} Returns the help text information. If there is no error message, it returns an undefined value
	* @example
	* ```typescript
	* const builder = univerAPI.newDataValidation().setOptions({
	*   showErrorMessage: true,
	*   error: 'Please enter a valid value'
	* });
	* console.log(builder.getHelpText()); // 'Please enter a valid value'
	* ```
	*/
	getHelpText() {
		return this._rule.error;
	}
	/**
	* Sets the data validation rule to require that the input is a boolean value; this value is rendered as a checkbox.
	* @param {string} [checkedValue] - The value assigned to a checked box.
	* @param {string} [uncheckedValue] - The value assigned to an unchecked box.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set the data validation for cell A1:A10 to require a checkbox with default 1 and 0 values
	* const fRange = fWorksheet.getRange('A1:A10');
	* const rule = univerAPI.newDataValidation()
	*   .requireCheckbox()
	*   .build();
	* fRange.setDataValidation(rule);
	*
	* // Set the data validation for cell B1:B10 to require a checkbox with 'Yes' and 'No' values
	* const fRange2 = fWorksheet.getRange('B1:B10');
	* const rule2 = univerAPI.newDataValidation()
	*   .requireCheckbox('Yes', 'No')
	*   .build();
	* fRange2.setDataValidation(rule2);
	* ```
	*/
	requireCheckbox(checkedValue, uncheckedValue) {
		this._rule.type = _univerjs_core.DataValidationType.CHECKBOX;
		this._rule.formula1 = checkedValue;
		this._rule.formula2 = uncheckedValue;
		return this;
	}
	/**
	* Set the data validation type to DATE and configure the validation rules to be after a specific date.
	* @param {Date} date - The latest unacceptable date.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set some date values in the range A1:B2
	* const fRange = fWorksheet.getRange('A1:B2');
	* fRange.setValues([
	*   ['2024-01-01', '2024-12-31'],
	*   ['2025-01-01', '2025-12-31']
	* ]);
	*
	* // Create a data validation rule that requires a date after 2025-01-01
	* const rule = univerAPI.newDataValidation()
	*   .requireDateAfter(new Date('2025-01-01'))
	*   .build();
	* fRange.setDataValidation(rule);
	*
	* // Get the validation status of the range
	* const status = await fRange.getValidatorStatus();
	* console.log(status); // [['invalid', 'invalid', 'invalid', 'valid']]
	* ```
	*/
	requireDateAfter(date) {
		this._rule.type = _univerjs_core.DataValidationType.DATE;
		this._rule.formula1 = date.toLocaleDateString();
		this._rule.operator = _univerjs_core.DataValidationOperator.GREATER_THAN;
		return this;
	}
	/**
	* Set the data validation type to DATE and configure the validation rules to be before a specific date.
	* @param {Date} date - The earliest unacceptable date.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set some date values in the range A1:B2
	* const fRange = fWorksheet.getRange('A1:B2');
	* fRange.setValues([
	*   ['2024-01-01', '2024-12-31'],
	*   ['2025-01-01', '2025-12-31']
	* ]);
	*
	* // Create a data validation rule that requires a date before 2025-01-01
	* const rule = univerAPI.newDataValidation()
	*   .requireDateBefore(new Date('2025-01-01'))
	*   .build();
	* fRange.setDataValidation(rule);
	*
	* // Get the validation status of the range
	* const status = await fRange.getValidatorStatus();
	* console.log(status); // [['valid', 'valid', 'invalid', 'invalid']]
	* ```
	*/
	requireDateBefore(date) {
		this._rule.type = _univerjs_core.DataValidationType.DATE;
		this._rule.formula1 = date.toLocaleDateString();
		this._rule.formula2 = void 0;
		this._rule.operator = _univerjs_core.DataValidationOperator.LESS_THAN;
		return this;
	}
	/**
	* Set the data validation type to DATE and configure the validation rules to be within a specific date range.
	* @param {Date} start - The earliest acceptable date.
	* @param {Date} end - The latest acceptable date.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set some date values in the range A1:B2
	* const fRange = fWorksheet.getRange('A1:B2');
	* fRange.setValues([
	*   ['2024-01-01', '2024-12-31'],
	*   ['2025-01-01', '2025-12-31']
	* ]);
	*
	* // Create a data validation rule that requires a date between 2024-06-01 and 2025-06-01
	* const rule = univerAPI.newDataValidation()
	*   .requireDateBetween(new Date('2024-06-01'), new Date('2025-06-01'))
	*   .build();
	* fRange.setDataValidation(rule);
	*
	* // Get the validation status of the range
	* const status = await fRange.getValidatorStatus();
	* console.log(status); // [['invalid', 'valid', 'valid', 'invalid']]
	* ```
	*/
	requireDateBetween(start, end) {
		this._rule.type = _univerjs_core.DataValidationType.DATE;
		this._rule.formula1 = start.toLocaleDateString();
		this._rule.formula2 = end.toLocaleDateString();
		this._rule.operator = _univerjs_core.DataValidationOperator.BETWEEN;
		return this;
	}
	/**
	* Set the data validation type to DATE and configure the validation rules to be equal to a specific date.
	* @param {Date} date - The sole acceptable date.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set some date values in the range A1:B2
	* const fRange = fWorksheet.getRange('A1:B2');
	* fRange.setValues([
	*   ['2024-01-01', '2024-12-31'],
	*   ['2025-01-01', '2025-12-31']
	* ]);
	*
	* // Create a data validation rule that requires a date equal to 2025-01-01
	* const rule = univerAPI.newDataValidation()
	*   .requireDateEqualTo(new Date('2025-01-01'))
	*   .build();
	* fRange.setDataValidation(rule);
	*
	* // Get the validation status of the cell A2
	* const status = await fWorksheet.getRange('A2').getValidatorStatus();
	* console.log(status?.[0]?.[0]); // 'valid'
	*
	* // Get the validation status of the cell B2
	* const status2 = await fWorksheet.getRange('B2').getValidatorStatus();
	* console.log(status2?.[0]?.[0]); // 'invalid'
	* ```
	*/
	requireDateEqualTo(date) {
		this._rule.type = _univerjs_core.DataValidationType.DATE;
		this._rule.formula1 = date.toLocaleDateString();
		this._rule.formula2 = void 0;
		this._rule.operator = _univerjs_core.DataValidationOperator.EQUAL;
		return this;
	}
	/**
	* Set the data validation type to DATE and configure the validation rules to be not within a specific date range.
	* @param {Date} start - The earliest unacceptable date.
	* @param {Date} end - The latest unacceptable date.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set some date values in the range A1:B2
	* const fRange = fWorksheet.getRange('A1:B2');
	* fRange.setValues([
	*   ['2024-01-01', '2024-12-31'],
	*   ['2025-01-01', '2025-12-31']
	* ]);
	*
	* // Create a data validation rule that requires a date not between 2024-06-01 and 2025-06-01
	* const rule = univerAPI.newDataValidation()
	*   .requireDateNotBetween(new Date('2024-06-01'), new Date('2025-06-01'))
	*   .build();
	* fRange.setDataValidation(rule);
	*
	* // Get the validation status of the range
	* const status = await fRange.getValidatorStatus();
	* console.log(status); // [['valid', 'invalid', 'invalid', 'valid']]
	* ```
	*/
	requireDateNotBetween(start, end) {
		this._rule.type = _univerjs_core.DataValidationType.DATE;
		this._rule.formula1 = start.toLocaleDateString();
		this._rule.formula2 = end.toLocaleDateString();
		this._rule.operator = _univerjs_core.DataValidationOperator.NOT_BETWEEN;
		return this;
	}
	/**
	* Set the data validation type to DATE and configure the validation rules to be on or after a specific date.
	* @param {Date} date - The earliest acceptable date.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set some date values in the range A1:B2
	* const fRange = fWorksheet.getRange('A1:B2');
	* fRange.setValues([
	*   ['2024-01-01', '2024-12-31'],
	*   ['2025-01-01', '2025-12-31']
	* ]);
	*
	* // Create a data validation rule that requires a date on or after 2025-01-01
	* const rule = univerAPI.newDataValidation()
	*   .requireDateOnOrAfter(new Date('2025-01-01'))
	*   .build();
	* fRange.setDataValidation(rule);
	*
	* // Get the validation status of the range
	* const status = await fRange.getValidatorStatus();
	* console.log(status); // [['invalid', 'invalid', 'valid', 'valid']]
	* ```
	*/
	requireDateOnOrAfter(date) {
		this._rule.type = _univerjs_core.DataValidationType.DATE;
		this._rule.formula1 = date.toLocaleDateString();
		this._rule.formula2 = void 0;
		this._rule.operator = _univerjs_core.DataValidationOperator.GREATER_THAN_OR_EQUAL;
		return this;
	}
	/**
	* Set the data validation type to DATE and configure the validation rules to be on or before a specific date.
	* @param {Date} date - The latest acceptable date.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set some date values in the range A1:B2
	* const fRange = fWorksheet.getRange('A1:B2');
	* fRange.setValues([
	*   ['2024-01-01', '2024-12-31'],
	*   ['2025-01-01', '2025-12-31']
	* ]);
	*
	* // Create a data validation rule that requires a date on or before 2025-01-01
	* const rule = univerAPI.newDataValidation()
	*   .requireDateOnOrBefore(new Date('2025-01-01'))
	*   .build();
	* fRange.setDataValidation(rule);
	*
	* // Get the validation status of the range
	* const status = await fRange.getValidatorStatus();
	* console.log(status); // [['valid', 'valid', 'valid', 'invalid']]
	* ```
	*/
	requireDateOnOrBefore(date) {
		this._rule.type = _univerjs_core.DataValidationType.DATE;
		this._rule.formula1 = date.toLocaleDateString();
		this._rule.formula2 = void 0;
		this._rule.operator = _univerjs_core.DataValidationOperator.LESS_THAN_OR_EQUAL;
		return this;
	}
	/**
	* Sets the data validation rule to require that the given formula evaluates to `true`.
	* @param {string} formula - The formula string that needs to be satisfied, formula result should be TRUE or FALSE, and references range will relative offset.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set some values in the range A1:B2 and C1:D2
	* const cell = fWorksheet.getRange('A1:B2');
	* cell.setValues([
	*   [4, 3],
	*   [2, 1]
	* ]);
	* const fRange = fWorksheet.getRange('C1:D2');
	* fRange.setValues([
	*   [1, 2],
	*   [3, 4]
	* ]);
	*
	* // Create a data validation rule that requires the formula '=A1>2' to be satisfied
	* const rule = univerAPI.newDataValidation()
	*   .requireFormulaSatisfied('=A1>2')
	*   .setOptions({
	*     showErrorMessage: true,
	*     error: 'Please enter a value equal to A1'
	*   })
	*   .build();
	* fRange.setDataValidation(rule);
	*
	* // Get the validation status of the range
	* const status = await fRange.getValidatorStatus();
	* console.log(status); // [['valid', 'valid', 'invalid', 'invalid']]
	* ```
	*/
	requireFormulaSatisfied(formula) {
		this._rule.type = _univerjs_core.DataValidationType.CUSTOM;
		this._rule.formula1 = formula;
		this._rule.formula2 = void 0;
		return this;
	}
	/**
	* Sets the data validation rule to require a number that falls between, or is either of, two specified numbers.
	* @param {number} start - The lowest acceptable value.
	* @param {number} end - The highest acceptable value.
	* @param {boolean} [isInteger] - Indicates whether the required number is an integer.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new data validation rule that requires a number between 1 and 10 for the range A1:B10
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireNumberBetween(1, 10)
	*   .setOptions({
	*     allowBlank: false,
	*     showErrorMessage: true,
	*     error: 'Please enter a number between 1 and 10'
	*   })
	*   .build();
	* fRange.setDataValidation(rule);
	* ```
	*/
	requireNumberBetween(start, end, isInteger) {
		this._rule.formula1 = `${start}`;
		this._rule.formula2 = `${end}`;
		this._rule.operator = _univerjs_core.DataValidationOperator.BETWEEN;
		this._rule.type = isInteger ? _univerjs_core.DataValidationType.WHOLE : _univerjs_core.DataValidationType.DECIMAL;
		return this;
	}
	/**
	* Sets the data validation rule to require a number equal to the given value.
	* @param {number} num - The sole acceptable value.
	* @param {boolean} [isInteger] - Indicates whether the required number is an integer.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new data validation rule that requires a number equal to 10 for the range A1:B10
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireNumberEqualTo(10)
	*   .setOptions({
	*     allowBlank: false,
	*     showErrorMessage: true,
	*     error: 'Please enter a number equal to 10'
	*   })
	*   .build();
	* fRange.setDataValidation(rule);
	* ```
	*/
	requireNumberEqualTo(num, isInteger) {
		this._rule.formula1 = `${num}`;
		this._rule.formula2 = void 0;
		this._rule.operator = _univerjs_core.DataValidationOperator.EQUAL;
		this._rule.type = isInteger ? _univerjs_core.DataValidationType.WHOLE : _univerjs_core.DataValidationType.DECIMAL;
		return this;
	}
	/**
	* Sets the data validation rule to require a number greater than the given value.
	* @param {number} num - The highest unacceptable value.
	* @param {boolean} [isInteger] - Indicates whether the required number is an integer.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new data validation rule that requires a number greater than 10 for the range A1:B10
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireNumberGreaterThan(10)
	*   .setOptions({
	*     allowBlank: false,
	*     showErrorMessage: true,
	*     error: 'Please enter a number greater than 10'
	*   })
	*   .build();
	* fRange.setDataValidation(rule);
	* ```
	*/
	requireNumberGreaterThan(num, isInteger) {
		this._rule.formula1 = `${num}`;
		this._rule.formula2 = void 0;
		this._rule.operator = _univerjs_core.DataValidationOperator.GREATER_THAN;
		this._rule.type = isInteger ? _univerjs_core.DataValidationType.WHOLE : _univerjs_core.DataValidationType.DECIMAL;
		return this;
	}
	/**
	* Sets the data validation rule to require a number greater than or equal to the given value.
	* @param {number} num - The lowest acceptable value.
	* @param {boolean} [isInteger] - Indicates whether the required number is an integer.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new data validation rule that requires a number greater than 10 or equal to 10 for the range A1:B10
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireNumberGreaterThanOrEqualTo(10)
	*   .setOptions({
	*     allowBlank: false,
	*     showErrorMessage: true,
	*     error: 'Please enter a number greater than 10 or equal to 10'
	*   })
	*   .build();
	* fRange.setDataValidation(rule);
	* ```
	*/
	requireNumberGreaterThanOrEqualTo(num, isInteger) {
		this._rule.formula1 = `${num}`;
		this._rule.formula2 = void 0;
		this._rule.operator = _univerjs_core.DataValidationOperator.GREATER_THAN_OR_EQUAL;
		this._rule.type = isInteger ? _univerjs_core.DataValidationType.WHOLE : _univerjs_core.DataValidationType.DECIMAL;
		return this;
	}
	/**
	* Sets the data validation rule to require a number less than the given value.
	* @param {number} num - The lowest unacceptable value.
	* @param {boolean} [isInteger] - Indicates whether the required number is an integer.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new data validation rule that requires a number less than 10 for the range A1:B10
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireNumberLessThan(10)
	*   .setOptions({
	*     allowBlank: false,
	*     showErrorMessage: true,
	*     error: 'Please enter a number less than 10'
	*   })
	*   .build();
	* fRange.setDataValidation(rule);
	* ```
	*/
	requireNumberLessThan(num, isInteger) {
		this._rule.formula1 = `${num}`;
		this._rule.formula2 = void 0;
		this._rule.operator = _univerjs_core.DataValidationOperator.LESS_THAN;
		this._rule.type = isInteger ? _univerjs_core.DataValidationType.WHOLE : _univerjs_core.DataValidationType.DECIMAL;
		return this;
	}
	/**
	* Sets the data validation rule to require a number less than or equal to the given value.
	* @param {number} num - The highest acceptable value.
	* @param {boolean} [isInteger] - Indicates whether the required number is an integer.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new data validation rule that requires a number less than 10 or equal to 10 for the range A1:B10
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireNumberLessThanOrEqualTo(10)
	*   .setOptions({
	*     allowBlank: false,
	*     showErrorMessage: true,
	*     error: 'Please enter a number less than 10 or equal to 10'
	*   })
	*   .build();
	* fRange.setDataValidation(rule);
	* ```
	*/
	requireNumberLessThanOrEqualTo(num, isInteger) {
		this._rule.formula1 = `${num}`;
		this._rule.formula2 = void 0;
		this._rule.operator = _univerjs_core.DataValidationOperator.LESS_THAN_OR_EQUAL;
		this._rule.type = isInteger ? _univerjs_core.DataValidationType.WHOLE : _univerjs_core.DataValidationType.DECIMAL;
		return this;
	}
	/**
	* Sets the data validation rule to require a number that does not fall between, and is neither of, two specified numbers.
	* @param {number} start - The lowest unacceptable value.
	* @param {number} end - The highest unacceptable value.
	* @param {boolean} [isInteger] - Optional parameter, indicating whether the number to be verified is an integer.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new data validation rule that requires a number not between 1 and 10 for the range A1:B10
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireNumberNotBetween(1, 10)
	*   .setOptions({
	*     allowBlank: false,
	*     showErrorMessage: true,
	*     error: 'Please enter a number not between 1 and 10'
	*   })
	*   .build();
	* fRange.setDataValidation(rule);
	* ```
	*/
	requireNumberNotBetween(start, end, isInteger) {
		this._rule.formula1 = `${start}`;
		this._rule.formula2 = `${end}`;
		this._rule.operator = _univerjs_core.DataValidationOperator.NOT_BETWEEN;
		this._rule.type = isInteger ? _univerjs_core.DataValidationType.WHOLE : _univerjs_core.DataValidationType.DECIMAL;
		return this;
	}
	/**
	* Sets the data validation rule to require a number not equal to the given value.
	* @param {number} num - The sole unacceptable value.
	* @param {boolean} [isInteger] - Indicates whether the required number is an integer.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new data validation rule that requires a number not equal to 10 for the range A1:B10
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireNumberNotEqualTo(10)
	*   .setOptions({
	*     allowBlank: false,
	*     showErrorMessage: true,
	*     error: 'Please enter a number not equal to 10'
	*   })
	*   .build();
	* fRange.setDataValidation(rule);
	* ```
	*/
	requireNumberNotEqualTo(num, isInteger) {
		this._rule.formula1 = `${num}`;
		this._rule.formula2 = void 0;
		this._rule.operator = _univerjs_core.DataValidationOperator.NOT_EQUAL;
		this._rule.type = isInteger ? _univerjs_core.DataValidationType.WHOLE : _univerjs_core.DataValidationType.DECIMAL;
		return this;
	}
	/**
	* Sets a data validation rule that requires the user to enter a value from a list of specific values.
	* The list can be displayed in a dropdown, and the user can choose multiple values according to the settings.
	* @param {string[]} values - An array of acceptable values.
	* @param {boolean} [multiple] - Optional parameter indicating whether the user can select multiple values.
	* @param {boolean} [showDropdown] - Optional parameter indicating whether to display the list in a dropdown.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new data validation rule that requires the user to enter a value from the list ['Yes', 'No'] for the range A1:B10
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireValueInList(['Yes', 'No'])
	*   .setOptions({
	*     allowBlank: true,
	*     showErrorMessage: true,
	*     error: 'Please enter a value from the list'
	*   })
	*   .build();
	* fRange.setDataValidation(rule);
	* ```
	*/
	requireValueInList(values, multiple, showDropdown) {
		this._rule.type = multiple ? _univerjs_core.DataValidationType.LIST_MULTIPLE : _univerjs_core.DataValidationType.LIST;
		this._rule.formula1 = (0, _univerjs_sheets.serializeListOptions)(values);
		this._rule.formula2 = void 0;
		this._rule.showDropDown = showDropdown !== null && showDropdown !== void 0 ? showDropdown : true;
		return this;
	}
	/**
	* Sets a data validation rule that requires the user to enter a value within a specific range.
	* The range is defined by an FRange object, which contains the unit ID, sheet name, and cell range.
	* @param {FRange} range - An FRange object representing the range of values that the user can enter.
	* @param {boolean} [multiple] - Optional parameter indicating whether the user can select multiple values.
	* @param {boolean} [showDropdown] - Optional parameter indicating whether to display the list in a dropdown.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set the values in the range B1:B2
	* const fRange = fWorksheet.getRange('B1:B2');
	* fRange.setValues([
	*   ['Yes'],
	*   ['No']
	* ]);
	*
	* // Create a new data validation rule that requires the user to enter a value from the range B1:B2 for the range A1:A10
	* const rule = univerAPI.newDataValidation()
	*   .requireValueInRange(fRange)
	*   .setOptions({
	*     allowBlank: false,
	*     showErrorMessage: true,
	*     error: 'Please enter a value from the list'
	*   })
	*   .build();
	* const cell = fWorksheet.getRange('A1');
	* cell.setDataValidation(rule);
	* ```
	*/
	requireValueInRange(range, multiple, showDropdown) {
		this._rule.type = multiple ? _univerjs_core.DataValidationType.LIST_MULTIPLE : _univerjs_core.DataValidationType.LIST;
		this._rule.formula1 = `=${(0, _univerjs_engine_formula.serializeRangeToRefString)({
			unitId: range.getUnitId(),
			sheetName: range.getSheetName(),
			range: range.getRange()
		})}`;
		this._rule.formula2 = void 0;
		this._rule.showDropDown = showDropdown !== null && showDropdown !== void 0 ? showDropdown : true;
		return this;
	}
	/**
	* Sets whether to allow invalid data and configures the error style.
	* If invalid data is not allowed, the error style will be set to STOP, indicating that data entry must stop upon encountering an error.
	* If invalid data is allowed, the error style will be set to WARNING, indicating that a warning will be displayed when invalid data is entered, but data entry can continue.
	* @param {boolean} allowInvalidData - Whether to allow invalid data.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set the data validation for cell A1:B2 to allow invalid data, so A1:B2 will display a warning when invalid data is entered
	* const fRange = fWorksheet.getRange('A1:B2');
	* const rule = univerAPI.newDataValidation()
	*   .requireValueInList(['Yes', 'No'])
	*   .setAllowInvalid(true)
	*   .build();
	* fRange.setDataValidation(rule);
	*
	* // Set the data validation for cell C1:D2 to not allow invalid data, so C1:D2 will stop data entry when invalid data is entered
	* const fRange2 = fWorksheet.getRange('C1:D2');
	* const rule2 = univerAPI.newDataValidation()
	*   .requireValueInList(['Yes', 'No'])
	*   .setAllowInvalid(false)
	*   .build();
	* fRange2.setDataValidation(rule2);
	* ```
	*/
	setAllowInvalid(allowInvalidData) {
		this._rule.errorStyle = !allowInvalidData ? _univerjs_core.DataValidationErrorStyle.STOP : _univerjs_core.DataValidationErrorStyle.WARNING;
		return this;
	}
	/**
	* Sets whether to allow blank values.
	* @param {boolean} allowBlank - Whether to allow blank values.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* // Assume current sheet is empty data
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set the data validation for cell A1:B2 to allow blank values
	* const fRange = fWorksheet.getRange('A1:B2');
	* const rule = univerAPI.newDataValidation()
	*   .requireValueInList(['Yes', 'No'])
	*   .setAllowBlank(true)
	*   .build();
	* fRange.setDataValidation(rule);
	*
	* // Set the data validation for cell C1:D2 to not allow blank values
	* const fRange2 = fWorksheet.getRange('C1:D2');
	* const rule2 = univerAPI.newDataValidation()
	*   .requireValueInList(['Yes', 'No'])
	*   .setAllowBlank(false)
	*   .build();
	* fRange2.setDataValidation(rule2);
	* ```
	*/
	setAllowBlank(allowBlank) {
		this._rule.allowBlank = allowBlank;
		return this;
	}
	/**
	* Sets the options for the data validation rule.
	* @param {Partial<IDataValidationRuleOptions>} options - The options to set for the data validation rule.
	* @returns {FDataValidationBuilder} The current instance for method chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new data validation rule that requires the user to enter a value from the list ['Yes', 'No'] for the range A1:B10
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireValueInList(['Yes', 'No'])
	*   .setOptions({
	*     allowBlank: true,
	*     showErrorMessage: true,
	*     error: 'Please enter a value from the list'
	*   })
	*   .build();
	* fRange.setDataValidation(rule);
	* ```
	*/
	setOptions(options) {
		Object.assign(this._rule, options);
		return this;
	}
};

//#endregion
//#region src/facade/f-data-validation.ts
/**
* @hideconstructor
*/
var FDataValidation = class {
	constructor(rule, worksheet, _injector) {
		_defineProperty(this, "rule", void 0);
		_defineProperty(this, "_worksheet", void 0);
		_defineProperty(this, "_injector", void 0);
		this._injector = _injector;
		this.rule = rule;
		this._worksheet = worksheet;
	}
	/**
	* Gets whether invalid data is allowed based on the error style value
	* @returns {boolean} true if invalid data is allowed, false otherwise
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const rules = fWorksheet.getDataValidations();
	* rules.forEach((rule) => {
	*   console.log(rule, rule.getAllowInvalid());
	* });
	* ```
	*/
	getAllowInvalid() {
		return this.rule.errorStyle !== _univerjs_core.DataValidationErrorStyle.STOP;
	}
	/**
	* Gets the data validation type of the rule
	* @returns {DataValidationType | string} The data validation type
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const rules = fWorksheet.getDataValidations();
	* rules.forEach((rule) => {
	*   console.log(rule, rule.getCriteriaType());
	* });
	* ```
	*/
	getCriteriaType() {
		return this.rule.type;
	}
	/**
	* Gets the values used for criteria evaluation
	* @returns {[string | undefined, string | undefined, string | undefined]} An array containing the operator, formula1, and formula2 values
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const rules = fWorksheet.getDataValidations();
	* rules.forEach((rule) => {
	*   console.log(rule);
	*   const criteriaValues = rule.getCriteriaValues();
	*   const [operator, formula1, formula2] = criteriaValues;
	*   console.log(operator, formula1, formula2);
	* });
	* ```
	*/
	getCriteriaValues() {
		return [
			this.rule.operator,
			this.rule.formula1,
			this.rule.formula2
		];
	}
	/**
	* Gets the help text information, which is used to provide users with guidance and support
	* @returns {string | undefined} Returns the help text information. If there is no error message, it returns an undefined value
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireNumberBetween(1, 10)
	*   .setOptions({
	*     allowBlank: true,
	*     showErrorMessage: true,
	*     error: 'Please enter a number between 1 and 10'
	*   })
	*   .build();
	* fRange.setDataValidation(rule);
	* console.log(fRange.getDataValidation().getHelpText()); // 'Please enter a number between 1 and 10'
	* ```
	*/
	getHelpText() {
		return this.rule.error;
	}
	/**
	* Creates a new instance of FDataValidationBuilder using the current rule object
	* @returns {FDataValidationBuilder} A new FDataValidationBuilder instance with the same rule configuration
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireNumberBetween(1, 10)
	*   .setOptions({
	*     allowBlank: true,
	*     showErrorMessage: true,
	*     error: 'Please enter a number between 1 and 10'
	*   })
	*   .build();
	* fRange.setDataValidation(rule);
	*
	* const builder = fRange.getDataValidation().copy();
	* const newRule = builder
	*   .requireNumberBetween(1, 5)
	*   .setOptions({
	*     error: 'Please enter a number between 1 and 5'
	*   })
	*   .build();
	* fRange.setDataValidation(newRule);
	* ```
	*/
	copy() {
		return new FDataValidationBuilder(this.rule);
	}
	/**
	* Gets whether the data validation rule is applied to the worksheet
	* @returns {boolean} true if the rule is applied, false otherwise
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const rules = fWorksheet.getDataValidations();
	* rules.forEach((rule) => {
	*   console.log(rule, rule.getApplied());
	* });
	*
	* const fRange = fWorksheet.getRange('A1:B10');
	* console.log(fRange.getDataValidation()?.getApplied());
	* ```
	*/
	getApplied() {
		if (!this._worksheet) return false;
		const currentRule = this._injector.get(_univerjs_data_validation.DataValidationModel).getRuleById(this._worksheet.getUnitId(), this._worksheet.getSheetId(), this.rule.uid);
		if (currentRule && currentRule.ranges.length) return true;
		return false;
	}
	/**
	* Gets the ranges to which the data validation rule is applied
	* @returns {FRange[]} An array of FRange objects representing the ranges to which the data validation rule is applied
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const rules = fWorksheet.getDataValidations();
	* rules.forEach((rule) => {
	*   console.log(rule);
	*   const ranges = rule.getRanges();
	*   ranges.forEach((range) => {
	*     console.log(range.getA1Notation());
	*   });
	* });
	* ```
	*/
	getRanges() {
		if (!this.getApplied()) return [];
		const workbook = this._injector.get(_univerjs_core.IUniverInstanceService).getUnit(this._worksheet.getUnitId());
		return this.rule.ranges.map((range) => this._injector.createInstance(_univerjs_sheets_facade.FRange, workbook, this._worksheet, range));
	}
	/**
	* Gets the unit ID of the worksheet
	* @returns {string | undefined} The unit ID of the worksheet
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const fRange = fWorksheet.getRange('A1:B10');
	* console.log(fRange.getDataValidation().getUnitId());
	* ```
	*/
	getUnitId() {
		var _this$_worksheet;
		return (_this$_worksheet = this._worksheet) === null || _this$_worksheet === void 0 ? void 0 : _this$_worksheet.getUnitId();
	}
	/**
	* Gets the sheet ID of the worksheet
	* @returns {string | undefined} The sheet ID of the worksheet
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const fRange = fWorksheet.getRange('A1:B10');
	* console.log(fRange.getDataValidation().getSheetId());
	* ```
	*/
	getSheetId() {
		var _this$_worksheet2;
		return (_this$_worksheet2 = this._worksheet) === null || _this$_worksheet2 === void 0 ? void 0 : _this$_worksheet2.getSheetId();
	}
	/**
	* Set Criteria for the data validation rule
	* @param {DataValidationType} type - The type of data validation criteria
	* @param {[DataValidationOperator, string, string]} values - An array containing the operator, formula1, and formula2 values
	* @param {boolean} [allowBlank] - Whether to allow blank values
	* @returns {FDataValidation} The current instance for method chaining
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new data validation rule that requires a number equal to 20 for the range A1:B10
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireNumberEqualTo(20)
	*   .build();
	* fRange.setDataValidation(rule);
	*
	* // Change the rule criteria to require a number between 1 and 10
	* fRange.getDataValidation().setCriteria(
	*   univerAPI.Enum.DataValidationType.DECIMAL,
	*   [univerAPI.Enum.DataValidationOperator.BETWEEN, '1', '10']
	* );
	* ```
	*/
	setCriteria(type, values, allowBlank = true) {
		if (this.getApplied()) {
			if (!this._injector.get(_univerjs_core.ICommandService).syncExecuteCommand(_univerjs_sheets_data_validation.UpdateSheetDataValidationSettingCommand.id, {
				unitId: this.getUnitId(),
				subUnitId: this.getSheetId(),
				ruleId: this.rule.uid,
				setting: {
					operator: values[0],
					formula1: values[1],
					formula2: values[2],
					type: this.rule.type,
					allowBlank
				}
			})) throw new Error("setCriteria failed");
		}
		this.rule.operator = values[0];
		this.rule.formula1 = values[1];
		this.rule.formula2 = values[2];
		this.rule.type = type;
		this.rule.allowBlank = allowBlank;
		return this;
	}
	/**
	* Set the options for the data validation rule
	* @param {Partial<IDataValidationRuleOptions>} options - The options to set for the data validation rule
	* @returns {FDataValidation} The current instance for method chaining
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new data validation rule that requires a number equal to 20 for the range A1:B10
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireNumberEqualTo(20)
	*   .build();
	* fRange.setDataValidation(rule);
	*
	* // Supplement the rule with additional options
	* fRange.getDataValidation().setOptions({
	*   allowBlank: true,
	*   showErrorMessage: true,
	*   error: 'Please enter a valid value'
	* });
	* ```
	*/
	setOptions(options) {
		if (this.getApplied()) {
			if (!this._injector.get(_univerjs_core.ICommandService).syncExecuteCommand(_univerjs_sheets_data_validation.UpdateSheetDataValidationOptionsCommand.id, {
				unitId: this.getUnitId(),
				subUnitId: this.getSheetId(),
				ruleId: this.rule.uid,
				options: {
					...(0, _univerjs_data_validation.getRuleOptions)(this.rule),
					...options
				}
			})) throw new Error("setOptions failed");
		}
		Object.assign(this.rule, options);
		return this;
	}
	/**
	* Set the ranges to the data validation rule
	* @param {FRange[]} ranges - New ranges array
	* @returns {FDataValidation} The current instance for method chaining
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new data validation rule that requires a number equal to 20 for the range A1:B10
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireNumberEqualTo(20)
	*   .build();
	* fRange.setDataValidation(rule);
	*
	* // Change the range to C1:D10
	* const newRuleRange = fWorksheet.getRange('C1:D10');
	* fRange.getDataValidation().setRanges([newRuleRange]);
	* ```
	*/
	setRanges(ranges) {
		if (this.getApplied()) {
			if (!this._injector.get(_univerjs_core.ICommandService).syncExecuteCommand(_univerjs_sheets_data_validation.UpdateSheetDataValidationRangeCommand.id, {
				unitId: this.getUnitId(),
				subUnitId: this.getSheetId(),
				ruleId: this.rule.uid,
				ranges: ranges.map((range) => range.getRange())
			})) throw new Error("setRanges failed");
		}
		this.rule.ranges = ranges.map((range) => range.getRange());
		return this;
	}
	/**
	* Delete the data validation rule from the worksheet
	* @returns {boolean} true if the rule is deleted successfully, false otherwise
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new data validation rule that requires a number equal to 20 for the range A1:B10
	* const fRange = fWorksheet.getRange('A1:B10');
	* const rule = univerAPI.newDataValidation()
	*   .requireNumberEqualTo(20)
	*   .build();
	* fRange.setDataValidation(rule);
	*
	* // Delete the data validation rule
	* fRange.getDataValidation().delete();
	* ```
	*/
	delete() {
		if (!this.getApplied()) return false;
		return this._injector.get(_univerjs_core.ICommandService).syncExecuteCommand(_univerjs_sheets_data_validation.RemoveSheetDataValidationCommand.id, {
			unitId: this.getUnitId(),
			subUnitId: this.getSheetId(),
			ruleId: this.rule.uid
		});
	}
};

//#endregion
//#region src/facade/f-range.ts
/**
* @ignore
*/
var FRangeSheetsDataValidationMixin = class extends _univerjs_sheets_facade.FRange {
	setDataValidation(rule) {
		if (!rule) {
			this._commandService.syncExecuteCommand(_univerjs_sheets_data_validation.ClearRangeDataValidationCommand.id, {
				unitId: this._workbook.getUnitId(),
				subUnitId: this._worksheet.getSheetId(),
				ranges: [this._range]
			});
			return this;
		}
		const params = {
			unitId: this._workbook.getUnitId(),
			subUnitId: this._worksheet.getSheetId(),
			rule: {
				...rule.rule,
				ranges: [this._range]
			}
		};
		this._commandService.syncExecuteCommand(_univerjs_sheets_data_validation.AddSheetDataValidationCommand.id, params);
		return this;
	}
	getDataValidation() {
		const rule = this._injector.get(_univerjs_sheets_data_validation.SheetsDataValidationValidatorService).getDataValidation(this._workbook.getUnitId(), this._worksheet.getSheetId(), [this._range]);
		if (rule) return new FDataValidation(rule, this._worksheet, this._injector);
		return rule;
	}
	getDataValidations() {
		return this._injector.get(_univerjs_sheets_data_validation.SheetsDataValidationValidatorService).getDataValidations(this._workbook.getUnitId(), this._worksheet.getSheetId(), [this._range]).map((rule) => new FDataValidation(rule, this._worksheet, this._injector));
	}
	async getValidatorStatus() {
		return this._injector.get(_univerjs_sheets_data_validation.SheetsDataValidationValidatorService).validatorRanges(this._workbook.getUnitId(), this._worksheet.getSheetId(), [this._range]);
	}
	async getDataValidationErrorAsync() {
		const unitId = this._workbook.getUnitId();
		const sheetId = this._worksheet.getSheetId();
		return this._collectValidationErrorsForRange(unitId, sheetId, [this._range]);
	}
	async _collectValidationErrorsForRange(unitId, sheetId, ranges) {
		if (!ranges.length) return [];
		const validatorService = this._injector.get(_univerjs_sheets_data_validation.SheetsDataValidationValidatorService);
		const worksheet = this._worksheet;
		const sheetName = worksheet.getName();
		const errors = [];
		for (const range of ranges) {
			const promises = [];
			for (let row = range.startRow; row <= range.endRow; row++) for (let col = range.startColumn; col <= range.endColumn; col++) promises.push((async () => {
				try {
					if (await validatorService.validatorCell(unitId, sheetId, row, col) !== _univerjs_core.DataValidationStatus.VALID) {
						const rule = this._injector.get(_univerjs_sheets_data_validation.SheetDataValidationModel).getRuleByLocation(unitId, sheetId, row, col);
						if (rule) {
							var _worksheet$getCell;
							const cellValue = ((_worksheet$getCell = worksheet.getCell(row, col)) === null || _worksheet$getCell === void 0 ? void 0 : _worksheet$getCell.v) || null;
							const error = this._createDataValidationError(sheetName, row, col, rule, cellValue);
							errors.push(error);
						}
					}
				} catch (e) {
					console.warn(`Failed to validate cell [${row}, ${col}]:`, e);
				}
			})());
			await Promise.all(promises);
		}
		return errors;
	}
	_createDataValidationError(sheetName, row, column, rule, inputValue) {
		return {
			sheetName,
			row,
			column,
			ruleId: rule.uid,
			inputValue,
			rule
		};
	}
};
_univerjs_sheets_facade.FRange.extend(FRangeSheetsDataValidationMixin);

//#endregion
//#region src/facade/f-univer.ts
var FUniverSheetsDataValidationMixin = class extends _univerjs_core_facade.FUniver {
	/**
	* @deprecated use `univerAPI.newDataValidation()` as instead.
	* @returns {FDataValidationBuilder} A new instance of the FDataValidationBuilder class
	*/
	static newDataValidation() {
		return new FDataValidationBuilder();
	}
	newDataValidation() {
		return new FDataValidationBuilder();
	}
	/**
	* @ignore
	*/
	_initialize(injector) {
		const commandService = injector.get(_univerjs_core.ICommandService);
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetDataValidationChanged, () => {
			if (!injector.has(_univerjs_sheets_data_validation.SheetDataValidationModel)) return { dispose: () => {} };
			return injector.get(_univerjs_sheets_data_validation.SheetDataValidationModel).ruleChange$.subscribe((ruleChange) => {
				const { unitId, subUnitId, rule, oldRule, type } = ruleChange;
				const target = this.getSheetTarget(unitId, subUnitId);
				if (!target) return;
				const { workbook, worksheet } = target;
				const eventParams = {
					origin: ruleChange,
					worksheet,
					workbook,
					changeType: type,
					oldRule,
					rule: new FDataValidation(rule, worksheet.getSheet(), this._injector)
				};
				this.fireEvent(this.Event.SheetDataValidationChanged, eventParams);
			});
		}));
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetDataValidatorStatusChanged, () => {
			if (!injector.has(_univerjs_sheets_data_validation.SheetDataValidationModel)) return { dispose: () => {} };
			return injector.get(_univerjs_sheets_data_validation.SheetDataValidationModel).validStatusChange$.subscribe((statusChange) => {
				const { unitId, subUnitId, ruleId, status, row, col } = statusChange;
				const target = this.getSheetTarget(unitId, subUnitId);
				if (!target) return;
				const { workbook, worksheet } = target;
				const rule = worksheet.getDataValidation(ruleId);
				if (!rule) return;
				const eventParams = {
					workbook,
					worksheet,
					row,
					column: col,
					rule,
					status
				};
				this.fireEvent(this.Event.SheetDataValidatorStatusChanged, eventParams);
			});
		}));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetDataValidationAdd, () => commandService.beforeCommandExecuted((commandInfo) => {
			if (commandInfo.id === _univerjs_sheets_data_validation.AddSheetDataValidationCommand.id) {
				const params = commandInfo.params;
				const target = this.getSheetTarget(params.unitId, params.subUnitId);
				if (!target) return;
				const { workbook, worksheet } = target;
				const eventParams = {
					worksheet,
					workbook,
					rule: params.rule
				};
				this.fireEvent(this.Event.BeforeSheetDataValidationAdd, eventParams);
				if (eventParams.cancel) throw new _univerjs_core.CanceledError();
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetDataValidationCriteriaUpdate, () => commandService.beforeCommandExecuted((commandInfo) => {
			if (commandInfo.id === _univerjs_sheets_data_validation.UpdateSheetDataValidationSettingCommand.id) {
				const params = commandInfo.params;
				const target = this.getSheetTarget(params.unitId, params.subUnitId);
				if (!target) return;
				const { workbook, worksheet } = target;
				const rule = worksheet.getDataValidation(params.ruleId);
				if (!rule) return;
				const eventParams = {
					worksheet,
					workbook,
					rule,
					ruleId: params.ruleId,
					newCriteria: params.setting
				};
				this.fireEvent(this.Event.BeforeSheetDataValidationCriteriaUpdate, eventParams);
				if (eventParams.cancel) throw new _univerjs_core.CanceledError();
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetDataValidationRangeUpdate, () => commandService.beforeCommandExecuted((commandInfo) => {
			if (commandInfo.id === _univerjs_sheets_data_validation.UpdateSheetDataValidationRangeCommand.id) {
				const params = commandInfo.params;
				const target = this.getSheetTarget(params.unitId, params.subUnitId);
				if (!target) return;
				const { workbook, worksheet } = target;
				const rule = worksheet.getDataValidation(params.ruleId);
				if (!rule) return;
				const eventParams = {
					worksheet,
					workbook,
					rule,
					ruleId: params.ruleId,
					newRanges: params.ranges
				};
				this.fireEvent(this.Event.BeforeSheetDataValidationRangeUpdate, eventParams);
				if (eventParams.cancel) throw new _univerjs_core.CanceledError();
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetDataValidationOptionsUpdate, () => commandService.beforeCommandExecuted((commandInfo) => {
			if (commandInfo.id === _univerjs_sheets_data_validation.UpdateSheetDataValidationOptionsCommand.id) {
				const params = commandInfo.params;
				const target = this.getSheetTarget(params.unitId, params.subUnitId);
				if (!target) return;
				const { workbook, worksheet } = target;
				const rule = worksheet.getDataValidation(params.ruleId);
				if (!rule) return;
				const eventParams = {
					worksheet,
					workbook,
					rule,
					ruleId: params.ruleId,
					newOptions: params.options
				};
				this.fireEvent(this.Event.BeforeSheetDataValidationOptionsUpdate, eventParams);
				if (eventParams.cancel) throw new _univerjs_core.CanceledError();
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetDataValidationDelete, () => commandService.beforeCommandExecuted((commandInfo) => {
			if (commandInfo.id === _univerjs_sheets_data_validation.RemoveSheetDataValidationCommand.id) {
				const params = commandInfo.params;
				const target = this.getSheetTarget(params.unitId, params.subUnitId);
				if (!target) return;
				const { workbook, worksheet } = target;
				const rule = worksheet.getDataValidation(params.ruleId);
				if (!rule) return;
				const eventParams = {
					worksheet,
					workbook,
					rule,
					ruleId: params.ruleId
				};
				this.fireEvent(this.Event.BeforeSheetDataValidationDelete, eventParams);
				if (eventParams.cancel) throw new _univerjs_core.CanceledError();
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetDataValidationDeleteAll, () => commandService.beforeCommandExecuted((commandInfo) => {
			if (commandInfo.id === _univerjs_sheets_data_validation.RemoveSheetAllDataValidationCommand.id) {
				const params = commandInfo.params;
				const target = this.getSheetTarget(params.unitId, params.subUnitId);
				if (!target) return;
				const { workbook, worksheet } = target;
				const eventParams = {
					worksheet,
					workbook,
					rules: worksheet.getDataValidations()
				};
				this.fireEvent(this.Event.BeforeSheetDataValidationDeleteAll, eventParams);
				if (eventParams.cancel) throw new _univerjs_core.CanceledError();
			}
		})));
	}
};
_univerjs_core_facade.FUniver.extend(FUniverSheetsDataValidationMixin);

//#endregion
//#region src/facade/f-workbook.ts
/**
* @ignore
*/
var FWorkbookSheetsDataValidationMixin = class extends _univerjs_sheets_facade.FWorkbook {
	_initialize() {
		Object.defineProperty(this, "_dataValidationModel", { get() {
			return this._injector.get(_univerjs_sheets_data_validation.SheetDataValidationModel);
		} });
	}
	getValidatorStatus() {
		return this._injector.get(_univerjs_sheets_data_validation.SheetsDataValidationValidatorService).validatorWorkbook(this._workbook.getUnitId());
	}
	async getAllDataValidationErrorAsync() {
		const unitId = this._workbook.getUnitId();
		const sheetIds = this._dataValidationModel.getSubUnitIds(unitId);
		const allErrors = [];
		for (const sheetId of sheetIds) {
			const sheetErrors = await this._collectValidationErrorsForSheet(unitId, sheetId);
			allErrors.push(...sheetErrors);
		}
		return allErrors;
	}
	async _collectValidationErrorsForSheet(unitId, sheetId) {
		const rules = this._dataValidationModel.getRules(unitId, sheetId);
		if (!rules.length) return [];
		const allRanges = rules.flatMap((rule) => rule.ranges);
		return this._collectValidationErrorsForRange(unitId, sheetId, allRanges);
	}
	async _collectValidationErrorsForRange(unitId, sheetId, ranges) {
		if (!ranges.length) return [];
		const validatorService = this._injector.get(_univerjs_sheets_data_validation.SheetsDataValidationValidatorService);
		const worksheet = this._workbook.getSheetBySheetId(sheetId);
		if (!worksheet) throw new Error(`Cannot find worksheet with sheetId: ${sheetId}`);
		const sheetName = worksheet.getName();
		const errors = [];
		for (const range of ranges) {
			const promises = [];
			for (let row = range.startRow; row <= range.endRow; row++) for (let col = range.startColumn; col <= range.endColumn; col++) promises.push((async () => {
				try {
					if (await validatorService.validatorCell(unitId, sheetId, row, col) !== _univerjs_core.DataValidationStatus.VALID) {
						const rule = this._dataValidationModel.getRuleByLocation(unitId, sheetId, row, col);
						if (rule) {
							var _worksheet$getCell;
							const cellValue = ((_worksheet$getCell = worksheet.getCell(row, col)) === null || _worksheet$getCell === void 0 ? void 0 : _worksheet$getCell.v) || null;
							const error = this._createDataValidationError(sheetName, row, col, rule, cellValue);
							errors.push(error);
						}
					}
				} catch (e) {
					console.warn(`Failed to validate cell [${row}, ${col}]:`, e);
				}
			})());
			await Promise.all(promises);
		}
		return errors;
	}
	_createDataValidationError(sheetName, row, column, rule, inputValue) {
		return {
			sheetName,
			row,
			column,
			ruleId: rule.uid,
			inputValue,
			rule
		};
	}
};
_univerjs_sheets_facade.FWorkbook.extend(FWorkbookSheetsDataValidationMixin);

//#endregion
//#region src/facade/f-worksheet.ts
/**
* @ignore
*/
var FWorksheetDataValidationMixin = class extends _univerjs_sheets_facade.FWorksheet {
	getDataValidations() {
		return this._injector.get(_univerjs_data_validation.DataValidationModel).getRules(this._workbook.getUnitId(), this._worksheet.getSheetId()).map((rule) => new FDataValidation(rule, this._worksheet, this._injector));
	}
	getValidatorStatusAsync() {
		return this._injector.get(_univerjs_sheets_data_validation.SheetsDataValidationValidatorService).validatorWorksheet(this._workbook.getUnitId(), this._worksheet.getSheetId());
	}
	getDataValidation(ruleId) {
		const rule = this._injector.get(_univerjs_data_validation.DataValidationModel).getRuleById(this._workbook.getUnitId(), this._worksheet.getSheetId(), ruleId);
		if (rule) return new FDataValidation(rule, this._worksheet, this._injector);
		return null;
	}
	async getAllDataValidationErrorAsync() {
		const unitId = this._workbook.getUnitId();
		const sheetId = this._worksheet.getSheetId();
		return this._collectValidationErrorsForSheet(unitId, sheetId);
	}
	async _collectValidationErrorsForSheet(unitId, sheetId) {
		const rules = this._injector.get(_univerjs_data_validation.DataValidationModel).getRules(unitId, sheetId);
		if (!rules.length) return [];
		const allRanges = rules.flatMap((rule) => rule.ranges);
		return this._collectValidationErrorsForRange(unitId, sheetId, allRanges);
	}
	async _collectValidationErrorsForRange(unitId, sheetId, ranges) {
		if (!ranges.length) return [];
		const validatorService = this._injector.get(_univerjs_sheets_data_validation.SheetsDataValidationValidatorService);
		const worksheet = this._worksheet;
		const sheetName = worksheet.getName();
		const errors = [];
		for (const range of ranges) {
			const promises = [];
			for (let row = range.startRow; row <= range.endRow; row++) for (let col = range.startColumn; col <= range.endColumn; col++) promises.push((async () => {
				try {
					if (await validatorService.validatorCell(unitId, sheetId, row, col) !== _univerjs_core.DataValidationStatus.VALID) {
						const rule = this._injector.get(_univerjs_sheets_data_validation.SheetDataValidationModel).getRuleByLocation(unitId, sheetId, row, col);
						if (rule) {
							var _worksheet$getCell;
							const cellValue = ((_worksheet$getCell = worksheet.getCell(row, col)) === null || _worksheet$getCell === void 0 ? void 0 : _worksheet$getCell.v) || null;
							const error = this._createDataValidationError(sheetName, row, col, rule, cellValue);
							errors.push(error);
						}
					}
				} catch (e) {
					console.warn(`Failed to validate cell [${row}, ${col}]:`, e);
				}
			})());
			await Promise.all(promises);
		}
		return errors;
	}
	_createDataValidationError(sheetName, row, column, rule, inputValue) {
		return {
			sheetName,
			row,
			column,
			ruleId: rule.uid,
			inputValue,
			rule
		};
	}
};
_univerjs_sheets_facade.FWorksheet.extend(FWorksheetDataValidationMixin);

//#endregion
//#region src/facade/f-event.ts
/**
* @ignore
*/
var FSheetsDataValidationEventNameMixin = class extends _univerjs_core_facade.FEventName {
	get SheetDataValidationChanged() {
		return "SheetDataValidationChanged";
	}
	get SheetDataValidatorStatusChanged() {
		return "SheetDataValidatorStatusChanged";
	}
	get BeforeSheetDataValidationAdd() {
		return "BeforeSheetDataValidationAdd";
	}
	get BeforeSheetDataValidationDelete() {
		return "BeforeSheetDataValidationDelete";
	}
	get BeforeSheetDataValidationDeleteAll() {
		return "BeforeSheetDataValidationDeleteAll";
	}
	get BeforeSheetDataValidationCriteriaUpdate() {
		return "BeforeSheetDataValidationCriteriaUpdate";
	}
	get BeforeSheetDataValidationRangeUpdate() {
		return "BeforeSheetDataValidationRangeUpdate";
	}
	get BeforeSheetDataValidationOptionsUpdate() {
		return "BeforeSheetDataValidationOptionsUpdate";
	}
};
_univerjs_core_facade.FEventName.extend(FSheetsDataValidationEventNameMixin);

//#endregion
exports.FDataValidation = FDataValidation;
exports.FDataValidationBuilder = FDataValidationBuilder;
