// @univerjs/rpc/lib/index.js
import { CommandType, Disposable, ICommandService, IConfigService, ILogService, IUniverInstanceService, Inject, Injector, Plugin, RxDisposable, UniverInstanceType, createIdentifier, merge, toDisposable } from "@univerjs/core";
import { filter, take, takeUntil } from "rxjs/operators";
import { BehaviorSubject, Observable, firstValueFrom, isObservable, of, shareReplay } from "rxjs";

//#region src/config/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const PLUGIN_CONFIG_KEY_MAIN_THREAD = "rpc.main-thread.config";
const configSymbolMainThread = Symbol(PLUGIN_CONFIG_KEY_MAIN_THREAD);
const defaultPluginMainThreadConfig = {};
const PLUGIN_CONFIG_KEY_WORKER_THREAD = "rpc.worker-thread.config";
const configSymbolWorkerThread = Symbol(PLUGIN_CONFIG_KEY_WORKER_THREAD);
const defaultPluginWorkerThreadConfig = {};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/services/remote-instance/remote-instance.service.ts
const RemoteSyncServiceName = "rpc.remote-sync.service";
/**
* This service is provided by the primary Univer.
*
* Replica Univer could call this service to update mutations back to the primary Univer.
*/
const IRemoteSyncService = createIdentifier(RemoteSyncServiceName);
let RemoteSyncPrimaryService = class RemoteSyncPrimaryService {
	constructor(_commandService) {
		this._commandService = _commandService;
	}
	async syncMutation(params, options) {
		const { fromCollab, ...restOptions } = options || {};
		return this._commandService.syncExecuteCommand(params.mutationInfo.id, params.mutationInfo.params, {
			...restOptions,
			onlyLocal: true,
			fromSync: true
		});
	}
};
RemoteSyncPrimaryService = __decorate([__decorateParam(0, ICommandService)], RemoteSyncPrimaryService);
const RemoteInstanceServiceName = "univer.remote-instance-service";
/**
* This service is provided by the replica Univer.
*
* Primary univer could call this service to init and dispose univer business instances
* and sync mutations to replica univer.
*/
const IRemoteInstanceService = createIdentifier(RemoteInstanceServiceName);
let WebWorkerRemoteInstanceService = class WebWorkerRemoteInstanceService {
	constructor(_univerInstanceService, _commandService, _logService) {
		this._univerInstanceService = _univerInstanceService;
		this._commandService = _commandService;
		this._logService = _logService;
	}
	whenReady() {
		return Promise.resolve(true);
	}
	async syncMutation(params, options) {
		return this._applyMutation(params.mutationInfo, options);
	}
	async createInstance(params) {
		this._logService.debug(`[WebWorkerRemoteInstanceService]: Creating instance with id ${params.unitID}`);
		const { type, snapshot } = params;
		try {
			switch (type) {
				case UniverInstanceType.UNIVER_SHEET:
					this._univerInstanceService.createUnit(UniverInstanceType.UNIVER_SHEET, snapshot);
					return true;
				default: throw new Error(`[WebWorkerRemoteInstanceService]: cannot create replica for document type: ${type}.`);
			}
		} catch (err) {
			if (err instanceof Error) throw err;
			else throw new TypeError(`${err}`);
		}
	}
	async disposeInstance(params) {
		this._logService.debug(`[WebWorkerRemoteInstanceService]: Disposing instance with id ${params.unitID}`);
		return this._univerInstanceService.disposeUnit(params.unitID);
	}
	_applyMutation(mutationInfo, options) {
		const { id, params: mutationParams } = mutationInfo;
		const { fromCollab, ...restOptions } = options || {};
		return this._commandService.syncExecuteCommand(id, mutationParams, {
			...restOptions,
			onlyLocal: true,
			fromSync: true
		});
	}
};
WebWorkerRemoteInstanceService = __decorate([
	__decorateParam(0, IUniverInstanceService),
	__decorateParam(1, ICommandService),
	__decorateParam(2, ILogService)
], WebWorkerRemoteInstanceService);

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region src/services/rpc/rpc.service.ts
/**
* Wrapper a service or a controller into a channel so it could be invoked by
* a remote client. When the protocol is called, it would forward to the
* underlying service or controller.
*
* @param module the wrapper service or controller
* @returns the channel instance
*/
function fromModule(module) {
	const handler = module;
	return new class {
		call(method, args) {
			const target = handler[method];
			if (typeof target === "function") {
				let res = args ? target.apply(handler, args) : target.call(handler);
				if (!(res instanceof Promise)) res = Promise.resolve(res);
				return res;
			}
			throw new Error(`[RPC]: method not found for ${method}!`);
		}
		subscribe(eventMethod, args) {
			const target = handler[eventMethod];
			if (typeof target === "function") {
				const res = args ? target.apply(handler, args) : target.call(handler);
				if (!isObservable(res)) return of(res);
				return res;
			}
			throw new Error(`[RPC]: observable method not found for ${eventMethod}!`);
		}
	}();
}
/**
* Wrap a channel into a service or a controller so it could be invoked by
* the upper layer modules. When the service or controller is called, it would
* request the remote server by calling the channel.
*
* @param channel
* @returns A proxy object that forwards calls/subscriptions to the remote channel.
*/
function toModule(channel) {
	return new Proxy({}, { get(_, propKey) {
		if (propKey === "dispose") return;
		return function(...args) {
			if (propertyIsEventSource(propKey)) return channel.subscribe(propKey, args);
			return channel.call(propKey, args);
		};
	} });
}
function propertyIsEventSource(name) {
	return name.endsWith("$");
}
/**
* This method provides implementation for `IChannel` and is responsible for
* transforming a local calling to a RPC calling.
*/
var ChannelClient = class extends RxDisposable {
	constructor(_protocol) {
		super();
		this._protocol = _protocol;
		_defineProperty(this, "_initialized", new BehaviorSubject(false));
		_defineProperty(this, "_lastRequestCounter", 0);
		_defineProperty(this, "_pendingRequests", /* @__PURE__ */ new Map());
		this._protocol.send({ type: 50 });
		this._protocol.onMessage.pipe(takeUntil(this.dispose$)).subscribe((message) => this._onMessage(message));
	}
	dispose() {
		this._pendingRequests.clear();
	}
	getChannel(channelName) {
		return {
			call: (method, args) => {
				if (this._disposed) return Promise.reject(/* @__PURE__ */ new Error("[ChannelClient]: client is disposed!"));
				return this._remoteCall(channelName, method, args);
			},
			subscribe: (eventMethod, args) => {
				if (this._disposed) throw new Error("[ChannelClient]: client is disposed!");
				return this._remoteSubscribe(channelName, eventMethod, args);
			}
		};
	}
	_whenReady() {
		return firstValueFrom(this._initialized.pipe(filter((v) => v), take(1)));
	}
	_remoteCall(channelName, method, args) {
		if (this._initialized.getValue()) return this._doRemoteCall(channelName, method, args);
		return this._whenReady().then(() => this._doRemoteCall(channelName, method, args));
	}
	_doRemoteCall(channelName, method, args) {
		const sequence = ++this._lastRequestCounter;
		const request = {
			seq: sequence,
			type: 100,
			channelName,
			method,
			args
		};
		const pendingRequests = this._pendingRequests;
		return new Promise((resolve, reject) => {
			this._pendingRequests.set(sequence, { handle(response) {
				switch (response.type) {
					case 201:
						pendingRequests.delete(sequence);
						resolve(response.data);
						break;
					case 202:
						pendingRequests.delete(sequence);
						reject(response.data);
						break;
					default: throw new Error("[ChannelClient]: unknown response type!");
				}
			} });
			this._sendRequest(request);
		});
	}
	_remoteSubscribe(channelName, method, args) {
		return new Observable((subscriber) => {
			let sequence = -1;
			const doSubscribe = () => {
				sequence = ++this._lastRequestCounter;
				const request = {
					seq: sequence,
					type: 101,
					channelName,
					method,
					args
				};
				this._pendingRequests.set(sequence, { handle(response) {
					switch (response.type) {
						case 300:
							subscriber.next(response.data);
							break;
						case 301:
							subscriber.error(response.data);
							break;
						case 302:
							subscriber.complete();
							break;
						default: throw new Error("[ChannelClient]: unknown response type!");
					}
				} });
				this._sendRequest(request);
			};
			if (this._initialized.getValue()) doSubscribe();
			else this._whenReady().then(doSubscribe);
			return () => {
				if (sequence === -1) return;
				const cancelSubscriptionRequest = {
					type: 102,
					seq: sequence,
					channelName,
					method
				};
				this._sendRequest(cancelSubscriptionRequest);
			};
		});
	}
	_sendRequest(request) {
		this._protocol.send(request);
	}
	_onMessage(response) {
		const { type: responseType, seq } = response;
		switch (responseType) {
			case 0:
				this._initialized.next(true);
				break;
			case 201:
			case 202:
			case 300:
			case 302:
			case 301: {
				var _pendingRequests$get;
				const { _pendingRequests } = this;
				(_pendingRequests$get = _pendingRequests.get(seq)) === null || _pendingRequests$get === void 0 || _pendingRequests$get.handle(response);
				responseType !== 300 && _pendingRequests.delete(seq);
				break;
			}
		}
	}
};
var ChannelServer = class extends RxDisposable {
	constructor(_protocol) {
		super();
		this._protocol = _protocol;
		_defineProperty(this, "_channels", /* @__PURE__ */ new Map());
		_defineProperty(this, "_subscriptions", /* @__PURE__ */ new Map());
		this._protocol.onMessage.pipe(takeUntil(this.dispose$)).subscribe((message) => this._onRequest(message));
		this._sendInitialize();
	}
	dispose() {
		super.dispose();
		this._subscriptions.clear();
		this._channels.clear();
	}
	registerChannel(channelName, channel) {
		this._channels.set(channelName, channel);
	}
	_onRequest(request) {
		switch (request.type) {
			case 50:
				this._sendInitialize();
				break;
			case 100:
				this._onMethodCall(request);
				break;
			case 101:
				this._onSubscribe(request);
				break;
			case 102:
				this._onUnsubscribe(request);
				break;
			default: break;
		}
	}
	_sendInitialize() {
		this._sendResponse({
			seq: -1,
			type: 0
		});
	}
	_onMethodCall(request) {
		const { channelName, method, args } = request;
		const channel = this._channels.get(channelName);
		let promise;
		try {
			if (!channel) throw new Error(`[ChannelServer]: Channel ${channelName} not found!`);
			promise = args ? channel.call(method, args) : channel.call(method);
		} catch (err) {
			promise = Promise.reject(err instanceof Error ? err : new Error(String(err)));
		}
		promise.then((data) => {
			this._sendResponse({
				seq: request.seq,
				type: 201,
				data
			});
		}).catch((err) => {
			if (err instanceof Error) this._sendResponse({
				seq: request.seq,
				type: 202,
				data: err.message
			});
			else this._sendResponse({
				seq: request.seq,
				type: 202,
				data: String(err)
			});
		});
	}
	_onSubscribe(request) {
		const { channelName, seq } = request;
		const channel = this._channels.get(channelName);
		try {
			if (!channel) throw new Error(`[ChannelServer]: Channel ${channelName} not found!`);
			const subscription = channel.subscribe(request.method, request.args).subscribe({
				next: (data) => {
					this._sendResponse({
						seq,
						type: 300,
						data
					});
				},
				error: (err) => {
					this._sendResponse({
						seq,
						type: 301,
						data: err.message
					});
					this._sendResponse({
						seq,
						type: 302
					});
				},
				complete: () => {
					this._sendResponse({
						seq,
						type: 302
					});
				}
			});
			this._subscriptions.set(request.seq, subscription);
		} catch (err) {
			if (err instanceof Error) this._sendResponse({
				seq: request.seq,
				type: 301,
				data: err.message
			});
			else this._sendResponse({
				seq: request.seq,
				type: 301,
				data: String(err)
			});
		}
	}
	_onUnsubscribe(request) {
		const subscription = this._subscriptions.get(request.seq);
		if (subscription) {
			subscription.unsubscribe();
			this._subscriptions.delete(request.seq);
		}
	}
	_sendResponse(response) {
		this._protocol.send(response);
	}
};

//#endregion
//#region src/services/rpc/channel.service.ts
const IRPCChannelService = createIdentifier("IRPCChannelService");
/**
* This service is responsible for managing the RPC channels.
*/
var ChannelService = class {
	constructor(_messageProtocol) {
		_defineProperty(this, "_client", void 0);
		_defineProperty(this, "_server", void 0);
		this._client = new ChannelClient(_messageProtocol);
		this._server = new ChannelServer(_messageProtocol);
	}
	dispose() {
		this._client.dispose();
		this._server.dispose();
	}
	requestChannel(name) {
		return this._client.getChannel(name);
	}
	registerChannel(name, channel) {
		this._server.registerChannel(name, channel);
	}
};

//#endregion
//#region src/controllers/data-sync/data-sync-primary.controller.ts
let DataSyncPrimaryController = class DataSyncPrimaryController extends RxDisposable {
	constructor(_injector, _commandService, _univerInstanceService, _rpcChannelService, _remoteSyncService) {
		super();
		this._injector = _injector;
		this._commandService = _commandService;
		this._univerInstanceService = _univerInstanceService;
		this._rpcChannelService = _rpcChannelService;
		this._remoteSyncService = _remoteSyncService;
		_defineProperty(this, "_remoteInstanceService", void 0);
		_defineProperty(this, "_syncingUnits", /* @__PURE__ */ new Set());
		_defineProperty(this, "_syncingMutations", /* @__PURE__ */ new Set());
		this._initRPCChannels();
		this._init();
	}
	registerSyncingMutations(mutation) {
		this._syncingMutations.add(mutation.id);
	}
	/**
	* Only spreadsheets would be synced to the web worker in normal situations. If you would like to
	* sync other types of documents, you should manually call this method with that document's id.
	*/
	syncUnit(unitId) {
		this._syncingUnits.add(unitId);
		return toDisposable(() => this._syncingUnits.delete(unitId));
	}
	_initRPCChannels() {
		this._rpcChannelService.registerChannel(RemoteSyncServiceName, fromModule(this._remoteSyncService));
		this._injector.add([IRemoteInstanceService, { useFactory: () => toModule(this._rpcChannelService.requestChannel(RemoteInstanceServiceName)) }]);
		this._remoteInstanceService = this._injector.get(IRemoteInstanceService);
	}
	_init() {
		this._univerInstanceService.getTypeOfUnitAdded$(UniverInstanceType.UNIVER_SHEET).pipe(takeUntil(this.dispose$)).subscribe((event) => {
			const { unit: sheet } = event;
			this._syncingUnits.add(sheet.getUnitId());
			this._remoteInstanceService.createInstance({
				unitID: sheet.getUnitId(),
				type: UniverInstanceType.UNIVER_SHEET,
				snapshot: sheet.getSnapshot()
			});
		});
		this._univerInstanceService.getTypeOfUnitDisposed$(UniverInstanceType.UNIVER_SHEET).pipe(takeUntil(this.dispose$)).subscribe((workbook) => {
			this._syncingUnits.delete(workbook.getUnitId());
			this._remoteInstanceService.disposeInstance({ unitID: workbook.getUnitId() });
		});
		this.disposeWithMe(this._commandService.onCommandExecuted((commandInfo, options) => {
			const { type, params, id } = commandInfo;
			const unitId = (params === null || params === void 0 ? void 0 : params.unitId) || "";
			if (type === CommandType.MUTATION && (!unitId || this._syncingUnits.has(unitId)) && !(options === null || options === void 0 ? void 0 : options.fromSync) && this._syncingMutations.has(id)) this._remoteInstanceService.syncMutation({ mutationInfo: commandInfo }, options);
		}));
	}
};
DataSyncPrimaryController = __decorate([
	__decorateParam(0, Inject(Injector)),
	__decorateParam(1, ICommandService),
	__decorateParam(2, IUniverInstanceService),
	__decorateParam(3, IRPCChannelService),
	__decorateParam(4, IRemoteSyncService)
], DataSyncPrimaryController);

//#endregion
//#region src/controllers/data-sync/data-sync-replica.controller.ts
let DataSyncReplicaController = class DataSyncReplicaController extends Disposable {
	constructor(_injector, _remoteInstanceService, _commandService, _rpcChannelService) {
		super();
		this._injector = _injector;
		this._remoteInstanceService = _remoteInstanceService;
		this._commandService = _commandService;
		this._rpcChannelService = _rpcChannelService;
		_defineProperty(this, "_remoteSyncService", void 0);
		this._initRPCChannels();
		this._init();
	}
	_initRPCChannels() {
		this._rpcChannelService.registerChannel(RemoteInstanceServiceName, fromModule(this._remoteInstanceService));
		this._injector.add([IRemoteSyncService, { useFactory: () => toModule(this._rpcChannelService.requestChannel(RemoteSyncServiceName)) }]);
		this._remoteSyncService = this._injector.get(IRemoteSyncService);
	}
	_init() {
		this.disposeWithMe(this._commandService.onCommandExecuted((commandInfo, options) => {
			if (commandInfo.type === CommandType.MUTATION && !(options === null || options === void 0 ? void 0 : options.fromSync)) this._remoteSyncService.syncMutation({ mutationInfo: commandInfo }, options);
		}));
	}
};
DataSyncReplicaController = __decorate([
	__decorateParam(0, Inject(Injector)),
	__decorateParam(1, IRemoteInstanceService),
	__decorateParam(2, ICommandService),
	__decorateParam(3, IRPCChannelService)
], DataSyncReplicaController);

//#endregion
//#region package.json
var name = "@univerjs/rpc";
var version = "0.25.1";

//#endregion
//#region src/services/rpc/implementations/web-worker-rpc.service.ts
/**
* Generate an `IMessageProtocol` on the web worker.
* @returns A protocol wrapper around worker global messaging APIs.
*/
function createWebWorkerMessagePortOnWorker() {
	return {
		send(message) {
			postMessage(message);
		},
		onMessage: new Observable((subscriber) => {
			const handler = (event) => {
				subscriber.next(event.data);
			};
			addEventListener("message", handler);
			return () => removeEventListener("message", handler);
		}).pipe(shareReplay({
			bufferSize: 1,
			refCount: true
		}))
	};
}
/**
* Generate an `IMessageProtocol` on the main thread side.
* @param worker The Web Worker object
* @returns A protocol wrapper around the given worker messaging APIs.
*/
function createWebWorkerMessagePortOnMain(worker) {
	return {
		send(message) {
			worker.postMessage(message);
		},
		onMessage: new Observable((subscriber) => {
			const handler = (event) => {
				subscriber.next(event.data);
			};
			worker.addEventListener("message", handler);
			return () => worker.removeEventListener("message", handler);
		}).pipe(shareReplay({
			bufferSize: 1,
			refCount: true
		}))
	};
}

//#endregion
//#region src/plugin.ts
let UniverRPCMainThreadPlugin = class UniverRPCMainThreadPlugin extends Plugin {
	constructor(_config = defaultPluginMainThreadConfig, _injector, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		_defineProperty(this, "_internalWorker", null);
		const { ...rest } = merge({}, defaultPluginMainThreadConfig, this._config);
		this._configService.setConfig(PLUGIN_CONFIG_KEY_MAIN_THREAD, rest);
	}
	dispose() {
		super.dispose();
		if (this._internalWorker) {
			this._internalWorker.terminate();
			this._internalWorker = null;
		}
	}
	onStarting() {
		const { workerURL } = this._config;
		if (!workerURL) throw new Error("[UniverRPCMainThreadPlugin]: The workerURL is required for the RPC main thread plugin.");
		const worker = workerURL instanceof Worker ? workerURL : new Worker(workerURL);
		this._internalWorker = workerURL instanceof Worker ? null : worker;
		const messageProtocol = createWebWorkerMessagePortOnMain(worker);
		[
			[IRPCChannelService, { useFactory: () => new ChannelService(messageProtocol) }],
			[DataSyncPrimaryController],
			[IRemoteSyncService, { useClass: RemoteSyncPrimaryService }]
		].forEach((dependency) => this._injector.add(dependency));
		this._injector.get(DataSyncPrimaryController);
	}
};
_defineProperty(UniverRPCMainThreadPlugin, "pluginName", "UNIVER_RPC_MAIN_THREAD_PLUGIN");
_defineProperty(UniverRPCMainThreadPlugin, "packageName", name);
_defineProperty(UniverRPCMainThreadPlugin, "version", version);
UniverRPCMainThreadPlugin = __decorate([__decorateParam(1, Inject(Injector)), __decorateParam(2, IConfigService)], UniverRPCMainThreadPlugin);
let UniverRPCWorkerThreadPlugin = class UniverRPCWorkerThreadPlugin extends Plugin {
	constructor(_config = defaultPluginWorkerThreadConfig, _injector, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		const { ...rest } = merge({}, defaultPluginWorkerThreadConfig, this._config);
		this._configService.setConfig(PLUGIN_CONFIG_KEY_WORKER_THREAD, rest);
	}
	onStarting() {
		[
			[DataSyncReplicaController],
			[IRPCChannelService, { useFactory: () => new ChannelService(createWebWorkerMessagePortOnWorker()) }],
			[IRemoteInstanceService, { useClass: WebWorkerRemoteInstanceService }]
		].forEach((dependency) => this._injector.add(dependency));
		this._injector.get(DataSyncReplicaController);
	}
};
_defineProperty(UniverRPCWorkerThreadPlugin, "pluginName", "UNIVER_RPC_WORKER_THREAD_PLUGIN");
_defineProperty(UniverRPCWorkerThreadPlugin, "packageName", name);
_defineProperty(UniverRPCWorkerThreadPlugin, "version", version);
UniverRPCWorkerThreadPlugin = __decorate([__decorateParam(1, Inject(Injector)), __decorateParam(2, IConfigService)], UniverRPCWorkerThreadPlugin);

//#endregion
export { ChannelClient, ChannelServer, ChannelService, DataSyncPrimaryController, DataSyncReplicaController, IRPCChannelService, IRemoteInstanceService, IRemoteSyncService, PLUGIN_CONFIG_KEY_MAIN_THREAD, PLUGIN_CONFIG_KEY_WORKER_THREAD, RemoteInstanceServiceName, RemoteSyncPrimaryService, RemoteSyncServiceName, UniverRPCMainThreadPlugin, UniverRPCWorkerThreadPlugin, WebWorkerRemoteInstanceService, fromModule, toModule };
