// @univerjs/drawing/lib/es/index.js
import { CommandType, Disposable, DrawingTypeEnum, ICommandService, IConfigService, IImageIoService, IURLImageService, ImageSourceType, ImageUploadStatusType, Inject, Injector, Plugin, createIdentifier, generateRandomId, merge, mergeOverrideWithDependencies, sortRules, sortRulesByDesc, toDisposable } from "@univerjs/core";
import * as json1 from "ot-json1";
import { Subject } from "rxjs";

//#region src/basics/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const DRAWING_IMAGE_WIDTH_LIMIT = 500;
const DRAWING_IMAGE_HEIGHT_LIMIT = 500;
const DRAWING_IMAGE_COUNT_LIMIT = 10;
let DRAWING_IMAGE_ALLOW_SIZE = 5 * 1024 * 1024;
function setDrawingImageAllowSize(size) {
	DRAWING_IMAGE_ALLOW_SIZE = size;
}
function getDrawingImageAllowSize() {
	return DRAWING_IMAGE_ALLOW_SIZE;
}
const DRAWING_IMAGE_ALLOW_IMAGE_LIST = [
	"image/png",
	"image/jpeg",
	"image/jpg",
	"image/gif",
	"image/bmp"
];

//#endregion
//#region src/services/drawing-manager.service.ts
const IDrawingManagerService = createIdentifier("univer.drawing-manager.service");

//#endregion
//#region src/commands/operations/set-drawing-selected.operation.ts
const SetDrawingSelectedOperation = {
	id: "drawing.operation.set-drawing-selected",
	type: CommandType.OPERATION,
	handler: (accessor, params) => {
		const drawingManagerService = accessor.get(IDrawingManagerService);
		if (params == null) return false;
		drawingManagerService.focusDrawing(params);
		return true;
	}
};

//#endregion
//#region package.json
var name = "@univerjs/drawing";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
const DRAWING_PLUGIN_CONFIG_KEY = "drawing.config";
const configSymbol = Symbol(DRAWING_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region src/services/drawing-manager-impl.service.ts
/**
* unitId -> subUnitId -> drawingId -> drawingParam
*/
var UnitDrawingService = class {
	constructor() {
		_defineProperty(this, "drawingManagerData", {});
		_defineProperty(this, "_oldDrawingManagerData", {});
		_defineProperty(this, "_focusDrawings", []);
		_defineProperty(this, "_remove$", new Subject());
		_defineProperty(this, "remove$", this._remove$.asObservable());
		_defineProperty(this, "_add$", new Subject());
		_defineProperty(this, "add$", this._add$.asObservable());
		_defineProperty(this, "_update$", new Subject());
		_defineProperty(this, "update$", this._update$.asObservable());
		_defineProperty(this, "_order$", new Subject());
		_defineProperty(this, "order$", this._order$.asObservable());
		_defineProperty(this, "_group$", new Subject());
		_defineProperty(this, "group$", this._group$.asObservable());
		_defineProperty(this, "_ungroup$", new Subject());
		_defineProperty(this, "ungroup$", this._ungroup$.asObservable());
		_defineProperty(this, "_refreshTransform$", new Subject());
		_defineProperty(this, "refreshTransform$", this._refreshTransform$.asObservable());
		_defineProperty(this, "_visible$", new Subject());
		_defineProperty(this, "visible$", this._visible$.asObservable());
		_defineProperty(this, "_focus$", new Subject());
		_defineProperty(this, "focus$", this._focus$.asObservable());
		_defineProperty(this, "_featurePluginUpdate$", new Subject());
		_defineProperty(this, "featurePluginUpdate$", this._featurePluginUpdate$.asObservable());
		_defineProperty(this, "_featurePluginAdd$", new Subject());
		_defineProperty(this, "featurePluginAdd$", this._featurePluginAdd$.asObservable());
		_defineProperty(this, "_featurePluginRemove$", new Subject());
		_defineProperty(this, "featurePluginRemove$", this._featurePluginRemove$.asObservable());
		_defineProperty(this, "_featurePluginOrderUpdate$", new Subject());
		_defineProperty(this, "featurePluginOrderUpdate$", this._featurePluginOrderUpdate$.asObservable());
		_defineProperty(this, "_featurePluginGroupUpdate$", new Subject());
		_defineProperty(this, "featurePluginGroupUpdate$", this._featurePluginGroupUpdate$.asObservable());
		_defineProperty(this, "_featurePluginUngroupUpdate$", new Subject());
		_defineProperty(this, "featurePluginUngroupUpdate$", this._featurePluginUngroupUpdate$.asObservable());
		_defineProperty(this, "_visible", true);
		_defineProperty(this, "_editable", true);
	}
	dispose() {
		this._remove$.complete();
		this._add$.complete();
		this._update$.complete();
		this._order$.complete();
		this._focus$.complete();
		this._featurePluginUpdate$.complete();
		this._featurePluginAdd$.complete();
		this._featurePluginRemove$.complete();
		this._featurePluginOrderUpdate$.complete();
		this.drawingManagerData = {};
		this._oldDrawingManagerData = {};
	}
	visibleNotification(visibleParams) {
		this._visible$.next(visibleParams);
	}
	refreshTransform(updateParams) {
		updateParams.forEach((updateParam) => {
			const param = this._getCurrentBySearch(updateParam);
			if (param == null) return;
			param.transform = updateParam.transform;
			param.transforms = updateParam.transforms;
			param.isMultiTransform = updateParam.isMultiTransform;
		});
		this.refreshTransformNotification(updateParams);
	}
	getDrawingDataForUnit(unitId) {
		return this.drawingManagerData[unitId] || {};
	}
	removeDrawingDataForUnit(unitId) {
		const subUnits = this.drawingManagerData[unitId];
		if (subUnits == null) return;
		delete this.drawingManagerData[unitId];
		const drawings = [];
		Object.keys(subUnits).forEach((subUnitId) => {
			const subUnit = subUnits[subUnitId];
			if ((subUnit === null || subUnit === void 0 ? void 0 : subUnit.data) == null) return;
			Object.keys(subUnit.data).forEach((drawingId) => {
				drawings.push({
					unitId,
					subUnitId,
					drawingId
				});
			});
		});
		if (drawings.length > 0) this.removeNotification(drawings);
	}
	registerDrawingData(unitId, data) {
		this.drawingManagerData[unitId] = data;
	}
	initializeNotification(unitId) {
		const drawings = [];
		const data = this.drawingManagerData[unitId];
		if (data == null) return;
		Object.keys(data).forEach((subUnitId) => {
			this._establishDrawingMap(unitId, subUnitId);
			const subUnitData = data[subUnitId];
			Object.keys(subUnitData.data).forEach((drawingId) => {
				const drawing = subUnitData.data[drawingId];
				drawing.unitId = unitId;
				drawing.subUnitId = subUnitId;
				drawings.push(drawing);
			});
		});
		if (drawings.length > 0) this.addNotification(drawings);
	}
	getDrawingData(unitId, subUnitId) {
		return this._getDrawingData(unitId, subUnitId);
	}
	setDrawingData(unitId, subUnitId, data) {
		this.drawingManagerData[unitId][subUnitId].data = data;
	}
	getBatchAddOp(insertParams) {
		const objects = [];
		const ops = [];
		const invertOps = [];
		insertParams.forEach((insertParam) => {
			const { op, invertOp } = this._addByParam(insertParam);
			objects.push({
				unitId: insertParam.unitId,
				subUnitId: insertParam.subUnitId,
				drawingId: insertParam.drawingId
			});
			ops.push(op);
			invertOps.push(invertOp);
		});
		const op = ops.reduce(json1.type.compose, null);
		const invertOp = invertOps.reduce(json1.type.compose, null);
		const { unitId, subUnitId } = insertParams[0];
		return {
			undo: invertOp,
			redo: op,
			unitId,
			subUnitId,
			objects
		};
	}
	getBatchRemoveOpInOder(removeParams) {
		if (removeParams.length === 0) return {
			undo: null,
			redo: null,
			unitId: "",
			subUnitId: "",
			objects: []
		};
		const orders = this.getDrawingOrder(removeParams[0].unitId, removeParams[0].subUnitId);
		const orderIndexMap = /* @__PURE__ */ new Map();
		orders.forEach((drawingId, index) => {
			orderIndexMap.set(drawingId, index);
		});
		removeParams.sort((a, b) => {
			var _orderIndexMap$get, _orderIndexMap$get2;
			return ((_orderIndexMap$get = orderIndexMap.get(a.drawingId)) !== null && _orderIndexMap$get !== void 0 ? _orderIndexMap$get : Number.NEGATIVE_INFINITY) - ((_orderIndexMap$get2 = orderIndexMap.get(b.drawingId)) !== null && _orderIndexMap$get2 !== void 0 ? _orderIndexMap$get2 : Number.NEGATIVE_INFINITY);
		});
		return this.getBatchRemoveOp(removeParams);
	}
	getBatchRemoveOp(removeParams) {
		var _allToRemove$;
		const seenIds = /* @__PURE__ */ new Set();
		const allToRemove = [];
		removeParams.forEach((removeParam) => {
			const drawing = this.getDrawingByParam(removeParam);
			if ((drawing === null || drawing === void 0 ? void 0 : drawing.drawingType) === DrawingTypeEnum.DRAWING_GROUP) {
				const nested = this.getDrawingsByGroupNested(removeParam);
				if (nested) {
					const { flatChildren, groups } = nested;
					[...flatChildren !== null && flatChildren !== void 0 ? flatChildren : [], ...groups].forEach((d) => {
						if (!seenIds.has(d.drawingId)) {
							seenIds.add(d.drawingId);
							allToRemove.push({
								unitId: d.unitId,
								subUnitId: d.subUnitId,
								drawingId: d.drawingId
							});
						}
					});
				} else if (!seenIds.has(removeParam.drawingId)) {
					seenIds.add(removeParam.drawingId);
					allToRemove.push(removeParam);
				}
			} else if (!seenIds.has(removeParam.drawingId)) {
				seenIds.add(removeParam.drawingId);
				allToRemove.push(removeParam);
			}
		});
		const { unitId, subUnitId } = (_allToRemove$ = allToRemove[0]) !== null && _allToRemove$ !== void 0 ? _allToRemove$ : removeParams[0];
		const orderArr = this._getDrawingOrder(unitId, subUnitId);
		const orderIndexMap = /* @__PURE__ */ new Map();
		orderArr.forEach((id, idx) => orderIndexMap.set(id, idx));
		allToRemove.sort((a, b) => {
			var _orderIndexMap$get3, _orderIndexMap$get4;
			return ((_orderIndexMap$get3 = orderIndexMap.get(a.drawingId)) !== null && _orderIndexMap$get3 !== void 0 ? _orderIndexMap$get3 : Number.NEGATIVE_INFINITY) - ((_orderIndexMap$get4 = orderIndexMap.get(b.drawingId)) !== null && _orderIndexMap$get4 !== void 0 ? _orderIndexMap$get4 : Number.NEGATIVE_INFINITY);
		});
		const ops = [];
		const invertOps = [];
		allToRemove.forEach((removeParam) => {
			const { op, invertOp } = this._removeByParam(removeParam);
			/**
			* ot-json compose case
			* two remove ops to does composition
			* ops: [[unit, sheetUnit, order, 0, { r: true }], [unit, sheetUnit, order, 1, { r: true }]]
			* We expected them to composed as [unit, sheetUnit, order, [0, { r: true }], [1, { r: true }]]
			* But extremely confusing to get [unit, sheetUnit, order, 0, { r: true }, 2, { r: true }]
			* And We apply this composed op to data, it's no item with index 2 can be removed.
			* So use unshift api instead of push here.
			*/
			ops.unshift(op);
			invertOps.push(invertOp);
		});
		const op = ops.reduce(json1.type.compose, null);
		return {
			undo: invertOps.reduce(json1.type.compose, null),
			redo: op,
			unitId,
			subUnitId,
			objects: allToRemove
		};
	}
	getBatchUpdateOp(updateParams) {
		const objects = [];
		const ops = [];
		const invertOps = [];
		updateParams.forEach((updateParam) => {
			const { op, invertOp } = this._updateByParam(updateParam);
			objects.push({
				unitId: updateParam.unitId,
				subUnitId: updateParam.subUnitId,
				drawingId: updateParam.drawingId
			});
			ops.push(op);
			invertOps.push(invertOp);
		});
		const op = ops.reduce(json1.type.compose, null);
		const invertOp = invertOps.reduce(json1.type.compose, null);
		const { unitId, subUnitId } = updateParams[0];
		return {
			undo: invertOp,
			redo: op,
			unitId,
			subUnitId,
			objects
		};
	}
	removeNotification(removeParams) {
		this._remove$.next(removeParams);
	}
	addNotification(insertParams) {
		this._add$.next(insertParams);
	}
	updateNotification(updateParams) {
		this._update$.next(updateParams);
	}
	orderNotification(orderParams) {
		this._order$.next(orderParams);
	}
	groupUpdateNotification(groupParams) {
		this._group$.next(groupParams);
	}
	ungroupUpdateNotification(groupParams) {
		this._ungroup$.next(groupParams);
	}
	refreshTransformNotification(refreshParams) {
		this._refreshTransform$.next(refreshParams);
	}
	getGroupDrawingOp(groupParams) {
		const ops = [];
		const { unitId, subUnitId } = groupParams[0].parent;
		groupParams.forEach((groupParam) => {
			ops.push(this._getGroupDrawingOp(groupParam));
		});
		const op = ops.reduce(json1.type.compose, null);
		return {
			undo: json1.type.invertWithDoc(op, this.drawingManagerData),
			redo: op,
			unitId,
			subUnitId,
			objects: groupParams
		};
	}
	getUngroupDrawingOp(groupParams) {
		const ops = [];
		const { unitId, subUnitId } = groupParams[0].parent;
		groupParams.forEach((groupParam) => {
			ops.push(this._getUngroupDrawingOp(groupParam));
		});
		const op = ops.reduce(json1.type.compose, null);
		return {
			undo: json1.type.invertWithDoc(op, this.drawingManagerData),
			redo: op,
			unitId,
			subUnitId,
			objects: groupParams
		};
	}
	getDrawingsByGroup(groupParam) {
		const { unitId, subUnitId, drawingId } = groupParam;
		if (this.getDrawingByParam({
			unitId,
			subUnitId,
			drawingId
		}) == null) return [];
		const drawings = this._getDrawingData(unitId, subUnitId);
		const children = [];
		Object.keys(drawings).forEach((key) => {
			const drawing = drawings[key];
			if (drawing.groupId === drawingId) children.push(drawing);
		});
		return children;
	}
	getDrawingsByGroupNested(groupSearch) {
		const { unitId, subUnitId } = groupSearch;
		const rootParam = this.getDrawingByParam(groupSearch);
		if (!rootParam) return null;
		const allDrawings = this._getDrawingData(unitId, subUnitId);
		const groupDerivedDrawingsIdMap = /* @__PURE__ */ new Map();
		Object.values(allDrawings).forEach((drawing) => {
			if (drawing.groupId != null) {
				if (!groupDerivedDrawingsIdMap.has(drawing.groupId)) groupDerivedDrawingsIdMap.set(drawing.groupId, []);
				groupDerivedDrawingsIdMap.get(drawing.groupId).push(drawing.drawingId);
			}
		});
		const flatChildren = [];
		const groups = [];
		const nestedIdRecord = {};
		const dfs = (param) => {
			var _groupDerivedDrawings;
			const { drawingId } = param;
			const childrenIds = (_groupDerivedDrawings = groupDerivedDrawingsIdMap.get(drawingId)) !== null && _groupDerivedDrawings !== void 0 ? _groupDerivedDrawings : [];
			nestedIdRecord[drawingId] = {
				drawingId,
				children: childrenIds
			};
			childrenIds.forEach((childId) => {
				const childParam = allDrawings[childId];
				if (!childParam) return;
				if (childParam.drawingType === DrawingTypeEnum.DRAWING_GROUP) {
					dfs(childParam);
					groups.push(childParam);
				} else flatChildren.push(childParam);
			});
		};
		dfs(rootParam);
		groups.push(rootParam);
		return {
			nestedIdRecord,
			flatChildren,
			groups
		};
	}
	_getGroupDrawingOp(groupParam) {
		const { parent, children } = groupParam;
		const { unitId: groupUnitId, subUnitId: groupSubUnitId, drawingId: groupDrawingId } = parent;
		const ops = [];
		ops.push(json1.insertOp([
			groupUnitId,
			groupSubUnitId,
			"data",
			groupDrawingId
		], parent));
		let maxChildIndex = Number.NEGATIVE_INFINITY;
		children.forEach((child) => {
			const { unitId, subUnitId, drawingId } = child;
			const index = this._hasDrawingOrder({
				unitId,
				subUnitId,
				drawingId
			});
			maxChildIndex = Math.max(maxChildIndex, index);
			ops.push(...this._getUpdateParamCompareOp(child, this.getDrawingByParam({
				unitId,
				subUnitId,
				drawingId
			})));
		});
		if (maxChildIndex === Number.NEGATIVE_INFINITY) maxChildIndex = this._getDrawingOrder(groupUnitId, groupSubUnitId).length;
		ops.push(json1.insertOp([
			groupUnitId,
			groupSubUnitId,
			"order",
			maxChildIndex
		], groupDrawingId));
		return ops.reduce(json1.type.compose, null);
	}
	_getUngroupDrawingOp(groupParam) {
		const { parent, children } = groupParam;
		const { unitId: groupUnitId, subUnitId: groupSubUnitId, drawingId: groupDrawingId } = parent;
		const ops = [];
		children.forEach((child) => {
			const { unitId, subUnitId, drawingId } = child;
			ops.push(...this._getUpdateParamCompareOp(child, this.getDrawingByParam({
				unitId,
				subUnitId,
				drawingId
			})));
		});
		ops.push(json1.removeOp([
			groupUnitId,
			groupSubUnitId,
			"data",
			groupDrawingId
		], true));
		ops.push(json1.removeOp([
			groupUnitId,
			groupSubUnitId,
			"order",
			this._getDrawingOrder(groupUnitId, groupSubUnitId).indexOf(groupDrawingId)
		], true));
		return ops.reduce(json1.type.compose, null);
	}
	applyJson1(unitId, subUnitId, jsonOp) {
		this._establishDrawingMap(unitId, subUnitId);
		this._oldDrawingManagerData = { ...this.drawingManagerData };
		this.drawingManagerData = json1.type.apply(this.drawingManagerData, jsonOp);
	}
	featurePluginUpdateNotification(updateParams) {
		this._featurePluginUpdate$.next(updateParams);
	}
	featurePluginOrderUpdateNotification(drawingOrderUpdateParam) {
		this._featurePluginOrderUpdate$.next(drawingOrderUpdateParam);
	}
	featurePluginAddNotification(insertParams) {
		this._featurePluginAdd$.next(insertParams);
	}
	featurePluginRemoveNotification(removeParams) {
		this._featurePluginRemove$.next(removeParams);
	}
	featurePluginGroupUpdateNotification(groupParams) {
		this._featurePluginGroupUpdate$.next(groupParams);
	}
	featurePluginUngroupUpdateNotification(groupParams) {
		this._featurePluginUngroupUpdate$.next(groupParams);
	}
	getDrawingByParam(param) {
		return this._getCurrentBySearch(param);
	}
	getOldDrawingByParam(param) {
		return this._getOldBySearch(param);
	}
	getDrawingOKey(oKey) {
		const [unitId, subUnitId, drawingId] = oKey.split("#-#");
		return this._getCurrentBySearch({
			unitId,
			subUnitId,
			drawingId
		});
	}
	focusDrawing(params) {
		if (params == null || params.length === 0) {
			this._focusDrawings = [];
			this._focus$.next([]);
			return;
		}
		const drawingParams = [];
		params.forEach((param) => {
			var _this$_getDrawingData;
			const { unitId, subUnitId, drawingId } = param;
			const item = (_this$_getDrawingData = this._getDrawingData(unitId, subUnitId)) === null || _this$_getDrawingData === void 0 ? void 0 : _this$_getDrawingData[drawingId];
			if (item != null) drawingParams.push(item);
		});
		if (drawingParams.length > 0) {
			this._focusDrawings = drawingParams;
			this._focus$.next(drawingParams);
		}
	}
	getFocusDrawings() {
		const drawingParams = [];
		this._focusDrawings.forEach((param) => {
			var _this$_getDrawingData2;
			const { unitId, subUnitId, drawingId } = param;
			const item = (_this$_getDrawingData2 = this._getDrawingData(unitId, subUnitId)) === null || _this$_getDrawingData2 === void 0 ? void 0 : _this$_getDrawingData2[drawingId];
			if (item != null) drawingParams.push(item);
		});
		return drawingParams;
	}
	getDrawingOrder(unitId, subUnitId) {
		return this._getDrawingOrder(unitId, subUnitId);
	}
	setDrawingOrder(unitId, subUnitId, order) {
		this.drawingManagerData[unitId][subUnitId].order = order;
	}
	orderUpdateNotification(orderParams) {
		this._order$.next(orderParams);
	}
	getForwardDrawingsOp(orderParams) {
		const { unitId, subUnitId, drawingIds } = orderParams;
		const ops = [];
		const orders = this.getDrawingOrder(unitId, subUnitId);
		const newIds = [...drawingIds];
		drawingIds.forEach((drawingId) => {
			const index = this._hasDrawingOrder({
				unitId,
				subUnitId,
				drawingId
			});
			if (index === -1 || index === orders.length - 1) return;
			const op = json1.moveOp([
				unitId,
				subUnitId,
				"order",
				index
			], [
				unitId,
				subUnitId,
				"order",
				index + 1
			]);
			ops.push(op);
			if (!newIds.includes(orders[index + 1])) newIds.push(orders[index + 1]);
		});
		const op = ops.reduce(json1.type.compose, null);
		return {
			undo: json1.type.invertWithDoc(op, this.drawingManagerData),
			redo: op,
			unitId,
			subUnitId,
			objects: {
				...orderParams,
				drawingIds: newIds
			}
		};
	}
	getBackwardDrawingOp(orderParams) {
		const { unitId, subUnitId, drawingIds } = orderParams;
		const ops = [];
		const orders = this.getDrawingOrder(unitId, subUnitId);
		const newIds = [...drawingIds];
		drawingIds.forEach((drawingId) => {
			const index = this._hasDrawingOrder({
				unitId,
				subUnitId,
				drawingId
			});
			if (index === -1 || index === 0) return;
			const op = json1.moveOp([
				unitId,
				subUnitId,
				"order",
				index
			], [
				unitId,
				subUnitId,
				"order",
				index - 1
			]);
			ops.push(op);
			if (!newIds.includes(orders[index - 1])) newIds.push(orders[index - 1]);
		});
		const op = ops.reduce(json1.type.compose, null);
		return {
			undo: json1.type.invertWithDoc(op, this.drawingManagerData),
			redo: op,
			unitId,
			subUnitId,
			objects: {
				...orderParams,
				drawingIds: newIds
			}
		};
	}
	getFrontDrawingsOp(orderParams) {
		const { unitId, subUnitId, drawingIds } = orderParams;
		const orderDrawingIds = this._getOrderFromSearchParams(unitId, subUnitId, drawingIds);
		const newIds = [...drawingIds];
		const orders = this.getDrawingOrder(unitId, subUnitId);
		const ops = [];
		orderDrawingIds.forEach((orderDrawingId) => {
			const { drawingId } = orderDrawingId;
			const index = this._getDrawingCount(unitId, subUnitId) - 1;
			const op = json1.moveOp([
				unitId,
				subUnitId,
				"order",
				this._getDrawingOrder(unitId, subUnitId).indexOf(drawingId)
			], [
				unitId,
				subUnitId,
				"order",
				index
			]);
			ops.push(op);
			if (!newIds.includes(orders[index])) newIds.push(orders[index]);
		});
		const op = ops.reduce(json1.type.compose, null);
		return {
			undo: json1.type.invertWithDoc(op, this.drawingManagerData),
			redo: op,
			unitId,
			subUnitId,
			objects: {
				...orderParams,
				drawingIds: newIds
			}
		};
	}
	getBackDrawingsOp(orderParams) {
		const { unitId, subUnitId, drawingIds } = orderParams;
		const orderSearchParams = this._getOrderFromSearchParams(unitId, subUnitId, drawingIds, true);
		const newIds = [...drawingIds];
		const orders = this.getDrawingOrder(unitId, subUnitId);
		const ops = [];
		orderSearchParams.forEach((orderSearchParam) => {
			const { drawingId } = orderSearchParam;
			const op = json1.moveOp([
				unitId,
				subUnitId,
				"order",
				this._getDrawingOrder(unitId, subUnitId).indexOf(drawingId)
			], [
				unitId,
				subUnitId,
				"order",
				0
			]);
			ops.push(op);
			if (!newIds.includes(orders[0])) newIds.push(orders[0]);
		});
		const op = ops.reduce(json1.type.compose, null);
		return {
			undo: json1.type.invertWithDoc(op, this.drawingManagerData),
			redo: op,
			unitId,
			subUnitId,
			objects: {
				...orderParams,
				drawingIds: newIds
			}
		};
	}
	_getDrawingCount(unitId, subUnitId) {
		return this.getDrawingOrder(unitId, subUnitId).length || 0;
	}
	_getOrderFromSearchParams(unitId, subUnitId, drawingIds, isDesc = false) {
		return drawingIds.map((drawingId) => {
			return {
				drawingId,
				zIndex: this._hasDrawingOrder({
					unitId,
					subUnitId,
					drawingId
				})
			};
		}).sort(isDesc === false ? sortRules : sortRulesByDesc);
	}
	_hasDrawingOrder(searchParam) {
		if (searchParam == null) return -1;
		const { unitId, subUnitId, drawingId } = searchParam;
		this._establishDrawingMap(unitId, subUnitId);
		return this._getDrawingOrder(unitId, subUnitId).indexOf(drawingId);
	}
	_getCurrentBySearch(searchParam) {
		var _this$drawingManagerD;
		if (searchParam == null) return;
		const { unitId, subUnitId, drawingId } = searchParam;
		return (_this$drawingManagerD = this.drawingManagerData[unitId]) === null || _this$drawingManagerD === void 0 || (_this$drawingManagerD = _this$drawingManagerD[subUnitId]) === null || _this$drawingManagerD === void 0 || (_this$drawingManagerD = _this$drawingManagerD.data) === null || _this$drawingManagerD === void 0 ? void 0 : _this$drawingManagerD[drawingId];
	}
	_getOldBySearch(searchParam) {
		var _this$_oldDrawingMana;
		if (searchParam == null) return;
		const { unitId, subUnitId, drawingId } = searchParam;
		return (_this$_oldDrawingMana = this._oldDrawingManagerData[unitId]) === null || _this$_oldDrawingMana === void 0 || (_this$_oldDrawingMana = _this$_oldDrawingMana[subUnitId]) === null || _this$_oldDrawingMana === void 0 || (_this$_oldDrawingMana = _this$_oldDrawingMana.data) === null || _this$_oldDrawingMana === void 0 ? void 0 : _this$_oldDrawingMana[drawingId];
	}
	_establishDrawingMap(unitId, subUnitId, drawingId) {
		var _this$drawingManagerD2;
		if (!this.drawingManagerData[unitId]) this.drawingManagerData[unitId] = {};
		if (!this.drawingManagerData[unitId][subUnitId]) this.drawingManagerData[unitId][subUnitId] = {
			data: {},
			order: []
		};
		if (drawingId == null) return null;
		return (_this$drawingManagerD2 = this.drawingManagerData[unitId][subUnitId].data) === null || _this$drawingManagerD2 === void 0 ? void 0 : _this$drawingManagerD2[drawingId];
	}
	_addByParam(insertParam) {
		const { unitId, subUnitId, drawingId } = insertParam;
		this._establishDrawingMap(unitId, subUnitId, drawingId);
		const op = [json1.insertOp([
			unitId,
			subUnitId,
			"data",
			drawingId
		], insertParam), json1.insertOp([
			unitId,
			subUnitId,
			"order",
			this._getDrawingOrder(unitId, subUnitId).length
		], drawingId)].reduce(json1.type.compose, null);
		return {
			op,
			invertOp: json1.type.invertWithDoc(op, this.drawingManagerData)
		};
	}
	_removeByParam(searchParam) {
		if (searchParam == null) return {
			op: [],
			invertOp: []
		};
		const { unitId, subUnitId, drawingId } = searchParam;
		if (this._establishDrawingMap(unitId, subUnitId, drawingId) == null) return {
			op: [],
			invertOp: []
		};
		const op = [json1.removeOp([
			unitId,
			subUnitId,
			"data",
			drawingId
		], true), json1.removeOp([
			unitId,
			subUnitId,
			"order",
			this._getDrawingOrder(unitId, subUnitId).indexOf(drawingId)
		], true)].reduce(json1.type.compose, null);
		return {
			op,
			invertOp: json1.type.invertWithDoc(op, this.drawingManagerData)
		};
	}
	_updateByParam(updateParam) {
		const { unitId, subUnitId, drawingId } = updateParam;
		const object = this._establishDrawingMap(unitId, subUnitId, drawingId);
		if (object == null) return {
			op: [],
			invertOp: []
		};
		const op = this._getUpdateParamCompareOp(updateParam, object).reduce(json1.type.compose, null);
		return {
			op,
			invertOp: json1.type.invertWithDoc(op, this.drawingManagerData)
		};
	}
	_getUpdateParamCompareOp(newParam, oldParam) {
		const { unitId, subUnitId, drawingId } = newParam;
		const ops = [];
		Object.keys(newParam).forEach((key) => {
			const newVal = newParam[key];
			const oldVal = oldParam[key];
			if (oldVal === newVal) return;
			ops.push(json1.replaceOp([
				unitId,
				subUnitId,
				"data",
				drawingId,
				key
			], oldVal, newVal));
		});
		return ops;
	}
	_getDrawingData(unitId, subUnitId) {
		var _this$drawingManagerD3;
		return ((_this$drawingManagerD3 = this.drawingManagerData[unitId]) === null || _this$drawingManagerD3 === void 0 || (_this$drawingManagerD3 = _this$drawingManagerD3[subUnitId]) === null || _this$drawingManagerD3 === void 0 ? void 0 : _this$drawingManagerD3.data) || {};
	}
	_getDrawingOrder(unitId, subUnitId) {
		var _this$drawingManagerD4;
		return ((_this$drawingManagerD4 = this.drawingManagerData[unitId]) === null || _this$drawingManagerD4 === void 0 || (_this$drawingManagerD4 = _this$drawingManagerD4[subUnitId]) === null || _this$drawingManagerD4 === void 0 ? void 0 : _this$drawingManagerD4.order) || [];
	}
	getDrawingVisible() {
		return this._visible;
	}
	getDrawingEditable() {
		return this._editable;
	}
	setDrawingVisible(visible) {
		this._visible = visible;
	}
	setDrawingEditable(editable) {
		this._editable = editable;
	}
};
var DrawingManagerService = class extends UnitDrawingService {};

//#endregion
//#region src/services/image-io-impl.service.ts
var ImageIoService = class {
	constructor() {
		_defineProperty(this, "_waitCount", 0);
		_defineProperty(this, "_change$", new Subject());
		_defineProperty(this, "change$", this._change$);
		_defineProperty(this, "_imageSourceCache", /* @__PURE__ */ new Map());
	}
	setWaitCount(count) {
		this._waitCount = count;
		this._change$.next(count);
	}
	getImageSourceCache(source, imageSourceType) {
		if (imageSourceType === ImageSourceType.BASE64) {
			const image = new Image();
			image.src = source;
			return image;
		}
		return this._imageSourceCache.get(source);
	}
	addImageSourceCache(source, imageSourceType, imageSource) {
		if (imageSourceType === ImageSourceType.BASE64 || imageSource == null) return;
		this._imageSourceCache.set(source, imageSource);
	}
	async getImage(imageId) {
		return Promise.resolve(imageId);
	}
	async saveImage(imageFile) {
		return new Promise((resolve, reject) => {
			if (!DRAWING_IMAGE_ALLOW_IMAGE_LIST.includes(imageFile.type)) {
				reject(new Error(ImageUploadStatusType.ERROR_IMAGE_TYPE));
				this._decreaseWaiting();
				return;
			}
			if (imageFile.size > getDrawingImageAllowSize()) {
				reject(new Error(ImageUploadStatusType.ERROR_EXCEED_SIZE));
				this._decreaseWaiting();
				return;
			}
			const reader = new FileReader();
			reader.readAsDataURL(imageFile);
			reader.onload = (evt) => {
				var _evt$target;
				const replaceSrc = (_evt$target = evt.target) === null || _evt$target === void 0 ? void 0 : _evt$target.result;
				if (replaceSrc == null) {
					reject(new Error(ImageUploadStatusType.ERROR_IMAGE));
					this._decreaseWaiting();
					return;
				}
				resolve({
					imageId: generateRandomId(6),
					imageSourceType: ImageSourceType.BASE64,
					source: replaceSrc,
					base64Cache: replaceSrc,
					status: ImageUploadStatusType.SUCCUSS
				});
				this._decreaseWaiting();
			};
		});
	}
	_decreaseWaiting() {
		this._waitCount -= 1;
		this._change$.next(this._waitCount);
	}
};

//#endregion
//#region src/services/url-image.service.ts
var URLImageService = class extends Disposable {
	constructor(..._args) {
		super(..._args);
		_defineProperty(this, "_urlImageDownloader", null);
	}
	registerURLImageDownloader(downloader) {
		this._urlImageDownloader = downloader;
		return toDisposable(() => {
			this._urlImageDownloader = null;
		});
	}
	async getImage(url) {
		if (this._urlImageDownloader) try {
			return await this._urlImageDownloader(url);
		} catch (error) {
			console.error(`Custom downloader failed for ${url}, falling back to default behavior:`, error);
		}
		return url;
	}
	async downloadImage(url) {
		if (this._urlImageDownloader) try {
			const base64 = await this._urlImageDownloader(url);
			return await (await fetch(base64)).blob();
		} catch (error) {
			console.error(`Custom downloader failed for ${url}, falling back to default fetch:`, error);
		}
		return await (await fetch(url)).blob();
	}
};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/plugin.ts
let UniverDrawingPlugin = class UniverDrawingPlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _injector, _configService, _commandService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		this._commandService = _commandService;
		const { ...rest } = merge({}, defaultPluginConfig, this._config);
		this._configService.setConfig(DRAWING_PLUGIN_CONFIG_KEY, rest);
		if (rest.allowImageSize !== void 0) setDrawingImageAllowSize(rest.allowImageSize);
	}
	onStarting() {
		this._initCommands();
		this._initDependencies();
	}
	_initDependencies() {
		var _this$_config;
		mergeOverrideWithDependencies([
			[IImageIoService, { useClass: ImageIoService }],
			[IURLImageService, { useClass: URLImageService }],
			[IDrawingManagerService, { useClass: DrawingManagerService }]
		], (_this$_config = this._config) === null || _this$_config === void 0 ? void 0 : _this$_config.override).forEach((d) => this._injector.add(d));
	}
	_initCommands() {
		[SetDrawingSelectedOperation].forEach((command) => this.disposeWithMe(this._commandService.registerCommand(command)));
	}
};
_defineProperty(UniverDrawingPlugin, "pluginName", "UNIVER_DRAWING_PLUGIN");
_defineProperty(UniverDrawingPlugin, "packageName", name);
_defineProperty(UniverDrawingPlugin, "version", version);
UniverDrawingPlugin = __decorate([
	__decorateParam(1, Inject(Injector)),
	__decorateParam(2, IConfigService),
	__decorateParam(3, ICommandService)
], UniverDrawingPlugin);

//#endregion
//#region src/utils/get-image-shape-key.ts
function getDrawingShapeKeyByDrawingSearch({ unitId, subUnitId, drawingId }, index) {
	return typeof index === "number" ? `${unitId}#-#${subUnitId}#-#${drawingId}#-#${index}` : `${unitId}#-#${subUnitId}#-#${drawingId}`;
}

//#endregion
//#region src/utils/get-image-size.ts
const getImageSize = async (src) => {
	return new Promise((resolve, reject) => {
		const image = new Image();
		image.src = src;
		image.onload = () => {
			resolve({
				width: image.width,
				height: image.height,
				image
			});
		};
		image.onerror = (error) => {
			reject(error);
		};
	});
};

//#endregion
export { DRAWING_IMAGE_ALLOW_IMAGE_LIST, DRAWING_IMAGE_COUNT_LIMIT, DRAWING_IMAGE_HEIGHT_LIMIT, DRAWING_IMAGE_WIDTH_LIMIT, DrawingManagerService, IDrawingManagerService, IImageIoService, ImageIoService, ImageSourceType, ImageUploadStatusType, SetDrawingSelectedOperation, URLImageService, UnitDrawingService, UniverDrawingPlugin, getDrawingImageAllowSize, getDrawingShapeKeyByDrawingSearch, getImageSize };
