// @univerjs/sheets-note/lib/facade.js
import { FEventName, FUniver } from "@univerjs/core/facade";
import { RemoveNoteMutation, SheetDeleteNoteCommand, SheetToggleNotePopupCommand, SheetUpdateNoteCommand, SheetsNoteModel, UpdateNoteMutation } from "@univerjs/sheets-note";
import { FRange, FWorksheet } from "@univerjs/sheets/facade";
import { CanceledError, ICommandService, IUniverInstanceService } from "@univerjs/core";
import { SheetsSelectionsService, getSheetCommandTarget } from "@univerjs/sheets";

//#region src/facade/f-event.ts
/**
* @ignore
*/
var FSheetsNoteEventNameMixin = class extends FEventName {
	get SheetNoteAdd() {
		return "SheetNoteAdd";
	}
	get SheetNoteDelete() {
		return "SheetNoteDelete";
	}
	get SheetNoteUpdate() {
		return "SheetNoteUpdate";
	}
	get SheetNoteShow() {
		return "SheetNoteShow";
	}
	get SheetNoteHide() {
		return "SheetNoteHide";
	}
	get BeforeSheetNoteAdd() {
		return "BeforeSheetNoteAdd";
	}
	get BeforeSheetNoteDelete() {
		return "BeforeSheetNoteDelete";
	}
	get BeforeSheetNoteUpdate() {
		return "BeforeSheetNoteUpdate";
	}
	get BeforeSheetNoteShow() {
		return "BeforeSheetNoteShow";
	}
	get BeforeSheetNoteHide() {
		return "BeforeSheetNoteHide";
	}
};
FEventName.extend(FSheetsNoteEventNameMixin);

//#endregion
//#region src/facade/f-range.ts
var FRangeSheetsNoteMixin = class extends FRange {
	createOrUpdateNote(note) {
		this._commandService.syncExecuteCommand(UpdateNoteMutation.id, {
			unitId: this.getUnitId(),
			sheetId: this.getSheetId(),
			row: this.getRow(),
			col: this.getColumn(),
			note
		});
		return this;
	}
	deleteNote() {
		this._commandService.syncExecuteCommand(RemoveNoteMutation.id, {
			unitId: this.getUnitId(),
			sheetId: this.getSheetId(),
			row: this.getRow(),
			col: this.getColumn()
		});
		return this;
	}
	getNote() {
		return this._injector.get(SheetsNoteModel).getNote(this.getUnitId(), this.getSheetId(), {
			row: this.getRow(),
			col: this.getColumn()
		});
	}
};
FRange.extend(FRangeSheetsNoteMixin);

//#endregion
//#region src/facade/f-univer.ts
var FUniverSheetsNoteMixin = class extends FUniver {
	_initialize(injector) {
		const commandService = injector.get(ICommandService);
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetNoteAdd, () => {
			return injector.get(SheetsNoteModel).change$.subscribe((change) => {
				if (change.type === "update" && !change.oldNote && change.newNote) {
					const { unitId, subUnitId, newNote } = change;
					const target = this.getSheetTarget(unitId, subUnitId);
					if (!target) return;
					const { workbook, worksheet } = target;
					const eventParams = {
						workbook,
						worksheet,
						row: newNote.row,
						col: newNote.col,
						note: newNote
					};
					this.fireEvent(this.Event.SheetNoteAdd, eventParams);
				}
			});
		}));
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetNoteDelete, () => {
			return injector.get(SheetsNoteModel).change$.subscribe((change) => {
				if (change.type === "update" && change.oldNote && !change.newNote) {
					const { unitId, subUnitId, oldNote } = change;
					const target = this.getSheetTarget(unitId, subUnitId);
					if (!target) return;
					const { workbook, worksheet } = target;
					const eventParams = {
						workbook,
						worksheet,
						row: oldNote.row,
						col: oldNote.col,
						oldNote
					};
					this.fireEvent(this.Event.SheetNoteDelete, eventParams);
				}
			});
		}));
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetNoteUpdate, () => {
			return injector.get(SheetsNoteModel).change$.subscribe((change) => {
				if (change.type === "update" && change.oldNote && change.newNote) {
					const { unitId, subUnitId, oldNote, newNote } = change;
					const target = this.getSheetTarget(unitId, subUnitId);
					if (!target) return;
					const { workbook, worksheet } = target;
					const eventParams = {
						workbook,
						worksheet,
						row: newNote.row,
						col: newNote.col,
						note: newNote,
						oldNote
					};
					this.fireEvent(this.Event.SheetNoteUpdate, eventParams);
				}
			});
		}));
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetNoteShow, () => {
			return injector.get(SheetsNoteModel).change$.subscribe((change) => {
				if (change.type === "update" && change.oldNote && change.newNote && !change.oldNote.show && change.newNote.show) {
					const { unitId, subUnitId, newNote } = change;
					const target = this.getSheetTarget(unitId, subUnitId);
					if (!target) return;
					const { workbook, worksheet } = target;
					const eventParams = {
						workbook,
						worksheet,
						row: newNote.row,
						col: newNote.col
					};
					this.fireEvent(this.Event.SheetNoteShow, eventParams);
				}
			});
		}));
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetNoteHide, () => {
			return injector.get(SheetsNoteModel).change$.subscribe((change) => {
				if (change.type === "update" && change.oldNote && change.newNote && change.oldNote.show && !change.newNote.show) {
					const { unitId, subUnitId, newNote } = change;
					const target = this.getSheetTarget(unitId, subUnitId);
					if (!target) return;
					const { workbook, worksheet } = target;
					const eventParams = {
						workbook,
						worksheet,
						row: newNote.row,
						col: newNote.col
					};
					this.fireEvent(this.Event.SheetNoteHide, eventParams);
				}
			});
		}));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetNoteAdd, () => commandService.beforeCommandExecuted((command) => {
			if (command.id === SheetUpdateNoteCommand.id) {
				const { unitId, sheetId, row, col, note } = command.params;
				const target = this.getSheetTarget(unitId, sheetId);
				if (!target) return;
				if (injector.get(SheetsNoteModel).getNote(unitId, sheetId, {
					noteId: note.id,
					row,
					col
				})) return;
				const { workbook, worksheet } = target;
				const eventParams = {
					workbook,
					worksheet,
					row,
					col,
					note
				};
				if (this.fireEvent(this.Event.BeforeSheetNoteAdd, eventParams)) throw new CanceledError();
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetNoteDelete, () => commandService.beforeCommandExecuted((command) => {
			if (command.id === SheetDeleteNoteCommand.id) {
				const { unitId, sheetId, row, col } = command.params;
				const target = this.getSheetTarget(unitId, sheetId);
				if (!target) return;
				if (row === void 0 || col === void 0) return;
				const oldNote = injector.get(SheetsNoteModel).getNote(unitId, sheetId, {
					row,
					col
				});
				if (!oldNote) return;
				const { workbook, worksheet } = target;
				const eventParams = {
					workbook,
					worksheet,
					row,
					col,
					oldNote
				};
				if (this.fireEvent(this.Event.BeforeSheetNoteDelete, eventParams)) throw new CanceledError();
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetNoteUpdate, () => commandService.beforeCommandExecuted((command) => {
			if (command.id === SheetUpdateNoteCommand.id) {
				const { unitId, sheetId, row, col, note } = command.params;
				const target = this.getSheetTarget(unitId, sheetId);
				if (!target) return;
				const oldNote = injector.get(SheetsNoteModel).getNote(unitId, sheetId, {
					row,
					col
				});
				if (!oldNote) return;
				const { workbook, worksheet } = target;
				const eventParams = {
					workbook,
					worksheet,
					row,
					col,
					note,
					oldNote
				};
				if (this.fireEvent(this.Event.BeforeSheetNoteUpdate, eventParams)) throw new CanceledError();
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetNoteShow, () => commandService.beforeCommandExecuted((command) => {
			if (command.id === SheetToggleNotePopupCommand.id) {
				const target = getSheetCommandTarget(injector.get(IUniverInstanceService));
				if (!target) return;
				const { unitId, subUnitId } = target;
				const workbook = this.getUniverSheet(unitId);
				if (!workbook) return;
				const worksheet = workbook.getSheetBySheetId(subUnitId);
				if (!worksheet) return;
				const selection = injector.get(SheetsSelectionsService).getCurrentLastSelection();
				if (!(selection === null || selection === void 0 ? void 0 : selection.primary)) return;
				const sheetsNoteModel = injector.get(SheetsNoteModel);
				const { actualColumn, actualRow } = selection.primary;
				const note = sheetsNoteModel.getNote(unitId, subUnitId, {
					row: actualRow,
					col: actualColumn
				});
				if (!note || note.show) return;
				const eventParams = {
					workbook,
					worksheet,
					row: actualRow,
					col: actualColumn
				};
				if (this.fireEvent(this.Event.BeforeSheetNoteShow, eventParams)) throw new CanceledError();
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeSheetNoteHide, () => commandService.beforeCommandExecuted((command) => {
			if (command.id === SheetToggleNotePopupCommand.id) {
				const target = getSheetCommandTarget(injector.get(IUniverInstanceService));
				if (!target) return;
				const { unitId, subUnitId } = target;
				const workbook = this.getUniverSheet(unitId);
				if (!workbook) return;
				const worksheet = workbook.getSheetBySheetId(subUnitId);
				if (!worksheet) return;
				const selection = injector.get(SheetsSelectionsService).getCurrentLastSelection();
				if (!(selection === null || selection === void 0 ? void 0 : selection.primary)) return;
				const sheetsNoteModel = injector.get(SheetsNoteModel);
				const { actualColumn, actualRow } = selection.primary;
				const note = sheetsNoteModel.getNote(unitId, subUnitId, {
					row: actualRow,
					col: actualColumn
				});
				if (!note || !note.show) return;
				const eventParams = {
					workbook,
					worksheet,
					row: actualRow,
					col: actualColumn
				};
				if (this.fireEvent(this.Event.BeforeSheetNoteHide, eventParams)) throw new CanceledError();
			}
		})));
	}
};
FUniver.extend(FUniverSheetsNoteMixin);

//#endregion
//#region src/facade/f-worksheet.ts
var FWorksheetNoteMixin = class extends FWorksheet {
	getNotes() {
		const notes = this._injector.get(SheetsNoteModel).getSheetNotes(this.getWorkbook().getUnitId(), this.getSheetId());
		if (!notes) return [];
		return Array.from(notes.values()).map((note) => ({ ...note }));
	}
};
FWorksheet.extend(FWorksheetNoteMixin);

//#endregion
export { FRangeSheetsNoteMixin, FSheetsNoteEventNameMixin, FUniverSheetsNoteMixin, FWorksheetNoteMixin };
