// @univerjs/sheets-note/lib/index.js
import { CommandType, DependentOn, Disposable, ICommandService, IConfigService, IResourceManagerService, IUndoRedoService, IUniverInstanceService, Inject, Injector, Plugin, UniverInstanceType, generateRandomId, merge, sequenceExecuteAsync, touchDependencies } from "@univerjs/core";
import { CopySheetCommand, RefRangeService, RemoveSheetCommand, SheetInterceptorService, SheetsSelectionsService, UniverSheetsPlugin, getSheetCommandTarget, handleCommonRangeChangeWithEffectRefCommandsSkipNoInterests } from "@univerjs/sheets";
import { Subject, filter, map } from "rxjs";

//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region src/models/sheets-note.model.ts
var SheetsNoteModel = class extends Disposable {
	constructor(..._args) {
		super(..._args);
		_defineProperty(this, "_notesMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_change$", new Subject());
		_defineProperty(this, "change$", this._change$.asObservable());
	}
	_ensureNotesMap(unitId, subUnitId) {
		let unitMap = this._notesMap.get(unitId);
		if (!unitMap) {
			unitMap = /* @__PURE__ */ new Map();
			this._notesMap.set(unitId, unitMap);
		}
		let subUnitMap = unitMap.get(subUnitId);
		if (!subUnitMap) {
			subUnitMap = /* @__PURE__ */ new Map();
			unitMap.set(subUnitId, subUnitMap);
		}
		return subUnitMap;
	}
	_getNoteByPosition(unitId, subUnitId, row, col) {
		const subUnitMap = this._ensureNotesMap(unitId, subUnitId);
		for (const [_, note] of subUnitMap) if (note.row === row && note.col === col) return note;
	}
	_getNoteById(unitId, subUnitId, id) {
		return this._ensureNotesMap(unitId, subUnitId).get(id);
	}
	_getNoteByParams(unitId, subUnitId, params) {
		const { noteId, row, col } = params;
		if (noteId) return this._getNoteById(unitId, subUnitId, noteId);
		if (row !== void 0 && col !== void 0) return this._getNoteByPosition(unitId, subUnitId, row, col);
		return null;
	}
	getSheetShowNotes$(unitId, subUnitId) {
		return this._change$.pipe(filter(({ unitId: u, subUnitId: s }) => u === unitId && s === subUnitId), map(() => {
			const subUnitMap = this._ensureNotesMap(unitId, subUnitId);
			const notes = [];
			for (const [_, note] of subUnitMap) if (note.show) notes.push({
				loc: {
					row: note.row,
					col: note.col,
					unitId,
					subUnitId
				},
				note
			});
			return notes;
		}));
	}
	getCellNoteChange$(unitId, subUnitId, row, col) {
		return this._change$.pipe(filter(({ unitId: u, subUnitId: s, oldNote }) => {
			if (u !== unitId || s !== subUnitId || !oldNote) return false;
			return oldNote.row === row && oldNote.col === col;
		}), map((newNote) => newNote));
	}
	updateNote(unitId, subUnitId, row, col, note, silent) {
		const oldNote = this._getNoteByParams(unitId, subUnitId, {
			noteId: note === null || note === void 0 ? void 0 : note.id,
			row,
			col
		});
		const subUnitMap = this._ensureNotesMap(unitId, subUnitId);
		const newNote = {
			...note,
			id: (oldNote === null || oldNote === void 0 ? void 0 : oldNote.id) || note.id || generateRandomId(6),
			row,
			col
		};
		subUnitMap.set(newNote.id, newNote);
		this._change$.next({
			unitId,
			subUnitId,
			oldNote,
			type: "update",
			newNote,
			silent
		});
	}
	removeNote(unitId, subUnitId, params) {
		const { noteId, row, col, silent } = params;
		const oldNote = this._getNoteByParams(unitId, subUnitId, {
			noteId,
			row,
			col
		});
		if (!oldNote) return;
		this._ensureNotesMap(unitId, subUnitId).delete(oldNote.id);
		this._change$.next({
			unitId,
			subUnitId,
			oldNote,
			type: "update",
			newNote: null,
			silent
		});
	}
	toggleNotePopup(unitId, subUnitId, params) {
		const { noteId, row, col, silent } = params;
		const oldNote = this._getNoteByParams(unitId, subUnitId, {
			noteId,
			row,
			col
		});
		if (!oldNote) return;
		const subUnitMap = this._ensureNotesMap(unitId, subUnitId);
		const newNote = {
			...oldNote,
			show: !oldNote.show
		};
		subUnitMap.set(newNote.id, newNote);
		this._change$.next({
			unitId,
			subUnitId,
			oldNote,
			type: "update",
			newNote,
			silent
		});
	}
	updateNotePosition(unitId, subUnitId, params) {
		const { noteId, row, col, newRow, newCol, silent } = params;
		const oldNote = this._getNoteByParams(unitId, subUnitId, {
			noteId,
			row,
			col
		});
		if (!oldNote) return;
		const subUnitMap = this._ensureNotesMap(unitId, subUnitId);
		const newNote = {
			...oldNote,
			row: newRow,
			col: newCol
		};
		subUnitMap.set(newNote.id, newNote);
		this._change$.next({
			unitId,
			subUnitId,
			oldNote,
			type: "ref",
			newNote,
			silent
		});
	}
	getNote(unitId, subUnitId, params) {
		return this._getNoteByParams(unitId, subUnitId, params);
	}
	getNotes() {
		return this._notesMap;
	}
	getUnitNotes(unitId) {
		return this._notesMap.get(unitId);
	}
	getSheetNotes(unitId, subUnitId) {
		const unitMap = this._notesMap.get(unitId);
		if (!unitMap) return;
		return unitMap.get(subUnitId);
	}
	deleteUnitNotes(unitId) {
		this._notesMap.delete(unitId);
	}
};

//#endregion
//#region src/commands/mutations/note.mutation.ts
const UpdateNoteMutation = {
	id: "sheet.mutation.update-note",
	type: CommandType.MUTATION,
	handler: (accessor, params) => {
		const { unitId, sheetId, row, col, note, silent } = params;
		accessor.get(SheetsNoteModel).updateNote(unitId, sheetId, row, col, note, silent);
		return true;
	}
};
const RemoveNoteMutation = {
	id: "sheet.mutation.remove-note",
	type: CommandType.MUTATION,
	handler: (accessor, params) => {
		const { unitId, sheetId, noteId, row, col, silent } = params;
		accessor.get(SheetsNoteModel).removeNote(unitId, sheetId, {
			noteId,
			row,
			col,
			silent
		});
		return true;
	}
};
const ToggleNotePopupMutation = {
	id: "sheet.mutation.toggle-note-popup",
	type: CommandType.MUTATION,
	handler: (accessor, params) => {
		const { unitId, sheetId, noteId, row, col, silent } = params;
		accessor.get(SheetsNoteModel).toggleNotePopup(unitId, sheetId, {
			noteId,
			row,
			col,
			silent
		});
		return true;
	}
};
const UpdateNotePositionMutation = {
	id: "sheet.mutation.update-note-position",
	type: CommandType.MUTATION,
	handler: (accessor, params) => {
		const { unitId, sheetId, noteId, row, col, newPosition, silent } = params;
		accessor.get(SheetsNoteModel).updateNotePosition(unitId, sheetId, {
			noteId,
			row,
			col,
			newRow: newPosition.row,
			newCol: newPosition.col,
			silent
		});
		return true;
	}
};

//#endregion
//#region src/commands/commands/note.command.ts
const SheetDeleteNoteCommand = {
	id: "sheet.command.delete-note",
	type: CommandType.COMMAND,
	handler: (accessor) => {
		const target = getSheetCommandTarget(accessor.get(IUniverInstanceService));
		if (!target) return false;
		const selection = accessor.get(SheetsSelectionsService).getCurrentLastSelection();
		if (!(selection === null || selection === void 0 ? void 0 : selection.primary)) return false;
		const sheetsNoteModel = accessor.get(SheetsNoteModel);
		const { unitId, subUnitId } = target;
		const { actualColumn, actualRow } = selection.primary;
		const note = sheetsNoteModel.getNote(unitId, subUnitId, {
			row: actualRow,
			col: actualColumn
		});
		if (!note) return false;
		const commandService = accessor.get(ICommandService);
		const undoRedoService = accessor.get(IUndoRedoService);
		const redoMutation = {
			id: RemoveNoteMutation.id,
			params: {
				unitId,
				sheetId: subUnitId,
				noteId: note.id
			}
		};
		const undoMutation = {
			id: UpdateNoteMutation.id,
			params: {
				unitId,
				sheetId: subUnitId,
				row: actualRow,
				col: actualColumn,
				note: { ...note }
			}
		};
		if (commandService.syncExecuteCommand(redoMutation.id, redoMutation.params)) {
			undoRedoService.pushUndoRedo({
				unitID: unitId,
				redoMutations: [redoMutation],
				undoMutations: [undoMutation]
			});
			return true;
		}
		return false;
	}
};
const SheetToggleNotePopupCommand = {
	id: "sheet.command.toggle-note-popup",
	type: CommandType.COMMAND,
	handler: (accessor) => {
		const target = getSheetCommandTarget(accessor.get(IUniverInstanceService));
		if (!target) return false;
		const selection = accessor.get(SheetsSelectionsService).getCurrentLastSelection();
		if (!(selection === null || selection === void 0 ? void 0 : selection.primary)) return false;
		const sheetsNoteModel = accessor.get(SheetsNoteModel);
		const { unitId, subUnitId } = target;
		const { actualColumn, actualRow } = selection.primary;
		const note = sheetsNoteModel.getNote(unitId, subUnitId, {
			row: actualRow,
			col: actualColumn
		});
		if (!note) return false;
		const commandService = accessor.get(ICommandService);
		const undoRedoService = accessor.get(IUndoRedoService);
		const redoMutation = {
			id: ToggleNotePopupMutation.id,
			params: {
				unitId,
				sheetId: subUnitId,
				noteId: note.id
			}
		};
		const undoMutation = {
			id: ToggleNotePopupMutation.id,
			params: {
				unitId,
				sheetId: subUnitId,
				noteId: note.id
			}
		};
		if (commandService.syncExecuteCommand(redoMutation.id, redoMutation.params)) {
			undoRedoService.pushUndoRedo({
				unitID: unitId,
				redoMutations: [redoMutation],
				undoMutations: [undoMutation]
			});
			return true;
		}
		return false;
	}
};
const SheetUpdateNoteCommand = {
	id: "sheet.command.update-note",
	type: CommandType.COMMAND,
	handler: (accessor, params) => {
		const target = getSheetCommandTarget(accessor.get(IUniverInstanceService), params);
		if (!target) return false;
		const commandService = accessor.get(ICommandService);
		const undoRedoService = accessor.get(IUndoRedoService);
		const sheetsNoteModel = accessor.get(SheetsNoteModel);
		const { unitId, subUnitId } = target;
		const { row, col, note } = params;
		const oldNote = sheetsNoteModel.getNote(unitId, subUnitId, {
			noteId: note.id,
			row,
			col
		});
		const redoMutation = {
			id: UpdateNoteMutation.id,
			params: {
				unitId,
				sheetId: subUnitId,
				row,
				col,
				note
			}
		};
		const undoMutations = [];
		if (oldNote) {
			const undoMutation = {
				id: UpdateNoteMutation.id,
				params: {
					unitId,
					sheetId: subUnitId,
					row,
					col,
					note: { ...oldNote }
				}
			};
			undoMutations.push(undoMutation);
		} else {
			const undoMutation = {
				id: RemoveNoteMutation.id,
				params: {
					unitId,
					sheetId: subUnitId,
					row,
					col
				}
			};
			undoMutations.push(undoMutation);
		}
		if (commandService.syncExecuteCommand(redoMutation.id, redoMutation.params)) {
			undoRedoService.pushUndoRedo({
				unitID: unitId,
				redoMutations: [redoMutation],
				undoMutations
			});
			return true;
		}
		return false;
	}
};

//#endregion
//#region src/const.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const PLUGIN_NAME = "SHEET_NOTE_PLUGIN";

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/controllers/sheets-note-resource.controller.ts
let SheetsNoteResourceController = class SheetsNoteResourceController extends Disposable {
	constructor(_resourceManagerService, _univerInstanceService, _sheetInterceptorService, _sheetsNoteModel) {
		super();
		this._resourceManagerService = _resourceManagerService;
		this._univerInstanceService = _univerInstanceService;
		this._sheetInterceptorService = _sheetInterceptorService;
		this._sheetsNoteModel = _sheetsNoteModel;
		this._initSnapshot();
		this._initSheetChange();
	}
	_initSnapshot() {
		const toJson = (unitId) => {
			const unitMap = this._sheetsNoteModel.getUnitNotes(unitId);
			if (!unitMap) return "";
			const result = {};
			unitMap.forEach((subUnitMap, subUnitId) => {
				const sheetNotes = {};
				subUnitMap.forEach((note) => {
					const { row, col } = note;
					if (!sheetNotes[row]) sheetNotes[row] = {};
					sheetNotes[row][col] = note;
				});
				if (Object.keys(sheetNotes).length > 0) result[subUnitId] = sheetNotes;
			});
			return JSON.stringify(result);
		};
		const parseJson = (json) => {
			if (!json) return {};
			try {
				return JSON.parse(json);
			} catch {
				return {};
			}
		};
		this.disposeWithMe(this._resourceManagerService.registerPluginResource({
			pluginName: PLUGIN_NAME,
			businesses: [UniverInstanceType.UNIVER_SHEET],
			toJson: (unitId) => toJson(unitId),
			parseJson: (json) => parseJson(json),
			onUnLoad: (unitId) => {
				this._sheetsNoteModel.deleteUnitNotes(unitId);
			},
			onLoad: (unitId, value) => {
				Object.entries(value).forEach(([sheetId, sheetNotes]) => {
					Object.entries(sheetNotes).forEach(([row, colNotes]) => {
						Object.entries(colNotes).forEach(([col, note]) => {
							this._sheetsNoteModel.updateNote(unitId, sheetId, Number(row), Number(col), note);
						});
					});
				});
			}
		}));
	}
	_initSheetChange() {
		this.disposeWithMe(this._sheetInterceptorService.interceptCommand({ getMutations: (commandInfo) => {
			if (commandInfo.id === RemoveSheetCommand.id) {
				var _getActiveSheet;
				const params = commandInfo.params;
				const unitId = params.unitId || this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_SHEET).getUnitId();
				const subUnitId = params.subUnitId || ((_getActiveSheet = this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_SHEET).getActiveSheet()) === null || _getActiveSheet === void 0 ? void 0 : _getActiveSheet.getSheetId());
				if (!unitId || !subUnitId) return {
					redos: [],
					undos: []
				};
				const sheetNotes = this._sheetsNoteModel.getSheetNotes(unitId, subUnitId);
				if (!sheetNotes) return {
					redos: [],
					undos: []
				};
				const redos = [];
				const undos = [];
				sheetNotes.forEach((note) => {
					redos.push({
						id: RemoveNoteMutation.id,
						params: {
							unitId,
							sheetId: subUnitId,
							noteId: note.id,
							row: note.row,
							col: note.col
						}
					});
					undos.push({
						id: UpdateNoteMutation.id,
						params: {
							unitId,
							sheetId: subUnitId,
							row: note.row,
							col: note.col,
							note
						}
					});
				});
				return {
					redos,
					undos
				};
			} else if (commandInfo.id === CopySheetCommand.id) {
				const { unitId, subUnitId, targetSubUnitId } = commandInfo.params;
				if (!unitId || !subUnitId || !targetSubUnitId) return {
					redos: [],
					undos: []
				};
				const sheetNotes = this._sheetsNoteModel.getSheetNotes(unitId, subUnitId);
				if (!sheetNotes) return {
					redos: [],
					undos: []
				};
				const redos = [];
				const undos = [];
				sheetNotes.forEach((note) => {
					const newNote = {
						...note,
						id: generateRandomId(6)
					};
					redos.push({
						id: UpdateNoteMutation.id,
						params: {
							unitId,
							sheetId: targetSubUnitId,
							row: newNote.row,
							col: newNote.col,
							note: newNote
						}
					});
					undos.push({
						id: RemoveNoteMutation.id,
						params: {
							unitId,
							sheetId: targetSubUnitId,
							noteId: newNote.id,
							row: newNote.row,
							col: newNote.col
						}
					});
				});
				return {
					redos,
					undos
				};
			}
			return {
				redos: [],
				undos: []
			};
		} }));
	}
};
SheetsNoteResourceController = __decorate([
	__decorateParam(0, IResourceManagerService),
	__decorateParam(1, IUniverInstanceService),
	__decorateParam(2, Inject(SheetInterceptorService)),
	__decorateParam(3, Inject(SheetsNoteModel))
], SheetsNoteResourceController);

//#endregion
//#region package.json
var name = "@univerjs/sheets-note";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const SHEETS_NOTE_PLUGIN_CONFIG_KEY = "sheets-note.config";
const configSymbol = Symbol(SHEETS_NOTE_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/controllers/sheets-note-ref-range.controller.ts
let SheetsNoteRefRangeController = class SheetsNoteRefRangeController extends Disposable {
	constructor(_refRangeService, _sheetsNoteModel, _selectionManagerService, _commandService) {
		super();
		this._refRangeService = _refRangeService;
		this._sheetsNoteModel = _sheetsNoteModel;
		this._selectionManagerService = _selectionManagerService;
		this._commandService = _commandService;
		_defineProperty(this, "_disposableMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_watcherMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_handleRangeChange", (unitId, subUnitId, note, row, col, resultRange, silent) => {
			if (!resultRange) return {
				redos: [{
					id: RemoveNoteMutation.id,
					params: {
						unitId,
						sheetId: subUnitId,
						noteId: note.id
					}
				}],
				undos: [{
					id: UpdateNoteMutation.id,
					params: {
						unitId,
						sheetId: subUnitId,
						row,
						col,
						note
					}
				}]
			};
			return {
				redos: [{
					id: UpdateNotePositionMutation.id,
					params: {
						unitId,
						sheetId: subUnitId,
						noteId: note.id,
						newPosition: {
							row: resultRange.startRow,
							col: resultRange.startColumn
						},
						silent
					}
				}],
				undos: [{
					id: UpdateNotePositionMutation.id,
					params: {
						unitId,
						sheetId: subUnitId,
						noteId: note.id,
						newPosition: {
							row,
							col
						},
						note,
						silent
					}
				}]
			};
		});
		this._initData();
		this._initRefRange();
	}
	_getIdWithUnitId(unitId, subUnitId, row, col) {
		return `${unitId}-${subUnitId}-${row}-${col}`;
	}
	_register(unitId, subUnitId, note, row, col) {
		const oldRange = {
			startColumn: col,
			endColumn: col,
			startRow: row,
			endRow: row
		};
		this._disposableMap.set(this._getIdWithUnitId(unitId, subUnitId, row, col), this._refRangeService.registerRefRange(oldRange, (commandInfo) => {
			const resultRanges = handleCommonRangeChangeWithEffectRefCommandsSkipNoInterests(oldRange, commandInfo, { selectionManagerService: this._selectionManagerService });
			const resultRange = Array.isArray(resultRanges) ? resultRanges[0] : resultRanges;
			if (resultRange && resultRange.startColumn === oldRange.startColumn && resultRange.startRow === oldRange.startRow) return {
				undos: [],
				redos: []
			};
			return this._handleRangeChange(unitId, subUnitId, note, row, col, resultRange, false);
		}, unitId, subUnitId));
	}
	_watch(unitId, subUnitId, note, row, col) {
		const oldRange = {
			startColumn: col,
			endColumn: col,
			startRow: row,
			endRow: row
		};
		this._watcherMap.set(this._getIdWithUnitId(unitId, subUnitId, row, col), this._refRangeService.watchRange(unitId, subUnitId, oldRange, (before, after) => {
			const { redos } = this._handleRangeChange(unitId, subUnitId, note, before.startRow, before.startColumn, after, true);
			sequenceExecuteAsync(redos, this._commandService, { onlyLocal: true });
		}, true));
	}
	_unwatch(unitId, subUnitId, row, col) {
		var _this$_watcherMap$get;
		const id = this._getIdWithUnitId(unitId, subUnitId, row, col);
		(_this$_watcherMap$get = this._watcherMap.get(id)) === null || _this$_watcherMap$get === void 0 || _this$_watcherMap$get.dispose();
		this._watcherMap.delete(id);
	}
	_unregister(unitId, subUnitId, row, col) {
		var _this$_disposableMap$;
		const id = this._getIdWithUnitId(unitId, subUnitId, row, col);
		(_this$_disposableMap$ = this._disposableMap.get(id)) === null || _this$_disposableMap$ === void 0 || _this$_disposableMap$.dispose();
		this._disposableMap.delete(id);
	}
	_initData() {
		const unitNotes = this._sheetsNoteModel.getNotes();
		for (const [unitId, unitNote] of unitNotes) for (const [subUnitId, notes] of unitNote) notes.forEach((note) => {
			this._register(unitId, subUnitId, note, note.row, note.col);
			this._watch(unitId, subUnitId, note, note.row, note.col);
		});
	}
	_initRefRange() {
		this.disposeWithMe(this._sheetsNoteModel.change$.subscribe((option) => {
			switch (option.type) {
				case "update": {
					const { unitId, subUnitId, oldNote, newNote } = option;
					const row = newNote ? newNote.row : oldNote.row;
					const col = newNote ? newNote.col : oldNote.col;
					const id = this._getIdWithUnitId(unitId, subUnitId, row, col);
					if (newNote) {
						if (!this._disposableMap.has(id)) {
							this._register(unitId, subUnitId, newNote, row, col);
							this._watch(unitId, subUnitId, newNote, row, col);
						}
					} else {
						this._unregister(unitId, subUnitId, row, col);
						this._unwatch(unitId, subUnitId, row, col);
					}
					break;
				}
				case "ref": {
					const { unitId, subUnitId, oldNote, newNote, silent } = option;
					const { row: oldRow, col: oldCol } = oldNote;
					const { row: newRow, col: newCol } = newNote;
					this._unregister(unitId, subUnitId, oldRow, oldCol);
					if (!silent) {
						this._unwatch(unitId, subUnitId, oldRow, oldCol);
						this._watch(unitId, subUnitId, newNote, newRow, newCol);
					}
					this._register(unitId, subUnitId, newNote, newRow, newCol);
					break;
				}
			}
		}));
	}
};
SheetsNoteRefRangeController = __decorate([
	__decorateParam(0, Inject(RefRangeService)),
	__decorateParam(1, Inject(SheetsNoteModel)),
	__decorateParam(2, Inject(SheetsSelectionsService)),
	__decorateParam(3, ICommandService)
], SheetsNoteRefRangeController);

//#endregion
//#region src/controllers/sheets.note.controller.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
let SheetsNoteController = class SheetsNoteController extends Disposable {
	constructor(_commandService) {
		super();
		this._commandService = _commandService;
		this._initialize();
	}
	_initialize() {
		[
			UpdateNotePositionMutation,
			ToggleNotePopupMutation,
			UpdateNoteMutation,
			RemoveNoteMutation,
			SheetDeleteNoteCommand,
			SheetToggleNotePopupCommand,
			SheetUpdateNoteCommand
		].forEach((command) => {
			this.disposeWithMe(this._commandService.registerCommand(command));
		});
	}
};
SheetsNoteController = __decorate([__decorateParam(0, ICommandService)], SheetsNoteController);

//#endregion
//#region src/plugin.ts
let UniverSheetsNotePlugin = class UniverSheetsNotePlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _configService, _injector) {
		super();
		this._config = _config;
		this._configService = _configService;
		this._injector = _injector;
		const { ...rest } = merge({}, defaultPluginConfig, this._config);
		this._configService.setConfig(SHEETS_NOTE_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		[
			[SheetsNoteModel],
			[SheetsNoteController],
			[SheetsNoteResourceController],
			[SheetsNoteRefRangeController]
		].forEach((dependency) => {
			this._injector.add(dependency);
		});
		touchDependencies(this._injector, [
			[SheetsNoteModel],
			[SheetsNoteController],
			[SheetsNoteResourceController]
		]);
	}
	onReady() {
		touchDependencies(this._injector, [[SheetsNoteRefRangeController]]);
	}
};
_defineProperty(UniverSheetsNotePlugin, "pluginName", PLUGIN_NAME);
_defineProperty(UniverSheetsNotePlugin, "packageName", name);
_defineProperty(UniverSheetsNotePlugin, "version", version);
_defineProperty(UniverSheetsNotePlugin, "type", UniverInstanceType.UNIVER_SHEET);
UniverSheetsNotePlugin = __decorate([
	DependentOn(UniverSheetsPlugin),
	__decorateParam(1, IConfigService),
	__decorateParam(2, Inject(Injector))
], UniverSheetsNotePlugin);

//#endregion
export { RemoveNoteMutation, SheetDeleteNoteCommand, SheetToggleNotePopupCommand, SheetUpdateNoteCommand, SheetsNoteModel, SheetsNoteResourceController, ToggleNotePopupMutation, UniverSheetsNotePlugin, UpdateNoteMutation, UpdateNotePositionMutation };
