// @univerjs/sheets-hyper-link/lib/index.js
import { BuildTextUtils, CellValueType, CommandType, CustomRangeType, DOCS_NORMAL_EDITOR_UNIT_ID_KEY, DependentOn, Disposable, DisposableCollection, ICommandService, IConfigService, IUndoRedoService, IUniverInstanceService, Inject, Injector, LocaleService, ObjectMatrix, Plugin, Range, Rectangle, TextX, Tools, UniverInstanceType, generateRandomId, getBodySlice, isValidRange, merge, registerDependencies, sequenceExecute, sequenceExecuteAsync, toDisposable, touchDependencies } from "@univerjs/core";
import { addCustomRangeBySelectionFactory, deleteCustomRangeFactory, replaceSelectionFactory } from "@univerjs/docs";
import { AFTER_CELL_EDIT, ClearSelectionAllCommand, ClearSelectionContentCommand, ClearSelectionFormatCommand, RefRangeService, RemoveSheetCommand, SetRangeValuesCommand, SetRangeValuesMutation, SetRangeValuesUndoMutationFactory, SheetInterceptorService, SheetsSelectionsService, UniverSheetsPlugin, getSheetCommandTarget, handleCommonRangeChangeWithEffectRefCommandsSkipNoInterests, handleDefaultRangeChangeWithEffectRefCommands, handleDefaultRangeChangeWithEffectRefCommandsSkipNoInterests } from "@univerjs/sheets";
import { Subject } from "rxjs";
import { IDefinedNamesService, deserializeRangeWithSheet, serializeRange, serializeRangeWithSheet } from "@univerjs/engine-formula";

//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/models/hyper-link.model.ts
let HyperLinkModel = class HyperLinkModel extends Disposable {
	constructor(_univerInstanceService) {
		super();
		this._univerInstanceService = _univerInstanceService;
		_defineProperty(this, "_linkUpdate$", new Subject());
		_defineProperty(this, "linkUpdate$", this._linkUpdate$.asObservable());
		_defineProperty(this, "_linkMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_linkPositionMap", /* @__PURE__ */ new Map());
		this.disposeWithMe({ dispose: () => {
			this._linkUpdate$.complete();
		} });
	}
	_ensureMap(unitId, subUnitId) {
		let unitMap = this._linkMap.get(unitId);
		if (!unitMap) {
			unitMap = /* @__PURE__ */ new Map();
			this._linkMap.set(unitId, unitMap);
		}
		let matrix = unitMap.get(subUnitId);
		if (!matrix) {
			matrix = new ObjectMatrix();
			unitMap.set(subUnitId, matrix);
		}
		let positionUnitMap = this._linkPositionMap.get(unitId);
		if (!positionUnitMap) {
			positionUnitMap = /* @__PURE__ */ new Map();
			this._linkPositionMap.set(unitId, positionUnitMap);
		}
		let positionSubUnitMap = positionUnitMap.get(subUnitId);
		if (!positionSubUnitMap) {
			positionSubUnitMap = /* @__PURE__ */ new Map();
			positionUnitMap.set(subUnitId, positionSubUnitMap);
		}
		return {
			matrix,
			positionMap: positionSubUnitMap
		};
	}
	addHyperLink(unitId, subUnitId, link) {
		const { matrix, positionMap } = this._ensureMap(unitId, subUnitId);
		matrix.setValue(link.row, link.column, link);
		positionMap.set(link.id, {
			row: link.row,
			column: link.column,
			link
		});
		this._linkUpdate$.next({
			unitId,
			subUnitId,
			payload: link,
			type: "add"
		});
		return true;
	}
	updateHyperLink(unitId, subUnitId, id, payload, silent = false) {
		const { matrix, positionMap } = this._ensureMap(unitId, subUnitId);
		const position = positionMap.get(id);
		if (!position) return true;
		const link = matrix.getValue(position.row, position.column);
		if (!link) return true;
		Object.assign(link, payload);
		this._linkUpdate$.next({
			unitId,
			subUnitId,
			payload: {
				display: link.display,
				payload: link.payload
			},
			id,
			type: "update",
			silent
		});
		return true;
	}
	updateHyperLinkRef(unitId, subUnitId, id, payload, silent = false) {
		const { matrix, positionMap } = this._ensureMap(unitId, subUnitId);
		const position = positionMap.get(id);
		if (!position) return true;
		let link = matrix.getValue(position.row, position.column);
		if (!link || link.id !== id) link = position.link;
		else matrix.realDeleteValue(position.row, position.column);
		Object.assign(link, payload);
		positionMap.set(id, {
			...payload,
			link
		});
		matrix.setValue(payload.row, payload.column, link);
		this._linkUpdate$.next({
			unitId,
			subUnitId,
			payload,
			id,
			type: "updateRef",
			silent
		});
		return true;
	}
	removeHyperLink(unitId, subUnitId, id) {
		const { matrix, positionMap } = this._ensureMap(unitId, subUnitId);
		const position = positionMap.get(id);
		if (!position) return false;
		positionMap.delete(id);
		const link = matrix.getValue(position.row, position.column);
		if (link && link.id === id) matrix.realDeleteValue(position.row, position.column);
		this._linkUpdate$.next({
			unitId,
			subUnitId,
			payload: position.link,
			type: "remove"
		});
		return true;
	}
	getHyperLink(unitId, subUnitId, id) {
		const { matrix, positionMap } = this._ensureMap(unitId, subUnitId);
		const position = positionMap.get(id);
		if (!position) return;
		return matrix.getValue(position.row, position.column);
	}
	getHyperLinkByLocation(unitId, subUnitId, row, column) {
		const { matrix } = this._ensureMap(unitId, subUnitId);
		return matrix.getValue(row, column);
	}
	getHyperLinkByLocationSync(unitId, subUnitId, row, column) {
		var _workbook$getSheetByS, _ref, _cell$v, _cell$p;
		const { matrix } = this._ensureMap(unitId, subUnitId);
		const workbook = this._univerInstanceService.getUnit(unitId, UniverInstanceType.UNIVER_SHEET);
		const cell = workbook === null || workbook === void 0 || (_workbook$getSheetByS = workbook.getSheetBySheetId(subUnitId)) === null || _workbook$getSheetByS === void 0 ? void 0 : _workbook$getSheetByS.getCellRaw(row, column);
		const cellValueStr = ((_ref = (_cell$v = cell === null || cell === void 0 ? void 0 : cell.v) !== null && _cell$v !== void 0 ? _cell$v : cell === null || cell === void 0 || (_cell$p = cell.p) === null || _cell$p === void 0 || (_cell$p = _cell$p.body) === null || _cell$p === void 0 ? void 0 : _cell$p.dataStream.slice(0, -2)) !== null && _ref !== void 0 ? _ref : "").toString();
		const link = matrix.getValue(row, column);
		if (!link) return;
		return {
			...link,
			display: cellValueStr
		};
	}
	getSubUnit(unitId, subUnitId) {
		const { matrix } = this._ensureMap(unitId, subUnitId);
		const links = [];
		matrix.forValue((row, col, value) => {
			if (value) links.push(value);
		});
		return links;
	}
	getUnit(unitId) {
		const unitMap = this._linkMap.get(unitId);
		if (!unitMap) return [];
		return Array.from(unitMap.keys()).map((subUnitId) => {
			return {
				unitId,
				subUnitId,
				links: this.getSubUnit(unitId, subUnitId)
			};
		});
	}
	deleteUnit(unitId) {
		const links = this.getUnit(unitId);
		this._linkMap.delete(unitId);
		this._linkPositionMap.delete(unitId);
		this._linkUpdate$.next({
			type: "unload",
			unitId,
			unitLinks: links
		});
	}
	getAll() {
		return Array.from(this._linkMap.keys()).map((unitId) => this.getUnit(unitId));
	}
};
HyperLinkModel = __decorate([__decorateParam(0, IUniverInstanceService)], HyperLinkModel);

//#endregion
//#region src/commands/mutations/add-hyper-link.mutation.ts
const AddHyperLinkMutation = {
	type: CommandType.MUTATION,
	id: "sheets.mutation.add-hyper-link",
	handler(accessor, params) {
		if (!params) return false;
		const model = accessor.get(HyperLinkModel);
		const { unitId, subUnitId, link } = params;
		return model.addHyperLink(unitId, subUnitId, link);
	}
};

//#endregion
//#region src/commands/mutations/remove-hyper-link.mutation.ts
const RemoveHyperLinkMutation = {
	type: CommandType.MUTATION,
	id: "sheets.mutation.remove-hyper-link",
	handler(accessor, params) {
		if (!params) return false;
		const model = accessor.get(HyperLinkModel);
		const { unitId, subUnitId, id } = params;
		return model.removeHyperLink(unitId, subUnitId, id);
	}
};

//#endregion
//#region src/commands/commands/add-hyper-link.command.ts
/**
* Command for add hyperlink
*/
const AddHyperLinkCommand = {
	type: CommandType.COMMAND,
	id: "sheets.command.add-hyper-link",
	async handler(accessor, params) {
		if (!params) return false;
		const commandService = accessor.get(ICommandService);
		const undoRedoService = accessor.get(IUndoRedoService);
		const univerInstanceService = accessor.get(IUniverInstanceService);
		const hyperLinkModel = accessor.get(HyperLinkModel);
		const sheetInterceptorService = accessor.get(SheetInterceptorService);
		const target = getSheetCommandTarget(univerInstanceService, params);
		if (!target) return false;
		const { unitId, subUnitId, workbook, worksheet } = target;
		const { link } = params;
		const { payload, display, row, column, id } = link;
		const cellData = worksheet.getCell(row, column);
		const doc = worksheet.getBlankCellDocumentModel(cellData, row, column);
		const snapshot = doc.documentModel.getSnapshot();
		const body = Tools.deepClone(snapshot.body);
		if (!body) return false;
		let textX;
		if (display) textX = BuildTextUtils.selection.replace({
			selection: {
				startOffset: 0,
				endOffset: body.dataStream.length - 2,
				collapsed: body.dataStream.length - 2 === 0
			},
			body: {
				dataStream: `${display}`,
				customRanges: [{
					startIndex: 0,
					endIndex: display.length - 1,
					rangeType: CustomRangeType.HYPERLINK,
					rangeId: id,
					properties: { url: payload }
				}]
			},
			doc: doc.documentModel
		});
		else textX = BuildTextUtils.customRange.add({
			body,
			ranges: [{
				startOffset: 0,
				endOffset: body.dataStream.length - 2,
				collapsed: false
			}],
			rangeId: id,
			rangeType: CustomRangeType.HYPERLINK,
			properties: {
				url: payload,
				refId: id
			}
		});
		if (!textX) return false;
		const newBody = TextX.apply(body, textX.serialize());
		const newCellData = {
			p: {
				...snapshot,
				body: newBody
			},
			t: CellValueType.STRING
		};
		const finalCellData = sheetInterceptorService.onWriteCell(workbook, worksheet, row, column, newCellData);
		const redoParams = {
			unitId,
			subUnitId,
			cellValue: { [link.row]: { [link.column]: finalCellData } }
		};
		const redo = {
			id: SetRangeValuesMutation.id,
			params: redoParams
		};
		const undoParams = SetRangeValuesUndoMutationFactory(accessor, redoParams);
		const undo = {
			id: SetRangeValuesMutation.id,
			params: undoParams
		};
		const redos = [redo];
		const undos = [undo];
		const modelLink = hyperLinkModel.getHyperLinkByLocation(unitId, subUnitId, row, column);
		if (modelLink) {
			redos.push({
				id: RemoveHyperLinkMutation.id,
				params: {
					unitId,
					subUnitId,
					id: modelLink.id
				}
			});
			undos.push({
				id: AddHyperLinkMutation.id,
				params: {
					unitId,
					subUnitId,
					link: modelLink
				}
			});
		}
		if (await sequenceExecute(redos, commandService)) {
			if (await sheetInterceptorService.onValidateCell(workbook, worksheet, row, column) === false) {
				sequenceExecute(undos, commandService);
				return false;
			}
			undoRedoService.pushUndoRedo({
				redoMutations: redos,
				undoMutations: undos,
				unitID: unitId
			});
			return true;
		}
		return false;
	}
};
const AddRichHyperLinkCommand = {
	id: "sheets.command.add-rich-hyper-link",
	type: CommandType.COMMAND,
	handler: async (accessor, params) => {
		if (!params) return false;
		const { documentId, link } = params;
		const commandService = accessor.get(ICommandService);
		const newId = generateRandomId();
		const { payload } = link;
		const replaceSelection = addCustomRangeBySelectionFactory(accessor, {
			unitId: documentId,
			rangeId: newId,
			rangeType: CustomRangeType.HYPERLINK,
			properties: {
				url: payload,
				refId: newId
			}
		});
		if (replaceSelection) return commandService.syncExecuteCommand(replaceSelection.id, replaceSelection.params);
		return false;
	}
};

//#endregion
//#region src/commands/commands/remove-hyper-link.command.ts
const CancelHyperLinkCommand = {
	type: CommandType.COMMAND,
	id: "sheets.command.cancel-hyper-link",
	handler(accessor, params) {
		var _snapshot$body;
		if (!params) return false;
		const commandService = accessor.get(ICommandService);
		const undoRedoService = accessor.get(IUndoRedoService);
		const instanceSrv = accessor.get(IUniverInstanceService);
		const hyperLinkModel = accessor.get(HyperLinkModel);
		const target = getSheetCommandTarget(instanceSrv, params);
		if (!target) return false;
		const { row, column, id } = params;
		const { unitId, subUnitId, worksheet } = target;
		const cellData = worksheet.getCell(row, column);
		if (!cellData) return false;
		const doc = worksheet.getCellDocumentModelWithFormula(cellData, row, column);
		if (!(doc === null || doc === void 0 ? void 0 : doc.documentModel)) return false;
		const snapshot = Tools.deepClone(doc.documentModel.getSnapshot());
		const range = (_snapshot$body = snapshot.body) === null || _snapshot$body === void 0 || (_snapshot$body = _snapshot$body.customRanges) === null || _snapshot$body === void 0 ? void 0 : _snapshot$body.find((range) => `${range.rangeId}` === id);
		if (!range) return false;
		const textX = BuildTextUtils.customRange.delete({
			documentDataModel: doc.documentModel,
			rangeId: range.rangeId
		});
		if (!textX) return false;
		const newBody = TextX.apply(snapshot.body, textX.serialize());
		const redos = [];
		const undos = [];
		const setRangeParams = {
			unitId,
			subUnitId,
			cellValue: { [row]: { [column]: {
				p: {
					...snapshot,
					body: newBody
				},
				t: CellValueType.STRING
			} } }
		};
		redos.push({
			id: SetRangeValuesMutation.id,
			params: setRangeParams
		});
		const undoParams = SetRangeValuesUndoMutationFactory(accessor, setRangeParams);
		undos.push({
			id: SetRangeValuesMutation.id,
			params: undoParams
		});
		const link = hyperLinkModel.getHyperLinkByLocation(unitId, subUnitId, row, column);
		if (link) {
			redos.push({
				id: RemoveHyperLinkMutation.id,
				params: {
					unitId,
					subUnitId,
					id: link.id
				}
			});
			undos.push({
				id: AddHyperLinkMutation.id,
				params: {
					unitId,
					subUnitId,
					link: { ...link }
				}
			});
		}
		if (sequenceExecute(redos, commandService).result) {
			undoRedoService.pushUndoRedo({
				redoMutations: redos,
				undoMutations: undos,
				unitID: unitId
			});
			return true;
		}
		return false;
	}
};
const CancelRichHyperLinkCommand = {
	type: CommandType.COMMAND,
	id: "sheets.command.cancel-rich-hyper-link",
	handler(accessor, params) {
		var _doc$getBody;
		if (!params) return false;
		const { id: linkId, documentId } = params;
		const commandService = accessor.get(ICommandService);
		const doc = accessor.get(IUniverInstanceService).getUnit(documentId, UniverInstanceType.UNIVER_DOC);
		const link = doc === null || doc === void 0 || (_doc$getBody = doc.getBody()) === null || _doc$getBody === void 0 || (_doc$getBody = _doc$getBody.customRanges) === null || _doc$getBody === void 0 ? void 0 : _doc$getBody.find((i) => i.rangeId === linkId);
		let insert = null;
		if (link && link.endIndex === doc.getBody().dataStream.length - 3) insert = { dataStream: " " };
		const doMutation = deleteCustomRangeFactory(accessor, {
			unitId: documentId,
			rangeId: linkId,
			insert
		});
		if (!doMutation) return false;
		return commandService.syncExecuteCommand(doMutation.id, doMutation.params);
	}
};

//#endregion
//#region src/commands/commands/update-hyper-link.command.ts
const UpdateHyperLinkCommand = {
	type: CommandType.COMMAND,
	id: "sheets.command.update-hyper-link",
	async handler(accessor, params) {
		var _snapshot$body, _oldBody$textRuns;
		if (!params) return false;
		const commandService = accessor.get(ICommandService);
		const undoRedoService = accessor.get(IUndoRedoService);
		const instanceSrv = accessor.get(IUniverInstanceService);
		const hyperLinkModel = accessor.get(HyperLinkModel);
		const interceptorService = accessor.get(SheetInterceptorService);
		const target = getSheetCommandTarget(instanceSrv, {
			unitId: params.unitId,
			subUnitId: params.subUnitId
		});
		if (!target) return false;
		const { payload: link, row, column, id } = params;
		const { workbook, worksheet, unitId, subUnitId } = target;
		const { payload, display = "" } = link;
		const cellData = worksheet.getCell(row, column);
		if (!cellData) return false;
		const doc = worksheet.getCellDocumentModelWithFormula(cellData, row, column);
		if (!(doc === null || doc === void 0 ? void 0 : doc.documentModel)) return false;
		const snapshot = doc.documentModel.getSnapshot();
		const range = (_snapshot$body = snapshot.body) === null || _snapshot$body === void 0 || (_snapshot$body = _snapshot$body.customRanges) === null || _snapshot$body === void 0 ? void 0 : _snapshot$body.find((range) => `${range.rangeId}` === id);
		if (!range) return false;
		const newId = generateRandomId();
		const textRun = (_oldBody$textRuns = getBodySlice(doc.documentModel.getBody(), range.startIndex, range.endIndex + 1).textRuns) === null || _oldBody$textRuns === void 0 ? void 0 : _oldBody$textRuns[0];
		if (textRun) textRun.ed = display.length + 1;
		const replaceSelection = replaceSelectionFactory(accessor, {
			unitId,
			body: {
				dataStream: `${display}`,
				customRanges: [{
					rangeId: newId,
					rangeType: CustomRangeType.HYPERLINK,
					startIndex: 0,
					endIndex: display.length - 1,
					properties: { url: payload }
				}],
				textRuns: textRun ? [textRun] : void 0
			},
			selection: {
				startOffset: range.startIndex,
				endOffset: range.endIndex + 1,
				collapsed: false
			},
			doc: doc.documentModel
		});
		if (!replaceSelection) return false;
		const newBody = TextX.apply(Tools.deepClone(snapshot.body), replaceSelection.textX.serialize());
		const newCellData = {
			p: {
				...snapshot,
				body: newBody
			},
			t: CellValueType.STRING
		};
		const finalCellData = interceptorService.onWriteCell(workbook, worksheet, row, column, newCellData);
		const redo = {
			id: SetRangeValuesMutation.id,
			params: {
				unitId,
				subUnitId,
				cellValue: { [row]: { [column]: finalCellData } }
			}
		};
		const undoParams = SetRangeValuesUndoMutationFactory(accessor, redo.params);
		const undo = {
			id: SetRangeValuesMutation.id,
			params: undoParams
		};
		const redos = [redo];
		const undos = [undo];
		const modelLink = hyperLinkModel.getHyperLinkByLocation(unitId, subUnitId, row, column);
		if (modelLink) {
			redos.push({
				id: RemoveHyperLinkMutation.id,
				params: {
					unitId,
					subUnitId,
					id: modelLink.id
				}
			});
			undos.push({
				id: AddHyperLinkMutation.id,
				params: {
					unitId,
					subUnitId,
					link: modelLink
				}
			});
		}
		if (sequenceExecute(redos, commandService).result) {
			if (await interceptorService.onValidateCell(workbook, worksheet, row, column) === false) {
				sequenceExecute(undos, commandService);
				return false;
			}
			undoRedoService.pushUndoRedo({
				redoMutations: redos,
				undoMutations: undos,
				unitID: unitId
			});
			return true;
		}
		return false;
	}
};
const UpdateRichHyperLinkCommand = {
	type: CommandType.COMMAND,
	id: "sheets.command.update-rich-hyper-link",
	handler: (accessor, params) => {
		var _doc$getBody, _params$payload$displ, _oldBody$textRuns2;
		if (!params) return false;
		const { documentId: unitId, payload, id: rangeId } = params;
		const univerInstanceService = accessor.get(IUniverInstanceService);
		const commandService = accessor.get(ICommandService);
		const doc = univerInstanceService.getUnit(unitId, UniverInstanceType.UNIVER_DOC);
		if (!doc) return false;
		const range = (_doc$getBody = doc.getBody()) === null || _doc$getBody === void 0 || (_doc$getBody = _doc$getBody.customRanges) === null || _doc$getBody === void 0 ? void 0 : _doc$getBody.find((range) => range.rangeId === rangeId);
		if (!range) return false;
		const display = (_params$payload$displ = params.payload.display) !== null && _params$payload$displ !== void 0 ? _params$payload$displ : "";
		const newId = generateRandomId();
		const textRun = (_oldBody$textRuns2 = getBodySlice(doc.getBody(), range.startIndex, range.endIndex + 1).textRuns) === null || _oldBody$textRuns2 === void 0 ? void 0 : _oldBody$textRuns2[0];
		if (textRun) textRun.ed = display.length + 1;
		const replaceSelection = replaceSelectionFactory(accessor, {
			unitId,
			body: {
				dataStream: `${display}`,
				customRanges: [{
					rangeId: newId,
					rangeType: CustomRangeType.HYPERLINK,
					startIndex: 0,
					endIndex: display.length - 1,
					properties: { url: payload.payload }
				}],
				textRuns: textRun ? [textRun] : void 0
			},
			selection: {
				startOffset: range.startIndex,
				endOffset: range.endIndex + 1,
				collapsed: false
			},
			doc
		});
		if (!replaceSelection) return false;
		return commandService.syncExecuteCommand(replaceSelection.id, replaceSelection.params);
	}
};

//#endregion
//#region src/commands/mutations/update-hyper-link.mutation.ts
const UpdateHyperLinkMutation = {
	type: CommandType.MUTATION,
	id: "sheets.mutation.update-hyper-link",
	handler(accessor, params) {
		if (!params) return false;
		const model = accessor.get(HyperLinkModel);
		const { unitId, subUnitId, payload, id } = params;
		return model.updateHyperLink(unitId, subUnitId, id, payload, false);
	}
};
const UpdateHyperLinkRefMutation = {
	type: CommandType.MUTATION,
	id: "sheets.mutation.update-hyper-link-ref",
	handler(accessor, params) {
		if (!params) return false;
		const model = accessor.get(HyperLinkModel);
		const { unitId, subUnitId, id, row, column, silent } = params;
		return model.updateHyperLinkRef(unitId, subUnitId, id, {
			row,
			column
		}, silent);
	}
};
const UpdateRichHyperLinkMutation = {
	type: CommandType.MUTATION,
	id: "sheets.mutation.update-rich-hyper-link",
	handler(accessor, params) {
		var _cell$p;
		if (!params) return false;
		const { unitId, subUnitId, row, col, id, url } = params;
		const sheetTarget = getSheetCommandTarget(accessor.get(IUniverInstanceService), {
			unitId,
			subUnitId
		});
		if (!sheetTarget) return false;
		const { worksheet } = sheetTarget;
		const cell = worksheet.getCellRaw(row, col);
		const link = cell === null || cell === void 0 || (_cell$p = cell.p) === null || _cell$p === void 0 || (_cell$p = _cell$p.body) === null || _cell$p === void 0 || (_cell$p = _cell$p.customRanges) === null || _cell$p === void 0 ? void 0 : _cell$p.find((range) => range.rangeType === CustomRangeType.HYPERLINK && range.rangeId === id);
		if (!link) return true;
		link.properties.url = url;
		return true;
	}
};

//#endregion
//#region package.json
var name = "@univerjs/sheets-hyper-link";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const SHEETS_HYPER_LINK_PLUGIN_CONFIG_KEY = "sheets-hyper-link.config";
const configSymbol = Symbol(SHEETS_HYPER_LINK_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/types/const.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const SHEET_HYPER_LINK_PLUGIN = "SHEET_HYPER_LINK_PLUGIN";
const ERROR_RANGE = "err";

//#endregion
//#region src/controllers/ref-range.controller.ts
let SheetsHyperLinkRefRangeController = class SheetsHyperLinkRefRangeController extends Disposable {
	constructor(_refRangeService, _hyperLinkModel, _selectionManagerService, _commandService) {
		super();
		this._refRangeService = _refRangeService;
		this._hyperLinkModel = _hyperLinkModel;
		this._selectionManagerService = _selectionManagerService;
		this._commandService = _commandService;
		_defineProperty(this, "_disposableMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_watchDisposableMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_rangeDisableMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_rangeWatcherMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_handlePositionChange", (unitId, subUnitId, link, resultRange, silent) => {
			const oldRange = {
				startColumn: link.column,
				endColumn: link.column,
				startRow: link.row,
				endRow: link.row
			};
			if (!resultRange) return {
				redos: [{
					id: RemoveHyperLinkMutation.id,
					params: {
						unitId,
						subUnitId,
						id: link.id
					}
				}],
				undos: [{
					id: AddHyperLinkMutation.id,
					params: {
						unitId,
						subUnitId,
						link
					}
				}]
			};
			return {
				redos: [{
					id: UpdateHyperLinkRefMutation.id,
					params: {
						unitId,
						subUnitId,
						id: link.id,
						row: resultRange.startRow,
						column: resultRange.startColumn,
						silent
					}
				}],
				undos: [{
					id: UpdateHyperLinkRefMutation.id,
					params: {
						unitId,
						subUnitId,
						id: link.id,
						row: oldRange.startRow,
						column: oldRange.startColumn,
						silent
					}
				}]
			};
		});
		this._initData();
		this._initRefRange();
	}
	_registerPosition(unitId, subUnitId, link) {
		const id = link.id;
		const oldRange = {
			startColumn: link.column,
			endColumn: link.column,
			startRow: link.row,
			endRow: link.row
		};
		const handleRefRangeChange = (commandInfo) => {
			const resultRanges = handleCommonRangeChangeWithEffectRefCommandsSkipNoInterests(oldRange, commandInfo, { selectionManagerService: this._selectionManagerService });
			const resultRange = Array.isArray(resultRanges) ? resultRanges[0] : resultRanges;
			if (resultRange && resultRange.startColumn === oldRange.startColumn && resultRange.startRow === oldRange.startRow) return {
				undos: [],
				redos: []
			};
			return this._handlePositionChange(unitId, subUnitId, link, resultRange, false);
		};
		this._disposableMap.set(id, this._refRangeService.registerRefRange(oldRange, handleRefRangeChange, unitId, subUnitId));
	}
	_watchPosition(unitId, subUnitId, link) {
		const id = link.id;
		const oldRange = {
			startColumn: link.column,
			endColumn: link.column,
			startRow: link.row,
			endRow: link.row
		};
		this._watchDisposableMap.set(id, this._refRangeService.watchRange(unitId, subUnitId, oldRange, (before, after) => {
			const { redos } = this._handlePositionChange(unitId, subUnitId, link, after, true);
			sequenceExecuteAsync(redos, this._commandService, { onlyLocal: true });
		}, true));
	}
	_unregisterPosition(id) {
		const disposable = this._disposableMap.get(id);
		disposable === null || disposable === void 0 || disposable.dispose();
		this._disposableMap.delete(id);
	}
	_unwatchPosition(id) {
		const disposable = this._watchDisposableMap.get(id);
		disposable === null || disposable === void 0 || disposable.dispose();
		this._watchDisposableMap.delete(id);
	}
	_registerRange(unitId, id, payload, silent = false) {
		if (payload.startsWith("#")) {
			var _search$get, _search$get2, _search$get3;
			const search = new URLSearchParams(payload.slice(1));
			const searchObj = {
				gid: (_search$get = search.get("gid")) !== null && _search$get !== void 0 ? _search$get : "",
				range: (_search$get2 = search.get("range")) !== null && _search$get2 !== void 0 ? _search$get2 : "",
				rangeid: (_search$get3 = search.get("rangeid")) !== null && _search$get3 !== void 0 ? _search$get3 : ""
			};
			if (searchObj.range && searchObj.gid) {
				const subUnitId = searchObj.gid;
				const range = deserializeRangeWithSheet(searchObj.range).range;
				if (isValidRange(range) && searchObj.range !== "err") {
					const handleRangeChange = (commandInfo) => {
						const resultRange = handleDefaultRangeChangeWithEffectRefCommandsSkipNoInterests(range, commandInfo, { selectionManagerService: this._selectionManagerService });
						if (resultRange && serializeRange(resultRange) === serializeRange(range)) return {
							redos: [],
							undos: []
						};
						return {
							redos: [{
								id: UpdateHyperLinkMutation.id,
								params: {
									unitId,
									subUnitId,
									id,
									payload: { payload: `#gid=${subUnitId}&range=${resultRange ? serializeRange(resultRange) : "err"}` }
								}
							}],
							undos: [{
								id: UpdateHyperLinkMutation.id,
								params: {
									unitId,
									subUnitId,
									id,
									payload: { payload }
								}
							}]
						};
					};
					this._rangeDisableMap.set(id, this._refRangeService.registerRefRange(range, handleRangeChange, unitId, subUnitId));
					if (!silent) this._rangeWatcherMap.set(id, this._refRangeService.watchRange(unitId, subUnitId, range, (before, after) => {
						this._hyperLinkModel.updateHyperLink(unitId, subUnitId, id, { payload: `#gid=${subUnitId}&range=${after ? serializeRange(after) : "err"}` }, true);
					}, true));
				}
			}
		}
	}
	_unregisterRange(id) {
		const disposable = this._rangeDisableMap.get(id);
		disposable === null || disposable === void 0 || disposable.dispose();
		this._rangeDisableMap.delete(id);
	}
	_unwatchRange(id) {
		const disposable = this._rangeWatcherMap.get(id);
		disposable === null || disposable === void 0 || disposable.dispose();
		this._rangeWatcherMap.delete(id);
	}
	_initData() {
		this._hyperLinkModel.getAll().forEach((unitData) => {
			unitData.forEach((subUnitData) => {
				const { unitId, subUnitId, links } = subUnitData;
				links.forEach((link) => {
					this._registerPosition(unitId, subUnitId, link);
					this._watchPosition(unitId, subUnitId, link);
					this._registerRange(unitId, link.id, link.payload);
				});
			});
		});
	}
	_initRefRange() {
		this.disposeWithMe(this._hyperLinkModel.linkUpdate$.subscribe((option) => {
			switch (option.type) {
				case "add":
					this._registerPosition(option.unitId, option.subUnitId, option.payload);
					this._watchPosition(option.unitId, option.subUnitId, option.payload);
					this._registerRange(option.unitId, option.payload.id, option.payload.payload);
					break;
				case "remove":
					this._unregisterPosition(option.payload.id);
					this._unwatchPosition(option.payload.id);
					this._unregisterRange(option.payload.id);
					this._unwatchRange(option.payload.id);
					break;
				case "updateRef": {
					const { unitId, subUnitId, id, silent } = option;
					const link = this._hyperLinkModel.getHyperLink(unitId, subUnitId, id);
					if (!link) return;
					this._unregisterPosition(id);
					this._registerPosition(unitId, subUnitId, link);
					if (!silent) {
						this._unwatchPosition(id);
						this._watchPosition(unitId, subUnitId, link);
					}
					break;
				}
				case "unload": {
					const { unitLinks } = option;
					unitLinks.forEach((subUnitData) => {
						const { links } = subUnitData;
						links.forEach((link) => {
							this._unregisterPosition(link.id);
							this._unwatchPosition(link.id);
							this._unregisterRange(link.id);
							this._unwatchRange(link.id);
						});
					});
					break;
				}
				case "update":
					if (!option.silent) this._unwatchRange(option.id);
					this._unregisterRange(option.id);
					this._registerRange(option.unitId, option.id, option.payload.payload, option.silent);
					break;
			}
		}));
		this.disposeWithMe(toDisposable(() => {
			this._disposableMap.forEach((item) => {
				item.dispose();
			});
			this._disposableMap.clear();
		}));
	}
};
SheetsHyperLinkRefRangeController = __decorate([
	__decorateParam(0, Inject(RefRangeService)),
	__decorateParam(1, Inject(HyperLinkModel)),
	__decorateParam(2, Inject(SheetsSelectionsService)),
	__decorateParam(3, ICommandService)
], SheetsHyperLinkRefRangeController);

//#endregion
//#region src/controllers/remove-sheet.controller.ts
let SheetsHyperLinkRemoveSheetController = class SheetsHyperLinkRemoveSheetController extends Disposable {
	constructor(_sheetInterceptorService, _univerInstanceService, _hyperLinkModel) {
		super();
		this._sheetInterceptorService = _sheetInterceptorService;
		this._univerInstanceService = _univerInstanceService;
		this._hyperLinkModel = _hyperLinkModel;
		this._initSheetChange();
	}
	_initSheetChange() {
		this.disposeWithMe(this._sheetInterceptorService.interceptCommand({ getMutations: (commandInfo) => {
			if (commandInfo.id === RemoveSheetCommand.id) {
				var _workbook$getActiveSh;
				const params = commandInfo.params;
				const workbook = params.unitId ? this._univerInstanceService.getUnit(params.unitId) : this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_SHEET);
				if (!workbook) return {
					redos: [],
					undos: []
				};
				const unitId = workbook.getUnitId();
				const subUnitId = params.subUnitId || ((_workbook$getActiveSh = workbook.getActiveSheet()) === null || _workbook$getActiveSh === void 0 ? void 0 : _workbook$getActiveSh.getSheetId());
				if (!subUnitId) return {
					redos: [],
					undos: []
				};
				const links = this._hyperLinkModel.getSubUnit(unitId, subUnitId);
				return {
					redos: links.map((link) => ({
						id: RemoveHyperLinkMutation.id,
						params: {
							unitId,
							subUnitId,
							id: link.id
						}
					})),
					undos: links.map((link) => ({
						id: AddHyperLinkMutation.id,
						params: {
							unitId,
							subUnitId,
							link
						}
					}))
				};
			}
			return {
				redos: [],
				undos: []
			};
		} }));
	}
};
SheetsHyperLinkRemoveSheetController = __decorate([
	__decorateParam(0, Inject(SheetInterceptorService)),
	__decorateParam(1, IUniverInstanceService),
	__decorateParam(2, Inject(HyperLinkModel))
], SheetsHyperLinkRemoveSheetController);

//#endregion
//#region src/controllers/rich-text-ref-range.controller.ts
let SheetsHyperLinkRichTextRefRangeController = class SheetsHyperLinkRichTextRefRangeController extends Disposable {
	constructor(_commandService, _univerInstanceService, _refRangeService) {
		super();
		this._commandService = _commandService;
		this._univerInstanceService = _univerInstanceService;
		this._refRangeService = _refRangeService;
		_defineProperty(this, "_refRangeMap", /* @__PURE__ */ new Map());
		this._initWorkbookLoad();
		this._initWorkbookUnload();
		this._initSetRangesListener();
	}
	_enusreMap(unitId, subUnitId) {
		let unitMap = this._refRangeMap.get(unitId);
		if (!unitMap) {
			unitMap = /* @__PURE__ */ new Map();
			this._refRangeMap.set(unitId, unitMap);
		}
		let subUnitMap = unitMap.get(subUnitId);
		if (!subUnitMap) {
			subUnitMap = new ObjectMatrix();
			unitMap.set(subUnitId, subUnitMap);
		}
		return subUnitMap;
	}
	_isLegalRangeUrl(unitId, payload) {
		const workbook = this._univerInstanceService.getUnit(unitId, UniverInstanceType.UNIVER_SHEET);
		if (!workbook) return null;
		if (payload && payload.startsWith("#")) {
			var _search$get, _search$get2, _search$get3;
			const search = new URLSearchParams(payload.slice(1));
			const searchObj = {
				gid: (_search$get = search.get("gid")) !== null && _search$get !== void 0 ? _search$get : "",
				range: (_search$get2 = search.get("range")) !== null && _search$get2 !== void 0 ? _search$get2 : "",
				rangeid: (_search$get3 = search.get("rangeid")) !== null && _search$get3 !== void 0 ? _search$get3 : ""
			};
			if (searchObj.range && searchObj.gid) {
				const subUnitId = searchObj.gid;
				const worksheet = workbook.getSheetBySheetId(subUnitId);
				if (!worksheet) return null;
				const range = deserializeRangeWithSheet(searchObj.range).range;
				if (isValidRange(range, worksheet) && searchObj.range !== "err") return {
					range,
					worksheet
				};
			}
		}
		return null;
	}
	_registerRange(unitId, subUnitId, row, col, p) {
		var _p$body;
		const map = this._enusreMap(unitId, subUnitId);
		if ((_p$body = p.body) === null || _p$body === void 0 || (_p$body = _p$body.customRanges) === null || _p$body === void 0 ? void 0 : _p$body.some((customRange) => {
			var _customRange$properti;
			return customRange.rangeType === CustomRangeType.HYPERLINK && this._isLegalRangeUrl(unitId, (_customRange$properti = customRange.properties) === null || _customRange$properti === void 0 ? void 0 : _customRange$properti.url);
		})) {
			var _p$body2;
			const disposableCollection = new DisposableCollection();
			let hasWatch = false;
			(_p$body2 = p.body) === null || _p$body2 === void 0 || (_p$body2 = _p$body2.customRanges) === null || _p$body2 === void 0 || _p$body2.forEach((customRange) => {
				if (customRange.rangeType === CustomRangeType.HYPERLINK) {
					var _customRange$properti2;
					const payload = (_customRange$properti2 = customRange.properties) === null || _customRange$properti2 === void 0 ? void 0 : _customRange$properti2.url;
					const rangeInfo = this._isLegalRangeUrl(unitId, payload);
					if (rangeInfo) {
						const { range, worksheet } = rangeInfo;
						hasWatch = true;
						disposableCollection.add(this._refRangeService.registerRefRange(range, (commandInfo) => {
							const newRange = handleDefaultRangeChangeWithEffectRefCommands(range, commandInfo);
							if (newRange && Rectangle.equals(newRange, range)) return {
								preRedos: [],
								preUndos: [],
								redos: [],
								undos: []
							};
							return {
								preRedos: [{
									id: UpdateRichHyperLinkMutation.id,
									params: {
										unitId,
										subUnitId,
										row,
										col,
										id: customRange.rangeId,
										url: `#gid=${subUnitId}&range=${newRange ? serializeRange(newRange) : "err"}`
									}
								}],
								undos: [{
									id: UpdateRichHyperLinkMutation.id,
									params: {
										unitId,
										subUnitId,
										row,
										col,
										id: customRange.rangeId,
										url: payload
									}
								}],
								redos: []
							};
						}, worksheet.getUnitId(), worksheet.getSheetId()));
					}
				}
			});
			if (hasWatch) map.setValue(row, col, disposableCollection);
		}
	}
	_initWorkbookLoad() {
		const handleWorkbook = (workbook) => {
			const unitId = workbook.getUnitId();
			workbook.getSheets().forEach((sheet) => {
				const subUnitId = sheet.getSheetId();
				const map = this._enusreMap(unitId, subUnitId);
				sheet.getCellMatrix().forValue((row, col, cell) => {
					const dispose = map.getValue(row, col);
					if (dispose) dispose.dispose();
					if (cell && cell.p) this._registerRange(unitId, subUnitId, row, col, cell.p);
				});
			});
		};
		this._univerInstanceService.getAllUnitsForType(UniverInstanceType.UNIVER_SHEET).forEach((workbook) => {
			handleWorkbook(workbook);
		});
		this.disposeWithMe(this._univerInstanceService.unitAdded$.subscribe((event) => {
			const { unit } = event;
			if (unit.type === UniverInstanceType.UNIVER_SHEET) handleWorkbook(unit);
		}));
	}
	_initWorkbookUnload() {
		this.disposeWithMe(this._univerInstanceService.unitDisposed$.subscribe((unit) => {
			if (unit.type === UniverInstanceType.UNIVER_SHEET) {
				const workbook = unit;
				const unitId = workbook.getUnitId();
				workbook.getSheets().forEach((sheet) => {
					const subUnitId = sheet.getSheetId();
					this._enusreMap(unitId, subUnitId).forValue((row, col, dispose) => {
						if (dispose) dispose.dispose();
					});
				});
				this._refRangeMap.delete(unitId);
			}
		}));
	}
	_initSetRangesListener() {
		this.disposeWithMe(this._commandService.onCommandExecuted((commandInfo) => {
			if (commandInfo.id === SetRangeValuesMutation.id) {
				const { unitId, subUnitId, cellValue } = commandInfo.params;
				const map = this._enusreMap(unitId, subUnitId);
				if (cellValue) new ObjectMatrix(cellValue).forValue((row, col, cell) => {
					const dispose = map.getValue(row, col);
					if (dispose) dispose.dispose();
					if (cell && cell.p) this._registerRange(unitId, subUnitId, row, col, cell.p);
				});
			}
		}));
		this.disposeWithMe(this._commandService.onCommandExecuted((commandInfo) => {
			if (commandInfo.id === UpdateRichHyperLinkMutation.id) {
				const { unitId, subUnitId, row, col } = commandInfo.params;
				const sheetTarget = getSheetCommandTarget(this._univerInstanceService, {
					unitId,
					subUnitId
				});
				const dispose = this._enusreMap(unitId, subUnitId).getValue(row, col);
				if (dispose) dispose.dispose();
				if (sheetTarget) {
					const { worksheet } = sheetTarget;
					const cell = worksheet.getCellRaw(row, col);
					if (cell && cell.p) this._registerRange(unitId, subUnitId, row, col, cell.p);
				}
			}
		}));
	}
};
SheetsHyperLinkRichTextRefRangeController = __decorate([
	__decorateParam(0, ICommandService),
	__decorateParam(1, IUniverInstanceService),
	__decorateParam(2, Inject(RefRangeService))
], SheetsHyperLinkRichTextRefRangeController);

//#endregion
//#region src/controllers/set-range.controller.ts
let SheetHyperLinkSetRangeController = class SheetHyperLinkSetRangeController extends Disposable {
	constructor(_sheetInterceptorService, _hyperLinkModel, _selectionManagerService, _univerInstanceService) {
		super();
		this._sheetInterceptorService = _sheetInterceptorService;
		this._hyperLinkModel = _hyperLinkModel;
		this._selectionManagerService = _selectionManagerService;
		this._univerInstanceService = _univerInstanceService;
		this._initCommandInterceptor();
		this._initAfterEditor();
	}
	_initCommandInterceptor() {
		this._initSetRangeValuesCommandInterceptor();
		this._initClearSelectionCommandInterceptor();
	}
	_initSetRangeValuesCommandInterceptor() {
		this.disposeWithMe(this._sheetInterceptorService.interceptCommand({ getMutations: (command) => {
			if (command.id === SetRangeValuesCommand.id) {
				const params = command.params;
				const { unitId, subUnitId } = params;
				const redos = [];
				const undos = [];
				if (params.cellValue) new ObjectMatrix(params.cellValue).forValue((row, col) => {
					const link = this._hyperLinkModel.getHyperLinkByLocation(unitId, subUnitId, row, col);
					if (link) {
						redos.push({
							id: RemoveHyperLinkMutation.id,
							params: {
								unitId,
								subUnitId,
								id: link.id
							}
						});
						undos.push({
							id: AddHyperLinkMutation.id,
							params: {
								unitId,
								subUnitId,
								link
							}
						});
					}
				});
				return {
					undos,
					redos
				};
			}
			return {
				redos: [],
				undos: []
			};
		} }));
	}
	_initClearSelectionCommandInterceptor() {
		this.disposeWithMe(this._sheetInterceptorService.interceptCommand({ getMutations: (command) => {
			if (command.id === ClearSelectionContentCommand.id || command.id === ClearSelectionAllCommand.id || command.id === ClearSelectionFormatCommand.id) {
				const redos = [];
				const undos = [];
				const selection = this._selectionManagerService.getCurrentLastSelection();
				const target = getSheetCommandTarget(this._univerInstanceService, command.params);
				if (selection && target) {
					const { unitId, subUnitId } = target;
					Range.foreach(selection.range, (row, col) => {
						const link = this._hyperLinkModel.getHyperLinkByLocation(unitId, subUnitId, row, col);
						if (link) {
							redos.push({
								id: RemoveHyperLinkMutation.id,
								params: {
									unitId,
									subUnitId,
									id: link.id
								}
							});
							undos.push({
								id: AddHyperLinkMutation.id,
								params: {
									unitId,
									subUnitId,
									link
								}
							});
						}
					});
				}
				return {
					redos,
					undos
				};
			}
			return {
				redos: [],
				undos: []
			};
		} }));
	}
	_initAfterEditor() {
		this.disposeWithMe(this._sheetInterceptorService.writeCellInterceptor.intercept(AFTER_CELL_EDIT, { handler: (cell, context, next) => {
			if (!cell || cell.p) return next(cell);
			if (typeof cell.v === "string" && Tools.isLegalUrl(cell.v) && cell.v[cell.v.length - 1] !== " ") {
				const { unitId, subUnitId, row, col } = context;
				const link = Tools.normalizeUrl(cell.v);
				const workbook = this._univerInstanceService.getUnit(unitId, UniverInstanceType.UNIVER_SHEET);
				const worksheet = workbook === null || workbook === void 0 ? void 0 : workbook.getSheetBySheetId(subUnitId);
				if (!worksheet) return next(cell);
				const doc = worksheet.getBlankCellDocumentModel(cell, row, col);
				if (!doc.documentModel) return next(cell);
				const textX = BuildTextUtils.selection.replace({
					selection: {
						startOffset: 0,
						endOffset: cell.v.length,
						collapsed: false
					},
					body: {
						dataStream: `${cell.v}`,
						customRanges: [{
							startIndex: 0,
							endIndex: cell.v.length - 1,
							rangeId: generateRandomId(),
							rangeType: CustomRangeType.HYPERLINK,
							properties: { url: link }
						}]
					},
					doc: doc.documentModel
				});
				if (!textX) return next(cell);
				const body = doc.documentModel.getBody();
				TextX.apply(body, textX.serialize());
				return next({
					...cell,
					p: {
						id: DOCS_NORMAL_EDITOR_UNIT_ID_KEY,
						body,
						documentStyle: { pageSize: {
							width: Infinity,
							height: Infinity
						} }
					}
				});
			}
			return next(cell);
		} }));
	}
};
SheetHyperLinkSetRangeController = __decorate([
	__decorateParam(0, Inject(SheetInterceptorService)),
	__decorateParam(1, Inject(HyperLinkModel)),
	__decorateParam(2, Inject(SheetsSelectionsService)),
	__decorateParam(3, IUniverInstanceService)
], SheetHyperLinkSetRangeController);

//#endregion
//#region src/controllers/sheet-hyper-link.controller.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
let SheetsHyperLinkController = class SheetsHyperLinkController extends Disposable {
	constructor(_commandService) {
		super();
		this._commandService = _commandService;
		this._registerCommands();
	}
	_registerCommands() {
		[
			AddHyperLinkCommand,
			UpdateHyperLinkCommand,
			CancelHyperLinkCommand,
			UpdateRichHyperLinkCommand,
			CancelRichHyperLinkCommand,
			AddRichHyperLinkCommand,
			AddHyperLinkMutation,
			UpdateHyperLinkMutation,
			RemoveHyperLinkMutation,
			UpdateHyperLinkRefMutation,
			UpdateRichHyperLinkMutation
		].forEach((command) => {
			this._commandService.registerCommand(command);
		});
	}
};
SheetsHyperLinkController = __decorate([__decorateParam(0, ICommandService)], SheetsHyperLinkController);

//#endregion
//#region src/types/enums/hyper-link-type.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
let SheetHyperLinkType = /* @__PURE__ */ function(SheetHyperLinkType) {
	SheetHyperLinkType["SHEET"] = "gid";
	SheetHyperLinkType["RANGE"] = "range";
	SheetHyperLinkType["DEFINE_NAME"] = "rangeid";
	SheetHyperLinkType["INVALID"] = "invalid";
	SheetHyperLinkType["URL"] = "url";
	return SheetHyperLinkType;
}({});

//#endregion
//#region src/services/parser.service.ts
let SheetsHyperLinkParserService = class SheetsHyperLinkParserService {
	constructor(_univerInstanceService, _localeService, _definedNamesService) {
		this._univerInstanceService = _univerInstanceService;
		this._localeService = _localeService;
		this._definedNamesService = _definedNamesService;
	}
	buildHyperLink(defineNameIdOrRange, sheetId) {
		if (typeof defineNameIdOrRange === "string") return `#${"rangeid"}=${defineNameIdOrRange}`;
		let result = `#${"gid"}=${sheetId}`;
		if (defineNameIdOrRange) result += `&${"range"}=${serializeRange(defineNameIdOrRange)}`;
		return result;
	}
	parseHyperLink(urlStr) {
		if (urlStr.startsWith("#")) {
			var _search$get, _search$get2, _search$get3, _search$get4;
			const search = new URLSearchParams(urlStr.slice(1));
			const searchObj = {
				gid: (_search$get = search.get("gid")) !== null && _search$get !== void 0 ? _search$get : "",
				range: (_search$get2 = search.get("range")) !== null && _search$get2 !== void 0 ? _search$get2 : "",
				rangeid: (_search$get3 = search.get("rangeid")) !== null && _search$get3 !== void 0 ? _search$get3 : "",
				unitid: (_search$get4 = search.get("unitid")) !== null && _search$get4 !== void 0 ? _search$get4 : ""
			};
			const urlInfo = this._getURLName(searchObj);
			return {
				type: urlInfo.type,
				name: urlInfo.name,
				url: urlStr,
				searchObj
			};
		} else return {
			type: "url",
			name: urlStr,
			url: urlStr,
			searchObj: null
		};
	}
	_getURLName(params) {
		var _sheet$getName;
		const { gid, range, rangeid, unitid } = params;
		const workbook = unitid ? this._univerInstanceService.getUnit(unitid, UniverInstanceType.UNIVER_SHEET) : this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_SHEET);
		const invalidLink = {
			type: "invalid",
			name: this._localeService.t("sheets-hyper-link.message.refError")
		};
		if (!workbook) return invalidLink;
		const sheet = gid ? workbook.getSheetBySheetId(gid) : workbook.getActiveSheet();
		const sheetName = (_sheet$getName = sheet === null || sheet === void 0 ? void 0 : sheet.getName()) !== null && _sheet$getName !== void 0 ? _sheet$getName : "";
		if (range) {
			if (!sheet) return invalidLink;
			const rangeObj = deserializeRangeWithSheet(range).range;
			if (isValidRange(rangeObj, sheet) && range !== "err") return {
				type: "range",
				name: serializeRangeWithSheet(sheetName, rangeObj)
			};
			return invalidLink;
		}
		if (rangeid) {
			const range = this._definedNamesService.getValueById(workbook.getUnitId(), rangeid);
			if (range) return {
				type: "rangeid",
				name: range.formulaOrRefString
			};
			return invalidLink;
		}
		if (gid) {
			const worksheet = workbook.getSheetBySheetId(gid);
			if (worksheet) return {
				type: "gid",
				name: worksheet.getName()
			};
			return invalidLink;
		}
		return invalidLink;
	}
};
SheetsHyperLinkParserService = __decorate([
	__decorateParam(0, IUniverInstanceService),
	__decorateParam(1, Inject(LocaleService)),
	__decorateParam(2, IDefinedNamesService)
], SheetsHyperLinkParserService);

//#endregion
//#region src/plugin.ts
let UniverSheetsHyperLinkPlugin = class UniverSheetsHyperLinkPlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _injector, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		const { ...rest } = merge({}, defaultPluginConfig, this._config);
		this._configService.setConfig(SHEETS_HYPER_LINK_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		registerDependencies(this._injector, [
			[HyperLinkModel],
			[SheetsHyperLinkParserService],
			[SheetsHyperLinkController],
			[SheetsHyperLinkRefRangeController],
			[SheetHyperLinkSetRangeController],
			[SheetsHyperLinkRemoveSheetController],
			[SheetsHyperLinkRichTextRefRangeController]
		]);
		touchDependencies(this._injector, [
			[SheetsHyperLinkRefRangeController],
			[SheetsHyperLinkController],
			[SheetHyperLinkSetRangeController],
			[SheetsHyperLinkRemoveSheetController],
			[SheetsHyperLinkRichTextRefRangeController]
		]);
	}
};
_defineProperty(UniverSheetsHyperLinkPlugin, "pluginName", SHEET_HYPER_LINK_PLUGIN);
_defineProperty(UniverSheetsHyperLinkPlugin, "packageName", name);
_defineProperty(UniverSheetsHyperLinkPlugin, "version", version);
_defineProperty(UniverSheetsHyperLinkPlugin, "type", UniverInstanceType.UNIVER_SHEET);
UniverSheetsHyperLinkPlugin = __decorate([
	DependentOn(UniverSheetsPlugin),
	__decorateParam(1, Inject(Injector)),
	__decorateParam(2, IConfigService)
], UniverSheetsHyperLinkPlugin);

//#endregion
export { AddHyperLinkCommand, AddHyperLinkMutation, AddRichHyperLinkCommand, CancelHyperLinkCommand, CancelRichHyperLinkCommand, ERROR_RANGE, HyperLinkModel, RemoveHyperLinkMutation, SheetHyperLinkType, SheetsHyperLinkParserService, UniverSheetsHyperLinkPlugin, UpdateHyperLinkCommand, UpdateHyperLinkMutation, UpdateHyperLinkRefMutation, UpdateRichHyperLinkCommand };
