// @univerjs/docs-drawing/lib/es/index.js
import { Disposable, IConfigService, IResourceManagerService, IUniverInstanceService, Inject, Injector, Plugin, UniverInstanceType, createIdentifier, merge, touchDependencies } from "@univerjs/core";
import { IDrawingManagerService, UnitDrawingService } from "@univerjs/drawing";

//#region src/services/doc-drawing.service.ts
var DocDrawingService = class extends UnitDrawingService {};
const IDocDrawingService = createIdentifier("univer.doc.plugin.doc-drawing.service");

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/controllers/doc-drawing.controller.ts
const DOCS_DRAWING_PLUGIN = "DOC_DRAWING_PLUGIN";
let DocDrawingController = class DocDrawingController extends Disposable {
	constructor(_docDrawingService, _drawingManagerService, _resourceManagerService, _univerInstanceService) {
		super();
		this._docDrawingService = _docDrawingService;
		this._drawingManagerService = _drawingManagerService;
		this._resourceManagerService = _resourceManagerService;
		this._univerInstanceService = _univerInstanceService;
		this._init();
	}
	_init() {
		this._initSnapshot();
	}
	_initSnapshot() {
		const toJson = (unitId) => {
			const doc = this._univerInstanceService.getUnit(unitId, UniverInstanceType.UNIVER_DOC);
			if (doc) {
				const drawings = doc.getSnapshot().drawings;
				const drawingOrder = doc.getSnapshot().drawingsOrder;
				return JSON.stringify({
					data: drawings !== null && drawings !== void 0 ? drawings : {},
					order: drawingOrder !== null && drawingOrder !== void 0 ? drawingOrder : []
				});
			}
			return "";
		};
		const parseJson = (json) => {
			if (!json) return {
				data: {},
				order: []
			};
			try {
				return JSON.parse(json);
			} catch (err) {
				return {
					data: {},
					order: []
				};
			}
		};
		this.disposeWithMe(this._resourceManagerService.registerPluginResource({
			pluginName: DOCS_DRAWING_PLUGIN,
			businesses: [UniverInstanceType.UNIVER_DOC],
			toJson: (unitId) => toJson(unitId),
			parseJson: (json) => parseJson(json),
			onUnLoad: (unitId) => {
				this._setDrawingDataForUnit(unitId, {
					data: {},
					order: []
				});
			},
			onLoad: (unitId, value) => {
				var _value$data, _value$order;
				this._setDrawingDataForUnit(unitId, {
					data: (_value$data = value.data) !== null && _value$data !== void 0 ? _value$data : {},
					order: (_value$order = value.order) !== null && _value$order !== void 0 ? _value$order : []
				});
			}
		}));
	}
	_setDrawingDataForUnit(unitId, drawingMapItem) {
		const documentDataModel = this._univerInstanceService.getUnit(unitId);
		if (documentDataModel == null) return;
		documentDataModel.resetDrawing(drawingMapItem.data, drawingMapItem.order);
		this.loadDrawingDataForUnit(unitId);
	}
	loadDrawingDataForUnit(unitId) {
		const dataModel = this._univerInstanceService.getUnit(unitId, UniverInstanceType.UNIVER_DOC);
		if (!dataModel) return false;
		const subUnitId = unitId;
		const drawingDataModels = dataModel.getDrawings();
		const drawingOrderModel = dataModel.getDrawingsOrder();
		if (!drawingDataModels || !drawingOrderModel) return false;
		Object.keys(drawingDataModels).forEach((drawingId) => {
			drawingDataModels[drawingId] = { ...drawingDataModels[drawingId] };
		});
		const subDrawings = { [subUnitId]: {
			unitId,
			subUnitId,
			data: drawingDataModels,
			order: drawingOrderModel
		} };
		this._docDrawingService.registerDrawingData(unitId, subDrawings);
		this._drawingManagerService.registerDrawingData(unitId, subDrawings);
		return true;
	}
};
DocDrawingController = __decorate([
	__decorateParam(0, IDocDrawingService),
	__decorateParam(1, IDrawingManagerService),
	__decorateParam(2, IResourceManagerService),
	__decorateParam(3, IUniverInstanceService)
], DocDrawingController);

//#endregion
//#region package.json
var name = "@univerjs/docs-drawing";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const DOCS_DRAWING_PLUGIN_CONFIG_KEY = "docs-drawing.config";
const configSymbol = Symbol(DOCS_DRAWING_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region src/plugin.ts
let UniverDocsDrawingPlugin = class UniverDocsDrawingPlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _injector, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		const { ...rest } = merge({}, defaultPluginConfig, this._config);
		this._configService.setConfig(DOCS_DRAWING_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		[
			[DocDrawingController],
			[DocDrawingService],
			[IDocDrawingService, { useClass: DocDrawingService }]
		].forEach((dependency) => this._injector.add(dependency));
		touchDependencies(this._injector, [[DocDrawingController]]);
	}
};
_defineProperty(UniverDocsDrawingPlugin, "pluginName", DOCS_DRAWING_PLUGIN);
_defineProperty(UniverDocsDrawingPlugin, "packageName", name);
_defineProperty(UniverDocsDrawingPlugin, "version", version);
_defineProperty(UniverDocsDrawingPlugin, "type", UniverInstanceType.UNIVER_DOC);
UniverDocsDrawingPlugin = __decorate([__decorateParam(1, Inject(Injector)), __decorateParam(2, IConfigService)], UniverDocsDrawingPlugin);

//#endregion
export { DOCS_DRAWING_PLUGIN, DocDrawingController, DocDrawingService, IDocDrawingService, UniverDocsDrawingPlugin };
