// @univerjs/docs-hyper-link/lib/cjs/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
let _univerjs_core = require("@univerjs/core");

//#region package.json
var name = "@univerjs/docs-hyper-link";
var version = "0.25.1";

//#endregion
//#region src/commands/mutations/hyper-link.mutation.ts
const AddHyperLinkMuatation = {
	id: "docs.mutation.add-hyper-link",
	type: _univerjs_core.CommandType.MUTATION,
	handler: () => {
		return true;
	}
};
const UpdateHyperLinkMuatation = {
	id: "docs.mutation.update-hyper-link",
	type: _univerjs_core.CommandType.MUTATION,
	handler: () => {
		return true;
	}
};
const DeleteHyperLinkMuatation = {
	id: "docs.mutation.delete-hyper-link",
	type: _univerjs_core.CommandType.MUTATION,
	handler: () => {
		return true;
	}
};

//#endregion
//#region src/config/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const DOCS_HYPER_LINK_PLUGIN_CONFIG_KEY = "docs-hyper-link.config";
const configSymbol = Symbol(DOCS_HYPER_LINK_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/controllers/resource.controller.ts
const DOC_HYPER_LINK_PLUGIN = "DOC_HYPER_LINK_PLUGIN";
let DocHyperLinkResourceController = class DocHyperLinkResourceController extends _univerjs_core.Disposable {
	constructor(_resourceManagerService, _univerInstanceService) {
		super();
		this._resourceManagerService = _resourceManagerService;
		this._univerInstanceService = _univerInstanceService;
		this._init();
	}
	_init() {
		this._resourceManagerService.registerPluginResource({
			pluginName: DOC_HYPER_LINK_PLUGIN,
			businesses: [_univerjs_core.UniverInstanceType.UNIVER_DOC],
			onLoad: (unitID, resource) => {
				const doc = this._univerInstanceService.getUnit(unitID, _univerjs_core.UniverInstanceType.UNIVER_DOC);
				if (!doc) return;
				const customRangeMap = /* @__PURE__ */ new Map();
				const handleDoc = (model) => {
					var _model$getBody;
					(_model$getBody = model.getBody()) === null || _model$getBody === void 0 || (_model$getBody = _model$getBody.customRanges) === null || _model$getBody === void 0 || _model$getBody.forEach((customRange) => {
						if (customRange.rangeType === _univerjs_core.CustomRangeType.HYPERLINK) customRangeMap.set(customRange.rangeId, customRange);
					});
					return customRangeMap;
				};
				doc.headerModelMap.forEach((headerModel) => {
					handleDoc(headerModel);
				});
				doc.footerModelMap.forEach((footerModel) => {
					handleDoc(footerModel);
				});
				handleDoc(doc);
				resource.links.forEach((link) => {
					const customRange = customRangeMap.get(link.id);
					if (customRange) customRange.properties = {
						...customRange.properties,
						url: link.payload
					};
				});
			},
			onUnLoad: (unitID) => {},
			toJson: (unitID) => {
				const doc = this._univerInstanceService.getUnit(unitID, _univerjs_core.UniverInstanceType.UNIVER_DOC);
				const links = [];
				if (doc) {
					const handleDoc = (model) => {
						var _model$getBody2;
						(_model$getBody2 = model.getBody()) === null || _model$getBody2 === void 0 || (_model$getBody2 = _model$getBody2.customRanges) === null || _model$getBody2 === void 0 || _model$getBody2.forEach((customRange) => {
							if (customRange.rangeType === _univerjs_core.CustomRangeType.HYPERLINK) {
								var _customRange$properti;
								links.push({
									id: customRange.rangeId,
									payload: ((_customRange$properti = customRange.properties) === null || _customRange$properti === void 0 ? void 0 : _customRange$properti.url) || ""
								});
							}
						});
					};
					doc.headerModelMap.forEach((headerModel) => {
						handleDoc(headerModel);
					});
					doc.footerModelMap.forEach((footerModel) => {
						handleDoc(footerModel);
					});
					handleDoc(doc);
				}
				return JSON.stringify({ links });
			},
			parseJson(bytes) {
				return JSON.parse(bytes);
			}
		});
	}
};
DocHyperLinkResourceController = __decorate([__decorateParam(0, (0, _univerjs_core.Inject)(_univerjs_core.IResourceManagerService)), __decorateParam(1, _univerjs_core.IUniverInstanceService)], DocHyperLinkResourceController);

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region src/plugin.ts
let UniverDocsHyperLinkPlugin = class UniverDocsHyperLinkPlugin extends _univerjs_core.Plugin {
	constructor(_config = defaultPluginConfig, _injector, _configService, _commandService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		this._commandService = _commandService;
		const { ...rest } = (0, _univerjs_core.merge)({}, defaultPluginConfig, this._config);
		this._configService.setConfig(DOCS_HYPER_LINK_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		[[DocHyperLinkResourceController]].forEach((dep) => this._injector.add(dep));
		[
			AddHyperLinkMuatation,
			DeleteHyperLinkMuatation,
			UpdateHyperLinkMuatation
		].forEach((mutation) => {
			this.disposeWithMe(this._commandService.registerCommand(mutation));
		});
		this._injector.get(DocHyperLinkResourceController);
	}
};
_defineProperty(UniverDocsHyperLinkPlugin, "pluginName", DOC_HYPER_LINK_PLUGIN);
_defineProperty(UniverDocsHyperLinkPlugin, "packageName", name);
_defineProperty(UniverDocsHyperLinkPlugin, "version", version);
_defineProperty(UniverDocsHyperLinkPlugin, "type", _univerjs_core.UniverInstanceType.UNIVER_DOC);
UniverDocsHyperLinkPlugin = __decorate([
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector)),
	__decorateParam(2, _univerjs_core.IConfigService),
	__decorateParam(3, _univerjs_core.ICommandService)
], UniverDocsHyperLinkPlugin);

//#endregion
Object.defineProperty(exports, 'UniverDocsHyperLinkPlugin', {
  enumerable: true,
  get: function () {
    return UniverDocsHyperLinkPlugin;
  }
});
