// @univerjs/preset-sheets-advanced/lib/index.js
import { UniverProFormulaEnginePlugin as a } from "@univerjs-pro/engine-formula";
export * from "@univerjs-pro/engine-formula";
import { UniverExchangeClientPlugin as u } from "@univerjs-pro/exchange-client";
export * from "@univerjs-pro/exchange-client";
import { UniverLicensePlugin as s } from "@univerjs-pro/license";
export * from "@univerjs-pro/license";
import { UniverSheetsChartPlugin as x } from "@univerjs-pro/sheets-chart";
export * from "@univerjs-pro/sheets-chart";
import { UniverSheetsChartUIPlugin as v } from "@univerjs-pro/sheets-chart-ui";
export * from "@univerjs-pro/sheets-chart-ui";
import { UniverSheetsExchangeClientPlugin as U } from "@univerjs-pro/sheets-exchange-client";
export * from "@univerjs-pro/sheets-exchange-client";
import { UniverSheetsOutlinePlugin as g } from "@univerjs-pro/sheets-outline";
export * from "@univerjs-pro/sheets-outline";
import { UniverSheetsOutlineUIPlugin as h } from "@univerjs-pro/sheets-outline-ui";
export * from "@univerjs-pro/sheets-outline-ui";
import { UniverSheetsPivotTablePlugin as c } from "@univerjs-pro/sheets-pivot";
export * from "@univerjs-pro/sheets-pivot";
import { UniverSheetsPivotTableUIPlugin as S } from "@univerjs-pro/sheets-pivot-ui";
export * from "@univerjs-pro/sheets-pivot-ui";
import { UniverSheetsPrintPlugin as P } from "@univerjs-pro/sheets-print";
export * from "@univerjs-pro/sheets-print";
import { UniverSheetsShapePlugin as d } from "@univerjs-pro/sheets-shape";
export * from "@univerjs-pro/sheets-shape";
import { UniverSheetsShapeUIPlugin as I } from "@univerjs-pro/sheets-shape-ui";
export * from "@univerjs-pro/sheets-shape-ui";
import { UniverSheetSparklinePlugin as k } from "@univerjs-pro/sheets-sparkline";
export * from "@univerjs-pro/sheets-sparkline";
import { UniverSheetSparklineUIPlugin as $ } from "@univerjs-pro/sheets-sparkline-ui";
export * from "@univerjs-pro/sheets-sparkline-ui";
import "@univerjs-pro/exchange-client/lib/facade";
import "@univerjs-pro/sheets-pivot/lib/facade";
import "@univerjs-pro/engine-formula/lib/facade";
import "@univerjs-pro/sheets-print/lib/facade";
import "@univerjs-pro/sheets-chart/lib/facade";
import "@univerjs-pro/sheets-outline/lib/facade";
import "@univerjs-pro/sheets-shape/lib/facade";
import "@univerjs-pro/sheets-sparkline/lib/facade";
export * from "@univerjs-pro/engine-chart";
export * from "@univerjs-pro/engine-shape";
export * from "@univerjs/sheets-graphics";
function R(m = {
  license: "",
  universerEndpoint: ""
}) {
  const {
    license: p,
    universerEndpoint: e,
    useWorker: n,
    formula: o,
    exchangeClientOptions: l,
    print: i,
    pivot: t
  } = m, r = e != null ? e : `${window.location.protocol}//${window.location.host}`;
  return {
    plugins: [
      [s, { license: p }],
      [c, {
        notExecuteFormula: n,
        maxLimitItemCount: t == null ? void 0 : t.maxLimitItemCount
      }],
      S,
      [a, {
        notExecuteFormula: n,
        function: o == null ? void 0 : o.function
      }],
      [P, {
        enforceWatermark: i == null ? void 0 : i.enforceWatermark
      }],
      x,
      v,
      g,
      h,
      d,
      I,
      k,
      $,
      [u, {
        uploadFileServerUrl: `${r}/universer-api/stream/file/upload`,
        getTaskServerUrl: `${r}/universer-api/exchange/task/{taskID}`,
        signUrlServerUrl: `${r}/universer-api/file/{fileID}/sign-url`,
        importServerUrl: `${r}/universer-api/exchange/{type}/import`,
        exportServerUrl: `${r}/universer-api/exchange/{type}/export`,
        downloadEndpointUrl: `${r}/`,
        options: l
      }],
      U
    ].filter((f) => !!f)
  };
}
export {
  R as UniverSheetsAdvancedPreset
};
