// @univerjs/sheets-drawing/lib/cjs/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
let _univerjs_sheets = require("@univerjs/sheets");
let _univerjs_core = require("@univerjs/core");
let _univerjs_drawing = require("@univerjs/drawing");
let _univerjs_engine_render = require("@univerjs/engine-render");

//#region src/basics/transform-position.ts
function drawingPositionToTransform(position, sheetSkeletonParam) {
	if (!sheetSkeletonParam) return;
	const { unitId, sheetId, skeleton } = sheetSkeletonParam;
	const { from, to, flipY = false, flipX = false, angle = 0, skewX = 0, skewY = 0 } = position;
	let { left, top, width, height } = (0, _univerjs_sheets.convertPositionSheetOverGridToAbsolute)(unitId, sheetId, {
		from,
		to
	}, skeleton);
	const sheetWidth = skeleton.rowHeaderWidth + skeleton.columnTotalWidth;
	const sheetHeight = skeleton.columnHeaderHeight + skeleton.rowTotalHeight;
	if (left + width > sheetWidth) left = sheetWidth - width;
	if (top + height > sheetHeight) top = sheetHeight - height;
	return {
		flipY,
		flipX,
		angle,
		skewX,
		skewY,
		left,
		top,
		width,
		height
	};
}
function transformToDrawingPosition(transform, skeleton) {
	const { left = 0, top = 0, width = 0, height = 0, flipY = false, flipX = false, angle = 0, skewX = 0, skewY = 0 } = transform;
	return {
		flipY,
		flipX,
		angle,
		skewX,
		skewY,
		from: skeleton.getCellIndexAndOffsetByPosition(left, top),
		to: skeleton.getCellIndexAndOffsetByPosition(left + width, top + height)
	};
}
/**
* In excel, the basic drawing with rotate bound use major axis switch, axis-aligned bound will bu used.That means the position bound of drawing element will save as nearly axis-aligned rectangle.
* Here is the rule to convert transform to axis-aligned position:
* [-45°, 45°):  use the original bound
* [45°, 135°): rotate the bound 90° clockwise,and the left, top, bottom,right will use the rotated bound.
* [135°, 225°): use the original bound
* [225°, 315°): rotate the bound 90° counterclockwise, and the left, top, bottom, right will use the rotated bound.
* @return The axis-aligned position of the drawing element.
*/
function transformToAxisAlignPosition(transform, skeleton) {
	const { left = 0, top = 0, width = 0, height = 0, angle = 0 } = transform;
	const norm = (angle % 360 + 360) % 360;
	if (!(norm >= 45 && norm < 135 || norm >= 225 && norm < 315)) return transformToDrawingPosition(transform, skeleton);
	return transformToDrawingPosition({
		...transform,
		left: left + width / 2 - height / 2,
		top: top + height / 2 - width / 2,
		width: height,
		height: width
	}, skeleton);
}

//#endregion
//#region src/services/sheet-drawing.service.ts
let SheetDrawingAnchorType = /* @__PURE__ */ function(SheetDrawingAnchorType) {
	/**
	* Only the position of the drawing follows the cell changes. When rows or columns are inserted or deleted, the position of the drawing changes, but the size remains the same.
	*/
	SheetDrawingAnchorType["Position"] = "0";
	/**
	* The size and position of the drawing follow the cell changes. When rows or columns are inserted or deleted, the size and position of the drawing change accordingly.
	*/
	SheetDrawingAnchorType["Both"] = "1";
	/**
	* The size and position of the drawing do not follow the cell changes. When rows or columns are inserted or deleted, the position and size of the drawing remain unchanged.
	*/
	SheetDrawingAnchorType["None"] = "2";
	return SheetDrawingAnchorType;
}({});
var SheetDrawingService = class extends _univerjs_drawing.UnitDrawingService {};
const ISheetDrawingService = (0, _univerjs_core.createIdentifier)("sheets-drawing.sheet-drawing.service");

//#endregion
//#region src/commands/mutations/set-drawing-apply.mutation.ts
let DrawingApplyType = /* @__PURE__ */ function(DrawingApplyType) {
	DrawingApplyType[DrawingApplyType["INSERT"] = 0] = "INSERT";
	DrawingApplyType[DrawingApplyType["REMOVE"] = 1] = "REMOVE";
	DrawingApplyType[DrawingApplyType["UPDATE"] = 2] = "UPDATE";
	DrawingApplyType[DrawingApplyType["ARRANGE"] = 3] = "ARRANGE";
	DrawingApplyType[DrawingApplyType["GROUP"] = 4] = "GROUP";
	DrawingApplyType[DrawingApplyType["UNGROUP"] = 5] = "UNGROUP";
	return DrawingApplyType;
}({});
const SetDrawingApplyMutation = {
	id: "sheet.mutation.set-drawing-apply",
	type: _univerjs_core.CommandType.MUTATION,
	handler: (accessor, params) => {
		const drawingManagerService = accessor.get(_univerjs_drawing.IDrawingManagerService);
		const sheetDrawingService = accessor.get(ISheetDrawingService);
		const { op, unitId, subUnitId, type, objects } = params;
		drawingManagerService.applyJson1(unitId, subUnitId, op);
		sheetDrawingService.applyJson1(unitId, subUnitId, op);
		switch (type) {
			case 0:
				drawingManagerService.addNotification(objects);
				sheetDrawingService.addNotification(objects);
				break;
			case 1:
				drawingManagerService.removeNotification(objects);
				sheetDrawingService.removeNotification(objects);
				break;
			case 2:
				drawingManagerService.updateNotification(objects);
				sheetDrawingService.updateNotification(objects);
				break;
			case 3:
				drawingManagerService.orderNotification(objects);
				sheetDrawingService.orderNotification(objects);
				break;
			case 4:
				drawingManagerService.groupUpdateNotification(objects);
				sheetDrawingService.groupUpdateNotification(objects);
				break;
			case 5:
				drawingManagerService.ungroupUpdateNotification(objects);
				sheetDrawingService.ungroupUpdateNotification(objects);
				break;
		}
		return true;
	}
};

//#endregion
//#region src/commands/operations/clear-drawing-transformer.operation.ts
const ClearSheetDrawingTransformerOperation = {
	id: "sheet.operation.clear-drawing-transformer",
	type: _univerjs_core.CommandType.MUTATION,
	handler: (accessor, params) => {
		const renderManagerService = accessor.get(_univerjs_engine_render.IRenderManagerService);
		params.forEach((unitId) => {
			var _renderManagerService;
			(_renderManagerService = renderManagerService.getRenderById(unitId)) === null || _renderManagerService === void 0 || (_renderManagerService = _renderManagerService.scene.getTransformer()) === null || _renderManagerService === void 0 || _renderManagerService.debounceRefreshControls();
		});
		return true;
	}
};

//#endregion
//#region src/commands/commands/insert-sheet-drawing.command.ts
const InsertSheetDrawingCommand = {
	id: "sheet.command.insert-sheet-image",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		var _intercepted$preRedos, _intercepted$preUndos;
		if (!params) return false;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const undoRedoService = accessor.get(_univerjs_core.IUndoRedoService);
		const sheetDrawingService = accessor.get(ISheetDrawingService);
		const sheetInterceptorService = accessor.get(_univerjs_sheets.SheetInterceptorService);
		const drawings = params.drawings;
		const { unitId, subUnitId, undo, redo, objects } = sheetDrawingService.getBatchAddOp(drawings);
		const intercepted = sheetInterceptorService.onCommandExecute({
			id: InsertSheetDrawingCommand.id,
			params
		});
		const redoMutations = [
			...(_intercepted$preRedos = intercepted.preRedos) !== null && _intercepted$preRedos !== void 0 ? _intercepted$preRedos : [],
			{
				id: SetDrawingApplyMutation.id,
				params: {
					unitId,
					subUnitId,
					op: redo,
					objects,
					type: 0
				}
			},
			{
				id: ClearSheetDrawingTransformerOperation.id,
				params: [unitId]
			},
			...intercepted.redos
		];
		const undoMutations = [
			...(_intercepted$preUndos = intercepted.preUndos) !== null && _intercepted$preUndos !== void 0 ? _intercepted$preUndos : [],
			{
				id: SetDrawingApplyMutation.id,
				params: {
					unitId,
					subUnitId,
					op: undo,
					objects,
					type: 1
				}
			},
			{
				id: ClearSheetDrawingTransformerOperation.id,
				params: [unitId]
			},
			...intercepted.undos
		];
		if ((0, _univerjs_core.sequenceExecute)(redoMutations, commandService).result) {
			undoRedoService.pushUndoRedo({
				unitID: unitId,
				undoMutations,
				redoMutations
			});
			return true;
		}
		return false;
	}
};

//#endregion
//#region src/commands/commands/remove-sheet-drawing.command.ts
const RemoveSheetDrawingCommand = {
	id: "sheet.command.remove-sheet-image",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		var _intercepted$preRedos, _intercepted$preUndos;
		if (!params) return false;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const undoRedoService = accessor.get(_univerjs_core.IUndoRedoService);
		const sheetInterceptorService = accessor.get(_univerjs_sheets.SheetInterceptorService);
		const sheetDrawingService = accessor.get(ISheetDrawingService);
		const { drawings } = params;
		const { unitId, subUnitId, undo, redo, objects } = sheetDrawingService.getBatchRemoveOp(drawings);
		const intercepted = sheetInterceptorService.onCommandExecute({
			id: RemoveSheetDrawingCommand.id,
			params
		});
		const redoMutations = [
			...(_intercepted$preRedos = intercepted.preRedos) !== null && _intercepted$preRedos !== void 0 ? _intercepted$preRedos : [],
			{
				id: SetDrawingApplyMutation.id,
				params: {
					unitId,
					subUnitId,
					op: redo,
					objects,
					type: 1
				}
			},
			{
				id: ClearSheetDrawingTransformerOperation.id,
				params: [unitId]
			},
			...intercepted.redos
		];
		const undoMutations = [
			...(_intercepted$preUndos = intercepted.preUndos) !== null && _intercepted$preUndos !== void 0 ? _intercepted$preUndos : [],
			{
				id: SetDrawingApplyMutation.id,
				params: {
					unitId,
					subUnitId,
					op: undo,
					objects,
					type: 0
				}
			},
			{
				id: ClearSheetDrawingTransformerOperation.id,
				params: [unitId]
			},
			...intercepted.undos
		];
		if ((0, _univerjs_core.sequenceExecute)(redoMutations, commandService).result) {
			undoRedoService.pushUndoRedo({
				unitID: unitId,
				undoMutations,
				redoMutations
			});
			return true;
		}
		return false;
	}
};

//#endregion
//#region src/commands/commands/set-drawing-arrange.command.ts
const SetDrawingArrangeCommand = {
	id: "sheet.command.set-drawing-arrange",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const undoRedoService = accessor.get(_univerjs_core.IUndoRedoService);
		if (!params) return false;
		const sheetDrawingService = accessor.get(ISheetDrawingService);
		const { unitId, subUnitId, drawingIds, arrangeType } = params;
		const drawingOrderMapParam = {
			unitId,
			subUnitId,
			drawingIds
		};
		let jsonOp;
		if (arrangeType === _univerjs_core.ArrangeTypeEnum.forward) jsonOp = sheetDrawingService.getForwardDrawingsOp(drawingOrderMapParam);
		else if (arrangeType === _univerjs_core.ArrangeTypeEnum.backward) jsonOp = sheetDrawingService.getBackwardDrawingOp(drawingOrderMapParam);
		else if (arrangeType === _univerjs_core.ArrangeTypeEnum.front) jsonOp = sheetDrawingService.getFrontDrawingsOp(drawingOrderMapParam);
		else if (arrangeType === _univerjs_core.ArrangeTypeEnum.back) jsonOp = sheetDrawingService.getBackDrawingsOp(drawingOrderMapParam);
		if (jsonOp == null) return false;
		const { objects, redo, undo } = jsonOp;
		if (commandService.syncExecuteCommand(SetDrawingApplyMutation.id, {
			op: redo,
			unitId,
			subUnitId,
			objects,
			type: 3
		})) {
			undoRedoService.pushUndoRedo({
				unitID: unitId,
				undoMutations: [{
					id: SetDrawingApplyMutation.id,
					params: {
						op: undo,
						unitId,
						subUnitId,
						objects,
						type: 3
					}
				}],
				redoMutations: [{
					id: SetDrawingApplyMutation.id,
					params: {
						op: redo,
						unitId,
						subUnitId,
						objects,
						type: 3
					}
				}]
			});
			return true;
		}
		return false;
	}
};

//#endregion
//#region src/commands/commands/set-sheet-drawing.command.ts
const SetSheetDrawingCommand = {
	id: "sheet.command.set-sheet-image",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		var _intercepted$preRedos, _intercepted$preUndos;
		if (!params) return false;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const undoRedoService = accessor.get(_univerjs_core.IUndoRedoService);
		const sheetDrawingService = accessor.get(ISheetDrawingService);
		const sheetInterceptorService = accessor.get(_univerjs_sheets.SheetInterceptorService);
		const { drawings } = params;
		const { unitId, subUnitId, undo, redo, objects } = sheetDrawingService.getBatchUpdateOp(drawings);
		const intercepted = sheetInterceptorService.onCommandExecute({
			id: SetSheetDrawingCommand.id,
			params
		});
		const redoMutations = [
			...(_intercepted$preRedos = intercepted.preRedos) !== null && _intercepted$preRedos !== void 0 ? _intercepted$preRedos : [],
			{
				id: SetDrawingApplyMutation.id,
				params: {
					unitId,
					subUnitId,
					op: redo,
					objects,
					type: 2
				}
			},
			{
				id: ClearSheetDrawingTransformerOperation.id,
				params: [unitId]
			},
			...intercepted.redos
		];
		const undoMutations = [
			...(_intercepted$preUndos = intercepted.preUndos) !== null && _intercepted$preUndos !== void 0 ? _intercepted$preUndos : [],
			{
				id: SetDrawingApplyMutation.id,
				params: {
					unitId,
					subUnitId,
					op: undo,
					objects,
					type: 2
				}
			},
			{
				id: ClearSheetDrawingTransformerOperation.id,
				params: [unitId]
			},
			...intercepted.undos
		];
		if ((0, _univerjs_core.sequenceExecute)(redoMutations, commandService).result) {
			undoRedoService.pushUndoRedo({
				unitID: unitId,
				undoMutations,
				redoMutations
			});
			return true;
		}
		return false;
	}
};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/controllers/sheet-drawing.controller.ts
const SHEET_DRAWING_PLUGIN = "SHEET_DRAWING_PLUGIN";
let SheetsDrawingLoadController = class SheetsDrawingLoadController extends _univerjs_core.Disposable {
	constructor(_sheetInterceptorService, _univerInstanceService, _commandService, _sheetDrawingService, _drawingManagerService, _resourceManagerService) {
		super();
		this._sheetInterceptorService = _sheetInterceptorService;
		this._univerInstanceService = _univerInstanceService;
		this._commandService = _commandService;
		this._sheetDrawingService = _sheetDrawingService;
		this._drawingManagerService = _drawingManagerService;
		this._resourceManagerService = _resourceManagerService;
		this._initCommands();
		this._initSnapshot();
		this._initSheetChange();
		this.disposeWithMe(this._commandService.registerCommand(SetDrawingApplyMutation));
	}
	_initCommands() {
		[
			SetSheetDrawingCommand,
			InsertSheetDrawingCommand,
			RemoveSheetDrawingCommand,
			SetDrawingArrangeCommand,
			ClearSheetDrawingTransformerOperation
		].forEach((command) => this.disposeWithMe(this._commandService.registerCommand(command)));
	}
	_initSnapshot() {
		const toJson = (unitId, model) => {
			const map = model || this._sheetDrawingService.getDrawingDataForUnit(unitId);
			if (map) return JSON.stringify(map);
			return "";
		};
		const parseJson = (json) => {
			if (!json) return {};
			try {
				return JSON.parse(json);
			} catch {
				return {};
			}
		};
		this.disposeWithMe(this._resourceManagerService.registerPluginResource({
			pluginName: SHEET_DRAWING_PLUGIN,
			businesses: [_univerjs_core.UniverInstanceType.UNIVER_SHEET],
			toJson: (unitId, model) => toJson(unitId, model),
			parseJson: (json) => parseJson(json),
			onUnLoad: (unitId) => {
				this._sheetDrawingService.removeDrawingDataForUnit(unitId);
				this._drawingManagerService.removeDrawingDataForUnit(unitId);
			},
			onLoad: (unitId, value) => {
				this._sheetDrawingService.registerDrawingData(unitId, value);
				this._drawingManagerService.registerDrawingData(unitId, value);
			}
		}));
	}
	_initSheetChange() {
		this.disposeWithMe(this._sheetInterceptorService.interceptCommand({ getMutations: (commandInfo) => {
			if (commandInfo.id === _univerjs_sheets.RemoveSheetCommand.id) {
				var _getActiveSheet;
				const params = commandInfo.params;
				const unitId = params.unitId || this._univerInstanceService.getCurrentUnitOfType(_univerjs_core.UniverInstanceType.UNIVER_SHEET).getUnitId();
				const subUnitId = params.subUnitId || ((_getActiveSheet = this._univerInstanceService.getCurrentUnitOfType(_univerjs_core.UniverInstanceType.UNIVER_SHEET).getActiveSheet()) === null || _getActiveSheet === void 0 ? void 0 : _getActiveSheet.getSheetId());
				if (!unitId || !subUnitId) return {
					redos: [],
					undos: []
				};
				const drawingData = this._sheetDrawingService.getDrawingData(unitId, subUnitId);
				const drawings = Object.values(drawingData).filter((drawing) => {
					if (drawing.drawingType === _univerjs_core.DrawingTypeEnum.DRAWING_CHART) return false;
					return true;
				});
				if (drawings.length === 0) return {
					redos: [],
					undos: []
				};
				const { unitId: jsonOpUnitId, subUnitId: jsonOpSubUnitId, undo, redo, objects } = this._sheetDrawingService.getBatchRemoveOp(drawings);
				return {
					redos: [{
						id: SetDrawingApplyMutation.id,
						params: {
							op: redo,
							unitId: jsonOpUnitId,
							subUnitId: jsonOpSubUnitId,
							objects,
							type: 1
						}
					}],
					undos: [{
						id: SetDrawingApplyMutation.id,
						params: {
							op: undo,
							unitId: jsonOpUnitId,
							subUnitId: jsonOpSubUnitId,
							objects,
							type: 0
						}
					}]
				};
			} else if (commandInfo.id === _univerjs_sheets.CopySheetCommand.id) {
				const { unitId, subUnitId, targetSubUnitId } = commandInfo.params;
				if (!unitId || !subUnitId || !targetSubUnitId) return {
					redos: [],
					undos: []
				};
				const drawingData = this._sheetDrawingService.getDrawingData(unitId, subUnitId);
				const drawings = Object.values(drawingData).filter((drawing) => {
					if (drawing.drawingType === _univerjs_core.DrawingTypeEnum.DRAWING_CHART) return false;
					return true;
				}).map((drawing) => {
					return {
						...drawing,
						subUnitId: targetSubUnitId,
						drawingId: (0, _univerjs_core.generateRandomId)(6)
					};
				});
				if (drawings.length === 0) return {
					redos: [],
					undos: []
				};
				const { unitId: jsonOpUnitId, subUnitId: jsonOpSubUnitId, undo, redo, objects } = this._sheetDrawingService.getBatchAddOp(drawings);
				return {
					redos: [{
						id: SetDrawingApplyMutation.id,
						params: {
							op: redo,
							unitId: jsonOpUnitId,
							subUnitId: jsonOpSubUnitId,
							objects,
							type: 0
						}
					}],
					undos: [{
						id: SetDrawingApplyMutation.id,
						params: {
							op: undo,
							unitId: jsonOpUnitId,
							subUnitId: jsonOpSubUnitId,
							objects,
							type: 1
						}
					}]
				};
			}
			return {
				redos: [],
				undos: []
			};
		} }));
	}
};
SheetsDrawingLoadController = __decorate([
	__decorateParam(0, (0, _univerjs_core.Inject)(_univerjs_sheets.SheetInterceptorService)),
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.IUniverInstanceService)),
	__decorateParam(2, _univerjs_core.ICommandService),
	__decorateParam(3, ISheetDrawingService),
	__decorateParam(4, _univerjs_drawing.IDrawingManagerService),
	__decorateParam(5, _univerjs_core.IResourceManagerService)
], SheetsDrawingLoadController);

//#endregion
//#region package.json
var name = "@univerjs/sheets-drawing";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const SHEETS_DRAWING_PLUGIN_CONFIG_KEY = "sheets-drawing.config";
const configSymbol = Symbol(SHEETS_DRAWING_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region src/plugin.ts
let UniverSheetsDrawingPlugin = class UniverSheetsDrawingPlugin extends _univerjs_core.Plugin {
	constructor(_config = defaultPluginConfig, _injector, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		const { ...rest } = (0, _univerjs_core.merge)({}, defaultPluginConfig, this._config);
		this._configService.setConfig(SHEETS_DRAWING_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		[[SheetsDrawingLoadController], [ISheetDrawingService, { useClass: SheetDrawingService }]].forEach((dependency) => this._injector.add(dependency));
		this._injector.get(SheetsDrawingLoadController);
	}
};
_defineProperty(UniverSheetsDrawingPlugin, "pluginName", SHEET_DRAWING_PLUGIN);
_defineProperty(UniverSheetsDrawingPlugin, "packageName", name);
_defineProperty(UniverSheetsDrawingPlugin, "version", version);
_defineProperty(UniverSheetsDrawingPlugin, "type", _univerjs_core.UniverInstanceType.UNIVER_SHEET);
UniverSheetsDrawingPlugin = __decorate([
	(0, _univerjs_core.DependentOn)(_univerjs_drawing.UniverDrawingPlugin),
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector)),
	__decorateParam(2, _univerjs_core.IConfigService)
], UniverSheetsDrawingPlugin);

//#endregion
exports.ClearSheetDrawingTransformerOperation = ClearSheetDrawingTransformerOperation;
exports.DrawingApplyType = DrawingApplyType;
exports.ISheetDrawingService = ISheetDrawingService;
exports.InsertSheetDrawingCommand = InsertSheetDrawingCommand;
exports.RemoveSheetDrawingCommand = RemoveSheetDrawingCommand;
exports.SHEET_DRAWING_PLUGIN = SHEET_DRAWING_PLUGIN;
exports.SetDrawingApplyMutation = SetDrawingApplyMutation;
exports.SetDrawingArrangeCommand = SetDrawingArrangeCommand;
exports.SetSheetDrawingCommand = SetSheetDrawingCommand;
exports.SheetDrawingAnchorType = SheetDrawingAnchorType;
Object.defineProperty(exports, 'UniverSheetsDrawingPlugin', {
  enumerable: true,
  get: function () {
    return UniverSheetsDrawingPlugin;
  }
});
exports.drawingPositionToTransform = drawingPositionToTransform;
exports.transformToAxisAlignPosition = transformToAxisAlignPosition;
exports.transformToDrawingPosition = transformToDrawingPosition;
