// @univerjs/docs/lib/index.js
import { BuildTextUtils, CommandType, DOCS_FORMULA_BAR_EDITOR_UNIT_ID_KEY, DOCS_NORMAL_EDITOR_UNIT_ID_KEY, Disposable, DisposableCollection, ICommandService, IConfigService, IUniverInstanceService, Inject, Injector, JSONX, LocaleService, Plugin, RxDisposable, UniverInstanceType, composeInterceptors, createInterceptorKey, isInternalEditorID, merge, remove, toDisposable } from "@univerjs/core";
import { DocumentSkeleton, DocumentViewModel, IRenderManagerService, NORMAL_TEXT_SELECTION_PLUGIN_STYLE } from "@univerjs/engine-render";
import { BehaviorSubject, Subject, takeUntil } from "rxjs";

//#region src/commands/operations/text-selection.operation.ts
const SetTextSelectionsOperation = {
	id: "doc.operation.set-selections",
	type: CommandType.OPERATION,
	handler: () => {
		return true;
	}
};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/services/doc-selection-manager.service.ts
let DocSelectionManagerService = class DocSelectionManagerService extends RxDisposable {
	constructor(_commandService, _univerInstanceService) {
		super();
		this._commandService = _commandService;
		this._univerInstanceService = _univerInstanceService;
		_defineProperty(this, "_currentSelection", null);
		_defineProperty(this, "_textSelectionInfo", /* @__PURE__ */ new Map());
		_defineProperty(this, "_textSelection$", new Subject());
		_defineProperty(this, "textSelection$", this._textSelection$.asObservable());
		_defineProperty(this, "_refreshSelection$", new BehaviorSubject(null));
		_defineProperty(this, "refreshSelection$", this._refreshSelection$.asObservable());
		this._listenCurrentUnit();
	}
	_listenCurrentUnit() {
		this._univerInstanceService.getCurrentTypeOfUnit$(UniverInstanceType.UNIVER_DOC).pipe(takeUntil(this.dispose$)).subscribe((documentModel) => {
			if (documentModel == null) return;
			const unitId = documentModel.getUnitId();
			this._setCurrentSelectionNotRefresh({
				unitId,
				subUnitId: unitId
			});
		});
	}
	__getCurrentSelection() {
		return this._currentSelection;
	}
	getSelectionInfo(params = this._currentSelection) {
		return this._getTextRanges(params);
	}
	refreshSelection(params = this._currentSelection) {
		if (params == null) return;
		this._refresh(params);
	}
	__TEST_ONLY_setCurrentSelection(param) {
		this._currentSelection = param;
		this._refresh(param);
	}
	getTextRanges(params = this._currentSelection) {
		var _this$_getTextRanges;
		return (_this$_getTextRanges = this._getTextRanges(params)) === null || _this$_getTextRanges === void 0 ? void 0 : _this$_getTextRanges.textRanges;
	}
	getRectRanges(params = this._currentSelection) {
		var _this$_getTextRanges2;
		return (_this$_getTextRanges2 = this._getTextRanges(params)) === null || _this$_getTextRanges2 === void 0 ? void 0 : _this$_getTextRanges2.rectRanges;
	}
	getDocRanges(params = this._currentSelection) {
		var _this$getTextRanges, _this$getRectRanges;
		const textRanges = (_this$getTextRanges = this.getTextRanges(params)) !== null && _this$getTextRanges !== void 0 ? _this$getTextRanges : [];
		const rectRanges = (_this$getRectRanges = this.getRectRanges(params)) !== null && _this$getRectRanges !== void 0 ? _this$getRectRanges : [];
		return [...textRanges, ...rectRanges].filter((range) => range.startOffset != null && range.endOffset != null).sort((a, b) => {
			if (a.startOffset > b.startOffset) return 1;
			else if (a.startOffset < b.startOffset) return -1;
			else return 0;
		});
	}
	getActiveTextRange() {
		const selectionInfo = this._getTextRanges(this._currentSelection);
		if (selectionInfo == null) return;
		const { textRanges } = selectionInfo;
		return textRanges.find((textRange) => textRange.isActive);
	}
	/**
	*
	* @deprecated
	*/
	getActiveRectRange() {
		const selectionInfo = this._getTextRanges(this._currentSelection);
		if (selectionInfo == null) return;
		const { rectRanges } = selectionInfo;
		return rectRanges.find((rectRange) => rectRange.isActive);
	}
	__TEST_ONLY_add(textRanges, isEditing = true) {
		if (this._currentSelection == null) return;
		this._addByParam({
			...this._currentSelection,
			textRanges,
			rectRanges: [],
			segmentId: "",
			segmentPage: -1,
			isEditing,
			style: NORMAL_TEXT_SELECTION_PLUGIN_STYLE
		});
	}
	/**
	* @deprecated pls use replaceDocRanges.
	*/
	replaceTextRanges(docRanges, isEditing = true, options) {
		return this.replaceDocRanges(docRanges, this._currentSelection, isEditing, options);
	}
	replaceDocRanges(docRanges, params = this._currentSelection, isEditing = true, options) {
		if (params == null) return;
		const { unitId, subUnitId } = params;
		this._refreshSelection$.next({
			unitId,
			subUnitId,
			docRanges,
			isEditing,
			options
		});
	}
	__replaceTextRangesWithNoRefresh(textSelectionInfo, search) {
		if (this._currentSelection == null) return;
		const params = {
			...textSelectionInfo,
			...search
		};
		this._replaceByParam(params);
		this._textSelection$.next(params);
		const { unitId, subUnitId, segmentId, style, textRanges, rectRanges, isEditing } = params;
		const ranges = [...textRanges, ...rectRanges].filter((range) => range.startOffset != null && range.endOffset != null).sort((a, b) => {
			if (a.startOffset > b.startOffset) return 1;
			else if (a.startOffset < b.startOffset) return -1;
			else return 0;
		});
		this._commandService.executeCommand(SetTextSelectionsOperation.id, {
			unitId,
			subUnitId,
			segmentId,
			style,
			isEditing,
			ranges
		});
	}
	dispose() {
		this._textSelection$.complete();
		this._refreshSelection$.complete();
	}
	_setCurrentSelectionNotRefresh(param) {
		this._currentSelection = param;
	}
	_getTextRanges(param) {
		var _this$_textSelectionI;
		if (param == null) return;
		const { unitId, subUnitId = "" } = param;
		return (_this$_textSelectionI = this._textSelectionInfo.get(unitId)) === null || _this$_textSelectionI === void 0 ? void 0 : _this$_textSelectionI.get(subUnitId);
	}
	_refresh(param) {
		const allTextSelectionInfo = this._getTextRanges(param);
		if (allTextSelectionInfo == null) return;
		const { textRanges, rectRanges } = allTextSelectionInfo;
		const docRanges = [...textRanges, ...rectRanges];
		const { unitId, subUnitId } = param;
		this._refreshSelection$.next({
			unitId,
			subUnitId,
			docRanges,
			isEditing: false
		});
	}
	_replaceByParam(insertParam) {
		const { unitId, subUnitId, ...selectionInsertParam } = insertParam;
		if (!this._textSelectionInfo.has(unitId)) this._textSelectionInfo.set(unitId, /* @__PURE__ */ new Map());
		this._textSelectionInfo.get(unitId).set(subUnitId, { ...selectionInsertParam });
	}
	_addByParam(insertParam) {
		const { unitId, subUnitId, ...selectionInsertParam } = insertParam;
		if (!this._textSelectionInfo.has(unitId)) this._textSelectionInfo.set(unitId, /* @__PURE__ */ new Map());
		const unitTextRange = this._textSelectionInfo.get(unitId);
		if (!unitTextRange.has(subUnitId)) unitTextRange.set(subUnitId, { ...selectionInsertParam });
		else unitTextRange.get(subUnitId).textRanges.push(...insertParam.textRanges);
	}
};
DocSelectionManagerService = __decorate([__decorateParam(0, ICommandService), __decorateParam(1, IUniverInstanceService)], DocSelectionManagerService);

//#endregion
//#region src/services/doc-skeleton-manager.service.ts
let DocSkeletonManagerService = class DocSkeletonManagerService extends RxDisposable {
	constructor(_context, _localeService, _univerInstanceService) {
		super();
		this._context = _context;
		this._localeService = _localeService;
		this._univerInstanceService = _univerInstanceService;
		_defineProperty(this, "_skeleton", void 0);
		_defineProperty(this, "_docViewModel", void 0);
		_defineProperty(this, "_currentSkeleton$", new BehaviorSubject(null));
		_defineProperty(this, "currentSkeleton$", this._currentSkeleton$.asObservable());
		_defineProperty(this, "_currentSkeletonBefore$", new BehaviorSubject(null));
		_defineProperty(this, "currentSkeletonBefore$", this._currentSkeletonBefore$.asObservable());
		_defineProperty(this, "_currentViewModel$", new BehaviorSubject(null));
		_defineProperty(this, "currentViewModel$", this._currentViewModel$.asObservable());
		this._init();
		this._univerInstanceService.getCurrentTypeOfUnit$(UniverInstanceType.UNIVER_DOC).pipe(takeUntil(this.dispose$)).subscribe((documentModel) => {
			if (documentModel && documentModel.getUnitId() === this._context.unitId) this._update(documentModel);
		});
	}
	dispose() {
		super.dispose();
		this._currentSkeletonBefore$.complete();
		this._currentSkeleton$.complete();
	}
	getSkeleton() {
		return this._skeleton;
	}
	getViewModel() {
		return this._docViewModel;
	}
	_init() {
		const documentDataModel = this._context.unit;
		this._update(documentDataModel);
	}
	_update(documentDataModel) {
		const unitId = this._context.unitId;
		if (documentDataModel.getBody() == null) return;
		if (this._docViewModel && isInternalEditorID(unitId)) {
			this._docViewModel.reset(documentDataModel);
			this._context.unit = documentDataModel;
		} else if (!this._docViewModel) this._docViewModel = this._buildDocViewModel(documentDataModel);
		if (!this._skeleton) this._skeleton = this._buildSkeleton(this._docViewModel);
		const skeleton = this._skeleton;
		skeleton.calculate();
		this._currentSkeletonBefore$.next(skeleton);
		this._currentSkeleton$.next(skeleton);
		this._currentViewModel$.next(this._docViewModel);
	}
	_buildSkeleton(documentViewModel) {
		return DocumentSkeleton.create(documentViewModel, this._localeService);
	}
	_buildDocViewModel(documentDataModel) {
		return new DocumentViewModel(documentDataModel);
	}
};
DocSkeletonManagerService = __decorate([__decorateParam(1, Inject(LocaleService)), __decorateParam(2, IUniverInstanceService)], DocSkeletonManagerService);

//#endregion
//#region src/services/doc-state-emit.service.ts
var DocStateEmitService = class extends RxDisposable {
	constructor() {
		super();
		_defineProperty(this, "_docStateChangeParams$", new BehaviorSubject(null));
		_defineProperty(this, "docStateChangeParams$", this._docStateChangeParams$.asObservable());
	}
	emitStateChangeInfo(params) {
		this._docStateChangeParams$.next(params);
	}
	dispose() {
		super.dispose();
		this._docStateChangeParams$.complete();
	}
};

//#endregion
//#region src/commands/mutations/core-editing.mutation.ts
const RichTextEditingMutationId = "doc.mutation.rich-text-editing";
/**
* The core mutator to change rich text actions. The execution result would be undo mutation params. Could be directly
* send to undo redo service (will be used by the triggering command).
*/
const RichTextEditingMutation = {
	id: RichTextEditingMutationId,
	type: CommandType.MUTATION,
	handler: (accessor, params) => {
		var _renderManagerService, _docSelectionManagerS;
		const { unitId, segmentId = "", actions, textRanges, prevTextRanges, trigger, noHistory, isCompositionEnd, noNeedSetTextRange, debounce, isEditing = true, isSync, syncer } = params;
		const univerInstanceService = accessor.get(IUniverInstanceService);
		const renderManagerService = accessor.get(IRenderManagerService);
		const docStateEmitService = accessor.get(DocStateEmitService);
		const documentDataModel = univerInstanceService.getUniverDocInstance(unitId);
		const documentViewModel = (_renderManagerService = renderManagerService.getRenderById(unitId)) === null || _renderManagerService === void 0 ? void 0 : _renderManagerService.with(DocSkeletonManagerService).getViewModel();
		if (documentDataModel == null || documentViewModel == null) throw new Error(`DocumentDataModel or documentViewModel not found for unitId: ${unitId}`);
		const docSelectionManagerService = accessor.get(DocSelectionManagerService);
		const docRanges = (_docSelectionManagerS = docSelectionManagerService.getDocRanges()) !== null && _docSelectionManagerS !== void 0 ? _docSelectionManagerS : [];
		const disabled = !!documentDataModel.getSnapshot().disabled;
		if (JSONX.isNoop(actions) || actions && actions.length === 0 || disabled) return {
			unitId,
			actions: [],
			textRanges: docRanges
		};
		const undoActions = JSONX.invertWithDoc(actions, documentDataModel.getSnapshot());
		documentDataModel.apply(actions);
		documentViewModel.reset(documentDataModel);
		if (!noNeedSetTextRange && textRanges && trigger != null && !isSync) queueMicrotask(() => {
			docSelectionManagerService.replaceDocRanges(textRanges, {
				unitId,
				subUnitId: unitId
			}, isEditing, params.options);
		});
		const changeState = {
			commandId: RichTextEditingMutationId,
			unitId,
			segmentId,
			trigger,
			noHistory,
			debounce,
			redoState: {
				actions,
				textRanges
			},
			undoState: {
				actions: undoActions,
				textRanges: prevTextRanges !== null && prevTextRanges !== void 0 ? prevTextRanges : docRanges
			},
			isCompositionEnd,
			isSync,
			syncer
		};
		docStateEmitService.emitStateChangeInfo(changeState);
		return {
			unitId,
			actions: undoActions,
			textRanges: docRanges
		};
	}
};

//#endregion
//#region package.json
var name = "@univerjs/docs";
var version = "0.25.1";

//#endregion
//#region src/commands/mutations/docs-rename.mutation.ts
const DocsRenameMutation = {
	id: "doc.mutation.rename-doc",
	type: CommandType.MUTATION,
	handler: (accessor, params) => {
		const doc = accessor.get(IUniverInstanceService).getUnit(params.unitId, UniverInstanceType.UNIVER_DOC);
		if (!doc) return false;
		doc.setName(params.name);
		return true;
	}
};

//#endregion
//#region src/config/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const DOCS_PLUGIN_CONFIG_KEY = "docs.config";
const configSymbol = Symbol(DOCS_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/controllers/custom-range.controller.ts
let DocCustomRangeController = class DocCustomRangeController extends Disposable {
	constructor(_commandService, _textSelectionManagerService, _univerInstanceService) {
		super();
		this._commandService = _commandService;
		this._textSelectionManagerService = _textSelectionManagerService;
		this._univerInstanceService = _univerInstanceService;
		this._initSelectionChange();
	}
	_transformCustomRange(doc, selection) {
		var _doc$getCustomRanges;
		const { startOffset, endOffset, collapsed } = selection;
		const customRanges = (_doc$getCustomRanges = doc.getCustomRanges()) === null || _doc$getCustomRanges === void 0 ? void 0 : _doc$getCustomRanges.filter((range) => {
			if (!range.wholeEntity) return false;
			if (startOffset <= range.startIndex && endOffset > range.endIndex) return false;
			if (collapsed) return range.startIndex < startOffset && range.endIndex >= endOffset;
			return BuildTextUtils.range.isIntersects(startOffset, endOffset - 1, range.startIndex, range.endIndex);
		});
		if (customRanges === null || customRanges === void 0 ? void 0 : customRanges.length) {
			let start = startOffset;
			let end = endOffset;
			customRanges.forEach((range) => {
				start = Math.min(range.startIndex, start);
				end = Math.max(range.endIndex + 1, end);
			});
			return {
				...selection,
				startOffset: start,
				endOffset: end,
				collapsed: start === end
			};
		}
		return selection;
	}
	_initSelectionChange() {
		this.disposeWithMe(this._commandService.onCommandExecuted((commandInfo) => {
			if (commandInfo.id === SetTextSelectionsOperation.id) {
				const { unitId, ranges, isEditing } = commandInfo.params;
				const doc = this._univerInstanceService.getUnit(unitId);
				if (!doc) return;
				const transformedRanges = ranges.map((range) => this._transformCustomRange(doc, range));
				if (transformedRanges.some((range, i) => ranges[i] !== range)) this._textSelectionManagerService.replaceTextRanges(transformedRanges, isEditing);
			}
		}));
	}
};
DocCustomRangeController = __decorate([
	__decorateParam(0, ICommandService),
	__decorateParam(1, Inject(DocSelectionManagerService)),
	__decorateParam(2, IUniverInstanceService)
], DocCustomRangeController);

//#endregion
//#region src/plugin.ts
let UniverDocsPlugin = class UniverDocsPlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _injector, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		const { ...rest } = merge({}, defaultPluginConfig, this._config);
		this._configService.setConfig(DOCS_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		this._initializeDependencies();
		this._initializeCommands();
	}
	_initializeCommands() {
		[
			RichTextEditingMutation,
			DocsRenameMutation,
			SetTextSelectionsOperation
		].forEach((command) => {
			this._injector.get(ICommandService).registerCommand(command);
		});
	}
	_initializeDependencies() {
		[
			[DocSelectionManagerService],
			[DocStateEmitService],
			[DocCustomRangeController]
		].forEach((d) => this._injector.add(d));
	}
	onReady() {
		this._injector.get(DocCustomRangeController);
	}
};
_defineProperty(UniverDocsPlugin, "pluginName", "DOCS_PLUGIN");
_defineProperty(UniverDocsPlugin, "packageName", name);
_defineProperty(UniverDocsPlugin, "version", version);
UniverDocsPlugin = __decorate([__decorateParam(1, Inject(Injector)), __decorateParam(2, IConfigService)], UniverDocsPlugin);

//#endregion
//#region src/services/doc-interceptor/interceptor-const.ts
const CUSTOM_RANGE = createInterceptorKey("CUSTOM_RANGE");
const CUSTOM_DECORATION = createInterceptorKey("CUSTOM_DECORATION");
const DOC_INTERCEPTOR_POINT = {
	CUSTOM_RANGE,
	CUSTOM_DECORATION
};

//#endregion
//#region src/services/doc-interceptor/doc-interceptor.service.ts
let DocInterceptorService = class DocInterceptorService extends Disposable {
	constructor(_context, _docSkeletonManagerService) {
		super();
		this._context = _context;
		this._docSkeletonManagerService = _docSkeletonManagerService;
		_defineProperty(this, "_interceptorsByName", /* @__PURE__ */ new Map());
		const viewModel = this._docSkeletonManagerService.getViewModel();
		const unitId = viewModel.getDataModel().getUnitId();
		if (unitId === DOCS_NORMAL_EDITOR_UNIT_ID_KEY || unitId === DOCS_FORMULA_BAR_EDITOR_UNIT_ID_KEY) return;
		this.disposeWithMe(this.interceptDocumentViewModel(viewModel));
		this.disposeWithMe(this.intercept(DOC_INTERCEPTOR_POINT.CUSTOM_RANGE, {
			priority: -1,
			handler: (data, pos, next) => {
				return next(data);
			}
		}));
		let disposableCollection = new DisposableCollection();
		viewModel.segmentViewModels$.subscribe((segmentViewModels) => {
			disposableCollection.dispose();
			disposableCollection = new DisposableCollection();
			segmentViewModels.forEach((segmentViewModel) => {
				disposableCollection.add(this.interceptDocumentViewModel(segmentViewModel));
			});
		});
		this.disposeWithMe(disposableCollection);
	}
	intercept(name, interceptor) {
		const key = name;
		if (!this._interceptorsByName.has(key)) this._interceptorsByName.set(key, []);
		const interceptors = this._interceptorsByName.get(key);
		interceptors.push(interceptor);
		this._interceptorsByName.set(key, interceptors.sort((a, b) => {
			var _b$priority, _a$priority;
			return ((_b$priority = b.priority) !== null && _b$priority !== void 0 ? _b$priority : 0) - ((_a$priority = a.priority) !== null && _a$priority !== void 0 ? _a$priority : 0);
		}));
		return this.disposeWithMe(toDisposable(() => remove(this._interceptorsByName.get(key), interceptor)));
	}
	fetchThroughInterceptors(name) {
		const key = name;
		return composeInterceptors(this._interceptorsByName.get(key) || []);
	}
	interceptDocumentViewModel(viewModel) {
		const disposableCollection = new DisposableCollection();
		disposableCollection.add(viewModel.registerCustomRangeInterceptor({
			getCustomRange: (index) => {
				var _viewModel$getDataMod;
				return this.fetchThroughInterceptors(DOC_INTERCEPTOR_POINT.CUSTOM_RANGE)(viewModel.getCustomRangeRaw(index), {
					index,
					unitId: viewModel.getDataModel().getUnitId(),
					customRanges: (_viewModel$getDataMod = viewModel.getDataModel().getCustomRanges()) !== null && _viewModel$getDataMod !== void 0 ? _viewModel$getDataMod : []
				});
			},
			getCustomDecoration: (index) => {
				var _viewModel$getDataMod2;
				return this.fetchThroughInterceptors(DOC_INTERCEPTOR_POINT.CUSTOM_DECORATION)(viewModel.getCustomDecorationRaw(index), {
					index,
					unitId: viewModel.getDataModel().getUnitId(),
					customDecorations: (_viewModel$getDataMod2 = viewModel.getDataModel().getCustomDecorations()) !== null && _viewModel$getDataMod2 !== void 0 ? _viewModel$getDataMod2 : []
				});
			}
		}));
		return disposableCollection;
	}
};
DocInterceptorService = __decorate([__decorateParam(1, Inject(DocSkeletonManagerService))], DocInterceptorService);

//#endregion
//#region src/utils/custom-range-factory.ts
/**
* @deprecated This is a duplication from docs-ui to avoid making too much breaking changes.
*/
function getRichTextEditPath(docDataModel, segmentId = "") {
	if (!segmentId) return ["body"];
	const { headers, footers } = docDataModel.getSnapshot();
	if (headers == null && footers == null) throw new Error("Document data model must have headers or footers when update by segment id");
	if ((headers === null || headers === void 0 ? void 0 : headers[segmentId]) != null) return [
		"headers",
		segmentId,
		"body"
	];
	else if ((footers === null || footers === void 0 ? void 0 : footers[segmentId]) != null) return [
		"footers",
		segmentId,
		"body"
	];
	else throw new Error("Segment id not found in headers or footers");
}
function addCustomRangeFactory(accessor, param, body) {
	const { unitId, segmentId } = param;
	const documentDataModel = accessor.get(IUniverInstanceService).getUnit(unitId);
	if (!documentDataModel) return false;
	const doMutation = {
		id: RichTextEditingMutation.id,
		params: {
			unitId: param.unitId,
			actions: [],
			textRanges: void 0
		}
	};
	const jsonX = JSONX.getInstance();
	const textX = BuildTextUtils.customRange.add({
		...param,
		body
	});
	if (!textX) return false;
	const path = getRichTextEditPath(documentDataModel, segmentId);
	doMutation.params.actions = jsonX.editOp(textX.serialize(), path);
	return doMutation;
}
function addCustomRangeBySelectionFactory(accessor, param) {
	var _selections$;
	const { rangeId, rangeType, wholeEntity, properties, unitId, selections: propSelection } = param;
	const docSelectionManagerService = accessor.get(DocSelectionManagerService);
	const univerInstanceService = accessor.get(IUniverInstanceService);
	const selections = propSelection !== null && propSelection !== void 0 ? propSelection : docSelectionManagerService.getTextRanges({
		unitId,
		subUnitId: unitId
	});
	const segmentId = selections === null || selections === void 0 || (_selections$ = selections[0]) === null || _selections$ === void 0 ? void 0 : _selections$.segmentId;
	if (!(selections === null || selections === void 0 ? void 0 : selections.length)) return false;
	const documentDataModel = univerInstanceService.getUnit(unitId, UniverInstanceType.UNIVER_DOC);
	if (!documentDataModel) return false;
	const body = documentDataModel.getSelfOrHeaderFooterModel(segmentId).getBody();
	if (!body) return false;
	const textX = BuildTextUtils.customRange.add({
		ranges: selections,
		rangeId,
		rangeType,
		segmentId,
		wholeEntity,
		properties,
		body
	});
	if (!textX) return false;
	const jsonX = JSONX.getInstance();
	const doMutation = {
		id: RichTextEditingMutation.id,
		params: {
			unitId,
			actions: [],
			textRanges: textX.selections,
			segmentId
		},
		textX
	};
	const path = getRichTextEditPath(documentDataModel, segmentId);
	doMutation.params.actions = jsonX.editOp(textX.serialize(), path);
	return doMutation;
}
function deleteCustomRangeFactory(accessor, params) {
	const { unitId, segmentId, insert } = params;
	const documentDataModel = accessor.get(IUniverInstanceService).getUnit(unitId);
	if (!documentDataModel) return false;
	const doMutation = {
		id: RichTextEditingMutation.id,
		params: {
			unitId: params.unitId,
			actions: [],
			textRanges: void 0,
			segmentId
		}
	};
	const jsonX = JSONX.getInstance();
	const textX = BuildTextUtils.customRange.delete({
		documentDataModel,
		rangeId: params.rangeId,
		insert,
		segmentId
	});
	if (!textX) return false;
	const path = getRichTextEditPath(documentDataModel, segmentId);
	doMutation.params.actions = jsonX.editOp(textX.serialize(), path);
	doMutation.params.textRanges = textX.selections;
	return doMutation;
}

//#endregion
//#region src/utils/replace-selection-factory.ts
function replaceSelectionFactory(accessor, params) {
	var _params$selection, _docDataModel$getSelf, _params$selection2, _params$textRanges;
	const { unitId, body: insertBody, doc } = params;
	let docDataModel = doc;
	if (!docDataModel) docDataModel = accessor.get(IUniverInstanceService).getUnit(unitId);
	if (!docDataModel) return false;
	const segmentId = (_params$selection = params.selection) === null || _params$selection === void 0 ? void 0 : _params$selection.segmentId;
	const body = (_docDataModel$getSelf = docDataModel.getSelfOrHeaderFooterModel(segmentId)) === null || _docDataModel$getSelf === void 0 ? void 0 : _docDataModel$getSelf.getBody();
	if (!body) return false;
	const docSelectionManagerService = accessor.get(DocSelectionManagerService);
	const selection = (_params$selection2 = params.selection) !== null && _params$selection2 !== void 0 ? _params$selection2 : docSelectionManagerService.getActiveTextRange();
	if (!selection || !body) return false;
	const textRanges = (_params$textRanges = params.textRanges) !== null && _params$textRanges !== void 0 ? _params$textRanges : [{
		startOffset: selection.startOffset + insertBody.dataStream.length,
		endOffset: selection.startOffset + insertBody.dataStream.length,
		collapsed: true,
		segmentId
	}];
	const textX = BuildTextUtils.selection.replace({
		selection,
		body: insertBody,
		doc: docDataModel
	});
	if (!textX) return false;
	const doMutation = {
		id: RichTextEditingMutation.id,
		params: {
			unitId,
			actions: [],
			textRanges,
			debounce: true,
			segmentId
		},
		textX
	};
	const jsonX = JSONX.getInstance();
	doMutation.params.actions = jsonX.editOp(textX.serialize());
	return doMutation;
}

//#endregion
export { DOC_INTERCEPTOR_POINT, DocInterceptorService, DocSelectionManagerService, DocSkeletonManagerService, DocStateEmitService, RichTextEditingMutation, SetTextSelectionsOperation, UniverDocsPlugin, addCustomRangeBySelectionFactory, addCustomRangeFactory, deleteCustomRangeFactory, replaceSelectionFactory };
