// @univerjs/sheets-thread-comment/lib/facade.js
import { CanceledError, ICommandService, IUniverInstanceService, Inject, Injector, Range, RichTextBuilder, RichTextValue, Tools, UniverInstanceType, UserManagerService, generateRandomId, toDisposable } from "@univerjs/core";
import { SheetsThreadCommentModel } from "@univerjs/sheets-thread-comment";
import { FRange, FWorkbook, FWorksheet } from "@univerjs/sheets/facade";
import { AddCommentCommand, DeleteCommentCommand, DeleteCommentTreeCommand, ResolveCommentCommand, ThreadCommentModel, UpdateCommentCommand, getDT } from "@univerjs/thread-comment";
import { deserializeRangeWithSheet } from "@univerjs/engine-formula";
import { filter } from "rxjs";
import { FEventName, FUniver } from "@univerjs/core/facade";

//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/facade/f-thread-comment.ts
var _FThreadComment;
/**
* An readonly class that represents a comment.
* @ignore
*/
var FTheadCommentItem = class FTheadCommentItem {
	/**
	* Create a new FTheadCommentItem
	* @param {IThreadComment|undefined} comment The comment
	* @returns {FTheadCommentItem} A new instance of FTheadCommentItem
	* @example
	* ```ts
	* const commentBuilder = univerAPI.newTheadComment();
	* console.log(commentBuilder);
	* ```
	*/
	static create(comment) {
		return new FTheadCommentItem(comment);
	}
	constructor(comment) {
		_defineProperty(this, "_comment", {
			id: generateRandomId(),
			ref: "",
			threadId: "",
			dT: "",
			personId: "",
			text: RichTextBuilder.newEmptyData().body,
			attachments: [],
			unitId: "",
			subUnitId: ""
		});
		if (comment) this._comment = comment;
	}
	/**
	* Get the person id of the comment
	* @returns {string} The person id of the comment
	* @example
	* ```ts
	* const commentBuilder = univerAPI.newTheadComment();
	* console.log(commentBuilder.personId);
	* ```
	*/
	get personId() {
		return this._comment.personId;
	}
	/**
	* Get the date time of the comment
	* @returns {string} The date time of the comment
	* @example
	* ```ts
	* const commentBuilder = univerAPI.newTheadComment();
	* console.log(commentBuilder.dateTime);
	* ```
	*/
	get dateTime() {
		return this._comment.dT;
	}
	/**
	* Get the content of the comment
	* @returns {RichTextValue} The content of the comment
	* @example
	* ```ts
	* const commentBuilder = univerAPI.newTheadComment();
	* console.log(commentBuilder.content);
	* ```
	*/
	get content() {
		return RichTextValue.createByBody(this._comment.text);
	}
	/**
	* Get the id of the comment
	* @returns {string} The id of the comment
	* @example
	* ```ts
	* const commentBuilder = univerAPI.newTheadComment();
	* console.log(commentBuilder.id);
	* ```
	*/
	get id() {
		return this._comment.id;
	}
	/**
	* Get the thread id of the comment
	* @returns {string} The thread id of the comment
	* @example
	* ```ts
	* const commentBuilder = univerAPI.newTheadComment();
	* console.log(commentBuilder.threadId);
	* ```
	*/
	get threadId() {
		return this._comment.threadId;
	}
	/**
	* Copy the comment
	* @returns {FTheadCommentBuilder} The comment builder
	* @example
	* ```ts
	* const commentBuilder = univerAPI.newTheadComment();
	* const newCommentBuilder = commentBuilder.copy();
	* console.log(newCommentBuilder);
	* ```
	*/
	copy() {
		return FTheadCommentBuilder.create(Tools.deepClone(this._comment));
	}
};
/**
* A builder for thread comment. use {@link FUniver} `univerAPI.newTheadComment()` to create a new builder.
*/
var FTheadCommentBuilder = class FTheadCommentBuilder extends FTheadCommentItem {
	static create(comment) {
		return new FTheadCommentBuilder(comment);
	}
	/**
	* Set the content of the comment
	* @param {IDocumentBody | RichTextValue} content The content of the comment
	* @returns {FTheadCommentBuilder} The comment builder for chaining
	* @example
	* ```ts
	* // Create a new comment
	* const richText = univerAPI.newRichText().insertText('hello univer');
	* const commentBuilder = univerAPI.newTheadComment()
	*   .setContent(richText);
	* console.log(commentBuilder.content);
	*
	* // Add the comment to the cell A1
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const cell = fWorksheet.getRange('A1');
	* const result = await cell.addCommentAsync(commentBuilder);
	* console.log(result);
	* ```
	*/
	setContent(content) {
		if (content instanceof RichTextValue) this._comment.text = content.getData().body;
		else this._comment.text = content;
		return this;
	}
	/**
	* Set the person id of the comment
	* @param {string} userId The person id of the comment
	* @returns {FTheadCommentBuilder} The comment builder for chaining
	* @example
	* ```ts
	* // Create a new comment
	* const richText = univerAPI.newRichText().insertText('hello univer');
	* const commentBuilder = univerAPI.newTheadComment()
	*   .setContent(richText)
	*   .setPersonId('mock-user-id');
	* console.log(commentBuilder.personId);
	*
	* // Add the comment to the cell A1
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const cell = fWorksheet.getRange('A1');
	* const result = await cell.addCommentAsync(commentBuilder);
	* console.log(result);
	* ```
	*/
	setPersonId(userId) {
		this._comment.personId = userId;
		return this;
	}
	/**
	* Set the date time of the comment
	* @param {Date} date The date time of the comment
	* @returns {FTheadCommentBuilder} The comment builder for chaining
	* @example
	* ```ts
	* // Create a new comment
	* const richText = univerAPI.newRichText().insertText('hello univer');
	* const commentBuilder = univerAPI.newTheadComment()
	*   .setContent(richText)
	*   .setDateTime(new Date('2025-02-21 14:22:22'));
	* console.log(commentBuilder.dateTime);
	*
	* // Add the comment to the cell A1
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const cell = fWorksheet.getRange('A1');
	* const result = await cell.addCommentAsync(commentBuilder);
	* console.log(result);
	* ```
	*/
	setDateTime(date) {
		this._comment.dT = getDT(date);
		return this;
	}
	/**
	* Set the id of the comment
	* @param {string} id The id of the comment
	* @returns {FTheadCommentBuilder} The comment builder for chaining
	* @example
	* ```ts
	* // Create a new comment
	* const richText = univerAPI.newRichText().insertText('hello univer');
	* const commentBuilder = univerAPI.newTheadComment()
	*   .setContent(richText)
	*   .setId('mock-comment-id');
	* console.log(commentBuilder.id);
	*
	* // Add the comment to the cell A1
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const cell = fWorksheet.getRange('A1');
	* const result = await cell.addCommentAsync(commentBuilder);
	* console.log(result);
	* ```
	*/
	setId(id) {
		this._comment.id = id;
		return this;
	}
	/**
	* Set the thread id of the comment
	* @param {string} threadId The thread id of the comment
	* @returns {FTheadCommentBuilder} The comment builder
	* @example
	* ```ts
	* // Create a new comment
	* const richText = univerAPI.newRichText().insertText('hello univer');
	* const commentBuilder = univerAPI.newTheadComment()
	*   .setContent(richText)
	*   .setThreadId('mock-thread-id');
	* console.log(commentBuilder.threadId);
	*
	* // Add the comment to the cell A1
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const cell = fWorksheet.getRange('A1');
	* const result = await cell.addCommentAsync(commentBuilder);
	* console.log(result);
	* ```
	*/
	setThreadId(threadId) {
		this._comment.threadId = threadId;
		return this;
	}
	/**
	* Build the comment
	* @returns {IThreadComment} The comment
	* @example
	* ```ts
	* const richText = univerAPI.newRichText().insertText('hello univer');
	* const comment = univerAPI.newTheadComment()
	*   .setContent(richText)
	*   .setPersonId('mock-user-id')
	*   .setDateTime(new Date('2025-02-21 14:22:22'))
	*   .setId('mock-comment-id')
	*   .setThreadId('mock-thread-id')
	*   .build();
	* console.log(comment);
	* ```
	*/
	build() {
		return this._comment;
	}
};
let FThreadComment = _FThreadComment = class FThreadComment {
	/**
	* @ignore
	*/
	constructor(_thread, _parent, _injector, _commandService, _univerInstanceService, _threadCommentModel, _userManagerService) {
		this._thread = _thread;
		this._parent = _parent;
		this._injector = _injector;
		this._commandService = _commandService;
		this._univerInstanceService = _univerInstanceService;
		this._threadCommentModel = _threadCommentModel;
		this._userManagerService = _userManagerService;
	}
	_getRef() {
		var _this$_parent;
		return deserializeRangeWithSheet(((_this$_parent = this._parent) === null || _this$_parent === void 0 ? void 0 : _this$_parent.ref) || this._thread.ref).range;
	}
	/**
	* Whether the comment is a root comment
	* @returns {boolean} Whether the comment is a root comment
	* @example
	* ```ts
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const comments = fWorksheet.getComments();
	* comments.forEach((comment) => {
	*   console.log(comment.getIsRoot());
	* });
	* ```
	*/
	getIsRoot() {
		return !this._parent;
	}
	/**
	* Get the comment data
	* @returns {IBaseComment} The comment data
	* @example
	* ```ts
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const comments = fWorksheet.getComments();
	* comments.forEach((comment) => {
	*   console.log(comment.getCommentData());
	* });
	* ```
	*/
	getCommentData() {
		const { children, ...comment } = this._thread;
		return comment;
	}
	/**
	* Get the replies of the comment
	* @returns {FThreadComment[]} the replies of the comment
	* @example
	* ```ts
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const comments = fWorksheet.getComments();
	* comments.forEach((comment) => {
	*   if (comment.getIsRoot()) {
	*     const replies = comment.getReplies();
	*     replies.forEach((reply) => {
	*       console.log(reply.getCommentData());
	*     });
	*   }
	* });
	* ```
	*/
	getReplies() {
		var _comments$children;
		const range = this._getRef();
		const comments = this._threadCommentModel.getCommentWithChildren(this._thread.unitId, this._thread.subUnitId, range.startRow, range.startColumn);
		return comments === null || comments === void 0 || (_comments$children = comments.children) === null || _comments$children === void 0 ? void 0 : _comments$children.map((child) => this._injector.createInstance(_FThreadComment, child));
	}
	/**
	* Get the range of the comment
	* @returns {FRange | null} The range of the comment
	* @example
	* ```ts
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const comments = fWorksheet.getComments();
	* comments.forEach((comment) => {
	*   console.log(comment.getRange().getA1Notation());
	* });
	* ```
	*/
	getRange() {
		const workbook = this._univerInstanceService.getUnit(this._thread.unitId, UniverInstanceType.UNIVER_SHEET);
		if (!workbook) return null;
		const worksheet = workbook.getSheetBySheetId(this._thread.subUnitId);
		if (!worksheet) return null;
		const range = this._getRef();
		return this._injector.createInstance(FRange, workbook, worksheet, range);
	}
	/**
	* @deprecated use `getRichText` as instead
	*/
	getContent() {
		return this._thread.text;
	}
	/**
	* Get the rich text of the comment
	* @returns {RichTextValue} The rich text of the comment
	* @example
	* ```ts
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const comments = fWorksheet.getComments();
	* comments.forEach((comment) => {
	*   console.log(comment.getRichText());
	* });
	* ```
	*/
	getRichText() {
		const body = this._thread.text;
		return RichTextValue.create({
			body,
			documentStyle: {},
			id: "d"
		});
	}
	/**
	* Delete the comment and it's replies
	* @returns {Promise<boolean>} Whether the comment is deleted successfully
	* @example
	* ```ts
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const comments = fWorksheet.getComments();
	*
	* // Delete the first comment
	* const result = await comments[0]?.deleteAsync();
	* console.log(result);
	* ```
	*/
	deleteAsync() {
		return this._commandService.executeCommand(this.getIsRoot() ? DeleteCommentTreeCommand.id : DeleteCommentCommand.id, {
			commentId: this._thread.id,
			unitId: this._thread.unitId,
			subUnitId: this._thread.subUnitId
		});
	}
	/**
	* @deprecated use `deleteAsync` as instead.
	*/
	delete() {
		return this.deleteAsync();
	}
	/**
	* @deprecated use `updateAsync` as instead
	*/
	async update(content) {
		return this.updateAsync(content);
	}
	/**
	* Update the comment content
	* @param {IDocumentBody | RichTextValue} content The new content of the comment
	* @returns {Promise<boolean>} Whether the comment is updated successfully
	* @example
	* ```ts
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new comment
	* const richText = univerAPI.newRichText().insertText('hello univer');
	* const commentBuilder = univerAPI.newTheadComment()
	*   .setContent(richText)
	*   .setId('mock-comment-id');
	* const cell = fWorksheet.getRange('A1');
	* await cell.addCommentAsync(commentBuilder);
	*
	* // Update the comment after 3 seconds
	* setTimeout(async () => {
	*   const comment = fWorksheet.getCommentById('mock-comment-id');
	*   const newRichText = univerAPI.newRichText().insertText('Hello Univer AI');
	*   const result = await comment.updateAsync(newRichText);
	*   console.log(result);
	* }, 3000);
	* ```
	*/
	async updateAsync(content) {
		const body = content instanceof RichTextValue ? content.getData().body : content;
		const dt = getDT();
		return await this._commandService.executeCommand(UpdateCommentCommand.id, {
			unitId: this._thread.unitId,
			subUnitId: this._thread.subUnitId,
			payload: {
				commentId: this._thread.id,
				text: body,
				updated: true,
				updateT: dt
			}
		});
	}
	/**
	* @deprecated use `resolveAsync` as instead
	*/
	resolve(resolved) {
		return this.resolveAsync(resolved);
	}
	/**
	* Resolve the comment
	* @param {boolean} resolved Whether the comment is resolved
	* @returns {Promise<boolean>} Set the comment to resolved or not operation result
	* @example
	* ```ts
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new comment
	* const richText = univerAPI.newRichText().insertText('hello univer');
	* const commentBuilder = univerAPI.newTheadComment()
	*   .setContent(richText)
	*   .setId('mock-comment-id');
	* const cell = fWorksheet.getRange('A1');
	* await cell.addCommentAsync(commentBuilder);
	*
	* // Resolve the comment after 3 seconds
	* setTimeout(async () => {
	*   const comment = fWorksheet.getCommentById('mock-comment-id');
	*   const result = await comment.resolveAsync(true);
	*   console.log(result);
	* }, 3000);
	* ```
	*/
	resolveAsync(resolved) {
		return this._commandService.executeCommand(ResolveCommentCommand.id, {
			unitId: this._thread.unitId,
			subUnitId: this._thread.subUnitId,
			commentId: this._thread.id,
			resolved: resolved !== null && resolved !== void 0 ? resolved : !this._thread.resolved
		});
	}
	/**
	* Reply to the comment
	* @param {FTheadCommentBuilder} comment The comment to reply to
	* @returns {Promise<boolean>} Whether the comment is replied successfully
	* @example
	* ```ts
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Create a new comment
	* const richText = univerAPI.newRichText().insertText('hello univer');
	* const commentBuilder = univerAPI.newTheadComment()
	*   .setContent(richText)
	*   .setId('mock-comment-id');
	* const cell = fWorksheet.getRange('A1');
	* await cell.addCommentAsync(commentBuilder);
	*
	* // Reply to the comment
	* const replyText = univerAPI.newRichText().insertText('Hello Univer AI');
	* const reply = univerAPI.newTheadComment().setContent(replyText);
	* const comment = fWorksheet.getCommentById('mock-comment-id');
	* const result = await comment.replyAsync(reply);
	* console.log(result);
	* ```
	*/
	replyAsync(comment) {
		var _this$_parent2;
		const commentData = comment.build();
		return this._commandService.executeCommand(AddCommentCommand.id, {
			unitId: this._thread.unitId,
			subUnitId: this._thread.subUnitId,
			comment: {
				id: generateRandomId(),
				parentId: this._thread.id,
				threadId: this._thread.threadId,
				ref: ((_this$_parent2 = this._parent) === null || _this$_parent2 === void 0 ? void 0 : _this$_parent2.ref) || this._thread.ref,
				unitId: this._thread.unitId,
				subUnitId: this._thread.subUnitId,
				text: commentData.text,
				attachments: commentData.attachments,
				dT: commentData.dT || getDT(),
				personId: commentData.personId || this._userManagerService.getCurrentUser().userID
			}
		});
	}
};
FThreadComment = _FThreadComment = __decorate([
	__decorateParam(2, Inject(Injector)),
	__decorateParam(3, ICommandService),
	__decorateParam(4, IUniverInstanceService),
	__decorateParam(5, Inject(SheetsThreadCommentModel)),
	__decorateParam(6, Inject(UserManagerService))
], FThreadComment);

//#endregion
//#region src/facade/f-range.ts
/**
* @ignore
*/
var FRangeSheetsThreadCommentMixin = class extends FRange {
	getComment() {
		const sheetsTheadCommentModel = this._injector.get(SheetsThreadCommentModel);
		const unitId = this._workbook.getUnitId();
		const sheetId = this._worksheet.getSheetId();
		const commentId = sheetsTheadCommentModel.getByLocation(unitId, sheetId, this._range.startRow, this._range.startColumn);
		if (!commentId) return null;
		const comment = sheetsTheadCommentModel.getComment(unitId, sheetId, commentId);
		if (comment) return this._injector.createInstance(FThreadComment, comment);
		return null;
	}
	getComments() {
		const sheetsTheadCommentModel = this._injector.get(SheetsThreadCommentModel);
		const unitId = this._workbook.getUnitId();
		const sheetId = this._worksheet.getSheetId();
		const comments = [];
		Range.foreach(this._range, (row, col) => {
			const commentId = sheetsTheadCommentModel.getByLocation(unitId, sheetId, row, col);
			if (commentId) {
				const comment = sheetsTheadCommentModel.getComment(unitId, sheetId, commentId);
				if (comment) comments.push(this._injector.createInstance(FThreadComment, comment));
			}
		});
		return comments;
	}
	addComment(content) {
		var _this$getComment;
		const injector = this._injector;
		const currentComment = (_this$getComment = this.getComment()) === null || _this$getComment === void 0 ? void 0 : _this$getComment.getCommentData();
		const commentService = injector.get(ICommandService);
		const userService = injector.get(UserManagerService);
		const unitId = this._workbook.getUnitId();
		const sheetId = this._worksheet.getSheetId();
		const refStr = `${Tools.chatAtABC(this._range.startColumn)}${this._range.startRow + 1}`;
		const currentUser = userService.getCurrentUser();
		const commentData = content instanceof FTheadCommentBuilder ? content.build() : { text: content };
		return commentService.executeCommand(AddCommentCommand.id, {
			unitId,
			subUnitId: sheetId,
			comment: {
				text: commentData.text,
				dT: commentData.dT || getDT(),
				attachments: [],
				id: commentData.id || generateRandomId(),
				ref: refStr,
				personId: commentData.personId || currentUser.userID,
				parentId: currentComment === null || currentComment === void 0 ? void 0 : currentComment.id,
				unitId,
				subUnitId: sheetId,
				threadId: (currentComment === null || currentComment === void 0 ? void 0 : currentComment.threadId) || generateRandomId()
			}
		});
	}
	clearComment() {
		var _this$getComment2;
		const injector = this._injector;
		const currentComment = (_this$getComment2 = this.getComment()) === null || _this$getComment2 === void 0 ? void 0 : _this$getComment2.getCommentData();
		const commentService = injector.get(ICommandService);
		const unitId = this._workbook.getUnitId();
		const sheetId = this._worksheet.getSheetId();
		if (currentComment) return commentService.executeCommand(DeleteCommentTreeCommand.id, {
			unitId,
			subUnitId: sheetId,
			threadId: currentComment.threadId,
			commentId: currentComment.id
		});
		return Promise.resolve(true);
	}
	clearComments() {
		const promises = this.getComments().map((comment) => comment.deleteAsync());
		return Promise.all(promises).then(() => true);
	}
	addCommentAsync(content) {
		return this.addComment(content);
	}
	clearCommentAsync() {
		return this.clearComment();
	}
	clearCommentsAsync() {
		return this.clearComments();
	}
};
FRange.extend(FRangeSheetsThreadCommentMixin);

//#endregion
//#region src/facade/f-workbook.ts
/**
* @ignore
*/
var FWorkbookSheetsThreadCommentMixin = class extends FWorkbook {
	/**
	* @ignore
	*/
	_initialize() {
		Object.defineProperty(this, "_threadCommentModel", { get() {
			return this._injector.get(ThreadCommentModel);
		} });
	}
	getComments() {
		return this._threadCommentModel.getUnit(this._workbook.getUnitId()).map((i) => this._injector.createInstance(FThreadComment, i.root));
	}
	clearComments() {
		const promises = this.getComments().map((comment) => comment.deleteAsync());
		return Promise.all(promises).then(() => true);
	}
	/**
	* @param callback
	* @deprecated
	*/
	onThreadCommentChange(callback) {
		return toDisposable(this._threadCommentModel.commentUpdate$.pipe(filter((change) => change.unitId === this._workbook.getUnitId())).subscribe(callback));
	}
	/**
	* @param callback
	* @deprecated
	*/
	onBeforeAddThreadComment(callback) {
		return toDisposable(this._commandService.beforeCommandExecuted((commandInfo, options) => {
			const params = commandInfo.params;
			if (commandInfo.id === AddCommentCommand.id) {
				if (params.unitId !== this._workbook.getUnitId()) return;
				if (callback(params, options) === false) throw new Error("Command is stopped by the hook onBeforeAddThreadComment");
			}
		}));
	}
	/**
	* @param callback
	* @deprecated
	*/
	onBeforeUpdateThreadComment(callback) {
		return toDisposable(this._commandService.beforeCommandExecuted((commandInfo, options) => {
			const params = commandInfo.params;
			if (commandInfo.id === UpdateCommentCommand.id) {
				if (params.unitId !== this._workbook.getUnitId()) return;
				if (callback(params, options) === false) throw new Error("Command is stopped by the hook onBeforeUpdateThreadComment");
			}
		}));
	}
	/**
	* @param callback
	* @deprecated
	*/
	onBeforeDeleteThreadComment(callback) {
		return toDisposable(this._commandService.beforeCommandExecuted((commandInfo, options) => {
			const params = commandInfo.params;
			if (commandInfo.id === DeleteCommentCommand.id || commandInfo.id === DeleteCommentTreeCommand.id) {
				if (params.unitId !== this._workbook.getUnitId()) return;
				if (callback(params, options) === false) throw new Error("Command is stopped by the hook onBeforeDeleteThreadComment");
			}
		}));
	}
};
FWorkbook.extend(FWorkbookSheetsThreadCommentMixin);

//#endregion
//#region src/facade/f-worksheet.ts
/**
* @ignore
*/
var FWorksheetCommentMixin = class extends FWorksheet {
	getComments() {
		return this._injector.get(SheetsThreadCommentModel).getSubUnitAll(this._workbook.getUnitId(), this._worksheet.getSheetId()).map((comment) => this._injector.createInstance(FThreadComment, comment));
	}
	clearComments() {
		const promises = this.getComments().map((comment) => comment.deleteAsync());
		return Promise.all(promises).then(() => true);
	}
	/**
	* Subscribe to comment events.
	* @param callback Callback function, param contains comment info and target cell.
	*/
	onCommented(callback) {
		return this._injector.get(ICommandService).onCommandExecuted((command) => {
			if (command.id === AddCommentCommand.id) {
				const params = command.params;
				callback(params);
			}
		});
	}
	getCommentById(commentId) {
		const comment = this._injector.get(SheetsThreadCommentModel).getComment(this._workbook.getUnitId(), this._worksheet.getSheetId(), commentId);
		if (comment) return this._injector.createInstance(FThreadComment, comment);
	}
};
FWorksheet.extend(FWorksheetCommentMixin);

//#endregion
//#region src/facade/f-event.ts
/**
* @ignore
*/
var FSheetsThreadCommentEventNameMixin = class extends FEventName {
	get CommentAdded() {
		return "CommentAdded";
	}
	get BeforeCommentAdd() {
		return "BeforeCommentAdd";
	}
	get CommentUpdated() {
		return "CommentUpdated";
	}
	get BeforeCommentUpdate() {
		return "BeforeCommentUpdate";
	}
	get CommentDeleted() {
		return "CommentDeleted";
	}
	get BeforeCommentDelete() {
		return "BeforeCommentDelete";
	}
	get CommentResolved() {
		return "CommentResolved";
	}
	get BeforeCommentResolve() {
		return "BeforeCommentResolve";
	}
};
FEventName.extend(FSheetsThreadCommentEventNameMixin);

//#endregion
//#region src/facade/f-univer.ts
/**
* @ignore
*/
var FUniverSheetsThreadCommentMixin = class extends FUniver {
	_getTargetSheet(params = {}) {
		var _this$getActiveWorkbo;
		const workbook = params.unitId ? this.getUniverSheet(params.unitId) : (_this$getActiveWorkbo = this.getActiveWorkbook) === null || _this$getActiveWorkbo === void 0 ? void 0 : _this$getActiveWorkbo.call(this);
		if (!workbook) return null;
		const worksheet = params.subUnitId ? workbook.getSheetBySheetId(params.subUnitId) : workbook.getActiveSheet();
		if (!worksheet) return null;
		return {
			workbook,
			worksheet
		};
	}
	_initialize(injector) {
		const commandService = injector.get(ICommandService);
		this.disposeWithMe(this.registerEventHandler(this.Event.CommentAdded, () => commandService.onCommandExecuted((commandInfo) => {
			var _threadComment$getRan, _threadComment$getRan2, _threadComment$getRan3, _threadComment$getRan4;
			if (commandInfo.id !== AddCommentCommand.id) return;
			const target = this._getTargetSheet(commandInfo.params);
			if (!target) return;
			const { workbook, worksheet } = target;
			const { comment } = commandInfo.params;
			const threadComment = worksheet.getCommentById(comment.id);
			if (!threadComment) return;
			const eventParams = {
				workbook,
				worksheet,
				row: (_threadComment$getRan = (_threadComment$getRan2 = threadComment.getRange()) === null || _threadComment$getRan2 === void 0 ? void 0 : _threadComment$getRan2.getRow()) !== null && _threadComment$getRan !== void 0 ? _threadComment$getRan : 0,
				col: (_threadComment$getRan3 = (_threadComment$getRan4 = threadComment.getRange()) === null || _threadComment$getRan4 === void 0 ? void 0 : _threadComment$getRan4.getColumn()) !== null && _threadComment$getRan3 !== void 0 ? _threadComment$getRan3 : 0,
				comment: threadComment
			};
			this.fireEvent(this.Event.CommentAdded, eventParams);
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.CommentUpdated, () => commandService.onCommandExecuted((commandInfo) => {
			var _threadComment$getRan5, _threadComment$getRan6, _threadComment$getRan7, _threadComment$getRan8;
			if (commandInfo.id !== UpdateCommentCommand.id) return;
			const target = this._getTargetSheet(commandInfo.params);
			if (!target) return;
			const { workbook, worksheet } = target;
			const { payload } = commandInfo.params;
			const threadComment = worksheet.getCommentById(payload.commentId);
			if (!threadComment) return;
			const eventParams = {
				workbook,
				worksheet,
				row: (_threadComment$getRan5 = (_threadComment$getRan6 = threadComment.getRange()) === null || _threadComment$getRan6 === void 0 ? void 0 : _threadComment$getRan6.getRow()) !== null && _threadComment$getRan5 !== void 0 ? _threadComment$getRan5 : 0,
				col: (_threadComment$getRan7 = (_threadComment$getRan8 = threadComment.getRange()) === null || _threadComment$getRan8 === void 0 ? void 0 : _threadComment$getRan8.getColumn()) !== null && _threadComment$getRan7 !== void 0 ? _threadComment$getRan7 : 0,
				comment: threadComment
			};
			this.fireEvent(this.Event.CommentUpdated, eventParams);
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.CommentDeleted, () => commandService.onCommandExecuted((commandInfo) => {
			if (commandInfo.id !== DeleteCommentCommand.id && commandInfo.id !== DeleteCommentTreeCommand.id) return;
			const target = this._getTargetSheet(commandInfo.params);
			if (!target) return;
			const { workbook, worksheet } = target;
			const { commentId } = commandInfo.params;
			const eventParams = {
				workbook,
				worksheet,
				commentId
			};
			this.fireEvent(this.Event.CommentDeleted, eventParams);
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.CommentResolved, () => commandService.onCommandExecuted((commandInfo) => {
			var _threadComment$getRan9, _threadComment$getRan10, _threadComment$getRan11, _threadComment$getRan12;
			if (commandInfo.id !== ResolveCommentCommand.id) return;
			const target = this._getTargetSheet(commandInfo.params);
			if (!target) return;
			const { workbook, worksheet } = target;
			const { commentId, resolved } = commandInfo.params;
			const threadComment = worksheet.getCommentById(commentId);
			if (!threadComment) return;
			const eventParams = {
				workbook,
				worksheet,
				row: (_threadComment$getRan9 = (_threadComment$getRan10 = threadComment.getRange()) === null || _threadComment$getRan10 === void 0 ? void 0 : _threadComment$getRan10.getRow()) !== null && _threadComment$getRan9 !== void 0 ? _threadComment$getRan9 : 0,
				col: (_threadComment$getRan11 = (_threadComment$getRan12 = threadComment.getRange()) === null || _threadComment$getRan12 === void 0 ? void 0 : _threadComment$getRan12.getColumn()) !== null && _threadComment$getRan11 !== void 0 ? _threadComment$getRan11 : 0,
				comment: threadComment,
				resolved
			};
			this.fireEvent(this.Event.CommentResolved, eventParams);
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeCommentAdd, () => commandService.beforeCommandExecuted((commandInfo) => {
			if (commandInfo.id !== AddCommentCommand.id) return;
			const target = this._getTargetSheet(commandInfo.params);
			if (!target) return;
			const { workbook, worksheet } = target;
			const { comment } = commandInfo.params;
			const { range } = deserializeRangeWithSheet(comment.ref);
			const eventParams = {
				workbook,
				worksheet,
				row: range.startRow,
				col: range.startColumn,
				comment: FTheadCommentItem.create(comment)
			};
			this.fireEvent(this.Event.BeforeCommentAdd, eventParams);
			if (eventParams.cancel) throw new CanceledError();
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeCommentUpdate, () => commandService.beforeCommandExecuted((commandInfo) => {
			var _threadComment$getRan13, _threadComment$getRan14, _threadComment$getRan15, _threadComment$getRan16;
			if (commandInfo.id !== UpdateCommentCommand.id) return;
			const target = this._getTargetSheet(commandInfo.params);
			if (!target) return;
			const { workbook, worksheet } = target;
			const { payload } = commandInfo.params;
			const threadComment = worksheet.getCommentById(payload.commentId);
			if (!threadComment) return;
			const eventParams = {
				workbook,
				worksheet,
				row: (_threadComment$getRan13 = (_threadComment$getRan14 = threadComment.getRange()) === null || _threadComment$getRan14 === void 0 ? void 0 : _threadComment$getRan14.getRow()) !== null && _threadComment$getRan13 !== void 0 ? _threadComment$getRan13 : 0,
				col: (_threadComment$getRan15 = (_threadComment$getRan16 = threadComment.getRange()) === null || _threadComment$getRan16 === void 0 ? void 0 : _threadComment$getRan16.getColumn()) !== null && _threadComment$getRan15 !== void 0 ? _threadComment$getRan15 : 0,
				comment: threadComment,
				newContent: RichTextValue.createByBody(payload.text)
			};
			this.fireEvent(this.Event.BeforeCommentUpdate, eventParams);
			if (eventParams.cancel) throw new CanceledError();
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeCommentDelete, () => commandService.beforeCommandExecuted((commandInfo) => {
			var _threadComment$getRan17, _threadComment$getRan18, _threadComment$getRan19, _threadComment$getRan20;
			if (commandInfo.id !== DeleteCommentCommand.id && commandInfo.id !== DeleteCommentTreeCommand.id) return;
			const target = this._getTargetSheet(commandInfo.params);
			if (!target) return;
			const { workbook, worksheet } = target;
			const { commentId } = commandInfo.params;
			const threadComment = worksheet.getCommentById(commentId);
			if (!threadComment) return;
			const eventParams = {
				workbook,
				worksheet,
				row: (_threadComment$getRan17 = (_threadComment$getRan18 = threadComment.getRange()) === null || _threadComment$getRan18 === void 0 ? void 0 : _threadComment$getRan18.getRow()) !== null && _threadComment$getRan17 !== void 0 ? _threadComment$getRan17 : 0,
				col: (_threadComment$getRan19 = (_threadComment$getRan20 = threadComment.getRange()) === null || _threadComment$getRan20 === void 0 ? void 0 : _threadComment$getRan20.getColumn()) !== null && _threadComment$getRan19 !== void 0 ? _threadComment$getRan19 : 0,
				comment: threadComment
			};
			this.fireEvent(this.Event.BeforeCommentDelete, eventParams);
			if (eventParams.cancel) throw new CanceledError();
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeCommentResolve, () => commandService.beforeCommandExecuted((commandInfo) => {
			var _getRow, _getColumn;
			if (commandInfo.id !== ResolveCommentCommand.id) return;
			const target = this._getTargetSheet(commandInfo.params);
			if (!target) return;
			const { workbook, worksheet } = target;
			const { commentId, resolved } = commandInfo.params;
			const threadComment = worksheet.getCommentById(commentId);
			if (!threadComment) return;
			const eventParams = {
				workbook,
				worksheet,
				row: (_getRow = threadComment.getRange().getRow()) !== null && _getRow !== void 0 ? _getRow : 0,
				col: (_getColumn = threadComment.getRange().getColumn()) !== null && _getColumn !== void 0 ? _getColumn : 0,
				comment: threadComment,
				resolved
			};
			this.fireEvent(this.Event.BeforeCommentResolve, eventParams);
			if (eventParams.cancel) throw new CanceledError();
		})));
	}
	/**
	* @ignore
	*/
	newTheadComment(comment) {
		return new FTheadCommentBuilder(comment);
	}
};
FUniver.extend(FUniverSheetsThreadCommentMixin);

//#endregion
export { FThreadComment };
