// @univerjs/network/lib/index.js
import { Disposable, DisposableCollection, IConfigService, ILogService, Inject, Injector, LookUp, Plugin, Quantity, createIdentifier, merge, mergeOverrideWithDependencies, noop, registerDependencies, remove, toDisposable } from "@univerjs/core";
import { Observable, catchError, firstValueFrom, of, throwError } from "rxjs";
import { concatMap, retry, share } from "rxjs/operators";

//#region package.json
var name = "@univerjs/network";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
const NETWORK_PLUGIN_CONFIG_KEY = "network.config";
const configSymbol = Symbol(NETWORK_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region src/services/http/headers.ts
const ApplicationJSONType = "application/json";
/**
* Check if the content type is application/json
* "application/json" or "application/json; charset=utf-8" or ["application/json"]
* @param contentType
*/
function isApplicationJSONType(contentType) {
	if (Array.isArray(contentType)) return contentType.some((type) => type.includes(ApplicationJSONType));
	return contentType.includes(ApplicationJSONType);
}
/**
* It wraps headers of HTTP requests' and responses' headers.
*/
var HTTPHeaders = class {
	constructor(headers) {
		_defineProperty(this, "_headers", /* @__PURE__ */ new Map());
		if (typeof headers === "string") this._handleHeadersString(headers);
		else if (headers instanceof Headers) this._handleHeaders(headers);
		else if (headers) this._handleHeadersConstructorProps(headers);
	}
	forEach(callback) {
		this._headers.forEach((v, key) => callback(key, v));
	}
	has(key) {
		return !!this._headers.has(key.toLowerCase());
	}
	get(key) {
		const k = key.toLowerCase();
		return this._headers.has(k) ? this._headers.get(k) : null;
	}
	set(key, value) {
		this._setHeader(key, value);
	}
	toHeadersInit(body) {
		var _headers$accept;
		const headers = {};
		this._headers.forEach((values, key) => {
			headers[key] = values.join(",");
		});
		(_headers$accept = headers.accept) !== null && _headers$accept !== void 0 || (headers.accept = "application/json, text/plain, */*");
		if (!(body instanceof FormData)) {
			var _headers$contentType;
			(_headers$contentType = headers["content-type"]) !== null && _headers$contentType !== void 0 || (headers["content-type"] = "application/json;charset=UTF-8");
		}
		return headers;
	}
	_setHeader(name, value) {
		const lowerCase = name.toLowerCase();
		if (this._headers.has(lowerCase)) this._headers.get(lowerCase).push(value.toString());
		else this._headers.set(lowerCase, [value.toString()]);
	}
	_handleHeadersString(headers) {
		headers.split("\n").forEach((header) => {
			const [name, value] = header.split(":");
			if (name && value) this._setHeader(name, value);
		});
	}
	_handleHeadersConstructorProps(headers) {
		Object.entries(headers).forEach(([name, value]) => this._setHeader(name, value));
	}
	_handleHeaders(headers) {
		headers.forEach((value, name) => this._setHeader(name, value));
	}
};

//#endregion
//#region src/services/http/implementations/implementation.ts
const IHTTPImplementation = createIdentifier("network.http-implementation");

//#endregion
//#region src/services/http/params.ts
var HTTPParams = class {
	constructor(params) {
		this.params = params;
	}
	toString() {
		if (!this.params) return "";
		return Object.keys(this.params).map((key) => {
			const value = this.params[key];
			if (Array.isArray(value)) return value.map((v) => `${key}=${v}`).join("&");
			return `${key}=${value}`;
		}).join("&");
	}
};

//#endregion
//#region src/services/http/request.ts
let HTTPRequestUID = 0;
var HTTPRequest = class {
	get headers() {
		return this.requestParams.headers;
	}
	get withCredentials() {
		return this.requestParams.withCredentials;
	}
	get responseType() {
		return this.requestParams.responseType;
	}
	constructor(method, url, requestParams) {
		this.method = method;
		this.url = url;
		this.requestParams = requestParams;
		_defineProperty(this, "uid", HTTPRequestUID++);
	}
	getUrlWithParams() {
		var _this$requestParams;
		const params = (_this$requestParams = this.requestParams) === null || _this$requestParams === void 0 || (_this$requestParams = _this$requestParams.params) === null || _this$requestParams === void 0 ? void 0 : _this$requestParams.toString();
		if (!params) return this.url;
		return `${this.url}${this.url.includes("?") ? "&" : "?"}${params}`;
	}
	getBody() {
		var _this$headers$get, _this$requestParams2;
		const contentType = (_this$headers$get = this.headers.get("Content-Type")) !== null && _this$headers$get !== void 0 ? _this$headers$get : ApplicationJSONType;
		const body = (_this$requestParams2 = this.requestParams) === null || _this$requestParams2 === void 0 ? void 0 : _this$requestParams2.body;
		if (body instanceof FormData) return body;
		if (isApplicationJSONType(contentType) && body && typeof body === "object") return JSON.stringify(body);
		return body ? `${body}` : null;
	}
	getHeadersInit() {
		var _this$requestParams3;
		return this.headers.toHeadersInit((_this$requestParams3 = this.requestParams) === null || _this$requestParams3 === void 0 ? void 0 : _this$requestParams3.body);
	}
};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/services/http/http.service.ts
let HTTPService = class HTTPService extends Disposable {
	constructor(_http) {
		super();
		this._http = _http;
		_defineProperty(this, "_interceptors", []);
		_defineProperty(this, "_pipe", void 0);
	}
	/**
	* Register an HTTP interceptor.
	*
	* @param interceptor the http interceptor
	* @returns a disposable handler to remove the interceptor
	*/
	registerHTTPInterceptor(interceptor) {
		if (this._interceptors.indexOf(interceptor) !== -1) throw new Error("[HTTPService]: The interceptor has already been registered!");
		this._interceptors.push(interceptor);
		this._interceptors = this._interceptors.sort((a, b) => {
			var _a$priority, _b$priority;
			return ((_a$priority = a.priority) !== null && _a$priority !== void 0 ? _a$priority : 0) - ((_b$priority = b.priority) !== null && _b$priority !== void 0 ? _b$priority : 0);
		});
		this._pipe = null;
		return toDisposable(() => remove(this._interceptors, interceptor));
	}
	get(url, params) {
		return this.request("GET", url, params);
	}
	post(url, params) {
		return this.request("POST", url, params);
	}
	put(url, params) {
		return this.request("PUT", url, params);
	}
	delete(url, params) {
		return this.request("DELETE", url, params);
	}
	patch(url, params) {
		return this.request("PATCH", url, params);
	}
	/**
	* The HTTP request implementations. Normally you should use the `get`, `post`, `put`, `delete`,
	* `patch` methods instead.
	* @param method HTTP request method, e.g. GET, POST, PUT, DELETE, etc.
	* @param url The URL to send the request to.
	* @param options Optional parameters for the request.
	* @returns A promise that resolves to the HTTP response.
	*/
	async request(method, url, options) {
		var _options$withCredenti, _options$responseType;
		return await firstValueFrom(of(new HTTPRequest(method, url, {
			headers: new HTTPHeaders(options === null || options === void 0 ? void 0 : options.headers),
			params: new HTTPParams(options === null || options === void 0 ? void 0 : options.params),
			withCredentials: (_options$withCredenti = options === null || options === void 0 ? void 0 : options.withCredentials) !== null && _options$withCredenti !== void 0 ? _options$withCredenti : false,
			responseType: (_options$responseType = options === null || options === void 0 ? void 0 : options.responseType) !== null && _options$responseType !== void 0 ? _options$responseType : "json",
			body: ["GET", "DELETE"].includes(method) ? void 0 : options === null || options === void 0 ? void 0 : options.body
		})).pipe(concatMap((request) => this._runInterceptorsAndImplementation(request))));
	}
	/**
	* Send an HTTP request. It returns an observable that emits HTTP events. For example, it can be used to
	* send Server-Sent Events (SSE) requests.
	* @deprecated Please use `stream` method instead.
	* @param method HTTP request method, e.g. GET, POST, PUT, DELETE, etc.
	* @param url The URL to send the request to.
	* @param _params Optional parameters for the request.
	* @returns An observable of the HTTP event.
	*/
	stream(method, url, _params) {
		return this.getSSE(method, url, _params);
	}
	/**
	* Send a Server-Sent Events (SSE) request. It returns an observable that emits HTTP events. It is the observable
	* pair of the `request` method.
	* @deprecated Please use `stream` method instead.
	* @param method HTTP request method, e.g. GET, POST, PUT, DELETE, etc.
	* @param url The URL to send the request to.
	* @param _params Optional parameters for the request.
	* @returns An observable of the HTTP event.
	*/
	getSSE(method, url, _params) {
		var _params$withCredentia, _params$responseType;
		return of(new HTTPRequest(method, url, {
			headers: new HTTPHeaders(_params === null || _params === void 0 ? void 0 : _params.headers),
			params: new HTTPParams(_params === null || _params === void 0 ? void 0 : _params.params),
			withCredentials: (_params$withCredentia = _params === null || _params === void 0 ? void 0 : _params.withCredentials) !== null && _params$withCredentia !== void 0 ? _params$withCredentia : false,
			reportProgress: true,
			responseType: (_params$responseType = _params === null || _params === void 0 ? void 0 : _params.responseType) !== null && _params$responseType !== void 0 ? _params$responseType : "json",
			body: ["GET", "DELETE"].includes(method) ? void 0 : _params === null || _params === void 0 ? void 0 : _params.body
		})).pipe(concatMap((request) => this._runInterceptorsAndImplementation(request)));
	}
	_runInterceptorsAndImplementation(request) {
		if (!this._pipe) this._pipe = this._interceptors.map((handler) => handler.interceptor).reduceRight((nextHandlerFunction, interceptorFunction) => chainInterceptorFn(nextHandlerFunction, interceptorFunction), (requestFromPrevInterceptor, finalHandler) => finalHandler(requestFromPrevInterceptor));
		return this._pipe(request, (requestToNext) => this._http.send(requestToNext));
	}
};
HTTPService = __decorate([__decorateParam(0, IHTTPImplementation)], HTTPService);
function chainInterceptorFn(afterInterceptorChain, currentInterceptorFn) {
	return (prevRequest, nextHandlerFn) => currentInterceptorFn(prevRequest, (nextRequest) => afterInterceptorChain(nextRequest, nextHandlerFn));
}

//#endregion
//#region src/services/http/http.ts
/**
* Http status codes.
*
* https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml
*/
let HTTPStatusCode = /* @__PURE__ */ function(HTTPStatusCode) {
	HTTPStatusCode[HTTPStatusCode["Continue"] = 100] = "Continue";
	HTTPStatusCode[HTTPStatusCode["SwitchingProtocols"] = 101] = "SwitchingProtocols";
	HTTPStatusCode[HTTPStatusCode["Processing"] = 102] = "Processing";
	HTTPStatusCode[HTTPStatusCode["EarlyHints"] = 103] = "EarlyHints";
	HTTPStatusCode[HTTPStatusCode["Ok"] = 200] = "Ok";
	HTTPStatusCode[HTTPStatusCode["Created"] = 201] = "Created";
	HTTPStatusCode[HTTPStatusCode["Accepted"] = 202] = "Accepted";
	HTTPStatusCode[HTTPStatusCode["NonAuthoritativeInformation"] = 203] = "NonAuthoritativeInformation";
	HTTPStatusCode[HTTPStatusCode["NoContent"] = 204] = "NoContent";
	HTTPStatusCode[HTTPStatusCode["ResetContent"] = 205] = "ResetContent";
	HTTPStatusCode[HTTPStatusCode["PartialContent"] = 206] = "PartialContent";
	HTTPStatusCode[HTTPStatusCode["MultiStatus"] = 207] = "MultiStatus";
	HTTPStatusCode[HTTPStatusCode["AlreadyReported"] = 208] = "AlreadyReported";
	HTTPStatusCode[HTTPStatusCode["ImUsed"] = 226] = "ImUsed";
	HTTPStatusCode[HTTPStatusCode["MultipleChoices"] = 300] = "MultipleChoices";
	HTTPStatusCode[HTTPStatusCode["MovedPermanently"] = 301] = "MovedPermanently";
	HTTPStatusCode[HTTPStatusCode["Found"] = 302] = "Found";
	HTTPStatusCode[HTTPStatusCode["SeeOther"] = 303] = "SeeOther";
	HTTPStatusCode[HTTPStatusCode["NotModified"] = 304] = "NotModified";
	HTTPStatusCode[HTTPStatusCode["UseProxy"] = 305] = "UseProxy";
	HTTPStatusCode[HTTPStatusCode["Unused"] = 306] = "Unused";
	HTTPStatusCode[HTTPStatusCode["TemporaryRedirect"] = 307] = "TemporaryRedirect";
	HTTPStatusCode[HTTPStatusCode["PermanentRedirect"] = 308] = "PermanentRedirect";
	HTTPStatusCode[HTTPStatusCode["BadRequest"] = 400] = "BadRequest";
	HTTPStatusCode[HTTPStatusCode["Unauthorized"] = 401] = "Unauthorized";
	HTTPStatusCode[HTTPStatusCode["PaymentRequired"] = 402] = "PaymentRequired";
	HTTPStatusCode[HTTPStatusCode["Forbidden"] = 403] = "Forbidden";
	HTTPStatusCode[HTTPStatusCode["NotFound"] = 404] = "NotFound";
	HTTPStatusCode[HTTPStatusCode["MethodNotAllowed"] = 405] = "MethodNotAllowed";
	HTTPStatusCode[HTTPStatusCode["NotAcceptable"] = 406] = "NotAcceptable";
	HTTPStatusCode[HTTPStatusCode["ProxyAuthenticationRequired"] = 407] = "ProxyAuthenticationRequired";
	HTTPStatusCode[HTTPStatusCode["RequestTimeout"] = 408] = "RequestTimeout";
	HTTPStatusCode[HTTPStatusCode["Conflict"] = 409] = "Conflict";
	HTTPStatusCode[HTTPStatusCode["Gone"] = 410] = "Gone";
	HTTPStatusCode[HTTPStatusCode["LengthRequired"] = 411] = "LengthRequired";
	HTTPStatusCode[HTTPStatusCode["PreconditionFailed"] = 412] = "PreconditionFailed";
	HTTPStatusCode[HTTPStatusCode["PayloadTooLarge"] = 413] = "PayloadTooLarge";
	HTTPStatusCode[HTTPStatusCode["UriTooLong"] = 414] = "UriTooLong";
	HTTPStatusCode[HTTPStatusCode["UnsupportedMediaType"] = 415] = "UnsupportedMediaType";
	HTTPStatusCode[HTTPStatusCode["RangeNotSatisfiable"] = 416] = "RangeNotSatisfiable";
	HTTPStatusCode[HTTPStatusCode["ExpectationFailed"] = 417] = "ExpectationFailed";
	HTTPStatusCode[HTTPStatusCode["ImATeapot"] = 418] = "ImATeapot";
	HTTPStatusCode[HTTPStatusCode["MisdirectedRequest"] = 421] = "MisdirectedRequest";
	HTTPStatusCode[HTTPStatusCode["UnprocessableEntity"] = 422] = "UnprocessableEntity";
	HTTPStatusCode[HTTPStatusCode["Locked"] = 423] = "Locked";
	HTTPStatusCode[HTTPStatusCode["FailedDependency"] = 424] = "FailedDependency";
	HTTPStatusCode[HTTPStatusCode["TooEarly"] = 425] = "TooEarly";
	HTTPStatusCode[HTTPStatusCode["UpgradeRequired"] = 426] = "UpgradeRequired";
	HTTPStatusCode[HTTPStatusCode["PreconditionRequired"] = 428] = "PreconditionRequired";
	HTTPStatusCode[HTTPStatusCode["TooManyRequests"] = 429] = "TooManyRequests";
	HTTPStatusCode[HTTPStatusCode["RequestHeaderFieldsTooLarge"] = 431] = "RequestHeaderFieldsTooLarge";
	HTTPStatusCode[HTTPStatusCode["UnavailableForLegalReasons"] = 451] = "UnavailableForLegalReasons";
	HTTPStatusCode[HTTPStatusCode["InternalServerError"] = 500] = "InternalServerError";
	HTTPStatusCode[HTTPStatusCode["NotImplemented"] = 501] = "NotImplemented";
	HTTPStatusCode[HTTPStatusCode["BadGateway"] = 502] = "BadGateway";
	HTTPStatusCode[HTTPStatusCode["ServiceUnavailable"] = 503] = "ServiceUnavailable";
	HTTPStatusCode[HTTPStatusCode["GatewayTimeout"] = 504] = "GatewayTimeout";
	HTTPStatusCode[HTTPStatusCode["HttpVersionNotSupported"] = 505] = "HttpVersionNotSupported";
	HTTPStatusCode[HTTPStatusCode["VariantAlsoNegotiates"] = 506] = "VariantAlsoNegotiates";
	HTTPStatusCode[HTTPStatusCode["InsufficientStorage"] = 507] = "InsufficientStorage";
	HTTPStatusCode[HTTPStatusCode["LoopDetected"] = 508] = "LoopDetected";
	HTTPStatusCode[HTTPStatusCode["NotExtended"] = 510] = "NotExtended";
	HTTPStatusCode[HTTPStatusCode["NetworkAuthenticationRequired"] = 511] = "NetworkAuthenticationRequired";
	return HTTPStatusCode;
}({});

//#endregion
//#region src/services/http/response.ts
let HTTPEventType = /* @__PURE__ */ function(HTTPEventType) {
	HTTPEventType[HTTPEventType["DownloadProgress"] = 0] = "DownloadProgress";
	HTTPEventType[HTTPEventType["Response"] = 1] = "Response";
	return HTTPEventType;
}({});
/**
* Wraps success response info.
*/
var HTTPResponse = class {
	constructor({ body, headers, status, statusText }) {
		_defineProperty(this, "type", 1);
		_defineProperty(this, "body", void 0);
		_defineProperty(this, "headers", void 0);
		_defineProperty(this, "status", void 0);
		_defineProperty(this, "statusText", void 0);
		this.body = body;
		this.headers = headers;
		this.status = status;
		this.statusText = statusText;
	}
};
/**
* Progress event for HTTP request. Usually used for reporting download/upload progress or SSE streaming.
*/
var HTTPProgress = class {
	constructor(total, loaded, partialText) {
		this.total = total;
		this.loaded = loaded;
		this.partialText = partialText;
		_defineProperty(this, "type", 0);
	}
};
var ResponseHeader = class {
	constructor(headers, status, statusText) {
		this.headers = headers;
		this.status = status;
		this.statusText = statusText;
	}
};
var HTTPResponseError = class {
	constructor({ request, headers, status, statusText, error }) {
		_defineProperty(this, "request", void 0);
		_defineProperty(this, "headers", void 0);
		_defineProperty(this, "status", void 0);
		_defineProperty(this, "statusText", void 0);
		_defineProperty(this, "error", void 0);
		this.request = request;
		this.headers = headers;
		this.status = status;
		this.statusText = statusText;
		this.error = error;
	}
};

//#endregion
//#region src/services/http/implementations/util.ts
function parseFetchParamsFromRequest(request) {
	return {
		method: request.method,
		headers: request.getHeadersInit(),
		body: request.getBody(),
		credentials: request.withCredentials ? "include" : void 0
	};
}

//#endregion
//#region src/services/http/implementations/fetch.ts
let FetchHTTPImplementation = class FetchHTTPImplementation {
	constructor(_logService) {
		this._logService = _logService;
	}
	send(request) {
		return new Observable((subscriber) => {
			const abortController = new AbortController();
			this._send(request, subscriber, abortController).catch((error) => {
				subscriber.error(new HTTPResponseError({
					error,
					request
				}));
			});
			return () => abortController.abort();
		});
	}
	async _send(request, subscriber, abortController) {
		let response;
		try {
			const fetchParams = parseFetchParamsFromRequest(request);
			const urlWithParams = request.getUrlWithParams();
			const fetchPromise = fetch(urlWithParams, {
				signal: abortController.signal,
				...fetchParams
			});
			this._logService.debug(`[FetchHTTPImplementation]: sending request to url ${urlWithParams} with params ${fetchParams}`);
			response = await fetchPromise;
		} catch (error) {
			var _error$status, _error$statusText;
			const e = new HTTPResponseError({
				request,
				error,
				status: (_error$status = error.status) !== null && _error$status !== void 0 ? _error$status : 0,
				statusText: (_error$statusText = error.statusText) !== null && _error$statusText !== void 0 ? _error$statusText : "Unknown Error",
				headers: error.headers
			});
			this._logService.error("[FetchHTTPImplementation]: network error", e);
			subscriber.error(e);
			return;
		}
		const responseHeaders = new HTTPHeaders(response.headers);
		const status = response.status;
		const statusText = response.statusText;
		let body = null;
		if (response.body) body = await this._readBody(request, response, subscriber);
		if (status >= 200 && status < 300) subscriber.next(new HTTPResponse({
			body,
			headers: responseHeaders,
			status,
			statusText
		}));
		else {
			const e = new HTTPResponseError({
				request,
				error: body,
				status,
				statusText,
				headers: responseHeaders
			});
			this._logService.error("[FetchHTTPImplementation]: network error", e);
			subscriber.error(e);
		}
		subscriber.complete();
	}
	async _readBody(request, response, subscriber) {
		var _request$requestParam;
		const chunks = [];
		const reader = response.body.getReader();
		const contentLength = response.headers.get("content-length");
		let receivedLength = 0;
		const reportProgress = (_request$requestParam = request.requestParams) === null || _request$requestParam === void 0 ? void 0 : _request$requestParam.reportProgress;
		const responseType = request.responseType;
		let partialText;
		let decoder;
		while (true) {
			const { done, value } = await reader.read();
			if (done) break;
			chunks.push(value);
			receivedLength += value.length;
			if (reportProgress && responseType === "text") {
				var _partialText, _decoder;
				partialText = ((_partialText = partialText) !== null && _partialText !== void 0 ? _partialText : "") + ((_decoder = decoder) !== null && _decoder !== void 0 ? _decoder : decoder = new TextDecoder()).decode(value, { stream: true });
				subscriber.next(new HTTPProgress(contentLength ? Number.parseInt(contentLength, 10) : void 0, receivedLength, partialText));
			}
		}
		const all = mergeChunks(chunks, receivedLength);
		try {
			var _response$headers$get;
			return deserialize(request, all, (_response$headers$get = response.headers.get("content-type")) !== null && _response$headers$get !== void 0 ? _response$headers$get : "");
		} catch (error) {
			const e = new HTTPResponseError({
				request,
				error,
				status: response.status,
				statusText: response.statusText,
				headers: new HTTPHeaders(response.headers)
			});
			this._logService.error("[FetchHTTPImplementation]: network error", e);
			subscriber.error(e);
			return null;
		}
	}
};
FetchHTTPImplementation = __decorate([__decorateParam(0, ILogService)], FetchHTTPImplementation);
function mergeChunks(chunks, totalLength) {
	const all = new Uint8Array(totalLength);
	let position = 0;
	for (const chunk of chunks) {
		all.set(chunk, position);
		position += chunk.length;
	}
	return all;
}
const XSSI_PREFIX = /^\)\]\}',?\n/;
function deserialize(request, bin, contentType) {
	switch (request.responseType) {
		case "json":
			const text = new TextDecoder().decode(bin).replace(XSSI_PREFIX, "");
			return text === "" ? null : JSON.parse(text);
		case "text": return new TextDecoder().decode(bin);
		case "blob": return new Blob([bin.buffer], { type: contentType });
		case "arraybuffer": return bin.buffer;
		default: throw new Error(`[FetchHTTPImplementation]: unknown response type: ${request.responseType}.`);
	}
}

//#endregion
//#region src/services/http/implementations/xhr.ts
let XHRHTTPImplementation = class XHRHTTPImplementation {
	constructor(_logService) {
		this._logService = _logService;
	}
	send(request) {
		return new Observable((observer) => {
			const xhr = new XMLHttpRequest();
			const urlWithParams = request.getUrlWithParams();
			const fetchParams = parseFetchParamsFromRequest(request);
			const { responseType } = request;
			xhr.open(request.method, urlWithParams);
			if (request.withCredentials) xhr.withCredentials = true;
			if (fetchParams.headers) Object.entries(fetchParams.headers).forEach(([key, value]) => xhr.setRequestHeader(key, value));
			const buildResponseHeader = () => {
				const statusText = xhr.statusText || "OK";
				return new ResponseHeader(new HTTPHeaders(xhr.getAllResponseHeaders()), xhr.status, statusText);
			};
			const onLoadHandler = () => {
				const { headers, statusText, status } = buildResponseHeader();
				let body = null;
				let error = null;
				if (status !== 204) body = typeof xhr.response === "undefined" ? xhr.responseText : xhr.response;
				let success = status >= 200 && status < 300;
				if (responseType === "json" && typeof body === "string") {
					const originalBody = body;
					try {
						body = body ? JSON.parse(body) : null;
					} catch (e) {
						success = false;
						body = originalBody;
						error = e;
					}
				}
				if (responseType === "blob" && !(body instanceof Blob)) {
					success = false;
					error = /* @__PURE__ */ new Error("Response is not a Blob object");
				}
				if (success) observer.next(new HTTPResponse({
					body,
					headers,
					status,
					statusText
				}));
				else {
					const e = new HTTPResponseError({
						request,
						error,
						headers,
						status,
						statusText
					});
					this._logService.error("[XHRHTTPImplementation]: network error", e);
					observer.error(e);
				}
			};
			const onErrorHandler = (error) => {
				const e = new HTTPResponseError({
					request,
					error,
					status: xhr.status || 0,
					statusText: xhr.statusText || "Unknown Error",
					headers: buildResponseHeader().headers
				});
				this._logService.error("[XHRHTTPImplementation]: network error", e);
				observer.error(e);
			};
			xhr.responseType = responseType || "";
			xhr.addEventListener("load", onLoadHandler);
			xhr.addEventListener("error", onErrorHandler);
			xhr.addEventListener("abort", onErrorHandler);
			xhr.addEventListener("timeout", onErrorHandler);
			const body = request.getBody();
			xhr.send(body);
			this._logService.debug("[XHRHTTPImplementation]", `sending request to url ${urlWithParams} with params ${fetchParams}`);
			return () => {
				if (xhr.readyState !== xhr.DONE) xhr.abort();
				xhr.removeEventListener("load", onLoadHandler);
				xhr.removeEventListener("error", onErrorHandler);
				xhr.removeEventListener("abort", onErrorHandler);
				xhr.removeEventListener("timeout", onErrorHandler);
			};
		});
	}
};
XHRHTTPImplementation = __decorate([__decorateParam(0, ILogService)], XHRHTTPImplementation);

//#endregion
//#region src/plugin.ts
let UniverNetworkPlugin = class UniverNetworkPlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _logger, _injector, _configService) {
		super();
		this._config = _config;
		this._logger = _logger;
		this._injector = _injector;
		this._configService = _configService;
		const { ...rest } = merge({}, defaultPluginConfig, this._config);
		this._configService.setConfig(NETWORK_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		var _this$_config, _this$_config2, _this$_config3;
		if (this._injector.get(HTTPService, Quantity.OPTIONAL, LookUp.SKIP_SELF) && !((_this$_config = this._config) === null || _this$_config === void 0 ? void 0 : _this$_config.forceUseNewInstance)) {
			this._logger.warn("[UniverNetworkPlugin]", "HTTPService is already registered in an ancestor interceptor. Skipping registration. If you want to force a new instance, set \"forceUseNewInstance\" to true in the plugin configuration.");
			return;
		}
		const impl = ((_this$_config2 = this._config) === null || _this$_config2 === void 0 ? void 0 : _this$_config2.useFetchImpl) ? FetchHTTPImplementation : typeof window !== "undefined" ? XHRHTTPImplementation : FetchHTTPImplementation;
		registerDependencies(this._injector, mergeOverrideWithDependencies([[HTTPService], [IHTTPImplementation, { useClass: impl }]], (_this$_config3 = this._config) === null || _this$_config3 === void 0 ? void 0 : _this$_config3.override));
	}
};
_defineProperty(UniverNetworkPlugin, "pluginName", "UNIVER_NETWORK_PLUGIN");
_defineProperty(UniverNetworkPlugin, "packageName", name);
_defineProperty(UniverNetworkPlugin, "version", version);
UniverNetworkPlugin = __decorate([
	__decorateParam(1, ILogService),
	__decorateParam(2, Inject(Injector)),
	__decorateParam(3, IConfigService)
], UniverNetworkPlugin);

//#endregion
//#region src/services/http/interceptors/auth-interceptor.ts
const AuthInterceptorFactory = (params) => {
	const { errorStatusCodes, onAuthError } = params;
	const authInterceptor = (request, next) => {
		return next(request).pipe(catchError((error) => {
			if (error instanceof HTTPResponseError && errorStatusCodes.some((c) => c === error.status)) onAuthError();
			return throwError(() => error);
		}));
	};
	return authInterceptor;
};

//#endregion
//#region src/services/http/interceptors/merge-interceptor.ts
const createDefaultFetchCheck = (time = 300) => {
	let cancel = noop;
	return (_currentConfig) => {
		return new Promise((res) => {
			cancel();
			const t = setTimeout(() => {
				res(true);
			}, time);
			cancel = () => {
				clearTimeout(t);
				res(false);
			};
		});
	};
};
const createDistributeResult = () => {
	return (result, list) => {
		return list.map((config) => ({
			config,
			result
		}));
	};
};
const MergeInterceptorFactory = (config, options = {}) => {
	const { isMatch, getParamsFromRequest, mergeParamsToRequest } = config;
	const { fetchCheck = createDefaultFetchCheck(300), distributeResult = createDistributeResult() } = options;
	const hookList = [];
	const getPlainList = (_list) => _list.map((item) => item.config);
	return (requestConfig, next) => {
		if (!isMatch(requestConfig)) return next(requestConfig);
		return new Observable((observer) => {
			const params = getParamsFromRequest(requestConfig);
			hookList.push({
				next: (v) => observer.next(v),
				error: (error) => observer.error(error),
				config: params
			});
			const list = getPlainList(hookList);
			fetchCheck(requestConfig).then((isFetch) => {
				if (isFetch) {
					const currentHookList = [];
					list.forEach((config) => {
						const index = hookList.findIndex((item) => item.config === config);
						if (index >= 0) {
							const [hook] = hookList.splice(index, 1);
							currentHookList.push(hook);
						}
					});
					next(mergeParamsToRequest(list, requestConfig)).subscribe({
						next: (e) => {
							if (e.type === 1) {
								const body = e.body;
								const configList = distributeResult(body, list);
								currentHookList.forEach((hookItem) => {
									const res = configList.find((item) => item.config === hookItem.config);
									if (res) {
										const response = new HTTPResponse({
											body: res.result,
											headers: e.headers,
											status: e.status,
											statusText: e.statusText
										});
										hookItem.next(response);
									} else hookItem.error("batch error");
								});
							}
						},
						complete: () => observer.complete(),
						error: (e) => observer.error(e)
					});
				}
			});
		});
	};
};

//#endregion
//#region src/services/http/interceptors/retry-interceptor.ts
const DEFAULT_MAX_RETRY_ATTEMPTS = 3;
const DELAY_INTERVAL = 1e3;
const RetryInterceptorFactory = (params) => {
	var _params$maxRetryAttem, _params$delayInterval;
	const maxRetryAttempts = (_params$maxRetryAttem = params === null || params === void 0 ? void 0 : params.maxRetryAttempts) !== null && _params$maxRetryAttem !== void 0 ? _params$maxRetryAttem : DEFAULT_MAX_RETRY_ATTEMPTS;
	const delayInterval = (_params$delayInterval = params === null || params === void 0 ? void 0 : params.delayInterval) !== null && _params$delayInterval !== void 0 ? _params$delayInterval : DELAY_INTERVAL;
	return (request, next) => next(request).pipe(retry({
		delay: delayInterval,
		count: maxRetryAttempts
	}));
};

//#endregion
//#region src/services/http/interceptors/threshold-interceptor.ts
const ThresholdInterceptorFactory = (params) => {
	/**
	* The local variable to store handles.
	*/
	const handlers = [];
	const ongoingHandlers = /* @__PURE__ */ new Set();
	const tick = () => {
		var _params$maxParallel;
		while (ongoingHandlers.size < ((_params$maxParallel = params === null || params === void 0 ? void 0 : params.maxParallel) !== null && _params$maxParallel !== void 0 ? _params$maxParallel : 1) && handlers.length > 0) {
			const handler = handlers.shift();
			ongoingHandlers.add(handler);
			handler();
		}
	};
	return (request, next) => {
		return new Observable((observer) => {
			const handler = () => next(request).subscribe({
				next: (event) => observer.next(event),
				error: (err) => observer.error(err),
				complete: () => observer.complete()
			});
			const teardown = () => {
				ongoingHandlers.delete(handler);
				remove(handlers, handler);
				tick();
			};
			handlers.push(handler);
			tick();
			return teardown;
		});
	};
};

//#endregion
//#region src/services/web-socket/web-socket.service.ts
/**
* This service is responsible for establishing bidi-directional connection to a remote server.
*/
const ISocketService = createIdentifier("univer.network.socket.service");
/**
* This service create a WebSocket connection to a remote server.
*/
var WebSocketService = class extends Disposable {
	createSocket(URL) {
		try {
			const connection = new WebSocket(URL);
			const disposables = new DisposableCollection();
			return {
				URL,
				close: (code, reason) => {
					connection.close(code, reason);
					disposables.dispose();
				},
				send: (data) => {
					connection.send(data);
				},
				open$: new Observable((subscriber) => {
					const callback = (event) => subscriber.next(event);
					connection.addEventListener("open", callback);
					disposables.add(toDisposable(() => connection.removeEventListener("open", callback)));
				}).pipe(share()),
				close$: new Observable((subscriber) => {
					const callback = (event) => subscriber.next(event);
					connection.addEventListener("close", callback);
					disposables.add(toDisposable(() => connection.removeEventListener("close", callback)));
				}).pipe(share()),
				error$: new Observable((subscriber) => {
					const callback = (event) => subscriber.next(event);
					connection.addEventListener("error", callback);
					disposables.add(toDisposable(() => connection.removeEventListener("error", callback)));
				}).pipe(share()),
				message$: new Observable((subscriber) => {
					const callback = (event) => subscriber.next(event);
					connection.addEventListener("message", callback);
					disposables.add(toDisposable(() => connection.removeEventListener("message", callback)));
				}).pipe(share())
			};
		} catch (e) {
			console.error(e);
			return null;
		}
	}
};

//#endregion
export { AuthInterceptorFactory, FetchHTTPImplementation, HTTPEventType, HTTPHeaders, HTTPProgress, HTTPRequest, HTTPResponse, HTTPResponseError, HTTPService, HTTPStatusCode, IHTTPImplementation, ISocketService, MergeInterceptorFactory, ResponseHeader, RetryInterceptorFactory, ThresholdInterceptorFactory, UniverNetworkPlugin, WebSocketService, XHRHTTPImplementation };
