// @univerjs/core/lib/cjs/facade.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
let _univerjs_core = require("@univerjs/core");
let rxjs = require("rxjs");

//#region src/facade/f-base.ts
/**
* `FBase` is a base class for all facade classes.
* It provides a way to extend classes with static and instance methods.
* The `_initialize` as a special method that will be called after the constructor. You should never call it directly.
* @ignore
*/
var FBase = class extends _univerjs_core.Disposable {
	/**
	* @ignore
	*/
	static extend(source) {
		Object.getOwnPropertyNames(source.prototype).forEach((name) => {
			if (name !== "constructor") this.prototype[name] = source.prototype[name];
		});
		Object.getOwnPropertyNames(source).forEach((name) => {
			if (name !== "prototype" && name !== "name" && name !== "length") this[name] = source[name];
		});
	}
};
/**
* @ignore
*/
const InitializerSymbol$1 = Symbol("initializers");
/**
* @ignore
*/
const ManualInitSymbol = Symbol("manualInit");
/**
* @ignore
* @hideconstructor
*/
var FBaseInitialable = class extends _univerjs_core.Disposable {
	constructor(_injector) {
		super();
		this._injector = _injector;
		const Ctor = this.constructor;
		if (Boolean(Ctor[ManualInitSymbol])) return;
		const self = this;
		const initializers = Object.getPrototypeOf(this)[InitializerSymbol$1];
		if (initializers) initializers.forEach(function(fn) {
			fn.apply(self, [_injector]);
		});
	}
	/**
	* @ignore
	*/
	_initialize(injector, ..._rest) {}
	_runInitializers(...args) {
		const initializers = Object.getPrototypeOf(this)[InitializerSymbol$1];
		if (initializers === null || initializers === void 0 ? void 0 : initializers.length) initializers.forEach((fn) => fn.apply(this, args));
	}
	static _enableManualInit() {
		this[ManualInitSymbol] = true;
	}
	/**
	* @ignore
	*/
	static extend(source) {
		Object.getOwnPropertyNames(source.prototype).forEach((name) => {
			if (name === "_initialize") {
				let initializers = this.prototype[InitializerSymbol$1];
				if (!initializers) {
					initializers = [];
					this.prototype[InitializerSymbol$1] = initializers;
				}
				initializers.push(source.prototype._initialize);
			} else if (name !== "constructor") this.prototype[name] = source.prototype[name];
		});
		Object.getOwnPropertyNames(source).forEach((name) => {
			if (name !== "prototype" && name !== "name" && name !== "length") this[name] = source[name];
		});
	}
};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/facade/f-blob.ts
var _FBlob;
let FBlob = _FBlob = class FBlob extends FBase {
	constructor(_blob, _injector) {
		super();
		this._blob = _blob;
		this._injector = _injector;
	}
	/**
	* Returns a copy of this blob.
	* @returns a new blob by copying the current blob
	* @example
	* ```ts
	* const blob = univerAPI.newBlob();
	* const newBlob = blob.copyBlob();
	* console.log(newBlob);
	* ```
	*/
	copyBlob() {
		return this._injector.createInstance(_FBlob, this._blob);
	}
	/**
	* Return the data inside this object as a blob converted to the specified content type.
	* @param contentType the content type refer to https://developer.mozilla.org/en-US/docs/Web/HTTP/MIME_types/Common_types
	* @returns a new blob by converting the current blob to the specified content type
	* @example
	* ```ts
	* const blob = univerAPI.newBlob();
	* const newBlob = blob.getAs('text/plain');
	* console.log(newBlob);
	* ```
	*/
	getAs(contentType) {
		const newBlob = this.copyBlob();
		newBlob.setContentType(contentType);
		return newBlob;
	}
	getDataAsString(charset) {
		if (this._blob === null) return Promise.resolve("");
		if (charset === void 0) return this._blob.text();
		return new Promise((resolve, reject) => {
			this._blob.arrayBuffer().then((arrayBuffer) => {
				resolve(new TextDecoder(charset).decode(arrayBuffer));
			}).catch((error) => {
				reject(/* @__PURE__ */ new Error(`Failed to read Blob as ArrayBuffer: ${error.message}`));
			});
		});
	}
	/**
	* Gets the data stored in this blob.
	* @returns the blob content as a byte array
	* @example
	* ```ts
	* const blob = univerAPI.newBlob();
	* const bytes = await blob.getBytes();
	* console.log(bytes);
	* ```
	*/
	getBytes() {
		if (!this._blob) return Promise.reject(/* @__PURE__ */ new Error("Blob is undefined or null."));
		return this._blob.arrayBuffer().then((buffer) => new Uint8Array(buffer));
	}
	/**
	* Sets the data stored in this blob.
	* @param bytes a byte array
	* @returns the blob object
	* @example
	* ```ts
	* const blob = univerAPI.newBlob();
	* const bytes = new Uint8Array(10);
	* blob.setBytes(bytes);
	* ```
	*/
	setBytes(bytes) {
		this._blob = new Blob([bytes.buffer]);
		return this;
	}
	setDataFromString(data, contentType) {
		const blob = new Blob([data], { type: contentType !== null && contentType !== void 0 ? contentType : "text/plain" });
		this._blob = blob;
		return this;
	}
	/**
	* Gets the content type of the data stored in this blob.
	* @returns the content type
	* @example
	* ```ts
	* const blob = univerAPI.newBlob();
	* const contentType = blob.getContentType();
	* console.log(contentType);
	* ```
	*/
	getContentType() {
		var _this$_blob;
		return (_this$_blob = this._blob) === null || _this$_blob === void 0 ? void 0 : _this$_blob.type;
	}
	/**
	* Sets the content type of the data stored in this blob.
	* @param contentType the content type refer to https://developer.mozilla.org/en-US/docs/Web/HTTP/MIME_types/Common_types
	* @returns the blob object
	* @example
	* ```ts
	* const blob = univerAPI.newBlob();
	* blob.setContentType('text/plain');
	* ```
	*/
	setContentType(contentType) {
		var _this$_blob2;
		this._blob = (_this$_blob2 = this._blob) === null || _this$_blob2 === void 0 ? void 0 : _this$_blob2.slice(0, this._blob.size, contentType);
		return this;
	}
};
FBlob = _FBlob = __decorate([__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector))], FBlob);

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region src/facade/f-enum.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
/**
* @hideconstructor
*/
var FEnum = class FEnum {
	static get() {
		if (this._instance) return this._instance;
		const instance = new FEnum();
		this._instance = instance;
		return instance;
	}
	/**
	* @ignore
	*/
	static extend(source) {
		Object.getOwnPropertyNames(source.prototype).forEach((name) => {
			if (name !== "constructor") this.prototype[name] = source.prototype[name];
		});
		Object.getOwnPropertyNames(source).forEach((name) => {
			if (name !== "prototype" && name !== "name" && name !== "length") this[name] = source[name];
		});
	}
	constructor() {
		for (const key in FEnum.prototype) this[key] = FEnum.prototype[key];
	}
	/**
	* Defines different types of absolute references
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.AbsoluteRefType);
	* ```
	*/
	get AbsoluteRefType() {
		return _univerjs_core.AbsoluteRefType;
	}
	/**
	* Defines different types of Univer instances
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.UniverInstanceType.UNIVER_SHEET);
	* ```
	*/
	get UniverInstanceType() {
		return _univerjs_core.UniverInstanceType;
	}
	/**
	* Represents different stages in the lifecycle
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.LifecycleStages.Rendered);
	* ```
	*/
	get LifecycleStages() {
		return _univerjs_core.LifecycleStages;
	}
	/**
	* Different types of data validation
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.DataValidationType.LIST);
	* ```
	*/
	get DataValidationType() {
		return _univerjs_core.DataValidationType;
	}
	/**
	* Different error display styles
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.DataValidationErrorStyle.WARNING);
	* ```
	*/
	get DataValidationErrorStyle() {
		return _univerjs_core.DataValidationErrorStyle;
	}
	/**
	* Different validation rendering modes
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.DataValidationRenderMode.TEXT);
	* ```
	*/
	get DataValidationRenderMode() {
		return _univerjs_core.DataValidationRenderMode;
	}
	/**
	* Different validation operators
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.DataValidationOperator.BETWEEN);
	* ```
	*/
	get DataValidationOperator() {
		return _univerjs_core.DataValidationOperator;
	}
	/**
	* Different validation states
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.DataValidationStatus.VALID);
	* ```
	*/
	get DataValidationStatus() {
		return _univerjs_core.DataValidationStatus;
	}
	/**
	* Different types of commands
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.CommandType.COMMAND);
	* ```
	*/
	get CommandType() {
		return _univerjs_core.CommandType;
	}
	/**
	* Different baseline offsets for text baseline positioning
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.BaselineOffset.SUPERSCRIPT);
	* ```
	*/
	get BaselineOffset() {
		return _univerjs_core.BaselineOffset;
	}
	/**
	* Boolean number representations
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.BooleanNumber.TRUE);
	* ```
	*/
	get BooleanNumber() {
		return _univerjs_core.BooleanNumber;
	}
	/**
	* Different horizontal text alignment options
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.HorizontalAlign.CENTER);
	* ```
	*/
	get HorizontalAlign() {
		return _univerjs_core.HorizontalAlign;
	}
	/**
	* Different text decoration styles
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.TextDecoration.DOUBLE);
	* ```
	*/
	get TextDecoration() {
		return _univerjs_core.TextDecoration;
	}
	/**
	* Different text direction options
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.TextDirection.LEFT_TO_RIGHT);
	* ```
	*/
	get TextDirection() {
		return _univerjs_core.TextDirection;
	}
	/**
	* Different vertical text alignment options
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.VerticalAlign.MIDDLE);
	* ```
	*/
	get VerticalAlign() {
		return _univerjs_core.VerticalAlign;
	}
	/**
	* Different wrap strategy options
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.WrapStrategy.WRAP);
	* ```
	*/
	get WrapStrategy() {
		return _univerjs_core.WrapStrategy;
	}
	/**
	* Different border types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.BorderType.OUTSIDE);
	* ```
	*/
	get BorderType() {
		return _univerjs_core.BorderType;
	}
	/**
	* Different border style types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.BorderStyleTypes.NONE);
	* ```
	*/
	get BorderStyleTypes() {
		return _univerjs_core.BorderStyleTypes;
	}
	/**
	* Auto fill series types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.AutoFillSeries.ALTERNATE_SERIES);
	* ```
	*/
	get AutoFillSeries() {
		return _univerjs_core.AutoFillSeries;
	}
	/**
	* Color types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.ColorType.RGB);
	* ```
	*/
	get ColorType() {
		return _univerjs_core.ColorType;
	}
	/**
	* Common hide types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.CommonHideTypes.ON);
	* ```
	*/
	get CommonHideTypes() {
		return _univerjs_core.CommonHideTypes;
	}
	/**
	* Copy paste types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.CopyPasteType.PASTE_VALUES);
	* ```
	*/
	get CopyPasteType() {
		return _univerjs_core.CopyPasteType;
	}
	/**
	* Delete direction types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.DeleteDirection.LEFT);
	* ```
	*/
	get DeleteDirection() {
		return _univerjs_core.DeleteDirection;
	}
	/**
	* Developer metadata visibility types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.DeveloperMetadataVisibility.DOCUMENT);
	* ```
	*/
	get DeveloperMetadataVisibility() {
		return _univerjs_core.DeveloperMetadataVisibility;
	}
	/**
	* Dimension types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.Dimension.ROWS);
	* ```
	*/
	get Dimension() {
		return _univerjs_core.Dimension;
	}
	/**
	* Direction types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.Direction.UP);
	* ```
	*/
	get Direction() {
		return _univerjs_core.Direction;
	}
	/**
	* Interpolation point types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.InterpolationPointType.NUMBER);
	* ```
	*/
	get InterpolationPointType() {
		return _univerjs_core.InterpolationPointType;
	}
	/**
	* Locale types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.LocaleType.EN_US);
	* ```
	*/
	get LocaleType() {
		return _univerjs_core.LocaleType;
	}
	/**
	* Mention types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.MentionType.PERSON);
	* ```
	*/
	get MentionType() {
		return _univerjs_core.MentionType;
	}
	/**
	* Protection types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.ProtectionType.RANGE);
	* ```
	*/
	get ProtectionType() {
		return _univerjs_core.ProtectionType;
	}
	/**
	* Relative date types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.RelativeDate.TODAY);
	* ```
	*/
	get RelativeDate() {
		return _univerjs_core.RelativeDate;
	}
	/**
	* Sheet types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.SheetTypes.GRID);
	* ```
	*/
	get SheetTypes() {
		return _univerjs_core.SheetTypes;
	}
	/**
	* Theme color types
	*
	* @example
	* ```ts
	* console.log(univerAPI.Enum.ThemeColorType.ACCENT1);
	* ```
	*/
	get ThemeColorType() {
		return _univerjs_core.ThemeColorType;
	}
};
_defineProperty(FEnum, "_instance", void 0);

//#endregion
//#region src/facade/f-event.ts
/**
* @hideconstructor
*/
var FEventName = class FEventName {
	static get() {
		if (this._instance) return this._instance;
		const instance = new FEventName();
		this._instance = instance;
		return instance;
	}
	/**
	* @ignore
	*/
	static extend(source) {
		Object.getOwnPropertyNames(source.prototype).forEach((name) => {
			if (name !== "constructor") this.prototype[name] = source.prototype[name];
		});
		Object.getOwnPropertyNames(source).forEach((name) => {
			if (name !== "prototype" && name !== "name" && name !== "length") this[name] = source[name];
		});
	}
	constructor() {
		for (const key in FEventName.prototype) this[key] = FEventName.prototype[key];
	}
	/**
	* Event fired when a document is created
	* @see {@link IDocCreatedParam}
	* @example
	* ```ts
	* const disposable = univerAPI.addEvent(univerAPI.Event.DocCreated, (params) => {
	*   const { unitId, type, doc, unit } = params;
	*   console.log('doc created', params);
	* });
	*
	* // Remove the event listener, use `disposable.dispose()`
	* ```
	*/
	get DocCreated() {
		return "DocCreated";
	}
	/**
	* Event fired when a document is disposed
	* @see {@link IDocDisposedEvent}
	* @example
	* ```ts
	* const disposable = univerAPI.addEvent(univerAPI.Event.DocDisposed, (params) => {
	*   const { unitId, unitType, snapshot } = params;
	*   console.log('doc disposed', params);
	* });
	*
	* // Remove the event listener, use `disposable.dispose()`
	* ```
	*/
	get DocDisposed() {
		return "DocDisposed";
	}
	/**
	* Event fired when life cycle is changed
	* @see {@link ILifeCycleChangedEvent}
	* @example
	* ```ts
	* const disposable = univerAPI.addEvent(univerAPI.Event.LifeCycleChanged, (params) => {
	*   const { stage } = params;
	*   console.log('life cycle changed', params);
	* });
	*
	* // Remove the event listener, use `disposable.dispose()`
	* ```
	*/
	get LifeCycleChanged() {
		return "LifeCycleChanged";
	}
	/**
	* Event fired when a redo command is executed
	* @see {@link ICommandEvent}
	* @example
	* ```ts
	* const disposable = univerAPI.addEvent(univerAPI.Event.Redo, (event) => {
	*   const { params, id, type } = event;
	*   console.log('redo command executed', event);
	* });
	*
	* // Remove the event listener, use `disposable.dispose()`
	* ```
	*/
	get Redo() {
		return "Redo";
	}
	/**
	* Event fired when an undo command is executed
	* @see {@link ICommandEvent}
	* @example
	* ```ts
	* const disposable = univerAPI.addEvent(univerAPI.Event.Undo, (event) => {
	*   const { params, id, type } = event;
	*   console.log('undo command executed', event);
	* });
	*
	* // Remove the event listener, use `disposable.dispose()`
	* ```
	*/
	get Undo() {
		return "Undo";
	}
	/**
	* Event fired before a redo command is executed
	* @see {@link ICommandEvent}
	* @example
	* ```ts
	* const disposable = univerAPI.addEvent(univerAPI.Event.BeforeRedo, (event) => {
	*   const { params, id, type } = event;
	*   console.log('before redo command executed', event);
	*
	*   // Cancel the redo operation
	*   event.cancel = true;
	* });
	*
	* // Remove the event listener, use `disposable.dispose()`
	* ```
	*/
	get BeforeRedo() {
		return "BeforeRedo";
	}
	/**
	* Event fired before an undo command is executed
	* @see {@link ICommandEvent}
	* @example
	* ```ts
	* const disposable = univerAPI.addEvent(univerAPI.Event.BeforeUndo, (event) => {
	*   const { params, id, type } = event;
	*   console.log('before undo command executed', event);
	*
	*   // Cancel the undo operation
	*   event.cancel = true;
	* });
	*
	* // Remove the event listener, use `disposable.dispose()`
	* ```
	*/
	get BeforeUndo() {
		return "BeforeUndo";
	}
	/**
	* Event fired when a command is executed
	* @see {@link ICommandEvent}
	* @example
	* ```ts
	* const disposable = univerAPI.addEvent(univerAPI.Event.CommandExecuted, (event) => {
	*   const { params, id, type, options } = event;
	*   console.log('command executed', event);
	* });
	*
	* // Remove the event listener, use `disposable.dispose()`
	* ```
	*/
	get CommandExecuted() {
		return "CommandExecuted";
	}
	/**
	* Event fired before a command is executed
	* @see {@link ICommandEvent}
	* @example
	* ```ts
	* const disposable = univerAPI.addEvent(univerAPI.Event.BeforeCommandExecute, (event) => {
	*   const { params, id, type, options } = event;
	*   console.log('before command executed', event);
	*
	*   // Cancel the command execution
	*   event.cancel = true;
	* });
	*
	* // Remove the event listener, use `disposable.dispose()`
	* ```
	*/
	get BeforeCommandExecute() {
		return "BeforeCommandExecute";
	}
};
_defineProperty(FEventName, "_instance", void 0);

//#endregion
//#region src/facade/f-hooks.ts
let FHooks = class FHooks extends FBase {
	constructor(_injector, _lifecycleService) {
		super();
		this._injector = _injector;
		this._lifecycleService = _lifecycleService;
	}
	/**
	* @param callback
	* @deprecated use `univerAPI.addEvent(univerAPI.Event.LifeCycleChanged, ({ stage }) => {})` as instead
	*/
	onStarting(callback) {
		return (0, _univerjs_core.toDisposable)(this._lifecycleService.lifecycle$.pipe((0, rxjs.filter)((lifecycle) => lifecycle === _univerjs_core.LifecycleStages.Starting)).subscribe(callback));
	}
	/**
	* @param callback
	* @deprecated use `univerAPI.addEvent(univerAPI.Event.LifeCycleChanged, ({ stage }) => {})` as instead
	*/
	onReady(callback) {
		return (0, _univerjs_core.toDisposable)(this._lifecycleService.lifecycle$.pipe((0, rxjs.filter)((lifecycle) => lifecycle === _univerjs_core.LifecycleStages.Ready)).subscribe(callback));
	}
	/**
	* @param callback
	* @deprecated use `univerAPI.addEvent(univerAPI.Event.LifeCycleChanged, ({ stage }) => {})` as instead
	*/
	onRendered(callback) {
		return (0, _univerjs_core.toDisposable)(this._lifecycleService.lifecycle$.pipe((0, rxjs.filter)((lifecycle) => lifecycle === _univerjs_core.LifecycleStages.Rendered)).subscribe(callback));
	}
	/**
	* @param callback
	* @deprecated use `univerAPI.addEvent(univerAPI.Event.LifeCycleChanged, ({ stage }) => {})` as instead
	*/
	onSteady(callback) {
		return (0, _univerjs_core.toDisposable)(this._lifecycleService.lifecycle$.pipe((0, rxjs.filter)((lifecycle) => lifecycle === _univerjs_core.LifecycleStages.Steady)).subscribe(callback));
	}
	/**
	* @param callback
	* @deprecated use `univerAPI.addEvent(univerAPI.Event.BeforeUndo, (event) => {})` as instead
	*/
	onBeforeUndo(callback) {
		return this._injector.get(_univerjs_core.ICommandService).beforeCommandExecuted((command) => {
			if (command.id === _univerjs_core.UndoCommand.id) {
				const action = this._injector.get(_univerjs_core.IUndoRedoService).pitchTopUndoElement();
				if (action) callback(action);
			}
		});
	}
	/**
	* @param callback
	* @deprecated use `univerAPI.addEvent(univerAPI.Event.Undo, (event) => {})` as instead
	*/
	onUndo(callback) {
		return this._injector.get(_univerjs_core.ICommandService).onCommandExecuted((command) => {
			if (command.id === _univerjs_core.UndoCommand.id) {
				const action = this._injector.get(_univerjs_core.IUndoRedoService).pitchTopUndoElement();
				if (action) callback(action);
			}
		});
	}
	/**
	* @param callback
	* @deprecated use `univerAPI.addEvent(univerAPI.Event.BeforeRedo, (event) => {})` as instead
	*/
	onBeforeRedo(callback) {
		return this._injector.get(_univerjs_core.ICommandService).beforeCommandExecuted((command) => {
			if (command.id === _univerjs_core.RedoCommand.id) {
				const action = this._injector.get(_univerjs_core.IUndoRedoService).pitchTopRedoElement();
				if (action) callback(action);
			}
		});
	}
	/**
	* @param callback
	* @deprecated use `univerAPI.addEvent(univerAPI.Event.Redo, (event) => {})` as instead
	*/
	onRedo(callback) {
		return this._injector.get(_univerjs_core.ICommandService).onCommandExecuted((command) => {
			if (command.id === _univerjs_core.RedoCommand.id) {
				const action = this._injector.get(_univerjs_core.IUndoRedoService).pitchTopRedoElement();
				if (action) callback(action);
			}
		});
	}
};
FHooks = __decorate([__decorateParam(0, (0, _univerjs_core.Inject)(_univerjs_core.Injector)), __decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.LifecycleService))], FHooks);

//#endregion
//#region src/facade/f-doc.ts
let FDoc = class FDoc extends FBaseInitialable {
	constructor(doc, _injector) {
		super(_injector);
		this.doc = doc;
	}
};
FDoc = __decorate([__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector))], FDoc);

//#endregion
//#region src/facade/f-event-registry.ts
var FEventRegistry = class {
	constructor() {
		_defineProperty(this, "_eventRegistry", /* @__PURE__ */ new Map());
		_defineProperty(this, "_eventHandlerMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_eventHandlerRegisted", /* @__PURE__ */ new Map());
	}
	_ensureEventRegistry(event) {
		if (!this._eventRegistry.has(event)) this._eventRegistry.set(event, new _univerjs_core.Registry());
		return this._eventRegistry.get(event);
	}
	registerEventHandler(event, handler) {
		const current = this._eventHandlerMap.get(event);
		if (current) current.add(handler);
		else this._eventHandlerMap.set(event, new Set([handler]));
		if (this._ensureEventRegistry(event).getData().length) this._initEventHandler(event);
		return (0, _univerjs_core.toDisposable)(() => {
			var _this$_eventHandlerMa, _this$_eventHandlerRe, _this$_eventHandlerRe2;
			(_this$_eventHandlerMa = this._eventHandlerMap.get(event)) === null || _this$_eventHandlerMa === void 0 || _this$_eventHandlerMa.delete(handler);
			(_this$_eventHandlerRe = this._eventHandlerRegisted.get(event)) === null || _this$_eventHandlerRe === void 0 || (_this$_eventHandlerRe = _this$_eventHandlerRe.get(handler)) === null || _this$_eventHandlerRe === void 0 || _this$_eventHandlerRe.dispose();
			(_this$_eventHandlerRe2 = this._eventHandlerRegisted.get(event)) === null || _this$_eventHandlerRe2 === void 0 || _this$_eventHandlerRe2.delete(handler);
		});
	}
	removeEvent(event, callback) {
		const map = this._ensureEventRegistry(event);
		map.delete(callback);
		if (map.getData().length === 0) {
			const disposable = this._eventHandlerRegisted.get(event);
			disposable === null || disposable === void 0 || disposable.forEach((d) => d.dispose());
			this._eventHandlerRegisted.delete(event);
		}
	}
	_initEventHandler(event) {
		let current = this._eventHandlerRegisted.get(event);
		const handlers = this._eventHandlerMap.get(event);
		if (!handlers) return;
		if (!current || current.size === 0) {
			current = /* @__PURE__ */ new Map();
			this._eventHandlerRegisted.set(event, current);
			handlers === null || handlers === void 0 || handlers.forEach((handler) => {
				current === null || current === void 0 || current.set(handler, (0, _univerjs_core.toDisposable)(handler()));
			});
		}
	}
	/**
	* Add an event listener
	* @param {string} event key of event
	* @param {(params: IEventParamConfig[typeof event]) => void} callback callback when event triggered
	* @returns {Disposable} The Disposable instance, for remove the listener
	* @example
	* ```ts
	* univerAPI.addEvent(univerAPI.Event.LifeCycleChanged, (params) => {
	*   const { stage } = params;
	*   console.log('life cycle changed', params);
	* });
	* ```
	*/
	addEvent(event, callback) {
		this._ensureEventRegistry(event).add(callback);
		this._initEventHandler(event);
		return (0, _univerjs_core.toDisposable)(() => this.removeEvent(event, callback));
	}
	/**
	* Fire an event, used in internal only.
	* @param {string} event key of event
	* @param {any} params params of event
	* @returns {boolean} should cancel
	* @example
	* ```ts
	* this.fireEvent(univerAPI.Event.LifeCycleChanged, params);
	* ```
	*/
	fireEvent(event, params) {
		var _this$_eventRegistry$;
		(_this$_eventRegistry$ = this._eventRegistry.get(event)) === null || _this$_eventRegistry$ === void 0 || _this$_eventRegistry$.getData().forEach((callback) => {
			callback(params);
		});
		return params.cancel;
	}
};

//#endregion
//#region src/facade/f-usermanager.ts
let FUserManager = class FUserManager extends FBase {
	constructor(_injector, _userManagerService) {
		super();
		this._injector = _injector;
		this._userManagerService = _userManagerService;
	}
	/**
	* Get current user info.
	* @returns {IUser} Current user info.
	* @example
	* ```typescript
	* univerAPI.getUserManager().getCurrentUser();
	* ```
	*/
	getCurrentUser() {
		return this._userManagerService.getCurrentUser();
	}
};
FUserManager = __decorate([__decorateParam(0, (0, _univerjs_core.Inject)(_univerjs_core.Injector)), __decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.UserManagerService))], FUserManager);

//#endregion
//#region src/facade/f-util.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
/**
* @ignore
*/
var FUtil = class FUtil {
	static get() {
		if (this._instance) return this._instance;
		const instance = new FUtil();
		this._instance = instance;
		return instance;
	}
	/**
	* @ignore
	*/
	static extend(source) {
		Object.getOwnPropertyNames(source.prototype).forEach((name) => {
			if (name !== "constructor") this.prototype[name] = source.prototype[name];
		});
		Object.getOwnPropertyNames(source).forEach((name) => {
			if (name !== "prototype" && name !== "name" && name !== "length") this[name] = source[name];
		});
	}
	/**
	* Rectangle utils, including range operations likes merge, subtract, split
	*
	* @example
	* ```ts
	* const ranges = [
	*   { startRow: 0, startColumn: 0, endRow: 1, endColumn: 1 },
	*   { startRow: 1, startColumn: 1, endRow: 2, endColumn: 2 }
	* ];
	* const merged = univerAPI.Util.rectangle.mergeRanges(ranges);
	* console.log(merged);
	* ```
	*/
	get rectangle() {
		return _univerjs_core.Rectangle;
	}
	/**
	* Number format utils, including parse and strigify about date, price, etc
	*
	* @example
	* ```ts
	* const text = univerAPI.Util.numfmt.format('#,##0.00', 1234.567);
	* console.log(text);
	* ```
	*/
	get numfmt() {
		return _univerjs_core.numfmt;
	}
	/**
	* common tools
	*
	* @example
	* ```ts
	* const key = univerAPI.Util.tools.generateRandomId(6);
	* console.log(key);
	* ```
	*/
	get tools() {
		return _univerjs_core.Tools;
	}
};
_defineProperty(FUtil, "_instance", void 0);

//#endregion
//#region src/facade/f-univer.ts
var _FUniver;
/**
* @ignore
*/
const InitializerSymbol = Symbol("initializers");
let FUniver = _FUniver = class FUniver extends _univerjs_core.Disposable {
	/**
	* Create an FUniver instance, if the injector is not provided, it will create a new Univer instance.
	* @static
	* @param {Univer | Injector} wrapped - The Univer instance or injector instance.
	* @returns {FUniver} - The FUniver instance.
	*
	* @example
	* ```ts
	* const univerAPI = FUniver.newAPI(univer);
	* ```
	*/
	static newAPI(wrapped) {
		return (wrapped instanceof _univerjs_core.Univer ? wrapped.__getInjector() : wrapped).createInstance(_FUniver);
	}
	/**
	* @ignore
	*/
	_initialize(injector) {}
	/**
	* @ignore
	*/
	static extend(source) {
		Object.getOwnPropertyNames(source.prototype).forEach((name) => {
			if (name === "_initialize") {
				let initializers = this.prototype[InitializerSymbol];
				if (!initializers) {
					initializers = [];
					this.prototype[InitializerSymbol] = initializers;
				}
				initializers.push(source.prototype._initialize);
			} else if (name !== "constructor") this.prototype[name] = source.prototype[name];
		});
		Object.getOwnPropertyNames(source).forEach((name) => {
			if (name !== "prototype" && name !== "name" && name !== "length") this[name] = source[name];
		});
	}
	constructor(_injector, _commandService, _univerInstanceService, _lifecycleService) {
		super();
		this._injector = _injector;
		this._commandService = _commandService;
		this._univerInstanceService = _univerInstanceService;
		this._lifecycleService = _lifecycleService;
		_defineProperty(this, "_eventRegistry", new FEventRegistry());
		_defineProperty(this, "registerEventHandler", (event, handler) => {
			return this._eventRegistry.registerEventHandler(event, handler);
		});
		this.disposeWithMe(this.registerEventHandler(this.Event.LifeCycleChanged, () => (0, _univerjs_core.toDisposable)(this._lifecycleService.lifecycle$.subscribe((stage) => {
			this.fireEvent(this.Event.LifeCycleChanged, { stage });
		}))));
		this._initUnitEvent(this._injector);
		this._initBeforeCommandEvent(this._injector);
		this._initCommandEvent(this._injector);
		this._injector.onDispose(() => {
			this.dispose();
		});
		const initializers = Object.getPrototypeOf(this)[InitializerSymbol];
		if (initializers) {
			const self = this;
			initializers.forEach(function(fn) {
				fn.apply(self, [_injector]);
			});
		}
	}
	_initCommandEvent(injector) {
		const commandService = injector.get(_univerjs_core.ICommandService);
		this.disposeWithMe(this.registerEventHandler(this.Event.Redo, () => commandService.onCommandExecuted((commandInfo) => {
			const { id, type: propType, params } = commandInfo;
			if (commandInfo.id === _univerjs_core.RedoCommand.id) {
				const eventParams = {
					id,
					type: propType,
					params
				};
				this.fireEvent(this.Event.Redo, eventParams);
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.Undo, () => commandService.onCommandExecuted((commandInfo) => {
			const { id, type: propType, params } = commandInfo;
			if (commandInfo.id === _univerjs_core.UndoCommand.id) {
				const eventParams = {
					id,
					type: propType,
					params
				};
				this.fireEvent(this.Event.Undo, eventParams);
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.CommandExecuted, () => commandService.onCommandExecuted((commandInfo, options) => {
			const { id, type: propType, params } = commandInfo;
			if (commandInfo.id !== _univerjs_core.RedoCommand.id && commandInfo.id !== _univerjs_core.UndoCommand.id) {
				const eventParams = {
					id,
					type: propType,
					params,
					options
				};
				this.fireEvent(this.Event.CommandExecuted, eventParams);
			}
		})));
	}
	_initBeforeCommandEvent(injector) {
		const commandService = injector.get(_univerjs_core.ICommandService);
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeRedo, () => commandService.beforeCommandExecuted((commandInfo) => {
			const { id, type: propType, params } = commandInfo;
			if (commandInfo.id === _univerjs_core.RedoCommand.id) {
				const eventParams = {
					id,
					type: propType,
					params
				};
				this.fireEvent(this.Event.BeforeRedo, eventParams);
				if (eventParams.cancel) throw new _univerjs_core.CanceledError();
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeUndo, () => commandService.beforeCommandExecuted((commandInfo) => {
			const { id, type: propType, params } = commandInfo;
			if (commandInfo.id === _univerjs_core.UndoCommand.id) {
				const eventParams = {
					id,
					type: propType,
					params
				};
				this.fireEvent(this.Event.BeforeUndo, eventParams);
				if (eventParams.cancel) throw new _univerjs_core.CanceledError();
			}
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.BeforeCommandExecute, () => commandService.beforeCommandExecuted((commandInfo, options) => {
			const { id, type: propType, params } = commandInfo;
			if (commandInfo.id !== _univerjs_core.RedoCommand.id && commandInfo.id !== _univerjs_core.UndoCommand.id) {
				const eventParams = {
					id,
					type: propType,
					params,
					options
				};
				this.fireEvent(this.Event.BeforeCommandExecute, eventParams);
				if (eventParams.cancel) throw new _univerjs_core.CanceledError();
			}
		})));
	}
	_initUnitEvent(injector) {
		const univerInstanceService = injector.get(_univerjs_core.IUniverInstanceService);
		this.disposeWithMe(this.registerEventHandler(this.Event.DocDisposed, () => univerInstanceService.unitDisposed$.subscribe((unit) => {
			if (unit.type === _univerjs_core.UniverInstanceType.UNIVER_DOC) this.fireEvent(this.Event.DocDisposed, {
				unitId: unit.getUnitId(),
				unitType: unit.type,
				snapshot: unit.getSnapshot()
			});
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.DocCreated, () => univerInstanceService.unitAdded$.subscribe((event) => {
			const { unit } = event;
			if (unit.type === _univerjs_core.UniverInstanceType.UNIVER_DOC) {
				const doc = unit;
				const docUnit = injector.createInstance(FDoc, doc);
				this.fireEvent(this.Event.DocCreated, {
					unitId: unit.getUnitId(),
					type: unit.type,
					doc: docUnit,
					unit: docUnit
				});
			}
		})));
	}
	/**
	* Dispose the UniverSheet by the `unitId`. The UniverSheet would be unload from the application.
	* @param unitId The unit id of the UniverSheet.
	* @returns Whether the Univer instance is disposed successfully.
	*
	* @example
	* ```ts
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const unitId = fWorkbook?.getId();
	*
	* if (unitId) {
	*   univerAPI.disposeUnit(unitId);
	* }
	* ```
	*/
	disposeUnit(unitId) {
		return this._univerInstanceService.disposeUnit(unitId);
	}
	/**
	* Get the current lifecycle stage.
	* @returns {LifecycleStages} - The current lifecycle stage.
	*
	* @example
	* ```ts
	* const stage = univerAPI.getCurrentLifecycleStage();
	* console.log(stage);
	* ```
	*/
	getCurrentLifecycleStage() {
		return this._injector.get(_univerjs_core.LifecycleService).stage;
	}
	/**
	* Undo an editing on the currently focused document.
	* @returns {Promise<boolean>} undo result
	*
	* @example
	* ```ts
	* await univerAPI.undo();
	* ```
	*/
	undo() {
		return this._commandService.executeCommand(_univerjs_core.UndoCommand.id);
	}
	/**
	* Redo an editing on the currently focused document.
	* @returns {Promise<boolean>} redo result
	*
	* @example
	* ```ts
	* await univerAPI.redo();
	* ```
	*/
	redo() {
		return this._commandService.executeCommand(_univerjs_core.RedoCommand.id);
	}
	/**
	* Toggle dark mode on or off.
	* @param {boolean} isDarkMode - Whether the dark mode is enabled.
	* @example
	* ```ts
	* univerAPI.toggleDarkMode(true);
	* ```
	*/
	toggleDarkMode(isDarkMode) {
		this._injector.get(_univerjs_core.ThemeService).setDarkMode(isDarkMode);
	}
	/**
	* Load locales for the given locale.
	* @description This method is utilized to load locales, which can be either built-in or custom-defined.
	* @param {string} locale - A unique locale identifier.
	* @param {ILanguagePack} locales  - The locales object containing the translations.
	* @example
	* ```ts
	* univerAPI.loadLocales('esES', {
	*   'Hello World': 'Hola Mundo',
	* });
	* ```
	*/
	loadLocales(locale, locales) {
		this._injector.get(_univerjs_core.LocaleService).load({ [locale]: locales });
	}
	/**
	* Set the current locale.
	* @param {string} locale - A unique locale identifier.
	* @example
	* ```ts
	* univerAPI.setLocale('esES');
	* ```
	*/
	setLocale(locale) {
		this._injector.get(_univerjs_core.LocaleService).setLocale(locale);
	}
	/**
	* Register a callback that will be triggered before invoking a command.
	* @deprecated use `univerAPI.addEvent(univerAPI.Event.BeforeCommandExecute, (event) => {})` instead.
	* @param {CommandListener} callback The callback.
	* @returns {IDisposable} The disposable instance.
	*/
	onBeforeCommandExecute(callback) {
		return this._commandService.beforeCommandExecuted((command, options) => {
			callback(command, options);
		});
	}
	/**
	* Register a callback that will be triggered when a command is invoked.
	* @deprecated use `univerAPI.addEvent(univerAPI.Event.CommandExecuted, (event) => {})` instead.
	* @param {CommandListener} callback The callback.
	* @returns {IDisposable} The disposable instance.
	*/
	onCommandExecuted(callback) {
		return this._commandService.onCommandExecuted((command, options) => {
			callback(command, options);
		});
	}
	/**
	* Execute a command with the given id and parameters.
	* @param id Identifier of the command.
	* @param params Parameters of this execution.
	* @param options Options of this execution.
	* @returns The result of the execution. It is a boolean value by default which indicates the command is executed.
	*
	* @example
	* ```ts
	* univerAPI.executeCommand('sheet.command.set-range-values', {
	*   value: { v: "Hello, Univer!" },
	*   range: { startRow: 0, startColumn: 0, endRow: 0, endColumn: 0 }
	* });
	* ```
	*/
	executeCommand(id, params, options) {
		return this._commandService.executeCommand(id, params, options);
	}
	/**
	* Execute a command with the given id and parameters synchronously.
	* @param id Identifier of the command.
	* @param params Parameters of this execution.
	* @param options Options of this execution.
	* @returns The result of the execution. It is a boolean value by default which indicates the command is executed.
	*
	* @example
	* ```ts
	* univerAPI.syncExecuteCommand('sheet.command.set-range-values', {
	*   value: { v: "Hello, Univer!" },
	*   range: { startRow: 0, startColumn: 0, endRow: 0, endColumn: 0 }
	* });
	* ```
	*/
	syncExecuteCommand(id, params, options) {
		return this._commandService.syncExecuteCommand(id, params, options);
	}
	/**
	* Get hooks.
	* @deprecated use `addEvent` instead.
	* @returns {FHooks} FHooks instance
	*/
	getHooks() {
		return this._injector.createInstance(FHooks);
	}
	get Enum() {
		return FEnum.get();
	}
	get Event() {
		return FEventName.get();
	}
	get Util() {
		return FUtil.get();
	}
	/**
	* Add an event listener
	* @param {string} event key of event
	* @param {(params: IEventParamConfig[typeof event]) => void} callback callback when event triggered
	* @returns {Disposable} The Disposable instance, for remove the listener
	* @example
	* ```ts
	* // Add life cycle changed event listener
	* const disposable = univerAPI.addEvent(univerAPI.Event.LifeCycleChanged, (params) => {
	*   const { stage } = params;
	*   console.log('life cycle changed', params);
	* });
	*
	* // Remove the event listener, use `disposable.dispose()`
	* ```
	*/
	addEvent(event, callback) {
		if (!event || !callback) throw new Error("Cannot add empty event");
		return this._eventRegistry.addEvent(event, callback);
	}
	/**
	* Fire an event, used in internal only.
	* @param {string} event key of event
	* @param {any} params params of event
	* @returns {boolean} should cancel
	* @example
	* ```ts
	* this.fireEvent(univerAPI.Event.LifeCycleChanged, params);
	* ```
	*/
	fireEvent(event, params) {
		return this._eventRegistry.fireEvent(event, params);
	}
	getUserManager() {
		return this._injector.createInstance(FUserManager);
	}
	/**
	* Create a new blob.
	* @returns {FBlob} The new blob instance
	* @example
	* ```ts
	* const blob = univerAPI.newBlob();
	* ```
	*/
	newBlob() {
		return this._injector.createInstance(FBlob, null);
	}
	/**
	* Create a new rich text.
	* @param {IDocumentData} data
	* @returns {RichTextBuilder} The new rich text instance
	* @example
	* ```ts
	* const richText = univerAPI.newRichText({ body: { dataStream: 'Hello World\r\n' } });
	* const fWorksheet = univerAPI.getActiveWorkbook().getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const range = fWorksheet.getRange('A1');
	* range.setRichTextValueForCell(richText);
	* ```
	*/
	newRichText(data) {
		return _univerjs_core.RichTextBuilder.create(data);
	}
	/**
	* Create a new rich text value.
	* @param {IDocumentData} data - The rich text data
	* @returns {RichTextValue} The new rich text value instance
	* @example
	* ```ts
	* const richTextValue = univerAPI.newRichTextValue({ body: { dataStream: 'Hello World\r\n' } });
	* const fWorksheet = univerAPI.getActiveWorkbook().getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const range = fWorksheet.getRange('A1');
	* range.setRichTextValueForCell(richTextValue);
	* ```
	*/
	newRichTextValue(data) {
		return _univerjs_core.RichTextValue.create(data);
	}
	/**
	* Create a new paragraph style.
	* @param {IParagraphStyle} style - The paragraph style
	* @returns {ParagraphStyleBuilder} The new paragraph style instance
	* @example
	* ```ts
	* const richText = univerAPI.newRichText({ body: { dataStream: 'Hello World\r\n' } });
	* const paragraphStyle = univerAPI.newParagraphStyle({ textStyle: { ff: 'Arial', fs: 12, it: univerAPI.Enum.BooleanNumber.TRUE, bl: univerAPI.Enum.BooleanNumber.TRUE } });
	* richText.insertParagraph(paragraphStyle);
	* const fWorksheet = univerAPI.getActiveWorkbook().getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const range = fWorksheet.getRange('A1');
	* range.setRichTextValueForCell(richText);
	* ```
	*/
	newParagraphStyle(style) {
		return _univerjs_core.ParagraphStyleBuilder.create(style);
	}
	/**
	* Create a new paragraph style value.
	* @param {IParagraphStyle} style - The paragraph style
	* @returns {ParagraphStyleValue} The new paragraph style value instance
	* @example
	* ```ts
	* const paragraphStyleValue = univerAPI.newParagraphStyleValue();
	* ```
	*/
	newParagraphStyleValue(style) {
		return _univerjs_core.ParagraphStyleValue.create(style);
	}
	/**
	* Create a new text style.
	* @param {ITextStyle} style - The text style
	* @returns {TextStyleBuilder} The new text style instance
	* @example
	* ```ts
	* const textStyle = univerAPI.newTextStyle();
	* ```
	*/
	newTextStyle(style) {
		return _univerjs_core.TextStyleBuilder.create(style);
	}
	/**
	* Create a new text style value.
	* @param {ITextStyle} style - The text style
	* @returns {TextStyleValue} The new text style value instance
	* @example
	* ```ts
	* const textStyleValue = univerAPI.newTextStyleValue();
	* ```
	*/
	newTextStyleValue(style) {
		return _univerjs_core.TextStyleValue.create(style);
	}
	/**
	* Create a new text decoration.
	* @param {ITextDecoration} decoration - The text decoration
	* @returns {TextDecorationBuilder} The new text decoration instance
	* @example
	* ```ts
	* const decoration = univerAPI.newTextDecoration();
	* ```
	*/
	newTextDecoration(decoration) {
		return new _univerjs_core.TextDecorationBuilder(decoration);
	}
};
FUniver = _FUniver = __decorate([
	__decorateParam(0, (0, _univerjs_core.Inject)(_univerjs_core.Injector)),
	__decorateParam(1, _univerjs_core.ICommandService),
	__decorateParam(2, _univerjs_core.IUniverInstanceService),
	__decorateParam(3, (0, _univerjs_core.Inject)(_univerjs_core.LifecycleService))
], FUniver);

//#endregion
exports.FBase = FBase;
exports.FBaseInitialable = FBaseInitialable;
Object.defineProperty(exports, 'FBlob', {
  enumerable: true,
  get: function () {
    return FBlob;
  }
});
exports.FEnum = FEnum;
exports.FEventName = FEventName;
Object.defineProperty(exports, 'FHooks', {
  enumerable: true,
  get: function () {
    return FHooks;
  }
});
Object.defineProperty(exports, 'FUniver', {
  enumerable: true,
  get: function () {
    return FUniver;
  }
});
exports.FUtil = FUtil;
