// @univerjs/sheets-numfmt/lib/cjs/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
let _univerjs_core = require("@univerjs/core");
let _univerjs_sheets = require("@univerjs/sheets");
let rxjs = require("rxjs");
let _univerjs_engine_formula = require("@univerjs/engine-formula");

//#region src/base/const/currency-symbols.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const localeCurrencySymbolMap = new Map([
	[_univerjs_core.LocaleType.EN_US, "$"],
	[_univerjs_core.LocaleType.RU_RU, "₽"],
	[_univerjs_core.LocaleType.VI_VN, "₫"],
	[_univerjs_core.LocaleType.ZH_CN, "¥"],
	[_univerjs_core.LocaleType.ZH_TW, "NT$"],
	[_univerjs_core.LocaleType.ZH_HK, "HK$"],
	[_univerjs_core.LocaleType.FR_FR, "€"],
	[_univerjs_core.LocaleType.FA_IR, "﷼"],
	[_univerjs_core.LocaleType.KO_KR, "₩"],
	[_univerjs_core.LocaleType.ES_ES, "€"],
	[_univerjs_core.LocaleType.CA_ES, "€"],
	[_univerjs_core.LocaleType.SK_SK, "€"],
	[_univerjs_core.LocaleType.JA_JP, "¥"],
	[_univerjs_core.LocaleType.PT_BR, "R$"],
	[_univerjs_core.LocaleType.DE_DE, "€"],
	[_univerjs_core.LocaleType.IT_IT, "€"],
	[_univerjs_core.LocaleType.ID_ID, "Rp"],
	[_univerjs_core.LocaleType.PL_PL, "zł"],
	[_univerjs_core.LocaleType.AR_SA, "﷼"]
]);
/**
* Get the currency symbol icon based on the locale.
*/
function getCurrencySymbolIconByLocale(locale) {
	switch (locale) {
		case _univerjs_core.LocaleType.CA_ES:
		case _univerjs_core.LocaleType.DE_DE:
		case _univerjs_core.LocaleType.ES_ES:
		case _univerjs_core.LocaleType.FR_FR:
		case _univerjs_core.LocaleType.IT_IT:
		case _univerjs_core.LocaleType.SK_SK: return {
			icon: "EuroIcon",
			symbol: localeCurrencySymbolMap.get(locale) || "€",
			locale
		};
		case _univerjs_core.LocaleType.RU_RU: return {
			icon: "RoubleIcon",
			symbol: localeCurrencySymbolMap.get(locale) || "₽",
			locale
		};
		case _univerjs_core.LocaleType.JA_JP:
		case _univerjs_core.LocaleType.ZH_CN: return {
			icon: "RmbIcon",
			symbol: localeCurrencySymbolMap.get(locale) || "¥",
			locale
		};
		case _univerjs_core.LocaleType.AR_SA:
		case _univerjs_core.LocaleType.EN_US:
		case _univerjs_core.LocaleType.FA_IR:
		case _univerjs_core.LocaleType.ID_ID:
		case _univerjs_core.LocaleType.KO_KR:
		case _univerjs_core.LocaleType.PL_PL:
		case _univerjs_core.LocaleType.PT_BR:
		case _univerjs_core.LocaleType.VI_VN:
		case _univerjs_core.LocaleType.ZH_HK:
		case _univerjs_core.LocaleType.ZH_TW:
		default: return {
			icon: "DollarIcon",
			symbol: localeCurrencySymbolMap.get(locale) || "$",
			locale
		};
	}
}
/**
* Get the currency symbol by locale.
*/
function getCurrencySymbolByLocale(locale) {
	return localeCurrencySymbolMap.get(locale) || "$";
}
/**
* Get the currency format string based on the locale and number of digits.
*/
function getCurrencyFormat(locale, numberDigits = 2) {
	let _numberDigits = numberDigits;
	if (numberDigits > 127) _numberDigits = 127;
	let decimal = "";
	if (_numberDigits > 0) decimal = `.${"0".repeat(_numberDigits)}`;
	return `"${getCurrencySymbolByLocale(locale)}"#,##0${decimal}_);[Red]("${getCurrencySymbolByLocale(locale)}"#,##0${decimal})`;
}

//#endregion
//#region src/base/const/formatdetail.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const DATEFMTLISG = [
	{
		label: "1930-08-05",
		suffix: "yyyy-MM-dd"
	},
	{
		label: "1930/08/05",
		suffix: "yyyy/MM/dd"
	},
	{
		label: "1930年08月05日",
		suffix: "yyyy\"年\"MM\"月\"dd\"日\""
	},
	{
		label: "08-05",
		suffix: "MM-dd"
	},
	{
		label: "8月5日",
		suffix: "M\"月\"d\"日\""
	},
	{
		label: "13:30:30",
		suffix: "h:mm:ss"
	},
	{
		label: "13:30",
		suffix: "h:mm"
	},
	{
		label: "下午01:30",
		suffix: "A/P hh:mm"
	},
	{
		label: "下午1:30",
		suffix: "A/P h:mm"
	},
	{
		label: "下午1:30:30",
		suffix: "A/P h:mm:ss"
	},
	{
		label: "08-05 下午 01:30",
		suffix: "MM-dd A/P hh:mm"
	}
];
const NUMBERFORMAT = [
	{
		label: "(1,235)",
		suffix: "#,##0_);(#,##0)"
	},
	{
		label: "(1,235) ",
		suffix: "#,##0_);[Red](#,##0)",
		color: "red"
	},
	{
		label: "1,234.56",
		suffix: "#,##0.00_);#,##0.00"
	},
	{
		label: "1,234.56",
		suffix: "#,##0.00_);[Red]#,##0.00",
		color: "red"
	},
	{
		label: "-1,234.56",
		suffix: "#,##0.00_);-#,##0.00"
	},
	{
		label: "-1,234.56",
		suffix: "#,##0.00_);[Red]-#,##0.00",
		color: "red"
	}
];
const CURRENCYFORMAT = [
	{
		label: (suffix) => `${suffix}1,235`,
		suffix: (suffix) => `"${suffix}"#,##0.00_);"${suffix}"#,##0.00`
	},
	{
		label: (suffix) => `${suffix}1,235`,
		suffix: (suffix) => `"${suffix}"#,##0.00_);[Red]"${suffix}"#,##0.00`,
		color: "red"
	},
	{
		label: (suffix) => `(${suffix}1,235)`,
		suffix: (suffix) => `"${suffix}"#,##0.00_);("${suffix}"#,##0.00)`
	},
	{
		label: (suffix) => `(${suffix}1,235)`,
		suffix: (suffix) => `"${suffix}"#,##0.00_);[Red]("${suffix}"#,##0.00)`,
		color: "red"
	},
	{
		label: (suffix) => `-${suffix}1,235`,
		suffix: (suffix) => `"${suffix}"#,##0.00_);-"${suffix}"#,##0.00`
	},
	{
		label: (suffix) => `-${suffix}1,235`,
		suffix: (suffix) => `"${suffix}"#,##0.00_);[Red]-"${suffix}"#,##0.00`,
		color: "red"
	}
];

//#endregion
//#region src/utils/decimal.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
/**
* the function decimal just use positive,negative configuration ignored
*/
const getDecimalFromPattern = (pattern, defaultValue = 0) => {
	var _info$maxDecimals;
	if (!pattern) return defaultValue;
	return (_info$maxDecimals = _univerjs_core.numfmt.getFormatInfo(pattern).maxDecimals) !== null && _info$maxDecimals !== void 0 ? _info$maxDecimals : defaultValue;
};
const getDecimalString = (length) => new Array(Math.min(Math.max(0, Number(length)), 30)).fill(0).join("");
const setPatternDecimal = (patterns, decimalLength) => {
	return patterns.split(";").map((pattern) => {
		if (/\.0?/.test(pattern)) return pattern.replace(/\.0*/g, `${decimalLength > 0 ? "." : ""}${getDecimalString(Number(decimalLength || 0))}`);
		if (/0([^0]?)|0$/.test(pattern)) return pattern.replace(/0([^0]+)|0$/, `0${decimalLength > 0 ? "." : ""}${getDecimalString(Number(decimalLength || 0))}$1`);
		return pattern;
	}).join(";");
};
const isPatternHasDecimal = (pattern) => /\.0?/.test(pattern) || /0([^0]?)|0$/.test(pattern);

//#endregion
//#region src/commands/commands/set-numfmt.command.ts
const SetNumfmtCommand = {
	id: "sheet.command.numfmt.set.numfmt",
	type: _univerjs_core.CommandType.COMMAND,
	handler: (accessor, params) => {
		if (!params) return false;
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const univerInstanceService = accessor.get(_univerjs_core.IUniverInstanceService);
		const undoRedoService = accessor.get(_univerjs_core.IUndoRedoService);
		const target = (0, _univerjs_sheets.getSheetCommandTarget)(univerInstanceService, params);
		if (!target) return false;
		const { unitId, subUnitId, worksheet } = target;
		const setCells = params.values.filter((value) => !!value.pattern);
		const removeCells = params.values.filter((value) => !value.pattern);
		const setRedos = (0, _univerjs_sheets.transformCellsToRange)(unitId, subUnitId, setCells);
		const removeRedos = {
			unitId,
			subUnitId,
			ranges: removeCells.map((cell) => ({
				startColumn: cell.col,
				startRow: cell.row,
				endColumn: cell.col,
				endRow: cell.row
			}))
		};
		const redos = [];
		const undos = [];
		if (setCells.length) {
			const setCellTypeObj = setCells.reduce((pre, cur) => {
				if ((0, _univerjs_core.isTextFormat)(cur.pattern)) pre.setValue(cur.row, cur.col, { t: _univerjs_core.CellValueType.STRING });
				const cell = worksheet.getCellRaw(cur.row, cur.col);
				if (cell) {
					const type = (0, _univerjs_sheets.checkCellValueType)(cell.v);
					if (type !== cell.t) pre.setValue(cur.row, cur.col, { t: type });
				}
				return pre;
			}, new _univerjs_core.ObjectMatrix()).getMatrix();
			const undoSetCellTypeObj = new _univerjs_core.ObjectMatrix();
			new _univerjs_core.ObjectMatrix(setCellTypeObj).forValue((row, col) => {
				const cell = worksheet.getCellRaw(row, col);
				if (cell) undoSetCellTypeObj.setValue(row, col, { t: cell.t });
				else undoSetCellTypeObj.setValue(row, col, { t: void 0 });
			});
			Object.keys(setRedos.values).forEach((key) => {
				const v = setRedos.values[key];
				v.ranges = (0, _univerjs_sheets.rangeMerge)(v.ranges);
			});
			redos.push({
				id: _univerjs_sheets.SetNumfmtMutation.id,
				params: setRedos
			});
			const undo = (0, _univerjs_sheets.factorySetNumfmtUndoMutation)(accessor, setRedos);
			undos.push(...undo);
		}
		if (removeCells.length) {
			removeRedos.ranges = (0, _univerjs_sheets.rangeMerge)(removeRedos.ranges);
			const setCellTypeObj = removeCells.reduce((pre, cur) => {
				const cell = worksheet.getCellRaw(cur.row, cur.col);
				if (cell) {
					const type = (0, _univerjs_sheets.checkCellValueType)(cell.v);
					if (type !== cell.t) pre.setValue(cur.row, cur.col, { t: type });
				}
				return pre;
			}, new _univerjs_core.ObjectMatrix()).getMatrix();
			const undoSetCellTypeObj = new _univerjs_core.ObjectMatrix();
			new _univerjs_core.ObjectMatrix(setCellTypeObj).forValue((row, col) => {
				const cell = worksheet.getCellRaw(row, col);
				if (cell) undoSetCellTypeObj.setValue(row, col, { t: cell.t });
				else undoSetCellTypeObj.setValue(row, col, { t: void 0 });
			});
			redos.push({
				id: _univerjs_sheets.RemoveNumfmtMutation.id,
				params: removeRedos
			}, {
				id: _univerjs_sheets.SetRangeValuesMutation.id,
				params: {
					unitId,
					subUnitId,
					cellValue: setCellTypeObj
				}
			});
			const undo = (0, _univerjs_sheets.factoryRemoveNumfmtUndoMutation)(accessor, removeRedos);
			undos.push({
				id: _univerjs_sheets.SetRangeValuesMutation.id,
				params: {
					unitId,
					subUnitId,
					cellValue: undoSetCellTypeObj.getMatrix()
				}
			}, ...undo);
		}
		const result = (0, _univerjs_core.sequenceExecute)(redos, commandService).result;
		if (result) undoRedoService.pushUndoRedo({
			unitID: unitId,
			undoMutations: undos,
			redoMutations: redos
		});
		return result;
	}
};

//#endregion
//#region src/commands/commands/add-decimal.command.ts
const AddDecimalCommand = {
	id: "sheet.command.numfmt.add.decimal.command",
	type: _univerjs_core.CommandType.COMMAND,
	handler: async (accessor) => {
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const selectionManagerService = accessor.get(_univerjs_sheets.SheetsSelectionsService);
		const numfmtService = accessor.get(_univerjs_sheets.INumfmtService);
		const univerInstanceService = accessor.get(_univerjs_core.IUniverInstanceService);
		const selections = selectionManagerService.getCurrentSelections();
		if (!selections || !selections.length) return false;
		const target = (0, _univerjs_sheets.getSheetCommandTarget)(univerInstanceService);
		if (!target) return false;
		const { unitId, subUnitId } = target;
		let maxDecimals = 0;
		selections.forEach((selection) => {
			_univerjs_core.Range.foreach(selection.range, (row, col) => {
				const numfmtValue = numfmtService.getValue(unitId, subUnitId, row, col);
				if (!numfmtValue) {
					const cell = target.worksheet.getCellRaw(row, col);
					if (!maxDecimals && cell && cell.t === _univerjs_core.CellValueType.NUMBER && cell.v) {
						const regResult = /\.(\d*)$/.exec(String(cell.v));
						if (regResult) {
							const length = regResult[1].length;
							if (!length) return;
							maxDecimals = Math.max(maxDecimals, length);
						}
					}
					return;
				}
				const decimals = getDecimalFromPattern(numfmtValue.pattern);
				maxDecimals = decimals > maxDecimals ? decimals : maxDecimals;
			});
		});
		const decimals = maxDecimals + 1;
		const defaultPattern = setPatternDecimal(`0${decimals > 0 ? ".0" : ""}`, decimals);
		const values = [];
		selections.forEach((selection) => {
			_univerjs_core.Range.foreach(selection.range, (row, col) => {
				const numfmtValue = numfmtService.getValue(unitId, subUnitId, row, col);
				if ((0, _univerjs_core.isDefaultFormat)(numfmtValue === null || numfmtValue === void 0 ? void 0 : numfmtValue.pattern)) values.push({
					row,
					col,
					pattern: defaultPattern
				});
				else {
					const decimals = getDecimalFromPattern(numfmtValue.pattern);
					const pattern = setPatternDecimal(numfmtValue.pattern, decimals + 1);
					pattern !== numfmtValue.pattern && values.push({
						row,
						col,
						pattern
					});
				}
			});
		});
		if (values.length) return await commandService.executeCommand(SetNumfmtCommand.id, { values });
		return false;
	}
};

//#endregion
//#region src/commands/commands/set-currency.command.ts
/**
* This command is triggered by clicking the currency symbol icon in the menu.
* So the currency format is determined by the currency symbol icon.
*/
const SetCurrencyCommand = {
	id: "sheet.command.numfmt.set.currency",
	type: _univerjs_core.CommandType.COMMAND,
	handler: async (accessor) => {
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const selectionManagerService = accessor.get(_univerjs_sheets.SheetsSelectionsService);
		const localeService = accessor.get(_univerjs_core.LocaleService);
		const selections = selectionManagerService.getCurrentSelections();
		if (!selections || !selections.length) return false;
		const values = [];
		const currencyFormat = getCurrencyFormat(getCurrencySymbolIconByLocale(localeService.getCurrentLocale()).locale);
		selections.forEach((selection) => {
			_univerjs_core.Range.foreach(selection.range, (row, col) => {
				values.push({
					row,
					col,
					pattern: currencyFormat,
					type: "currency"
				});
			});
		});
		return await commandService.executeCommand(SetNumfmtCommand.id, { values });
	}
};

//#endregion
//#region src/commands/commands/set-percent.command.ts
const SetPercentCommand = {
	id: "sheet.command.numfmt.set.percent",
	type: _univerjs_core.CommandType.COMMAND,
	handler: async (accessor) => {
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const selections = accessor.get(_univerjs_sheets.SheetsSelectionsService).getCurrentSelections();
		if (!selections || !selections.length) return false;
		const values = [];
		const suffix = "0%";
		selections.forEach((selection) => {
			_univerjs_core.Range.foreach(selection.range, (row, col) => {
				values.push({
					row,
					col,
					pattern: suffix,
					type: "percent"
				});
			});
		});
		return await commandService.executeCommand(SetNumfmtCommand.id, { values });
	}
};

//#endregion
//#region src/commands/commands/subtract-decimal.command.ts
const SubtractDecimalCommand = {
	id: "sheet.command.numfmt.subtract.decimal.command",
	type: _univerjs_core.CommandType.COMMAND,
	handler: async (accessor) => {
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const selectionManagerService = accessor.get(_univerjs_sheets.SheetsSelectionsService);
		const numfmtService = accessor.get(_univerjs_sheets.INumfmtService);
		const univerInstanceService = accessor.get(_univerjs_core.IUniverInstanceService);
		const selections = selectionManagerService.getCurrentSelections();
		if (!selections || !selections.length) return false;
		const target = (0, _univerjs_sheets.getSheetCommandTarget)(univerInstanceService);
		if (!target) return false;
		const { unitId, subUnitId } = target;
		let maxDecimals = 0;
		selections.forEach((selection) => {
			_univerjs_core.Range.foreach(selection.range, (row, col) => {
				const numfmtValue = numfmtService.getValue(unitId, subUnitId, row, col);
				if (!numfmtValue) {
					const cell = target.worksheet.getCellRaw(row, col);
					if (!maxDecimals && cell && cell.t === _univerjs_core.CellValueType.NUMBER && cell.v) {
						const regResult = /\.(\d*)$/.exec(String(cell.v));
						if (regResult) {
							const length = regResult[1].length;
							if (!length) return;
							maxDecimals = Math.max(maxDecimals, length);
						}
					}
					return;
				}
				const decimals = getDecimalFromPattern(numfmtValue.pattern);
				maxDecimals = decimals > maxDecimals ? decimals : maxDecimals;
			});
		});
		const decimals = maxDecimals - 1;
		const defaultPattern = setPatternDecimal(`0${decimals > 0 ? ".0" : "."}`, decimals);
		const values = [];
		selections.forEach((selection) => {
			_univerjs_core.Range.foreach(selection.range, (row, col) => {
				const numfmtValue = numfmtService.getValue(unitId, subUnitId, row, col);
				if ((0, _univerjs_core.isDefaultFormat)(numfmtValue === null || numfmtValue === void 0 ? void 0 : numfmtValue.pattern)) values.push({
					row,
					col,
					pattern: defaultPattern
				});
				else {
					const decimals = getDecimalFromPattern(numfmtValue.pattern);
					values.push({
						row,
						col,
						pattern: setPatternDecimal(numfmtValue.pattern, decimals - 1)
					});
				}
			});
		});
		return await commandService.executeCommand(SetNumfmtCommand.id, { values });
	}
};

//#endregion
//#region src/config/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const SHEETS_NUMFMT_PLUGIN_CONFIG_KEY = "sheets-numfmt.config";
const configSymbol = Symbol(SHEETS_NUMFMT_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/utils/pattern.ts
const getPatternType = (pattern) => _univerjs_core.numfmt.getFormatInfo(pattern).type || "unknown";
const getPatternPreview = (pattern, value, locale = "en") => {
	try {
		const formatColor = _univerjs_core.numfmt.formatColor(pattern, value);
		const color = formatColor ? String(formatColor) : void 0;
		const result = _univerjs_core.numfmt.format(pattern, value, {
			locale,
			throws: false
		});
		if (value < 0) return {
			result,
			color
		};
		return { result };
	} catch (e) {
		console.warn("getPatternPreview error:", pattern, e);
	}
	return { result: String(value) };
};
const getPatternPreviewIgnoreGeneral = (pattern, value, locale) => {
	if (pattern === _univerjs_core.DEFAULT_NUMBER_FORMAT) return { result: String((0, _univerjs_engine_formula.stripErrorMargin)(value)) };
	return getPatternPreview(pattern, value, locale);
};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/controllers/numfmt-cell-content.controller.ts
const TEXT_FORMAT_MARK = { tl: {
	size: 6,
	color: "#409f11"
} };
let SheetsNumfmtCellContentController = class SheetsNumfmtCellContentController extends _univerjs_core.Disposable {
	constructor(_instanceService, _sheetInterceptorService, _themeService, _commandService, _numfmtService, _localeService, _configService) {
		super();
		this._instanceService = _instanceService;
		this._sheetInterceptorService = _sheetInterceptorService;
		this._themeService = _themeService;
		this._commandService = _commandService;
		this._numfmtService = _numfmtService;
		this._localeService = _localeService;
		this._configService = _configService;
		_defineProperty(this, "_locale$", new rxjs.BehaviorSubject("en"));
		_defineProperty(this, "locale$", this._locale$.asObservable());
		this._initInterceptorCellContent();
	}
	get locale() {
		const _locale = this._locale$.getValue();
		if (_locale) return _locale;
		switch (this._localeService.getCurrentLocale()) {
			case _univerjs_core.LocaleType.FR_FR: return "fr";
			case _univerjs_core.LocaleType.RU_RU: return "ru";
			case _univerjs_core.LocaleType.VI_VN: return "vi";
			case _univerjs_core.LocaleType.ZH_CN: return "zh-CN";
			case _univerjs_core.LocaleType.KO_KR: return "ko";
			case _univerjs_core.LocaleType.ZH_TW: return "zh-TW";
			case _univerjs_core.LocaleType.ZH_HK: return "zh-HK";
			case _univerjs_core.LocaleType.ES_ES:
			case _univerjs_core.LocaleType.CA_ES: return "es";
			case _univerjs_core.LocaleType.SK_SK: return "sk";
			case _univerjs_core.LocaleType.JA_JP: return "ja";
			case _univerjs_core.LocaleType.PT_BR: return "pt";
			case _univerjs_core.LocaleType.DE_DE: return "de";
			case _univerjs_core.LocaleType.IT_IT: return "it";
			case _univerjs_core.LocaleType.ID_ID: return "id";
			case _univerjs_core.LocaleType.PL_PL: return "pl";
			case _univerjs_core.LocaleType.AR_SA: return "ar";
			case _univerjs_core.LocaleType.EN_US:
			case _univerjs_core.LocaleType.FA_IR:
			default: return "en";
		}
	}
	_initInterceptorCellContent() {
		const renderCache = new _univerjs_core.ObjectMatrix();
		this.disposeWithMe((0, rxjs.merge)(this._locale$, this._localeService.currentLocale$).subscribe(() => {
			renderCache.reset();
		}));
		this.disposeWithMe(this._sheetInterceptorService.intercept(_univerjs_sheets.INTERCEPTOR_POINT.CELL_CONTENT, {
			effect: _univerjs_core.InterceptorEffectEnum.Value | _univerjs_core.InterceptorEffectEnum.Style,
			handler: (cell, location, next) => {
				if (!cell || cell.v === void 0 || cell.v === null || cell.t === _univerjs_core.CellValueType.BOOLEAN || cell.t === _univerjs_core.CellValueType.FORCE_STRING) return next(cell);
				const unitId = location.unitId;
				const sheetId = location.subUnitId;
				let numfmtValue;
				if (cell === null || cell === void 0 ? void 0 : cell.s) {
					const style = location.workbook.getStyles().get(cell.s);
					if (style === null || style === void 0 ? void 0 : style.n) numfmtValue = style.n;
				}
				if (!numfmtValue) numfmtValue = this._numfmtService.getValue(unitId, sheetId, location.row, location.col);
				if ((0, _univerjs_core.isDefaultFormat)(numfmtValue === null || numfmtValue === void 0 ? void 0 : numfmtValue.pattern)) return next(cell);
				if (cell.t !== _univerjs_core.CellValueType.NUMBER) {
					if ((0, _univerjs_sheets.checkCellValueType)(cell.v, cell.t) !== _univerjs_core.CellValueType.NUMBER) return next(cell);
				}
				const originCellValue = cell;
				if (!cell || cell === location.rawData) cell = { ...location.rawData };
				if ((0, _univerjs_core.isTextFormat)(numfmtValue === null || numfmtValue === void 0 ? void 0 : numfmtValue.pattern)) {
					var _this$_configService$;
					if ((_this$_configService$ = this._configService.getConfig("sheets-numfmt.config")) === null || _this$_configService$ === void 0 ? void 0 : _this$_configService$.disableTextFormatMark) {
						cell.t = _univerjs_core.CellValueType.STRING;
						return next(cell);
					}
					cell.t = _univerjs_core.CellValueType.STRING;
					cell.markers = {
						...cell === null || cell === void 0 ? void 0 : cell.markers,
						...TEXT_FORMAT_MARK
					};
					return next(cell);
				}
				let numfmtRes = "";
				const cache = renderCache.getValue(location.row, location.col);
				if (cache && cache.parameters === `${originCellValue.v}_${numfmtValue === null || numfmtValue === void 0 ? void 0 : numfmtValue.pattern}`) return next({
					...cell,
					...cache.result
				});
				const info = getPatternPreviewIgnoreGeneral(numfmtValue === null || numfmtValue === void 0 ? void 0 : numfmtValue.pattern, Number(originCellValue.v), this.locale);
				numfmtRes = info.result;
				if (!numfmtRes) return next(cell);
				const res = {
					v: numfmtRes,
					t: _univerjs_core.CellValueType.NUMBER
				};
				if (info.color) {
					var _this$_themeService$g;
					const color = (_this$_themeService$g = this._themeService.getColorFromTheme(`${info.color}.500`)) !== null && _this$_themeService$g !== void 0 ? _this$_themeService$g : info.color;
					if (color) res.interceptorStyle = { cl: { rgb: color } };
				}
				renderCache.setValue(location.row, location.col, {
					result: res,
					parameters: `${originCellValue.v}_${numfmtValue === null || numfmtValue === void 0 ? void 0 : numfmtValue.pattern}`
				});
				Object.assign(cell, res);
				return next(cell);
			},
			priority: _univerjs_sheets.InterceptCellContentPriority.NUMFMT
		}));
		this.disposeWithMe(this._commandService.onCommandExecuted((commandInfo) => {
			if (commandInfo.id === _univerjs_sheets.SetNumfmtMutation.id) {
				const params = commandInfo.params;
				Object.keys(params.values).forEach((key) => {
					params.values[key].ranges.forEach((range) => {
						_univerjs_core.Range.foreach(range, (row, col) => {
							renderCache.realDeleteValue(row, col);
						});
					});
				});
			} else if (commandInfo.id === _univerjs_sheets.SetRangeValuesMutation.id) {
				const params = commandInfo.params;
				new _univerjs_core.ObjectMatrix(params.cellValue).forValue((row, col) => {
					renderCache.realDeleteValue(row, col);
				});
			}
		}));
		this.disposeWithMe(this._instanceService.getCurrentTypeOfUnit$(_univerjs_core.UniverInstanceType.UNIVER_SHEET).pipe((0, rxjs.switchMap)((workbook) => {
			var _workbook$activeSheet;
			return (_workbook$activeSheet = workbook === null || workbook === void 0 ? void 0 : workbook.activeSheet$) !== null && _workbook$activeSheet !== void 0 ? _workbook$activeSheet : (0, rxjs.of)(null);
		}), (0, rxjs.skip)(1)).subscribe(() => renderCache.reset()));
	}
	setNumfmtLocal(locale) {
		this._locale$.next(locale);
	}
};
SheetsNumfmtCellContentController = __decorate([
	__decorateParam(0, _univerjs_core.IUniverInstanceService),
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_sheets.SheetInterceptorService)),
	__decorateParam(2, (0, _univerjs_core.Inject)(_univerjs_core.ThemeService)),
	__decorateParam(3, (0, _univerjs_core.Inject)(_univerjs_core.ICommandService)),
	__decorateParam(4, (0, _univerjs_core.Inject)(_univerjs_sheets.INumfmtService)),
	__decorateParam(5, (0, _univerjs_core.Inject)(_univerjs_core.LocaleService)),
	__decorateParam(6, _univerjs_core.IConfigService)
], SheetsNumfmtCellContentController);

//#endregion
//#region package.json
var name = "@univerjs/sheets-numfmt";
var version = "0.25.1";

//#endregion
//#region src/base/const/plugin-name.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const SHEET_NUMFMT_PLUGIN = "SHEET_NUMFMT_PLUGIN";

//#endregion
//#region src/plugin.ts
let UniverSheetsNumfmtPlugin = class UniverSheetsNumfmtPlugin extends _univerjs_core.Plugin {
	constructor(_config = defaultPluginConfig, _injector, _configService, _commandService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		this._commandService = _commandService;
		const { ...rest } = (0, _univerjs_core.merge)({}, defaultPluginConfig, this._config);
		this._configService.setConfig(SHEETS_NUMFMT_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		(0, _univerjs_core.registerDependencies)(this._injector, [[SheetsNumfmtCellContentController]]);
		(0, _univerjs_core.touchDependencies)(this._injector, [[SheetsNumfmtCellContentController]]);
		[
			AddDecimalCommand,
			SubtractDecimalCommand,
			SetCurrencyCommand,
			SetPercentCommand,
			SetNumfmtCommand
		].forEach((config) => {
			this.disposeWithMe(this._commandService.registerCommand(config));
		});
	}
};
_defineProperty(UniverSheetsNumfmtPlugin, "pluginName", SHEET_NUMFMT_PLUGIN);
_defineProperty(UniverSheetsNumfmtPlugin, "packageName", name);
_defineProperty(UniverSheetsNumfmtPlugin, "version", version);
_defineProperty(UniverSheetsNumfmtPlugin, "type", _univerjs_core.UniverInstanceType.UNIVER_SHEET);
UniverSheetsNumfmtPlugin = __decorate([
	(0, _univerjs_core.DependentOn)(_univerjs_sheets.UniverSheetsPlugin),
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector)),
	__decorateParam(2, _univerjs_core.IConfigService),
	__decorateParam(3, _univerjs_core.ICommandService)
], UniverSheetsNumfmtPlugin);

//#endregion
//#region src/utils/currency.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const getCurrencyType = (pattern) => {
	return _univerjs_core.currencySymbols.find((code) => pattern.includes(code));
};

//#endregion
//#region src/utils/options.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const getCurrencyOptions = () => _univerjs_core.currencySymbols.map((item) => ({
	label: item,
	value: item
}));
const getCurrencyFormatOptions = (suffix) => CURRENCYFORMAT.map((item) => ({
	label: item.label(suffix),
	value: item.suffix(suffix),
	color: item.color
}));
const getDateFormatOptions = () => DATEFMTLISG.map((item) => ({
	label: item.label,
	value: item.suffix
}));
const getNumberFormatOptions = () => NUMBERFORMAT.map((item) => ({
	label: item.label,
	value: item.suffix,
	color: item.color
}));

//#endregion
exports.AddDecimalCommand = AddDecimalCommand;
exports.CURRENCYFORMAT = CURRENCYFORMAT;
exports.DATEFMTLISG = DATEFMTLISG;
exports.NUMBERFORMAT = NUMBERFORMAT;
exports.SHEETS_NUMFMT_PLUGIN_CONFIG_KEY = SHEETS_NUMFMT_PLUGIN_CONFIG_KEY;
exports.SetCurrencyCommand = SetCurrencyCommand;
exports.SetNumfmtCommand = SetNumfmtCommand;
exports.SetPercentCommand = SetPercentCommand;
Object.defineProperty(exports, 'SheetsNumfmtCellContentController', {
  enumerable: true,
  get: function () {
    return SheetsNumfmtCellContentController;
  }
});
exports.SubtractDecimalCommand = SubtractDecimalCommand;
Object.defineProperty(exports, 'UniverSheetsNumfmtPlugin', {
  enumerable: true,
  get: function () {
    return UniverSheetsNumfmtPlugin;
  }
});
exports.getCurrencyFormat = getCurrencyFormat;
exports.getCurrencyFormatOptions = getCurrencyFormatOptions;
exports.getCurrencyOptions = getCurrencyOptions;
exports.getCurrencySymbolByLocale = getCurrencySymbolByLocale;
exports.getCurrencySymbolIconByLocale = getCurrencySymbolIconByLocale;
exports.getCurrencyType = getCurrencyType;
exports.getDateFormatOptions = getDateFormatOptions;
exports.getDecimalFromPattern = getDecimalFromPattern;
exports.getDecimalString = getDecimalString;
exports.getNumberFormatOptions = getNumberFormatOptions;
exports.getPatternPreview = getPatternPreview;
exports.getPatternPreviewIgnoreGeneral = getPatternPreviewIgnoreGeneral;
exports.getPatternType = getPatternType;
exports.isPatternHasDecimal = isPatternHasDecimal;
exports.localeCurrencySymbolMap = localeCurrencySymbolMap;
exports.setPatternDecimal = setPatternDecimal;
