// @univerjs/sheets-table-ui/lib/cjs/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
let _univerjs_core = require("@univerjs/core");
let _univerjs_engine_render = require("@univerjs/engine-render");
let _univerjs_sheets_table = require("@univerjs/sheets-table");
let _univerjs_sheets_ui = require("@univerjs/sheets-ui");
let _univerjs_ui = require("@univerjs/ui");
let rxjs = require("rxjs");
let _univerjs_design = require("@univerjs/design");
let _univerjs_icons = require("@univerjs/icons");
let _univerjs_sheets = require("@univerjs/sheets");
let _univerjs_sheets_sort = require("@univerjs/sheets-sort");
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");
let _univerjs_engine_formula = require("@univerjs/engine-formula");
let _univerjs_sheets_formula_ui = require("@univerjs/sheets-formula-ui");

//#region package.json
var name = "@univerjs/sheets-table-ui";
var version = "0.25.1";

//#endregion
//#region src/const.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const PLUGIN_NAME = "SHEET_TABLE_UI_PLUGIN";
const SHEETS_TABLE_FILTER_PANEL_OPENED_KEY = "SHEETS_TABLE_FILTER_PANEL_OPENED_KEY";
const UNIVER_SHEET_TABLE_FILTER_PANEL_ID = "UNIVER_SHEET_Table_FILTER_PANEL_ID";
const TABLE_TOOLBAR_BUTTON = "TABLE_TOOLBAR_BUTTON";
const TABLE_SELECTOR_DIALOG = "TABLE_SELECTOR_DIALOG";
const SHEET_TABLE_RENAME_DIALOG = "SHEET_TABLE_RENAME_DIALOG";
const SHEET_TABLE_RENAME_DIALOG_ID = "SHEET_TABLE_RENAME_DIALOG_ID";
const SHEET_TABLE_THEME_PANEL_ID = "SHEET_TABLE_THEME_PANEL_ID";
const SHEET_TABLE_THEME_PANEL = "SHEET_TABLE_THEME_PANEL";
const TABLE_CUSTOM_NAME_PREFIX = "table-custom-";
const TABLE_DEFAULT_NAME_PREFIX = "table-default-";
const TABLE_DEFAULT_BG_COLOR = "rgb(255, 255, 255)";
const TABLE_BORDER_NONE = "none";
const TABLE_BORDER_DEFAULT = "1px solid rgb(var(--grey-200))";

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/services/sheets-table-ui.service.ts
let SheetsTableUiService = class SheetsTableUiService extends _univerjs_core.Disposable {
	constructor(_tableManager, _sheetTableService, _univerInstanceService, _commandService, _localeService) {
		super();
		this._tableManager = _tableManager;
		this._sheetTableService = _sheetTableService;
		this._univerInstanceService = _univerInstanceService;
		this._commandService = _commandService;
		this._localeService = _localeService;
		_defineProperty(this, "_itemsCache", /* @__PURE__ */ new Map());
		this._registerTableFilterChangeEvent();
	}
	_registerTableFilterChangeEvent() {
		this._commandService.onCommandExecuted((command) => {
			if (command.id === _univerjs_sheets.SetRangeValuesMutation.id) {
				const { unitId, subUnitId, cellValue } = command.params;
				const tables = this._tableManager.getTablesBySubunitId(unitId, subUnitId);
				if (!tables.length) return;
				new _univerjs_core.ObjectMatrix(cellValue).forValue((row, col, _value) => {
					const cellRange = (0, _univerjs_core.cellToRange)(row, col);
					const overlapTable = tables.find((table) => {
						const tableRange = table.getTableFilterRange();
						return _univerjs_core.Rectangle.intersects(tableRange, cellRange);
					});
					if (overlapTable) {
						const colIndex = col - overlapTable.getRange().startColumn;
						this._itemsCache.delete(overlapTable.getId() + colIndex);
					}
				});
			} else if (command.id === _univerjs_sheets_table.SetSheetTableFilterCommand.id) {
				const { unitId, tableId } = command.params;
				const table = this._tableManager.getTable(unitId, tableId);
				if (!table) return;
				const subUnitId = table.getSubunitId();
				this._tableManager.getTablesBySubunitId(unitId, subUnitId).forEach((table) => {
					const range = table.getRange();
					for (let i = range.startColumn; i <= range.endColumn; i++) this._itemsCache.delete(table.getId() + i);
				});
			}
		});
	}
	getTableFilterPanelInitProps(unitId, subUnitId, tableId, column) {
		const table = this._tableManager.getTable(unitId, tableId);
		const tableRange = table.getRange();
		const tableFilter = table.getTableFilterColumn(column - tableRange.startColumn);
		return {
			unitId,
			subUnitId,
			tableFilter,
			currentFilterBy: (0, _univerjs_sheets_table.isConditionFilter)(tableFilter) ? "condition" : "items",
			tableId,
			columnIndex: column - tableRange.startColumn
		};
	}
	getTableFilterCheckedItems(unitId, tableId, columnIndex) {
		const table = this._tableManager.getTable(unitId, tableId);
		const checkedItems = [];
		if (table) {
			const filter = table.getTableFilterColumn(columnIndex);
			if (filter && (0, _univerjs_sheets_table.isManualTableFilter)(filter)) checkedItems.push(...filter.values.map((value) => value === _univerjs_sheets_table.TABLE_FILTER_EMPTY_VALUE ? this._localeService.t("sheets-table-ui.condition.empty") : value));
		}
		return checkedItems;
	}
	setTableFilter(unitId, tableId, columnIndex, tableFilter) {
		if (!this._tableManager.getTable(unitId, tableId)) return;
		const setTableFilterParams = {
			unitId,
			tableId,
			column: columnIndex,
			tableFilter
		};
		this._commandService.executeCommand(_univerjs_sheets_table.SetSheetTableFilterCommand.id, setTableFilterParams);
	}
	getTableFilterItems(unitId, subUnitId, tableId, columnIndex) {
		var _this$_univerInstance;
		if (this._itemsCache.has(tableId + columnIndex)) return this._itemsCache.get(tableId + columnIndex) || {
			data: [],
			itemsCountMap: /* @__PURE__ */ new Map(),
			allItemsCount: 0
		};
		const table = this._tableManager.getTable(unitId, tableId);
		if (!table) return {
			data: [],
			itemsCountMap: /* @__PURE__ */ new Map(),
			allItemsCount: 0
		};
		const tableRange = table.getTableFilterRange();
		const { startRow, endRow, startColumn } = tableRange;
		const column = startColumn + columnIndex;
		const worksheet = (_this$_univerInstance = this._univerInstanceService.getUnit(unitId)) === null || _this$_univerInstance === void 0 ? void 0 : _this$_univerInstance.getSheetBySheetId(subUnitId);
		if (!worksheet) return {
			data: [],
			itemsCountMap: /* @__PURE__ */ new Map(),
			allItemsCount: 0
		};
		const data = [];
		const map = /* @__PURE__ */ new Map();
		const filteredRowsByOtherColumns = /* @__PURE__ */ new Set();
		const tableFilters = table.getTableFilters();
		for (let i = tableRange.startColumn; i <= tableRange.endColumn; i++) {
			const currentColumnIndex = i - tableRange.startColumn;
			if (currentColumnIndex !== columnIndex && table.getTableFilterColumn(currentColumnIndex)) tableFilters.doColumnFilter(worksheet, tableRange, currentColumnIndex, filteredRowsByOtherColumns);
		}
		let allItemsCount = 0;
		for (let row = startRow; row <= endRow; row++) {
			if (filteredRowsByOtherColumns.has(row)) continue;
			let stringItem = this._sheetTableService.getCellValueWithConditionType(worksheet, row, column);
			if (stringItem == null) stringItem = this._localeService.t("sheets-table-ui.condition.empty");
			if (!map.has(stringItem)) data.push({
				title: stringItem,
				key: `${column}_${row}`,
				leaf: true
			});
			allItemsCount++;
			map.set(stringItem, (map.get(stringItem) || 0) + 1);
		}
		this._itemsCache.set(tableId + columnIndex, {
			data,
			itemsCountMap: map,
			allItemsCount
		});
		return {
			data,
			itemsCountMap: map,
			allItemsCount
		};
	}
};
SheetsTableUiService = __decorate([
	__decorateParam(0, (0, _univerjs_core.Inject)(_univerjs_sheets_table.TableManager)),
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_sheets_table.SheetTableService)),
	__decorateParam(2, (0, _univerjs_core.Inject)(_univerjs_core.IUniverInstanceService)),
	__decorateParam(3, _univerjs_core.ICommandService),
	__decorateParam(4, (0, _univerjs_core.Inject)(_univerjs_core.LocaleService))
], SheetsTableUiService);

//#endregion
//#region src/views/components/util.ts
function getCascaderListOptions(injector) {
	const t = injector.get(_univerjs_core.LocaleService).t;
	return [
		{
			value: _univerjs_sheets_table.TableConditionTypeEnum.String,
			label: t(`sheets-table-ui.condition.${_univerjs_sheets_table.TableConditionTypeEnum.String}`),
			children: [
				{
					value: _univerjs_sheets_table.TableStringCompareTypeEnum.Equal,
					label: t(`sheets-table-ui.string.compare.${_univerjs_sheets_table.TableStringCompareTypeEnum.Equal}`)
				},
				{
					value: _univerjs_sheets_table.TableStringCompareTypeEnum.NotEqual,
					label: t(`sheets-table-ui.string.compare.${_univerjs_sheets_table.TableStringCompareTypeEnum.NotEqual}`)
				},
				{
					value: _univerjs_sheets_table.TableStringCompareTypeEnum.Contains,
					label: t(`sheets-table-ui.string.compare.${_univerjs_sheets_table.TableStringCompareTypeEnum.Contains}`)
				},
				{
					value: _univerjs_sheets_table.TableStringCompareTypeEnum.NotContains,
					label: t(`sheets-table-ui.string.compare.${_univerjs_sheets_table.TableStringCompareTypeEnum.NotContains}`)
				},
				{
					value: _univerjs_sheets_table.TableStringCompareTypeEnum.StartsWith,
					label: t(`sheets-table-ui.string.compare.${_univerjs_sheets_table.TableStringCompareTypeEnum.StartsWith}`)
				},
				{
					value: _univerjs_sheets_table.TableStringCompareTypeEnum.EndsWith,
					label: t(`sheets-table-ui.string.compare.${_univerjs_sheets_table.TableStringCompareTypeEnum.EndsWith}`)
				}
			]
		},
		{
			value: _univerjs_sheets_table.TableConditionTypeEnum.Number,
			label: t(`sheets-table-ui.condition.${_univerjs_sheets_table.TableConditionTypeEnum.Number}`),
			children: [
				{
					value: _univerjs_sheets_table.TableNumberCompareTypeEnum.Equal,
					label: t(`sheets-table-ui.number.compare.${_univerjs_sheets_table.TableNumberCompareTypeEnum.Equal}`)
				},
				{
					value: _univerjs_sheets_table.TableNumberCompareTypeEnum.NotEqual,
					label: t(`sheets-table-ui.number.compare.${_univerjs_sheets_table.TableNumberCompareTypeEnum.NotEqual}`)
				},
				{
					value: _univerjs_sheets_table.TableNumberCompareTypeEnum.GreaterThan,
					label: t(`sheets-table-ui.number.compare.${_univerjs_sheets_table.TableNumberCompareTypeEnum.GreaterThan}`)
				},
				{
					value: _univerjs_sheets_table.TableNumberCompareTypeEnum.GreaterThanOrEqual,
					label: t(`sheets-table-ui.number.compare.${_univerjs_sheets_table.TableNumberCompareTypeEnum.GreaterThanOrEqual}`)
				},
				{
					value: _univerjs_sheets_table.TableNumberCompareTypeEnum.LessThan,
					label: t(`sheets-table-ui.number.compare.${_univerjs_sheets_table.TableNumberCompareTypeEnum.LessThan}`)
				},
				{
					value: _univerjs_sheets_table.TableNumberCompareTypeEnum.LessThanOrEqual,
					label: t(`sheets-table-ui.number.compare.${_univerjs_sheets_table.TableNumberCompareTypeEnum.LessThanOrEqual}`)
				},
				{
					value: _univerjs_sheets_table.TableNumberCompareTypeEnum.Between,
					label: t(`sheets-table-ui.number.compare.${_univerjs_sheets_table.TableNumberCompareTypeEnum.Between}`)
				},
				{
					value: _univerjs_sheets_table.TableNumberCompareTypeEnum.NotBetween,
					label: t(`sheets-table-ui.number.compare.${_univerjs_sheets_table.TableNumberCompareTypeEnum.NotBetween}`)
				},
				{
					value: _univerjs_sheets_table.TableNumberCompareTypeEnum.Above,
					label: t(`sheets-table-ui.number.compare.${_univerjs_sheets_table.TableNumberCompareTypeEnum.Above}`)
				},
				{
					value: _univerjs_sheets_table.TableNumberCompareTypeEnum.Below,
					label: t(`sheets-table-ui.number.compare.${_univerjs_sheets_table.TableNumberCompareTypeEnum.Below}`)
				}
			]
		},
		{
			value: _univerjs_sheets_table.TableConditionTypeEnum.Date,
			label: t(`sheets-table-ui.condition.${_univerjs_sheets_table.TableConditionTypeEnum.Date}`),
			children: [
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.Equal,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.Equal}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.NotEqual,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.NotEqual}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.After,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.After}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.AfterOrEqual,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.AfterOrEqual}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.Before,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.Before}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.BeforeOrEqual,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.BeforeOrEqual}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.Between,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.Between}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.NotBetween,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.NotBetween}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.Today,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.Today}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.Yesterday,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.Yesterday}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.Tomorrow,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.Tomorrow}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.ThisWeek,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.ThisWeek}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.LastWeek,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.LastWeek}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.NextWeek,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.NextWeek}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.ThisMonth,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.ThisMonth}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.LastMonth,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.LastMonth}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.NextMonth,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.NextMonth}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.ThisYear,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.ThisYear}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.LastYear,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.LastYear}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.NextYear,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.NextYear}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.Quarter,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.Quarter}`)
				},
				{
					value: _univerjs_sheets_table.TableDateCompareTypeEnum.Month,
					label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.Month}`)
				}
			]
		}
	];
}
function getConditionDateSelect(injector, dateType) {
	if (!dateType) return [];
	const t = injector.get(_univerjs_core.LocaleService).t;
	switch (dateType) {
		case _univerjs_sheets_table.TableDateCompareTypeEnum.Quarter: return [
			{
				value: _univerjs_sheets_table.TableDateCompareTypeEnum.Q1,
				label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.Q1}`)
			},
			{
				value: _univerjs_sheets_table.TableDateCompareTypeEnum.Q2,
				label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.Q2}`)
			},
			{
				value: _univerjs_sheets_table.TableDateCompareTypeEnum.Q3,
				label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.Q3}`)
			},
			{
				value: _univerjs_sheets_table.TableDateCompareTypeEnum.Q4,
				label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.Q4}`)
			}
		];
		case _univerjs_sheets_table.TableDateCompareTypeEnum.Month: return [
			{
				value: _univerjs_sheets_table.TableDateCompareTypeEnum.M1,
				label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.M1}`)
			},
			{
				value: _univerjs_sheets_table.TableDateCompareTypeEnum.M2,
				label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.M2}`)
			},
			{
				value: _univerjs_sheets_table.TableDateCompareTypeEnum.M3,
				label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.M3}`)
			},
			{
				value: _univerjs_sheets_table.TableDateCompareTypeEnum.M4,
				label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.M4}`)
			},
			{
				value: _univerjs_sheets_table.TableDateCompareTypeEnum.M5,
				label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.M5}`)
			},
			{
				value: _univerjs_sheets_table.TableDateCompareTypeEnum.M6,
				label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.M6}`)
			},
			{
				value: _univerjs_sheets_table.TableDateCompareTypeEnum.M7,
				label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.M7}`)
			},
			{
				value: _univerjs_sheets_table.TableDateCompareTypeEnum.M8,
				label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.M8}`)
			},
			{
				value: _univerjs_sheets_table.TableDateCompareTypeEnum.M9,
				label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.M9}`)
			},
			{
				value: _univerjs_sheets_table.TableDateCompareTypeEnum.M10,
				label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.M10}`)
			},
			{
				value: _univerjs_sheets_table.TableDateCompareTypeEnum.M11,
				label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.M11}`)
			},
			{
				value: _univerjs_sheets_table.TableDateCompareTypeEnum.M12,
				label: t(`sheets-table-ui.date.compare.${_univerjs_sheets_table.TableDateCompareTypeEnum.M12}`)
			}
		];
		default: return [];
	}
}
const datePickerSet = new Set([
	_univerjs_sheets_table.TableDateCompareTypeEnum.Equal,
	_univerjs_sheets_table.TableDateCompareTypeEnum.NotEqual,
	_univerjs_sheets_table.TableDateCompareTypeEnum.After,
	_univerjs_sheets_table.TableDateCompareTypeEnum.AfterOrEqual,
	_univerjs_sheets_table.TableDateCompareTypeEnum.Before,
	_univerjs_sheets_table.TableDateCompareTypeEnum.BeforeOrEqual
]);
function getSubComponentType(type, compare) {
	if (!compare) return "None";
	if (type === _univerjs_sheets_table.TableConditionTypeEnum.String) return "Input";
	else if (type === _univerjs_sheets_table.TableConditionTypeEnum.Number) if (compare === _univerjs_sheets_table.TableNumberCompareTypeEnum.Between || compare === _univerjs_sheets_table.TableNumberCompareTypeEnum.NotBetween) return "Inputs";
	else return "Input";
	else if (type === _univerjs_sheets_table.TableConditionTypeEnum.Date) {
		if (compare === _univerjs_sheets_table.TableDateCompareTypeEnum.Between || compare === _univerjs_sheets_table.TableDateCompareTypeEnum.NotBetween) return "DateRange";
		else if (compare === _univerjs_sheets_table.TableDateCompareTypeEnum.Quarter || compare === _univerjs_sheets_table.TableDateCompareTypeEnum.Month) return "Select";
		else if (datePickerSet.has(compare)) return "DatePicker";
		return "None";
	}
	return "None";
}
function getInitConditionInfo(tableFilter) {
	if (!tableFilter || tableFilter.filterType !== "condition") return {
		type: _univerjs_sheets_table.TableConditionTypeEnum.String,
		compareType: _univerjs_sheets_table.TableStringCompareTypeEnum.Equal,
		info: {}
	};
	const filterInfo = tableFilter.filterInfo;
	const { conditionType, compareType } = filterInfo;
	if (conditionType === _univerjs_sheets_table.TableConditionTypeEnum.Date) if (compareType === _univerjs_sheets_table.TableDateCompareTypeEnum.Between || compareType === _univerjs_sheets_table.TableDateCompareTypeEnum.NotBetween) {
		let dateRange;
		if (Array.isArray(filterInfo.expectedValue)) dateRange = filterInfo.expectedValue.map((i) => typeof i === "string" ? new Date(i) : i);
		return {
			type: conditionType,
			compare: compareType,
			info: { dateRange }
		};
	} else if (compareType === _univerjs_sheets_table.TableDateCompareTypeEnum.Today || compareType === _univerjs_sheets_table.TableDateCompareTypeEnum.Yesterday || compareType === _univerjs_sheets_table.TableDateCompareTypeEnum.Tomorrow || compareType === _univerjs_sheets_table.TableDateCompareTypeEnum.ThisWeek || compareType === _univerjs_sheets_table.TableDateCompareTypeEnum.LastWeek || compareType === _univerjs_sheets_table.TableDateCompareTypeEnum.NextWeek || compareType === _univerjs_sheets_table.TableDateCompareTypeEnum.ThisMonth || compareType === _univerjs_sheets_table.TableDateCompareTypeEnum.LastMonth || compareType === _univerjs_sheets_table.TableDateCompareTypeEnum.NextMonth || compareType === _univerjs_sheets_table.TableDateCompareTypeEnum.ThisYear || compareType === _univerjs_sheets_table.TableDateCompareTypeEnum.LastYear || compareType === _univerjs_sheets_table.TableDateCompareTypeEnum.NextYear) return {
		type: conditionType,
		compare: compareType,
		info: {}
	};
	else if (datePickerSet.has(compareType)) {
		let date;
		if (typeof filterInfo.expectedValue === "string") date = new Date(filterInfo.expectedValue);
		else if (Array.isArray(filterInfo.expectedValue)) {
			for (let i = 0; i < filterInfo.expectedValue.length; i++) if (typeof filterInfo.expectedValue[i] === "string") filterInfo.expectedValue[i] = new Date(filterInfo.expectedValue[i]);
		}
		return {
			type: conditionType,
			compare: compareType,
			info: { date }
		};
	} else if (new Set([
		_univerjs_sheets_table.TableDateCompareTypeEnum.Q1,
		_univerjs_sheets_table.TableDateCompareTypeEnum.Q2,
		_univerjs_sheets_table.TableDateCompareTypeEnum.Q3,
		_univerjs_sheets_table.TableDateCompareTypeEnum.Q4
	]).has(compareType)) return {
		type: conditionType,
		compare: _univerjs_sheets_table.TableDateCompareTypeEnum.Quarter,
		info: { dateSelect: filterInfo.compareType }
	};
	else return {
		type: conditionType,
		compare: _univerjs_sheets_table.TableDateCompareTypeEnum.Month,
		info: { dateSelect: filterInfo.compareType }
	};
	else if (conditionType === _univerjs_sheets_table.TableConditionTypeEnum.Number) if (compareType === _univerjs_sheets_table.TableNumberCompareTypeEnum.Between || compareType === _univerjs_sheets_table.TableNumberCompareTypeEnum.NotBetween) return {
		type: conditionType,
		compare: compareType,
		info: { numberRange: filterInfo.expectedValue }
	};
	else return {
		type: conditionType,
		compare: compareType,
		info: { number: filterInfo.expectedValue }
	};
	else if (conditionType === _univerjs_sheets_table.TableConditionTypeEnum.String) return {
		type: conditionType,
		compare: compareType,
		info: { string: filterInfo.expectedValue }
	};
	return {
		type: _univerjs_sheets_table.TableConditionTypeEnum.String,
		compare: _univerjs_sheets_table.TableStringCompareTypeEnum.Equal,
		info: {}
	};
}

//#endregion
//#region src/views/components/SheetTableConditionPanel.tsx
const SheetTableConditionPanel = (props) => {
	var _conditionInfo$info$d, _conditionInfo$info$d2, _conditionInfo$info$d3, _conditionInfo$info$d4, _conditionInfo$info$d5, _conditionInfo$info$n, _conditionInfo$info$n3, _conditionInfo$info$d6;
	const { conditionInfo, onChange } = props;
	const localeService = (0, _univerjs_ui.useDependency)(_univerjs_core.LocaleService);
	const [conditionVisible, setConditionVisible] = (0, react.useState)(false);
	const injector = (0, _univerjs_ui.useDependency)(_univerjs_core.Injector);
	const cascaderOptions = getCascaderListOptions(injector);
	const handleConditionInfo = (info, type, compare) => {
		onChange({
			type: type !== null && type !== void 0 ? type : conditionInfo.type,
			compare: compare !== null && compare !== void 0 ? compare : conditionInfo.compare,
			info
		});
	};
	const handleChange = (value) => {
		var _ref;
		const type = value[0];
		const compare = value[1];
		if (compare) setConditionVisible(false);
		const info = {};
		if (type === _univerjs_sheets_table.TableConditionTypeEnum.Date) if (compare === _univerjs_sheets_table.TableDateCompareTypeEnum.Quarter) info.dateSelect = _univerjs_sheets_table.TableDateCompareTypeEnum.Q1;
		else if (compare === _univerjs_sheets_table.TableDateCompareTypeEnum.Month) info.dateSelect = _univerjs_sheets_table.TableDateCompareTypeEnum.M1;
		else if (datePickerSet.has(compare)) info.date = /* @__PURE__ */ new Date();
		else info.dateRange = [/* @__PURE__ */ new Date(), /* @__PURE__ */ new Date()];
		else if (type === _univerjs_sheets_table.TableConditionTypeEnum.Number) info.number = 0;
		else if (type === _univerjs_sheets_table.TableConditionTypeEnum.String) info.string = "";
		handleConditionInfo(info, value[0], (_ref = value[1]) !== null && _ref !== void 0 ? _ref : _univerjs_sheets_table.TableStringCompareTypeEnum.Equal);
	};
	const subComponentType = getSubComponentType(conditionInfo.type, conditionInfo.compare);
	let selectType = "";
	if (conditionInfo.compare) selectType = `${localeService.t(`sheets-table-ui.condition.${conditionInfo.type}`)} - ${localeService.t(`sheets-table-ui.${conditionInfo.type}.compare.${conditionInfo.compare}`)}`;
	else selectType = localeService.t(`sheets-table-ui.condition.${conditionInfo.type}`);
	const conditionDateOptions = getConditionDateSelect(injector, conditionInfo.compare);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Dropdown, {
		align: "start",
		open: conditionVisible,
		onOpenChange: setConditionVisible,
		overlay: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.CascaderList, {
			value: [conditionInfo.type, conditionInfo.compare],
			options: cascaderOptions,
			onChange: handleChange,
			contentClassName: "univer-flex-1",
			wrapperClassName: "!univer-h-[150px]"
		}),
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: (0, _univerjs_design.clsx)("univer-box-border univer-flex univer-h-8 univer-w-full univer-items-center univer-justify-between univer-rounded-md univer-bg-white univer-px-2 univer-text-sm univer-transition-colors univer-duration-200 hover:univer-border-primary-600 focus:univer-border-primary-600 focus:univer-outline-none focus:univer-ring-2 dark:!univer-bg-gray-700 dark:!univer-text-white", _univerjs_design.borderClassName),
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { children: selectType }), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_icons.MoreDownIcon, {})]
		})
	}), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: "univer-mt-3 univer-w-full",
		children: [
			subComponentType === "Input" && /* @__PURE__ */ (0, react_jsx_runtime.jsx)(react_jsx_runtime.Fragment, { children: conditionInfo.type === _univerjs_sheets_table.TableConditionTypeEnum.String ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Input, {
				className: "univer-w-full",
				placeholder: "请输入",
				value: conditionInfo.info.string,
				onChange: (v) => handleConditionInfo({ string: v })
			}) : /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.InputNumber, {
				className: "univer-h-7 univer-w-full",
				value: conditionInfo.info.number,
				controls: false,
				onChange: (v) => {
					if (v !== null) handleConditionInfo({ number: v });
				}
			}) }),
			!!(subComponentType === "DatePicker") && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				id: "univer-table-date-picker-wrapper",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.DatePicker, {
					className: "univer-w-full",
					value: (_conditionInfo$info$d = conditionInfo.info.date) !== null && _conditionInfo$info$d !== void 0 ? _conditionInfo$info$d : /* @__PURE__ */ new Date(),
					onValueChange: (v) => handleConditionInfo({ date: v })
				})
			}),
			!!(subComponentType === "DateRange") && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				id: "univer-table-date-range-wrapper",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.DateRangePicker, {
					className: "univer-w-full",
					value: [(_conditionInfo$info$d2 = (_conditionInfo$info$d3 = conditionInfo.info.dateRange) === null || _conditionInfo$info$d3 === void 0 ? void 0 : _conditionInfo$info$d3[0]) !== null && _conditionInfo$info$d2 !== void 0 ? _conditionInfo$info$d2 : /* @__PURE__ */ new Date(), (_conditionInfo$info$d4 = (_conditionInfo$info$d5 = conditionInfo.info.dateRange) === null || _conditionInfo$info$d5 === void 0 ? void 0 : _conditionInfo$info$d5[1]) !== null && _conditionInfo$info$d4 !== void 0 ? _conditionInfo$info$d4 : /* @__PURE__ */ new Date()],
					onValueChange: (v) => {
						if (v) handleConditionInfo({ dateRange: v });
						else handleConditionInfo({});
					}
				})
			}),
			subComponentType === "Inputs" && /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: "univer-flex univer-items-center univer-gap-2",
				children: [
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.InputNumber, {
						className: "univer-w-full",
						value: (_conditionInfo$info$n = conditionInfo.info.numberRange) === null || _conditionInfo$info$n === void 0 ? void 0 : _conditionInfo$info$n[0],
						onChange: (v) => {
							if (v !== null) {
								var _conditionInfo$info$n2;
								handleConditionInfo({ numberRange: [v, (_conditionInfo$info$n2 = conditionInfo.info.numberRange) === null || _conditionInfo$info$n2 === void 0 ? void 0 : _conditionInfo$info$n2[1]] });
							}
						},
						controls: false
					}),
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { children: " - " }),
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.InputNumber, {
						className: "univer-w-full",
						value: (_conditionInfo$info$n3 = conditionInfo.info.numberRange) === null || _conditionInfo$info$n3 === void 0 ? void 0 : _conditionInfo$info$n3[1],
						controls: false,
						onChange: (v) => {
							if (v !== null) {
								var _conditionInfo$info$n4;
								handleConditionInfo({ numberRange: [(_conditionInfo$info$n4 = conditionInfo.info.numberRange) === null || _conditionInfo$info$n4 === void 0 ? void 0 : _conditionInfo$info$n4[0], v] });
							}
						}
					})
				]
			}),
			!!(subComponentType === "Select") && /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Select, {
				className: "univer-w-full",
				value: (_conditionInfo$info$d6 = conditionInfo.info.dateSelect) !== null && _conditionInfo$info$d6 !== void 0 ? _conditionInfo$info$d6 : conditionDateOptions[0].value,
				options: conditionDateOptions,
				onChange: (v) => handleConditionInfo({ dateSelect: v })
			})
		]
	})] });
};

//#endregion
//#region src/views/components/SheetTableItemsFilterPanel.tsx
const getCheckedCount = (map) => {
	let count = 0;
	map.forEach((value) => {
		count += value;
	});
	return count;
};
function SheetTableItemsFilterPanel(props) {
	const { unitId, tableId, subUnitId, columnIndex, checkedItemSet, setCheckedItemSet, tableFilter } = props;
	const localeService = (0, _univerjs_ui.useDependency)(_univerjs_core.LocaleService);
	const { data: items, itemsCountMap, allItemsCount } = (0, _univerjs_ui.useDependency)(SheetsTableUiService).getTableFilterItems(unitId, subUnitId, tableId, columnIndex);
	const [allChecked, setAllChecked] = (0, react.useState)(tableFilter === void 0 ? true : checkedItemSet.size === itemsCountMap.size);
	const [checkedCount, setCheckedCount] = (0, react.useState)(allChecked ? allItemsCount : getCheckedCount(itemsCountMap));
	const indeterminate = !allChecked && checkedItemSet.size > 0;
	const [searchText, setSearchText] = (0, react.useState)("");
	const displayItems = (0, react.useMemo)(() => {
		return searchText ? items.filter((item) => {
			return String(item.title).toLowerCase().includes(searchText.toLowerCase());
		}) : items;
	}, [searchText, items]);
	const onCheckAllToggled = (0, react.useCallback)(() => {
		if (allChecked) {
			checkedItemSet.clear();
			setCheckedItemSet(new Set(checkedItemSet));
			setAllChecked(false);
		} else {
			displayItems.forEach((item) => {
				checkedItemSet.add(item.title);
			});
			setCheckedItemSet(new Set(checkedItemSet));
			setAllChecked(true);
		}
	}, [allChecked]);
	const onSearchValueChange = (0, react.useCallback)((str) => {
		if (str === "") {
			setAllChecked(true);
			items.forEach((item) => {
				checkedItemSet.add(item.title);
			});
			setCheckedCount(allItemsCount);
		} else {
			checkedItemSet.clear();
			setAllChecked(false);
			setCheckedCount(0);
		}
		setSearchText(str);
	}, []);
	const onCheckItemToggled = (key) => {
		if (allChecked) {
			setAllChecked(false);
			const newSet = /* @__PURE__ */ new Set();
			for (const { title } of items) if (key !== title) newSet.add(title);
			setCheckedCount(allItemsCount - itemsCountMap.get(key));
			setCheckedItemSet(newSet);
		} else {
			if (checkedItemSet.has(key)) {
				checkedItemSet.delete(key);
				setCheckedCount(checkedCount - itemsCountMap.get(key));
			} else {
				checkedItemSet.add(key);
				setCheckedCount(checkedCount + itemsCountMap.get(key));
			}
			setCheckedItemSet(new Set(checkedItemSet));
		}
	};
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: "univer-flex univer-h-full univer-flex-col",
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Input, {
			autoFocus: true,
			value: searchText,
			placeholder: localeService.t("sheets-table-ui.filter.search-placeholder"),
			onChange: onSearchValueChange
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: (0, _univerjs_design.clsx)("univer-mt-2 univer-box-border univer-flex univer-h-[180px] univer-max-h-[180px] univer-flex-grow univer-flex-col univer-overflow-hidden univer-rounded-md univer-py-1.5 univer-pl-2", _univerjs_design.borderClassName),
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: (0, _univerjs_design.clsx)("univer-h-40 univer-min-w-0 univer-overflow-y-auto univer-py-1 univer-pl-2", _univerjs_design.scrollbarClassName),
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
					className: "univer-h-full",
					children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
						className: "univer-flex univer-items-center univer-px-2 univer-py-1",
						children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Checkbox, {
							className: "univer-min-w-0 univer-flex-1",
							contentClassName: "univer-flex-1 univer-min-w-0",
							indeterminate,
							disabled: items.length === 0,
							checked: allChecked,
							onChange: onCheckAllToggled,
							children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
								className: "univer-flex univer-h-5 univer-flex-1 univer-items-center univer-text-sm",
								children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
									className: "univer-flex-1 univer-truncate",
									children: `${localeService.t("sheets-table-ui.filter.select-all")}`
								}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
									className: "univer-ml univer-text-gray-400",
									children: `(${checkedCount}/${searchText ? displayItems.length : allItemsCount})`
								})]
							})
						})
					}), displayItems.map((item) => {
						return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
							className: "univer-flex univer-items-center univer-px-2 univer-py-1",
							children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Checkbox, {
								className: "univer-min-w-0 univer-flex-1",
								contentClassName: "univer-flex-1 univer-min-w-0",
								checked: allChecked || checkedItemSet.has(item.title),
								onChange: () => {
									onCheckItemToggled(item.title);
								},
								children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("span", {
									className: "univer-flex univer-h-5 univer-flex-1 univer-items-center univer-text-sm",
									children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
										className: "univer-flex-1 univer-truncate",
										children: item.title
									}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
										className: "univer-ml-1 univer-inline-flex univer-h-full univer-items-center univer-text-gray-400",
										children: `(${itemsCountMap.get(item.title) || 0})`
									})]
								})
							})
						}, item.key);
					})]
				})
			})
		})]
	});
}

//#endregion
//#region src/views/components/SheetTableFilterPanel.tsx
function SheetTableFilterPanel() {
	var _permissionService$ge;
	const localeService = (0, _univerjs_ui.useDependency)(_univerjs_core.LocaleService);
	const filterByItems = useFilterByOptions(localeService);
	const tableUiService = (0, _univerjs_ui.useDependency)(SheetsTableUiService);
	const tableManager = (0, _univerjs_ui.useDependency)(_univerjs_sheets_table.TableManager);
	const commandService = (0, _univerjs_ui.useDependency)(_univerjs_core.ICommandService);
	const permissionService = (0, _univerjs_ui.useDependency)(_univerjs_core.IPermissionService);
	const sheetsTableComponentController = (0, _univerjs_ui.useDependency)(SheetsTableComponentController);
	const tableFilterPanelInfo = sheetsTableComponentController.getCurrentTableFilterInfo();
	const props = tableUiService.getTableFilterPanelInitProps(tableFilterPanelInfo.unitId, tableFilterPanelInfo.subUnitId, tableFilterPanelInfo.tableId, tableFilterPanelInfo.column);
	const { unitId, subUnitId, tableId, tableFilter, currentFilterBy, columnIndex } = props;
	const { data } = tableUiService.getTableFilterItems(unitId, subUnitId, tableId, columnIndex);
	const checkedItems = tableUiService.getTableFilterCheckedItems(unitId, tableId, columnIndex);
	const [checkedItemSet, setCheckedItemSet] = (0, react.useState)(new Set(checkedItems));
	const [filterBy, setFilterBy] = (0, react.useState)(currentFilterBy || "items");
	const [conditionInfo, setConditionInfo] = (0, react.useState)(() => {
		const tableFilter = props.tableFilter;
		return getInitConditionInfo(tableFilter);
	});
	const table = tableManager.getTable(unitId, tableId);
	if (!table) return null;
	const tableFilters = table.getTableFilters();
	const tableRange = table.getRange();
	const sortState = tableFilters.getSortState();
	sortState.columnIndex === columnIndex && (sortState.sortState, _univerjs_sheets_table.SheetsTableSortStateEnum.Asc);
	sortState.columnIndex === columnIndex && (sortState.sortState, _univerjs_sheets_table.SheetsTableSortStateEnum.Desc);
	const absoluteColumn = tableFilterPanelInfo.column;
	const canDeleteColumn = tableRange.endColumn > tableRange.startColumn;
	const closeDialog = () => {
		sheetsTableComponentController.closeFilterPanel();
	};
	const onCancel = () => {
		closeDialog();
	};
	const applySort = (asc) => {
		const range = table.getTableFilterRange();
		commandService.executeCommand(_univerjs_sheets_sort.SortRangeCommand.id, {
			unitId,
			subUnitId,
			range,
			orderRules: [{
				colIndex: columnIndex + range.startColumn,
				type: asc ? _univerjs_sheets_sort.SortType.ASC : _univerjs_sheets_sort.SortType.DESC
			}],
			hasTitle: false
		});
		tableFilters.setSortState(columnIndex, asc ? _univerjs_sheets_table.SheetsTableSortStateEnum.Asc : _univerjs_sheets_table.SheetsTableSortStateEnum.Desc);
		closeDialog();
	};
	const insertColumn = (side) => {
		commandService.executeCommand(_univerjs_sheets_table.SheetTableInsertColumnAtCommand.id, {
			unitId,
			subUnitId,
			tableId,
			index: side === "left" ? absoluteColumn : absoluteColumn + 1,
			count: 1
		});
		closeDialog();
	};
	const deleteColumn = () => {
		if (!canDeleteColumn) return;
		commandService.executeCommand(_univerjs_sheets_table.SheetTableRemoveColumnAtCommand.id, {
			unitId,
			subUnitId,
			tableId,
			index: absoluteColumn,
			count: 1
		});
		closeDialog();
	};
	const onApply = () => {
		if (filterBy === "items") {
			const filteredItems = [];
			const emptyLabel = localeService.t("sheets-table-ui.condition.empty");
			for (const itemInfo of data) if (checkedItemSet.has(itemInfo.title)) filteredItems.push(itemInfo.title === emptyLabel ? _univerjs_sheets_table.TABLE_FILTER_EMPTY_VALUE : itemInfo.title);
			const originFilter = table.getTableFilterColumn(columnIndex);
			if (originFilter) {
				if (originFilter.values.join(",") === filteredItems.join(",")) {
					closeDialog();
					return;
				}
			} else if (filteredItems.length === 0) {
				closeDialog();
				return;
			}
			const tableFilter = {
				filterType: _univerjs_sheets_table.TableColumnFilterTypeEnum.manual,
				values: filteredItems
			};
			tableUiService.setTableFilter(unitId, tableId, columnIndex, tableFilter);
		} else {
			let filterInfo;
			if (conditionInfo.compare === _univerjs_sheets_table.TableDateCompareTypeEnum.Quarter || conditionInfo.compare === _univerjs_sheets_table.TableDateCompareTypeEnum.Month) filterInfo = {
				conditionType: conditionInfo.type,
				compareType: Object.values(conditionInfo.info)[0]
			};
			else filterInfo = {
				conditionType: conditionInfo.type,
				compareType: conditionInfo.compare,
				expectedValue: Object.values(conditionInfo.info)[0]
			};
			const tableFilter = {
				filterType: _univerjs_sheets_table.TableColumnFilterTypeEnum.condition,
				filterInfo
			};
			tableUiService.setTableFilter(unitId, tableId, columnIndex, tableFilter);
		}
		closeDialog();
	};
	const onClearFilter = () => {
		tableUiService.setTableFilter(unitId, tableId, columnIndex, void 0);
		closeDialog();
	};
	const workbookEditableId = new _univerjs_sheets.WorkbookEditablePermission(unitId).id;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: "univer-box-border univer-flex univer-w-[400px] univer-flex-col univer-rounded-[10px] univer-bg-white univer-p-4 univer-shadow-lg dark:!univer-border-gray-600 dark:!univer-bg-gray-700",
		children: [
			((_permissionService$ge = permissionService.getPermissionPoint(workbookEditableId)) === null || _permissionService$ge === void 0 ? void 0 : _permissionService$ge.value) && /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react_jsx_runtime.Fragment, { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: "-univer-mx-4 -univer-mt-2 univer-mb-3 univer-border-0 univer-border-b univer-border-solid univer-border-gray-200 univer-py-1",
				children: [
					/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("button", {
						type: "button",
						className: "univer-box-border univer-flex univer-h-10 univer-w-full univer-cursor-pointer univer-items-center univer-gap-3 univer-border-none univer-bg-transparent univer-px-4 univer-text-left univer-text-sm univer-text-gray-900 hover:univer-bg-gray-100 disabled:univer-cursor-not-allowed disabled:univer-text-gray-400 dark:!univer-text-white dark:hover:!univer-bg-gray-600",
						onClick: () => insertColumn("left"),
						children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_icons.LeftInsertColumnDoubleIcon, {
							className: "univer-size-5",
							extend: { colorChannel1: "var(--univer-primary-600)" }
						}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { children: localeService.t("sheets-table-ui.columnMenu.insert-left") })]
					}),
					/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("button", {
						type: "button",
						className: "univer-box-border univer-flex univer-h-10 univer-w-full univer-cursor-pointer univer-items-center univer-gap-3 univer-border-none univer-bg-transparent univer-px-4 univer-text-left univer-text-sm univer-text-gray-900 hover:univer-bg-gray-100 disabled:univer-cursor-not-allowed disabled:univer-text-gray-400 dark:!univer-text-white dark:hover:!univer-bg-gray-600",
						onClick: () => insertColumn("right"),
						children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_icons.RightInsertColumnDoubleIcon, {
							className: "univer-size-5",
							extend: { colorChannel1: "var(--univer-primary-600)" }
						}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { children: localeService.t("sheets-table-ui.columnMenu.insert-right") })]
					}),
					/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("button", {
						type: "button",
						className: "univer-box-border univer-flex univer-h-10 univer-w-full univer-cursor-pointer univer-items-center univer-gap-3 univer-border-none univer-bg-transparent univer-px-4 univer-text-left univer-text-sm univer-text-gray-900 hover:univer-bg-gray-100 disabled:univer-cursor-not-allowed disabled:univer-text-gray-400 dark:!univer-text-white dark:hover:!univer-bg-gray-600",
						disabled: !canDeleteColumn,
						onClick: deleteColumn,
						children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_icons.DeleteColumnDoubleIcon, {
							className: "univer-size-5",
							extend: { colorChannel1: "var(--univer-primary-600)" }
						}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { children: localeService.t("sheets-table-ui.columnMenu.delete") })]
					})
				]
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "univer-mb-3 univer-flex",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(_univerjs_design.ButtonGroup, {
					className: "univer-mb-3 !univer-flex univer-w-full",
					children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)(_univerjs_design.Button, {
						className: "univer-w-1/2",
						onClick: () => applySort(true),
						children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_icons.AscendingIcon, { className: "univer-mr-1" }), localeService.t("sheets-table-ui.sort.sort-asc")]
					}), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(_univerjs_design.Button, {
						className: "univer-w-1/2",
						onClick: () => applySort(false),
						children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_icons.DescendingIcon, { className: "univer-mr-1" }), localeService.t("sheets-table-ui.sort.sort-desc")]
					})]
				})
			})] }),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "univer-w-full",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Segmented, {
					value: filterBy,
					items: filterByItems,
					onChange: (value) => setFilterBy(value)
				})
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "univer-z-10 univer-h-60",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
					className: "univer-mt-3 univer-size-full",
					children: filterBy === "items" ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)(SheetTableItemsFilterPanel, {
						tableFilter,
						unitId,
						subUnitId,
						tableId,
						columnIndex,
						checkedItemSet,
						setCheckedItemSet
					}) : /* @__PURE__ */ (0, react_jsx_runtime.jsx)(SheetTableConditionPanel, {
						tableFilter,
						unitId,
						subUnitId,
						tableId,
						columnIndex,
						conditionInfo,
						onChange: setConditionInfo
					})
				})
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: "univer-flex-wrap-nowrap univer-mt-4 univer-inline-flex univer-flex-shrink-0 univer-flex-grow-0 univer-justify-between univer-gap-6 univer-overflow-hidden",
				children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Button, {
					disabled: tableFilter === void 0,
					onClick: onClearFilter,
					children: localeService.t("sheets-table-ui.filter.clear-filter")
				}), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Button, {
					className: "univer-mr-2",
					onClick: onCancel,
					children: localeService.t("sheets-table-ui.filter.cancel")
				}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Button, {
					variant: "primary",
					onClick: onApply,
					children: localeService.t("sheets-table-ui.filter.confirm")
				})] })]
			})
		]
	});
}
function useFilterByOptions(localeService) {
	return (0, react.useMemo)(() => [{
		label: localeService.t("sheets-table-ui.filter.by-values"),
		value: "items"
	}, {
		label: localeService.t("sheets-table-ui.filter.by-conditions"),
		value: "condition"
	}], [localeService.getCurrentLocale(), localeService]);
}

//#endregion
//#region src/views/components/SheetTableRenameDialog.tsx
function SheetTableRenameDialog(props) {
	var _table$getDisplayName;
	const { unitId, tableId, onClose } = props;
	const localeService = (0, _univerjs_ui.useDependency)(_univerjs_core.LocaleService);
	const commandService = (0, _univerjs_ui.useDependency)(_univerjs_core.ICommandService);
	const tableManager = (0, _univerjs_ui.useDependency)(_univerjs_sheets_table.TableManager);
	const univerInstanceService = (0, _univerjs_ui.useDependency)(_univerjs_core.IUniverInstanceService);
	const definedNamesService = (0, _univerjs_ui.useDependency)(_univerjs_engine_formula.IDefinedNamesService);
	const table = tableManager.getTableById(unitId, tableId);
	const [value, setValue] = (0, react.useState)((_table$getDisplayName = table === null || table === void 0 ? void 0 : table.getDisplayName()) !== null && _table$getDisplayName !== void 0 ? _table$getDisplayName : "");
	const [error, setError] = (0, react.useState)("");
	const existingNames = (0, react.useMemo)(() => {
		const names = (0, _univerjs_sheets_table.getExistingNamesSet)(unitId, {
			univerInstanceService,
			tableManager,
			definedNamesService
		});
		const currentName = table === null || table === void 0 ? void 0 : table.getDisplayName().toLowerCase();
		if (currentName) names.delete(currentName);
		return names;
	}, [
		definedNamesService,
		table,
		tableManager,
		unitId,
		univerInstanceService
	]);
	const handleConfirm = () => {
		const nextName = value.trim();
		if (!table || nextName === table.getDisplayName()) {
			onClose();
			return;
		}
		if (!(0, _univerjs_sheets_table.validateSheetTableName)(nextName, existingNames).valid) {
			setError(localeService.t("sheets-table-ui.tableNameError"));
			return;
		}
		commandService.executeCommand(_univerjs_sheets_table.SetSheetTableCommand.id, {
			unitId,
			tableId,
			name: nextName
		});
		onClose();
	};
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: "univer-box-border univer-flex univer-w-full univer-flex-col univer-gap-4 univer-pb-3 univer-pt-2",
		children: [
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Input, {
				size: "middle",
				value,
				placeholder: localeService.t("sheets-table-ui.renamePlaceholder"),
				onChange: (nextValue) => {
					setValue(nextValue);
					setError("");
				},
				onKeyDown: (event) => {
					if (event.key === "Enter") handleConfirm();
				},
				autoFocus: true
			}),
			error ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "-univer-mt-2 univer-text-sm univer-text-red-500",
				children: error
			}) : null,
			/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: "univer-flex univer-w-full univer-items-center univer-justify-end univer-gap-2",
				children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Button, {
					className: "univer-min-w-16",
					onClick: onClose,
					children: localeService.t("sheets-table-ui.cancel")
				}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Button, {
					className: "univer-min-w-16",
					variant: "primary",
					onClick: handleConfirm,
					children: localeService.t("sheets-table-ui.confirm")
				})]
			})
		]
	});
}

//#endregion
//#region src/controllers/sheet-table-component.controller.ts
let SheetsTableComponentController = class SheetsTableComponentController extends _univerjs_core.Disposable {
	constructor(_componentManager, _contextService, _sheetCanvasPopupService, _dialogService) {
		super();
		this._componentManager = _componentManager;
		this._contextService = _contextService;
		this._sheetCanvasPopupService = _sheetCanvasPopupService;
		this._dialogService = _dialogService;
		_defineProperty(this, "_popupDisposable", void 0);
		_defineProperty(this, "_currentTableFilterInfo", null);
		this._initComponents();
		this._initUIPopup();
	}
	setCurrentTableFilterInfo(info) {
		this._currentTableFilterInfo = info;
	}
	openOrToggleFilterPanel(info) {
		const opened = this._contextService.getContextValue(SHEETS_TABLE_FILTER_PANEL_OPENED_KEY);
		if (opened && this._isSameFilterPanelInfo(this._currentTableFilterInfo, info)) {
			this.closeFilterPanel();
			return;
		}
		this.setCurrentTableFilterInfo(info);
		if (opened) {
			var _this$_popupDisposabl;
			(_this$_popupDisposabl = this._popupDisposable) === null || _this$_popupDisposabl === void 0 || _this$_popupDisposabl.dispose();
			this._popupDisposable = null;
			this._openFilterPopup();
			return;
		}
		this._contextService.setContextValue(SHEETS_TABLE_FILTER_PANEL_OPENED_KEY, true);
	}
	clearCurrentTableFilterInfo() {
		this._currentTableFilterInfo = null;
	}
	getCurrentTableFilterInfo() {
		return this._currentTableFilterInfo;
	}
	_initComponents() {
		[[SHEETS_TABLE_FILTER_PANEL_OPENED_KEY, SheetTableFilterPanel], [SHEET_TABLE_RENAME_DIALOG, SheetTableRenameDialog]].forEach(([key, comp]) => {
			this.disposeWithMe(this._componentManager.register(key, comp));
		});
	}
	_initUIPopup() {
		this.disposeWithMe(this._contextService.subscribeContextValue$(SHEETS_TABLE_FILTER_PANEL_OPENED_KEY).pipe((0, rxjs.startWith)(void 0), (0, rxjs.distinctUntilChanged)()).subscribe((open) => {
			if (open) this._openFilterPopup();
			else if (open === false) this._closeFilterPopup();
		}));
	}
	closeFilterPanel() {
		this._contextService.setContextValue(SHEETS_TABLE_FILTER_PANEL_OPENED_KEY, false);
	}
	_openFilterPopup() {
		const currentFilterModel = this._currentTableFilterInfo;
		if (!currentFilterModel) throw new Error("[SheetsFilterUIController]: no filter model when opening filter popup!");
		const { row: startRow, column: col } = currentFilterModel;
		this._popupDisposable = this._sheetCanvasPopupService.attachPopupToCell(startRow, col, {
			componentKey: SHEETS_TABLE_FILTER_PANEL_OPENED_KEY,
			direction: "horizontal",
			onClickOutside: () => {
				this._dialogService.close(UNIVER_SHEET_TABLE_FILTER_PANEL_ID);
				this._contextService.setContextValue(SHEETS_TABLE_FILTER_PANEL_OPENED_KEY, false);
			},
			offset: [5, 0],
			portal: true
		});
	}
	_closeFilterPopup() {
		var _this$_popupDisposabl2;
		(_this$_popupDisposabl2 = this._popupDisposable) === null || _this$_popupDisposabl2 === void 0 || _this$_popupDisposabl2.dispose();
		this._popupDisposable = null;
		this.clearCurrentTableFilterInfo();
	}
	_isSameFilterPanelInfo(a, b) {
		return Boolean(a && a.unitId === b.unitId && a.subUnitId === b.subUnitId && a.tableId === b.tableId && a.column === b.column && a.row === b.row);
	}
};
SheetsTableComponentController = __decorate([
	__decorateParam(0, (0, _univerjs_core.Inject)(_univerjs_ui.ComponentManager)),
	__decorateParam(1, _univerjs_core.IContextService),
	__decorateParam(2, (0, _univerjs_core.Inject)(_univerjs_sheets_ui.SheetCanvasPopManagerService)),
	__decorateParam(3, (0, _univerjs_core.Inject)(_univerjs_ui.IDialogService))
], SheetsTableComponentController);

//#endregion
//#region src/commands/operations/open-table-filter-dialog.opration.ts
const OpenTableFilterPanelOperation = {
	type: _univerjs_core.CommandType.OPERATION,
	id: "sheet.operation.open-table-filter-panel",
	async handler(accessor, params) {
		if (!params) return false;
		const { row, col, unitId, subUnitId, tableId } = params;
		const tableManager = accessor.get(_univerjs_sheets_table.TableManager);
		const sheetsTableComponentController = accessor.get(SheetsTableComponentController);
		if (!tableManager.getTable(unitId, tableId)) return false;
		sheetsTableComponentController.openOrToggleFilterPanel({
			unitId,
			subUnitId,
			row,
			tableId,
			column: col
		});
		return true;
	}
};

//#endregion
//#region src/commands/operations/open-table-selector.operation.ts
const OpenTableSelectorOperation = {
	type: _univerjs_core.CommandType.OPERATION,
	id: "sheet.operation.open-table-selector",
	async handler(accessor) {
		var _lastSelection$range;
		const univerInstanceService = accessor.get(_univerjs_core.IUniverInstanceService);
		const commandService = accessor.get(_univerjs_core.ICommandService);
		const target = (0, _univerjs_sheets.getSheetCommandTarget)(univerInstanceService);
		if (!target) return false;
		const { unitId, subUnitId, worksheet } = target;
		const lastSelection = accessor.get(_univerjs_sheets.SheetsSelectionsService).getCurrentLastSelection();
		const range = (_lastSelection$range = lastSelection === null || lastSelection === void 0 ? void 0 : lastSelection.range) !== null && _lastSelection$range !== void 0 ? _lastSelection$range : {
			startRow: 0,
			endRow: 0,
			startColumn: 0,
			endColumn: 0
		};
		const rangeInfo = await openRangeSelector(accessor, unitId, subUnitId, (0, _univerjs_sheets.isSingleCellSelection)(lastSelection) ? (0, _univerjs_sheets.expandToContinuousRange)(range, {
			up: true,
			left: true,
			right: true,
			down: true
		}, worksheet) : range);
		if (!rangeInfo) return false;
		commandService.executeCommand(_univerjs_sheets_table.AddSheetTableCommand.id, { ...rangeInfo });
		return true;
	}
};
async function openRangeSelector(accessor, unitId, subUnitId, range, tableId) {
	const dialogService = accessor.get(_univerjs_ui.IDialogService);
	const localeService = accessor.get(_univerjs_core.LocaleService);
	return new Promise((resolve) => {
		const dialogProps = {
			unitId,
			subUnitId,
			range,
			tableId,
			onConfirm: (info) => {
				resolve(info);
				dialogService.close(TABLE_SELECTOR_DIALOG);
			},
			onCancel: () => {
				resolve(null);
				dialogService.close(TABLE_SELECTOR_DIALOG);
			}
		};
		dialogService.open({
			id: TABLE_SELECTOR_DIALOG,
			title: { title: localeService.t("sheets-table-ui.selectRange") },
			draggable: true,
			destroyOnClose: true,
			mask: false,
			maskClosable: false,
			children: { label: {
				name: TABLE_SELECTOR_DIALOG,
				props: dialogProps
			} },
			width: 300,
			onClose: () => {
				resolve(null);
				dialogService.close(TABLE_SELECTOR_DIALOG);
			}
		});
	});
}

//#endregion
//#region src/config/config.ts
const SHEETS_TABLE_UI_PLUGIN_CONFIG_KEY = "sheets-table-ui.config";
const configSymbol = Symbol(SHEETS_TABLE_UI_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {
	anchorHeight: 24,
	anchorBackgroundColor: "rgb(134,139,156)"
};

//#endregion
//#region src/views/widgets/table-controls-util.ts
const TABLE_CONTROL_MENU_WIDTH = 168;
const TABLE_CONTROL_MENU_ITEM_HEIGHT = 32;
const TABLE_CONTROL_MENU_ACTIONS = [
	"rename",
	"update-range",
	"set-theme",
	"delete"
];
function isPointInTableControlRegion(region, x, y) {
	return x >= region.left && x <= region.left + region.width && y >= region.top && y <= region.top + region.height;
}
function hitTestTableControl(regions, x, y) {
	for (let i = regions.length - 1; i >= 0; i--) if (isPointInTableControlRegion(regions[i], x, y)) return regions[i];
	return null;
}
function buildTableMenuRegions(tableId, left, top) {
	return TABLE_CONTROL_MENU_ACTIONS.map((action, index) => ({
		type: "menu-item",
		tableId,
		action,
		left,
		top: top + index * 32,
		width: 168,
		height: 32
	}));
}
function buildCenteredPlusSegments(centerX, centerY, size) {
	const halfSize = size / 2;
	return [{
		fromX: centerX - halfSize,
		fromY: centerY,
		toX: centerX + halfSize,
		toY: centerY
	}, {
		fromX: centerX,
		fromY: centerY - halfSize,
		toX: centerX,
		toY: centerY + halfSize
	}];
}

//#endregion
//#region src/views/widgets/table-controls.shape.ts
const ANCHOR_MIN_WIDTH = 122;
const ANCHOR_MAX_WIDTH = 240;
const ANCHOR_PADDING_X = 12;
const ANCHOR_TOGGLE_WIDTH = 30;
const ANCHOR_OFFSET_Y = 0;
const ANCHOR_BORDER = "rgba(0, 0, 0, 0.22)";
const ANCHOR_DIVIDER = "rgba(0, 0, 0, 0.20)";
const ANCHOR_TOGGLE_BG_ACTIVE = "rgba(0, 0, 0, 0.12)";
const MENU_RADIUS = 8;
const MENU_BORDER = "#d9dee7";
const MENU_HOVER_BG = "#f1f3f4";
const INSERT_BUTTON_VISUAL_SIZE = 18;
const INSERT_BUTTON_PLUS_SIZE = 8;
var SheetTableControlsShape = class extends _univerjs_engine_render.Shape {
	constructor(key, _getSkeleton) {
		super(key, {
			evented: true,
			fill: "rgba(0, 0, 0, 0)",
			zIndex: 5001
		});
		this._getSkeleton = _getSkeleton;
		_defineProperty(this, "_items", []);
		_defineProperty(this, "_regions", []);
		_defineProperty(this, "_openedMenuTableId", null);
		_defineProperty(this, "_hoveredRegion", null);
		_defineProperty(this, "_hoveredInsertRegion", null);
		_defineProperty(this, "_menuLabels", {
			rename: "Rename table",
			"update-range": "Update range",
			"set-theme": "Set theme",
			delete: "Remove table"
		});
	}
	setItems(items) {
		this._items = items;
		this.makeDirty(true);
	}
	setMenuLabels(labels) {
		this._menuLabels = labels;
		this.makeDirty(true);
	}
	setOpenedMenuTableId(tableId) {
		if (this._openedMenuTableId === tableId) return;
		this._openedMenuTableId = tableId;
		this.makeDirty(true);
	}
	getOpenedMenuTableId() {
		return this._openedMenuTableId;
	}
	setHoveredRegion(region) {
		if (this._hoveredRegion === region) return;
		this._hoveredRegion = region;
		this.makeDirty(true);
	}
	setHoveredInsertRegion(region) {
		if (this._hoveredInsertRegion === region) return;
		this._hoveredInsertRegion = region;
		this.makeDirty(true);
	}
	hitTest(x, y) {
		return hitTestTableControl(this._regions, x, y);
	}
	isHit(coord) {
		return this.hitTest(coord.x, coord.y) != null;
	}
	refreshBounds() {
		const skeleton = this._getSkeleton();
		if (!skeleton) {
			this.hide();
			return;
		}
		this.show();
		this.transformByState({
			left: 0,
			top: 0,
			width: skeleton.rowHeaderWidth + skeleton.columnTotalWidth,
			height: skeleton.columnHeaderHeight + skeleton.rowTotalHeight
		});
	}
	_draw(ctx) {
		this._regions = [];
		const skeleton = this._getSkeleton();
		if (!skeleton) return;
		ctx.save();
		ctx.textBaseline = "middle";
		for (const item of this._items) this._drawAnchor(ctx, skeleton, item);
		if (this._hoveredInsertRegion) {
			var _item$fill;
			const item = this._items.find((renderItem) => {
				var _this$_hoveredInsertR;
				return renderItem.tableId === ((_this$_hoveredInsertR = this._hoveredInsertRegion) === null || _this$_hoveredInsertR === void 0 ? void 0 : _this$_hoveredInsertR.tableId);
			});
			this._drawInsertButton(ctx, this._hoveredInsertRegion, (_item$fill = item === null || item === void 0 ? void 0 : item.fill) !== null && _item$fill !== void 0 ? _item$fill : "#355bb7");
			this._regions.push(this._hoveredInsertRegion);
		}
		ctx.restore();
	}
	_drawAnchor(ctx, skeleton, item) {
		const position = skeleton.getNoMergeCellWithCoordByIndex(item.range.startRow, item.range.startColumn);
		const left = position.startX;
		const rawTop = position.startY - 28 - ANCHOR_OFFSET_Y;
		const top = Math.max(0, rawTop);
		const width = Math.max(ANCHOR_MIN_WIDTH, Math.min(ANCHOR_MAX_WIDTH, item.tableName.length * 8.5 + ANCHOR_PADDING_X * 2 + ANCHOR_TOGGLE_WIDTH));
		const toggleRegion = {
			type: "anchor-menu-toggle",
			tableId: item.tableId,
			left: left + width - ANCHOR_TOGGLE_WIDTH,
			top,
			width: ANCHOR_TOGGLE_WIDTH,
			height: 28
		};
		ctx.save();
		ctx.translateWithPrecision(left, top);
		this._drawTopRoundedRect(ctx, width, 28, 14, item.fill, ANCHOR_BORDER);
		this._drawAnchorToggle(ctx, width, item.text, this._openedMenuTableId === item.tableId || this._isSameRegion(this._hoveredRegion, toggleRegion));
		ctx.font = `600 13px ${_univerjs_engine_render.DEFAULT_FONTFACE_PLANE}`;
		ctx.fillStyle = item.text;
		ctx.textAlign = "left";
		ctx.fillText(item.tableName, ANCHOR_PADDING_X, 28 / 2);
		ctx.restore();
		this._regions.push({
			type: "anchor-main",
			tableId: item.tableId,
			left,
			top,
			width,
			height: 28
		});
		this._regions.push(toggleRegion);
		if (this._openedMenuTableId === item.tableId) this._drawMenu(ctx, item.tableId, left, top + 28);
	}
	_drawAnchorToggle(ctx, anchorWidth, color, active) {
		const toggleLeft = anchorWidth - ANCHOR_TOGGLE_WIDTH;
		if (active) this._drawRightTopRoundedRect(ctx, toggleLeft, anchorWidth, 28, 14, ANCHOR_TOGGLE_BG_ACTIVE);
		ctx.save();
		ctx.beginPath();
		ctx.strokeStyle = ANCHOR_DIVIDER;
		ctx.lineWidth = 1;
		ctx.moveTo(toggleLeft + .5, 5);
		ctx.lineTo(toggleLeft + .5, 28 - 5);
		ctx.stroke();
		ctx.restore();
		const centerX = anchorWidth - ANCHOR_TOGGLE_WIDTH / 2;
		const centerY = 28 / 2;
		ctx.save();
		ctx.beginPath();
		ctx.strokeStyle = color;
		ctx.lineWidth = 1.6;
		ctx.lineCap = "round";
		ctx.moveTo(centerX - 5, centerY - 4);
		ctx.lineTo(centerX + 5, centerY - 4);
		ctx.moveTo(centerX - 5, centerY);
		ctx.lineTo(centerX + 5, centerY);
		ctx.moveTo(centerX - 5, centerY + 4);
		ctx.lineTo(centerX + 5, centerY + 4);
		ctx.stroke();
		ctx.restore();
	}
	_drawTopRoundedRect(ctx, width, height, radius, fill, stroke) {
		const r = Math.min(radius, width / 2, height);
		ctx.beginPath();
		ctx.moveTo(0, height);
		ctx.lineTo(0, r);
		ctx.arcTo(0, 0, r, 0, r);
		ctx.lineTo(width - r, 0);
		ctx.arcTo(width, 0, width, r, r);
		ctx.lineTo(width, height);
		ctx.closePath();
		ctx.fillStyle = fill;
		ctx.fill();
		if (stroke) {
			ctx.strokeStyle = stroke;
			ctx.lineWidth = 1;
			ctx.stroke();
		}
	}
	_drawRightTopRoundedRect(ctx, left, width, height, radius, fill) {
		const r = Math.min(radius, width - left, height);
		ctx.save();
		ctx.beginPath();
		ctx.moveTo(left, height);
		ctx.lineTo(left, 0);
		ctx.lineTo(width - r, 0);
		ctx.arcTo(width, 0, width, r, r);
		ctx.lineTo(width, height);
		ctx.closePath();
		ctx.fillStyle = fill;
		ctx.fill();
		ctx.restore();
	}
	_drawMenu(ctx, tableId, left, top) {
		const regions = buildTableMenuRegions(tableId, left, top);
		ctx.save();
		ctx.translateWithPrecision(left, top);
		_univerjs_engine_render.Rect.drawWith(ctx, {
			width: 168,
			height: regions.length * 32,
			radius: MENU_RADIUS,
			fill: "#fff",
			stroke: MENU_BORDER
		});
		ctx.restore();
		for (const region of regions) {
			if (this._isSameRegion(this._hoveredRegion, region)) {
				ctx.save();
				ctx.fillStyle = MENU_HOVER_BG;
				ctx.fillRectByPrecision(region.left, region.top, region.width, region.height);
				ctx.restore();
			}
			ctx.save();
			ctx.font = `12px ${_univerjs_engine_render.DEFAULT_FONTFACE_PLANE}`;
			ctx.fillStyle = region.action === "delete" ? "#d92d20" : "#344054";
			ctx.textAlign = "left";
			ctx.fillText(this._menuLabels[region.action], region.left + 12, region.top + region.height / 2);
			ctx.restore();
		}
		this._regions.push(...regions);
	}
	_drawInsertButton(ctx, region, fill) {
		const centerX = region.left + region.width / 2;
		const centerY = region.top + region.height / 2;
		const radius = INSERT_BUTTON_VISUAL_SIZE / 2;
		ctx.save();
		ctx.beginPath();
		ctx.arc(centerX, centerY, radius, 0, Math.PI * 2);
		ctx.fillStyle = "#fff";
		ctx.fill();
		ctx.strokeStyle = fill;
		ctx.stroke();
		ctx.beginPath();
		ctx.strokeStyle = fill;
		ctx.lineWidth = 1;
		ctx.lineCap = "round";
		for (const segment of buildCenteredPlusSegments(centerX, centerY, INSERT_BUTTON_PLUS_SIZE)) {
			ctx.moveTo(segment.fromX, segment.fromY);
			ctx.lineTo(segment.toX, segment.toY);
		}
		ctx.stroke();
		ctx.restore();
	}
	_isSameRegion(a, b) {
		return Boolean(a && a.type === b.type && a.tableId === b.tableId && a.action === b.action && a.index === b.index);
	}
};

//#endregion
//#region src/controllers/sheet-table-theme-ui.controller.ts
let SheetTableThemeUIController = class SheetTableThemeUIController extends _univerjs_core.Disposable {
	constructor(_commandService) {
		super();
		this._commandService = _commandService;
		_defineProperty(this, "_refreshTable", new rxjs.Subject());
		_defineProperty(this, "refreshTable$", this._refreshTable.asObservable());
		this._initListener();
	}
	_initListener() {
		this.disposeWithMe(this._commandService.onCommandExecuted((command) => {
			if (command.id === _univerjs_sheets.SetRangeThemeMutation.id) {
				const { styleName } = command.params;
				if (styleName.startsWith(_univerjs_sheets_table.SHEET_TABLE_CUSTOM_THEME_PREFIX)) this._refreshTable.next(Math.random());
			}
		}));
	}
};
SheetTableThemeUIController = __decorate([__decorateParam(0, (0, _univerjs_core.Inject)(_univerjs_core.ICommandService))], SheetTableThemeUIController);

//#endregion
//#region src/controllers/sheet-table-controls-render.controller.ts
const TABLE_CONTROLS_LAYER_INDEX = 5002;
const TABLE_CONTROL_GAP_ROW = 0;
const TABLE_RENDER_REFRESH_COMMANDS = new Set([_univerjs_sheets_ui.SetScrollOperation.id, _univerjs_sheets_ui.SetZoomRatioOperation.id]);
function isSameTopGap(left, right) {
	if (left === null || right === null) return left === right;
	return left.size === right.size && left.color === right.color && left.stripeColor === right.stripeColor;
}
let SheetTableControlsRenderController = class SheetTableControlsRenderController extends _univerjs_core.Disposable {
	constructor(_context, _injector, _sheetSkeletonManagerService, _commandService, _tableManager, _rangeThemeModel, _workbookPermissionService, _permissionService, _sheetsSelectionsService, _selectionRenderService, _sheetTableThemeUIController, _localeService, _dialogService, _sidebarService) {
		super();
		this._context = _context;
		this._injector = _injector;
		this._sheetSkeletonManagerService = _sheetSkeletonManagerService;
		this._commandService = _commandService;
		this._tableManager = _tableManager;
		this._rangeThemeModel = _rangeThemeModel;
		this._workbookPermissionService = _workbookPermissionService;
		this._permissionService = _permissionService;
		this._sheetsSelectionsService = _sheetsSelectionsService;
		this._selectionRenderService = _selectionRenderService;
		this._sheetTableThemeUIController = _sheetTableThemeUIController;
		this._localeService = _localeService;
		this._dialogService = _dialogService;
		this._sidebarService = _sidebarService;
		_defineProperty(this, "_shape", void 0);
		_defineProperty(this, "_topGapBaseBySkeleton", /* @__PURE__ */ new WeakMap());
		this._shape = new SheetTableControlsShape("SheetTableControlsShape", () => this._sheetSkeletonManagerService.getCurrentSkeleton() || null);
		this._initShape();
		this._initRefresh();
		this._refresh();
	}
	_initShape() {
		var _this$_context$compon, _this$_context$compon2;
		this._context.scene.addObjects([this._shape], TABLE_CONTROLS_LAYER_INDEX);
		this.disposeWithMe((0, _univerjs_core.toDisposable)(() => {
			this._context.scene.removeObjects([this._shape]);
		}));
		this.disposeWithMe(this._shape.onPointerMove$.subscribeEvent((evt, state) => {
			this._handlePointerMove(evt, state);
		}));
		this.disposeWithMe(this._shape.onPointerLeave$.subscribeEvent((_evt, state) => {
			this._handlePointerLeave(state);
		}));
		this.disposeWithMe(this._shape.onPointerDown$.subscribeEvent((evt, state) => {
			this._handlePointerDown(evt, state);
		}));
		this.disposeWithMe((_this$_context$compon = (_this$_context$compon2 = this._context.components.get(_univerjs_sheets_ui.SHEET_VIEW_KEY.MAIN)) === null || _this$_context$compon2 === void 0 ? void 0 : _this$_context$compon2.onPointerMove$.subscribeEvent((evt) => {
			const point = this._getLocalPoint(evt);
			const insertRegion = this._getInsertRegionFromPoint(point.x, point.y);
			this._shape.setHoveredInsertRegion(insertRegion);
		})) !== null && _this$_context$compon !== void 0 ? _this$_context$compon : (0, _univerjs_core.toDisposable)(() => {}));
	}
	_initRefresh() {
		const commandExecuted$ = (0, _univerjs_core.fromCallback)(this._commandService.onCommandExecuted.bind(this._commandService)).pipe((0, rxjs.filter)(([command]) => {
			if (command.type === _univerjs_core.CommandType.OPERATION && TABLE_RENDER_REFRESH_COMMANDS.has(command.id)) {
				this._closeFloatingControls();
				return true;
			}
			return command.type === _univerjs_core.CommandType.MUTATION || command.type === _univerjs_core.CommandType.COMMAND;
		}));
		this.disposeWithMe((0, rxjs.merge)(this._context.unit.activeSheet$, this._sheetSkeletonManagerService.currentSkeleton$, this._tableManager.tableAdd$, this._tableManager.tableDelete$, this._tableManager.tableNameChanged$, this._tableManager.tableRangeChanged$, this._tableManager.tableThemeChanged$, this._sheetTableThemeUIController.refreshTable$, this._workbookPermissionService.unitPermissionInitStateChange$, this._permissionService.permissionPointUpdate$, this._sheetsSelectionsService.selectionChanged$, commandExecuted$).subscribe(() => {
			this._closeFloatingControls();
			this._refresh();
		}));
	}
	_refresh() {
		const skeleton = this._sheetSkeletonManagerService.getCurrentSkeleton();
		const worksheet = this._context.unit.getActiveSheet();
		if (!skeleton || !worksheet || !this._canEditWorkbook()) {
			this._shape.setItems([]);
			this._shape.refreshBounds();
			this._context.scene.makeDirty();
			return;
		}
		this._syncTopTableGap(skeleton);
		this._shape.setMenuLabels({
			rename: this._localeService.t("sheets-table-ui.rename"),
			"update-range": this._localeService.t("sheets-table-ui.updateRange"),
			"set-theme": this._localeService.t("sheets-table-ui.setTheme"),
			delete: this._localeService.t("sheets-table-ui.removeTable")
		});
		const unitId = this._context.unit.getUnitId();
		const subUnitId = worksheet.getSheetId();
		const items = this._tableManager.getTablesBySubunitId(unitId, subUnitId).map((table) => {
			var _rangeTheme$getHeader, _rangeTheme$getHeader2, _rangeTheme$getHeader3, _rangeTheme$getHeader4;
			const rangeTheme = this._rangeThemeModel.getRangeThemeStyle(unitId, table.getTableStyleId());
			return {
				tableId: table.getId(),
				tableName: table.getDisplayName(),
				range: table.getRange(),
				fill: (_rangeTheme$getHeader = rangeTheme === null || rangeTheme === void 0 || (_rangeTheme$getHeader2 = rangeTheme.getHeaderRowStyle()) === null || _rangeTheme$getHeader2 === void 0 || (_rangeTheme$getHeader2 = _rangeTheme$getHeader2.bg) === null || _rangeTheme$getHeader2 === void 0 ? void 0 : _rangeTheme$getHeader2.rgb) !== null && _rangeTheme$getHeader !== void 0 ? _rangeTheme$getHeader : "rgb(53,91,183)",
				text: (_rangeTheme$getHeader3 = rangeTheme === null || rangeTheme === void 0 || (_rangeTheme$getHeader4 = rangeTheme.getHeaderRowStyle()) === null || _rangeTheme$getHeader4 === void 0 || (_rangeTheme$getHeader4 = _rangeTheme$getHeader4.cl) === null || _rangeTheme$getHeader4 === void 0 ? void 0 : _rangeTheme$getHeader4.rgb) !== null && _rangeTheme$getHeader3 !== void 0 ? _rangeTheme$getHeader3 : "rgb(255,255,255)"
			};
		});
		this._shape.setItems(items);
		this._shape.refreshBounds();
		this._shape.makeDirty(true);
		this._context.scene.makeDirty();
	}
	_canEditWorkbook() {
		var _this$_permissionServ;
		const unitId = this._context.unit.getUnitId();
		return ((_this$_permissionServ = this._permissionService.getPermissionPoint(new _univerjs_sheets.WorkbookEditablePermission(unitId).id)) === null || _this$_permissionServ === void 0 ? void 0 : _this$_permissionServ.value) !== false;
	}
	_handlePointerMove(evt, state) {
		const point = this._getLocalPoint(evt);
		const hit = this._shape.hitTest(point.x, point.y);
		const insertRegion = this._isInsertHit(hit) ? hit : hit ? null : this._getInsertRegionFromPoint(point.x, point.y);
		const activeHit = hit !== null && hit !== void 0 ? hit : insertRegion;
		this._shape.setHoveredRegion(this._isInsertHit(hit) ? null : hit);
		this._shape.setHoveredInsertRegion(insertRegion);
		if (activeHit) {
			state.stopPropagation();
			this._context.scene.setCursor(_univerjs_engine_render.CURSOR_TYPE.POINTER);
		} else this._context.scene.resetCursor();
	}
	_isInsertHit(hit) {
		return (hit === null || hit === void 0 ? void 0 : hit.type) === "insert-row" || (hit === null || hit === void 0 ? void 0 : hit.type) === "insert-column";
	}
	_handlePointerLeave(state) {
		state.stopPropagation();
		this._shape.setHoveredRegion(null);
		this._shape.setHoveredInsertRegion(null);
		this._context.scene.resetCursor();
	}
	_handlePointerDown(evt, state) {
		var _this$_shape$hitTest;
		if (evt.button === 2) return;
		const point = this._getLocalPoint(evt);
		const hit = (_this$_shape$hitTest = this._shape.hitTest(point.x, point.y)) !== null && _this$_shape$hitTest !== void 0 ? _this$_shape$hitTest : this._getInsertRegionFromPoint(point.x, point.y);
		if (!hit) {
			this._closeFloatingControls();
			return;
		}
		state.stopPropagation();
		evt.stopPropagation();
		evt.preventDefault();
		this._handleHit(hit);
	}
	_handleHit(hit) {
		const worksheet = this._context.unit.getActiveSheet();
		if (!worksheet) return;
		const unitId = this._context.unit.getUnitId();
		const subUnitId = worksheet.getSheetId();
		if (hit.type === "anchor-menu-toggle" || hit.type === "anchor-main") {
			this._shape.setOpenedMenuTableId(this._shape.getOpenedMenuTableId() === hit.tableId ? null : hit.tableId);
			return;
		}
		if (hit.type === "insert-row") {
			this._commandService.executeCommand(_univerjs_sheets_table.SheetTableInsertRowAtCommand.id, {
				unitId,
				subUnitId,
				tableId: hit.tableId,
				index: hit.index,
				count: 1
			});
			this._closeFloatingControls();
			return;
		}
		if (hit.type === "insert-column") {
			this._commandService.executeCommand(_univerjs_sheets_table.SheetTableInsertColumnAtCommand.id, {
				unitId,
				subUnitId,
				tableId: hit.tableId,
				index: hit.index,
				count: 1
			});
			this._closeFloatingControls();
			return;
		}
		if (hit.type !== "menu-item") return;
		switch (hit.action) {
			case "rename":
				this._openRenameDialog(unitId, hit.tableId);
				break;
			case "update-range":
				this._openRangeSelector(unitId, subUnitId, hit.tableId);
				break;
			case "set-theme":
				this._openThemePanel(unitId, subUnitId, hit.tableId);
				break;
			case "delete":
				this._commandService.executeCommand(_univerjs_sheets_table.DeleteSheetTableCommand.id, {
					tableId: hit.tableId,
					subUnitId,
					unitId
				});
				break;
		}
		this._closeFloatingControls();
	}
	_openRenameDialog(unitId, tableId) {
		this._dialogService.open({
			id: SHEET_TABLE_RENAME_DIALOG_ID,
			title: { title: this._localeService.t("sheets-table-ui.rename") },
			draggable: true,
			destroyOnClose: true,
			mask: true,
			children: { label: {
				name: SHEET_TABLE_RENAME_DIALOG,
				props: {
					unitId,
					tableId,
					onClose: () => this._dialogService.close(SHEET_TABLE_RENAME_DIALOG_ID)
				}
			} },
			width: 360,
			onClose: () => this._dialogService.close(SHEET_TABLE_RENAME_DIALOG_ID)
		});
	}
	async _openRangeSelector(unitId, subUnitId, tableId) {
		const table = this._tableManager.getTableById(unitId, tableId);
		if (!table) return;
		const selection = await openRangeSelector(this._injector, unitId, subUnitId, table.getRange(), tableId);
		if (!selection) return;
		this._commandService.executeCommand(_univerjs_sheets_table.SetSheetTableCommand.id, {
			tableId,
			unitId,
			updateRange: { newRange: selection.range }
		});
	}
	_openThemePanel(unitId, subUnitId, tableId) {
		const table = this._tableManager.getTableById(unitId, tableId);
		if (!table) return;
		this._sidebarService.open({
			id: SHEET_TABLE_THEME_PANEL_ID,
			header: { title: this._localeService.t("sheets-table-ui.tableStyle") },
			children: {
				label: SHEET_TABLE_THEME_PANEL,
				oldConfig: table.getTableConfig(),
				unitId,
				subUnitId,
				tableId
			},
			width: 330
		});
	}
	_getInsertRegionFromPoint(x, y) {
		const skeleton = this._sheetSkeletonManagerService.getCurrentSkeleton();
		const worksheet = this._context.unit.getActiveSheet();
		if (!skeleton || !worksheet) return null;
		const unitId = this._context.unit.getUnitId();
		const subUnitId = worksheet.getSheetId();
		const tables = this._tableManager.getTablesBySubunitId(unitId, subUnitId);
		for (const table of tables) {
			const range = table.getRange();
			const tableBounds = this._getRangeBounds(skeleton, range);
			if (x < tableBounds.left || x > tableBounds.right || y < tableBounds.top || y > tableBounds.bottom) continue;
			if (y > this._getRangeBounds(skeleton, {
				...range,
				endRow: range.startRow
			}).bottom) for (let row = range.startRow + 1; row <= range.endRow; row++) {
				const cell = skeleton.getNoMergeCellWithCoordByIndex(row, range.startColumn);
				if (y >= cell.startY && y <= cell.endY) return {
					type: "insert-row",
					tableId: table.getId(),
					index: row + 1,
					left: tableBounds.left - 22 / 2,
					top: cell.endY - 22 / 2,
					width: 22,
					height: 22
				};
			}
		}
		return null;
	}
	_getRangeBounds(skeleton, range) {
		const startCell = skeleton.getNoMergeCellWithCoordByIndex(range.startRow, range.startColumn);
		const endCell = skeleton.getNoMergeCellWithCoordByIndex(range.endRow, range.endColumn);
		return {
			left: startCell.startX,
			top: startCell.startY,
			right: endCell.endX,
			bottom: endCell.endY
		};
	}
	_syncTopTableGap(skeleton) {
		const worksheet = this._context.unit.getActiveSheet();
		if (!worksheet) return;
		const unitId = this._context.unit.getUnitId();
		const subUnitId = worksheet.getSheetId();
		const hasTopTable = this._tableManager.getTablesBySubunitId(unitId, subUnitId).some((table) => table.getRange().startRow === 0);
		const current = skeleton.gapConfig;
		const rowGaps = { ...current.rowGaps };
		const previousTopGap = rowGaps[TABLE_CONTROL_GAP_ROW] ? { ...rowGaps[TABLE_CONTROL_GAP_ROW] } : null;
		let shouldSync = false;
		if (hasTopTable) {
			var _baseGap$size;
			if (!this._topGapBaseBySkeleton.has(skeleton)) this._topGapBaseBySkeleton.set(skeleton, rowGaps[TABLE_CONTROL_GAP_ROW] ? { ...rowGaps[TABLE_CONTROL_GAP_ROW] } : null);
			const baseGap = this._topGapBaseBySkeleton.get(skeleton);
			rowGaps[TABLE_CONTROL_GAP_ROW] = {
				...baseGap !== null && baseGap !== void 0 ? baseGap : rowGaps[TABLE_CONTROL_GAP_ROW],
				size: ((_baseGap$size = baseGap === null || baseGap === void 0 ? void 0 : baseGap.size) !== null && _baseGap$size !== void 0 ? _baseGap$size : 0) + 32
			};
			shouldSync = true;
		} else if (this._topGapBaseBySkeleton.has(skeleton)) {
			const baseGap = this._topGapBaseBySkeleton.get(skeleton);
			if (baseGap) rowGaps[TABLE_CONTROL_GAP_ROW] = { ...baseGap };
			else delete rowGaps[TABLE_CONTROL_GAP_ROW];
			this._topGapBaseBySkeleton.delete(skeleton);
			shouldSync = true;
		}
		if (!shouldSync) return;
		if (isSameTopGap(previousTopGap, rowGaps[TABLE_CONTROL_GAP_ROW] ? { ...rowGaps[TABLE_CONTROL_GAP_ROW] } : null)) return;
		skeleton.setGapConfig({
			...current,
			rowGaps
		});
		this._refreshSelections();
	}
	_refreshSelections() {
		this._selectionRenderService.resetSelectionsByModelData(this._sheetsSelectionsService.getCurrentSelections());
	}
	_closeFloatingControls() {
		this._shape.setOpenedMenuTableId(null);
		this._shape.setHoveredInsertRegion(null);
		this._shape.setHoveredRegion(null);
	}
	_getLocalPoint(evt) {
		const skeleton = this._sheetSkeletonManagerService.getCurrentSkeleton();
		if (skeleton) return (0, _univerjs_sheets_ui.getTransformCoord)(evt.offsetX, evt.offsetY, this._context.scene, skeleton);
		return {
			x: evt.offsetX,
			y: evt.offsetY
		};
	}
};
SheetTableControlsRenderController = __decorate([
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector)),
	__decorateParam(2, (0, _univerjs_core.Inject)(_univerjs_sheets_ui.SheetSkeletonManagerService)),
	__decorateParam(3, _univerjs_core.ICommandService),
	__decorateParam(4, (0, _univerjs_core.Inject)(_univerjs_sheets_table.TableManager)),
	__decorateParam(5, (0, _univerjs_core.Inject)(_univerjs_sheets.SheetRangeThemeModel)),
	__decorateParam(6, (0, _univerjs_core.Inject)(_univerjs_sheets.WorkbookPermissionService)),
	__decorateParam(7, (0, _univerjs_core.Inject)(_univerjs_core.IPermissionService)),
	__decorateParam(8, (0, _univerjs_core.Inject)(_univerjs_sheets.SheetsSelectionsService)),
	__decorateParam(9, _univerjs_sheets_ui.ISheetSelectionRenderService),
	__decorateParam(10, (0, _univerjs_core.Inject)(SheetTableThemeUIController)),
	__decorateParam(11, (0, _univerjs_core.Inject)(_univerjs_core.LocaleService)),
	__decorateParam(12, _univerjs_ui.IDialogService),
	__decorateParam(13, _univerjs_ui.ISidebarService)
], SheetTableControlsRenderController);

//#endregion
//#region src/views/widgets/table-filter-button.shape.ts
const FILTER_ICON_SIZE = 16;
const FILTER_TRIGGER_HOVER_RADIUS = 4;
let SheetsTableFilterButtonShape = class SheetsTableFilterButtonShape extends _univerjs_engine_render.Shape {
	constructor(key, props, _commandService) {
		super(key, props);
		this._commandService = _commandService;
		_defineProperty(this, "_cellWidth", 0);
		_defineProperty(this, "_cellHeight", 0);
		_defineProperty(this, "_filterParams", void 0);
		_defineProperty(this, "_iconColor", "#fff");
		_defineProperty(this, "_hoverBackground", "rgba(255, 255, 255, 0.92)");
		_defineProperty(this, "_hoverIconColor", "#202124");
		_defineProperty(this, "_hovered", false);
		this.setShapeProps(props);
		this.onPointerDown$.subscribeEvent((evt) => this.onPointerDown(evt));
		this.onPointerEnter$.subscribeEvent(() => this.onPointerEnter());
		this.onPointerLeave$.subscribeEvent(() => this.onPointerLeave());
	}
	setShapeProps(props) {
		if (typeof props.cellHeight !== "undefined") this._cellHeight = props.cellHeight;
		if (typeof props.cellWidth !== "undefined") this._cellWidth = props.cellWidth;
		if (typeof props.filterParams !== "undefined") this._filterParams = props.filterParams;
		if (typeof props.iconColor !== "undefined") this._iconColor = props.iconColor;
		if (typeof props.hoverBackground !== "undefined") this._hoverBackground = props.hoverBackground;
		if (typeof props.hoverIconColor !== "undefined") this._hoverIconColor = props.hoverIconColor;
		this.transformByState({
			width: props.width,
			height: props.height
		});
	}
	_draw(ctx) {
		const cellHeight = this._cellHeight;
		const cellWidth = this._cellWidth;
		const left = 16 - cellWidth;
		const top = 16 - cellHeight;
		ctx.save();
		const cellRegion = new Path2D();
		cellRegion.rect(left, top, cellWidth, cellHeight);
		ctx.clip(cellRegion);
		if (this._hovered) {
			var _ctx$roundRect;
			ctx.save();
			ctx.fillStyle = this._hoverBackground;
			ctx.beginPath();
			(_ctx$roundRect = ctx.roundRect) === null || _ctx$roundRect === void 0 || _ctx$roundRect.call(ctx, 0, 0, 16, 16, FILTER_TRIGGER_HOVER_RADIUS);
			if (!ctx.roundRect) ctx.rect(0, 0, 16, 16);
			ctx.fill();
			ctx.restore();
		}
		this._drawChevron(ctx, this._hovered ? this._hoverIconColor : this._iconColor);
		ctx.restore();
	}
	_drawChevron(ctx, color) {
		const centerX = 16 / 2;
		const centerY = 16 / 2;
		ctx.save();
		ctx.beginPath();
		ctx.strokeStyle = color;
		ctx.lineWidth = 1.8;
		ctx.lineCap = "round";
		ctx.lineJoin = "round";
		ctx.moveTo(centerX - 4.5, centerY - 2.5);
		ctx.lineTo(centerX, 10);
		ctx.lineTo(12.5, centerY - 2.5);
		ctx.stroke();
		ctx.restore();
	}
	onPointerDown(evt) {
		if (evt.button === 2) return;
		const { row, col, unitId, subUnitId, tableId } = this._filterParams;
		if (!this._commandService.hasCommand(OpenTableFilterPanelOperation.id)) return;
		setTimeout(() => {
			const cmdParams = {
				row,
				col,
				unitId,
				subUnitId,
				tableId
			};
			this._commandService.executeCommand(OpenTableFilterPanelOperation.id, cmdParams);
		}, 200);
	}
	onPointerEnter() {
		this._hovered = true;
		this.makeDirty(true);
	}
	onPointerLeave() {
		this._hovered = false;
		this.makeDirty(true);
	}
};
SheetsTableFilterButtonShape = __decorate([__decorateParam(2, _univerjs_core.ICommandService)], SheetsTableFilterButtonShape);

//#endregion
//#region src/controllers/sheet-table-filter-button-render.controller.ts
const SHEETS_FILTER_BUTTON_Z_INDEX = 5e3;
const computeIconTop = (startY, endY, cellHeight, verticalAlign) => {
	switch (verticalAlign) {
		case _univerjs_core.VerticalAlign.TOP: return startY + 1;
		case _univerjs_core.VerticalAlign.MIDDLE: return startY + Math.max(0, (cellHeight - 16) / 2);
		case _univerjs_core.VerticalAlign.BOTTOM:
		default: return endY - 16 - 1;
	}
};
let SheetsTableFilterButtonRenderController = class SheetsTableFilterButtonRenderController extends _univerjs_core.RxDisposable {
	constructor(_context, _injector, _sheetSkeletonManagerService, _sheetInterceptorService, _tableManager, _rangeThemeModel, _commandService) {
		super();
		this._context = _context;
		this._injector = _injector;
		this._sheetSkeletonManagerService = _sheetSkeletonManagerService;
		this._sheetInterceptorService = _sheetInterceptorService;
		this._tableManager = _tableManager;
		this._rangeThemeModel = _rangeThemeModel;
		this._commandService = _commandService;
		_defineProperty(this, "_buttonRenderDisposable", null);
		_defineProperty(this, "_tableFilterButtonShapes", []);
		this._initRenderer();
		this._initCommandExecuted();
	}
	dispose() {
		super.dispose();
		this._disposeRendering();
	}
	_initRenderer() {
		const tableManager = this._tableManager;
		this._sheetSkeletonManagerService.currentSkeleton$.pipe((0, rxjs.switchMap)((skeletonParams) => {
			var _workbook$getActiveSh;
			if (!skeletonParams) return (0, rxjs.of)(null);
			const { unit: workbook, unitId } = this._context;
			const worksheetId = ((_workbook$getActiveSh = workbook.getActiveSheet()) === null || _workbook$getActiveSh === void 0 ? void 0 : _workbook$getActiveSh.getSheetId()) || "";
			const getParams = () => ({
				unitId,
				worksheetId,
				tableFilterRanges: this._tableManager.getSheetFilterRangeWithState(workbook.getUnitId(), worksheetId),
				skeleton: skeletonParams.skeleton
			});
			return (0, rxjs.merge)(tableManager.tableAdd$, tableManager.tableNameChanged$, tableManager.tableRangeChanged$, tableManager.tableThemeChanged$, tableManager.tableDelete$, tableManager.tableFilterChanged$).pipe((0, rxjs.map)(() => getParams()), (0, rxjs.startWith)(getParams()));
		}), (0, rxjs.takeUntil)(this.dispose$)).subscribe((renderParams) => {
			this._disposeRendering();
			if (!renderParams || !renderParams.tableFilterRanges) return;
			this._renderButtons(renderParams);
		});
	}
	_initCommandExecuted() {
		this.disposeWithMe(this._commandService.onCommandExecuted((command) => {
			var _workbook$getActiveSh2;
			if (command.id !== _univerjs_sheets.SetVerticalTextAlignCommand.id) return;
			const { unit: workbook, unitId } = this._context;
			const worksheetId = ((_workbook$getActiveSh2 = workbook.getActiveSheet()) === null || _workbook$getActiveSh2 === void 0 ? void 0 : _workbook$getActiveSh2.getSheetId()) || "";
			const skeleton = this._sheetSkeletonManagerService.getCurrentSkeleton();
			if (!skeleton) return;
			const renderParams = {
				unitId,
				worksheetId,
				tableFilterRanges: this._tableManager.getSheetFilterRangeWithState(workbook.getUnitId(), worksheetId),
				skeleton
			};
			this._disposeRendering();
			if (!renderParams || !renderParams.tableFilterRanges) return;
			this._renderButtons(renderParams);
		}));
	}
	_renderButtons(params) {
		const { tableFilterRanges, unitId, skeleton, worksheetId } = params;
		const { unit: workbook, scene } = this._context;
		const worksheet = workbook.getSheetBySheetId(worksheetId);
		if (!worksheet) return;
		for (const { range, states, tableId } of tableFilterRanges) {
			var _this$_rangeThemeMode, _headerStyle$cl$rgb, _headerStyle$cl, _headerStyle$bg$rgb, _headerStyle$bg;
			const { startRow, startColumn, endColumn } = range;
			const table = this._tableManager.getTableById(unitId, tableId);
			const headerStyle = table ? (_this$_rangeThemeMode = this._rangeThemeModel.getRangeThemeStyle(unitId, table.getTableStyleId())) === null || _this$_rangeThemeMode === void 0 ? void 0 : _this$_rangeThemeMode.getHeaderRowStyle() : null;
			const iconColor = (_headerStyle$cl$rgb = headerStyle === null || headerStyle === void 0 || (_headerStyle$cl = headerStyle.cl) === null || _headerStyle$cl === void 0 ? void 0 : _headerStyle$cl.rgb) !== null && _headerStyle$cl$rgb !== void 0 ? _headerStyle$cl$rgb : "#fff";
			const hoverIconColor = (_headerStyle$bg$rgb = headerStyle === null || headerStyle === void 0 || (_headerStyle$bg = headerStyle.bg) === null || _headerStyle$bg === void 0 ? void 0 : _headerStyle$bg.rgb) !== null && _headerStyle$bg$rgb !== void 0 ? _headerStyle$bg$rgb : "#202124";
			this._interceptCellContent(unitId, worksheetId, range);
			for (let col = startColumn; col <= endColumn; col++) {
				const key = `sheets-table-filter-button-${startRow}-${col}`;
				const startPosition = (0, _univerjs_sheets_ui.getCoordByCell)(startRow, col, scene, skeleton);
				const cellStyle = worksheet.getCellStyle(startRow, col);
				const verticalAlign = (cellStyle === null || cellStyle === void 0 ? void 0 : cellStyle.vt) || _univerjs_core.VerticalAlign.BOTTOM;
				const { startX, startY, endX, endY } = startPosition;
				const cellWidth = endX - startX;
				const cellHeight = endY - startY;
				if (cellHeight <= 1 || cellWidth <= 1) continue;
				const state = states[col - startColumn];
				const props = {
					left: endX - 16 - 1,
					top: computeIconTop(startY, endY, cellHeight, verticalAlign),
					height: 16,
					width: 16,
					zIndex: SHEETS_FILTER_BUTTON_Z_INDEX,
					iconColor,
					hoverBackground: iconColor,
					hoverIconColor,
					cellHeight,
					cellWidth,
					filterParams: {
						unitId,
						subUnitId: worksheetId,
						row: startRow,
						col,
						buttonState: state,
						tableId
					}
				};
				const buttonShape = this._injector.createInstance(SheetsTableFilterButtonShape, key, props);
				this._tableFilterButtonShapes.push(buttonShape);
			}
		}
		scene.addObjects(this._tableFilterButtonShapes);
		scene.makeDirty();
	}
	_interceptCellContent(workbookId, worksheetId, range) {
		const { startRow, startColumn, endColumn } = range;
		this._buttonRenderDisposable = this._sheetInterceptorService.intercept(_univerjs_sheets.INTERCEPTOR_POINT.CELL_CONTENT, {
			effect: _univerjs_core.InterceptorEffectEnum.Style,
			handler: (cell, pos, next) => {
				const { row, col, unitId, subUnitId } = pos;
				if (unitId !== workbookId || subUnitId !== worksheetId || row !== startRow || col < startColumn || col > endColumn) return next(cell);
				if (!cell || cell === pos.rawData) cell = { ...pos.rawData };
				cell.fontRenderExtension = {
					...cell === null || cell === void 0 ? void 0 : cell.fontRenderExtension,
					rightOffset: 16 + 1 + 2
				};
				return next(cell);
			},
			priority: 10
		});
	}
	_disposeRendering() {
		var _this$_buttonRenderDi;
		this._tableFilterButtonShapes.forEach((s) => s.dispose());
		(_this$_buttonRenderDi = this._buttonRenderDisposable) === null || _this$_buttonRenderDi === void 0 || _this$_buttonRenderDi.dispose();
		this._buttonRenderDisposable = null;
		this._tableFilterButtonShapes = [];
	}
};
SheetsTableFilterButtonRenderController = __decorate([
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector)),
	__decorateParam(2, (0, _univerjs_core.Inject)(_univerjs_sheets_ui.SheetSkeletonManagerService)),
	__decorateParam(3, (0, _univerjs_core.Inject)(_univerjs_sheets.SheetInterceptorService)),
	__decorateParam(4, (0, _univerjs_core.Inject)(_univerjs_sheets_table.TableManager)),
	__decorateParam(5, (0, _univerjs_core.Inject)(_univerjs_sheets.SheetRangeThemeModel)),
	__decorateParam(6, _univerjs_core.ICommandService)
], SheetsTableFilterButtonRenderController);

//#endregion
//#region src/controllers/sheet-table-render.controller.ts
let SheetsTableRenderController = class SheetsTableRenderController extends _univerjs_core.RxDisposable {
	constructor(_context, _injector, _sheetSkeletonManagerService, _tableManager, _sheetTableThemeUIController) {
		super();
		this._context = _context;
		this._injector = _injector;
		this._sheetSkeletonManagerService = _sheetSkeletonManagerService;
		this._tableManager = _tableManager;
		this._sheetTableThemeUIController = _sheetTableThemeUIController;
		this._initListener();
	}
	_dirtySkeleton() {
		var _this$_context$mainCo;
		(_this$_context$mainCo = this._context.mainComponent) === null || _this$_context$mainCo === void 0 || _this$_context$mainCo.makeDirty();
		const currentParam = this._sheetSkeletonManagerService.getCurrentParam();
		if (currentParam) {
			const param = {
				...currentParam,
				dirty: true
			};
			this._sheetSkeletonManagerService.reCalculate(param);
		}
	}
	_initListener() {
		const tableManager = this._tableManager;
		const dirtySkeleton = this._dirtySkeleton.bind(this);
		this.disposeWithMe((0, rxjs.merge)(tableManager.tableAdd$, tableManager.tableDelete$, tableManager.tableNameChanged$, tableManager.tableRangeChanged$, tableManager.tableThemeChanged$, tableManager.tableFilterChanged$, tableManager.tableInitStatus$, this._sheetTableThemeUIController.refreshTable$).subscribe(dirtySkeleton));
	}
};
SheetsTableRenderController = __decorate([
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector)),
	__decorateParam(2, (0, _univerjs_core.Inject)(_univerjs_sheets_ui.SheetSkeletonManagerService)),
	__decorateParam(3, (0, _univerjs_core.Inject)(_univerjs_sheets_table.TableManager)),
	__decorateParam(4, (0, _univerjs_core.Inject)(SheetTableThemeUIController))
], SheetsTableRenderController);

//#endregion
//#region src/controllers/sheet-table-selection.controller.ts
let SheetTableSelectionController = class SheetTableSelectionController extends _univerjs_core.Disposable {
	constructor(_sheetInterceptorService, _univerInstanceService, _tableManager) {
		super();
		this._sheetInterceptorService = _sheetInterceptorService;
		this._univerInstanceService = _univerInstanceService;
		this._tableManager = _tableManager;
		this._initSelectionChange();
	}
	_initSelectionChange() {
		this.disposeWithMe(this._sheetInterceptorService.interceptCommand({ getMutations: (command) => {
			if (command.id === _univerjs_sheets_ui.SelectAllCommand.id) {
				const target = (0, _univerjs_sheets.getSheetCommandTarget)(this._univerInstanceService);
				if (!target) return {
					redos: [],
					undos: []
				};
				const { range } = command.params;
				const { unitId, subUnitId, worksheet } = target;
				const overlapTable = this._tableManager.getTablesBySubunitId(unitId, subUnitId).find((table) => {
					const tableRange = table.getRange();
					return _univerjs_core.Rectangle.contains(tableRange, range);
				});
				if (overlapTable) {
					const tableRange = overlapTable.getRange();
					const tableRangeWithoutHeader = {
						...tableRange,
						startRow: tableRange.startRow + 1
					};
					if (_univerjs_core.Rectangle.equals(tableRange, range)) return {
						undos: [],
						redos: []
					};
					else if (_univerjs_core.Rectangle.equals(tableRangeWithoutHeader, range)) return {
						undos: [],
						redos: [{
							id: _univerjs_sheets.SetSelectionsOperation.id,
							params: {
								unitId,
								subUnitId,
								selections: [{
									range: tableRange,
									primary: (0, _univerjs_sheets.getPrimaryForRange)(tableRange, worksheet)
								}]
							}
						}]
					};
					else return {
						undos: [],
						redos: [{
							id: _univerjs_sheets.SetSelectionsOperation.id,
							params: {
								unitId,
								subUnitId,
								selections: [{
									range: tableRangeWithoutHeader,
									primary: (0, _univerjs_sheets.getPrimaryForRange)(tableRangeWithoutHeader, worksheet)
								}]
							}
						}]
					};
				}
			}
			return {
				redos: [],
				undos: []
			};
		} }));
	}
};
SheetTableSelectionController = __decorate([
	__decorateParam(0, (0, _univerjs_core.Inject)(_univerjs_sheets.SheetInterceptorService)),
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.IUniverInstanceService)),
	__decorateParam(2, (0, _univerjs_core.Inject)(_univerjs_sheets_table.TableManager))
], SheetTableSelectionController);

//#endregion
//#region src/views/components/SheetTableSelector.tsx
const SheetTableSelector = (props) => {
	const { unitId, subUnitId, range, onCancel, onConfirm, tableId } = props;
	const tableManager = (0, _univerjs_ui.useDependency)(_univerjs_sheets_table.TableManager);
	const [selectedRange, setSelectedRange] = (0, react.useState)(range);
	const [rangeError, setRangeError] = (0, react.useState)("");
	const localeService = (0, _univerjs_ui.useDependency)(_univerjs_core.LocaleService);
	const univerInstanceService = (0, _univerjs_ui.useDependency)(_univerjs_core.IUniverInstanceService);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react_jsx_runtime.Fragment, { children: [
		/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_sheets_formula_ui.RangeSelector, {
			maxRangeCount: 1,
			unitId,
			subUnitId,
			initialValue: (0, _univerjs_engine_formula.serializeRange)(range),
			onChange: (_, text) => {
				const originValue = (0, _univerjs_engine_formula.serializeRange)(range);
				const newRange = (0, _univerjs_engine_formula.deserializeRangeWithSheet)(text).range;
				const target = (0, _univerjs_sheets.getSheetCommandTarget)(univerInstanceService, {
					unitId,
					subUnitId
				});
				if (!target) return;
				if (target.worksheet.getMergeData().some((merge) => {
					return _univerjs_core.Rectangle.intersects(newRange, merge);
				})) {
					setRangeError(localeService.t("sheets-table-ui.tableRangeWithMergeError"));
					return;
				}
				if (tableManager.getTablesBySubunitId(unitId, subUnitId).some((table) => {
					if (table.getId() === tableId) return false;
					const tableRange = table.getRange();
					return _univerjs_core.Rectangle.intersects(newRange, tableRange);
				})) {
					setRangeError(localeService.t("sheets-table-ui.tableRangeWithOtherTableError"));
					return;
				}
				const { startRow, endRow } = newRange;
				if (startRow === endRow) {
					setRangeError(localeService.t("sheets-table-ui.tableRangeSingleRowError"));
					return;
				}
				if (originValue === text) return;
				if (tableId) {
					const table = tableManager.getTableById(unitId, tableId);
					if (table) {
						const oldRange = table.getRange();
						if (_univerjs_core.Rectangle.intersects(newRange, oldRange) && oldRange.startRow === newRange.startRow) {
							setSelectedRange(newRange);
							setRangeError("");
							onConfirm({
								unitId,
								subUnitId,
								range: newRange
							});
							return;
						} else {
							setRangeError(localeService.t("sheets-table-ui.updateError"));
							return;
						}
					}
				}
				setSelectedRange(newRange);
				setRangeError("");
			},
			supportAcrossSheet: false
		}),
		rangeError && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "univer-mt-1 univer-text-xs univer-text-red-500",
			children: rangeError
		}),
		/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "univer-mt-4 univer-flex univer-justify-end",
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Button, {
				onClick: onCancel,
				children: localeService.t("sheets-table-ui.cancel")
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Button, {
				variant: "primary",
				onClick: () => {
					if (rangeError) return;
					onConfirm({
						unitId,
						subUnitId,
						range: selectedRange
					});
				},
				className: "univer-ml-2",
				children: localeService.t("sheets-table-ui.confirm")
			})]
		})
	] });
};

//#endregion
//#region src/views/components/SheetTableThemePanel.tsx
const SheetTableThemePanel = (props) => {
	var _customStyle$getHeade, _customStyle$getHeade2, _customStyle$getFirst, _customStyle$getFirst2, _customStyle$getSecon, _customStyle$getSecon2, _customStyle$getLastR, _customStyle$getLastR2;
	const { unitId, subUnitId, tableId } = props;
	const commandService = (0, _univerjs_ui.useDependency)(_univerjs_core.ICommandService);
	const localeService = (0, _univerjs_ui.useDependency)(_univerjs_core.LocaleService);
	const tableManager = (0, _univerjs_ui.useDependency)(_univerjs_sheets_table.TableManager);
	const table = tableManager.getTableById(unitId, tableId);
	const rangeThemeModel = (0, _univerjs_ui.useDependency)(_univerjs_sheets.SheetRangeThemeModel);
	const sheetTableThemeUIController = (0, _univerjs_ui.useDependency)(SheetTableThemeUIController);
	const rangeThemeMapChanged = (0, _univerjs_ui.useObservable)(rangeThemeModel.rangeThemeMapChange$);
	const tableRefresh = (0, _univerjs_ui.useObservable)(sheetTableThemeUIController.refreshTable$);
	const errorService = (0, _univerjs_ui.useDependency)(_univerjs_core.ErrorService);
	const [, setRefresh] = (0, react.useState)(Math.random());
	const themeConfig = (0, _univerjs_ui.useObservable)(tableManager.tableThemeChanged$, {
		theme: table === null || table === void 0 ? void 0 : table.getTableStyleId(),
		oldTheme: table === null || table === void 0 ? void 0 : table.getTableStyleId(),
		unitId,
		subUnitId,
		tableId
	});
	const defaultRangeThemes = rangeThemeModel.getRegisteredRangeThemes().filter((item) => item === null || item === void 0 ? void 0 : item.startsWith(TABLE_DEFAULT_NAME_PREFIX));
	const customRangeThemes = rangeThemeModel.getALLRegisteredTheme(unitId).filter((item) => item === null || item === void 0 ? void 0 : item.startsWith(TABLE_CUSTOM_NAME_PREFIX));
	const selectedTheme = table === null || table === void 0 ? void 0 : table.getTableStyleId();
	const customSelected = customRangeThemes.find((item) => item === selectedTheme);
	const customStyleName = customSelected || customRangeThemes[0];
	const customStyle = rangeThemeModel.getCustomRangeThemeStyle(unitId, customStyleName);
	const headerBg = (_customStyle$getHeade = customStyle === null || customStyle === void 0 || (_customStyle$getHeade2 = customStyle.getHeaderRowStyle()) === null || _customStyle$getHeade2 === void 0 || (_customStyle$getHeade2 = _customStyle$getHeade2.bg) === null || _customStyle$getHeade2 === void 0 ? void 0 : _customStyle$getHeade2.rgb) !== null && _customStyle$getHeade !== void 0 ? _customStyle$getHeade : TABLE_DEFAULT_BG_COLOR;
	const firstRowBg = (_customStyle$getFirst = customStyle === null || customStyle === void 0 || (_customStyle$getFirst2 = customStyle.getFirstRowStyle()) === null || _customStyle$getFirst2 === void 0 || (_customStyle$getFirst2 = _customStyle$getFirst2.bg) === null || _customStyle$getFirst2 === void 0 ? void 0 : _customStyle$getFirst2.rgb) !== null && _customStyle$getFirst !== void 0 ? _customStyle$getFirst : TABLE_DEFAULT_BG_COLOR;
	const secondRowBg = (_customStyle$getSecon = customStyle === null || customStyle === void 0 || (_customStyle$getSecon2 = customStyle.getSecondRowStyle()) === null || _customStyle$getSecon2 === void 0 || (_customStyle$getSecon2 = _customStyle$getSecon2.bg) === null || _customStyle$getSecon2 === void 0 ? void 0 : _customStyle$getSecon2.rgb) !== null && _customStyle$getSecon !== void 0 ? _customStyle$getSecon : TABLE_DEFAULT_BG_COLOR;
	const lastRowBg = (_customStyle$getLastR = customStyle === null || customStyle === void 0 || (_customStyle$getLastR2 = customStyle.getLastRowStyle()) === null || _customStyle$getLastR2 === void 0 || (_customStyle$getLastR2 = _customStyle$getLastR2.bg) === null || _customStyle$getLastR2 === void 0 ? void 0 : _customStyle$getLastR2.rgb) !== null && _customStyle$getLastR !== void 0 ? _customStyle$getLastR : TABLE_DEFAULT_BG_COLOR;
	const [hoverCustomId, setHoverCustomId] = (0, react.useState)(null);
	const handleThemeChange = (theme) => {
		commandService.executeCommand(_univerjs_sheets_table.SetSheetTableCommand.id, {
			unitId,
			tableId,
			theme
		});
	};
	const handleAddCustomTheme = () => {
		if (customRangeThemes.length >= 11) {
			errorService.emit(localeService.t("sheets-table-ui.customTooMore"));
			return;
		}
		const lastCustomTheme = customRangeThemes[customRangeThemes.length - 1];
		let newThemeName = `${TABLE_CUSTOM_NAME_PREFIX}1`;
		if (lastCustomTheme) newThemeName = `${TABLE_CUSTOM_NAME_PREFIX}${Number(lastCustomTheme.split("-")[2]) + 1}`;
		const newTheme = new _univerjs_sheets.RangeThemeStyle(newThemeName, { ..._univerjs_sheets_table.customEmptyThemeWithBorderStyle });
		commandService.executeCommand(_univerjs_sheets_table.AddTableThemeCommand.id, {
			unitId,
			tableId,
			themeStyle: newTheme
		});
	};
	const setCustomTheme = (themeName, tableThemeStyle) => {
		commandService.executeCommand(_univerjs_sheets.SetRangeThemeMutation.id, {
			unitId,
			subUnitId,
			styleName: themeName,
			style: tableThemeStyle
		});
	};
	const removeCustomTheme = (themeName) => {
		commandService.executeCommand(_univerjs_sheets_table.RemoveTableThemeCommand.id, {
			unitId,
			tableId,
			themeName
		});
	};
	(0, react.useEffect)(() => {
		setRefresh(Math.random());
	}, [rangeThemeMapChanged, tableRefresh]);
	if (!table) return null;
	const headerBgIsDark = new _univerjs_core.ColorKit(headerBg).isDark();
	const firstRowBgIsDark = new _univerjs_core.ColorKit(firstRowBg).isDark();
	const secondRowBgIsDark = new _univerjs_core.ColorKit(secondRowBg).isDark();
	const lastRowBgIsDark = new _univerjs_core.ColorKit(lastRowBg).isDark();
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", { children: [
		/* @__PURE__ */ (0, react_jsx_runtime.jsx)("h5", { children: localeService.t("sheets-table-ui.defaultStyle") }),
		/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "univer-flex univer-gap-2",
			children: defaultRangeThemes.map((item) => {
				var _rangeThemeItem$getHe, _rangeThemeItem$getFi, _rangeThemeItem$getSe, _rangeThemeItem$getLa;
				const rangeThemeItem = rangeThemeModel.getDefaultRangeThemeStyle(item);
				const headerRowBg = (rangeThemeItem === null || rangeThemeItem === void 0 || (_rangeThemeItem$getHe = rangeThemeItem.getHeaderRowStyle()) === null || _rangeThemeItem$getHe === void 0 || (_rangeThemeItem$getHe = _rangeThemeItem$getHe.bg) === null || _rangeThemeItem$getHe === void 0 ? void 0 : _rangeThemeItem$getHe.rgb) || "rgb(255, 255, 255)";
				const firstRowBg = (rangeThemeItem === null || rangeThemeItem === void 0 || (_rangeThemeItem$getFi = rangeThemeItem.getFirstRowStyle()) === null || _rangeThemeItem$getFi === void 0 || (_rangeThemeItem$getFi = _rangeThemeItem$getFi.bg) === null || _rangeThemeItem$getFi === void 0 ? void 0 : _rangeThemeItem$getFi.rgb) || "rgb(255, 255, 255)";
				const secondRowBg = (rangeThemeItem === null || rangeThemeItem === void 0 || (_rangeThemeItem$getSe = rangeThemeItem.getSecondRowStyle()) === null || _rangeThemeItem$getSe === void 0 || (_rangeThemeItem$getSe = _rangeThemeItem$getSe.bg) === null || _rangeThemeItem$getSe === void 0 ? void 0 : _rangeThemeItem$getSe.rgb) || "rgb(255, 255, 255)";
				const lastRowBg = (rangeThemeItem === null || rangeThemeItem === void 0 || (_rangeThemeItem$getLa = rangeThemeItem.getLastRowStyle()) === null || _rangeThemeItem$getLa === void 0 || (_rangeThemeItem$getLa = _rangeThemeItem$getLa.bg) === null || _rangeThemeItem$getLa === void 0 ? void 0 : _rangeThemeItem$getLa.rgb) || "rgb(255, 255, 255)";
				return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
					className: (0, _univerjs_design.clsx)("univer-h-10 univer-w-8 univer-cursor-pointer univer-border univer-border-solid univer-border-gray-200 univer-p-px [&>div]:univer-box-border [&>div]:univer-h-2.5", { "univer-border-blue-500": item === themeConfig.theme }),
					onClick: () => handleThemeChange(item),
					children: [
						/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { style: {
							background: headerRowBg,
							border: `${headerRowBg ? TABLE_BORDER_NONE : TABLE_BORDER_DEFAULT}`
						} }),
						/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { style: {
							background: firstRowBg,
							border: `${firstRowBg ? TABLE_BORDER_NONE : TABLE_BORDER_DEFAULT}`
						} }),
						/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { style: {
							background: secondRowBg,
							border: `${secondRowBg ? TABLE_BORDER_NONE : TABLE_BORDER_DEFAULT}`
						} }),
						/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { style: {
							background: lastRowBg,
							border: `${lastRowBg ? TABLE_BORDER_NONE : TABLE_BORDER_DEFAULT}`
						} })
					]
				}, item);
			})
		}),
		/* @__PURE__ */ (0, react_jsx_runtime.jsx)("h5", { children: localeService.t("sheets-table-ui.customStyle") }),
		/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: (0, _univerjs_design.clsx)("univer-w-full univer-rounded-sm", _univerjs_design.borderClassName),
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: "univer-flex univer-flex-wrap univer-gap-2 univer-p-2",
				children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
					className: (0, _univerjs_design.clsx)("univer-h-10 univer-w-8 univer-cursor-pointer univer-p-px univer-text-center univer-leading-10", _univerjs_design.borderClassName),
					onClick: handleAddCustomTheme,
					children: "+"
				}), customRangeThemes.map((item) => {
					var _rangeThemeItem$getHe2, _rangeThemeItem$getFi2, _rangeThemeItem$getSe2, _rangeThemeItem$getLa2;
					const rangeThemeItem = rangeThemeModel.getCustomRangeThemeStyle(unitId, item);
					const headerRowBg = rangeThemeItem === null || rangeThemeItem === void 0 || (_rangeThemeItem$getHe2 = rangeThemeItem.getHeaderRowStyle()) === null || _rangeThemeItem$getHe2 === void 0 || (_rangeThemeItem$getHe2 = _rangeThemeItem$getHe2.bg) === null || _rangeThemeItem$getHe2 === void 0 ? void 0 : _rangeThemeItem$getHe2.rgb;
					const firstRowBg = rangeThemeItem === null || rangeThemeItem === void 0 || (_rangeThemeItem$getFi2 = rangeThemeItem.getFirstRowStyle()) === null || _rangeThemeItem$getFi2 === void 0 || (_rangeThemeItem$getFi2 = _rangeThemeItem$getFi2.bg) === null || _rangeThemeItem$getFi2 === void 0 ? void 0 : _rangeThemeItem$getFi2.rgb;
					const secondRowBg = rangeThemeItem === null || rangeThemeItem === void 0 || (_rangeThemeItem$getSe2 = rangeThemeItem.getSecondRowStyle()) === null || _rangeThemeItem$getSe2 === void 0 || (_rangeThemeItem$getSe2 = _rangeThemeItem$getSe2.bg) === null || _rangeThemeItem$getSe2 === void 0 ? void 0 : _rangeThemeItem$getSe2.rgb;
					const lastRowBg = rangeThemeItem === null || rangeThemeItem === void 0 || (_rangeThemeItem$getLa2 = rangeThemeItem.getLastRowStyle()) === null || _rangeThemeItem$getLa2 === void 0 || (_rangeThemeItem$getLa2 = _rangeThemeItem$getLa2.bg) === null || _rangeThemeItem$getLa2 === void 0 ? void 0 : _rangeThemeItem$getLa2.rgb;
					return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
						className: (0, _univerjs_design.clsx)("univer-relative univer-h-10 univer-w-8 univer-cursor-pointer univer-border univer-border-solid univer-border-gray-200 univer-p-px", { "univer-border-blue-500": item === themeConfig.theme }),
						onClick: () => handleThemeChange(item),
						onMouseEnter: () => setHoverCustomId(item),
						onMouseLeave: () => setHoverCustomId(null),
						children: [
							/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
								className: "univer-box-border univer-h-2.5",
								style: {
									background: headerRowBg !== null && headerRowBg !== void 0 ? headerRowBg : TABLE_BORDER_NONE,
									border: `${headerRowBg ? TABLE_BORDER_NONE : TABLE_BORDER_DEFAULT}`
								}
							}),
							/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
								className: "univer-box-border univer-h-2.5",
								style: {
									background: firstRowBg !== null && firstRowBg !== void 0 ? firstRowBg : TABLE_BORDER_NONE,
									border: `${firstRowBg ? TABLE_BORDER_NONE : TABLE_BORDER_DEFAULT}`
								}
							}),
							/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
								className: "univer-box-border univer-h-2.5",
								style: {
									background: secondRowBg !== null && secondRowBg !== void 0 ? secondRowBg : TABLE_BORDER_NONE,
									border: `${secondRowBg ? TABLE_BORDER_NONE : TABLE_BORDER_DEFAULT}`
								}
							}),
							/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
								className: "univer-box-border univer-h-2.5",
								style: {
									background: lastRowBg !== null && lastRowBg !== void 0 ? lastRowBg : TABLE_BORDER_NONE,
									border: `${lastRowBg ? TABLE_BORDER_NONE : TABLE_BORDER_DEFAULT}`
								}
							}),
							/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
								className: "univer-absolute univer-right-[-3px] univer-top-[-3px] univer-size-3 univer-rounded-md univer-bg-gray-200 univer-text-center univer-text-xs univer-leading-[10px]",
								style: { display: hoverCustomId === item ? "block" : "none" },
								onClick: (e) => {
									e.stopPropagation();
									removeCustomTheme(item);
								},
								children: "x"
							})
						]
					}, item);
				})]
			}), customSelected && /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react_jsx_runtime.Fragment, { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { className: "univer-h-px univer-w-full univer-bg-gray-200" }), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: "univer-flex univer-flex-col univer-gap-2 univer-p-2",
				children: [
					/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
						className: "univer-flex univer-h-9 univer-gap-2",
						children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
							className: (0, _univerjs_design.clsx)("univer-box-border univer-h-full univer-w-52 univer-rounded-sm univer-text-center univer-leading-9", _univerjs_design.borderClassName, {
								"univer-text-white": headerBgIsDark,
								"univer-text-gray-900": !headerBgIsDark
							}),
							style: { background: headerBg },
							children: localeService.t("sheets-table-ui.header")
						}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Dropdown, {
							overlay: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
								className: "univer-p-2",
								children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.ColorPicker, {
									value: headerBg,
									onChange: (val) => {
										const headerRowStyle = (0, _univerjs_sheets_table.processStyleWithBorderStyle)("headerRowStyle", {
											bg: { rgb: val },
											cl: { rgb: new _univerjs_core.ColorKit(val).isDark() ? "#fff" : "#000" }
										});
										setCustomTheme(table.getTableStyleId(), { headerRowStyle });
									}
								})
							}),
							children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
								className: (0, _univerjs_design.clsx)("univer-flex univer-cursor-pointer univer-items-center univer-gap-2 univer-rounded-sm univer-bg-white univer-p-1", _univerjs_design.borderClassName),
								children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
									className: (0, _univerjs_design.clsx)("univer-size-4 univer-rounded-lg univer-bg-gray-400", _univerjs_design.borderClassName, {
										"univer-text-white": headerBgIsDark,
										"univer-text-gray-900": !headerBgIsDark
									}),
									style: { background: headerBg }
								}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_icons.DropdownIcon, { className: "univer-size-2" })]
							})
						})]
					}),
					/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
						className: "univer-flex univer-h-9 univer-gap-2",
						children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
							className: (0, _univerjs_design.clsx)("univer-box-border univer-h-full univer-w-52 univer-rounded-sm univer-text-center univer-leading-9", _univerjs_design.borderClassName, {
								"univer-text-white": firstRowBgIsDark,
								"univer-text-gray-900": !firstRowBgIsDark
							}),
							style: { background: firstRowBg },
							children: localeService.t("sheets-table-ui.firstLine")
						}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Dropdown, {
							overlay: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
								className: "univer-p-2",
								children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.ColorPicker, {
									value: firstRowBg,
									onChange: (val) => {
										setCustomTheme(table.getTableStyleId(), { firstRowStyle: {
											bg: { rgb: val },
											cl: { rgb: new _univerjs_core.ColorKit(val).isDark() ? "#fff" : "#000" }
										} });
									}
								})
							}),
							children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
								className: (0, _univerjs_design.clsx)("univer-flex univer-cursor-pointer univer-items-center univer-gap-2 univer-rounded-sm univer-bg-white univer-p-1", _univerjs_design.borderClassName),
								children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
									className: (0, _univerjs_design.clsx)("univer-size-4 univer-rounded-lg univer-bg-gray-400", _univerjs_design.borderClassName),
									style: { background: firstRowBg }
								}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_icons.DropdownIcon, { className: "univer-size-2" })]
							})
						})]
					}),
					/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
						className: "univer-flex univer-h-9 univer-gap-2",
						children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
							className: (0, _univerjs_design.clsx)("univer-box-border univer-h-full univer-w-52 univer-rounded-sm univer-text-center univer-leading-9", _univerjs_design.borderClassName, {
								"univer-text-white": secondRowBgIsDark,
								"univer-text-gray-900": !secondRowBgIsDark
							}),
							style: { background: secondRowBg },
							children: localeService.t("sheets-table-ui.secondLine")
						}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Dropdown, {
							overlay: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
								className: "univer-p-2",
								children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.ColorPicker, {
									value: secondRowBg,
									onChange: (val) => setCustomTheme(table.getTableStyleId(), { secondRowStyle: {
										bg: { rgb: val },
										cl: { rgb: new _univerjs_core.ColorKit(val).isDark() ? "#fff" : "#000" }
									} })
								})
							}),
							children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
								className: (0, _univerjs_design.clsx)("univer-flex univer-cursor-pointer univer-items-center univer-gap-2 univer-rounded-sm univer-bg-white univer-p-1", _univerjs_design.borderClassName),
								children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
									className: (0, _univerjs_design.clsx)("univer-size-4 univer-rounded-lg univer-bg-gray-400", _univerjs_design.borderClassName),
									style: { background: secondRowBg }
								}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_icons.DropdownIcon, { className: "univer-size-2" })]
							})
						})]
					}),
					/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
						className: "univer-flex univer-h-9 univer-gap-2",
						children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
							className: (0, _univerjs_design.clsx)("univer-box-border univer-h-full univer-w-52 univer-rounded-sm univer-text-center univer-leading-9", _univerjs_design.borderClassName, {
								"univer-text-white": lastRowBgIsDark,
								"univer-text-gray-900": !lastRowBgIsDark
							}),
							style: { background: lastRowBg },
							children: localeService.t("sheets-table-ui.footer")
						}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.Dropdown, {
							overlay: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
								className: "univer-p-2",
								children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_design.ColorPicker, {
									value: lastRowBg,
									onChange: (val) => {
										const lastRowStyle = (0, _univerjs_sheets_table.processStyleWithBorderStyle)("lastRowStyle", {
											bg: { rgb: val },
											cl: { rgb: new _univerjs_core.ColorKit(val).isDark() ? "#fff" : "#000" }
										});
										setCustomTheme(table.getTableStyleId(), { lastRowStyle });
									}
								})
							}),
							children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
								className: (0, _univerjs_design.clsx)("univer-flex univer-cursor-pointer univer-items-center univer-gap-2 univer-rounded-sm univer-bg-white univer-p-1", _univerjs_design.borderClassName),
								children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
									className: (0, _univerjs_design.clsx)("univer-size-4 univer-rounded-lg univer-bg-gray-400", _univerjs_design.borderClassName),
									style: { background: lastRowBg }
								}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_univerjs_icons.DropdownIcon, { className: "univer-size-2" })]
							})
						})]
					})
				]
			})] })]
		})
	] });
};

//#endregion
//#region src/menu/menu.ts
const SHEET_TABLE_CONTEXT_INSERT_MENU_ID = "sheet.table.context-insert_menu-id";
const SHEET_TABLE_CONTEXT_REMOVE_MENU_ID = "sheet.table.context-remove_menu-id";
function sheetTableToolbarInsertMenuFactory(accessor) {
	return {
		id: OpenTableSelectorOperation.id,
		type: _univerjs_ui.MenuItemType.BUTTON,
		icon: TABLE_TOOLBAR_BUTTON,
		tooltip: "sheets-table-ui.title",
		title: "sheets-table-ui.title",
		hidden$: (0, _univerjs_ui.getMenuHiddenObservable)(accessor, _univerjs_core.UniverInstanceType.UNIVER_SHEET),
		disabled$: (0, _univerjs_sheets_ui.getCurrentRangeDisable$)(accessor, {}, true)
	};
}
function SheetTableInsertContextMenuFactory(accessor) {
	return {
		id: SHEET_TABLE_CONTEXT_INSERT_MENU_ID,
		type: _univerjs_ui.MenuItemType.SUBITEMS,
		icon: "InsertDoubleIcon",
		title: "sheets-table-ui.insert.main",
		hidden$: getSheetTableRowColOperationHidden$(accessor)
	};
}
function SheetTableRemoveContextMenuFactory(accessor) {
	return {
		id: SHEET_TABLE_CONTEXT_REMOVE_MENU_ID,
		type: _univerjs_ui.MenuItemType.SUBITEMS,
		icon: "ReduceDoubleIcon",
		title: "sheets-table-ui.remove.main",
		hidden$: getSheetTableRowColOperationHidden$(accessor)
	};
}
function SheetTableInsertRowMenuFactory(accessor) {
	return {
		id: _univerjs_sheets_table.SheetTableInsertRowCommand.id,
		type: _univerjs_ui.MenuItemType.BUTTON,
		title: "sheets-table-ui.insert.row",
		hidden$: getSheetTableHeaderOperationHidden$(accessor)
	};
}
function SheetTableInsertColMenuFactory(accessor) {
	return {
		id: _univerjs_sheets_table.SheetTableInsertColCommand.id,
		title: "sheets-table-ui.insert.col",
		type: _univerjs_ui.MenuItemType.BUTTON
	};
}
function SheetTableRemoveRowMenuFactory(accessor) {
	return {
		id: _univerjs_sheets_table.SheetTableRemoveRowCommand.id,
		type: _univerjs_ui.MenuItemType.BUTTON,
		title: "sheets-table-ui.remove.row",
		hidden$: getSheetTableHeaderOperationHidden$(accessor)
	};
}
function SheetTableRemoveColMenuFactory(accessor) {
	return {
		id: _univerjs_sheets_table.SheetTableRemoveColCommand.id,
		title: "sheets-table-ui.remove.col",
		type: _univerjs_ui.MenuItemType.BUTTON
	};
}
function getSheetTableRowColOperationHidden$(accessor) {
	const sheetsSelectionsService = accessor.get(_univerjs_sheets.SheetsSelectionsService);
	return accessor.get(_univerjs_core.IUniverInstanceService).getCurrentTypeOfUnit$(_univerjs_core.UniverInstanceType.UNIVER_SHEET).pipe((0, rxjs.switchMap)((workbook) => {
		if (!workbook) return (0, rxjs.of)(true);
		return workbook.activeSheet$.pipe((0, rxjs.switchMap)((sheet) => {
			if (!sheet) return (0, rxjs.of)(true);
			return sheetsSelectionsService.selectionMoveEnd$.pipe((0, rxjs.switchMap)((selections) => {
				if (!selections.length || selections.length > 1) return (0, rxjs.of)(true);
				const range = selections[0].range;
				return (0, rxjs.of)(!accessor.get(_univerjs_sheets_table.SheetsTableController).getContainerTableWithRange(workbook.getUnitId(), sheet.getSheetId(), range));
			}));
		}));
	}));
}
function getSheetTableHeaderOperationHidden$(accessor) {
	const sheetsSelectionsService = accessor.get(_univerjs_sheets.SheetsSelectionsService);
	return accessor.get(_univerjs_core.IUniverInstanceService).getCurrentTypeOfUnit$(_univerjs_core.UniverInstanceType.UNIVER_SHEET).pipe((0, rxjs.switchMap)((workbook) => {
		if (!workbook) return (0, rxjs.of)(true);
		return workbook.activeSheet$.pipe((0, rxjs.switchMap)((sheet) => {
			if (!sheet) return (0, rxjs.of)(true);
			return sheetsSelectionsService.selectionMoveEnd$.pipe((0, rxjs.switchMap)((selections) => {
				if (!selections.length || selections.length > 1) return (0, rxjs.of)(true);
				const range = selections[0].range;
				const isInTable = accessor.get(_univerjs_sheets_table.SheetsTableController).getContainerTableWithRange(workbook.getUnitId(), sheet.getSheetId(), range);
				if (!isInTable) return (0, rxjs.of)(true);
				const tableRange = isInTable.getRange();
				if (range.startRow === tableRange.startRow) return (0, rxjs.of)(true);
				return (0, rxjs.of)(false);
			}));
		}));
	}));
}

//#endregion
//#region src/menu/schema.ts
const menuSchema = {
	[_univerjs_ui.RibbonDataGroup.ORGANIZATION]: { [OpenTableSelectorOperation.id]: {
		order: 0,
		menuItemFactory: sheetTableToolbarInsertMenuFactory
	} },
	[_univerjs_ui.ContextMenuPosition.MAIN_AREA]: { [_univerjs_ui.ContextMenuGroup.LAYOUT]: {
		[SHEET_TABLE_CONTEXT_INSERT_MENU_ID]: {
			order: 5,
			menuItemFactory: SheetTableInsertContextMenuFactory,
			[_univerjs_sheets_table.SheetTableInsertRowCommand.id]: {
				order: 1,
				menuItemFactory: SheetTableInsertRowMenuFactory
			},
			[_univerjs_sheets_table.SheetTableInsertColCommand.id]: {
				order: 2,
				menuItemFactory: SheetTableInsertColMenuFactory
			}
		},
		[SHEET_TABLE_CONTEXT_REMOVE_MENU_ID]: {
			order: 6,
			menuItemFactory: SheetTableRemoveContextMenuFactory,
			[_univerjs_sheets_table.SheetTableRemoveRowCommand.id]: {
				order: 1,
				menuItemFactory: SheetTableRemoveRowMenuFactory
			},
			[_univerjs_sheets_table.SheetTableRemoveColCommand.id]: {
				order: 2,
				menuItemFactory: SheetTableRemoveColMenuFactory
			}
		}
	} }
};

//#endregion
//#region src/menu/sheet-table-menu.controller.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
let SheetTableMenuController = class SheetTableMenuController extends _univerjs_core.Disposable {
	constructor(_componentManager, _menuManagerService) {
		super();
		this._componentManager = _componentManager;
		this._menuManagerService = _menuManagerService;
		this._initComponents();
		this._initMenu();
	}
	_initComponents() {
		[
			[TABLE_TOOLBAR_BUTTON, _univerjs_icons.TableIcon],
			[TABLE_SELECTOR_DIALOG, SheetTableSelector],
			[SHEET_TABLE_THEME_PANEL, SheetTableThemePanel]
		].forEach(([key, comp]) => {
			this.disposeWithMe(this._componentManager.register(key, comp));
		});
	}
	_initMenu() {
		this._menuManagerService.mergeMenu(menuSchema);
	}
};
SheetTableMenuController = __decorate([__decorateParam(0, (0, _univerjs_core.Inject)(_univerjs_ui.ComponentManager)), __decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_ui.IMenuManagerService))], SheetTableMenuController);

//#endregion
//#region src/plugin.ts
let UniverSheetsTableUIPlugin = class UniverSheetsTableUIPlugin extends _univerjs_core.Plugin {
	constructor(_config = defaultPluginConfig, _injector, _commandService, _configService, _renderManagerService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._commandService = _commandService;
		this._configService = _configService;
		this._renderManagerService = _renderManagerService;
		const { menu, ...rest } = (0, _univerjs_core.merge)({}, defaultPluginConfig, this._config);
		if (menu) this._configService.setConfig("menu", menu, { merge: true });
		this._configService.setConfig(SHEETS_TABLE_UI_PLUGIN_CONFIG_KEY, rest);
		this._initRegisterCommand();
	}
	onStarting() {
		(0, _univerjs_core.registerDependencies)(this._injector, [
			[SheetsTableComponentController],
			[SheetsTableUiService],
			[SheetTableMenuController],
			[SheetTableThemeUIController],
			[SheetTableSelectionController]
		]);
	}
	onReady() {
		(0, _univerjs_core.touchDependencies)(this._injector, [
			[SheetsTableComponentController],
			[SheetsTableUiService],
			[SheetTableMenuController],
			[SheetTableThemeUIController],
			[SheetTableSelectionController]
		]);
	}
	onRendered() {
		this._registerRenderModules();
	}
	_registerRenderModules() {
		const renderDependencies = [];
		if (this._config.hideAnchor !== true) renderDependencies.push([SheetTableControlsRenderController]);
		renderDependencies.push([SheetsTableFilterButtonRenderController], [SheetsTableRenderController]);
		renderDependencies.forEach((m) => {
			this.disposeWithMe(this._renderManagerService.registerRenderModule(_univerjs_core.UniverInstanceType.UNIVER_SHEET, m));
		});
	}
	_initRegisterCommand() {
		[OpenTableFilterPanelOperation, OpenTableSelectorOperation].forEach((m) => this._commandService.registerCommand(m));
	}
};
_defineProperty(UniverSheetsTableUIPlugin, "pluginName", PLUGIN_NAME);
_defineProperty(UniverSheetsTableUIPlugin, "packageName", name);
_defineProperty(UniverSheetsTableUIPlugin, "version", version);
_defineProperty(UniverSheetsTableUIPlugin, "type", _univerjs_core.UniverInstanceType.UNIVER_SHEET);
UniverSheetsTableUIPlugin = __decorate([
	(0, _univerjs_core.DependentOn)(_univerjs_sheets_table.UniverSheetsTablePlugin),
	__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector)),
	__decorateParam(2, (0, _univerjs_core.Inject)(_univerjs_core.ICommandService)),
	__decorateParam(3, _univerjs_core.IConfigService),
	__decorateParam(4, _univerjs_engine_render.IRenderManagerService)
], UniverSheetsTableUIPlugin);

//#endregion
Object.defineProperty(exports, 'UniverSheetsTableUIPlugin', {
  enumerable: true,
  get: function () {
    return UniverSheetsTableUIPlugin;
  }
});
