// @univerjs/sheets-filter/lib/es/facade.js
import { ICommandService, Inject, Injector } from "@univerjs/core";
import { FEnum, FEventName, FUniver } from "@univerjs/core/facade";
import { ClearSheetsFilterCriteriaCommand, CustomFilterOperator, RemoveSheetFilterCommand, SetSheetFilterRangeCommand, SetSheetsFilterCriteriaCommand, SheetsFilterService } from "@univerjs/sheets-filter";
import { FRange, FWorksheet } from "@univerjs/sheets/facade";

//#region src/facade/f-univer.ts
var FUniverSheetsFilterMixin = class extends FUniver {
	/**
	* @ignore
	*/
	_initialize(injector) {
		const commandService = injector.get(ICommandService);
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetBeforeRangeFilter, () => commandService.beforeCommandExecuted((commandInfo) => {
			if (commandInfo.id === SetSheetsFilterCriteriaCommand.id) this._beforeRangeFilter(commandInfo);
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetBeforeRangeFilterClear, () => commandService.beforeCommandExecuted((commandInfo) => {
			if (commandInfo.id === ClearSheetsFilterCriteriaCommand.id) this._beforeRangeFilterClear();
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetRangeFiltered, () => commandService.onCommandExecuted((commandInfo) => {
			if (commandInfo.id === SetSheetsFilterCriteriaCommand.id) this._onRangeFiltered(commandInfo);
		})));
		this.disposeWithMe(this.registerEventHandler(this.Event.SheetRangeFilterCleared, () => commandService.onCommandExecuted((commandInfo) => {
			if (commandInfo.id === ClearSheetsFilterCriteriaCommand.id) this._onRangeFilterCleared();
		})));
	}
	_beforeRangeFilter(commandInfo) {
		const params = commandInfo.params;
		const fWorkbook = this.getUniverSheet(params.unitId);
		const eventParams = {
			workbook: fWorkbook,
			worksheet: fWorkbook.getSheetBySheetId(params.subUnitId),
			col: params.col,
			criteria: params.criteria
		};
		this.fireEvent(this.Event.SheetBeforeRangeFilter, eventParams);
		if (eventParams.cancel) throw new Error("SetSheetsFilterCriteriaCommand canceled.");
	}
	_onRangeFiltered(commandInfo) {
		const params = commandInfo.params;
		const fWorkbook = this.getUniverSheet(params.unitId);
		const eventParams = {
			workbook: fWorkbook,
			worksheet: fWorkbook.getSheetBySheetId(params.subUnitId),
			col: params.col,
			criteria: params.criteria
		};
		this.fireEvent(this.Event.SheetRangeFiltered, eventParams);
	}
	_beforeRangeFilterClear() {
		const fWorkbook = this.getActiveWorkbook();
		if (!fWorkbook) return;
		const eventParams = {
			workbook: fWorkbook,
			worksheet: fWorkbook.getActiveSheet()
		};
		this.fireEvent(this.Event.SheetBeforeRangeFilterClear, eventParams);
		if (eventParams.cancel) throw new Error("SetSheetsFilterCriteriaCommand canceled.");
	}
	_onRangeFilterCleared() {
		const fWorkbook = this.getActiveWorkbook();
		if (!fWorkbook) return;
		const eventParams = {
			workbook: fWorkbook,
			worksheet: fWorkbook.getActiveSheet()
		};
		this.fireEvent(this.Event.SheetRangeFilterCleared, eventParams);
	}
};
FUniver.extend(FUniverSheetsFilterMixin);

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/facade/f-filter.ts
let FFilter = class FFilter {
	constructor(_workbook, _worksheet, _filterModel, _injector, _commandSrv) {
		this._workbook = _workbook;
		this._worksheet = _worksheet;
		this._filterModel = _filterModel;
		this._injector = _injector;
		this._commandSrv = _commandSrv;
	}
	/**
	* Get the filtered out rows by this filter.
	* @returns {number[]} Filtered out rows by this filter.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set some values of the range C1:F10
	* const fRange = fWorksheet.getRange('C1:F10');
	* fRange.setValues([
	*   [1, 2, 3, 4],
	*   [2, 3, 4, 5],
	*   [3, 4, 5, 6],
	*   [4, 5, 6, 7],
	*   [5, 6, 7, 8],
	*   [6, 7, 8, 9],
	*   [7, 8, 9, 10],
	*   [8, 9, 10, 11],
	*   [9, 10, 11, 12],
	*   [10, 11, 12, 13],
	* ]);
	*
	* // Create a filter on the range C1:F10
	* let fFilter = fRange.createFilter();
	*
	* // If the filter already exists, remove it and create a new one
	* if (!fFilter) {
	*   fRange.getFilter().remove();
	*   fFilter = fRange.createFilter();
	* }
	*
	* // Set the filter criteria of the column C, filter out the rows that are not 1, 5, 9
	* const column = fWorksheet.getRange('C:C').getColumn();
	* fFilter.setColumnFilterCriteria(column, {
	*   colId: 0,
	*   filters: {
	*     filters: ['1', '5', '9'],
	*   },
	* });
	*
	* // Get the filtered out rows
	* console.log(fFilter.getFilteredOutRows()); // [1, 2, 3, 5, 6, 7, 9]
	* ```
	*/
	getFilteredOutRows() {
		return Array.from(this._filterModel.filteredOutRows).sort();
	}
	/**
	* Get the filter criteria of a column.
	* @param {number} column - The column index.
	* @returns {Nullable<IFilterColumn>} The filter criteria of the column.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set some values of the range C1:F10
	* const fRange = fWorksheet.getRange('C1:F10');
	* fRange.setValues([
	*   [1, 2, 3, 4],
	*   [2, 3, 4, 5],
	*   [3, 4, 5, 6],
	*   [4, 5, 6, 7],
	*   [5, 6, 7, 8],
	*   [6, 7, 8, 9],
	*   [7, 8, 9, 10],
	*   [8, 9, 10, 11],
	*   [9, 10, 11, 12],
	*   [10, 11, 12, 13],
	* ]);
	*
	* // Create a filter on the range C1:F10
	* let fFilter = fRange.createFilter();
	*
	* // If the filter already exists, remove it and create a new one
	* if (!fFilter) {
	*   fRange.getFilter().remove();
	*   fFilter = fRange.createFilter();
	* }
	*
	* // Set the filter criteria of the column C, filter out the rows that are not 1, 5, 9
	* const column = fWorksheet.getRange('C:C').getColumn();
	* fFilter.setColumnFilterCriteria(column, {
	*   colId: 0,
	*   filters: {
	*     filters: ['1', '5', '9'],
	*   },
	* });
	*
	* // Print the filter criteria of the column C and D
	* console.log(fFilter.getColumnFilterCriteria(column)); // { colId: 0, filters: { filters: ['1', '5', '9'] } }
	* console.log(fFilter.getColumnFilterCriteria(column + 1)); // undefined
	* ```
	*/
	getColumnFilterCriteria(column) {
		var _this$_filterModel$ge;
		return (_this$_filterModel$ge = this._filterModel.getFilterColumn(column)) === null || _this$_filterModel$ge === void 0 ? void 0 : _this$_filterModel$ge.getColumnData();
	}
	/**
	* Clear the filter criteria of a column.
	* @param {number} column - The column index.
	* @returns {FFilter} The FFilter instance for chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set some values of the range C1:F10
	* const fRange = fWorksheet.getRange('C1:F10');
	* fRange.setValues([
	*   [1, 2, 3, 4],
	*   [2, 3, 4, 5],
	*   [3, 4, 5, 6],
	*   [4, 5, 6, 7],
	*   [5, 6, 7, 8],
	*   [6, 7, 8, 9],
	*   [7, 8, 9, 10],
	*   [8, 9, 10, 11],
	*   [9, 10, 11, 12],
	*   [10, 11, 12, 13],
	* ]);
	*
	* // Create a filter on the range C1:F10
	* let fFilter = fRange.createFilter();
	*
	* // If the filter already exists, remove it and create a new one
	* if (!fFilter) {
	*   fRange.getFilter().remove();
	*   fFilter = fRange.createFilter();
	* }
	*
	* // Set the filter criteria of the column C, filter out the rows that are not 1, 5, 9
	* const column = fWorksheet.getRange('C:C').getColumn();
	* fFilter.setColumnFilterCriteria(column, {
	*   colId: 0,
	*   filters: {
	*     filters: ['1', '5', '9'],
	*   },
	* });
	*
	* // Clear the filter criteria of the column C after 3 seconds
	* setTimeout(() => {
	*   fFilter.removeColumnFilterCriteria(column);
	* }, 3000);
	* ```
	*/
	removeColumnFilterCriteria(column) {
		this._commandSrv.syncExecuteCommand(SetSheetsFilterCriteriaCommand.id, {
			unitId: this._workbook.getUnitId(),
			subUnitId: this._worksheet.getSheetId(),
			col: column,
			criteria: null
		});
		return this;
	}
	/**
	* Set the filter criteria of a column.
	* @param {number} column - The column index.
	* @param {ISetSheetsFilterCriteriaCommandParams['criteria']} criteria - The new filter criteria.
	* @returns {FFilter} The FFilter instance for chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set some values of the range C1:F10
	* const fRange = fWorksheet.getRange('C1:F10');
	* fRange.setValues([
	*   [1, 2, 3, 4],
	*   [2, 3, 4, 5],
	*   [3, 4, 5, 6],
	*   [4, 5, 6, 7],
	*   [5, 6, 7, 8],
	*   [6, 7, 8, 9],
	*   [7, 8, 9, 10],
	*   [8, 9, 10, 11],
	*   [9, 10, 11, 12],
	*   [10, 11, 12, 13],
	* ]);
	*
	* // Create a filter on the range C1:F10
	* let fFilter = fRange.createFilter();
	*
	* // If the filter already exists, remove it and create a new one
	* if (!fFilter) {
	*   fRange.getFilter().remove();
	*   fFilter = fRange.createFilter();
	* }
	*
	* // Set the filter criteria of the column C, filter out the rows that are not 1, 5, 9
	* const column = fWorksheet.getRange('C:C').getColumn();
	* fFilter.setColumnFilterCriteria(column, {
	*   colId: 0,
	*   filters: {
	*     filters: ['1', '5', '9'],
	*   },
	* });
	* ```
	*/
	setColumnFilterCriteria(column, criteria) {
		this._commandSrv.syncExecuteCommand(SetSheetsFilterCriteriaCommand.id, {
			unitId: this._workbook.getUnitId(),
			subUnitId: this._worksheet.getSheetId(),
			col: column,
			criteria
		});
		return this;
	}
	/**
	* Get the range of the filter.
	* @returns {FRange} The range of the filter.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const fFilter = fWorksheet.getFilter();
	* console.log(fFilter?.getRange().getA1Notation());
	* ```
	*/
	getRange() {
		const range = this._filterModel.getRange();
		return this._injector.createInstance(FRange, this._workbook, this._worksheet, range);
	}
	/**
	* Remove the filter criteria of all columns.
	* @returns {FFilter} The FFilter instance for chaining.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	*
	* // Set some values of the range C1:F10
	* const fRange = fWorksheet.getRange('C1:F10');
	* fRange.setValues([
	*   [1, 2, 3, 4],
	*   [2, 3, 4, 5],
	*   [3, 4, 5, 6],
	*   [4, 5, 6, 7],
	*   [5, 6, 7, 8],
	*   [6, 7, 8, 9],
	*   [7, 8, 9, 10],
	*   [8, 9, 10, 11],
	*   [9, 10, 11, 12],
	*   [10, 11, 12, 13],
	* ]);
	*
	* // Create a filter on the range C1:F10
	* let fFilter = fRange.createFilter();
	*
	* // If the filter already exists, remove it and create a new one
	* if (!fFilter) {
	*   fRange.getFilter().remove();
	*   fFilter = fRange.createFilter();
	* }
	*
	* // Set the filter criteria of the column C, filter out the rows that are not 1, 5, 9
	* const column = fWorksheet.getRange('C:C').getColumn();
	* fFilter.setColumnFilterCriteria(column, {
	*   colId: 0,
	*   filters: {
	*     filters: ['1', '5', '9'],
	*   },
	* });
	*
	* // Clear the filter criteria of all columns after 3 seconds
	* setTimeout(() => {
	*   fFilter.removeFilterCriteria();
	* }, 3000);
	* ```
	*/
	removeFilterCriteria() {
		this._commandSrv.syncExecuteCommand(ClearSheetsFilterCriteriaCommand.id);
		return this;
	}
	/**
	* Remove the filter from the worksheet.
	* @returns {boolean} True if the filter is removed successfully; otherwise, false.
	* @example
	* ```typescript
	* const fWorkbook = univerAPI.getActiveWorkbook();
	* const fWorksheet = fWorkbook.getSheetByName('Sheet1');
	* if (!fWorksheet) return;
	* const fRange = fWorksheet.getRange('A1:D14');
	* let fFilter = fRange.createFilter();
	*
	* // If the worksheet already has a filter, remove it and create a new filter.
	* if (!fFilter) {
	*   fWorksheet.getFilter().remove();
	*   fFilter = fRange.createFilter();
	* }
	* console.log(fFilter);
	* ```
	*/
	remove() {
		return this._commandSrv.syncExecuteCommand(RemoveSheetFilterCommand.id, {
			unitId: this._workbook.getUnitId(),
			subUnitId: this._worksheet.getSheetId()
		});
	}
};
FFilter = __decorate([__decorateParam(3, Inject(Injector)), __decorateParam(4, ICommandService)], FFilter);

//#endregion
//#region src/facade/f-range.ts
var FRangeSheetsFilterMixin = class extends FRange {
	createFilter() {
		if (this._getFilterModel()) return null;
		if (!this._commandService.syncExecuteCommand(SetSheetFilterRangeCommand.id, {
			unitId: this._workbook.getUnitId(),
			subUnitId: this._worksheet.getSheetId(),
			range: this._range
		})) return null;
		return this.getFilter();
	}
	/**
	* Get the filter for the current range's worksheet.
	* @returns {FFilter | null} The interface class to handle the filter. If the worksheet does not have a filter,
	* this method would return `null`.
	*/
	getFilter() {
		const filterModel = this._getFilterModel();
		if (!filterModel) return null;
		return this._injector.createInstance(FFilter, this._workbook, this._worksheet, filterModel);
	}
	_getFilterModel() {
		return this._injector.get(SheetsFilterService).getFilterModel(this._workbook.getUnitId(), this._worksheet.getSheetId());
	}
};
FRange.extend(FRangeSheetsFilterMixin);

//#endregion
//#region src/facade/f-worksheet.ts
var FWorksheetFilterMixin = class extends FWorksheet {
	getFilter() {
		const filterModel = this._getFilterModel();
		if (!filterModel) return null;
		return this._injector.createInstance(FFilter, this._workbook, this._worksheet, filterModel);
	}
	_getFilterModel() {
		return this._injector.get(SheetsFilterService).getFilterModel(this._workbook.getUnitId(), this._worksheet.getSheetId());
	}
};
FWorksheet.extend(FWorksheetFilterMixin);

//#endregion
//#region src/facade/f-enum.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
var FSheetsFilterEnumMixin = class extends FEnum {
	get CustomFilterOperator() {
		return CustomFilterOperator;
	}
};
FEnum.extend(FSheetsFilterEnumMixin);

//#endregion
//#region src/facade/f-event.ts
var FSheetsFilterEventNameMixin = class extends FEventName {
	get SheetBeforeRangeFilter() {
		return "SheetBeforeRangeFilter";
	}
	get SheetRangeFiltered() {
		return "SheetRangeFiltered";
	}
	get SheetRangeFilterCleared() {
		return "SheetRangeFilterCleared";
	}
	get SheetBeforeRangeFilterClear() {
		return "SheetBeforeRangeFilterClear";
	}
};
FEventName.extend(FSheetsFilterEventNameMixin);

//#endregion
export { FFilter };
