// @univerjs/docs-thread-comment-ui/lib/index.js
import { BuildTextUtils, CommandType, CustomDecorationType, DependentOn, Disposable, ICommandService, IConfigService, IUniverInstanceService, Inject, Injector, Plugin, SHEET_EDITOR_UNITS, UniverInstanceType, UserManagerService, isInternalEditorID, merge, sequenceExecute } from "@univerjs/core";
import { DocBackScrollRenderController, DocRenderController, DocSelectionRenderService, addCustomDecorationBySelectionFactory, deleteCustomDecorationFactory } from "@univerjs/docs-ui";
import { AddCommentMutation, IThreadCommentDataSourceService, ThreadCommentModel, getDT } from "@univerjs/thread-comment";
import { SetActiveCommentOperation, ThreadCommentPanel, ThreadCommentPanelService, UniverThreadCommentUIPlugin } from "@univerjs/thread-comment-ui";
import { DOC_INTERCEPTOR_POINT, DocInterceptorService, DocSelectionManagerService, DocSkeletonManagerService, RichTextEditingMutation, SetTextSelectionsOperation } from "@univerjs/docs";
import { DocumentEditArea, IRenderManagerService, withCurrentTypeOfRenderer } from "@univerjs/engine-render";
import { ComponentManager, ContextMenuGroup, ContextMenuPosition, IMenuManagerService, ISidebarService, MenuItemType, RibbonInsertGroup, getMenuHiddenObservable, useDependency, useObservable } from "@univerjs/ui";
import { BehaviorSubject, Observable, debounceTime, filter } from "rxjs";
import { CommentIcon } from "@univerjs/icons";
import { useEffect, useMemo, useState } from "react";
import { jsx } from "react/jsx-runtime";

//#region src/common/const.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const DOCS_THREAD_COMMENT_PANEL = "univer.doc.thread-comment-panel";
const PLUGIN_NAME = "DOC_THREAD_COMMENT_UI_PLUGIN";
const DEFAULT_DOC_SUBUNIT_ID = "default_doc";

//#endregion
//#region src/commands/commands/add-doc-comment.command.ts
const AddDocCommentComment = {
	id: "docs.command.add-comment",
	type: CommandType.COMMAND,
	async handler(accessor, params) {
		if (!params) return false;
		const { comment: originComment, unitId } = params;
		const comment = await accessor.get(IThreadCommentDataSourceService).addComment(originComment);
		const commandService = accessor.get(ICommandService);
		const doMutation = addCustomDecorationBySelectionFactory(accessor, {
			id: comment.threadId,
			type: CustomDecorationType.COMMENT,
			unitId
		});
		if (doMutation) return (await sequenceExecute([
			{
				id: AddCommentMutation.id,
				params: {
					unitId,
					subUnitId: DEFAULT_DOC_SUBUNIT_ID,
					comment
				}
			},
			doMutation,
			{
				id: SetActiveCommentOperation.id,
				params: {
					unitId,
					subUnitId: DEFAULT_DOC_SUBUNIT_ID,
					commentId: comment.id
				}
			}
		], commandService)).result;
		return false;
	}
};

//#endregion
//#region src/commands/commands/delete-doc-comment.command.ts
const DeleteDocCommentComment = {
	id: "docs.command.delete-comment",
	type: CommandType.COMMAND,
	async handler(accessor, params) {
		if (!params) return false;
		const { commentId, unitId } = params;
		const commandService = accessor.get(ICommandService);
		const doMutation = deleteCustomDecorationFactory(accessor, {
			id: commentId,
			unitId
		});
		if (doMutation) return (await sequenceExecute([doMutation], commandService)).result;
		return false;
	}
};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/services/doc-thread-comment.service.ts
let DocThreadCommentService = class DocThreadCommentService extends Disposable {
	get addingComment() {
		return this._addingComment$.getValue();
	}
	constructor(_sidebarService, _threadCommentPanelService) {
		super();
		this._sidebarService = _sidebarService;
		this._threadCommentPanelService = _threadCommentPanelService;
		_defineProperty(this, "_addingComment$", new BehaviorSubject(void 0));
		_defineProperty(this, "addingComment$", this._addingComment$.asObservable());
		this.disposeWithMe(() => {
			this._addingComment$.complete();
		});
	}
	startAdd(comment) {
		this._addingComment$.next(comment);
	}
	endAdd() {
		this._addingComment$.next(void 0);
	}
};
DocThreadCommentService = __decorate([__decorateParam(0, ISidebarService), __decorateParam(1, Inject(ThreadCommentPanelService))], DocThreadCommentService);

//#endregion
//#region src/commands/operations/show-comment-panel.operation.ts
const ShowCommentPanelOperation = {
	id: "docs.operation.show-comment-panel",
	type: CommandType.OPERATION,
	handler(accessor, params) {
		var _sidebarService$optio;
		const panelService = accessor.get(ThreadCommentPanelService);
		const sidebarService = accessor.get(ISidebarService);
		if (!panelService.panelVisible || ((_sidebarService$optio = sidebarService.options.children) === null || _sidebarService$optio === void 0 ? void 0 : _sidebarService$optio.label) !== "univer.doc.thread-comment-panel") {
			sidebarService.open({
				header: { title: "docs-thread-comment-ui.panel.title" },
				children: { label: DOCS_THREAD_COMMENT_PANEL },
				width: 320,
				onClose: () => panelService.setPanelVisible(false)
			});
			panelService.setPanelVisible(true);
		}
		if (params) panelService.setActiveComment(params === null || params === void 0 ? void 0 : params.activeComment);
		return true;
	}
};
const ToggleCommentPanelOperation = {
	id: "docs.operation.toggle-comment-panel",
	type: CommandType.OPERATION,
	handler(accessor) {
		var _sidebarService$optio2;
		const panelService = accessor.get(ThreadCommentPanelService);
		const sidebarService = accessor.get(ISidebarService);
		if (!panelService.panelVisible || ((_sidebarService$optio2 = sidebarService.options.children) === null || _sidebarService$optio2 === void 0 ? void 0 : _sidebarService$optio2.label) !== "univer.doc.thread-comment-panel") {
			sidebarService.open({
				header: { title: "docs-thread-comment-ui.panel.title" },
				children: { label: DOCS_THREAD_COMMENT_PANEL },
				width: 320,
				onClose: () => panelService.setPanelVisible(false)
			});
			panelService.setPanelVisible(true);
		} else {
			sidebarService.close();
			panelService.setPanelVisible(false);
			panelService.setActiveComment(null);
		}
		return true;
	}
};
const StartAddCommentOperation = {
	id: "docs.operation.start-add-comment",
	type: CommandType.OPERATION,
	handler(accessor) {
		var _renderManagerService, _doc$getBody$dataStre, _doc$getBody;
		const panelService = accessor.get(ThreadCommentPanelService);
		const doc = accessor.get(IUniverInstanceService).getCurrentUnitOfType(UniverInstanceType.UNIVER_DOC);
		const docSelectionManagerService = accessor.get(DocSelectionManagerService);
		const renderManagerService = accessor.get(IRenderManagerService);
		const userManagerService = accessor.get(UserManagerService);
		const docCommentService = accessor.get(DocThreadCommentService);
		const commandService = accessor.get(ICommandService);
		const sidebarService = accessor.get(ISidebarService);
		const textRange = docSelectionManagerService.getActiveTextRange();
		if (!doc || !textRange) return false;
		const docSelectionRenderManager = (_renderManagerService = renderManagerService.getRenderById(doc.getUnitId())) === null || _renderManagerService === void 0 ? void 0 : _renderManagerService.with(DocSelectionRenderService);
		docSelectionRenderManager === null || docSelectionRenderManager === void 0 || docSelectionRenderManager.setReserveRangesStatus(true);
		if (textRange.collapsed) {
			if (panelService.panelVisible) {
				panelService.setPanelVisible(false);
				sidebarService.close();
			} else commandService.executeCommand(ShowCommentPanelOperation.id);
			return true;
		}
		commandService.executeCommand(ShowCommentPanelOperation.id);
		const unitId = doc.getUnitId();
		const dataStream = ((_doc$getBody$dataStre = (_doc$getBody = doc.getBody()) === null || _doc$getBody === void 0 ? void 0 : _doc$getBody.dataStream) !== null && _doc$getBody$dataStre !== void 0 ? _doc$getBody$dataStre : "").slice(textRange.startOffset, textRange.endOffset);
		const text = BuildTextUtils.transform.getPlainText(dataStream);
		const subUnitId = DEFAULT_DOC_SUBUNIT_ID;
		const commentId = "";
		const comment = {
			unitId,
			subUnitId,
			id: commentId,
			ref: text,
			dT: getDT(),
			personId: userManagerService.getCurrentUser().userID,
			text: { dataStream: "\r\n" },
			startOffset: textRange.startOffset,
			endOffset: textRange.endOffset,
			collapsed: true,
			threadId: commentId
		};
		docSelectionRenderManager === null || docSelectionRenderManager === void 0 || docSelectionRenderManager.blur();
		docCommentService.startAdd(comment);
		panelService.setActiveComment({
			unitId,
			subUnitId,
			commentId
		});
		return true;
	}
};

//#endregion
//#region package.json
var name = "@univerjs/docs-thread-comment-ui";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
const DOCS_THREAD_COMMENT_UI_PLUGIN_CONFIG_KEY = "docs-thread-comment-ui.config";
const configSymbol = Symbol(DOCS_THREAD_COMMENT_UI_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/controllers/doc-thread-comment-selection.controller.ts
let DocThreadCommentSelectionController = class DocThreadCommentSelectionController extends Disposable {
	constructor(_threadCommentPanelService, _univerInstanceService, _commandService, _docThreadCommentService, _renderManagerService, _threadCommentModel) {
		super();
		this._threadCommentPanelService = _threadCommentPanelService;
		this._univerInstanceService = _univerInstanceService;
		this._commandService = _commandService;
		this._docThreadCommentService = _docThreadCommentService;
		this._renderManagerService = _renderManagerService;
		this._threadCommentModel = _threadCommentModel;
		this._initSelectionChange();
		this._initActiveCommandChange();
	}
	_initSelectionChange() {
		let lastSelection;
		this.disposeWithMe(this._commandService.onCommandExecuted((commandInfo) => {
			if (commandInfo.id === SetTextSelectionsOperation.id) {
				const { unitId, ranges } = commandInfo.params;
				if (isInternalEditorID(unitId)) return;
				const doc = this._univerInstanceService.getUnit(unitId, UniverInstanceType.UNIVER_DOC);
				const primary = ranges[0];
				if ((lastSelection === null || lastSelection === void 0 ? void 0 : lastSelection.startOffset) === (primary === null || primary === void 0 ? void 0 : primary.startOffset) && (lastSelection === null || lastSelection === void 0 ? void 0 : lastSelection.endOffset) === (primary === null || primary === void 0 ? void 0 : primary.endOffset)) return;
				lastSelection = primary;
				if (primary && doc) {
					const { startOffset, endOffset, collapsed } = primary;
					let customRange;
					if (collapsed) {
						var _doc$getBody;
						customRange = (_doc$getBody = doc.getBody()) === null || _doc$getBody === void 0 || (_doc$getBody = _doc$getBody.customDecorations) === null || _doc$getBody === void 0 ? void 0 : _doc$getBody.find((value) => value.startIndex <= startOffset && value.endIndex >= endOffset - 1);
					} else {
						var _doc$getBody2;
						customRange = (_doc$getBody2 = doc.getBody()) === null || _doc$getBody2 === void 0 || (_doc$getBody2 = _doc$getBody2.customDecorations) === null || _doc$getBody2 === void 0 ? void 0 : _doc$getBody2.find((value) => value.startIndex <= startOffset && value.endIndex >= endOffset - 1);
					}
					if (customRange) {
						const comment = this._threadCommentModel.getComment(unitId, DEFAULT_DOC_SUBUNIT_ID, customRange.id);
						if (comment && !comment.resolved) this._commandService.executeCommand(ShowCommentPanelOperation.id, { activeComment: {
							unitId,
							subUnitId: DEFAULT_DOC_SUBUNIT_ID,
							commentId: customRange.id
						} });
						return;
					}
				}
				if (!this._threadCommentPanelService.activeCommentId) return;
				const addingComment = this._docThreadCommentService.addingComment;
				const activeComment = this._threadCommentPanelService.activeCommentId;
				if (addingComment && (activeComment === null || activeComment === void 0 ? void 0 : activeComment.unitId) === addingComment.unitId && (activeComment === null || activeComment === void 0 ? void 0 : activeComment.subUnitId) === "default_doc" && (activeComment === null || activeComment === void 0 ? void 0 : activeComment.commentId) === addingComment.id) return;
				this._commandService.executeCommand(SetActiveCommentOperation.id);
			}
		}));
	}
	_initActiveCommandChange() {
		this.disposeWithMe(this._threadCommentPanelService.activeCommentId$.subscribe((activeComment) => {
			var _this$_docThreadComme;
			if (activeComment) {
				const doc = this._univerInstanceService.getUnit(activeComment.unitId);
				if (doc) {
					var _this$_renderManagerS, _doc$getBody3;
					const backScrollController = (_this$_renderManagerS = this._renderManagerService.getRenderById(activeComment.unitId)) === null || _this$_renderManagerS === void 0 ? void 0 : _this$_renderManagerS.with(DocBackScrollRenderController);
					const customRange = (_doc$getBody3 = doc.getBody()) === null || _doc$getBody3 === void 0 || (_doc$getBody3 = _doc$getBody3.customDecorations) === null || _doc$getBody3 === void 0 ? void 0 : _doc$getBody3.find((range) => range.id === activeComment.commentId);
					if (customRange && backScrollController) backScrollController.scrollToRange({
						startOffset: customRange.startIndex,
						endOffset: customRange.endIndex,
						collapsed: false
					});
				}
			}
			if (!activeComment || activeComment.commentId !== ((_this$_docThreadComme = this._docThreadCommentService.addingComment) === null || _this$_docThreadComme === void 0 ? void 0 : _this$_docThreadComme.id)) this._docThreadCommentService.endAdd();
		}));
	}
};
DocThreadCommentSelectionController = __decorate([
	__decorateParam(0, Inject(ThreadCommentPanelService)),
	__decorateParam(1, IUniverInstanceService),
	__decorateParam(2, ICommandService),
	__decorateParam(3, Inject(DocThreadCommentService)),
	__decorateParam(4, IRenderManagerService),
	__decorateParam(5, Inject(ThreadCommentModel))
], DocThreadCommentSelectionController);

//#endregion
//#region src/menu/menu.ts
const shouldDisableAddComment = (accessor) => {
	var _withCurrentTypeOfRen;
	const renderManagerService = accessor.get(IRenderManagerService);
	const docSelectionManagerService = accessor.get(DocSelectionManagerService);
	const skeleton = (_withCurrentTypeOfRen = withCurrentTypeOfRenderer(UniverInstanceType.UNIVER_DOC, DocSkeletonManagerService, accessor.get(IUniverInstanceService), renderManagerService)) === null || _withCurrentTypeOfRen === void 0 ? void 0 : _withCurrentTypeOfRen.getSkeleton();
	const editArea = skeleton === null || skeleton === void 0 ? void 0 : skeleton.getViewModel().getEditArea();
	if (editArea === DocumentEditArea.FOOTER || editArea === DocumentEditArea.HEADER) return true;
	const range = docSelectionManagerService.getActiveTextRange();
	if (range == null || range.collapsed) return true;
	return false;
};
function AddDocCommentMenuItemFactory(accessor) {
	return {
		id: StartAddCommentOperation.id,
		type: MenuItemType.BUTTON,
		icon: "CommentIcon",
		title: "docs-thread-comment-ui.panel.addComment",
		tooltip: "docs-thread-comment-ui.panel.addComment",
		hidden$: getMenuHiddenObservable(accessor, UniverInstanceType.UNIVER_DOC, void 0, SHEET_EDITOR_UNITS),
		disabled$: new Observable(function(subscribe) {
			const observer = accessor.get(DocSelectionManagerService).textSelection$.pipe(debounceTime(16)).subscribe(() => {
				subscribe.next(shouldDisableAddComment(accessor));
			});
			return () => {
				observer.unsubscribe();
			};
		})
	};
}
function ToolbarDocCommentMenuItemFactory(accessor) {
	return {
		id: ToggleCommentPanelOperation.id,
		type: MenuItemType.BUTTON,
		icon: "CommentIcon",
		title: "docs-thread-comment-ui.panel.addComment",
		tooltip: "docs-thread-comment-ui.panel.addComment",
		hidden$: getMenuHiddenObservable(accessor, UniverInstanceType.UNIVER_DOC)
	};
}

//#endregion
//#region src/menu/schema.ts
const menuSchema = {
	[RibbonInsertGroup.MEDIA]: { [ToggleCommentPanelOperation.id]: {
		order: 3,
		menuItemFactory: ToolbarDocCommentMenuItemFactory
	} },
	[ContextMenuPosition.MAIN_AREA]: { [ContextMenuGroup.DATA]: { [StartAddCommentOperation.id]: {
		order: 1,
		menuItemFactory: AddDocCommentMenuItemFactory
	} } }
};

//#endregion
//#region src/views/doc-thread-comment-panel/index.tsx
const DocThreadCommentPanel = () => {
	const univerInstanceService = useDependency(IUniverInstanceService);
	const injector = useDependency(Injector);
	const doc = useObservable(useMemo(() => univerInstanceService.getCurrentTypeOfUnit$(UniverInstanceType.UNIVER_DOC).pipe(filter((doc) => !!doc && !isInternalEditorID(doc.getUnitId()))), [univerInstanceService]));
	const subUnitId$ = useMemo(() => new Observable((sub) => sub.next(DEFAULT_DOC_SUBUNIT_ID)), []);
	const docSelectionManagerService = useDependency(DocSelectionManagerService);
	useObservable(useMemo(() => docSelectionManagerService.textSelection$.pipe(debounceTime(16)), [docSelectionManagerService.textSelection$]));
	const commandService = useDependency(ICommandService);
	const docCommentService = useDependency(DocThreadCommentService);
	const tempComment = useObservable(docCommentService.addingComment$);
	const [commentIds, setCommentIds] = useState([]);
	useEffect(() => {
		var _customRanges$map$fil;
		const set = /* @__PURE__ */ new Set();
		const customRanges = doc === null || doc === void 0 ? void 0 : doc.getCustomDecorations();
		setCommentIds((_customRanges$map$fil = customRanges === null || customRanges === void 0 ? void 0 : customRanges.map((r) => r.id).filter((i) => {
			const hasRepeat = set.has(i);
			set.add(i);
			return !hasRepeat;
		})) !== null && _customRanges$map$fil !== void 0 ? _customRanges$map$fil : []);
		const dispose = commandService.onCommandExecuted((command) => {
			if (command.id === RichTextEditingMutation.id) {
				var _customRanges$map$fil2;
				const set = /* @__PURE__ */ new Set();
				const customRanges = doc === null || doc === void 0 ? void 0 : doc.getCustomDecorations();
				setCommentIds((_customRanges$map$fil2 = customRanges === null || customRanges === void 0 ? void 0 : customRanges.map((r) => r.id).filter((i) => {
					const hasRepeat = set.has(i);
					set.add(i);
					return !hasRepeat;
				})) !== null && _customRanges$map$fil2 !== void 0 ? _customRanges$map$fil2 : []);
			}
		});
		return () => {
			dispose.dispose();
		};
	}, [commandService, doc]);
	if (!doc) return null;
	const isInValidSelection = shouldDisableAddComment(injector);
	const unitId = doc.getUnitId();
	return /* @__PURE__ */ jsx(ThreadCommentPanel, {
		unitId,
		subUnitId$,
		type: UniverInstanceType.UNIVER_DOC,
		onAdd: () => {
			commandService.executeCommand(StartAddCommentOperation.id);
		},
		getSubUnitName: () => "",
		disableAdd: isInValidSelection,
		tempComment,
		onAddComment: (comment) => {
			if (!comment.parentId) {
				const params = {
					unitId,
					range: tempComment,
					comment
				};
				commandService.executeCommand(AddDocCommentComment.id, params);
				docCommentService.endAdd();
				return false;
			}
			return true;
		},
		onDeleteComment: (comment) => {
			if (!comment.parentId) {
				const params = {
					unitId,
					commentId: comment.id
				};
				commandService.executeCommand(DeleteDocCommentComment.id, params);
				return false;
			}
			return true;
		},
		showComments: commentIds
	});
};

//#endregion
//#region src/controllers/doc-thread-comment-ui.controller.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
let DocThreadCommentUIController = class DocThreadCommentUIController extends Disposable {
	constructor(_commandService, _menuManagerService, _componentManager) {
		super();
		this._commandService = _commandService;
		this._menuManagerService = _menuManagerService;
		this._componentManager = _componentManager;
		this._initCommands();
		this._initMenus();
		this._initComponents();
	}
	_initCommands() {
		[
			AddDocCommentComment,
			DeleteDocCommentComment,
			ShowCommentPanelOperation,
			StartAddCommentOperation,
			ToggleCommentPanelOperation
		].forEach((command) => {
			this.disposeWithMe(this._commandService.registerCommand(command));
		});
	}
	_initMenus() {
		this._menuManagerService.mergeMenu(menuSchema);
	}
	_initComponents() {
		[[DOCS_THREAD_COMMENT_PANEL, DocThreadCommentPanel], ["CommentIcon", CommentIcon]].forEach(([id, comp]) => {
			this.disposeWithMe(this._componentManager.register(id, comp));
		});
	}
};
DocThreadCommentUIController = __decorate([
	__decorateParam(0, ICommandService),
	__decorateParam(1, IMenuManagerService),
	__decorateParam(2, Inject(ComponentManager))
], DocThreadCommentUIController);

//#endregion
//#region src/controllers/render-controllers/render.controller.ts
let DocThreadCommentRenderController = class DocThreadCommentRenderController extends Disposable {
	constructor(_context, _docInterceptorService, _threadCommentPanelService, _docRenderController, _univerInstanceService, _threadCommentModel, _commandService) {
		super();
		this._context = _context;
		this._docInterceptorService = _docInterceptorService;
		this._threadCommentPanelService = _threadCommentPanelService;
		this._docRenderController = _docRenderController;
		this._univerInstanceService = _univerInstanceService;
		this._threadCommentModel = _threadCommentModel;
		this._commandService = _commandService;
		this._interceptorViewModel();
		this._initReRender();
		this._initSyncComments();
	}
	_initReRender() {
		this.disposeWithMe(this._threadCommentPanelService.activeCommentId$.subscribe((activeComment) => {
			var _this$_univerInstance;
			if (activeComment) {
				this._docRenderController.reRender(activeComment.unitId);
				return;
			}
			const unitId = (_this$_univerInstance = this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_DOC)) === null || _this$_univerInstance === void 0 ? void 0 : _this$_univerInstance.getUnitId();
			if (unitId) this._docRenderController.reRender(unitId);
		}));
		this.disposeWithMe(this._threadCommentModel.commentUpdate$.subscribe((update) => {
			if (update.type === "resolve") this._docRenderController.reRender(update.unitId);
		}));
	}
	_interceptorViewModel() {
		this._docInterceptorService.intercept(DOC_INTERCEPTOR_POINT.CUSTOM_DECORATION, { handler: (data, pos, next) => {
			if (!data) return next(data);
			const { unitId, index, customDecorations } = pos;
			const { commentId, unitId: commentUnitID } = this._threadCommentPanelService.activeCommentId || {};
			const activeCustomDecoration = customDecorations.find((i) => i.id === commentId);
			const comment = this._threadCommentModel.getComment(unitId, DEFAULT_DOC_SUBUNIT_ID, data.id);
			if (!comment) return next({
				...data,
				show: false
			});
			const isActiveIndex = activeCustomDecoration && index >= activeCustomDecoration.startIndex && index <= activeCustomDecoration.endIndex;
			const isActive = commentUnitID === unitId && data.id === commentId;
			return next({
				...data,
				active: isActive || isActiveIndex,
				show: !comment.resolved
			});
		} });
	}
	_initSyncComments() {
		var _this$_context$unit$g, _this$_context$unit$g2;
		const unitId = this._context.unit.getUnitId();
		const subUnitId = DEFAULT_DOC_SUBUNIT_ID;
		const threadIds = (_this$_context$unit$g = (_this$_context$unit$g2 = this._context.unit.getBody()) === null || _this$_context$unit$g2 === void 0 || (_this$_context$unit$g2 = _this$_context$unit$g2.customDecorations) === null || _this$_context$unit$g2 === void 0 ? void 0 : _this$_context$unit$g2.filter((i) => i.type === CustomDecorationType.COMMENT).map((i) => i.id)) !== null && _this$_context$unit$g !== void 0 ? _this$_context$unit$g : [];
		threadIds.length && this._threadCommentModel.syncThreadComments(this._context.unit.getUnitId(), "default_doc", threadIds);
		let prevThreadIds = threadIds.sort();
		this.disposeWithMe(this._commandService.onCommandExecuted((commandInfo) => {
			if (commandInfo.id === RichTextEditingMutation.id) {
				var _this$_context$unit$g3, _this$_context$unit$g4;
				if (commandInfo.params.unitId !== this._context.unit.getUnitId()) return;
				const currentThreadIds = (_this$_context$unit$g3 = (_this$_context$unit$g4 = this._context.unit.getBody()) === null || _this$_context$unit$g4 === void 0 || (_this$_context$unit$g4 = _this$_context$unit$g4.customDecorations) === null || _this$_context$unit$g4 === void 0 ? void 0 : _this$_context$unit$g4.filter((i) => i.type === CustomDecorationType.COMMENT).map((i) => i.id)) !== null && _this$_context$unit$g3 !== void 0 ? _this$_context$unit$g3 : [];
				const currentThreadIdsSorted = currentThreadIds.sort();
				if (JSON.stringify(prevThreadIds) !== JSON.stringify(currentThreadIdsSorted)) {
					const preIds = new Set(prevThreadIds);
					const addIds = /* @__PURE__ */ new Set();
					currentThreadIds.forEach((id) => {
						if (!preIds.has(id)) addIds.add(id);
					});
					prevThreadIds = currentThreadIdsSorted;
					this._threadCommentModel.syncThreadComments(unitId, subUnitId, [...addIds]);
				}
			}
		}));
	}
};
DocThreadCommentRenderController = __decorate([
	__decorateParam(1, Inject(DocInterceptorService)),
	__decorateParam(2, Inject(ThreadCommentPanelService)),
	__decorateParam(3, Inject(DocRenderController)),
	__decorateParam(4, IUniverInstanceService),
	__decorateParam(5, Inject(ThreadCommentModel)),
	__decorateParam(6, ICommandService)
], DocThreadCommentRenderController);

//#endregion
//#region src/plugin.ts
let UniverDocsThreadCommentUIPlugin = class UniverDocsThreadCommentUIPlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _injector, _renderManagerSrv, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._renderManagerSrv = _renderManagerSrv;
		this._configService = _configService;
		const { menu, ...rest } = merge({}, defaultPluginConfig, this._config);
		if (menu) this._configService.setConfig("menu", menu, { merge: true });
		this._configService.setConfig(DOCS_THREAD_COMMENT_UI_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		[
			[DocThreadCommentUIController],
			[DocThreadCommentSelectionController],
			[DocThreadCommentService]
		].forEach((dep) => {
			this._injector.add(dep);
		});
	}
	onRendered() {
		this._initRenderModule();
		this._injector.get(DocThreadCommentSelectionController);
		this._injector.get(DocThreadCommentUIController);
	}
	_initRenderModule() {
		[DocThreadCommentRenderController].forEach((dep) => {
			this._renderManagerSrv.registerRenderModule(UniverInstanceType.UNIVER_DOC, dep);
		});
	}
};
_defineProperty(UniverDocsThreadCommentUIPlugin, "pluginName", PLUGIN_NAME);
_defineProperty(UniverDocsThreadCommentUIPlugin, "packageName", name);
_defineProperty(UniverDocsThreadCommentUIPlugin, "version", version);
_defineProperty(UniverDocsThreadCommentUIPlugin, "type", UniverInstanceType.UNIVER_DOC);
UniverDocsThreadCommentUIPlugin = __decorate([
	DependentOn(UniverThreadCommentUIPlugin),
	__decorateParam(1, Inject(Injector)),
	__decorateParam(2, IRenderManagerService),
	__decorateParam(3, IConfigService)
], UniverDocsThreadCommentUIPlugin);

//#endregion
export { AddDocCommentComment, DeleteDocCommentComment, ShowCommentPanelOperation, StartAddCommentOperation, UniverDocsThreadCommentUIPlugin };
