// @univerjs/sheets-graphics/lib/es/index.js
import { Disposable, IConfigService, Inject, Injector, Plugin, Range, UniverInstanceType, merge } from "@univerjs/core";
import { SheetPrintInterceptorService } from "@univerjs/sheets-ui";
import { IRenderManagerService, SheetExtension } from "@univerjs/engine-render";

//#region src/common/const.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const UNIQUE_KEY = "SheetGraphicsExtension";

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region src/views/extensions/graphics.extension.ts
var Graphics = class Graphics extends SheetExtension {
	constructor(..._args) {
		super(..._args);
		_defineProperty(this, "uKey", UNIQUE_KEY);
		_defineProperty(this, "Z_INDEX", 35);
		_defineProperty(this, "_graphicsRenderMap", /* @__PURE__ */ new Map());
	}
	registerRenderer(key, renderer) {
		this._graphicsRenderMap.set(key, renderer);
	}
	draw(ctx, _parentScale, skeleton, diffBounds, { viewRanges }) {
		const featureTypes = Array.from(this._graphicsRenderMap.keys());
		viewRanges.forEach((range) => {
			Range.foreach(range, (row, col) => {
				const cellPosition = skeleton.getCellByIndexWithNoHeader(row, col);
				if (!cellPosition) return;
				featureTypes.forEach((featureType) => {
					const renderer = this._graphicsRenderMap.get(featureType);
					renderer === null || renderer === void 0 || renderer(ctx, skeleton, cellPosition);
				});
			});
		});
	}
	dispose() {
		this._graphicsRenderMap.clear();
	}
	copy() {
		const newGraphics = new Graphics();
		this._graphicsRenderMap.forEach((renderer, key) => {
			newGraphics.registerRenderer(key, renderer);
		});
		return newGraphics;
	}
};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/controllers/graphics-render.controller.ts
let SheetGraphicsRenderController = class SheetGraphicsRenderController extends Disposable {
	constructor(_context, _sheetPrintInterceptorService) {
		super();
		this._context = _context;
		this._sheetPrintInterceptorService = _sheetPrintInterceptorService;
		_defineProperty(this, "_graphicsExtensionInstance", null);
		this._initRender();
		this._initPrinting();
	}
	_initRender() {
		const spreadsheet = this._context.mainComponent;
		if (spreadsheet && !spreadsheet.getExtensionByKey("SheetGraphicsExtension")) {
			this._graphicsExtensionInstance = new Graphics();
			spreadsheet.register(this._graphicsExtensionInstance);
		}
	}
	_initPrinting() {
		this.disposeWithMe(this._sheetPrintInterceptorService.interceptor.intercept(this._sheetPrintInterceptorService.interceptor.getInterceptPoints().PRINTING_COMPONENT_COLLECT, { handler: (component, context, next) => {
			const { spreadsheet } = context;
			if (this._graphicsExtensionInstance) spreadsheet.register(this._graphicsExtensionInstance.copy());
			return next(component);
		} }));
	}
	registerRenderer(key, renderer) {
		var _this$_graphicsExtens;
		(_this$_graphicsExtens = this._graphicsExtensionInstance) === null || _this$_graphicsExtens === void 0 || _this$_graphicsExtens.registerRenderer(key, renderer);
	}
};
SheetGraphicsRenderController = __decorate([__decorateParam(1, Inject(SheetPrintInterceptorService))], SheetGraphicsRenderController);

//#endregion
//#region package.json
var name = "@univerjs/sheets-graphics";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
const PLUGIN_CONFIG_KEY = "graphics.config";
const configSymbol = Symbol(PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/plugin.ts
let UniverSheetsGraphicsPlugin = class UniverSheetsGraphicsPlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _injector, _configService, _renderManagerService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		this._renderManagerService = _renderManagerService;
		const { ...rest } = merge({}, defaultPluginConfig, this._config);
		this._configService.setConfig(PLUGIN_CONFIG_KEY, rest);
	}
	onRendered() {
		[[SheetGraphicsRenderController]].forEach((dep) => {
			this._renderManagerService.registerRenderModule(UniverInstanceType.UNIVER_SHEET, dep);
		});
	}
};
_defineProperty(UniverSheetsGraphicsPlugin, "pluginName", "UNIVER_SHEET_DRAWING_PLUGIN");
_defineProperty(UniverSheetsGraphicsPlugin, "packageName", name);
_defineProperty(UniverSheetsGraphicsPlugin, "version", version);
UniverSheetsGraphicsPlugin = __decorate([
	__decorateParam(1, Inject(Injector)),
	__decorateParam(2, IConfigService),
	__decorateParam(3, IRenderManagerService)
], UniverSheetsGraphicsPlugin);

//#endregion
export { SheetGraphicsRenderController, UniverSheetsGraphicsPlugin };
