// @univerjs/rpc-node/lib/cjs/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") {
		for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
			key = keys[i];
			if (!__hasOwnProp.call(to, key) && key !== except) {
				__defProp(to, key, {
					get: ((k) => from[k]).bind(null, key),
					enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
				});
			}
		}
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
let node_child_process = require("node:child_process");
let node_process = require("node:process");
node_process = __toESM(node_process);
let _univerjs_core = require("@univerjs/core");
let _univerjs_rpc = require("@univerjs/rpc");
let rxjs = require("rxjs");

//#region package.json
var name = "@univerjs/rpc-node";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const PLUGIN_CONFIG_KEY_MAIN_THREAD = "rpc-node.main-thread.config";
const configSymbolMainThread = Symbol(PLUGIN_CONFIG_KEY_MAIN_THREAD);
const defaultPluginMainThreadConfig = {};
const PLUGIN_CONFIG_KEY_WORKER_THREAD = "rpc-node.worker-thread.config";
const configSymbolWorkerThread = Symbol(PLUGIN_CONFIG_KEY_WORKER_THREAD);
const defaultPluginWorkerThreadConfig = {};

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/plugin.ts
let UniverRPCNodeMainPlugin = class UniverRPCNodeMainPlugin extends _univerjs_core.Plugin {
	constructor(_config = defaultPluginMainThreadConfig, _injector, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		_defineProperty(this, "_child", null);
		const { ...rest } = (0, _univerjs_core.merge)({}, defaultPluginMainThreadConfig, this._config);
		this._configService.setConfig(PLUGIN_CONFIG_KEY_MAIN_THREAD, rest);
	}
	onStarting() {
		const { workerSrc } = this._config;
		if (!workerSrc) throw new Error("[UniverRPCNodeMainPlugin] workerSrc is required for UniverRPCNodeMainPlugin");
		const [messageProtocol, child] = createNodeMessagePortOnMain(this._injector, workerSrc);
		[
			[_univerjs_rpc.IRPCChannelService, { useFactory: () => new _univerjs_rpc.ChannelService(messageProtocol) }],
			[_univerjs_rpc.DataSyncPrimaryController],
			[_univerjs_rpc.IRemoteSyncService, { useClass: _univerjs_rpc.RemoteSyncPrimaryService }]
		].forEach((dependency) => this._injector.add(dependency));
		this._injector.get(_univerjs_rpc.DataSyncPrimaryController);
		this._child = child;
	}
	dispose() {
		super.dispose();
		if (this._child) {
			try {
				this._child.kill();
			} catch (e) {
				console.error("Failed to kill child process:", e);
			}
			this._child = null;
		}
	}
};
_defineProperty(UniverRPCNodeMainPlugin, "pluginName", "UNIVER_RPC_NODE_MAIN_PLUGIN");
_defineProperty(UniverRPCNodeMainPlugin, "packageName", name);
_defineProperty(UniverRPCNodeMainPlugin, "version", version);
UniverRPCNodeMainPlugin = __decorate([__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector)), __decorateParam(2, _univerjs_core.IConfigService)], UniverRPCNodeMainPlugin);
let UniverRPCNodeWorkerPlugin = class UniverRPCNodeWorkerPlugin extends _univerjs_core.Plugin {
	constructor(_config = defaultPluginWorkerThreadConfig, _injector, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._configService = _configService;
		const { ...rest } = (0, _univerjs_core.merge)({}, defaultPluginWorkerThreadConfig, this._config);
		this._configService.setConfig(PLUGIN_CONFIG_KEY_WORKER_THREAD, rest);
	}
	onStarting() {
		[
			[_univerjs_rpc.DataSyncReplicaController],
			[_univerjs_rpc.IRPCChannelService, { useFactory: () => new _univerjs_rpc.ChannelService(createNodeWorkerMessageProtocol()) }],
			[_univerjs_rpc.IRemoteInstanceService, { useClass: _univerjs_rpc.WebWorkerRemoteInstanceService }]
		].forEach((d) => this._injector.add(d));
		this._injector.get(_univerjs_rpc.DataSyncReplicaController);
	}
};
_defineProperty(UniverRPCNodeWorkerPlugin, "pluginName", "UNIVER_RPC_NODE_WORKER_PLUGIN");
_defineProperty(UniverRPCNodeWorkerPlugin, "packageName", name);
_defineProperty(UniverRPCNodeWorkerPlugin, "version", version);
UniverRPCNodeWorkerPlugin = __decorate([__decorateParam(1, (0, _univerjs_core.Inject)(_univerjs_core.Injector)), __decorateParam(2, _univerjs_core.IConfigService)], UniverRPCNodeWorkerPlugin);
function createNodeMessagePortOnMain(injector, path) {
	const logService = injector.get(_univerjs_core.ILogService);
	const child = (0, node_child_process.fork)(path);
	child.on("spawn", () => logService.log("Child computing process spawned!"));
	child.on("error", (error) => logService.error(error));
	return [{
		send(message) {
			child.send(message);
		},
		onMessage: new rxjs.Observable((subscriber) => {
			const handler = (message) => {
				subscriber.next(message);
			};
			child.on("message", handler);
			return () => child.off("message", handler);
		}).pipe((0, rxjs.shareReplay)({
			bufferSize: 1,
			refCount: true
		}))
	}, child];
}
function createNodeWorkerMessageProtocol() {
	return {
		send(message) {
			node_process.default.send(message);
		},
		onMessage: new rxjs.Observable((subscriber) => {
			const handler = (event) => {
				subscriber.next(event);
			};
			node_process.default.on("message", handler);
			return () => node_process.default.off("message", handler);
		}).pipe((0, rxjs.shareReplay)({
			bufferSize: 1,
			refCount: true
		}))
	};
}

//#endregion
Object.defineProperty(exports, 'UniverRPCNodeMainPlugin', {
  enumerable: true,
  get: function () {
    return UniverRPCNodeMainPlugin;
  }
});
Object.defineProperty(exports, 'UniverRPCNodeWorkerPlugin', {
  enumerable: true,
  get: function () {
    return UniverRPCNodeWorkerPlugin;
  }
});
