// @univerjs/sheets-data-validation/lib/index.js
import { CellValueType, CommandType, DataValidationOperator, DataValidationRenderMode, DataValidationStatus, DataValidationType, DependentOn, Disposable, ICommandService, IConfigService, IPermissionService, IUndoRedoService, IUniverInstanceService, Inject, Injector, LifecycleService, LifecycleStages, ObjectMatrix, Plugin, RBush, Range, Rectangle, RxDisposable, Tools, UniverInstanceType, WrapStrategy, bufferDebounceTime, dateKit, debounce, generateRandomId, getIntersectRange, getOriginCellValue, isFormulaString, isRangesEqual, merge, numfmt, sequenceExecute, toDisposable } from "@univerjs/core";
import { AddDataValidationMutation, BaseDataValidator, DataValidationModel, DataValidatorRegistryService, RemoveDataValidationMutation, UniverDataValidationPlugin, UpdateDataValidationMutation, UpdateRuleType, getRuleOptions, getRuleSetting } from "@univerjs/data-validation";
import { ERROR_TYPE_SET, LexerTreeBuilder, OtherFormulaBizType, RegisterOtherFormulaService, deserializeRangeWithSheet, deserializeRangeWithSheetWithCache, isReferenceString, operatorToken, sequenceNodeType } from "@univerjs/engine-formula";
import { ClearSelectionAllCommand, CopySheetCommand, RefRangeService, RemoveSheetCommand, RemoveSheetMutation, SetRangeValuesMutation, SetRangeValuesUndoMutationFactory, SheetInterceptorService, SheetsSelectionsService, WorksheetViewPermission, deserializeListOptions, getSheetCommandTarget, handleCommonDefaultRangeChangeWithEffectRefCommands } from "@univerjs/sheets";
import { Subject, bufferWhen, filter, skip } from "rxjs";
import { UnitAction } from "@univerjs/protocol";
import { FormulaRefRangeService, UniverSheetsFormulaPlugin } from "@univerjs/sheets-formula";

//#region \0@oxc-project+runtime@0.133.0/helpers/esm/typeof.js
function _typeof(o) {
	"@babel/helpers - typeof";
	return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(o) {
		return typeof o;
	} : function(o) {
		return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
	}, _typeof(o);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPrimitive.js
function toPrimitive(t, r) {
	if ("object" != _typeof(t) || !t) return t;
	var e = t[Symbol.toPrimitive];
	if (void 0 !== e) {
		var i = e.call(t, r || "default");
		if ("object" != _typeof(i)) return i;
		throw new TypeError("@@toPrimitive must return a primitive value.");
	}
	return ("string" === r ? String : Number)(t);
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/toPropertyKey.js
function toPropertyKey(t) {
	var i = toPrimitive(t, "string");
	return "symbol" == _typeof(i) ? i : i + "";
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/defineProperty.js
function _defineProperty(e, r, t) {
	return (r = toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
		value: t,
		enumerable: !0,
		configurable: !0,
		writable: !0
	}) : e[r] = t, e;
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorateParam.js
function __decorateParam(paramIndex, decorator) {
	return function(target, key) {
		decorator(target, key, paramIndex);
	};
}

//#endregion
//#region \0@oxc-project+runtime@0.133.0/helpers/esm/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}

//#endregion
//#region src/services/dv-cache.service.ts
let DataValidationCacheService = class DataValidationCacheService extends Disposable {
	constructor(_commandService, _univerInstanceService, _sheetDataValidationModel) {
		super();
		this._commandService = _commandService;
		this._univerInstanceService = _univerInstanceService;
		this._sheetDataValidationModel = _sheetDataValidationModel;
		_defineProperty(this, "_cacheMatrix", /* @__PURE__ */ new Map());
		_defineProperty(this, "_dirtyRanges$", new Subject());
		_defineProperty(this, "dirtyRanges$", this._dirtyRanges$.asObservable());
		this._initDirtyRanges();
		this._initSheetRemove();
	}
	_initDirtyRanges() {
		this.disposeWithMe(this._commandService.onCommandExecuted((commandInfo, options) => {
			if (commandInfo.id === SetRangeValuesMutation.id && !(options === null || options === void 0 ? void 0 : options.onlyLocal)) {
				const { cellValue, unitId, subUnitId } = commandInfo.params;
				if (cellValue) {
					const range = new ObjectMatrix(cellValue).getDataRange();
					if (range.endRow === -1) return;
					const rules = this._sheetDataValidationModel.getRules(unitId, subUnitId);
					const ranges = [];
					for (const rule of rules) ranges.push(...rule.ranges);
					const intersectsRanges = [];
					for (const ruleRange of ranges) {
						const intersect = getIntersectRange(ruleRange, range);
						if (intersect) intersectsRanges.push(intersect);
					}
					if (intersectsRanges.length) this.markRangeDirty(unitId, subUnitId, intersectsRanges, true);
				}
			}
		}));
	}
	_initSheetRemove() {
		this.disposeWithMe(this._commandService.onCommandExecuted((commandInfo) => {
			if (commandInfo.id === RemoveSheetMutation.id) {
				var _this$_cacheMatrix$ge;
				const { unitId, subUnitId } = commandInfo.params;
				(_this$_cacheMatrix$ge = this._cacheMatrix.get(unitId)) === null || _this$_cacheMatrix$ge === void 0 || _this$_cacheMatrix$ge.delete(subUnitId);
			}
		}));
		this.disposeWithMe(this._univerInstanceService.unitDisposed$.subscribe((univerInstance) => {
			if (univerInstance.type === UniverInstanceType.UNIVER_SHEET) this._cacheMatrix.delete(univerInstance.getUnitId());
		}));
	}
	_ensureCache(unitId, subUnitId) {
		let unitMap = this._cacheMatrix.get(unitId);
		if (!unitMap) {
			unitMap = /* @__PURE__ */ new Map();
			this._cacheMatrix.set(unitId, unitMap);
		}
		let cacheMatrix = unitMap.get(subUnitId);
		if (!cacheMatrix) {
			cacheMatrix = new ObjectMatrix();
			unitMap.set(subUnitId, cacheMatrix);
		}
		return cacheMatrix;
	}
	ensureCache(unitId, subUnitId) {
		return this._ensureCache(unitId, subUnitId);
	}
	addRule(unitId, subUnitId, rule) {
		this.markRangeDirty(unitId, subUnitId, rule.ranges);
	}
	removeRule(unitId, subUnitId, rule) {
		this._deleteRange(unitId, subUnitId, rule.ranges);
	}
	markRangeDirty(unitId, subUnitId, ranges, isSetRange) {
		const cache = this._ensureCache(unitId, subUnitId);
		ranges.forEach((range) => {
			Range.foreach(range, (row, col) => {
				if (cache.getValue(row, col) !== void 0) cache.setValue(row, col, void 0);
			});
		});
		this._dirtyRanges$.next({
			unitId,
			subUnitId,
			ranges,
			isSetRange
		});
	}
	_deleteRange(unitId, subUnitId, ranges) {
		const cache = this._ensureCache(unitId, subUnitId);
		ranges.forEach((range) => {
			Range.foreach(range, (row, col) => {
				cache.realDeleteValue(row, col);
			});
		});
		this._dirtyRanges$.next({
			unitId,
			subUnitId,
			ranges
		});
	}
	getValue(unitId, subUnitId, row, col) {
		return this._ensureCache(unitId, subUnitId).getValue(row, col);
	}
};
DataValidationCacheService = __decorate([
	__decorateParam(0, Inject(ICommandService)),
	__decorateParam(1, Inject(IUniverInstanceService)),
	__decorateParam(2, Inject(DataValidationModel))
], DataValidationCacheService);

//#endregion
//#region src/utils/formula.ts
function getFormulaResult(result) {
	var _result$;
	return result === null || result === void 0 || (_result$ = result[0]) === null || _result$ === void 0 || (_result$ = _result$[0]) === null || _result$ === void 0 ? void 0 : _result$.v;
}
function getFormulaCellData(result) {
	var _result$2;
	return result === null || result === void 0 || (_result$2 = result[0]) === null || _result$2 === void 0 ? void 0 : _result$2[0];
}
function isLegalFormulaResult(res) {
	return !ERROR_TYPE_SET.has(res);
}
/**
* Judge if the data-validation's formula need to be offseted by ranges
*/
function shouldOffsetFormulaByRange(type, validatorRegistryService) {
	var _validator$offsetForm;
	const validator = validatorRegistryService.getValidatorItem(type);
	return (_validator$offsetForm = validator === null || validator === void 0 ? void 0 : validator.offsetFormulaByRange) !== null && _validator$offsetForm !== void 0 ? _validator$offsetForm : false;
}

//#endregion
//#region src/services/dv-custom-formula.service.ts
let DataValidationCustomFormulaService = class DataValidationCustomFormulaService extends Disposable {
	constructor(_instanceSrv, _registerOtherFormulaService, _dataValidationModel, _dataValidationCacheService, _validatorRegistryService) {
		super();
		this._instanceSrv = _instanceSrv;
		this._registerOtherFormulaService = _registerOtherFormulaService;
		this._dataValidationModel = _dataValidationModel;
		this._dataValidationCacheService = _dataValidationCacheService;
		this._validatorRegistryService = _validatorRegistryService;
		_defineProperty(this, "_ruleFormulaMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_ruleFormulaMap2", /* @__PURE__ */ new Map());
		this._initFormulaResultHandler();
		this._initDirtyRanges();
	}
	dispose() {
		super.dispose();
		this._ruleFormulaMap.clear();
		this._ruleFormulaMap2.clear();
	}
	_initFormulaResultHandler() {
		this.disposeWithMe(this._registerOtherFormulaService.formulaResult$.subscribe((resultMap) => {
			for (const unitId in resultMap) {
				const unitMap = resultMap[unitId];
				if (this._instanceSrv.getUnitType(unitId) !== UniverInstanceType.UNIVER_SHEET) continue;
				for (const subUnitId in unitMap) {
					const results = unitMap[subUnitId];
					const { ruleFormulaMap } = this._ensureMaps(unitId, subUnitId);
					results.forEach((result) => {
						var _result$extra, _result$extra2;
						const ruleInfo = ruleFormulaMap.get((_result$extra = result.extra) === null || _result$extra === void 0 ? void 0 : _result$extra.ruleId);
						const rule = this._dataValidationModel.getRuleById(unitId, subUnitId, (_result$extra2 = result.extra) === null || _result$extra2 === void 0 ? void 0 : _result$extra2.ruleId);
						if (rule && ruleInfo) this._dataValidationCacheService.markRangeDirty(unitId, subUnitId, rule.ranges);
					});
				}
			}
		}));
	}
	_ensureMaps(unitId, subUnitId) {
		let ruleFormulaUnitMap = this._ruleFormulaMap.get(unitId);
		let ruleFormulaUnitMap2 = this._ruleFormulaMap2.get(unitId);
		if (!ruleFormulaUnitMap) {
			ruleFormulaUnitMap = /* @__PURE__ */ new Map();
			this._ruleFormulaMap.set(unitId, ruleFormulaUnitMap);
		}
		if (!ruleFormulaUnitMap2) {
			ruleFormulaUnitMap2 = /* @__PURE__ */ new Map();
			this._ruleFormulaMap2.set(unitId, ruleFormulaUnitMap2);
		}
		let ruleFormulaMap = ruleFormulaUnitMap.get(subUnitId);
		if (!ruleFormulaMap) {
			ruleFormulaMap = /* @__PURE__ */ new Map();
			ruleFormulaUnitMap.set(subUnitId, ruleFormulaMap);
		}
		let ruleFormulaMap2 = ruleFormulaUnitMap2.get(subUnitId);
		if (!ruleFormulaMap2) {
			ruleFormulaMap2 = /* @__PURE__ */ new Map();
			ruleFormulaUnitMap2.set(subUnitId, ruleFormulaMap2);
		}
		return {
			ruleFormulaMap,
			ruleFormulaMap2
		};
	}
	_registerFormula(unitId, subUnitId, ruleId, formulaString, ranges) {
		return this._registerOtherFormulaService.registerFormulaWithRange(unitId, subUnitId, formulaString, ranges, { ruleId }, OtherFormulaBizType.DATA_VALIDATION_CUSTOM, ruleId);
	}
	_handleDirtyRanges(unitId, subUnitId, ranges) {
		this._dataValidationModel.getRules(unitId, subUnitId).forEach((rule) => {
			const ruleRanges = rule.ranges;
			if (Rectangle.doAnyRangesIntersect(ruleRanges, ranges)) this.makeRuleDirty(unitId, subUnitId, rule.uid);
		});
	}
	_initDirtyRanges() {
		this.disposeWithMe(this._dataValidationCacheService.dirtyRanges$.subscribe((data) => {
			if (data.isSetRange) this._handleDirtyRanges(data.unitId, data.subUnitId, data.ranges);
		}));
	}
	deleteByRuleId(unitId, subUnitId, ruleId) {
		const { ruleFormulaMap, ruleFormulaMap2 } = this._ensureMaps(unitId, subUnitId);
		const rule = this._dataValidationModel.getRuleById(unitId, subUnitId, ruleId);
		const formulaInfo = ruleFormulaMap.get(ruleId);
		if (!rule || !formulaInfo) return;
		const current = ruleFormulaMap.get(ruleId);
		if (current) {
			ruleFormulaMap.delete(ruleId);
			this._registerOtherFormulaService.deleteFormula(unitId, subUnitId, [current.formulaId]);
		}
		const current2 = ruleFormulaMap2.get(ruleId);
		if (current2) {
			ruleFormulaMap2.delete(ruleId);
			this._registerOtherFormulaService.deleteFormula(unitId, subUnitId, [current2.formulaId]);
		}
	}
	_addFormulaByRange(unitId, subUnitId, ruleId, formula, formula2, ranges) {
		const { ruleFormulaMap, ruleFormulaMap2 } = this._ensureMaps(unitId, subUnitId);
		const originRow = ranges[0].startRow;
		const originCol = ranges[0].startColumn;
		if (formula && isFormulaString(formula)) {
			const formulaId = this._registerFormula(unitId, subUnitId, ruleId, formula, ranges);
			ruleFormulaMap.set(ruleId, {
				formula,
				originCol,
				originRow,
				formulaId
			});
		}
		if (formula2 && isFormulaString(formula2)) {
			const formulaId2 = this._registerFormula(unitId, subUnitId, ruleId, formula2, ranges);
			ruleFormulaMap2.set(ruleId, {
				formula: formula2,
				originCol,
				originRow,
				formulaId: formulaId2
			});
		}
	}
	addRule(unitId, subUnitId, rule) {
		if (shouldOffsetFormulaByRange(rule.type, this._validatorRegistryService)) {
			const { ranges, formula1, formula2, uid: ruleId } = rule;
			this._addFormulaByRange(unitId, subUnitId, ruleId, formula1, formula2, ranges);
		}
	}
	async getCellFormulaValue(unitId, subUnitId, ruleId, row, column) {
		var _result$result;
		const { ruleFormulaMap } = this._ensureMaps(unitId, subUnitId);
		const current = ruleFormulaMap.get(ruleId);
		if (!current) return Promise.resolve(void 0);
		const result = await this._registerOtherFormulaService.getFormulaValue(unitId, subUnitId, current.formulaId);
		const { originRow, originCol } = current;
		const offsetRow = row - originRow;
		const offsetCol = column - originCol;
		return getFormulaCellData(result === null || result === void 0 || (_result$result = result.result) === null || _result$result === void 0 || (_result$result = _result$result[offsetRow]) === null || _result$result === void 0 ? void 0 : _result$result[offsetCol]);
	}
	async getCellFormula2Value(unitId, subUnitId, ruleId, row, column) {
		var _result$result2;
		const { ruleFormulaMap2 } = this._ensureMaps(unitId, subUnitId);
		const current = ruleFormulaMap2.get(ruleId);
		if (!current) return Promise.resolve(void 0);
		const result = await this._registerOtherFormulaService.getFormulaValue(unitId, subUnitId, current.formulaId);
		const { originRow, originCol } = current;
		const offsetRow = row - originRow;
		const offsetCol = column - originCol;
		return getFormulaCellData(result === null || result === void 0 || (_result$result2 = result.result) === null || _result$result2 === void 0 || (_result$result2 = _result$result2[offsetRow]) === null || _result$result2 === void 0 ? void 0 : _result$result2[offsetCol]);
	}
	getCellFormulaValueSync(unitId, subUnitId, ruleId, row, column) {
		var _result$result3;
		const { ruleFormulaMap } = this._ensureMaps(unitId, subUnitId);
		const current = ruleFormulaMap.get(ruleId);
		if (!current) return;
		const result = this._registerOtherFormulaService.getFormulaValueSync(unitId, subUnitId, current.formulaId);
		const { originRow, originCol } = current;
		const offsetRow = row - originRow;
		const offsetCol = column - originCol;
		return getFormulaCellData(result === null || result === void 0 || (_result$result3 = result.result) === null || _result$result3 === void 0 || (_result$result3 = _result$result3[offsetRow]) === null || _result$result3 === void 0 ? void 0 : _result$result3[offsetCol]);
	}
	getCellFormula2ValueSync(unitId, subUnitId, ruleId, row, column) {
		var _result$result4;
		const { ruleFormulaMap2 } = this._ensureMaps(unitId, subUnitId);
		const current = ruleFormulaMap2.get(ruleId);
		if (!current) return;
		const result = this._registerOtherFormulaService.getFormulaValueSync(unitId, subUnitId, current.formulaId);
		const { originRow, originCol } = current;
		const offsetRow = row - originRow;
		const offsetCol = column - originCol;
		return getFormulaCellData(result === null || result === void 0 || (_result$result4 = result.result) === null || _result$result4 === void 0 || (_result$result4 = _result$result4[offsetRow]) === null || _result$result4 === void 0 ? void 0 : _result$result4[offsetCol]);
	}
	getRuleFormulaInfo(unitId, subUnitId, ruleId) {
		const { ruleFormulaMap } = this._ensureMaps(unitId, subUnitId);
		return ruleFormulaMap.get(ruleId);
	}
	makeRuleDirty(unitId, subUnitId, ruleId) {
		var _this$_ruleFormulaMap, _this$_ruleFormulaMap2;
		const formula1 = (_this$_ruleFormulaMap = this._ruleFormulaMap.get(unitId)) === null || _this$_ruleFormulaMap === void 0 || (_this$_ruleFormulaMap = _this$_ruleFormulaMap.get(subUnitId)) === null || _this$_ruleFormulaMap === void 0 ? void 0 : _this$_ruleFormulaMap.get(ruleId);
		const formula2 = (_this$_ruleFormulaMap2 = this._ruleFormulaMap2.get(unitId)) === null || _this$_ruleFormulaMap2 === void 0 || (_this$_ruleFormulaMap2 = _this$_ruleFormulaMap2.get(subUnitId)) === null || _this$_ruleFormulaMap2 === void 0 ? void 0 : _this$_ruleFormulaMap2.get(ruleId);
		if (formula1) this._registerOtherFormulaService.markFormulaDirty(unitId, subUnitId, formula1.formulaId);
		if (formula2) this._registerOtherFormulaService.markFormulaDirty(unitId, subUnitId, formula2.formulaId);
	}
};
DataValidationCustomFormulaService = __decorate([
	__decorateParam(0, IUniverInstanceService),
	__decorateParam(1, Inject(RegisterOtherFormulaService)),
	__decorateParam(2, Inject(DataValidationModel)),
	__decorateParam(3, Inject(DataValidationCacheService)),
	__decorateParam(4, Inject(DataValidatorRegistryService))
], DataValidationCustomFormulaService);

//#endregion
//#region src/utils/get-cell-data-origin.ts
function getCellValueOrigin(cell) {
	return getOriginCellValue(cell);
}
function getStringCellValue(cell) {
	var _getCellValueOrigin;
	return String((_getCellValueOrigin = getCellValueOrigin(cell)) !== null && _getCellValueOrigin !== void 0 ? _getCellValueOrigin : "");
}

//#endregion
//#region src/services/dv-list-cache.service.ts
let DataValidationListCacheService = class DataValidationListCacheService extends Disposable {
	constructor(_injector, _dataValidationModel) {
		super();
		this._injector = _injector;
		this._dataValidationModel = _dataValidationModel;
		_defineProperty(this, "_cache", /* @__PURE__ */ new Map());
		this._initRuleChangeListener();
	}
	_initRuleChangeListener() {
		this.disposeWithMe(this._dataValidationModel.ruleChange$.subscribe((change) => {
			if (change.type === "remove" || change.type === "update") this.markRuleDirty(change.unitId, change.subUnitId, change.rule.uid);
		}));
	}
	/**
	* Get cached list data or compute and cache it if not exists.
	*/
	getOrCompute(unitId, subUnitId, rule) {
		const cached = this.getCache(unitId, subUnitId, rule.uid);
		if (cached) return cached;
		const results = this._injector.get(DataValidationFormulaService).getRuleFormulaResultSync(unitId, subUnitId, rule.uid);
		return this.computeAndCache(unitId, subUnitId, rule, results);
	}
	_ensureCache(unitId, subUnitId) {
		let unitMap = this._cache.get(unitId);
		if (!unitMap) {
			unitMap = /* @__PURE__ */ new Map();
			this._cache.set(unitId, unitMap);
		}
		let subUnitMap = unitMap.get(subUnitId);
		if (!subUnitMap) {
			subUnitMap = /* @__PURE__ */ new Map();
			unitMap.set(subUnitId, subUnitMap);
		}
		return subUnitMap;
	}
	/**
	* Get cached list data for a rule. Returns undefined if not cached.
	*/
	getCache(unitId, subUnitId, ruleId) {
		var _this$_cache$get;
		return (_this$_cache$get = this._cache.get(unitId)) === null || _this$_cache$get === void 0 || (_this$_cache$get = _this$_cache$get.get(subUnitId)) === null || _this$_cache$get === void 0 ? void 0 : _this$_cache$get.get(ruleId);
	}
	/**
	* Set cache for a rule.
	*/
	setCache(unitId, subUnitId, ruleId, item) {
		this._ensureCache(unitId, subUnitId).set(ruleId, item);
	}
	/**
	* Mark a rule's cache as dirty (invalidate it).
	* Called when formula results change.
	*/
	markRuleDirty(unitId, subUnitId, ruleId) {
		var _this$_cache$get2;
		(_this$_cache$get2 = this._cache.get(unitId)) === null || _this$_cache$get2 === void 0 || (_this$_cache$get2 = _this$_cache$get2.get(subUnitId)) === null || _this$_cache$get2 === void 0 || _this$_cache$get2.delete(ruleId);
	}
	/**
	* Clear all caches.
	*/
	clear() {
		this._cache.clear();
	}
	/**
	* Compute list data from formula result and cache it.
	*/
	computeAndCache(unitId, subUnitId, rule, formulaResult) {
		var _formulaResult$;
		const { formula1 = "", formula2 = "" } = rule;
		const list = isFormulaString(formula1) ? this._getRuleFormulaResultSet(formulaResult === null || formulaResult === void 0 || (_formulaResult$ = formulaResult[0]) === null || _formulaResult$ === void 0 || (_formulaResult$ = _formulaResult$.result) === null || _formulaResult$ === void 0 || (_formulaResult$ = _formulaResult$[0]) === null || _formulaResult$ === void 0 ? void 0 : _formulaResult$[0]) : deserializeListOptions(formula1);
		const colorList = formula2.split(",");
		const listWithColor = list.map((label, i) => ({
			label,
			color: colorList[i] || ""
		}));
		const colorMap = {};
		for (const item of listWithColor) if (item.color) colorMap[item.label] = item.color;
		const cacheItem = {
			list,
			listWithColor,
			colorMap,
			set: new Set(list)
		};
		this.setCache(unitId, subUnitId, rule.uid, cacheItem);
		return cacheItem;
	}
	/**
	* Extract string list from formula result cells.
	*/
	_getRuleFormulaResultSet(result) {
		if (!result) return [];
		const resultSet = /* @__PURE__ */ new Set();
		for (let i = 0, rowLen = result.length; i < rowLen; i++) {
			const row = result[i];
			if (!row) continue;
			for (let j = 0, colLen = row.length; j < colLen; j++) {
				const cell = row[j];
				const value = getCellValueOrigin(cell);
				if (value !== null && value !== void 0) {
					var _cell$s;
					if (typeof value !== "string" && typeof (cell === null || cell === void 0 ? void 0 : cell.s) === "object" && ((_cell$s = cell.s) === null || _cell$s === void 0 || (_cell$s = _cell$s.n) === null || _cell$s === void 0 ? void 0 : _cell$s.pattern)) {
						resultSet.add(numfmt.format(cell.s.n.pattern, value, { throws: false }));
						continue;
					}
					const valueStr = typeof value === "string" ? value : String(value);
					if (isLegalFormulaResult(valueStr)) resultSet.add(valueStr);
				}
			}
		}
		return [...resultSet];
	}
};
DataValidationListCacheService = __decorate([__decorateParam(0, Inject(Injector)), __decorateParam(1, Inject(DataValidationModel))], DataValidationListCacheService);

//#endregion
//#region src/services/dv-formula.service.ts
let DataValidationFormulaService = class DataValidationFormulaService extends Disposable {
	constructor(_instanceService, _registerOtherFormulaService, _dataValidationCacheService, _dataValidationModel, _validatorRegistryService, _listCacheService) {
		super();
		this._instanceService = _instanceService;
		this._registerOtherFormulaService = _registerOtherFormulaService;
		this._dataValidationCacheService = _dataValidationCacheService;
		this._dataValidationModel = _dataValidationModel;
		this._validatorRegistryService = _validatorRegistryService;
		this._listCacheService = _listCacheService;
		_defineProperty(this, "_formulaRuleMap", /* @__PURE__ */ new Map());
		this._initFormulaResultHandler();
	}
	_initFormulaResultHandler() {
		this.disposeWithMe(this._registerOtherFormulaService.formulaResult$.subscribe((resultMap) => {
			for (const unitId in resultMap) {
				const unitMap = resultMap[unitId];
				if (this._instanceService.getUnitType(unitId) !== UniverInstanceType.UNIVER_SHEET) continue;
				for (const subUnitId in unitMap) {
					const results = unitMap[subUnitId];
					const formulaMap = this._ensureRuleFormulaMap(unitId, subUnitId);
					results.forEach((result) => {
						var _result$extra;
						const ruleId = (_result$extra = result.extra) === null || _result$extra === void 0 ? void 0 : _result$extra.ruleId;
						if (ruleId && formulaMap.get(ruleId)) {
							const rule = this._dataValidationModel.getRuleById(unitId, subUnitId, ruleId);
							if (rule) {
								this._listCacheService.markRuleDirty(unitId, subUnitId, ruleId);
								this._dataValidationCacheService.markRangeDirty(unitId, subUnitId, rule.ranges);
							}
						}
					});
				}
			}
		}));
	}
	_ensureRuleFormulaMap(unitId, subUnitId) {
		let unitMap = this._formulaRuleMap.get(unitId);
		if (!unitMap) {
			unitMap = /* @__PURE__ */ new Map();
			this._formulaRuleMap.set(unitId, unitMap);
		}
		let subUnitMap = unitMap.get(subUnitId);
		if (!subUnitMap) {
			subUnitMap = /* @__PURE__ */ new Map();
			unitMap.set(subUnitId, subUnitMap);
		}
		return subUnitMap;
	}
	_registerSingleFormula(unitId, subUnitId, formula, ruleId) {
		return this._registerOtherFormulaService.registerFormulaWithRange(unitId, subUnitId, formula, [{
			startColumn: 0,
			endColumn: 0,
			startRow: 0,
			endRow: 0
		}], { ruleId }, OtherFormulaBizType.DATA_VALIDATION, ruleId);
	}
	addRule(unitId, subUnitId, rule) {
		if (!shouldOffsetFormulaByRange(rule.type, this._validatorRegistryService) && rule.type !== DataValidationType.CHECKBOX) {
			const { formula1, formula2, uid: ruleId } = rule;
			const isFormula1Legal = isFormulaString(formula1);
			const isFormula2Legal = isFormulaString(formula2);
			if (!isFormula1Legal && !isFormula2Legal) return;
			const formulaRuleMap = this._ensureRuleFormulaMap(unitId, subUnitId);
			const item = [void 0, void 0];
			if (isFormula1Legal) item[0] = {
				id: this._registerSingleFormula(unitId, subUnitId, formula1, ruleId),
				text: formula1
			};
			if (isFormula2Legal) item[1] = {
				id: this._registerSingleFormula(unitId, subUnitId, formula2, ruleId),
				text: formula2
			};
			formulaRuleMap.set(ruleId, item);
		}
	}
	removeRule(unitId, subUnitId, ruleId) {
		const item = this._ensureRuleFormulaMap(unitId, subUnitId).get(ruleId);
		if (!item) return;
		const [formula1, formula2] = item;
		const idList = [formula1 === null || formula1 === void 0 ? void 0 : formula1.id, formula2 === null || formula2 === void 0 ? void 0 : formula2.id].filter(Boolean);
		idList.length && this._registerOtherFormulaService.deleteFormula(unitId, subUnitId, idList);
	}
	getRuleFormulaResult(unitId, subUnitId, ruleId) {
		const formulaInfo = this._ensureRuleFormulaMap(unitId, subUnitId).get(ruleId);
		if (!formulaInfo) return Promise.resolve(null);
		const getResult = async (info) => info && this._registerOtherFormulaService.getFormulaValue(unitId, subUnitId, info.id);
		return Promise.all([getResult(formulaInfo[0]), getResult(formulaInfo[1])]);
	}
	getRuleFormulaResultSync(unitId, subUnitId, ruleId) {
		const formulaInfo = this._ensureRuleFormulaMap(unitId, subUnitId).get(ruleId);
		if (!formulaInfo) return;
		return formulaInfo.map((i) => {
			if (i) return this._registerOtherFormulaService.getFormulaValueSync(unitId, subUnitId, i.id);
		});
	}
	getRuleFormulaInfo(unitId, subUnitId, ruleId) {
		return this._ensureRuleFormulaMap(unitId, subUnitId).get(ruleId);
	}
};
DataValidationFormulaService = __decorate([
	__decorateParam(0, IUniverInstanceService),
	__decorateParam(1, Inject(RegisterOtherFormulaService)),
	__decorateParam(2, Inject(DataValidationCacheService)),
	__decorateParam(3, Inject(DataValidationModel)),
	__decorateParam(4, Inject(DataValidatorRegistryService)),
	__decorateParam(5, Inject(DataValidationListCacheService))
], DataValidationFormulaService);

//#endregion
//#region src/models/rule-matrix.ts
var RuleMatrix = class RuleMatrix {
	constructor(value, _unitId, _subUnitId, _univerInstanceService, _disableTree = false) {
		this._unitId = _unitId;
		this._subUnitId = _subUnitId;
		this._univerInstanceService = _univerInstanceService;
		this._disableTree = _disableTree;
		_defineProperty(this, "_map", void 0);
		_defineProperty(this, "_tree", new RBush());
		_defineProperty(this, "_dirty", true);
		_defineProperty(this, "_buildTree", () => {
			if (!this._dirty || this._disableTree) return;
			this._tree.clear();
			const items = [];
			this._map.forEach((ranges, ruleId) => {
				ranges.forEach((range) => {
					items.push({
						minX: range.startRow,
						maxX: range.endRow,
						minY: range.startColumn,
						maxY: range.endColumn,
						ruleId
					});
				});
			});
			this._tree.load(items);
			this._dirty = false;
		});
		_defineProperty(this, "_debonceBuildTree", debounce(this._buildTree, 0));
		this._map = value;
		this._buildTree();
	}
	get _worksheet() {
		var _this$_univerInstance;
		return (_this$_univerInstance = this._univerInstanceService.getUnit(this._unitId, UniverInstanceType.UNIVER_SHEET)) === null || _this$_univerInstance === void 0 ? void 0 : _this$_univerInstance.getSheetBySheetId(this._subUnitId);
	}
	_addRule(ruleId, _ranges) {
		if (!this._worksheet) return;
		const ranges = Rectangle.mergeRanges(_ranges.map((range) => Range.transformRange(range, this._worksheet)));
		this._map.forEach((value, key) => {
			const newRanges = Rectangle.subtractMulti(value, ranges);
			if (newRanges.length === 0) this._map.delete(key);
			else this._map.set(key, newRanges);
		});
		this._dirty = true;
		this._map.set(ruleId, ranges);
		this._debonceBuildTree();
	}
	addRule(rule) {
		this._addRule(rule.uid, rule.ranges);
	}
	removeRange(_ranges) {
		if (!this._worksheet) return;
		const ranges = _ranges.map((range) => Range.transformRange(range, this._worksheet));
		this._map.forEach((value, key) => {
			const newRanges = Rectangle.subtractMulti(value, ranges);
			if (newRanges.length === 0) this._map.delete(key);
			else this._map.set(key, newRanges);
		});
		this._dirty = true;
		this._debonceBuildTree();
	}
	_removeRule(ruleId) {
		this._map.delete(ruleId);
		this._dirty = true;
		this._debonceBuildTree();
	}
	removeRule(rule) {
		this._removeRule(rule.uid);
	}
	updateRange(ruleId, _newRanges) {
		this._removeRule(ruleId);
		this._addRule(ruleId, _newRanges);
	}
	addRangeRules(rules) {
		rules.forEach(({ id: ruleId, ranges }) => {
			if (!ranges.length) return;
			let current = this._map.get(ruleId);
			if (!current) {
				current = ranges;
				this._map.set(ruleId, current);
			} else {
				this._map.set(ruleId, Rectangle.mergeRanges([...current, ...ranges]));
				current = this._map.get(ruleId);
			}
			this._map.forEach((value, key) => {
				if (key === ruleId) return;
				const newRanges = Rectangle.subtractMulti(value, ranges);
				if (newRanges.length === 0) this._map.delete(key);
				else this._map.set(key, newRanges);
			});
		});
		this._dirty = true;
		this._debonceBuildTree();
	}
	diff(rules) {
		const mutations = [];
		let deleteIndex = 0;
		rules.forEach((rule, index) => {
			var _this$_map$get;
			const newRanges = (_this$_map$get = this._map.get(rule.uid)) !== null && _this$_map$get !== void 0 ? _this$_map$get : [];
			const oldRanges = rule.ranges;
			if (newRanges.length !== 0 && (newRanges.length !== oldRanges.length || newRanges.some((range, i) => !Rectangle.equals(range, oldRanges[i])))) mutations.push({
				type: "update",
				ruleId: rule.uid,
				oldRanges,
				newRanges: Rectangle.sort(newRanges),
				rule
			});
			if (newRanges.length === 0) {
				mutations.push({
					type: "delete",
					rule,
					index: index - deleteIndex
				});
				deleteIndex++;
			}
		});
		return mutations;
	}
	diffWithAddition(rules, additionRules) {
		const mutations = [];
		let deleteIndex = 0;
		rules.forEach((rule, index) => {
			var _this$_map$get2;
			const newRanges = (_this$_map$get2 = this._map.get(rule.uid)) !== null && _this$_map$get2 !== void 0 ? _this$_map$get2 : [];
			const oldRanges = rule.ranges;
			if (newRanges.length !== 0 && (newRanges.length !== oldRanges.length || newRanges.some((range, i) => !Rectangle.equals(range, oldRanges[i])))) mutations.push({
				type: "update",
				ruleId: rule.uid,
				oldRanges,
				newRanges: Rectangle.sort(newRanges),
				rule
			});
			if (newRanges.length === 0) {
				mutations.push({
					type: "delete",
					rule,
					index: index - deleteIndex
				});
				deleteIndex++;
			}
		});
		Array.from(additionRules).forEach((rule) => {
			var _this$_map$get3;
			const newRanges = (_this$_map$get3 = this._map.get(rule.uid)) !== null && _this$_map$get3 !== void 0 ? _this$_map$get3 : [];
			mutations.push({
				type: "add",
				rule: {
					...rule,
					ranges: Rectangle.sort(newRanges)
				}
			});
		});
		return mutations;
	}
	clone() {
		return new RuleMatrix(new Map(Tools.deepClone(Array.from(this._map.entries()))), this._unitId, this._subUnitId, this._univerInstanceService, true);
	}
	getValue(row, col) {
		if (this._dirty) this._buildTree();
		const result = this._tree.search({
			minX: row,
			maxX: row,
			minY: col,
			maxY: col
		});
		return result.length > 0 ? result[0].ruleId : void 0;
	}
};

//#endregion
//#region src/models/sheet-data-validation-model.ts
let SheetDataValidationModel = class SheetDataValidationModel extends Disposable {
	constructor(_dataValidationModel, _univerInstanceService, _dataValidatorRegistryService, _dataValidationCacheService, _dataValidationFormulaService, _dataValidationCustomFormulaService, _commandService) {
		super();
		this._dataValidationModel = _dataValidationModel;
		this._univerInstanceService = _univerInstanceService;
		this._dataValidatorRegistryService = _dataValidatorRegistryService;
		this._dataValidationCacheService = _dataValidationCacheService;
		this._dataValidationFormulaService = _dataValidationFormulaService;
		this._dataValidationCustomFormulaService = _dataValidationCustomFormulaService;
		this._commandService = _commandService;
		_defineProperty(this, "_ruleMatrixMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "_validStatusChange$", new Subject());
		_defineProperty(this, "_ruleChange$", new Subject());
		_defineProperty(this, "ruleChange$", this._ruleChange$.asObservable());
		_defineProperty(this, "validStatusChange$", this._validStatusChange$.asObservable());
		this._initRuleUpdateListener();
		this.disposeWithMe(() => {
			this._ruleChange$.complete();
			this._validStatusChange$.complete();
		});
		this._initUniverInstanceListener();
	}
	_initUniverInstanceListener() {
		this.disposeWithMe(this._univerInstanceService.unitDisposed$.subscribe((unit) => {
			this._ruleMatrixMap.delete(unit.getUnitId());
		}));
		this.disposeWithMe(this._commandService.onCommandExecuted((command) => {
			if (command.id === RemoveSheetMutation.id) {
				const { unitId, subUnitId } = command.params;
				const subUnitMap = this._ruleMatrixMap.get(unitId);
				if (subUnitMap) subUnitMap.delete(subUnitId);
			}
		}));
	}
	_initRuleUpdateListener() {
		const allRules = this._dataValidationModel.getAll();
		for (const [unitId, subUnitMap] of allRules) for (const [subUnitId, rules] of subUnitMap) for (const rule of rules) {
			this._addRule(unitId, subUnitId, rule);
			this._ruleChange$.next({
				type: "add",
				unitId,
				subUnitId,
				rule,
				source: "patched"
			});
		}
		this.disposeWithMe(this._dataValidationModel.ruleChange$.subscribe((ruleChange) => {
			switch (ruleChange.type) {
				case "add":
					this._addRule(ruleChange.unitId, ruleChange.subUnitId, ruleChange.rule);
					break;
				case "update":
					this._updateRule(ruleChange.unitId, ruleChange.subUnitId, ruleChange.rule.uid, ruleChange.oldRule, ruleChange.updatePayload);
					break;
				case "remove":
					this._removeRule(ruleChange.unitId, ruleChange.subUnitId, ruleChange.rule);
					break;
			}
			this._ruleChange$.next(ruleChange);
		}));
	}
	_ensureRuleMatrix(unitId, subUnitId) {
		let unitMap = this._ruleMatrixMap.get(unitId);
		if (!unitMap) {
			unitMap = /* @__PURE__ */ new Map();
			this._ruleMatrixMap.set(unitId, unitMap);
		}
		let matrix = unitMap.get(subUnitId);
		if (!matrix) {
			matrix = new RuleMatrix(/* @__PURE__ */ new Map(), unitId, subUnitId, this._univerInstanceService);
			unitMap.set(subUnitId, matrix);
		}
		return matrix;
	}
	_addRuleSideEffect(unitId, subUnitId, rule) {
		this._ensureRuleMatrix(unitId, subUnitId).addRule(rule);
		this._dataValidationCacheService.addRule(unitId, subUnitId, rule);
		this._dataValidationFormulaService.addRule(unitId, subUnitId, rule);
		this._dataValidationCustomFormulaService.addRule(unitId, subUnitId, rule);
	}
	_addRule(unitId, subUnitId, rule) {
		(Array.isArray(rule) ? rule : [rule]).forEach((item) => {
			this._addRuleSideEffect(unitId, subUnitId, item);
		});
	}
	_updateRule(unitId, subUnitId, ruleId, oldRule, payload) {
		const ruleMatrix = this._ensureRuleMatrix(unitId, subUnitId);
		const newRule = {
			...oldRule,
			...payload.payload
		};
		if (payload.type === UpdateRuleType.RANGE) ruleMatrix.updateRange(ruleId, payload.payload);
		else if (payload.type === UpdateRuleType.ALL) ruleMatrix.updateRange(ruleId, payload.payload.ranges);
		this._dataValidationCacheService.removeRule(unitId, subUnitId, oldRule);
		this._dataValidationCacheService.addRule(unitId, subUnitId, newRule);
		this._dataValidationFormulaService.removeRule(unitId, subUnitId, oldRule.uid);
		this._dataValidationFormulaService.addRule(unitId, subUnitId, newRule);
		this._dataValidationCustomFormulaService.deleteByRuleId(unitId, subUnitId, ruleId);
		this._dataValidationCustomFormulaService.addRule(unitId, subUnitId, newRule);
	}
	_removeRule(unitId, subUnitId, oldRule) {
		this._ensureRuleMatrix(unitId, subUnitId).removeRule(oldRule);
		this._dataValidationCacheService.removeRule(unitId, subUnitId, oldRule);
		this._dataValidationCustomFormulaService.deleteByRuleId(unitId, subUnitId, oldRule.uid);
	}
	getValidator(type) {
		return this._dataValidatorRegistryService.getValidatorItem(type);
	}
	getRuleIdByLocation(unitId, subUnitId, row, col) {
		return this._ensureRuleMatrix(unitId, subUnitId).getValue(row, col);
	}
	getRuleByLocation(unitId, subUnitId, row, col) {
		const ruleId = this.getRuleIdByLocation(unitId, subUnitId, row, col);
		if (!ruleId) return;
		return this._dataValidationModel.getRuleById(unitId, subUnitId, ruleId);
	}
	validator(rule, pos, _onCompete) {
		const { col, row, unitId, subUnitId, worksheet } = pos;
		const onCompete = (status, changed) => {
			if (_onCompete) _onCompete(status, changed);
			if (changed) this._validStatusChange$.next({
				unitId,
				subUnitId,
				ruleId: rule.uid,
				status,
				row,
				col
			});
		};
		const cell = worksheet.getCellValueOnly(row, col);
		const validator = this.getValidator(rule.type);
		const cellRaw = worksheet.getCellRaw(row, col);
		const cellValue = getCellValueOrigin(cellRaw);
		if (validator) {
			const cache = this._dataValidationCacheService.ensureCache(unitId, subUnitId);
			const current = cache.getValue(row, col);
			if (current === null || current === void 0) {
				cache.setValue(row, col, DataValidationStatus.VALIDATING);
				validator.validator({
					value: cellValue,
					unitId,
					subUnitId,
					row,
					column: col,
					worksheet: pos.worksheet,
					workbook: pos.workbook,
					interceptValue: getCellValueOrigin(cell),
					t: cellRaw === null || cellRaw === void 0 ? void 0 : cellRaw.t
				}, rule).then((status) => {
					const realStatus = status ? DataValidationStatus.VALID : DataValidationStatus.INVALID;
					const now = cache.getValue(row, col);
					if (realStatus === DataValidationStatus.VALID) cache.realDeleteValue(row, col);
					else cache.setValue(row, col, realStatus);
					onCompete(realStatus, current !== now);
				});
				return DataValidationStatus.VALIDATING;
			}
			onCompete(current !== null && current !== void 0 ? current : DataValidationStatus.VALID, false);
			return current !== null && current !== void 0 ? current : DataValidationStatus.VALID;
		} else {
			onCompete(DataValidationStatus.VALID, false);
			return DataValidationStatus.VALID;
		}
	}
	getRuleObjectMatrix(unitId, subUnitId) {
		return this._ensureRuleMatrix(unitId, subUnitId);
	}
	getRuleById(unitId, subUnitId, ruleId) {
		return this._dataValidationModel.getRuleById(unitId, subUnitId, ruleId);
	}
	getRuleIndex(unitId, subUnitId, ruleId) {
		return this._dataValidationModel.getRuleIndex(unitId, subUnitId, ruleId);
	}
	getRules(unitId, subUnitId) {
		return [...this._dataValidationModel.getRules(unitId, subUnitId)];
	}
	getUnitRules(unitId) {
		return this._dataValidationModel.getUnitRules(unitId);
	}
	deleteUnitRules(unitId) {
		return this._dataValidationModel.deleteUnitRules(unitId);
	}
	getSubUnitIds(unitId) {
		return this._dataValidationModel.getSubUnitIds(unitId);
	}
	getAll() {
		return this._dataValidationModel.getAll();
	}
};
SheetDataValidationModel = __decorate([
	__decorateParam(0, Inject(DataValidationModel)),
	__decorateParam(1, IUniverInstanceService),
	__decorateParam(2, Inject(DataValidatorRegistryService)),
	__decorateParam(3, Inject(DataValidationCacheService)),
	__decorateParam(4, Inject(DataValidationFormulaService)),
	__decorateParam(5, Inject(DataValidationCustomFormulaService)),
	__decorateParam(6, ICommandService)
], SheetDataValidationModel);

//#endregion
//#region src/types/const/operator-text-map.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const OperatorTextMap = {
	[DataValidationOperator.BETWEEN]: "sheets-data-validation.operators.between",
	[DataValidationOperator.EQUAL]: "sheets-data-validation.operators.equal",
	[DataValidationOperator.GREATER_THAN]: "sheets-data-validation.operators.greaterThan",
	[DataValidationOperator.GREATER_THAN_OR_EQUAL]: "sheets-data-validation.operators.greaterThanOrEqual",
	[DataValidationOperator.LESS_THAN]: "sheets-data-validation.operators.lessThan",
	[DataValidationOperator.LESS_THAN_OR_EQUAL]: "sheets-data-validation.operators.lessThanOrEqual",
	[DataValidationOperator.NOT_BETWEEN]: "sheets-data-validation.operators.notBetween",
	[DataValidationOperator.NOT_EQUAL]: "sheets-data-validation.operators.notEqual"
};
const OperatorTitleMap = {
	[DataValidationOperator.BETWEEN]: "sheets-data-validation.ruleName.between",
	[DataValidationOperator.EQUAL]: "sheets-data-validation.ruleName.equal",
	[DataValidationOperator.GREATER_THAN]: "sheets-data-validation.ruleName.greaterThan",
	[DataValidationOperator.GREATER_THAN_OR_EQUAL]: "sheets-data-validation.ruleName.greaterThanOrEqual",
	[DataValidationOperator.LESS_THAN]: "sheets-data-validation.ruleName.lessThan",
	[DataValidationOperator.LESS_THAN_OR_EQUAL]: "sheets-data-validation.ruleName.lessThanOrEqual",
	[DataValidationOperator.NOT_BETWEEN]: "sheets-data-validation.ruleName.notBetween",
	[DataValidationOperator.NOT_EQUAL]: "sheets-data-validation.ruleName.notEqual",
	NONE: "sheets-data-validation.ruleName.legal"
};
const OperatorErrorTitleMap = {
	[DataValidationOperator.BETWEEN]: "sheets-data-validation.errorMsg.between",
	[DataValidationOperator.EQUAL]: "sheets-data-validation.errorMsg.equal",
	[DataValidationOperator.GREATER_THAN]: "sheets-data-validation.errorMsg.greaterThan",
	[DataValidationOperator.GREATER_THAN_OR_EQUAL]: "sheets-data-validation.errorMsg.greaterThanOrEqual",
	[DataValidationOperator.LESS_THAN]: "sheets-data-validation.errorMsg.lessThan",
	[DataValidationOperator.LESS_THAN_OR_EQUAL]: "sheets-data-validation.errorMsg.lessThanOrEqual",
	[DataValidationOperator.NOT_BETWEEN]: "sheets-data-validation.errorMsg.notBetween",
	[DataValidationOperator.NOT_EQUAL]: "sheets-data-validation.errorMsg.notEqual",
	NONE: "sheets-data-validation.errorMsg.legal"
};
const TextLengthErrorTitleMap = {
	[DataValidationOperator.BETWEEN]: "sheets-data-validation.textLength.errorMsg.between",
	[DataValidationOperator.EQUAL]: "sheets-data-validation.textLength.errorMsg.equal",
	[DataValidationOperator.GREATER_THAN]: "sheets-data-validation.textLength.errorMsg.greaterThan",
	[DataValidationOperator.GREATER_THAN_OR_EQUAL]: "sheets-data-validation.textLength.errorMsg.greaterThanOrEqual",
	[DataValidationOperator.LESS_THAN]: "sheets-data-validation.textLength.errorMsg.lessThan",
	[DataValidationOperator.LESS_THAN_OR_EQUAL]: "sheets-data-validation.textLength.errorMsg.lessThanOrEqual",
	[DataValidationOperator.NOT_BETWEEN]: "sheets-data-validation.textLength.errorMsg.notBetween",
	[DataValidationOperator.NOT_EQUAL]: "sheets-data-validation.textLength.errorMsg.notEqual"
};

//#endregion
//#region src/validators/base-sheet-validator.ts
const FORMULA1$1 = "{FORMULA1}";
const FORMULA2$1 = "{FORMULA2}";
const TYPE = "{TYPE}";
var BaseSheetValidator = class extends BaseDataValidator {
	get operatorNames() {
		return this.operators.map((operator) => this.localeService.t(OperatorTextMap[operator]));
	}
	generateRuleName(rule) {
		var _rule$formula, _rule$formula2;
		if (!rule.operator) return this.localeService.t(OperatorTitleMap.NONE).replace(TYPE, this.titleStr);
		const ruleName = this.localeService.t(OperatorTitleMap[rule.operator]).replace(FORMULA1$1, (_rule$formula = rule.formula1) !== null && _rule$formula !== void 0 ? _rule$formula : "").replace(FORMULA2$1, (_rule$formula2 = rule.formula2) !== null && _rule$formula2 !== void 0 ? _rule$formula2 : "");
		return `${this.titleStr} ${ruleName}`;
	}
	generateRuleErrorMessage(rule, _position) {
		var _rule$formula3, _rule$formula4;
		if (!rule.operator) return this.localeService.t(OperatorErrorTitleMap.NONE).replace(TYPE, this.titleStr);
		return `${this.localeService.t(OperatorErrorTitleMap[rule.operator]).replace(FORMULA1$1, (_rule$formula3 = rule.formula1) !== null && _rule$formula3 !== void 0 ? _rule$formula3 : "").replace(FORMULA2$1, (_rule$formula4 = rule.formula2) !== null && _rule$formula4 !== void 0 ? _rule$formula4 : "")}`;
	}
};

//#endregion
//#region src/validators/checkbox-validator.ts
const CHECKBOX_FORMULA_1 = 1;
const CHECKBOX_FORMULA_2 = 0;
function getFailMessage(formula, localeService) {
	if (Tools.isBlank(formula)) return localeService.t("sheets-data-validation.validFail.value");
	if (isFormulaString(formula)) return localeService.t("sheets-data-validation.validFail.primitive");
	return "";
}
const transformCheckboxValue = (value) => Tools.isDefine(value) && String(value).toLowerCase() === "true" ? "1" : String(value).toLowerCase() === "false" ? "0" : value;
var CheckboxValidator = class extends BaseSheetValidator {
	constructor(..._args) {
		super(..._args);
		_defineProperty(this, "id", DataValidationType.CHECKBOX);
		_defineProperty(this, "title", "sheets-data-validation.checkbox.title");
		_defineProperty(this, "operators", []);
		_defineProperty(this, "scopes", ["sheet"]);
		_defineProperty(this, "order", 41);
		_defineProperty(this, "offsetFormulaByRange", false);
		_defineProperty(this, "_formulaService", this.injector.get(DataValidationFormulaService));
		_defineProperty(this, "skipDefaultFontRender", (rule, cellValue, pos) => {
			const { unitId, subUnitId } = pos;
			const { formula1, formula2 } = this.parseFormulaSync(rule, unitId, subUnitId);
			const valueStr = `${cellValue !== null && cellValue !== void 0 ? cellValue : ""}`;
			return !valueStr || valueStr === `${formula1}` || valueStr === `${formula2}`;
		});
	}
	validatorFormula(rule, unitId, subUnitId) {
		const { formula1, formula2 } = rule;
		const isEqual = formula1 === formula2;
		if (Tools.isBlank(formula1) && Tools.isBlank(formula2)) return { success: true };
		if (isEqual) return {
			success: false,
			formula1: this.localeService.t("sheets-data-validation.validFail.checkboxEqual"),
			formula2: this.localeService.t("sheets-data-validation.validFail.checkboxEqual")
		};
		const error1 = getFailMessage(formula1, this.localeService);
		const error2 = getFailMessage(formula2, this.localeService);
		return {
			success: !error1 && !error2,
			formula1: error1,
			formula2: error2
		};
	}
	async parseFormula(rule, unitId, subUnitId) {
		var _results$, _results$2;
		const { formula1 = 1, formula2 = 0 } = rule;
		const results = await this._formulaService.getRuleFormulaResult(unitId, subUnitId, rule.uid);
		const originFormula1 = isFormulaString(formula1) ? getFormulaResult(results === null || results === void 0 || (_results$ = results[0]) === null || _results$ === void 0 || (_results$ = _results$.result) === null || _results$ === void 0 ? void 0 : _results$[0][0]) : formula1;
		const originFormula2 = isFormulaString(formula2) ? getFormulaResult(results === null || results === void 0 || (_results$2 = results[1]) === null || _results$2 === void 0 || (_results$2 = _results$2.result) === null || _results$2 === void 0 ? void 0 : _results$2[0][0]) : formula2;
		const isFormulaValid = isLegalFormulaResult(String(originFormula1)) && isLegalFormulaResult(String(originFormula2));
		return {
			formula1: transformCheckboxValue(originFormula1),
			formula2: transformCheckboxValue(originFormula2),
			originFormula1,
			originFormula2,
			isFormulaValid
		};
	}
	getExtraStyle(rule, value) {
		return { tb: WrapStrategy.CLIP };
	}
	parseFormulaSync(rule, unitId, subUnitId) {
		var _results$3, _results$4;
		const { formula1 = 1, formula2 = 0 } = rule;
		const results = this._formulaService.getRuleFormulaResultSync(unitId, subUnitId, rule.uid);
		const originFormula1 = isFormulaString(formula1) ? getFormulaResult(results === null || results === void 0 || (_results$3 = results[0]) === null || _results$3 === void 0 || (_results$3 = _results$3.result) === null || _results$3 === void 0 ? void 0 : _results$3[0][0]) : formula1;
		const originFormula2 = isFormulaString(formula2) ? getFormulaResult(results === null || results === void 0 || (_results$4 = results[1]) === null || _results$4 === void 0 || (_results$4 = _results$4.result) === null || _results$4 === void 0 ? void 0 : _results$4[0][0]) : formula2;
		const isFormulaValid = isLegalFormulaResult(String(originFormula1)) && isLegalFormulaResult(String(originFormula2));
		return {
			formula1: transformCheckboxValue(originFormula1),
			formula2: transformCheckboxValue(originFormula2),
			originFormula1,
			originFormula2,
			isFormulaValid
		};
	}
	async isValidType(cellInfo, formula, rule) {
		const { value, unitId, subUnitId } = cellInfo;
		const { formula1, formula2, originFormula1, originFormula2 } = await this.parseFormula(rule, unitId, subUnitId);
		if (!Tools.isDefine(formula1) || !Tools.isDefine(formula2)) return true;
		return Tools.isDefine(value) && (String(value) === String(formula1) || String(value) === String(formula2) || String(value) === String(originFormula1 !== null && originFormula1 !== void 0 ? originFormula1 : "") || String(value) === String(originFormula2 !== null && originFormula2 !== void 0 ? originFormula2 : ""));
	}
	generateRuleErrorMessage(rule) {
		return this.localeService.t("sheets-data-validation.checkbox.error");
	}
	generateRuleName(rule) {
		return this.titleStr;
	}
};

//#endregion
//#region src/common/date-text-map.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const DateOperatorNameMap = {
	[DataValidationOperator.BETWEEN]: "sheets-data-validation.date.operators.between",
	[DataValidationOperator.EQUAL]: "sheets-data-validation.date.operators.equal",
	[DataValidationOperator.GREATER_THAN]: "sheets-data-validation.date.operators.greaterThan",
	[DataValidationOperator.GREATER_THAN_OR_EQUAL]: "sheets-data-validation.date.operators.greaterThanOrEqual",
	[DataValidationOperator.LESS_THAN]: "sheets-data-validation.date.operators.lessThan",
	[DataValidationOperator.LESS_THAN_OR_EQUAL]: "sheets-data-validation.date.operators.lessThanOrEqual",
	[DataValidationOperator.NOT_BETWEEN]: "sheets-data-validation.date.operators.notBetween",
	[DataValidationOperator.NOT_EQUAL]: "sheets-data-validation.date.operators.notEqual"
};
const DateOperatorTextMap = {
	[DataValidationOperator.BETWEEN]: "sheets-data-validation.date.operators.between",
	[DataValidationOperator.EQUAL]: "sheets-data-validation.date.operators.equal",
	[DataValidationOperator.GREATER_THAN]: "sheets-data-validation.date.operators.greaterThan",
	[DataValidationOperator.GREATER_THAN_OR_EQUAL]: "sheets-data-validation.date.operators.greaterThanOrEqual",
	[DataValidationOperator.LESS_THAN]: "sheets-data-validation.date.operators.lessThan",
	[DataValidationOperator.LESS_THAN_OR_EQUAL]: "sheets-data-validation.date.operators.lessThanOrEqual",
	[DataValidationOperator.NOT_BETWEEN]: "sheets-data-validation.date.operators.notBetween",
	[DataValidationOperator.NOT_EQUAL]: "sheets-data-validation.date.operators.notEqual"
};
const DateOperatorTitleMap = {
	[DataValidationOperator.BETWEEN]: "sheets-data-validation.date.ruleName.between",
	[DataValidationOperator.EQUAL]: "sheets-data-validation.date.ruleName.equal",
	[DataValidationOperator.GREATER_THAN]: "sheets-data-validation.date.ruleName.greaterThan",
	[DataValidationOperator.GREATER_THAN_OR_EQUAL]: "sheets-data-validation.date.ruleName.greaterThanOrEqual",
	[DataValidationOperator.LESS_THAN]: "sheets-data-validation.date.ruleName.lessThan",
	[DataValidationOperator.LESS_THAN_OR_EQUAL]: "sheets-data-validation.date.ruleName.lessThanOrEqual",
	[DataValidationOperator.NOT_BETWEEN]: "sheets-data-validation.date.ruleName.notBetween",
	[DataValidationOperator.NOT_EQUAL]: "sheets-data-validation.date.ruleName.notEqual",
	NONE: "sheets-data-validation.date.ruleName.legal"
};
const DateOperatorErrorTitleMap = {
	[DataValidationOperator.BETWEEN]: "sheets-data-validation.date.errorMsg.between",
	[DataValidationOperator.EQUAL]: "sheets-data-validation.date.errorMsg.equal",
	[DataValidationOperator.GREATER_THAN]: "sheets-data-validation.date.errorMsg.greaterThan",
	[DataValidationOperator.GREATER_THAN_OR_EQUAL]: "sheets-data-validation.date.errorMsg.greaterThanOrEqual",
	[DataValidationOperator.LESS_THAN]: "sheets-data-validation.date.errorMsg.lessThan",
	[DataValidationOperator.LESS_THAN_OR_EQUAL]: "sheets-data-validation.date.errorMsg.lessThanOrEqual",
	[DataValidationOperator.NOT_BETWEEN]: "sheets-data-validation.date.errorMsg.notBetween",
	[DataValidationOperator.NOT_EQUAL]: "sheets-data-validation.date.errorMsg.notEqual",
	NONE: "sheets-data-validation.date.errorMsg.legal"
};

//#endregion
//#region src/types/const/two-formula-operators.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const TWO_FORMULA_OPERATOR_COUNT = [DataValidationOperator.BETWEEN, DataValidationOperator.NOT_BETWEEN];

//#endregion
//#region src/validators/const.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const FORMULA1 = "{FORMULA1}";
const FORMULA2 = "{FORMULA2}";

//#endregion
//#region src/validators/util.ts
function getDataValidationCellValue(cellData) {
	const cellValue = getCellValueOrigin(cellData);
	if (cellValue === void 0 || cellValue === null) return "";
	return cellValue.toString();
}
function getTransformedFormula(lexerTreeBuilder, rule, position) {
	const { formula1, formula2 } = rule;
	const originStartRow = rule.ranges[0].startRow;
	const originStartColumn = rule.ranges[0].startColumn;
	const offsetRow = position.row - originStartRow;
	const offsetColumn = position.col - originStartColumn;
	return {
		transformedFormula1: isFormulaString(formula1) ? lexerTreeBuilder.moveFormulaRefOffset(formula1, offsetColumn, offsetRow, true) : formula1,
		transformedFormula2: isFormulaString(formula2) ? lexerTreeBuilder.moveFormulaRefOffset(formula2, offsetColumn, offsetRow, true) : formula2
	};
}

//#endregion
//#region src/validators/date-validator.ts
const transformDate2SerialNumber = (value) => {
	var _numfmt$parseDate, _numfmt$parseDate2;
	if (value === void 0 || value === null || typeof value === "boolean") return;
	if (typeof value === "number" || !Number.isNaN(+value)) return +value;
	const v = (_numfmt$parseDate = numfmt.parseDate(value)) === null || _numfmt$parseDate === void 0 ? void 0 : _numfmt$parseDate.v;
	if (Tools.isDefine(v)) return v;
	return (_numfmt$parseDate2 = numfmt.parseDate(dateKit(value).format("YYYY-MM-DD HH:mm:ss"))) === null || _numfmt$parseDate2 === void 0 ? void 0 : _numfmt$parseDate2.v;
};
var DateValidator = class extends BaseSheetValidator {
	constructor(..._args) {
		super(..._args);
		_defineProperty(this, "id", DataValidationType.DATE);
		_defineProperty(this, "title", "sheets-data-validation.date.title");
		_defineProperty(this, "order", 40);
		_defineProperty(this, "operators", [
			DataValidationOperator.BETWEEN,
			DataValidationOperator.EQUAL,
			DataValidationOperator.GREATER_THAN,
			DataValidationOperator.GREATER_THAN_OR_EQUAL,
			DataValidationOperator.LESS_THAN,
			DataValidationOperator.LESS_THAN_OR_EQUAL,
			DataValidationOperator.NOT_BETWEEN,
			DataValidationOperator.NOT_EQUAL
		]);
		_defineProperty(this, "scopes", ["sheet"]);
		_defineProperty(this, "_customFormulaService", this.injector.get(DataValidationCustomFormulaService));
		_defineProperty(this, "_lexerTreeBuilder", this.injector.get(LexerTreeBuilder));
	}
	async parseFormula(rule, unitId, subUnitId, row, column) {
		const formulaResult1 = await this._customFormulaService.getCellFormulaValue(unitId, subUnitId, rule.uid, row, column);
		const formulaResult2 = await this._customFormulaService.getCellFormula2Value(unitId, subUnitId, rule.uid, row, column);
		const { formula1, formula2 } = rule;
		const isFormulaValid = isLegalFormulaResult(String(formulaResult1 === null || formulaResult1 === void 0 ? void 0 : formulaResult1.v)) && isLegalFormulaResult(String(formulaResult2 === null || formulaResult2 === void 0 ? void 0 : formulaResult2.v));
		return {
			formula1: transformDate2SerialNumber(isFormulaString(formula1) ? formulaResult1 === null || formulaResult1 === void 0 ? void 0 : formulaResult1.v : formula1),
			formula2: transformDate2SerialNumber(isFormulaString(formula2) ? formulaResult2 === null || formulaResult2 === void 0 ? void 0 : formulaResult2.v : formula2),
			isFormulaValid
		};
	}
	async isValidType(info) {
		const { interceptValue, value } = info;
		if (typeof value === "number" && typeof interceptValue === "string") return Boolean(numfmt.parseDate(interceptValue));
		if (typeof interceptValue === "string") return Boolean(numfmt.parseDate(interceptValue));
		return false;
	}
	_validatorSingleFormula(formula) {
		return !Tools.isBlank(formula) && (isFormulaString(formula) || !Number.isNaN(+formula) || Boolean(formula && numfmt.parseDate(formula)));
	}
	validatorFormula(rule, unitId, subUnitId) {
		const operator = rule.operator;
		if (!operator) return { success: true };
		const formula1Success = this._validatorSingleFormula(rule.formula1);
		const errorMsg = this.localeService.t("sheets-data-validation.validFail.date");
		if (TWO_FORMULA_OPERATOR_COUNT.includes(operator)) {
			const formula2Success = this._validatorSingleFormula(rule.formula2);
			return {
				success: formula1Success && formula2Success,
				formula1: formula1Success ? void 0 : errorMsg,
				formula2: formula2Success ? void 0 : errorMsg
			};
		}
		return {
			success: formula1Success,
			formula1: formula1Success ? void 0 : errorMsg
		};
	}
	normalizeFormula(rule, _unitId, _subUnitId) {
		const { formula1, formula2, bizInfo } = rule;
		const normlizeSingleFormula = (formula) => {
			if (!formula) return formula;
			let date;
			if (!Number.isNaN(+formula)) date = numfmt.dateFromSerial(+formula);
			else {
				var _numfmt$parseDate3;
				const res = (_numfmt$parseDate3 = numfmt.parseDate(formula)) === null || _numfmt$parseDate3 === void 0 ? void 0 : _numfmt$parseDate3.v;
				if (res === void 0 || res === null) return "";
				date = numfmt.dateFromSerial(res);
			}
			return dateKit(`${date[0]}/${date[1]}/${date[2]} ${date[3]}:${date[4]}:${date[5]}`).format((bizInfo === null || bizInfo === void 0 ? void 0 : bizInfo.showTime) ? "YYYY-MM-DD HH:mm:ss" : "YYYY-MM-DD");
		};
		return {
			formula1: isFormulaString(formula1) ? formula1 : normlizeSingleFormula(`${formula1}`),
			formula2: isFormulaString(formula2) ? formula2 : normlizeSingleFormula(`${formula2}`)
		};
	}
	transform(cellInfo, _formula, _rule) {
		const { value } = cellInfo;
		return {
			...cellInfo,
			value: transformDate2SerialNumber(value)
		};
	}
	get operatorNames() {
		return this.operators.map((operator) => this.localeService.t(DateOperatorNameMap[operator]));
	}
	generateRuleName(rule) {
		var _rule$formula, _rule$formula2;
		if (!rule.operator) return this.localeService.t(DateOperatorTitleMap.NONE);
		const ruleName = this.localeService.t(DateOperatorTitleMap[rule.operator]).replace(FORMULA1, (_rule$formula = rule.formula1) !== null && _rule$formula !== void 0 ? _rule$formula : "").replace(FORMULA2, (_rule$formula2 = rule.formula2) !== null && _rule$formula2 !== void 0 ? _rule$formula2 : "");
		return `${this.titleStr} ${ruleName}`;
	}
	generateRuleErrorMessage(rule, pos) {
		if (!rule.operator) return this.titleStr;
		const { transformedFormula1, transformedFormula2 } = getTransformedFormula(this._lexerTreeBuilder, rule, pos);
		return `${this.localeService.t(DateOperatorErrorTitleMap[rule.operator]).replace(FORMULA1, transformedFormula1 !== null && transformedFormula1 !== void 0 ? transformedFormula1 : "").replace(FORMULA2, transformedFormula2 !== null && transformedFormula2 !== void 0 ? transformedFormula2 : "")}`;
	}
};

//#endregion
//#region src/validators/decimal-validator.ts
function getCellValueNumber(cellValue) {
	return +cellValue;
}
var DecimalValidator = class extends BaseSheetValidator {
	constructor(..._args) {
		super(..._args);
		_defineProperty(this, "_customFormulaService", this.injector.get(DataValidationCustomFormulaService));
		_defineProperty(this, "id", DataValidationType.DECIMAL);
		_defineProperty(this, "_lexerTreeBuilder", this.injector.get(LexerTreeBuilder));
		_defineProperty(this, "title", "sheets-data-validation.decimal.title");
		_defineProperty(this, "order", 20);
		_defineProperty(this, "operators", [
			DataValidationOperator.BETWEEN,
			DataValidationOperator.EQUAL,
			DataValidationOperator.GREATER_THAN,
			DataValidationOperator.GREATER_THAN_OR_EQUAL,
			DataValidationOperator.LESS_THAN,
			DataValidationOperator.LESS_THAN_OR_EQUAL,
			DataValidationOperator.NOT_BETWEEN,
			DataValidationOperator.NOT_EQUAL
		]);
		_defineProperty(this, "scopes", ["sheet"]);
	}
	_isFormulaOrNumber(formula) {
		return !Tools.isBlank(formula) && (isFormulaString(formula) || !Number.isNaN(+formula));
	}
	async isValidType(cellInfo, formula, rule) {
		const { value: cellValue } = cellInfo;
		return !Number.isNaN(getCellValueNumber(cellValue));
	}
	transform(cellInfo, formula, rule) {
		const { value: cellValue } = cellInfo;
		return {
			...cellInfo,
			value: getCellValueNumber(cellValue)
		};
	}
	_parseNumber(formula) {
		if (formula === void 0 || formula === null) return NaN;
		return +formula;
	}
	async parseFormula(rule, unitId, subUnitId, row, column) {
		const formulaResult1 = await this._customFormulaService.getCellFormulaValue(unitId, subUnitId, rule.uid, row, column);
		const formulaResult2 = await this._customFormulaService.getCellFormula2Value(unitId, subUnitId, rule.uid, row, column);
		const { formula1, formula2 } = rule;
		const isFormulaValid = isLegalFormulaResult(String(formulaResult1 === null || formulaResult1 === void 0 ? void 0 : formulaResult1.v)) && isLegalFormulaResult(String(formulaResult2 === null || formulaResult2 === void 0 ? void 0 : formulaResult2.v));
		return {
			formula1: this._parseNumber(isFormulaString(formula1) ? formulaResult1 === null || formulaResult1 === void 0 ? void 0 : formulaResult1.v : formula1),
			formula2: this._parseNumber(isFormulaString(formula2) ? formulaResult2 === null || formulaResult2 === void 0 ? void 0 : formulaResult2.v : formula2),
			isFormulaValid
		};
	}
	validatorFormula(rule, unitId, subUnitId) {
		const operator = rule.operator;
		if (!operator) return { success: true };
		const formula1Success = Tools.isDefine(rule.formula1) && this._isFormulaOrNumber(rule.formula1);
		const formula2Success = Tools.isDefine(rule.formula2) && this._isFormulaOrNumber(rule.formula2);
		const isTwoFormula = TWO_FORMULA_OPERATOR_COUNT.includes(operator);
		const errorMsg = this.localeService.t("sheets-data-validation.validFail.number");
		if (isTwoFormula) return {
			success: formula1Success && formula2Success,
			formula1: formula1Success ? void 0 : errorMsg,
			formula2: formula2Success ? void 0 : errorMsg
		};
		return {
			success: formula1Success,
			formula1: formula1Success ? "" : errorMsg
		};
	}
	generateRuleErrorMessage(rule, position) {
		if (!rule.operator) return this.localeService.t(OperatorErrorTitleMap.NONE).replace("{TYPE}", this.titleStr);
		const { transformedFormula1, transformedFormula2 } = getTransformedFormula(this._lexerTreeBuilder, rule, position);
		return `${this.localeService.t(OperatorErrorTitleMap[rule.operator]).replace(FORMULA1, transformedFormula1 !== null && transformedFormula1 !== void 0 ? transformedFormula1 : "").replace(FORMULA2, transformedFormula2 !== null && transformedFormula2 !== void 0 ? transformedFormula2 : "")}`;
	}
};

//#endregion
//#region src/validators/list-validator.ts
function getRuleFormulaResultSet(result) {
	if (!result) return [];
	const resultSet = /* @__PURE__ */ new Set();
	for (let i = 0, rowLen = result.length; i < rowLen; i++) {
		const row = result[i];
		if (!row) continue;
		for (let j = 0, colLen = row.length; j < colLen; j++) {
			const cell = row[j];
			const value = getCellValueOrigin(cell);
			if (value !== null && value !== void 0) {
				var _cell$s;
				if (typeof value !== "string" && typeof (cell === null || cell === void 0 ? void 0 : cell.s) === "object" && ((_cell$s = cell.s) === null || _cell$s === void 0 || (_cell$s = _cell$s.n) === null || _cell$s === void 0 ? void 0 : _cell$s.pattern)) {
					resultSet.add(numfmt.format(cell.s.n.pattern, value, { throws: false }));
					continue;
				}
				const valueStr = typeof value === "string" ? value : String(value);
				if (isLegalFormulaResult(valueStr)) resultSet.add(valueStr);
			}
		}
	}
	return [...resultSet];
}
const supportedFormula = [
	"if",
	"indirect",
	"choose",
	"offset"
];
function isValidListFormula(formula, lexer) {
	if (!isFormulaString(formula)) return true;
	if (isReferenceString(formula.slice(1))) return true;
	const nodes = lexer.sequenceNodesBuilder(formula);
	return nodes && nodes.some((node) => typeof node === "object" && node.nodeType === sequenceNodeType.FUNCTION && supportedFormula.indexOf(node.token.toLowerCase()) > -1);
}
function isRuleIntersects(rule, sheetName) {
	const { formula1 = "", ranges } = rule;
	if (isReferenceString(formula1.slice(1))) {
		const refRange = deserializeRangeWithSheet(formula1.slice(1));
		if ((!refRange.sheetName || refRange.sheetName === sheetName) && ranges.some((range) => Rectangle.intersects(range, refRange.range))) return true;
	}
	return false;
}
var ListValidator = class extends BaseSheetValidator {
	constructor(..._args) {
		super(..._args);
		_defineProperty(this, "formulaService", this.injector.get(DataValidationFormulaService));
		_defineProperty(this, "_lexer", this.injector.get(LexerTreeBuilder));
		_defineProperty(this, "_univerInstanceService", this.injector.get(IUniverInstanceService));
		_defineProperty(this, "_listCacheService", this.injector.get(DataValidationListCacheService));
		_defineProperty(this, "order", 50);
		_defineProperty(this, "offsetFormulaByRange", false);
		_defineProperty(this, "id", DataValidationType.LIST);
		_defineProperty(this, "title", "sheets-data-validation.list.title");
		_defineProperty(this, "operators", []);
		_defineProperty(this, "scopes", ["sheet"]);
		_defineProperty(this, "skipDefaultFontRender", (rule) => {
			return rule.renderMode !== DataValidationRenderMode.TEXT;
		});
	}
	validatorFormula(rule, unitId, subUnitId) {
		var _rule$formula, _this$_univerInstance;
		const success = !Tools.isBlank(rule.formula1);
		const valid = isValidListFormula((_rule$formula = rule.formula1) !== null && _rule$formula !== void 0 ? _rule$formula : "", this._lexer);
		const sheetName = (_this$_univerInstance = this._univerInstanceService.getUnit(unitId, UniverInstanceType.UNIVER_SHEET)) === null || _this$_univerInstance === void 0 || (_this$_univerInstance = _this$_univerInstance.getSheetBySheetId(subUnitId)) === null || _this$_univerInstance === void 0 ? void 0 : _this$_univerInstance.getName();
		const isIntersects = isRuleIntersects(rule, sheetName !== null && sheetName !== void 0 ? sheetName : "");
		return {
			success: Boolean(success && valid && !isIntersects),
			formula1: success ? valid ? !isIntersects ? void 0 : this.localeService.t("sheets-data-validation.validFail.listIntersects") : this.localeService.t("sheets-data-validation.validFail.listInvalid") : this.localeService.t("sheets-data-validation.validFail.list")
		};
	}
	getExtraStyle(rule, value, { style: defaultStyle }) {
		var _ref;
		const tb = (_ref = defaultStyle.tb !== WrapStrategy.OVERFLOW ? defaultStyle.tb : WrapStrategy.CLIP) !== null && _ref !== void 0 ? _ref : WrapStrategy.WRAP;
		if (rule.type === DataValidationType.LIST && (rule.renderMode === DataValidationRenderMode.ARROW || rule.renderMode === DataValidationRenderMode.TEXT)) {
			const color = this.getListWithColorMap(rule)[`${value !== null && value !== void 0 ? value : ""}`];
			if (color) return {
				bg: { rgb: color },
				tb
			};
		}
		return { tb };
	}
	parseCellValue(cellValue) {
		return deserializeListOptions(cellValue.toString());
	}
	async parseFormula(rule, unitId, subUnitId) {
		var _results$;
		const results = await this.formulaService.getRuleFormulaResult(unitId, subUnitId, rule.uid);
		const formulaResult1 = getFormulaResult(results === null || results === void 0 || (_results$ = results[0]) === null || _results$ === void 0 || (_results$ = _results$.result) === null || _results$ === void 0 ? void 0 : _results$[0][0]);
		return {
			formula1: void 0,
			formula2: void 0,
			isFormulaValid: isLegalFormulaResult(String(formulaResult1))
		};
	}
	async isValidType(cellInfo, formula, rule) {
		const { value, unitId, subUnitId } = cellInfo;
		const { formula1 = "" } = rule;
		const formula1Result = isFormulaString(formula1) ? this._listCacheService.getOrCompute(unitId, subUnitId, rule).list : deserializeListOptions(formula1);
		return (rule.type === DataValidationType.LIST ? value === void 0 || value === null || value === "" ? [] : [`${value}`] : this.parseCellValue(value)).every((i) => formula1Result.includes(i));
	}
	generateRuleName() {
		return this.localeService.t("sheets-data-validation.list.name");
	}
	generateRuleErrorMessage() {
		return this.localeService.t("sheets-data-validation.list.error");
	}
	_getUnitAndSubUnit(currentUnitId, currentSubUnitId) {
		var _ref2, _ref3;
		const workbook = (_ref2 = currentUnitId ? this._univerInstanceService.getUniverSheetInstance(currentUnitId) : void 0) !== null && _ref2 !== void 0 ? _ref2 : this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_SHEET);
		if (!workbook) return null;
		const worksheet = (_ref3 = currentSubUnitId ? workbook.getSheetBySheetId(currentSubUnitId) : void 0) !== null && _ref3 !== void 0 ? _ref3 : workbook.getActiveSheet();
		if (!worksheet) return null;
		return {
			unitId: workbook.getUnitId(),
			subUnitId: worksheet.getSheetId()
		};
	}
	getList(rule, currentUnitId, currentSubUnitId) {
		const location = this._getUnitAndSubUnit(currentUnitId, currentSubUnitId);
		if (!location) return [];
		const { unitId, subUnitId } = location;
		return this._listCacheService.getOrCompute(unitId, subUnitId, rule).list;
	}
	async getListAsync(rule, currentUnitId, currentSubUnitId) {
		var _results$2;
		const { formula1 = "" } = rule;
		const location = this._getUnitAndSubUnit(currentUnitId, currentSubUnitId);
		if (!location) return [];
		const { unitId, subUnitId } = location;
		const results = await this.formulaService.getRuleFormulaResult(unitId, subUnitId, rule.uid);
		return isFormulaString(formula1) ? getRuleFormulaResultSet(results === null || results === void 0 || (_results$2 = results[0]) === null || _results$2 === void 0 || (_results$2 = _results$2.result) === null || _results$2 === void 0 ? void 0 : _results$2[0][0]) : deserializeListOptions(formula1);
	}
	getListWithColor(rule, currentUnitId, currentSubUnitId) {
		const location = this._getUnitAndSubUnit(currentUnitId, currentSubUnitId);
		if (!location) return [];
		const { unitId, subUnitId } = location;
		return this._listCacheService.getOrCompute(unitId, subUnitId, rule).listWithColor;
	}
	getListWithColorMap(rule, currentUnitId, currentSubUnitId) {
		const location = this._getUnitAndSubUnit(currentUnitId, currentSubUnitId);
		if (!location) return {};
		const { unitId, subUnitId } = location;
		return this._listCacheService.getOrCompute(unitId, subUnitId, rule).colorMap;
	}
};

//#endregion
//#region src/validators/text-length-validator.ts
var TextLengthValidator = class extends BaseSheetValidator {
	constructor(..._args) {
		super(..._args);
		_defineProperty(this, "id", DataValidationType.TEXT_LENGTH);
		_defineProperty(this, "title", "sheets-data-validation.textLength.title");
		_defineProperty(this, "_lexerTreeBuilder", this.injector.get(LexerTreeBuilder));
		_defineProperty(this, "order", 30);
		_defineProperty(this, "operators", [
			DataValidationOperator.BETWEEN,
			DataValidationOperator.EQUAL,
			DataValidationOperator.GREATER_THAN,
			DataValidationOperator.GREATER_THAN_OR_EQUAL,
			DataValidationOperator.LESS_THAN,
			DataValidationOperator.LESS_THAN_OR_EQUAL,
			DataValidationOperator.NOT_BETWEEN,
			DataValidationOperator.NOT_EQUAL
		]);
		_defineProperty(this, "scopes", ["sheet"]);
		_defineProperty(this, "_customFormulaService", this.injector.get(DataValidationCustomFormulaService));
	}
	_isFormulaOrInt(formula) {
		return !Tools.isBlank(formula) && (isFormulaString(formula) || !Number.isNaN(+formula) && Number.isInteger(+formula));
	}
	validatorFormula(rule, _unitId, _subUnitId) {
		const operator = rule.operator;
		if (!operator) return { success: false };
		const formula1Success = Tools.isDefine(rule.formula1) && this._isFormulaOrInt(rule.formula1);
		const formula2Success = Tools.isDefine(rule.formula2) && this._isFormulaOrInt(rule.formula2);
		const isTwoFormula = TWO_FORMULA_OPERATOR_COUNT.includes(operator);
		const errorMsg = this.localeService.t("sheets-data-validation.validFail.number");
		if (isTwoFormula) return {
			success: formula1Success && formula2Success,
			formula1: formula1Success ? void 0 : errorMsg,
			formula2: formula2Success ? void 0 : errorMsg
		};
		return {
			success: formula1Success,
			formula1: errorMsg
		};
	}
	_parseNumber(formula) {
		if (formula === void 0 || formula === null) return NaN;
		return +formula;
	}
	async parseFormula(rule, unitId, subUnitId, row, column) {
		const formulaResult1 = await this._customFormulaService.getCellFormulaValue(unitId, subUnitId, rule.uid, row, column);
		const formulaResult2 = await this._customFormulaService.getCellFormula2Value(unitId, subUnitId, rule.uid, row, column);
		const { formula1, formula2 } = rule;
		const isFormulaValid = isLegalFormulaResult(String(formulaResult1 === null || formulaResult1 === void 0 ? void 0 : formulaResult1.v)) && isLegalFormulaResult(String(formulaResult2 === null || formulaResult2 === void 0 ? void 0 : formulaResult2.v));
		return {
			formula1: this._parseNumber(isFormulaString(formula1) ? formulaResult1 === null || formulaResult1 === void 0 ? void 0 : formulaResult1.v : formula1),
			formula2: this._parseNumber(isFormulaString(formula2) ? formulaResult2 === null || formulaResult2 === void 0 ? void 0 : formulaResult2.v : formula2),
			isFormulaValid
		};
	}
	transform(cellInfo, _formula, _rule) {
		return {
			...cellInfo,
			value: cellInfo.value.toString().length
		};
	}
	async isValidType(cellInfo, _formula, _rule) {
		const { value: cellValue } = cellInfo;
		return typeof cellValue === "string" || typeof cellValue === "number";
	}
	generateRuleErrorMessage(rule, pos) {
		if (!rule.operator) return this.titleStr;
		const { transformedFormula1, transformedFormula2 } = getTransformedFormula(this._lexerTreeBuilder, rule, pos);
		return `${this.localeService.t(TextLengthErrorTitleMap[rule.operator]).replace(FORMULA1, transformedFormula1 !== null && transformedFormula1 !== void 0 ? transformedFormula1 : "").replace(FORMULA2, transformedFormula2 !== null && transformedFormula2 !== void 0 ? transformedFormula2 : "")}`;
	}
};

//#endregion
//#region src/commands/commands/data-validation.command.ts
function isBlankCell(cellData) {
	var _cellData$p$body$data, _cellData$p$body;
	if (!cellData) return true;
	if (!cellData.p) return Tools.isBlank(cellData.v);
	return !((_cellData$p$body$data = (_cellData$p$body = cellData.p.body) === null || _cellData$p$body === void 0 ? void 0 : _cellData$p$body.dataStream) !== null && _cellData$p$body$data !== void 0 ? _cellData$p$body$data : "").slice(0, -2).trim();
}
function getDataValidationDiffMutations(unitId, subUnitId, diffs, accessor, source = "command", fillDefaultValue = true) {
	const lexerTreeBuilder = accessor.get(LexerTreeBuilder);
	const validatorRegistryService = accessor.get(DataValidatorRegistryService);
	const redoMutations = [];
	const undoMutations = [];
	const sheetDataValidationModel = accessor.get(SheetDataValidationModel);
	const target = getSheetCommandTarget(accessor.get(IUniverInstanceService), {
		unitId,
		subUnitId
	});
	if (!target) return {
		redoMutations,
		undoMutations
	};
	const { worksheet } = target;
	const redoMatrix = new ObjectMatrix();
	let setRangeValue = false;
	function setRangesDefaultValue(ranges, defaultValue) {
		if (!fillDefaultValue) return;
		ranges.forEach((range) => {
			Range.foreach(range, (row, column) => {
				const cellData = worksheet.getCellRaw(row, column);
				const value = getStringCellValue(cellData);
				if ((isBlankCell(cellData) || value === defaultValue) && !(cellData === null || cellData === void 0 ? void 0 : cellData.p)) {
					setRangeValue = true;
					redoMatrix.setValue(row, column, {
						v: defaultValue,
						p: null
					});
				}
			});
		});
	}
	diffs.forEach((diff) => {
		switch (diff.type) {
			case "delete":
				redoMutations.push({
					id: RemoveDataValidationMutation.id,
					params: {
						unitId,
						subUnitId,
						ruleId: diff.rule.uid,
						source
					}
				});
				undoMutations.unshift({
					id: AddDataValidationMutation.id,
					params: {
						unitId,
						subUnitId,
						rule: diff.rule,
						index: diff.index,
						source
					}
				});
				break;
			case "update": {
				if (shouldOffsetFormulaByRange(diff.rule.type, validatorRegistryService)) {
					const originRow = diff.oldRanges[0].startRow;
					const originColumn = diff.oldRanges[0].startColumn;
					const newRow = diff.newRanges[0].startRow;
					const newColumn = diff.newRanges[0].startColumn;
					const rowDiff = newRow - originRow;
					const columnDiff = newColumn - originColumn;
					const newFormula = isFormulaString(diff.rule.formula1) ? lexerTreeBuilder.moveFormulaRefOffset(diff.rule.formula1, columnDiff, rowDiff) : diff.rule.formula1;
					const newFormula2 = isFormulaString(diff.rule.formula2) ? lexerTreeBuilder.moveFormulaRefOffset(diff.rule.formula2, columnDiff, rowDiff) : diff.rule.formula2;
					if (newFormula !== diff.rule.formula1 || newFormula2 !== diff.rule.formula2 || !isRangesEqual(diff.newRanges, diff.oldRanges)) {
						redoMutations.push({
							id: UpdateDataValidationMutation.id,
							params: {
								unitId,
								subUnitId,
								ruleId: diff.ruleId,
								payload: {
									type: UpdateRuleType.ALL,
									payload: {
										formula1: newFormula,
										formula2: newFormula2,
										ranges: diff.newRanges
									}
								}
							}
						});
						undoMutations.unshift({
							id: UpdateDataValidationMutation.id,
							params: {
								unitId,
								subUnitId,
								ruleId: diff.ruleId,
								payload: {
									type: UpdateRuleType.ALL,
									payload: {
										formula1: diff.rule.formula1,
										formula2: diff.rule.formula2,
										ranges: diff.oldRanges
									}
								}
							}
						});
					} else {
						redoMutations.push({
							id: UpdateDataValidationMutation.id,
							params: {
								unitId,
								subUnitId,
								ruleId: diff.ruleId,
								payload: {
									type: UpdateRuleType.RANGE,
									payload: diff.newRanges
								},
								source
							}
						});
						undoMutations.unshift({
							id: UpdateDataValidationMutation.id,
							params: {
								unitId,
								subUnitId,
								ruleId: diff.ruleId,
								payload: {
									type: UpdateRuleType.RANGE,
									payload: diff.oldRanges
								},
								source
							}
						});
					}
				} else {
					redoMutations.push({
						id: UpdateDataValidationMutation.id,
						params: {
							unitId,
							subUnitId,
							ruleId: diff.ruleId,
							payload: {
								type: UpdateRuleType.RANGE,
								payload: diff.newRanges
							},
							source
						}
					});
					undoMutations.unshift({
						id: UpdateDataValidationMutation.id,
						params: {
							unitId,
							subUnitId,
							ruleId: diff.ruleId,
							payload: {
								type: UpdateRuleType.RANGE,
								payload: diff.oldRanges
							},
							source
						}
					});
				}
				const rule = sheetDataValidationModel.getRuleById(unitId, subUnitId, diff.ruleId);
				if (rule && rule.type === DataValidationType.CHECKBOX) {
					const formula = sheetDataValidationModel.getValidator(DataValidationType.CHECKBOX).parseFormulaSync(rule, unitId, subUnitId);
					setRangesDefaultValue(diff.newRanges, formula.formula2);
				}
				break;
			}
			case "add":
				redoMutations.push({
					id: AddDataValidationMutation.id,
					params: {
						unitId,
						subUnitId,
						rule: diff.rule,
						source
					}
				});
				undoMutations.unshift({
					id: RemoveDataValidationMutation.id,
					params: {
						unitId,
						subUnitId,
						ruleId: diff.rule.uid,
						source
					}
				});
				if (diff.rule.type === DataValidationType.CHECKBOX) {
					const formula = sheetDataValidationModel.getValidator(DataValidationType.CHECKBOX).parseFormulaSync(diff.rule, unitId, subUnitId);
					setRangesDefaultValue(diff.rule.ranges, formula.originFormula2);
				}
				break;
			default: break;
		}
	});
	if (setRangeValue) {
		const redoSetRangeValues = {
			id: SetRangeValuesMutation.id,
			params: {
				unitId,
				subUnitId,
				cellValue: redoMatrix.getData()
			}
		};
		const undoSetRangeValues = {
			id: SetRangeValuesMutation.id,
			params: SetRangeValuesUndoMutationFactory(accessor, redoSetRangeValues.params)
		};
		redoMutations.push(redoSetRangeValues);
		undoMutations.push(undoSetRangeValues);
	}
	return {
		redoMutations,
		undoMutations
	};
}
const UpdateSheetDataValidationRangeCommand = {
	type: CommandType.COMMAND,
	id: "sheet.command.updateDataValidationRuleRange",
	handler(accessor, params) {
		if (!params) return false;
		const { unitId, subUnitId, ranges, ruleId } = params;
		const sheetDataValidationModel = accessor.get(SheetDataValidationModel);
		const commandService = accessor.get(ICommandService);
		const undoRedoService = accessor.get(IUndoRedoService);
		if (!sheetDataValidationModel.getRuleById(unitId, subUnitId, ruleId)) return false;
		const matrix = sheetDataValidationModel.getRuleObjectMatrix(unitId, subUnitId).clone();
		matrix.updateRange(ruleId, ranges);
		const { redoMutations, undoMutations } = getDataValidationDiffMutations(unitId, subUnitId, matrix.diff(sheetDataValidationModel.getRules(unitId, subUnitId)), accessor);
		undoRedoService.pushUndoRedo({
			undoMutations,
			redoMutations,
			unitID: unitId
		});
		sequenceExecute(redoMutations, commandService);
		return true;
	}
};
const AddSheetDataValidationCommand = {
	type: CommandType.COMMAND,
	id: "sheet.command.addDataValidation",
	handler(accessor, params) {
		if (!params) return false;
		const { unitId, subUnitId, rule } = params;
		const sheetDataValidationModel = accessor.get(SheetDataValidationModel);
		const commandService = accessor.get(ICommandService);
		const undoRedoService = accessor.get(IUndoRedoService);
		const matrix = sheetDataValidationModel.getRuleObjectMatrix(unitId, subUnitId).clone();
		matrix.addRule(rule);
		const diffs = matrix.diff(sheetDataValidationModel.getRules(unitId, subUnitId));
		const validator = sheetDataValidationModel.getValidator(rule.type);
		const mutationParams = {
			unitId,
			subUnitId,
			rule: {
				...rule,
				...validator === null || validator === void 0 ? void 0 : validator.normalizeFormula(rule, unitId, subUnitId)
			}
		};
		const { redoMutations, undoMutations } = getDataValidationDiffMutations(unitId, subUnitId, diffs, accessor);
		redoMutations.push({
			id: AddDataValidationMutation.id,
			params: mutationParams
		});
		undoMutations.unshift({
			id: RemoveDataValidationMutation.id,
			params: {
				unitId,
				subUnitId,
				ruleId: rule.uid
			}
		});
		undoRedoService.pushUndoRedo({
			unitID: unitId,
			redoMutations,
			undoMutations
		});
		sequenceExecute(redoMutations, commandService);
		return true;
	}
};
const UpdateSheetDataValidationSettingCommand = {
	type: CommandType.COMMAND,
	id: "sheets.command.update-data-validation-setting",
	handler(accessor, params) {
		if (!params) return false;
		const commandService = accessor.get(ICommandService);
		const redoUndoService = accessor.get(IUndoRedoService);
		const sheetDataValidationModel = accessor.get(SheetDataValidationModel);
		const dataValidatorRegistryService = accessor.get(DataValidatorRegistryService);
		const { unitId, subUnitId, ruleId, setting } = params;
		const validator = dataValidatorRegistryService.getValidatorItem(setting.type);
		if (!validator) return false;
		const rule = sheetDataValidationModel.getRuleById(unitId, subUnitId, ruleId);
		if (!rule) return false;
		const newRule = {
			...rule,
			...setting
		};
		if (!validator.validatorFormula(newRule, unitId, subUnitId).success) return false;
		const mutationParams = {
			unitId,
			subUnitId,
			ruleId,
			payload: {
				type: UpdateRuleType.SETTING,
				payload: {
					...setting,
					...validator.normalizeFormula(newRule, unitId, subUnitId)
				}
			}
		};
		const redoMutations = [{
			id: UpdateDataValidationMutation.id,
			params: mutationParams
		}];
		const undoMutationParams = {
			unitId,
			subUnitId,
			ruleId,
			payload: {
				type: UpdateRuleType.SETTING,
				payload: getRuleSetting(rule)
			}
		};
		const undoMutations = [{
			id: UpdateDataValidationMutation.id,
			params: undoMutationParams
		}];
		if (setting.type === DataValidationType.CHECKBOX) {
			const ranges = rule.ranges;
			const target = getSheetCommandTarget(accessor.get(IUniverInstanceService), {
				unitId,
				subUnitId
			});
			if (target) {
				const redoMatrix = new ObjectMatrix();
				const { worksheet } = target;
				const { formula2: oldFormula2 = 0, formula1: oldFormula1 = 1 } = rule;
				const { formula2 = 0, formula1 = 1 } = setting;
				let setted = false;
				ranges.forEach((range) => {
					Range.foreach(range, (row, column) => {
						const cellData = worksheet.getCellRaw(row, column);
						const value = getStringCellValue(cellData);
						if ((isBlankCell(cellData) || value === String(oldFormula2)) && !(cellData === null || cellData === void 0 ? void 0 : cellData.p)) {
							redoMatrix.setValue(row, column, {
								v: formula2,
								p: null
							});
							setted = true;
						} else if (value === String(oldFormula1) && !(cellData === null || cellData === void 0 ? void 0 : cellData.p)) {
							redoMatrix.setValue(row, column, {
								v: formula1,
								p: null
							});
							setted = true;
						}
					});
				});
				if (setted) {
					const redoSetRangeValues = {
						id: SetRangeValuesMutation.id,
						params: {
							unitId,
							subUnitId,
							cellValue: redoMatrix.getData()
						}
					};
					const undoSetRangeValues = {
						id: SetRangeValuesMutation.id,
						params: SetRangeValuesUndoMutationFactory(accessor, redoSetRangeValues.params)
					};
					redoMutations.push(redoSetRangeValues);
					undoMutations.push(undoSetRangeValues);
				}
			}
		}
		if (sequenceExecute(redoMutations, commandService).result) {
			redoUndoService.pushUndoRedo({
				unitID: unitId,
				redoMutations,
				undoMutations
			});
			return true;
		}
		return false;
	}
};
const UpdateSheetDataValidationOptionsCommand = {
	type: CommandType.COMMAND,
	id: "sheets.command.update-data-validation-options",
	handler(accessor, params) {
		if (!params) return false;
		const commandService = accessor.get(ICommandService);
		const redoUndoService = accessor.get(IUndoRedoService);
		const sheetDataValidationModel = accessor.get(SheetDataValidationModel);
		const { unitId, subUnitId, ruleId, options } = params;
		const rule = sheetDataValidationModel.getRuleById(unitId, subUnitId, ruleId);
		if (!rule) return false;
		const mutationParams = {
			unitId,
			subUnitId,
			ruleId,
			payload: {
				type: UpdateRuleType.OPTIONS,
				payload: options
			}
		};
		const redoMutations = [{
			id: UpdateDataValidationMutation.id,
			params: mutationParams
		}];
		const undoMutationParams = {
			unitId,
			subUnitId,
			ruleId,
			payload: {
				type: UpdateRuleType.OPTIONS,
				payload: getRuleOptions(rule)
			}
		};
		const undoMutations = [{
			id: UpdateDataValidationMutation.id,
			params: undoMutationParams
		}];
		redoUndoService.pushUndoRedo({
			unitID: unitId,
			redoMutations,
			undoMutations
		});
		commandService.executeCommand(UpdateDataValidationMutation.id, mutationParams);
		return true;
	}
};
const ClearRangeDataValidationCommand = {
	type: CommandType.COMMAND,
	id: "sheets.command.clear-range-data-validation",
	handler(accessor, params) {
		if (!params) return false;
		const { unitId, subUnitId, ranges } = params;
		const commandService = accessor.get(ICommandService);
		const target = getSheetCommandTarget(accessor.get(IUniverInstanceService), {
			unitId,
			subUnitId
		});
		const sheetDataValidationModel = accessor.get(SheetDataValidationModel);
		if (!target) return false;
		const undoRedoService = accessor.get(IUndoRedoService);
		const matrix = sheetDataValidationModel.getRuleObjectMatrix(unitId, subUnitId).clone();
		matrix.removeRange(ranges);
		const { redoMutations, undoMutations } = getDataValidationDiffMutations(unitId, subUnitId, matrix.diff(sheetDataValidationModel.getRules(unitId, subUnitId)), accessor);
		undoRedoService.pushUndoRedo({
			unitID: unitId,
			redoMutations,
			undoMutations
		});
		return sequenceExecute(redoMutations, commandService).result;
	}
};
const RemoveSheetAllDataValidationCommand = {
	type: CommandType.COMMAND,
	id: "sheet.command.remove-all-data-validation",
	handler(accessor, params) {
		if (!params) return false;
		const { unitId, subUnitId } = params;
		const commandService = accessor.get(ICommandService);
		const sheetDataValidationModel = accessor.get(SheetDataValidationModel);
		const undoRedoService = accessor.get(IUndoRedoService);
		const currentRules = [...sheetDataValidationModel.getRules(unitId, subUnitId)];
		const redoParams = {
			unitId,
			subUnitId,
			ruleId: currentRules.map((rule) => rule.uid)
		};
		const redoMutations = [{
			id: RemoveDataValidationMutation.id,
			params: redoParams
		}];
		const undoMutations = [{
			id: AddDataValidationMutation.id,
			params: {
				unitId,
				subUnitId,
				rule: currentRules
			}
		}];
		undoRedoService.pushUndoRedo({
			redoMutations,
			undoMutations,
			unitID: unitId
		});
		commandService.executeCommand(RemoveDataValidationMutation.id, redoParams);
		return true;
	}
};
const removeDataValidationUndoFactory = (accessor, redoParams) => {
	const sheetDataValidationModel = accessor.get(SheetDataValidationModel);
	const { unitId, subUnitId, ruleId, source } = redoParams;
	if (Array.isArray(ruleId)) {
		const rules = ruleId.map((id) => sheetDataValidationModel.getRuleById(unitId, subUnitId, id)).filter(Boolean);
		return [{
			id: AddDataValidationMutation.id,
			params: {
				unitId,
				subUnitId,
				rule: rules,
				source
			}
		}];
	}
	return [{
		id: AddDataValidationMutation.id,
		params: {
			unitId,
			subUnitId,
			rule: { ...sheetDataValidationModel.getRuleById(unitId, subUnitId, ruleId) },
			index: sheetDataValidationModel.getRuleIndex(unitId, subUnitId, ruleId)
		}
	}];
};
const RemoveSheetDataValidationCommand = {
	type: CommandType.COMMAND,
	id: "sheet.command.remove-data-validation-rule",
	handler(accessor, params) {
		if (!params) return false;
		const { unitId, subUnitId, ruleId } = params;
		const commandService = accessor.get(ICommandService);
		const undoRedoService = accessor.get(IUndoRedoService);
		const sheetDataValidationModel = accessor.get(SheetDataValidationModel);
		const redoMutations = [{
			id: RemoveDataValidationMutation.id,
			params
		}];
		const undoMutations = [{
			id: AddDataValidationMutation.id,
			params: {
				unitId,
				subUnitId,
				rule: { ...sheetDataValidationModel.getRuleById(unitId, subUnitId, ruleId) },
				index: sheetDataValidationModel.getRuleIndex(unitId, subUnitId, ruleId)
			}
		}];
		undoRedoService.pushUndoRedo({
			undoMutations,
			redoMutations,
			unitID: params.unitId
		});
		commandService.executeCommand(RemoveDataValidationMutation.id, params);
		return true;
	}
};

//#endregion
//#region src/common/const.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const DATA_VALIDATION_PLUGIN_NAME = "SHEET_DATA_VALIDATION_PLUGIN";

//#endregion
//#region src/controllers/dv-formula.controller.ts
let DataValidationFormulaController = class DataValidationFormulaController extends Disposable {
	constructor(_univerInstanceService, _permissionService, _lexerTreeBuilder) {
		super();
		this._univerInstanceService = _univerInstanceService;
		this._permissionService = _permissionService;
		this._lexerTreeBuilder = _lexerTreeBuilder;
	}
	getFormulaRefCheck(formulaString) {
		const sequenceNodes = this._lexerTreeBuilder.sequenceNodesBuilder(formulaString);
		if (!sequenceNodes) return true;
		for (let i = 0; i < sequenceNodes.length; i++) {
			const node = sequenceNodes[i];
			if (typeof node === "string") continue;
			const { token } = node;
			const sequenceGrid = deserializeRangeWithSheetWithCache(token);
			const workbook = this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_SHEET);
			let targetSheet = workbook.getActiveSheet();
			const unitId = workbook.getUnitId();
			if (sequenceGrid.sheetName) {
				targetSheet = workbook.getSheetBySheetName(sequenceGrid.sheetName);
				if (!targetSheet) return false;
				const subUnitId = targetSheet === null || targetSheet === void 0 ? void 0 : targetSheet.getSheetId();
				if (!this._permissionService.getPermissionPoint(new WorksheetViewPermission(unitId, subUnitId).id)) return false;
			}
			if (!targetSheet) return false;
			const { startRow, endRow, startColumn, endColumn } = sequenceGrid.range;
			for (let i = startRow; i <= endRow; i++) for (let j = startColumn; j <= endColumn; j++) {
				var _targetSheet$getCell;
				const permission = (_targetSheet$getCell = targetSheet.getCell(i, j)) === null || _targetSheet$getCell === void 0 || (_targetSheet$getCell = _targetSheet$getCell.selectionProtection) === null || _targetSheet$getCell === void 0 ? void 0 : _targetSheet$getCell[0];
				if ((permission === null || permission === void 0 ? void 0 : permission[UnitAction.View]) === false) return false;
			}
		}
		return true;
	}
};
DataValidationFormulaController = __decorate([
	__decorateParam(0, IUniverInstanceService),
	__decorateParam(1, IPermissionService),
	__decorateParam(2, Inject(LexerTreeBuilder))
], DataValidationFormulaController);

//#endregion
//#region package.json
var name = "@univerjs/sheets-data-validation";
var version = "0.25.1";

//#endregion
//#region src/config/config.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
const SHEETS_DATA_VALIDATION_PLUGIN_CONFIG_KEY = "sheets-data-validation.config";
const configSymbol = Symbol(SHEETS_DATA_VALIDATION_PLUGIN_CONFIG_KEY);
const defaultPluginConfig = {};

//#endregion
//#region src/controllers/dv-formula-ref-range.controller.ts
let DataValidationFormulaRefRangeController = class DataValidationFormulaRefRangeController extends Disposable {
	constructor(_dataValidationModel, _formulaRefRangeService, _validatorRegistryService) {
		super();
		this._dataValidationModel = _dataValidationModel;
		this._formulaRefRangeService = _formulaRefRangeService;
		this._validatorRegistryService = _validatorRegistryService;
		_defineProperty(this, "_disposableMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "registerRule", (unitId, subUnitId, rule) => {
			if (!shouldOffsetFormulaByRange(rule.type, this._validatorRegistryService)) return;
			this.register(unitId, subUnitId, rule);
		});
		this._initRefRange();
	}
	_getIdWithUnitId(unitID, subUnitId, ruleId) {
		return `${unitID}_${subUnitId}_${ruleId}`;
	}
	register(unitId, subUnitId, rule) {
		const oldRanges = rule.ranges;
		const oldFormula1 = rule.formula1;
		const oldFormula2 = rule.formula2;
		const disposable = this._formulaRefRangeService.registerRangeFormula(unitId, subUnitId, oldRanges, [oldFormula1 !== null && oldFormula1 !== void 0 ? oldFormula1 : "", oldFormula2 !== null && oldFormula2 !== void 0 ? oldFormula2 : ""], (res) => {
			if (res.length === 0) return {
				undos: [{
					id: AddDataValidationMutation.id,
					params: {
						unitId,
						subUnitId,
						rule,
						source: "patched"
					}
				}],
				redos: [{
					id: RemoveDataValidationMutation.id,
					params: {
						unitId,
						subUnitId,
						ruleId: rule.uid,
						source: "patched"
					}
				}]
			};
			const redos = [];
			const undos = [];
			const first = res[0];
			redos.push({
				id: UpdateDataValidationMutation.id,
				params: {
					unitId,
					subUnitId,
					ruleId: rule.uid,
					payload: {
						type: UpdateRuleType.ALL,
						payload: {
							ranges: first.ranges,
							formula1: first.formulas[0],
							formula2: first.formulas[1]
						}
					},
					source: "patched"
				}
			});
			undos.push({
				id: UpdateDataValidationMutation.id,
				params: {
					unitId,
					subUnitId,
					ruleId: rule.uid,
					payload: {
						type: UpdateRuleType.ALL,
						payload: {
							ranges: oldRanges,
							formula1: oldFormula1,
							formula2: oldFormula2
						}
					},
					source: "patched"
				}
			});
			for (let i = 1; i < res.length; i++) {
				const item = res[i];
				const id = generateRandomId();
				redos.push({
					id: AddDataValidationMutation.id,
					params: {
						unitId,
						subUnitId,
						rule: {
							...rule,
							uid: id,
							formula1: item.formulas[0],
							formula2: item.formulas[1],
							ranges: item.ranges
						},
						source: "patched"
					}
				});
				undos.push({
					id: RemoveDataValidationMutation.id,
					params: {
						unitId,
						subUnitId,
						ruleId: id,
						source: "patched"
					}
				});
			}
			return {
				undos,
				redos
			};
		});
		const id = this._getIdWithUnitId(unitId, subUnitId, rule.uid);
		this._disposableMap.set(id, disposable);
	}
	_initRefRange() {
		const allRules = this._dataValidationModel.getAll();
		for (const [unitId, subUnitMap] of allRules) for (const [subUnitId, rules] of subUnitMap) for (const rule of rules) this.registerRule(unitId, subUnitId, rule);
		this.disposeWithMe(this._dataValidationModel.ruleChange$.subscribe((option) => {
			const { unitId, subUnitId, rule } = option;
			switch (option.type) {
				case "add": {
					const rule = option.rule;
					this.registerRule(option.unitId, option.subUnitId, rule);
					break;
				}
				case "remove": {
					const disposeSet = this._disposableMap.get(this._getIdWithUnitId(unitId, subUnitId, rule.uid));
					if (disposeSet) disposeSet.dispose();
					break;
				}
				case "update": {
					const rule = option.rule;
					const disposeSet = this._disposableMap.get(this._getIdWithUnitId(unitId, subUnitId, rule.uid));
					if (disposeSet) disposeSet.dispose();
					this.registerRule(option.unitId, option.subUnitId, rule);
					break;
				}
			}
		}));
		this.disposeWithMe(toDisposable(() => {
			this._disposableMap.forEach((item) => {
				item.dispose();
			});
			this._disposableMap.clear();
		}));
	}
};
DataValidationFormulaRefRangeController = __decorate([
	__decorateParam(0, Inject(SheetDataValidationModel)),
	__decorateParam(1, Inject(FormulaRefRangeService)),
	__decorateParam(2, Inject(DataValidatorRegistryService))
], DataValidationFormulaRefRangeController);

//#endregion
//#region src/controllers/dv-ref-range.controller.ts
let DataValidationRefRangeController = class DataValidationRefRangeController extends Disposable {
	constructor(_dataValidationModel, _injector, _refRangeService, _dataValidationFormulaService, _formulaRefRangeService, _validatorRegistryService) {
		super();
		this._dataValidationModel = _dataValidationModel;
		this._injector = _injector;
		this._refRangeService = _refRangeService;
		this._dataValidationFormulaService = _dataValidationFormulaService;
		this._formulaRefRangeService = _formulaRefRangeService;
		this._validatorRegistryService = _validatorRegistryService;
		_defineProperty(this, "_disposableMap", /* @__PURE__ */ new Map());
		_defineProperty(this, "registerRule", (unitId, subUnitId, rule) => {
			if (shouldOffsetFormulaByRange(rule.type, this._validatorRegistryService)) return;
			this.register(unitId, subUnitId, rule);
			this.registerFormula(unitId, subUnitId, rule);
		});
		this._initRefRange();
	}
	_getIdWithUnitId(unitID, subUnitId, ruleId) {
		return `${unitID}_${subUnitId}_${ruleId}`;
	}
	registerFormula(unitId, subUnitId, rule) {
		var _this$_disposableMap$;
		const ruleId = rule.uid;
		const id = this._getIdWithUnitId(unitId, subUnitId, ruleId);
		const disposeSet = (_this$_disposableMap$ = this._disposableMap.get(id)) !== null && _this$_disposableMap$ !== void 0 ? _this$_disposableMap$ : /* @__PURE__ */ new Set();
		const handleFormulaChange = (type, formulaString) => {
			const oldRule = this._dataValidationModel.getRuleById(unitId, subUnitId, ruleId);
			if (!oldRule) return {
				redos: [],
				undos: []
			};
			const oldFormula = oldRule[type];
			if (!oldFormula || oldFormula === formulaString) return {
				redos: [],
				undos: []
			};
			const redoParams = {
				unitId,
				subUnitId,
				ruleId: rule.uid,
				payload: {
					type: UpdateRuleType.SETTING,
					payload: {
						type: oldRule.type,
						formula1: oldRule.formula1,
						formula2: oldRule.formula2,
						[type]: formulaString
					}
				},
				source: "patched"
			};
			const undoParams = {
				unitId,
				subUnitId,
				ruleId: rule.uid,
				payload: {
					type: UpdateRuleType.SETTING,
					payload: {
						type: oldRule.type,
						formula1: oldRule.formula1,
						formula2: oldRule.formula2
					}
				},
				source: "patched"
			};
			return {
				redos: [{
					id: UpdateDataValidationMutation.id,
					params: redoParams
				}],
				undos: [{
					id: UpdateDataValidationMutation.id,
					params: undoParams
				}]
			};
		};
		const currentFormula = this._dataValidationFormulaService.getRuleFormulaInfo(unitId, subUnitId, ruleId);
		if (currentFormula) {
			const [formula1, formula2] = currentFormula;
			if (formula1) {
				const disposable = this._formulaRefRangeService.registerFormula(unitId, subUnitId, formula1.text, (newFormulaString) => handleFormulaChange("formula1", newFormulaString));
				disposeSet.add(() => disposable.dispose());
			}
			if (formula2) {
				const disposable = this._formulaRefRangeService.registerFormula(unitId, subUnitId, formula2.text, (newFormulaString) => handleFormulaChange("formula2", newFormulaString));
				disposeSet.add(() => disposable.dispose());
			}
		}
	}
	register(unitId, subUnitId, rule) {
		var _this$_disposableMap$2;
		const handleRangeChange = (commandInfo) => {
			const oldRanges = [...rule.ranges];
			const resultRanges = oldRanges.map((range) => {
				return handleCommonDefaultRangeChangeWithEffectRefCommands(range, commandInfo);
			}).filter((range) => !!range).flat();
			if (isRangesEqual(resultRanges, oldRanges)) return {
				redos: [],
				undos: []
			};
			if (resultRanges.length) {
				const redoParams = {
					unitId,
					subUnitId,
					ruleId: rule.uid,
					payload: {
						type: UpdateRuleType.RANGE,
						payload: resultRanges
					},
					source: "patched"
				};
				return {
					redos: [{
						id: UpdateDataValidationMutation.id,
						params: redoParams
					}],
					undos: [{
						id: UpdateDataValidationMutation.id,
						params: {
							unitId,
							subUnitId,
							ruleId: rule.uid,
							payload: {
								type: UpdateRuleType.RANGE,
								payload: oldRanges
							},
							source: "patched"
						}
					}]
				};
			} else {
				const redoParams = {
					unitId,
					subUnitId,
					ruleId: rule.uid
				};
				return {
					redos: [{
						id: RemoveDataValidationMutation.id,
						params: redoParams
					}],
					undos: removeDataValidationUndoFactory(this._injector, redoParams)
				};
			}
		};
		const disposeList = [];
		rule.ranges.forEach((range) => {
			const disposable = this._refRangeService.registerRefRange(range, handleRangeChange, unitId, subUnitId);
			disposeList.push(() => disposable.dispose());
		});
		const id = this._getIdWithUnitId(unitId, subUnitId, rule.uid);
		const current = (_this$_disposableMap$2 = this._disposableMap.get(id)) !== null && _this$_disposableMap$2 !== void 0 ? _this$_disposableMap$2 : /* @__PURE__ */ new Set();
		current.add(() => disposeList.forEach((dispose) => dispose()));
		this._disposableMap.set(id, current);
	}
	_initRefRange() {
		const allRules = this._dataValidationModel.getAll();
		for (const [unitId, subUnitMap] of allRules) for (const [subUnitId, rules] of subUnitMap) for (const rule of rules) this.registerRule(unitId, subUnitId, rule);
		this.disposeWithMe(this._dataValidationModel.ruleChange$.subscribe((option) => {
			const { unitId, subUnitId, rule } = option;
			switch (option.type) {
				case "add": {
					const rule = option.rule;
					this.registerRule(option.unitId, option.subUnitId, rule);
					break;
				}
				case "remove": {
					const disposeSet = this._disposableMap.get(this._getIdWithUnitId(unitId, subUnitId, rule.uid));
					if (disposeSet) disposeSet.forEach((dispose) => dispose());
					break;
				}
				case "update": {
					const rule = option.rule;
					const disposeSet = this._disposableMap.get(this._getIdWithUnitId(unitId, subUnitId, rule.uid));
					if (disposeSet) disposeSet.forEach((dispose) => dispose());
					this.registerRule(option.unitId, option.subUnitId, rule);
					break;
				}
			}
		}));
		this.disposeWithMe(toDisposable(() => {
			this._disposableMap.forEach((item) => {
				item.forEach((dispose) => dispose());
			});
			this._disposableMap.clear();
		}));
	}
};
DataValidationRefRangeController = __decorate([
	__decorateParam(0, Inject(SheetDataValidationModel)),
	__decorateParam(1, Inject(Injector)),
	__decorateParam(2, Inject(RefRangeService)),
	__decorateParam(3, Inject(DataValidationFormulaService)),
	__decorateParam(4, Inject(FormulaRefRangeService)),
	__decorateParam(5, Inject(DataValidatorRegistryService))
], DataValidationRefRangeController);

//#endregion
//#region src/controllers/dv-sheet.controller.ts
let SheetDataValidationSheetController = class SheetDataValidationSheetController extends Disposable {
	constructor(_sheetInterceptorService, _univerInstanceService, _sheetDataValidationModel) {
		super();
		this._sheetInterceptorService = _sheetInterceptorService;
		this._univerInstanceService = _univerInstanceService;
		this._sheetDataValidationModel = _sheetDataValidationModel;
		this._initSheetChange();
	}
	_initSheetChange() {
		this.disposeWithMe(this._sheetInterceptorService.interceptCommand({ getMutations: (commandInfo) => {
			if (commandInfo.id === RemoveSheetCommand.id) {
				var _workbook$getActiveSh;
				const params = commandInfo.params;
				const unitId = params.unitId || this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_SHEET).getUnitId();
				const workbook = this._univerInstanceService.getUniverSheetInstance(unitId);
				if (!workbook) return {
					redos: [],
					undos: []
				};
				const subUnitId = params.subUnitId || ((_workbook$getActiveSh = workbook.getActiveSheet()) === null || _workbook$getActiveSh === void 0 ? void 0 : _workbook$getActiveSh.getSheetId());
				if (!subUnitId) return {
					redos: [],
					undos: []
				};
				const rules = this._sheetDataValidationModel.getRules(unitId, subUnitId);
				if (rules.length === 0) return {
					redos: [],
					undos: []
				};
				const redoParams = {
					unitId,
					subUnitId,
					ruleId: rules.map((i) => i.uid),
					source: "patched"
				};
				const undoParams = {
					unitId,
					subUnitId,
					rule: [...rules],
					source: "patched"
				};
				return {
					redos: [{
						id: RemoveDataValidationMutation.id,
						params: redoParams
					}],
					undos: [{
						id: AddDataValidationMutation.id,
						params: undoParams
					}]
				};
			} else if (commandInfo.id === CopySheetCommand.id) {
				const { unitId, subUnitId, targetSubUnitId } = commandInfo.params;
				if (!unitId || !subUnitId || !targetSubUnitId) return {
					redos: [],
					undos: []
				};
				const rules = this._sheetDataValidationModel.getRules(unitId, subUnitId);
				if (rules.length === 0) return {
					redos: [],
					undos: []
				};
				const newRules = rules.map((rule) => {
					return {
						...rule,
						uid: generateRandomId(6)
					};
				});
				return {
					redos: [{
						id: AddDataValidationMutation.id,
						params: {
							unitId,
							subUnitId: targetSubUnitId,
							rule: newRules,
							source: "patched"
						}
					}],
					undos: [{
						id: RemoveDataValidationMutation.id,
						params: {
							unitId,
							subUnitId: targetSubUnitId,
							ruleId: newRules.map((r) => r.uid),
							source: "patched"
						}
					}]
				};
			}
			return {
				redos: [],
				undos: []
			};
		} }));
	}
};
SheetDataValidationSheetController = __decorate([
	__decorateParam(0, Inject(SheetInterceptorService)),
	__decorateParam(1, Inject(IUniverInstanceService)),
	__decorateParam(2, Inject(SheetDataValidationModel))
], SheetDataValidationSheetController);

//#endregion
//#region src/validators/any-validator.ts
var AnyValidator = class extends BaseSheetValidator {
	constructor(..._args) {
		super(..._args);
		_defineProperty(this, "id", DataValidationType.ANY);
		_defineProperty(this, "title", "sheets-data-validation.any.title");
		_defineProperty(this, "operators", []);
		_defineProperty(this, "scopes", ["sheet"]);
		_defineProperty(this, "order", 0);
		_defineProperty(this, "offsetFormulaByRange", false);
	}
	async parseFormula(rule, unitId, subUnitId) {
		return {
			formula1: rule.formula1,
			formula2: rule.formula2,
			isFormulaValid: true
		};
	}
	validatorFormula(rule, unitId, subUnitId) {
		return { success: true };
	}
	async isValidType(cellInfo, formula, rule) {
		return true;
	}
	generateRuleErrorMessage(rule) {
		return this.localeService.t("sheets-data-validation.any.error");
	}
};

//#endregion
//#region src/validators/custom-validator.ts
var CustomFormulaValidator = class extends BaseSheetValidator {
	constructor(..._args) {
		super(..._args);
		_defineProperty(this, "id", DataValidationType.CUSTOM);
		_defineProperty(this, "title", "sheets-data-validation.custom.title");
		_defineProperty(this, "operators", []);
		_defineProperty(this, "scopes", ["sheet"]);
		_defineProperty(this, "order", 60);
		_defineProperty(this, "_customFormulaService", this.injector.get(DataValidationCustomFormulaService));
		_defineProperty(this, "_lexerTreeBuilder", this.injector.get(LexerTreeBuilder));
	}
	validatorFormula(rule, unitId, subUnitId) {
		var _rule$formula;
		const success = isFormulaString(rule.formula1);
		const formulaText = (_rule$formula = rule.formula1) !== null && _rule$formula !== void 0 ? _rule$formula : "";
		const valid = this._lexerTreeBuilder.checkIfAddBracket(formulaText) === 0 && formulaText.startsWith(operatorToken.EQUALS);
		return {
			success: success && valid,
			formula1: success && valid ? "" : this.localeService.t("sheets-data-validation.validFail.formula")
		};
	}
	async parseFormula(_rule, _unitId, _subUnitId) {
		return {
			formula1: void 0,
			formula2: void 0,
			isFormulaValid: true
		};
	}
	async isValidType(cellInfo, _formula, _rule) {
		const { column, row, unitId, subUnitId } = cellInfo;
		const cellData = await this._customFormulaService.getCellFormulaValue(unitId, subUnitId, _rule.uid, row, column);
		const formulaResult = cellData === null || cellData === void 0 ? void 0 : cellData.v;
		if (!isLegalFormulaResult(String(formulaResult))) return false;
		if (Tools.isDefine(formulaResult) && formulaResult !== "") {
			if (cellData.t === CellValueType.BOOLEAN) return Boolean(formulaResult);
			if (typeof formulaResult === "boolean") return formulaResult;
			if (typeof formulaResult === "number") return Boolean(formulaResult);
			if (typeof formulaResult === "string") return isLegalFormulaResult(formulaResult);
			return Boolean(formulaResult);
		}
		return false;
	}
	generateRuleErrorMessage(rule) {
		return this.localeService.t("sheets-data-validation.custom.error");
	}
	generateRuleName(rule) {
		var _rule$formula2;
		return this.localeService.t("sheets-data-validation.custom.ruleName").replace("{FORMULA1}", (_rule$formula2 = rule.formula1) !== null && _rule$formula2 !== void 0 ? _rule$formula2 : "");
	}
};

//#endregion
//#region src/validators/list-multiple-validator.ts
/**
* Copyright 2023-present DreamNum Co., Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
var ListMultipleValidator = class extends ListValidator {
	constructor(..._args) {
		super(..._args);
		_defineProperty(this, "id", DataValidationType.LIST_MULTIPLE);
		_defineProperty(this, "title", "sheets-data-validation.listMultiple.title");
		_defineProperty(this, "offsetFormulaByRange", false);
		_defineProperty(this, "skipDefaultFontRender", () => {
			return true;
		});
	}
};

//#endregion
//#region src/validators/whole-validator.ts
var WholeValidator = class extends BaseSheetValidator {
	constructor(..._args) {
		super(..._args);
		_defineProperty(this, "_customFormulaService", this.injector.get(DataValidationCustomFormulaService));
		_defineProperty(this, "_lexerTreeBuilder", this.injector.get(LexerTreeBuilder));
		_defineProperty(this, "id", DataValidationType.WHOLE);
		_defineProperty(this, "title", "sheets-data-validation.whole.title");
		_defineProperty(this, "order", 10);
		_defineProperty(this, "operators", [
			DataValidationOperator.BETWEEN,
			DataValidationOperator.EQUAL,
			DataValidationOperator.GREATER_THAN,
			DataValidationOperator.GREATER_THAN_OR_EQUAL,
			DataValidationOperator.LESS_THAN,
			DataValidationOperator.LESS_THAN_OR_EQUAL,
			DataValidationOperator.NOT_BETWEEN,
			DataValidationOperator.NOT_EQUAL
		]);
		_defineProperty(this, "scopes", ["sheet"]);
	}
	_isFormulaOrInt(formula) {
		return !Tools.isBlank(formula) && (isFormulaString(formula) || !Number.isNaN(+formula) && Number.isInteger(+formula));
	}
	async isValidType(cellInfo, _formula, _rule) {
		const { value: cellValue } = cellInfo;
		const num = getCellValueNumber(cellValue);
		return !Number.isNaN(num) && Number.isInteger(num);
	}
	transform(cellInfo, _formula, _rule) {
		const { value: cellValue } = cellInfo;
		return {
			...cellInfo,
			value: getCellValueNumber(cellValue)
		};
	}
	_parseNumber(formula) {
		if (formula === void 0 || formula === null) return NaN;
		return +formula;
	}
	async parseFormula(rule, unitId, subUnitId, row, column) {
		const res1 = await this._customFormulaService.getCellFormulaValue(unitId, subUnitId, rule.uid, row, column);
		const res2 = await this._customFormulaService.getCellFormula2Value(unitId, subUnitId, rule.uid, row, column);
		const { formula1, formula2 } = rule;
		const formula1Result = isFormulaString(formula1) ? res1 === null || res1 === void 0 ? void 0 : res1.v : formula1;
		const formula2Result = isFormulaString(formula2) ? res2 === null || res2 === void 0 ? void 0 : res2.v : formula2;
		const isFormulaValid = isLegalFormulaResult(`${formula1Result}`) && isLegalFormulaResult(`${formula2Result}`);
		return {
			formula1: this._parseNumber(formula1Result),
			formula2: this._parseNumber(formula2Result),
			isFormulaValid
		};
	}
	validatorFormula(rule, _unitId, _subUnitId) {
		const operator = rule.operator;
		if (!operator) return { success: true };
		const formula1Success = Tools.isDefine(rule.formula1) && this._isFormulaOrInt(rule.formula1);
		const formula2Success = Tools.isDefine(rule.formula2) && this._isFormulaOrInt(rule.formula2);
		const isTwoFormula = TWO_FORMULA_OPERATOR_COUNT.includes(operator);
		const errorMsg = this.localeService.t("sheets-data-validation.validFail.number");
		if (isTwoFormula) return {
			success: formula1Success && formula2Success,
			formula1: formula1Success ? void 0 : errorMsg,
			formula2: formula2Success ? void 0 : errorMsg
		};
		return {
			success: formula1Success,
			formula1: errorMsg
		};
	}
	generateRuleErrorMessage(rule, position) {
		if (!rule.operator) return this.localeService.t(OperatorErrorTitleMap.NONE).replace("{TYPE}", this.titleStr);
		const { transformedFormula1, transformedFormula2 } = getTransformedFormula(this._lexerTreeBuilder, rule, position);
		return `${this.localeService.t(OperatorErrorTitleMap[rule.operator]).replace(FORMULA1, transformedFormula1 !== null && transformedFormula1 !== void 0 ? transformedFormula1 : "").replace(FORMULA2, transformedFormula2 !== null && transformedFormula2 !== void 0 ? transformedFormula2 : "")}`;
	}
};

//#endregion
//#region src/controllers/dv.controller.ts
let DataValidationController = class DataValidationController extends RxDisposable {
	constructor(_univerInstanceService, _dataValidatorRegistryService, _injector, _selectionManagerService, _sheetInterceptorService, _sheetDataValidationModel) {
		super();
		this._univerInstanceService = _univerInstanceService;
		this._dataValidatorRegistryService = _dataValidatorRegistryService;
		this._injector = _injector;
		this._selectionManagerService = _selectionManagerService;
		this._sheetInterceptorService = _sheetInterceptorService;
		this._sheetDataValidationModel = _sheetDataValidationModel;
		this._init();
	}
	_init() {
		this._registerValidators();
		this._initCommandInterceptor();
	}
	_registerValidators() {
		[
			AnyValidator,
			DecimalValidator,
			WholeValidator,
			TextLengthValidator,
			DateValidator,
			CheckboxValidator,
			ListValidator,
			ListMultipleValidator,
			CustomFormulaValidator
		].forEach((Validator) => {
			const validator = this._injector.createInstance(Validator);
			this.disposeWithMe(this._dataValidatorRegistryService.register(validator));
			this.disposeWithMe(toDisposable(() => this._injector.delete(Validator)));
		});
	}
	_initCommandInterceptor() {
		this._sheetInterceptorService.interceptCommand({ getMutations: (commandInfo) => {
			if (commandInfo.id === ClearSelectionAllCommand.id) {
				var _this$_selectionManag;
				const workbook = this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_SHEET);
				const unitId = workbook.getUnitId();
				const worksheet = workbook.getActiveSheet();
				if (!worksheet) throw new Error("No active sheet found");
				const subUnitId = worksheet.getSheetId();
				const selections = (_this$_selectionManag = this._selectionManagerService.getCurrentSelections()) === null || _this$_selectionManag === void 0 ? void 0 : _this$_selectionManag.map((s) => s.range);
				const ruleMatrix = this._sheetDataValidationModel.getRuleObjectMatrix(unitId, subUnitId).clone();
				if (selections) ruleMatrix.removeRange(selections);
				const { redoMutations, undoMutations } = getDataValidationDiffMutations(unitId, subUnitId, ruleMatrix.diff(this._sheetDataValidationModel.getRules(unitId, subUnitId)), this._injector, "patched");
				return {
					undos: undoMutations,
					redos: redoMutations
				};
			}
			return {
				undos: [],
				redos: []
			};
		} });
	}
};
DataValidationController = __decorate([
	__decorateParam(0, IUniverInstanceService),
	__decorateParam(1, Inject(DataValidatorRegistryService)),
	__decorateParam(2, Inject(Injector)),
	__decorateParam(3, Inject(SheetsSelectionsService)),
	__decorateParam(4, Inject(SheetInterceptorService)),
	__decorateParam(5, Inject(SheetDataValidationModel))
], DataValidationController);

//#endregion
//#region src/services/dv-validator.service.ts
let SheetsDataValidationValidatorService = class SheetsDataValidationValidatorService extends Disposable {
	constructor(_univerInstanceService, _sheetDataValidationModel, _dataValidationCacheService, _lifecycleService) {
		super();
		this._univerInstanceService = _univerInstanceService;
		this._sheetDataValidationModel = _sheetDataValidationModel;
		this._dataValidationCacheService = _dataValidationCacheService;
		this._lifecycleService = _lifecycleService;
		this._initRecalculate();
	}
	_initRecalculate() {
		const handleDirtyRanges = (ranges) => {
			if (ranges.length === 0) return;
			const workbook = this._univerInstanceService.getCurrentUnitOfType(UniverInstanceType.UNIVER_SHEET);
			const worksheet = workbook === null || workbook === void 0 ? void 0 : workbook.getActiveSheet();
			const map = {};
			ranges.flat().forEach((range) => {
				if (!map[range.unitId]) map[range.unitId] = {};
				if (!map[range.unitId][range.subUnitId]) map[range.unitId][range.subUnitId] = [];
				const workbook = this._univerInstanceService.getUnit(range.unitId, UniverInstanceType.UNIVER_SHEET);
				const worksheet = workbook === null || workbook === void 0 ? void 0 : workbook.getSheetBySheetId(range.subUnitId);
				if (!worksheet) return;
				map[range.unitId][range.subUnitId].push(...range.ranges.map((range) => Range.transformRange(range, worksheet)));
			});
			Object.entries(map).forEach(([unitId, subUnitMap]) => {
				Object.entries(subUnitMap).forEach(([subUnitId, ranges]) => {
					if ((workbook === null || workbook === void 0 ? void 0 : workbook.getUnitId()) === unitId && (worksheet === null || worksheet === void 0 ? void 0 : worksheet.getSheetId()) === subUnitId) this.validatorRanges(unitId, subUnitId, ranges);
					else requestIdleCallback(() => {
						this.validatorRanges(unitId, subUnitId, ranges);
					});
				});
			});
		};
		this.disposeWithMe(this._dataValidationCacheService.dirtyRanges$.pipe(bufferWhen(() => this._lifecycleService.lifecycle$.pipe(skip(1), filter((stage) => stage === LifecycleStages.Rendered)))).subscribe(handleDirtyRanges));
		this.disposeWithMe(this._dataValidationCacheService.dirtyRanges$.pipe(filter(() => this._lifecycleService.stage >= LifecycleStages.Rendered), bufferDebounceTime(20)).subscribe(handleDirtyRanges));
	}
	async _validatorByCell(workbook, worksheet, row, col) {
		const unitId = workbook.getUnitId();
		const subUnitId = worksheet.getSheetId();
		if (!Tools.isDefine(row) || !Tools.isDefine(col)) throw new Error(`row or col is not defined, row: ${row}, col: ${col}`);
		let _row = row;
		let _col = col;
		const mergedCell = worksheet.getMergedCell(row, col);
		if (mergedCell) {
			_row = mergedCell.startRow;
			_col = mergedCell.startColumn;
		}
		const rule = this._sheetDataValidationModel.getRuleByLocation(unitId, subUnitId, _row, _col);
		if (!rule) return DataValidationStatus.VALID;
		return new Promise((resolve) => {
			this._sheetDataValidationModel.validator(rule, {
				unitId,
				subUnitId,
				row: _row,
				col: _col,
				worksheet,
				workbook
			}, (status) => {
				resolve(status);
			});
		});
	}
	async validatorCell(unitId, subUnitId, row, col) {
		const workbook = this._univerInstanceService.getUnit(unitId, UniverInstanceType.UNIVER_SHEET);
		if (!workbook) throw new Error(`cannot find current workbook, unitId: ${unitId}`);
		const worksheet = workbook.getSheetBySheetId(subUnitId);
		if (!worksheet) throw new Error(`cannot find current worksheet, sheetId: ${subUnitId}`);
		return this._validatorByCell(workbook, worksheet, row, col);
	}
	async validatorRanges(unitId, subUnitId, ranges) {
		if (!ranges.length) return Promise.resolve([]);
		const workbook = this._univerInstanceService.getUnit(unitId, UniverInstanceType.UNIVER_SHEET);
		if (!workbook) throw new Error(`cannot find current workbook, unitId: ${unitId}`);
		const worksheet = workbook.getSheetBySheetId(subUnitId);
		if (!worksheet) throw new Error(`cannot find current worksheet, sheetId: ${subUnitId}`);
		const allRules = this._sheetDataValidationModel.getRules(unitId, subUnitId);
		const ruleRanges = [];
		for (const rule of allRules) ruleRanges.push(...rule.ranges);
		const intersectRanges = [];
		for (const range of ranges) for (const ruleRange of ruleRanges) {
			const intersect = getIntersectRange(range, ruleRange);
			if (intersect) intersectRanges.push(intersect);
		}
		const mergeCells = [];
		const result = await Promise.all(intersectRanges.map((range, resultRowIndex) => {
			const promises = [];
			for (let row = range.startRow; row <= range.endRow; row++) for (let col = range.startColumn; col <= range.endColumn; col++) {
				promises.push(this._validatorByCell(workbook, worksheet, row, col));
				const mergedCell = worksheet.getMergedCell(row, col);
				if (mergedCell) mergeCells.push({
					resultRowIndex,
					resultColIndex: promises.length - 1,
					row: mergedCell.startRow,
					col: mergedCell.startColumn
				});
			}
			return Promise.all(promises);
		}));
		/**
		* If this range has merged cells, the validation status of merged cells should be the same as the main cell, so we need to update the status of merged cells here after all validations are done.
		* Because during the validation process, merged cells are all marked as VALIDATING status in cache.
		*/
		if (mergeCells.length) mergeCells.forEach(({ resultRowIndex, resultColIndex, row, col }) => {
			if (result[resultRowIndex][resultColIndex] === DataValidationStatus.VALIDATING) {
				var _this$_dataValidation;
				result[resultRowIndex][resultColIndex] = (_this$_dataValidation = this._dataValidationCacheService.getValue(unitId, subUnitId, row, col)) !== null && _this$_dataValidation !== void 0 ? _this$_dataValidation : DataValidationStatus.VALID;
			}
		});
		return result;
	}
	async validatorWorksheet(unitId, subUnitId) {
		const workbook = this._univerInstanceService.getUnit(unitId, UniverInstanceType.UNIVER_SHEET);
		if (!workbook) throw new Error(`cannot find current workbook, unitId: ${unitId}`);
		const worksheet = workbook.getSheetBySheetId(subUnitId);
		if (!worksheet) throw new Error(`cannot find current worksheet, sheetId: ${subUnitId}`);
		const rules = this._sheetDataValidationModel.getRules(unitId, subUnitId);
		await Promise.all(rules.map((rule) => {
			return Promise.all(rule.ranges.map((range) => {
				const promises = [];
				Range.foreach(range, (row, col) => {
					promises.push(this._validatorByCell(workbook, worksheet, row, col));
				});
				return Promise.all(promises);
			}));
		}));
		return this._dataValidationCacheService.ensureCache(unitId, subUnitId);
	}
	async validatorWorkbook(unitId) {
		const sheetIds = this._sheetDataValidationModel.getSubUnitIds(unitId);
		const results = await Promise.all(sheetIds.map((id) => this.validatorWorksheet(unitId, id)));
		const map = {};
		results.forEach((result, i) => {
			map[sheetIds[i]] = result;
		});
		return map;
	}
	getDataValidations(unitId, subUnitId, ranges) {
		const ruleMatrix = this._sheetDataValidationModel.getRuleObjectMatrix(unitId, subUnitId);
		const ruleIdSet = /* @__PURE__ */ new Set();
		ranges.forEach((range) => {
			Range.foreach(range, (row, col) => {
				const ruleId = ruleMatrix.getValue(row, col);
				if (ruleId) ruleIdSet.add(ruleId);
			});
		});
		const rules = [];
		for (const id of ruleIdSet) {
			const rule = this._sheetDataValidationModel.getRuleById(unitId, subUnitId, id);
			if (rule) rules.push(rule);
		}
		return rules;
	}
	getDataValidation(unitId, subUnitId, ranges) {
		return this.getDataValidations(unitId, subUnitId, ranges)[0];
	}
};
SheetsDataValidationValidatorService = __decorate([
	__decorateParam(0, IUniverInstanceService),
	__decorateParam(1, Inject(SheetDataValidationModel)),
	__decorateParam(2, Inject(DataValidationCacheService)),
	__decorateParam(3, Inject(LifecycleService))
], SheetsDataValidationValidatorService);

//#endregion
//#region src/plugin.ts
let UniverSheetsDataValidationPlugin = class UniverSheetsDataValidationPlugin extends Plugin {
	constructor(_config = defaultPluginConfig, _injector, _commandService, _configService) {
		super();
		this._config = _config;
		this._injector = _injector;
		this._commandService = _commandService;
		this._configService = _configService;
		const { ...rest } = merge({}, defaultPluginConfig, this._config);
		this._configService.setConfig(SHEETS_DATA_VALIDATION_PLUGIN_CONFIG_KEY, rest);
	}
	onStarting() {
		[
			[DataValidationCacheService],
			[DataValidationListCacheService],
			[DataValidationFormulaService],
			[DataValidationCustomFormulaService],
			[SheetsDataValidationValidatorService],
			[SheetDataValidationModel],
			[DataValidationController],
			[DataValidationFormulaController],
			[SheetDataValidationSheetController],
			[DataValidationRefRangeController],
			[DataValidationFormulaRefRangeController]
		].forEach((dep) => {
			this._injector.add(dep);
		});
		[
			AddSheetDataValidationCommand,
			UpdateSheetDataValidationRangeCommand,
			UpdateSheetDataValidationSettingCommand,
			UpdateSheetDataValidationOptionsCommand,
			RemoveSheetDataValidationCommand,
			RemoveSheetAllDataValidationCommand,
			ClearRangeDataValidationCommand
		].forEach((command) => {
			this._commandService.registerCommand(command);
		});
		this._injector.get(DataValidationCacheService);
		this._injector.get(SheetsDataValidationValidatorService);
		this._injector.get(DataValidationController);
		this._injector.get(DataValidationFormulaRefRangeController);
		this._injector.get(DataValidationRefRangeController);
	}
	onReady() {
		this._injector.get(SheetDataValidationSheetController);
	}
	onRendered() {
		this._injector.get(DataValidationFormulaController);
	}
};
_defineProperty(UniverSheetsDataValidationPlugin, "pluginName", DATA_VALIDATION_PLUGIN_NAME);
_defineProperty(UniverSheetsDataValidationPlugin, "packageName", name);
_defineProperty(UniverSheetsDataValidationPlugin, "version", version);
_defineProperty(UniverSheetsDataValidationPlugin, "type", UniverInstanceType.UNIVER_SHEET);
UniverSheetsDataValidationPlugin = __decorate([
	DependentOn(UniverSheetsFormulaPlugin, UniverDataValidationPlugin),
	__decorateParam(1, Inject(Injector)),
	__decorateParam(2, ICommandService),
	__decorateParam(3, IConfigService)
], UniverSheetsDataValidationPlugin);

//#endregion
//#region src/utils/create.ts
function createDefaultNewRule(accessor) {
	const currentRanges = accessor.get(SheetsSelectionsService).getCurrentSelections().map((s) => s.range);
	return {
		uid: generateRandomId(6),
		type: DataValidationType.DECIMAL,
		operator: DataValidationOperator.EQUAL,
		formula1: "100",
		ranges: currentRanges !== null && currentRanges !== void 0 ? currentRanges : [{
			startColumn: 0,
			endColumn: 0,
			startRow: 0,
			endRow: 0
		}]
	};
}

//#endregion
//#region src/index.ts
const CUSTOM_FORMULA_INPUT_NAME = "sheets-data-validation.custom-formula-input";
const BASE_FORMULA_INPUT_NAME = "sheets-data-validation.formula-input";
const LIST_FORMULA_INPUT_NAME = "sheets-data-validation.list-formula-input";
const CHECKBOX_FORMULA_INPUT_NAME = "sheets-data-validation.checkbox-formula-input";

//#endregion
export { AddSheetDataValidationCommand, BASE_FORMULA_INPUT_NAME, CHECKBOX_FORMULA_1, CHECKBOX_FORMULA_2, CHECKBOX_FORMULA_INPUT_NAME, CUSTOM_FORMULA_INPUT_NAME, CheckboxValidator, ClearRangeDataValidationCommand, DATA_VALIDATION_PLUGIN_NAME, DataValidationCacheService, DataValidationCustomFormulaService, DataValidationFormulaController, DataValidationFormulaService, DataValidationListCacheService, DateValidator, LIST_FORMULA_INPUT_NAME, ListMultipleValidator, ListValidator, RemoveSheetAllDataValidationCommand, RemoveSheetDataValidationCommand, SheetDataValidationModel, SheetsDataValidationValidatorService, UniverSheetsDataValidationPlugin, UpdateSheetDataValidationOptionsCommand, UpdateSheetDataValidationRangeCommand, UpdateSheetDataValidationSettingCommand, createDefaultNewRule, getCellValueNumber, getCellValueOrigin, getDataValidationCellValue, getDataValidationDiffMutations, getFormulaCellData, getFormulaResult, getTransformedFormula, isLegalFormulaResult, transformCheckboxValue };
